{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alvin Oligo Generator\n",
    "\n",
    "Formatting mutations to generate WT and alt sequence for ML-based pegRNA generation.\n",
    "\n",
    "- Going to do this with: \n",
    "    - Diego's IDR library\n",
    "        - Need to filter out a few INS/DEL that aren't pure INS/DEL (they're complex indels)\n",
    "    - p53 follow-up library\n",
    "        - Not going to worry about the random indels\n",
    "        - I can include the efficient pegRNAs that I have for a subset of these (e.g. I254_I255insH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import regex as re\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import Bio.Seq\n",
    "import warnings\n",
    "import pegg\n",
    "import gffutils\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#loading in reference genome\n",
    "#and genome annotations\n",
    "\n",
    "filepath = '/Users/samgould/Desktop/FSR Lab/reference files/GRCh37/ncbi-genomes-2022-03-17/GCF_000001405.25_GRCh37.p13_genomic.fna.gz'\n",
    "records, index_list = pegg.genome_loader(filepath)\n",
    "\n",
    "file = '/Users/samgould/Desktop/FSR Lab/reference files/gencode_v19.db'\n",
    "db = gffutils.FeatureDB(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "def df_formatter(df, context_size = 60):\n",
    "\n",
    "    \"\"\" \n",
    "    Takes in variants (in cBioPortal format)\n",
    "    and outputs dataframe with WT and ALT oligos with designated context_size\n",
    "    context_size = the amount of nt on either side of the variant e.g. AAA(A/G)AAA = context_size of 3\n",
    "    \"\"\"\n",
    "\n",
    "    wt_w_context = []\n",
    "    alt_w_context = []\n",
    "\n",
    "    seq_start = []\n",
    "    seq_end = []\n",
    "\n",
    "    for i, val in df.iterrows():\n",
    "        vt = val['Variant_Type']\n",
    "        s = val['Start_Position']\n",
    "        e = val['End_Position']\n",
    "        ref = val['Reference_Allele']\n",
    "        alt = val['Tumor_Seq_Allele2']\n",
    "        chrom = val['Chromosome']\n",
    "\n",
    "        if chrom == 'X':\n",
    "            chrom = 22\n",
    "        elif chrom=='Y':\n",
    "            chrom = 23\n",
    "        else:\n",
    "            chrom = int(chrom)-1\n",
    "\n",
    "        chr_seq = records[index_list[chrom]].seq.upper()\n",
    "\n",
    "        if vt in ['SNP', 'ONP', 'DNP']:\n",
    "            ref = ref\n",
    "            alt = alt\n",
    "            #assert ref == chr_seq[s-1:e], print(ref, chr_seq[s-1:e])\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        elif vt =='INS':\n",
    "            ref = ''\n",
    "            alt = alt\n",
    "            #left_context = chr_seq[s-1-context_size:s+1] #need to do this since INS reference alleles are blank\n",
    "            left_context = chr_seq[s-1-context_size:s]\n",
    "            right_context = chr_seq[e-1:e+context_size]\n",
    "\n",
    "        elif vt=='DEL':\n",
    "            ref = ref\n",
    "            alt = ''\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        wt_seq = left_context + ref + right_context\n",
    "        alt_seq = left_context + alt + right_context\n",
    "\n",
    "        wt_w_context.append(str(wt_seq))\n",
    "        alt_w_context.append(str(alt_seq))\n",
    "\n",
    "        start = s-context_size\n",
    "        end = e+context_size\n",
    "\n",
    "        seq_start.append(start)\n",
    "        seq_end.append(end)\n",
    "\n",
    "        assert str(chr_seq[start-1:end])==str(wt_seq), print(chr_seq[start-1:end] + '\\n' + str(wt_seq))\n",
    "                                                            \n",
    "\n",
    "    cols_to_save = ['COUNT', 'Hugo_Symbol', 'Chromosome', 'Start_Position', 'End_Position', 'Consequence', 'Variant_Classification', 'Variant_Type', 'Reference_Allele', 'Tumor_Seq_Allele2', 'Tumor_Sample_Barcode', 'RefSeq', 'Protein_position', 'Exon_Number', 'HGVSc', 'HGVSp', 'HGVSp_Short', 'transcript_id_TRUE']\n",
    "    df_new = df[cols_to_save]\n",
    "\n",
    "    df_new['seq_start'] = seq_start\n",
    "    df_new['seq_end'] = seq_end\n",
    "    df_new['wt_w_context'] = wt_w_context\n",
    "    df_new['alt_w_context'] = alt_w_context\n",
    "    df_new = df_new.reset_index()\n",
    "    \n",
    "    return df_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1809.0</td>\n",
       "      <td>EIF1AX</td>\n",
       "      <td>X</td>\n",
       "      <td>20156758</td>\n",
       "      <td>20156758</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>Intron</td>\n",
       "      <td>DEL</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-COLU-951-1</td>\n",
       "      <td>NM_001412.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000379607.5:c.17-11del</td>\n",
       "      <td>NaN</td>\n",
       "      <td>p.*6*</td>\n",
       "      <td>ENST00000379607.5</td>\n",
       "      <td>20156698</td>\n",
       "      <td>20156818</td>\n",
       "      <td>TCATTCTCATTCTTACCCCTGCGTCTGTTTTTACCTCCTTTACCTG...</td>\n",
       "      <td>TCATTCTCATTCTTACCCCTGCGTCTGTTTTTACCTCCTTTACCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1512.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-JHU-00113-00335</td>\n",
       "      <td>NM_001126112.2</td>\n",
       "      <td>282.0</td>\n",
       "      <td>8/11</td>\n",
       "      <td>ENST00000269305.4:c.844C&gt;T</td>\n",
       "      <td>p.Arg282Trp</td>\n",
       "      <td>p.R282W</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577034</td>\n",
       "      <td>7577154</td>\n",
       "      <td>CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...</td>\n",
       "      <td>CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>983.0</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>17</td>\n",
       "      <td>56435161</td>\n",
       "      <td>56435161</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-DFCI-449465-4787362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>659.0</td>\n",
       "      <td>9/10</td>\n",
       "      <td>ENST00000407977.2:c.1976del</td>\n",
       "      <td>p.Gly659ValfsTer41</td>\n",
       "      <td>p.G659Vfs*41</td>\n",
       "      <td>ENST00000407977.2</td>\n",
       "      <td>56435101</td>\n",
       "      <td>56435221</td>\n",
       "      <td>TGGCAAGCTGGGTGCACAGTTGCATCCTGGGGCCGAGAGCCAGGGG...</td>\n",
       "      <td>TGGCAAGCTGGGTGCACAGTTGCATCCTGGGGCCGAGAGCCAGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>777.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175639</td>\n",
       "      <td>112175639</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00223-00468</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4348C&gt;T</td>\n",
       "      <td>p.Arg1450Ter</td>\n",
       "      <td>p.R1450*</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175579</td>\n",
       "      <td>112175699</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>762.0</td>\n",
       "      <td>ASXL1</td>\n",
       "      <td>20</td>\n",
       "      <td>31022441</td>\n",
       "      <td>31022442</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-003409-1958</td>\n",
       "      <td>NM_015338.5</td>\n",
       "      <td>642.0</td>\n",
       "      <td>13/13</td>\n",
       "      <td>ENST00000375687.4:c.1934dup</td>\n",
       "      <td>p.Gly646TrpfsTer12</td>\n",
       "      <td>p.G646Wfs*12</td>\n",
       "      <td>ENST00000375687.4</td>\n",
       "      <td>31022381</td>\n",
       "      <td>31022502</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9090</th>\n",
       "      <td>9090</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ERCC5</td>\n",
       "      <td>13</td>\n",
       "      <td>103527930</td>\n",
       "      <td>103527931</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>DNP</td>\n",
       "      <td>GG</td>\n",
       "      <td>CA</td>\n",
       "      <td>GENIE-PROV-9b34c1d29d-5652092fc3</td>\n",
       "      <td>NM_000123.3</td>\n",
       "      <td>1080.0</td>\n",
       "      <td>15/15</td>\n",
       "      <td>ENST00000355739.4:c.3238_3239delinsCA</td>\n",
       "      <td>p.Gly1080Gln</td>\n",
       "      <td>p.G1080Q</td>\n",
       "      <td>ENST00000355739.4</td>\n",
       "      <td>103527870</td>\n",
       "      <td>103527991</td>\n",
       "      <td>ATAACAAATACCTTAGAAGAGTCATCAAGCCTGAAAAGAAAGAGGC...</td>\n",
       "      <td>ATAACAAATACCTTAGAAGAGTCATCAAGCCTGAAAAGAAAGAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9091</th>\n",
       "      <td>9091</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>10</td>\n",
       "      <td>63850742</td>\n",
       "      <td>63850742</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-PROV-1a9954b812-1203eb4b3a</td>\n",
       "      <td>NM_032199.2</td>\n",
       "      <td>507.0</td>\n",
       "      <td>10/10</td>\n",
       "      <td>ENST00000279873.7:c.1520C&gt;T</td>\n",
       "      <td>p.Pro507Leu</td>\n",
       "      <td>p.P507L</td>\n",
       "      <td>ENST00000279873.7</td>\n",
       "      <td>63850682</td>\n",
       "      <td>63850802</td>\n",
       "      <td>CTCTCCCAGCAGCAGACATGAAGAAAAAAATAGAAGGGTATCAGGA...</td>\n",
       "      <td>CTCTCCCAGCAGCAGACATGAAGAAAAAAATAGAAGGGTATCAGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9092</th>\n",
       "      <td>9092</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ABL2</td>\n",
       "      <td>1</td>\n",
       "      <td>179078007</td>\n",
       "      <td>179078007</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>GENIE-PROV-6a280d78a3-155ebf48a8</td>\n",
       "      <td>NM_001168237.1</td>\n",
       "      <td>799.0</td>\n",
       "      <td>12/12</td>\n",
       "      <td>ENST00000502732.1:c.2395A&gt;G</td>\n",
       "      <td>p.Met799Val</td>\n",
       "      <td>p.M799V</td>\n",
       "      <td>ENST00000502732.1</td>\n",
       "      <td>179077947</td>\n",
       "      <td>179078067</td>\n",
       "      <td>AGGTGGACACTGTCCTTTCCAGCTGGAGTTTGGACCTCTGGCAGTT...</td>\n",
       "      <td>AGGTGGACACTGTCCTTTCCAGCTGGAGTTTGGACCTCTGGCAGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9093</th>\n",
       "      <td>9093</td>\n",
       "      <td>5.0</td>\n",
       "      <td>SOX10</td>\n",
       "      <td>22</td>\n",
       "      <td>38369885</td>\n",
       "      <td>38369885</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-PROV-238a11a4e2-b283f19879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>340.0</td>\n",
       "      <td>5/5</td>\n",
       "      <td>ENST00000360880.2:c.1018G&gt;A</td>\n",
       "      <td>p.Val340Met</td>\n",
       "      <td>p.V340M</td>\n",
       "      <td>ENST00000360880.2</td>\n",
       "      <td>38369825</td>\n",
       "      <td>38369945</td>\n",
       "      <td>TCTCTGTCTTCACCTGGGCTTTGGCATCCACACCAGGTGGTGAGAC...</td>\n",
       "      <td>TCTCTGTCTTCACCTGGGCTTTGGCATCCACACCAGGTGGTGAGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9094</th>\n",
       "      <td>9094</td>\n",
       "      <td>5.0</td>\n",
       "      <td>AMER1</td>\n",
       "      <td>X</td>\n",
       "      <td>63410992</td>\n",
       "      <td>63410992</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>GENIE-PROV-26142c27e9-be569ca40c</td>\n",
       "      <td>NM_152424.3</td>\n",
       "      <td>725.0</td>\n",
       "      <td>2/2</td>\n",
       "      <td>ENST00000330258.3:c.2175T&gt;G</td>\n",
       "      <td>p.Asp725Glu</td>\n",
       "      <td>p.D725E</td>\n",
       "      <td>ENST00000330258.3</td>\n",
       "      <td>63410932</td>\n",
       "      <td>63411052</td>\n",
       "      <td>GTAGGCCCTCCTGGGAGATCCTCCAAAATTTGCTTCTTGCATGTCT...</td>\n",
       "      <td>GTAGGCCCTCCTGGGAGATCCTCCAAAATTTGCTTCTTGCATGTCT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9095 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index   COUNT Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0         0  1809.0      EIF1AX          X        20156758      20156758   \n",
       "1         1  1512.0        TP53         17         7577094       7577094   \n",
       "2         2   983.0       RNF43         17        56435161      56435161   \n",
       "3         3   777.0         APC          5       112175639     112175639   \n",
       "4         4   762.0       ASXL1         20        31022441      31022442   \n",
       "...     ...     ...         ...        ...             ...           ...   \n",
       "9090   9090     5.0       ERCC5         13       103527930     103527931   \n",
       "9091   9091     5.0      ARID5B         10        63850742      63850742   \n",
       "9092   9092     5.0        ABL2          1       179078007     179078007   \n",
       "9093   9093     5.0       SOX10         22        38369885      38369885   \n",
       "9094   9094     5.0       AMER1          X        63410992      63410992   \n",
       "\n",
       "             Consequence Variant_Classification Variant_Type Reference_Allele  \\\n",
       "0         intron_variant                 Intron          DEL                A   \n",
       "1       missense_variant      Missense_Mutation          SNP                G   \n",
       "2     frameshift_variant        Frame_Shift_Del          DEL                C   \n",
       "3            stop_gained      Nonsense_Mutation          SNP                C   \n",
       "4     frameshift_variant        Frame_Shift_Ins          INS                -   \n",
       "...                  ...                    ...          ...              ...   \n",
       "9090    missense_variant      Missense_Mutation          DNP               GG   \n",
       "9091    missense_variant      Missense_Mutation          SNP                C   \n",
       "9092    missense_variant      Missense_Mutation          SNP                T   \n",
       "9093    missense_variant      Missense_Mutation          SNP                C   \n",
       "9094    missense_variant      Missense_Mutation          SNP                A   \n",
       "\n",
       "     Tumor_Seq_Allele2              Tumor_Sample_Barcode          RefSeq  \\\n",
       "0                    -                  GENIE-COLU-951-1     NM_001412.3   \n",
       "1                    A             GENIE-JHU-00113-00335  NM_001126112.2   \n",
       "2                    -         GENIE-DFCI-449465-4787362             NaN   \n",
       "3                    T             GENIE-JHU-00223-00468     NM_000038.5   \n",
       "4                    G            GENIE-DFCI-003409-1958     NM_015338.5   \n",
       "...                ...                               ...             ...   \n",
       "9090                CA  GENIE-PROV-9b34c1d29d-5652092fc3     NM_000123.3   \n",
       "9091                 T  GENIE-PROV-1a9954b812-1203eb4b3a     NM_032199.2   \n",
       "9092                 C  GENIE-PROV-6a280d78a3-155ebf48a8  NM_001168237.1   \n",
       "9093                 T  GENIE-PROV-238a11a4e2-b283f19879             NaN   \n",
       "9094                 C  GENIE-PROV-26142c27e9-be569ca40c     NM_152424.3   \n",
       "\n",
       "      Protein_position Exon_Number                                  HGVSc  \\\n",
       "0                  NaN         NaN           ENST00000379607.5:c.17-11del   \n",
       "1                282.0        8/11             ENST00000269305.4:c.844C>T   \n",
       "2                659.0        9/10            ENST00000407977.2:c.1976del   \n",
       "3               1450.0       16/16            ENST00000257430.4:c.4348C>T   \n",
       "4                642.0       13/13            ENST00000375687.4:c.1934dup   \n",
       "...                ...         ...                                    ...   \n",
       "9090            1080.0       15/15  ENST00000355739.4:c.3238_3239delinsCA   \n",
       "9091             507.0       10/10            ENST00000279873.7:c.1520C>T   \n",
       "9092             799.0       12/12            ENST00000502732.1:c.2395A>G   \n",
       "9093             340.0         5/5            ENST00000360880.2:c.1018G>A   \n",
       "9094             725.0         2/2            ENST00000330258.3:c.2175T>G   \n",
       "\n",
       "                   HGVSp   HGVSp_Short transcript_id_TRUE  seq_start  \\\n",
       "0                    NaN         p.*6*  ENST00000379607.5   20156698   \n",
       "1            p.Arg282Trp       p.R282W  ENST00000269305.4    7577034   \n",
       "2     p.Gly659ValfsTer41  p.G659Vfs*41  ENST00000407977.2   56435101   \n",
       "3           p.Arg1450Ter      p.R1450*  ENST00000257430.4  112175579   \n",
       "4     p.Gly646TrpfsTer12  p.G646Wfs*12  ENST00000375687.4   31022381   \n",
       "...                  ...           ...                ...        ...   \n",
       "9090        p.Gly1080Gln      p.G1080Q  ENST00000355739.4  103527870   \n",
       "9091         p.Pro507Leu       p.P507L  ENST00000279873.7   63850682   \n",
       "9092         p.Met799Val       p.M799V  ENST00000502732.1  179077947   \n",
       "9093         p.Val340Met       p.V340M  ENST00000360880.2   38369825   \n",
       "9094         p.Asp725Glu       p.D725E  ENST00000330258.3   63410932   \n",
       "\n",
       "        seq_end                                       wt_w_context  \\\n",
       "0      20156818  TCATTCTCATTCTTACCCCTGCGTCTGTTTTTACCTCCTTTACCTG...   \n",
       "1       7577154  CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...   \n",
       "2      56435221  TGGCAAGCTGGGTGCACAGTTGCATCCTGGGGCCGAGAGCCAGGGG...   \n",
       "3     112175699  ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...   \n",
       "4      31022502  GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...   \n",
       "...         ...                                                ...   \n",
       "9090  103527991  ATAACAAATACCTTAGAAGAGTCATCAAGCCTGAAAAGAAAGAGGC...   \n",
       "9091   63850802  CTCTCCCAGCAGCAGACATGAAGAAAAAAATAGAAGGGTATCAGGA...   \n",
       "9092  179078067  AGGTGGACACTGTCCTTTCCAGCTGGAGTTTGGACCTCTGGCAGTT...   \n",
       "9093   38369945  TCTCTGTCTTCACCTGGGCTTTGGCATCCACACCAGGTGGTGAGAC...   \n",
       "9094   63411052  GTAGGCCCTCCTGGGAGATCCTCCAAAATTTGCTTCTTGCATGTCT...   \n",
       "\n",
       "                                          alt_w_context  \n",
       "0     TCATTCTCATTCTTACCCCTGCGTCTGTTTTTACCTCCTTTACCTG...  \n",
       "1     CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...  \n",
       "2     TGGCAAGCTGGGTGCACAGTTGCATCCTGGGGCCGAGAGCCAGGGG...  \n",
       "3     ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...  \n",
       "4     GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...  \n",
       "...                                                 ...  \n",
       "9090  ATAACAAATACCTTAGAAGAGTCATCAAGCCTGAAAAGAAAGAGGC...  \n",
       "9091  CTCTCCCAGCAGCAGACATGAAGAAAAAAATAGAAGGGTATCAGGA...  \n",
       "9092  AGGTGGACACTGTCCTTTCCAGCTGGAGTTTGGACCTCTGGCAGTT...  \n",
       "9093  TCTCTGTCTTCACCTGGGCTTTGGCATCCACACCAGGTGGTGAGAC...  \n",
       "9094  GTAGGCCCTCCTGGGAGATCCTCCAAAATTTGCTTCTTGCATGTCT...  \n",
       "\n",
       "[9095 rows x 23 columns]"
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idr_oligos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [],
   "source": [
    "def frame_determiner(exon_subs):\n",
    "    \"\"\" \n",
    "    Determines the frame of the WT/ALT sequence (0,1, or 2)\n",
    "    And selects the CDS subset + puts it in the correct orientation\n",
    "    \"\"\"\n",
    "    \n",
    "    frame_list = []\n",
    "    coding_start_list = []\n",
    "    coding_end_list = []\n",
    "    coding_alt_start_list = []\n",
    "    coding_alt_end_list = []\n",
    "    CDS_WT_correct_orientation = []\n",
    "    CDS_alt_correct_orientation = []\n",
    "    strand_list = []\n",
    "\n",
    "    for index1, val in exon_subs.iterrows():\n",
    "        hg = val['HGVSp_Short']\n",
    "\n",
    "        tx = val['transcript_id_TRUE']\n",
    "        cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "        start_end_cds = [[i.start, i.end] for i in cds]\n",
    "        strand = db[tx].strand\n",
    "        strand_list.append(strand)\n",
    "        #print(strand)\n",
    "        chrom = val['Chromosome']\n",
    "\n",
    "        if chrom == 'X':\n",
    "            chrom = 22\n",
    "        elif chrom=='Y':\n",
    "            chrom = 23\n",
    "        else:\n",
    "            chrom = int(chrom)-1\n",
    "\n",
    "        chr_seq = records[index_list[chrom]].seq.upper()\n",
    "\n",
    "\n",
    "        wt_dna = ''\n",
    "        codon_locs = []\n",
    "        for i in start_end_cds:\n",
    "            wt_dna += chr_seq[i[0]-1:i[1]]\n",
    "\n",
    "            for j in range(i[0], i[1]+1):\n",
    "                codon_locs.append(j)\n",
    "\n",
    "        #and add in the last/stop codon\n",
    "        #if strand == '+':\n",
    "\n",
    "        # wt_dna += chr_seq[start_end_cds[-1][1]:start_end_cds[-1][1]+3]\n",
    "        #elif strand == '-':\n",
    "        #   wt_dna = chr_seq[start_end_cds[0][0]-3:start_end_cds[0][0]] +  wt_dna\n",
    "\n",
    "        #check location in cds\n",
    "        #get information about the wt/alt sequence\n",
    "        start = val['seq_start']\n",
    "        end = val['seq_end']\n",
    "        wt_seq = val['wt_w_context']\n",
    "        alt_seq = val['alt_w_context']\n",
    "\n",
    "        inc = []\n",
    "        dna_dict = dict(zip(list(range(start, end+1)), list(range(len(wt_seq)))))\n",
    "        for i in range(start, end+1):\n",
    "            if i in codon_locs:\n",
    "                inc.append(i)\n",
    "\n",
    "        if len(inc)==0:\n",
    "            frame = 'None'\n",
    "            frame_list.append(frame)\n",
    "            coding_start_list.append('None')\n",
    "            coding_end_list.append('None')\n",
    "            CDS_WT_correct_orientation.append('None')\n",
    "            CDS_alt_correct_orientation.append('None')\n",
    "            coding_alt_start_list.append('None')\n",
    "            coding_alt_end_list.append('None')\n",
    "            #means it's not in a coding region (could be splice or intron)\n",
    "        \n",
    "        else:\n",
    "            inc_start = inc[0]\n",
    "            inc_end = inc[-1]\n",
    "\n",
    "            #------and then double check that the frame is correct--------------\n",
    "            #by checking if the subsequence is located in the WT sequence\n",
    "\n",
    "            wt_start = dna_dict[inc_start]\n",
    "            wt_end = dna_dict[inc_end] + 1\n",
    "\n",
    "            #record what part of the subsequence is part of a CDS\n",
    "            coding_start_list.append(wt_start)\n",
    "            coding_end_list.append(wt_end)\n",
    "\n",
    "            full_prot = str(Bio.Seq.Seq(wt_dna).transcribe().translate())\n",
    "\n",
    "            subseq = Bio.Seq.Seq(wt_seq[wt_start:wt_end])\n",
    "\n",
    "            #also generate the alternate sequence:\n",
    "            diff = len(alt_seq)-len(wt_seq)\n",
    "            #if strand=='+':\n",
    "            alt_start = wt_start\n",
    "            alt_end = wt_end + diff\n",
    "            #elif strand =='-':\n",
    "            #    alt_start = wt_start - diff\n",
    "            #    alt_end = wt_end\n",
    "                \n",
    "            alt_subseq = Bio.Seq.Seq(alt_seq[alt_start:alt_end])\n",
    "            coding_alt_start_list.append(alt_start)\n",
    "            coding_alt_end_list.append(alt_end)\n",
    "\n",
    "            if strand =='-':\n",
    "                subseq = subseq.reverse_complement()\n",
    "                alt_subseq = alt_subseq.reverse_complement()\n",
    "                inc_start = inc[-1]\n",
    "                inc_end = inc[0]\n",
    "                full_prot = str(Bio.Seq.Seq(wt_dna).reverse_complement().transcribe().translate())\n",
    "                #have to reverse this list as well if the transcript is in the - orientation\n",
    "                codon_locs = codon_locs[::-1]\n",
    "\n",
    "            CDS_WT_correct_orientation.append(str(subseq))\n",
    "            CDS_alt_correct_orientation.append(str(alt_subseq))\n",
    "\n",
    "            #determine the frame\n",
    "            start_idx = codon_locs.index(inc_start)\n",
    "\n",
    "            frame = (3 - (start_idx % 3)) % 3\n",
    "            frame_list.append(frame)\n",
    "            \n",
    "            #and then confirm whether the subsequence is in the protein\n",
    "            sub1 = str(Bio.Seq.Seq(subseq[frame:]).transcribe().translate())\n",
    "            \n",
    "            #assert sub1 in full_prot, print(sub1)\n",
    "            if sub1 not in full_prot:\n",
    "                print('error')\n",
    "                print(index1)\n",
    "                print(val['Variant_Classification'])\n",
    "                print(val['HGVSp_Short'])\n",
    "                #print(strand)\n",
    "                #print(frame)\n",
    "            \n",
    "    exon_subs['Strand'] = strand_list\n",
    "    exon_subs['Frame'] = frame_list\n",
    "    exon_subs['CDS_wt_start'] = coding_start_list\n",
    "    exon_subs['CDS_wt_end'] = coding_end_list\n",
    "    exon_subs['CDS_wt_correct_orientation'] = CDS_WT_correct_orientation\n",
    "    exon_subs['CDS_alt_start'] = coding_alt_start_list\n",
    "    exon_subs['CDS_alt_end'] = coding_alt_end_list\n",
    "    exon_subs['CDS_alt_correct_orientation'] = CDS_alt_correct_orientation\n",
    "\n",
    "    return exon_subs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Diego's library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele1</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>dbSNP_RS</th>\n",
       "      <th>dbSNP_Val_Status</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Matched_Norm_Sample_Barcode</th>\n",
       "      <th>Match_Norm_Seq_Allele1</th>\n",
       "      <th>Match_Norm_Seq_Allele2</th>\n",
       "      <th>Tumor_Validation_Allele1</th>\n",
       "      <th>Tumor_Validation_Allele2</th>\n",
       "      <th>Match_Norm_Validation_Allele1</th>\n",
       "      <th>Match_Norm_Validation_Allele2</th>\n",
       "      <th>...</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>gnomAD_AF</th>\n",
       "      <th>gnomAD_AFR_AF</th>\n",
       "      <th>gnomAD_AMR_AF</th>\n",
       "      <th>gnomAD_ASJ_AF</th>\n",
       "      <th>gnomAD_EAS_AF</th>\n",
       "      <th>gnomAD_FIN_AF</th>\n",
       "      <th>gnomAD_NFE_AF</th>\n",
       "      <th>gnomAD_OTH_AF</th>\n",
       "      <th>gnomAD_SAS_AF</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>Polyphen_Prediction</th>\n",
       "      <th>Polyphen_Score</th>\n",
       "      <th>SIFT_Prediction</th>\n",
       "      <th>SIFT_Score</th>\n",
       "      <th>SWISSPROT</th>\n",
       "      <th>n_depth</th>\n",
       "      <th>t_depth</th>\n",
       "      <th>Annotation_Status</th>\n",
       "      <th>mutationInCis_Flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1809.0</td>\n",
       "      <td>EIF1AX</td>\n",
       "      <td>1964.0</td>\n",
       "      <td>COLU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>X</td>\n",
       "      <td>20156758</td>\n",
       "      <td>20156758</td>\n",
       "      <td>+</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>Intron</td>\n",
       "      <td>DEL</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-COLU-951-1</td>\n",
       "      <td>NORMAL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>p.*6*</td>\n",
       "      <td>ENST00000379607</td>\n",
       "      <td>NM_001412.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1512.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7157.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>17</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>rs28934574</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-JHU-00113-00335</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>p.R282W</td>\n",
       "      <td>ENST00000269305</td>\n",
       "      <td>NM_001126112.2</td>\n",
       "      <td>282.0</td>\n",
       "      <td>Cgg/Tgg</td>\n",
       "      <td>8/11</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PASS</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>1.000</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>728.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>983.0</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>54894.0</td>\n",
       "      <td>DFCI</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>17</td>\n",
       "      <td>56435161</td>\n",
       "      <td>56435161</td>\n",
       "      <td>+</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>rs781215815</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-DFCI-449465-4787362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>p.G659Vfs*41</td>\n",
       "      <td>ENST00000407977</td>\n",
       "      <td>NaN</td>\n",
       "      <td>659.0</td>\n",
       "      <td>gGt/gt</td>\n",
       "      <td>9/10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>263.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>777.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>324.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>5</td>\n",
       "      <td>112175639</td>\n",
       "      <td>112175639</td>\n",
       "      <td>+</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>rs121913332</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-JHU-00223-00468</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>p.R1450*</td>\n",
       "      <td>ENST00000257430</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>16/16</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1391.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>762.0</td>\n",
       "      <td>ASXL1</td>\n",
       "      <td>171023.0</td>\n",
       "      <td>DFCI</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>20</td>\n",
       "      <td>31022441</td>\n",
       "      <td>31022442</td>\n",
       "      <td>+</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>rs756958159</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-DFCI-003409-1958</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>p.G646Wfs*12</td>\n",
       "      <td>ENST00000375687</td>\n",
       "      <td>NM_015338.5</td>\n",
       "      <td>642.0</td>\n",
       "      <td>-/G</td>\n",
       "      <td>13/13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>83.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9090</th>\n",
       "      <td>5.0</td>\n",
       "      <td>ERCC5</td>\n",
       "      <td>2073.0</td>\n",
       "      <td>PROV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>13</td>\n",
       "      <td>103527930</td>\n",
       "      <td>103527931</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>DNP</td>\n",
       "      <td>GG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CA</td>\n",
       "      <td>rs587778291</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-PROV-9b34c1d29d-5652092fc3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>p.G1080Q</td>\n",
       "      <td>ENST00000355739</td>\n",
       "      <td>NM_000123.3</td>\n",
       "      <td>1080.0</td>\n",
       "      <td>GGa/CAa</td>\n",
       "      <td>15/15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.000</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.83</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>789.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9091</th>\n",
       "      <td>5.0</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>84159.0</td>\n",
       "      <td>PROV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>10</td>\n",
       "      <td>63850742</td>\n",
       "      <td>63850742</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-PROV-1a9954b812-1203eb4b3a</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>p.P507L</td>\n",
       "      <td>ENST00000279873</td>\n",
       "      <td>NM_032199.2</td>\n",
       "      <td>507.0</td>\n",
       "      <td>cCc/cTc</td>\n",
       "      <td>10/10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.003</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1204.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9092</th>\n",
       "      <td>5.0</td>\n",
       "      <td>ABL2</td>\n",
       "      <td>27.0</td>\n",
       "      <td>PROV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>179078007</td>\n",
       "      <td>179078007</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>rs781504947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-PROV-6a280d78a3-155ebf48a8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>p.M799V</td>\n",
       "      <td>ENST00000502732</td>\n",
       "      <td>NM_001168237.1</td>\n",
       "      <td>799.0</td>\n",
       "      <td>Atg/Gtg</td>\n",
       "      <td>12/12</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>PASS</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.003</td>\n",
       "      <td>tolerated_low_confidence</td>\n",
       "      <td>1.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1230.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9093</th>\n",
       "      <td>5.0</td>\n",
       "      <td>SOX10</td>\n",
       "      <td>6663.0</td>\n",
       "      <td>PROV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>22</td>\n",
       "      <td>38369885</td>\n",
       "      <td>38369885</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>rs549034055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-PROV-238a11a4e2-b283f19879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>p.V340M</td>\n",
       "      <td>ENST00000360880</td>\n",
       "      <td>NaN</td>\n",
       "      <td>340.0</td>\n",
       "      <td>Gtg/Atg</td>\n",
       "      <td>5/5</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>PASS</td>\n",
       "      <td>possibly_damaging</td>\n",
       "      <td>0.622</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>641.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9094</th>\n",
       "      <td>5.0</td>\n",
       "      <td>AMER1</td>\n",
       "      <td>139285.0</td>\n",
       "      <td>PROV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>X</td>\n",
       "      <td>63410992</td>\n",
       "      <td>63410992</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>rs369699027</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-PROV-26142c27e9-be569ca40c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>p.D725E</td>\n",
       "      <td>ENST00000330258</td>\n",
       "      <td>NM_152424.3</td>\n",
       "      <td>725.0</td>\n",
       "      <td>gaT/gaG</td>\n",
       "      <td>2/2</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0.000293</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PASS</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.028</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>741.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9095 rows Ã— 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       COUNT Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  \\\n",
       "0     1809.0      EIF1AX          1964.0   COLU     GRCh37          X   \n",
       "1     1512.0        TP53          7157.0    JHU     GRCh37         17   \n",
       "2      983.0       RNF43         54894.0   DFCI     GRCh37         17   \n",
       "3      777.0         APC           324.0    JHU     GRCh37          5   \n",
       "4      762.0       ASXL1        171023.0   DFCI     GRCh37         20   \n",
       "...      ...         ...             ...    ...        ...        ...   \n",
       "9090     5.0       ERCC5          2073.0   PROV     GRCh37         13   \n",
       "9091     5.0      ARID5B         84159.0   PROV     GRCh37         10   \n",
       "9092     5.0        ABL2            27.0   PROV     GRCh37          1   \n",
       "9093     5.0       SOX10          6663.0   PROV     GRCh37         22   \n",
       "9094     5.0       AMER1        139285.0   PROV     GRCh37          X   \n",
       "\n",
       "      Start_Position  End_Position Strand         Consequence  \\\n",
       "0           20156758      20156758      +      intron_variant   \n",
       "1            7577094       7577094      +    missense_variant   \n",
       "2           56435161      56435161      +  frameshift_variant   \n",
       "3          112175639     112175639      +         stop_gained   \n",
       "4           31022441      31022442      +  frameshift_variant   \n",
       "...              ...           ...    ...                 ...   \n",
       "9090       103527930     103527931      +    missense_variant   \n",
       "9091        63850742      63850742      +    missense_variant   \n",
       "9092       179078007     179078007      +    missense_variant   \n",
       "9093        38369885      38369885      +    missense_variant   \n",
       "9094        63410992      63410992      +    missense_variant   \n",
       "\n",
       "     Variant_Classification Variant_Type Reference_Allele Tumor_Seq_Allele1  \\\n",
       "0                    Intron          DEL                A                 A   \n",
       "1         Missense_Mutation          SNP                G               NaN   \n",
       "2           Frame_Shift_Del          DEL                C                 C   \n",
       "3         Nonsense_Mutation          SNP                C               NaN   \n",
       "4           Frame_Shift_Ins          INS                -                 -   \n",
       "...                     ...          ...              ...               ...   \n",
       "9090      Missense_Mutation          DNP               GG               NaN   \n",
       "9091      Missense_Mutation          SNP                C               NaN   \n",
       "9092      Missense_Mutation          SNP                T               NaN   \n",
       "9093      Missense_Mutation          SNP                C               NaN   \n",
       "9094      Missense_Mutation          SNP                A               NaN   \n",
       "\n",
       "     Tumor_Seq_Allele2     dbSNP_RS  dbSNP_Val_Status  \\\n",
       "0                    -        novel               NaN   \n",
       "1                    A   rs28934574               NaN   \n",
       "2                    -  rs781215815               NaN   \n",
       "3                    T  rs121913332               NaN   \n",
       "4                    G  rs756958159               NaN   \n",
       "...                ...          ...               ...   \n",
       "9090                CA  rs587778291               NaN   \n",
       "9091                 T          NaN               NaN   \n",
       "9092                 C  rs781504947               NaN   \n",
       "9093                 T  rs549034055               NaN   \n",
       "9094                 C  rs369699027               NaN   \n",
       "\n",
       "                  Tumor_Sample_Barcode Matched_Norm_Sample_Barcode  \\\n",
       "0                     GENIE-COLU-951-1                      NORMAL   \n",
       "1                GENIE-JHU-00113-00335                         NaN   \n",
       "2            GENIE-DFCI-449465-4787362                         NaN   \n",
       "3                GENIE-JHU-00223-00468                         NaN   \n",
       "4               GENIE-DFCI-003409-1958                         NaN   \n",
       "...                                ...                         ...   \n",
       "9090  GENIE-PROV-9b34c1d29d-5652092fc3                         NaN   \n",
       "9091  GENIE-PROV-1a9954b812-1203eb4b3a                         NaN   \n",
       "9092  GENIE-PROV-6a280d78a3-155ebf48a8                         NaN   \n",
       "9093  GENIE-PROV-238a11a4e2-b283f19879                         NaN   \n",
       "9094  GENIE-PROV-26142c27e9-be569ca40c                         NaN   \n",
       "\n",
       "     Match_Norm_Seq_Allele1 Match_Norm_Seq_Allele2  Tumor_Validation_Allele1  \\\n",
       "0                       NaN                    NaN                       NaN   \n",
       "1                       NaN                    NaN                       NaN   \n",
       "2                       NaN                    NaN                       NaN   \n",
       "3                       NaN                    NaN                       NaN   \n",
       "4                       NaN                    NaN                       NaN   \n",
       "...                     ...                    ...                       ...   \n",
       "9090                    NaN                    NaN                       NaN   \n",
       "9091                    NaN                    NaN                       NaN   \n",
       "9092                    NaN                    NaN                       NaN   \n",
       "9093                    NaN                    NaN                       NaN   \n",
       "9094                    NaN                    NaN                       NaN   \n",
       "\n",
       "      Tumor_Validation_Allele2  Match_Norm_Validation_Allele1  \\\n",
       "0                          NaN                            NaN   \n",
       "1                          NaN                            NaN   \n",
       "2                          NaN                            NaN   \n",
       "3                          NaN                            NaN   \n",
       "4                          NaN                            NaN   \n",
       "...                        ...                            ...   \n",
       "9090                       NaN                            NaN   \n",
       "9091                       NaN                            NaN   \n",
       "9092                       NaN                            NaN   \n",
       "9093                       NaN                            NaN   \n",
       "9094                       NaN                            NaN   \n",
       "\n",
       "      Match_Norm_Validation_Allele2  ...   HGVSp_Short    Transcript_ID  \\\n",
       "0                               NaN  ...         p.*6*  ENST00000379607   \n",
       "1                               NaN  ...       p.R282W  ENST00000269305   \n",
       "2                               NaN  ...  p.G659Vfs*41  ENST00000407977   \n",
       "3                               NaN  ...      p.R1450*  ENST00000257430   \n",
       "4                               NaN  ...  p.G646Wfs*12  ENST00000375687   \n",
       "...                             ...  ...           ...              ...   \n",
       "9090                            NaN  ...      p.G1080Q  ENST00000355739   \n",
       "9091                            NaN  ...       p.P507L  ENST00000279873   \n",
       "9092                            NaN  ...       p.M799V  ENST00000502732   \n",
       "9093                            NaN  ...       p.V340M  ENST00000360880   \n",
       "9094                            NaN  ...       p.D725E  ENST00000330258   \n",
       "\n",
       "              RefSeq  Protein_position   Codons  Exon_Number  gnomAD_AF  \\\n",
       "0        NM_001412.3               NaN      NaN          NaN        NaN   \n",
       "1     NM_001126112.2             282.0  Cgg/Tgg         8/11   0.000004   \n",
       "2                NaN             659.0   gGt/gt         9/10        NaN   \n",
       "3        NM_000038.5            1450.0  Cga/Tga        16/16   0.000000   \n",
       "4        NM_015338.5             642.0      -/G        13/13        NaN   \n",
       "...              ...               ...      ...          ...        ...   \n",
       "9090     NM_000123.3            1080.0  GGa/CAa        15/15        NaN   \n",
       "9091     NM_032199.2             507.0  cCc/cTc        10/10        NaN   \n",
       "9092  NM_001168237.1             799.0  Atg/Gtg        12/12   0.000044   \n",
       "9093             NaN             340.0  Gtg/Atg          5/5   0.000020   \n",
       "9094     NM_152424.3             725.0  gaT/gaG          2/2   0.000142   \n",
       "\n",
       "      gnomAD_AFR_AF  gnomAD_AMR_AF  gnomAD_ASJ_AF  gnomAD_EAS_AF  \\\n",
       "0               NaN            NaN            NaN            NaN   \n",
       "1               0.0       0.000000            0.0       0.000000   \n",
       "2               NaN            NaN            NaN            NaN   \n",
       "3               0.0       0.000000            0.0       0.000000   \n",
       "4               NaN            NaN            NaN            NaN   \n",
       "...             ...            ...            ...            ...   \n",
       "9090            NaN            NaN            NaN            NaN   \n",
       "9091            NaN            NaN            NaN            NaN   \n",
       "9092            0.0       0.000029            0.0       0.000000   \n",
       "9093            0.0       0.000029            0.0       0.000055   \n",
       "9094            0.0       0.000000            0.0       0.000000   \n",
       "\n",
       "      gnomAD_FIN_AF  gnomAD_NFE_AF gnomAD_OTH_AF gnomAD_SAS_AF FILTER  \\\n",
       "0               NaN            NaN           NaN           NaN   PASS   \n",
       "1          0.000000       0.000009      0.000000      0.000000   PASS   \n",
       "2               NaN            NaN           NaN           NaN   PASS   \n",
       "3          0.000000       0.000000      0.000000      0.000000   PASS   \n",
       "4               NaN            NaN           NaN           NaN   PASS   \n",
       "...             ...            ...           ...           ...    ...   \n",
       "9090            NaN            NaN           NaN           NaN   PASS   \n",
       "9091            NaN            NaN           NaN           NaN   PASS   \n",
       "9092       0.000092       0.000053      0.000163      0.000033   PASS   \n",
       "9093       0.000000       0.000009      0.000000      0.000065   PASS   \n",
       "9094       0.000125       0.000293      0.000000      0.000000   PASS   \n",
       "\n",
       "     Polyphen_Prediction Polyphen_Score           SIFT_Prediction SIFT_Score  \\\n",
       "0                    NaN            NaN                       NaN        NaN   \n",
       "1      probably_damaging          1.000               deleterious       0.00   \n",
       "2                    NaN            NaN                       NaN        NaN   \n",
       "3                    NaN            NaN                       NaN        NaN   \n",
       "4                    NaN            NaN                       NaN        NaN   \n",
       "...                  ...            ...                       ...        ...   \n",
       "9090              benign          0.000                 tolerated       0.83   \n",
       "9091              benign          0.003                 tolerated       0.25   \n",
       "9092              benign          0.003  tolerated_low_confidence       1.00   \n",
       "9093   possibly_damaging          0.622                 tolerated       0.26   \n",
       "9094              benign          0.028               deleterious       0.04   \n",
       "\n",
       "     SWISSPROT  n_depth  t_depth  Annotation_Status  mutationInCis_Flag  \n",
       "0          NaN      NaN      NaN            SUCCESS               False  \n",
       "1          NaN      NaN    728.0            SUCCESS               False  \n",
       "2          NaN      NaN    263.0            SUCCESS               False  \n",
       "3          NaN      NaN   1391.0            SUCCESS               False  \n",
       "4          NaN      NaN     83.0            SUCCESS               False  \n",
       "...        ...      ...      ...                ...                 ...  \n",
       "9090       NaN      NaN    789.0            SUCCESS               False  \n",
       "9091       NaN      NaN   1204.0            SUCCESS               False  \n",
       "9092       NaN      NaN   1230.0            SUCCESS               False  \n",
       "9093       NaN      NaN    641.0            SUCCESS               False  \n",
       "9094       NaN      NaN    741.0            SUCCESS               False  \n",
       "\n",
       "[9095 rows x 65 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idr = pd.read_csv('filtered_idr_mutations_5count_6nt_indels.csv')\n",
    "\n",
    "#filtering out the complex indels (i.e. non-pure INS or DEL)\n",
    "dels1 = idr[idr['Variant_Type']=='DEL']\n",
    "bad_del_idx = list(dels1[dels1['Tumor_Seq_Allele2']!='-'].index)\n",
    "\n",
    "\n",
    "ins1 = idr[idr['Variant_Type']=='INS']\n",
    "bad_ins_idx = list(ins1[ins1['Reference_Allele']!='-'].index)\n",
    "\n",
    "bad_idxs = bad_del_idx + bad_ins_idx\n",
    "\n",
    "idr_filtered = idr.drop(index=bad_idxs).reset_index().drop(columns='index')\n",
    "\n",
    "idr_filtered\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([\"3'UTR\", \"5'UTR\", 'Frame_Shift_Del', 'Frame_Shift_Ins',\n",
       "       'In_Frame_Del', 'In_Frame_Ins', 'Intron', 'Missense_Mutation',\n",
       "       'Nonsense_Mutation', 'Silent', 'Splice_Region', 'Splice_Site',\n",
       "       'Translation_Start_Site'], dtype=object)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(idr_filtered['Variant_Classification'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#also add in the transcript information\n",
    "tx_new = []\n",
    "for i, val in idr_filtered.iterrows():\n",
    "    h = val['HGVSc']\n",
    "    tx_new.append(h.split(':')[0])\n",
    "\n",
    "idr_filtered['transcript_id_TRUE'] = tx_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "idr_oligos = df_formatter(idr_filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error\n",
      "3266\n",
      "Intron\n",
      "p.*1354*\n"
     ]
    }
   ],
   "source": [
    "idr_oligos = frame_determiner(idr_oligos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP\n",
      "-\n",
      "0\n",
      "p.M1?\n",
      "MAEAPQVVEIDPDFEPLPRP\n",
      "VAEAPQVVEIDPDFEPLPRP\n"
     ]
    }
   ],
   "source": [
    "#and spot checking\n",
    "i = 2609\n",
    "\n",
    "#i = list(idr_oligos[idr_oligos['Variant_Type']=='DEL'][0:50].index)[10]\n",
    "\n",
    "wt1 = idr_oligos.iloc[i]['CDS_wt_correct_orientation']\n",
    "alt1 = idr_oligos.loc[i]['CDS_alt_correct_orientation']\n",
    "frame = idr_oligos.iloc[i]['Frame']\n",
    "hgvsp = idr_oligos.iloc[i]['HGVSp_Short']\n",
    "s = idr_oligos.iloc[i]['Strand']\n",
    "v = idr_oligos.iloc[i]['Variant_Type']\n",
    "print(v)\n",
    "print(s)\n",
    "print(frame)\n",
    "print(hgvsp)\n",
    "print(Bio.Seq.Seq(wt1[frame:]).transcribe().translate())\n",
    "print(Bio.Seq.Seq(alt1[frame:]).transcribe().translate())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([\"3'UTR\", \"5'UTR\", 'Frame_Shift_Del', 'Frame_Shift_Ins',\n",
       "       'In_Frame_Del', 'In_Frame_Ins', 'Intron', 'Missense_Mutation',\n",
       "       'Nonsense_Mutation', 'Silent', 'Splice_Region', 'Splice_Site',\n",
       "       'Translation_Start_Site'], dtype=object)"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#last step = noting which of these oligos are coding, and thus amenable to MMR\n",
    "np.unique(idr_oligos['Variant_Classification'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {},
   "outputs": [],
   "source": [
    "#last step = noting which of these oligos are coding, and thus amenable to MMR\n",
    "\n",
    "cols = ['Frame_Shift_Del', 'Frame_Shift_Ins','In_Frame_Del', 'In_Frame_Ins','Missense_Mutation','Nonsense_Mutation', 'Silent','Translation_Start_Site']\n",
    "#cols2 = ['Translation_Start_Site']\n",
    "#idr_oligos[idr_oligos['Variant_Classification'].isin(cols)]\n",
    "\n",
    "mmr_evasive_amenable = []\n",
    "for i, val in idr_oligos.iterrows():\n",
    "    cl = val['Variant_Classification']\n",
    "    if cl in cols:\n",
    "        mmr_evasive_amenable.append(True)\n",
    "    else:\n",
    "        mmr_evasive_amenable.append(False)\n",
    "\n",
    "idr_oligos['MMR_evasive_amenable']=mmr_evasive_amenable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>CDS_wt_start</th>\n",
       "      <th>CDS_wt_end</th>\n",
       "      <th>CDS_wt_correct_orientation</th>\n",
       "      <th>CDS_alt_start</th>\n",
       "      <th>CDS_alt_end</th>\n",
       "      <th>CDS_alt_correct_orientation</th>\n",
       "      <th>MMR_evasive_amenable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1512.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-JHU-00113-00335</td>\n",
       "      <td>NM_001126112.2</td>\n",
       "      <td>282.0</td>\n",
       "      <td>8/11</td>\n",
       "      <td>ENST00000269305.4:c.844C&gt;T</td>\n",
       "      <td>p.Arg282Trp</td>\n",
       "      <td>p.R282W</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577034</td>\n",
       "      <td>7577154</td>\n",
       "      <td>CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...</td>\n",
       "      <td>CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>GGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCT...</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>GGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCT...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>983.0</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>17</td>\n",
       "      <td>56435161</td>\n",
       "      <td>56435161</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-DFCI-449465-4787362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>659.0</td>\n",
       "      <td>9/10</td>\n",
       "      <td>ENST00000407977.2:c.1976del</td>\n",
       "      <td>p.Gly659ValfsTer41</td>\n",
       "      <td>p.G659Vfs*41</td>\n",
       "      <td>ENST00000407977.2</td>\n",
       "      <td>56435101</td>\n",
       "      <td>56435221</td>\n",
       "      <td>TGGCAAGCTGGGTGCACAGTTGCATCCTGGGGCCGAGAGCCAGGGG...</td>\n",
       "      <td>TGGCAAGCTGGGTGCACAGTTGCATCCTGGGGCCGAGAGCCAGGGG...</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>TGTTCAACTTGCAAAAATCCAGCCTCTCTGCCCGACACCCACAGAG...</td>\n",
       "      <td>0</td>\n",
       "      <td>120</td>\n",
       "      <td>TGTTCAACTTGCAAAAATCCAGCCTCTCTGCCCGACACCCACAGAG...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>777.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175639</td>\n",
       "      <td>112175639</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00223-00468</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4348C&gt;T</td>\n",
       "      <td>p.Arg1450Ter</td>\n",
       "      <td>p.R1450*</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175579</td>\n",
       "      <td>112175699</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>762.0</td>\n",
       "      <td>ASXL1</td>\n",
       "      <td>20</td>\n",
       "      <td>31022441</td>\n",
       "      <td>31022442</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-003409-1958</td>\n",
       "      <td>NM_015338.5</td>\n",
       "      <td>642.0</td>\n",
       "      <td>13/13</td>\n",
       "      <td>ENST00000375687.4:c.1934dup</td>\n",
       "      <td>p.Gly646TrpfsTer12</td>\n",
       "      <td>p.G646Wfs*12</td>\n",
       "      <td>ENST00000375687.4</td>\n",
       "      <td>31022381</td>\n",
       "      <td>31022502</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>122</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "      <td>0</td>\n",
       "      <td>123</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>645.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175952</td>\n",
       "      <td>112175953</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-JHU-00198-00378</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1554.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4666dup</td>\n",
       "      <td>p.Thr1556AsnfsTer3</td>\n",
       "      <td>p.T1556Nfs*3</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175892</td>\n",
       "      <td>112176013</td>\n",
       "      <td>GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...</td>\n",
       "      <td>GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>122</td>\n",
       "      <td>GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...</td>\n",
       "      <td>0</td>\n",
       "      <td>123</td>\n",
       "      <td>GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9090</th>\n",
       "      <td>9090</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ERCC5</td>\n",
       "      <td>13</td>\n",
       "      <td>103527930</td>\n",
       "      <td>103527931</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>DNP</td>\n",
       "      <td>GG</td>\n",
       "      <td>CA</td>\n",
       "      <td>GENIE-PROV-9b34c1d29d-5652092fc3</td>\n",
       "      <td>NM_000123.3</td>\n",
       "      <td>1080.0</td>\n",
       "      <td>15/15</td>\n",
       "      <td>ENST00000355739.4:c.3238_3239delinsCA</td>\n",
       "      <td>p.Gly1080Gln</td>\n",
       "      <td>p.G1080Q</td>\n",
       "      <td>ENST00000355739.4</td>\n",
       "      <td>103527870</td>\n",
       "      <td>103527991</td>\n",
       "      <td>ATAACAAATACCTTAGAAGAGTCATCAAGCCTGAAAAGAAAGAGGC...</td>\n",
       "      <td>ATAACAAATACCTTAGAAGAGTCATCAAGCCTGAAAAGAAAGAGGC...</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>122</td>\n",
       "      <td>ATAACAAATACCTTAGAAGAGTCATCAAGCCTGAAAAGAAAGAGGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>122</td>\n",
       "      <td>ATAACAAATACCTTAGAAGAGTCATCAAGCCTGAAAAGAAAGAGGC...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9091</th>\n",
       "      <td>9091</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>10</td>\n",
       "      <td>63850742</td>\n",
       "      <td>63850742</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-PROV-1a9954b812-1203eb4b3a</td>\n",
       "      <td>NM_032199.2</td>\n",
       "      <td>507.0</td>\n",
       "      <td>10/10</td>\n",
       "      <td>ENST00000279873.7:c.1520C&gt;T</td>\n",
       "      <td>p.Pro507Leu</td>\n",
       "      <td>p.P507L</td>\n",
       "      <td>ENST00000279873.7</td>\n",
       "      <td>63850682</td>\n",
       "      <td>63850802</td>\n",
       "      <td>CTCTCCCAGCAGCAGACATGAAGAAAAAAATAGAAGGGTATCAGGA...</td>\n",
       "      <td>CTCTCCCAGCAGCAGACATGAAGAAAAAAATAGAAGGGTATCAGGA...</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>CTCTCCCAGCAGCAGACATGAAGAAAAAAATAGAAGGGTATCAGGA...</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>CTCTCCCAGCAGCAGACATGAAGAAAAAAATAGAAGGGTATCAGGA...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9092</th>\n",
       "      <td>9092</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ABL2</td>\n",
       "      <td>1</td>\n",
       "      <td>179078007</td>\n",
       "      <td>179078007</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>GENIE-PROV-6a280d78a3-155ebf48a8</td>\n",
       "      <td>NM_001168237.1</td>\n",
       "      <td>799.0</td>\n",
       "      <td>12/12</td>\n",
       "      <td>ENST00000502732.1:c.2395A&gt;G</td>\n",
       "      <td>p.Met799Val</td>\n",
       "      <td>p.M799V</td>\n",
       "      <td>ENST00000502732.1</td>\n",
       "      <td>179077947</td>\n",
       "      <td>179078067</td>\n",
       "      <td>AGGTGGACACTGTCCTTTCCAGCTGGAGTTTGGACCTCTGGCAGTT...</td>\n",
       "      <td>AGGTGGACACTGTCCTTTCCAGCTGGAGTTTGGACCTCTGGCAGTT...</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>CCAAGGTCAAACTCTACATCTTCCATGTCCTCAGGGCTTCCAGAGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>CCAAGGTCAAACTCTACATCTTCCATGTCCTCAGGGCTTCCAGAGC...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9093</th>\n",
       "      <td>9093</td>\n",
       "      <td>5.0</td>\n",
       "      <td>SOX10</td>\n",
       "      <td>22</td>\n",
       "      <td>38369885</td>\n",
       "      <td>38369885</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-PROV-238a11a4e2-b283f19879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>340.0</td>\n",
       "      <td>5/5</td>\n",
       "      <td>ENST00000360880.2:c.1018G&gt;A</td>\n",
       "      <td>p.Val340Met</td>\n",
       "      <td>p.V340M</td>\n",
       "      <td>ENST00000360880.2</td>\n",
       "      <td>38369825</td>\n",
       "      <td>38369945</td>\n",
       "      <td>TCTCTGTCTTCACCTGGGCTTTGGCATCCACACCAGGTGGTGAGAC...</td>\n",
       "      <td>TCTCTGTCTTCACCTGGGCTTTGGCATCCACACCAGGTGGTGAGAC...</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>CTGGGCAGTGCCCTGGCCGTGGCCAGTGGACACTCCGCCTGGATCT...</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>CTGGGCAGTGCCCTGGCCGTGGCCAGTGGACACTCCGCCTGGATCT...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9094</th>\n",
       "      <td>9094</td>\n",
       "      <td>5.0</td>\n",
       "      <td>AMER1</td>\n",
       "      <td>X</td>\n",
       "      <td>63410992</td>\n",
       "      <td>63410992</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>GENIE-PROV-26142c27e9-be569ca40c</td>\n",
       "      <td>NM_152424.3</td>\n",
       "      <td>725.0</td>\n",
       "      <td>2/2</td>\n",
       "      <td>ENST00000330258.3:c.2175T&gt;G</td>\n",
       "      <td>p.Asp725Glu</td>\n",
       "      <td>p.D725E</td>\n",
       "      <td>ENST00000330258.3</td>\n",
       "      <td>63410932</td>\n",
       "      <td>63411052</td>\n",
       "      <td>GTAGGCCCTCCTGGGAGATCCTCCAAAATTTGCTTCTTGCATGTCT...</td>\n",
       "      <td>GTAGGCCCTCCTGGGAGATCCTCCAAAATTTGCTTCTTGCATGTCT...</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>TGAAGGAACCTGCTCCAAGAAAGATCAAAGTACCTGCCTGATGCAG...</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>TGAAGGAACCTGCTCCAAGAAAGATCAAAGTACCTGCCTGATGCAG...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8567 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index   COUNT Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "1         1  1512.0        TP53         17         7577094       7577094   \n",
       "2         2   983.0       RNF43         17        56435161      56435161   \n",
       "3         3   777.0         APC          5       112175639     112175639   \n",
       "4         4   762.0       ASXL1         20        31022441      31022442   \n",
       "5         5   645.0         APC          5       112175952     112175953   \n",
       "...     ...     ...         ...        ...             ...           ...   \n",
       "9090   9090     5.0       ERCC5         13       103527930     103527931   \n",
       "9091   9091     5.0      ARID5B         10        63850742      63850742   \n",
       "9092   9092     5.0        ABL2          1       179078007     179078007   \n",
       "9093   9093     5.0       SOX10         22        38369885      38369885   \n",
       "9094   9094     5.0       AMER1          X        63410992      63410992   \n",
       "\n",
       "             Consequence Variant_Classification Variant_Type Reference_Allele  \\\n",
       "1       missense_variant      Missense_Mutation          SNP                G   \n",
       "2     frameshift_variant        Frame_Shift_Del          DEL                C   \n",
       "3            stop_gained      Nonsense_Mutation          SNP                C   \n",
       "4     frameshift_variant        Frame_Shift_Ins          INS                -   \n",
       "5     frameshift_variant        Frame_Shift_Ins          INS                -   \n",
       "...                  ...                    ...          ...              ...   \n",
       "9090    missense_variant      Missense_Mutation          DNP               GG   \n",
       "9091    missense_variant      Missense_Mutation          SNP                C   \n",
       "9092    missense_variant      Missense_Mutation          SNP                T   \n",
       "9093    missense_variant      Missense_Mutation          SNP                C   \n",
       "9094    missense_variant      Missense_Mutation          SNP                A   \n",
       "\n",
       "     Tumor_Seq_Allele2              Tumor_Sample_Barcode          RefSeq  \\\n",
       "1                    A             GENIE-JHU-00113-00335  NM_001126112.2   \n",
       "2                    -         GENIE-DFCI-449465-4787362             NaN   \n",
       "3                    T             GENIE-JHU-00223-00468     NM_000038.5   \n",
       "4                    G            GENIE-DFCI-003409-1958     NM_015338.5   \n",
       "5                    A             GENIE-JHU-00198-00378     NM_000038.5   \n",
       "...                ...                               ...             ...   \n",
       "9090                CA  GENIE-PROV-9b34c1d29d-5652092fc3     NM_000123.3   \n",
       "9091                 T  GENIE-PROV-1a9954b812-1203eb4b3a     NM_032199.2   \n",
       "9092                 C  GENIE-PROV-6a280d78a3-155ebf48a8  NM_001168237.1   \n",
       "9093                 T  GENIE-PROV-238a11a4e2-b283f19879             NaN   \n",
       "9094                 C  GENIE-PROV-26142c27e9-be569ca40c     NM_152424.3   \n",
       "\n",
       "      Protein_position Exon_Number                                  HGVSc  \\\n",
       "1                282.0        8/11             ENST00000269305.4:c.844C>T   \n",
       "2                659.0        9/10            ENST00000407977.2:c.1976del   \n",
       "3               1450.0       16/16            ENST00000257430.4:c.4348C>T   \n",
       "4                642.0       13/13            ENST00000375687.4:c.1934dup   \n",
       "5               1554.0       16/16            ENST00000257430.4:c.4666dup   \n",
       "...                ...         ...                                    ...   \n",
       "9090            1080.0       15/15  ENST00000355739.4:c.3238_3239delinsCA   \n",
       "9091             507.0       10/10            ENST00000279873.7:c.1520C>T   \n",
       "9092             799.0       12/12            ENST00000502732.1:c.2395A>G   \n",
       "9093             340.0         5/5            ENST00000360880.2:c.1018G>A   \n",
       "9094             725.0         2/2            ENST00000330258.3:c.2175T>G   \n",
       "\n",
       "                   HGVSp   HGVSp_Short transcript_id_TRUE  seq_start  \\\n",
       "1            p.Arg282Trp       p.R282W  ENST00000269305.4    7577034   \n",
       "2     p.Gly659ValfsTer41  p.G659Vfs*41  ENST00000407977.2   56435101   \n",
       "3           p.Arg1450Ter      p.R1450*  ENST00000257430.4  112175579   \n",
       "4     p.Gly646TrpfsTer12  p.G646Wfs*12  ENST00000375687.4   31022381   \n",
       "5     p.Thr1556AsnfsTer3  p.T1556Nfs*3  ENST00000257430.4  112175892   \n",
       "...                  ...           ...                ...        ...   \n",
       "9090        p.Gly1080Gln      p.G1080Q  ENST00000355739.4  103527870   \n",
       "9091         p.Pro507Leu       p.P507L  ENST00000279873.7   63850682   \n",
       "9092         p.Met799Val       p.M799V  ENST00000502732.1  179077947   \n",
       "9093         p.Val340Met       p.V340M  ENST00000360880.2   38369825   \n",
       "9094         p.Asp725Glu       p.D725E  ENST00000330258.3   63410932   \n",
       "\n",
       "        seq_end                                       wt_w_context  \\\n",
       "1       7577154  CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...   \n",
       "2      56435221  TGGCAAGCTGGGTGCACAGTTGCATCCTGGGGCCGAGAGCCAGGGG...   \n",
       "3     112175699  ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...   \n",
       "4      31022502  GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...   \n",
       "5     112176013  GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...   \n",
       "...         ...                                                ...   \n",
       "9090  103527991  ATAACAAATACCTTAGAAGAGTCATCAAGCCTGAAAAGAAAGAGGC...   \n",
       "9091   63850802  CTCTCCCAGCAGCAGACATGAAGAAAAAAATAGAAGGGTATCAGGA...   \n",
       "9092  179078067  AGGTGGACACTGTCCTTTCCAGCTGGAGTTTGGACCTCTGGCAGTT...   \n",
       "9093   38369945  TCTCTGTCTTCACCTGGGCTTTGGCATCCACACCAGGTGGTGAGAC...   \n",
       "9094   63411052  GTAGGCCCTCCTGGGAGATCCTCCAAAATTTGCTTCTTGCATGTCT...   \n",
       "\n",
       "                                          alt_w_context Strand Frame  \\\n",
       "1     CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...      -     0   \n",
       "2     TGGCAAGCTGGGTGCACAGTTGCATCCTGGGGCCGAGAGCCAGGGG...      -     2   \n",
       "3     ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...      +     0   \n",
       "4     GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...      +     1   \n",
       "5     GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...      +     2   \n",
       "...                                                 ...    ...   ...   \n",
       "9090  ATAACAAATACCTTAGAAGAGTCATCAAGCCTGAAAAGAAAGAGGC...      +     0   \n",
       "9091  CTCTCCCAGCAGCAGACATGAAGAAAAAAATAGAAGGGTATCAGGA...      +     2   \n",
       "9092  AGGTGGACACTGTCCTTTCCAGCTGGAGTTTGGACCTCTGGCAGTT...      -     0   \n",
       "9093  TCTCTGTCTTCACCTGGGCTTTGGCATCCACACCAGGTGGTGAGAC...      -     0   \n",
       "9094  GTAGGCCCTCCTGGGAGATCCTCCAAAATTTGCTTCTTGCATGTCT...      -     1   \n",
       "\n",
       "     CDS_wt_start CDS_wt_end  \\\n",
       "1               0        121   \n",
       "2               0        121   \n",
       "3               0        121   \n",
       "4               0        122   \n",
       "5               0        122   \n",
       "...           ...        ...   \n",
       "9090            0        122   \n",
       "9091            0        121   \n",
       "9092            0        121   \n",
       "9093            0        121   \n",
       "9094            0        121   \n",
       "\n",
       "                             CDS_wt_correct_orientation CDS_alt_start  \\\n",
       "1     GGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCT...             0   \n",
       "2     TGTTCAACTTGCAAAAATCCAGCCTCTCTGCCCGACACCCACAGAG...             0   \n",
       "3     ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...             0   \n",
       "4     GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...             0   \n",
       "5     GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...             0   \n",
       "...                                                 ...           ...   \n",
       "9090  ATAACAAATACCTTAGAAGAGTCATCAAGCCTGAAAAGAAAGAGGC...             0   \n",
       "9091  CTCTCCCAGCAGCAGACATGAAGAAAAAAATAGAAGGGTATCAGGA...             0   \n",
       "9092  CCAAGGTCAAACTCTACATCTTCCATGTCCTCAGGGCTTCCAGAGC...             0   \n",
       "9093  CTGGGCAGTGCCCTGGCCGTGGCCAGTGGACACTCCGCCTGGATCT...             0   \n",
       "9094  TGAAGGAACCTGCTCCAAGAAAGATCAAAGTACCTGCCTGATGCAG...             0   \n",
       "\n",
       "     CDS_alt_end                        CDS_alt_correct_orientation  \\\n",
       "1            121  GGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCT...   \n",
       "2            120  TGTTCAACTTGCAAAAATCCAGCCTCTCTGCCCGACACCCACAGAG...   \n",
       "3            121  ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...   \n",
       "4            123  GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...   \n",
       "5            123  GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...   \n",
       "...          ...                                                ...   \n",
       "9090         122  ATAACAAATACCTTAGAAGAGTCATCAAGCCTGAAAAGAAAGAGGC...   \n",
       "9091         121  CTCTCCCAGCAGCAGACATGAAGAAAAAAATAGAAGGGTATCAGGA...   \n",
       "9092         121  CCAAGGTCAAACTCTACATCTTCCATGTCCTCAGGGCTTCCAGAGC...   \n",
       "9093         121  CTGGGCAGTGCCCTGGCCGTGGCCAGTGGACACTCCGCCTGGATCT...   \n",
       "9094         121  TGAAGGAACCTGCTCCAAGAAAGATCAAAGTACCTGCCTGATGCAG...   \n",
       "\n",
       "      MMR_evasive_amenable  \n",
       "1                     True  \n",
       "2                     True  \n",
       "3                     True  \n",
       "4                     True  \n",
       "5                     True  \n",
       "...                    ...  \n",
       "9090                  True  \n",
       "9091                  True  \n",
       "9092                  True  \n",
       "9093                  True  \n",
       "9094                  True  \n",
       "\n",
       "[8567 rows x 32 columns]"
      ]
     },
     "execution_count": 259,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idr_oligos[idr_oligos['MMR_evasive_amenable']==True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [],
   "source": [
    "#idr_oligos.to_csv('Diego_IDR_library_oligos.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# p53 library generation\n",
    "\n",
    "- Working off of the initial/original p53 input library\n",
    "- Going to get all of the CORRECT insertion/deletions this time..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {},
   "outputs": [],
   "source": [
    "# step 1 = filter out large INS/DEL (over 10 nt)\n",
    "# step 2 = filter out complex indels (non-pure INS/DEL)\n",
    "\n",
    "p53 = pd.read_csv('p53_combined_FINAL2.csv')\n",
    "\n",
    "cutoff_size = 10\n",
    "\n",
    "exclude_list = []\n",
    "#step 1 filtration\n",
    "for i, val in p53.iterrows():\n",
    "    vt = val['Variant_Type']\n",
    "    ref = val['Reference_Allele']\n",
    "    alt = val['Tumor_Seq_Allele2']\n",
    "    if vt =='INS':\n",
    "        size = len(alt)\n",
    "        if size >cutoff_size:\n",
    "            exclude_list.append(i)\n",
    "    elif vt=='DEL':\n",
    "        size = len(ref)\n",
    "        if size >cutoff_size:\n",
    "            exclude_list.append(i)\n",
    "\n",
    "p53 = p53.drop(index=exclude_list).reset_index().drop(columns='index')\n",
    "\n",
    "#step 2 filtration\n",
    "dels1 = p53[p53['Variant_Type']=='DEL']\n",
    "bad_del_idx = list(dels1[dels1['Tumor_Seq_Allele2']!='-'].index)\n",
    "\n",
    "\n",
    "ins1 = p53[p53['Variant_Type']=='INS']\n",
    "bad_ins_idx = list(ins1[ins1['Reference_Allele']!='-'].index)\n",
    "\n",
    "bad_idxs = bad_del_idx + bad_ins_idx\n",
    "\n",
    "p53_filtered = p53.drop(index=bad_idxs).reset_index().drop(columns=['index', 'Unnamed: 0'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Hugo_Symbol',\n",
       " 'Chromosome',\n",
       " 'Start_Position',\n",
       " 'End_Position',\n",
       " 'Variant_Type',\n",
       " 'Reference_Allele',\n",
       " 'Tumor_Seq_Allele2',\n",
       " 'Strand',\n",
       " 'HGVSp',\n",
       " 'num_occurences',\n",
       " 'codon',\n",
       " 'ref_aa',\n",
       " 'mut_aa',\n",
       " 'Most frequent HGVSc',\n",
       " 'classification',\n",
       " 'transcript_id_TRUE']"
      ]
     },
     "execution_count": 294,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p53_filtered['transcript_id_TRUE'] = 'ENST00000269305.4'\n",
    "p53_filtered\n",
    "\n",
    "#rename the appropriate columns to match up with the formula, then generate...\n",
    "#or just make a different version of the function; probably easier...\n",
    "cols_to_save = list(p53_filtered.keys())\n",
    "cols_to_save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {},
   "outputs": [],
   "source": [
    "def df_formatter(df, cols_to_save, context_size = 60):\n",
    "\n",
    "    \"\"\" \n",
    "    Takes in variants (in cBioPortal format)\n",
    "    and outputs dataframe with WT and ALT oligos with designated context_size\n",
    "    context_size = the amount of nt on either side of the variant e.g. AAA(A/G)AAA = context_size of 3\n",
    "    \"\"\"\n",
    "\n",
    "    wt_w_context = []\n",
    "    alt_w_context = []\n",
    "\n",
    "    seq_start = []\n",
    "    seq_end = []\n",
    "\n",
    "    for i, val in df.iterrows():\n",
    "        vt = val['Variant_Type']\n",
    "        s = val['Start_Position']\n",
    "        e = val['End_Position']\n",
    "        ref = val['Reference_Allele']\n",
    "        alt = val['Tumor_Seq_Allele2']\n",
    "        chrom = val['Chromosome']\n",
    "\n",
    "        if chrom == 'X':\n",
    "            chrom = 22\n",
    "        elif chrom=='Y':\n",
    "            chrom = 23\n",
    "        else:\n",
    "            chrom = int(chrom)-1\n",
    "\n",
    "        chr_seq = records[index_list[chrom]].seq.upper()\n",
    "\n",
    "        if vt in ['SNP', 'ONP', 'DNP']:\n",
    "            ref = ref\n",
    "            alt = alt\n",
    "            #assert ref == chr_seq[s-1:e], print(ref, chr_seq[s-1:e])\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        elif vt =='INS':\n",
    "            ref = ''\n",
    "            alt = alt\n",
    "            #left_context = chr_seq[s-1-context_size:s+1] #need to do this since INS reference alleles are blank\n",
    "            left_context = chr_seq[s-1-context_size:s]\n",
    "            right_context = chr_seq[e-1:e+context_size]\n",
    "\n",
    "        elif vt=='DEL':\n",
    "            ref = ref\n",
    "            alt = ''\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        wt_seq = left_context + ref + right_context\n",
    "        alt_seq = left_context + alt + right_context\n",
    "\n",
    "        wt_w_context.append(str(wt_seq))\n",
    "        alt_w_context.append(str(alt_seq))\n",
    "\n",
    "        start = s-context_size\n",
    "        end = e+context_size\n",
    "\n",
    "        seq_start.append(start)\n",
    "        seq_end.append(end)\n",
    "\n",
    "        assert str(chr_seq[start-1:end])==str(wt_seq), print(chr_seq[start-1:end] + '\\n' + str(wt_seq))\n",
    "                                                            \n",
    "\n",
    "    #cols_to_save = ['COUNT', 'Hugo_Symbol', 'Chromosome', 'Start_Position', 'End_Position', 'Consequence', 'Variant_Classification', 'Variant_Type', 'Reference_Allele', 'Tumor_Seq_Allele2', 'Tumor_Sample_Barcode', 'RefSeq', 'Protein_position', 'Exon_Number', 'HGVSc', 'HGVSp', 'HGVSp_Short', 'transcript_id_TRUE']\n",
    "    df_new = df[cols_to_save]\n",
    "\n",
    "    df_new['seq_start'] = seq_start\n",
    "    df_new['seq_end'] = seq_end\n",
    "    df_new['wt_w_context'] = wt_w_context\n",
    "    df_new['alt_w_context'] = alt_w_context\n",
    "    df_new = df_new.reset_index()\n",
    "    \n",
    "    return df_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {},
   "outputs": [],
   "source": [
    "p53_oligos = df_formatter(p53_filtered, cols_to_save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {},
   "outputs": [],
   "source": [
    "p53_oligos = p53_oligos.rename(columns = {'HGVSp':'HGVSp_Short'}).drop(columns = 'Strand')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {},
   "outputs": [],
   "source": [
    "p53_oligos = frame_determiner(p53_oligos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEL\n",
      "-\n",
      "1\n",
      "p.Asn263IlefsTer82\n",
      "GNLLGRNSFEVRVCACPGRDR\n",
      "GIYWDGTALRCVFVPVLGETG\n"
     ]
    }
   ],
   "source": [
    "#and spot checking\n",
    "i = 1106\n",
    "\n",
    "wt1 = p53_oligos.iloc[i]['CDS_wt_correct_orientation']\n",
    "alt1 = p53_oligos.loc[i]['CDS_alt_correct_orientation']\n",
    "frame = p53_oligos.iloc[i]['Frame']\n",
    "hgvsp = p53_oligos.iloc[i]['HGVSp_Short']\n",
    "s = p53_oligos.iloc[i]['Strand']\n",
    "v = p53_oligos.iloc[i]['Variant_Type']\n",
    "print(v)\n",
    "print(s)\n",
    "print(frame)\n",
    "print(hgvsp)\n",
    "print(Bio.Seq.Seq(wt1[frame:]).transcribe().translate())\n",
    "print(Bio.Seq.Seq(alt1[frame:]).transcribe().translate())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {},
   "outputs": [],
   "source": [
    "p53_oligos['MMR_evasive_amenable']=True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "metadata": {},
   "outputs": [],
   "source": [
    "#p53_oligos.to_csv('p53_library_oligos.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
