{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PEGG 2.0\n",
    "\n",
    "- Updates to PEGG:\n",
    "    - more flexible data input (following PRIDICT/PrimeDesign format)\n",
    "    - more flexible PAM sequence searching\n",
    "    - inclusion of base editing sensor module\n",
    "    - fixing INS/DEL error\n",
    "    - G+19 instead of G+20\n",
    "    - Better fetching of MIT specificity and Rule Set 2/3 information...\n",
    "    - Automated generation of WT/edited sequence generation\n",
    "        - Use this to determine errors in pegRNAs...\n",
    "    - Improved distance to nick and homology overhang sizing\n",
    "    - Also automate finding of silent variants for MMR-evasion...\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import regex as re\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import Bio.Seq\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## improved PAM finder for various PAM sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "substitution_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GGC/TTG)AGAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "ins_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGGGA(/GTAA)GAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "del_eaxmple = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGGGAG(AGAC/)CCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#introduce error messages for each of these\n",
    "\n",
    "start = substitution_example.find(\"(\")\n",
    "\n",
    "end = substitution_example.find(\")\")\n",
    "\n",
    "replace_seq = substitution_example[start:end+1]\n",
    "#if '/' in replace_seq:\n",
    "    #if '+' in replace_seq:\n",
    "        #throw an error\n",
    "\n",
    "loc_replace = replace_seq.find('/')\n",
    "wt_replace = replace_seq[1:loc_replace]\n",
    "mut_replace = replace_seq[loc_replace+1:-1]\n",
    "\n",
    "wt_seq = substitution_example[:start] + wt_replace + substitution_example[end+1:]\n",
    "\n",
    "wt_start = start\n",
    "wt_end = start+len(wt_replace)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "replace_seq = '(/AGGG)'\n",
    "loc_replace = replace_seq.find('/')\n",
    "wt_replace = replace_seq[1:loc_replace]\n",
    "mut_replace = replace_seq[loc_replace+1:-1]\n",
    "wt_replace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "103"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "context_size = 50\n",
    "\n",
    "wt_w_context = wt_seq[wt_start-context_size:wt_start] + wt_seq[wt_start:wt_end] + wt_seq[wt_end:wt_end+context_size]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mut_formatter(pridict_format, context_size = 60):\n",
    "    \"\"\" \n",
    "    Takes in mutations in format of AAA(AA/GC)ATAGC\n",
    "    and converts it into a format that allows pegRNAs to be generated\n",
    "\n",
    "    Parameters\n",
    "    _____\n",
    "    pridict_format = list of sequences in e.g. AAA(AA/GC)ATAGC format\n",
    "    context_size = amount of nt on either side of mutation to select\n",
    "    \"\"\"\n",
    "\n",
    "    original_start = []\n",
    "    original_end = []\n",
    "\n",
    "    wt = []\n",
    "    mut = []\n",
    "    \n",
    "    wt_context = []\n",
    "    mut_context = []\n",
    "\n",
    "    left_context_length = []\n",
    "    right_context_length = []\n",
    "\n",
    "    replace_start = []\n",
    "    replace_end = []\n",
    "\n",
    "    for k in pridict_format:\n",
    "        #find mutation\n",
    "        start = k.find(\"(\")\n",
    "        end = k.find(\")\")\n",
    "\n",
    "        #get out the mutant and WT allele\n",
    "        replace_seq = k[start:end+1]\n",
    "        loc_replace = replace_seq.find('/')\n",
    "        wt_replace = replace_seq[1:loc_replace]\n",
    "        mut_replace = replace_seq[loc_replace+1:-1]\n",
    "\n",
    "        wt.append(wt_replace)\n",
    "        mut.append(mut_replace)\n",
    "\n",
    "        #generate full WT sequence\n",
    "        wt_seq = substitution_example[:start] + wt_replace + substitution_example[end+1:]\n",
    "        wt_start = start\n",
    "        wt_end = start+len(wt_replace)\n",
    "        original_start.append(wt_start)\n",
    "        original_end.append(wt_end)\n",
    "\n",
    "        #and just the subset\n",
    "        left_context = wt_seq[wt_start-context_size:wt_start]\n",
    "        print(left_context)\n",
    "        right_context = wt_seq[wt_end:wt_end+context_size]\n",
    "\n",
    "        wt_w_context = left_context + wt_replace + right_context\n",
    "        mut_w_context = left_context + mut_replace + right_context\n",
    "\n",
    "        wt_context.append(wt_w_context)\n",
    "        mut_context.append(mut_w_context)\n",
    "\n",
    "        left_context_length.append(len(left_context))\n",
    "        right_context_length.append(len(right_context))\n",
    "\n",
    "        replace_start.append(len(left_context))\n",
    "        replace_end.append(len(left_context)+len(wt_replace))\n",
    "\n",
    "    col_labels = ['Original_start', 'Original_end', 'WT', 'Mutant', 'WT_context', 'Mutant_context', 'Replace_start', 'Replace_end', 'Left_context_length', 'Right_context_length']\n",
    "    cols = [original_start, original_end, wt, mut, wt_context, mut_context, replace_start, replace_end, left_context_length, right_context_length]\n",
    "\n",
    "    df = pd.DataFrame(dict(zip(col_labels, cols)))\n",
    "\n",
    "    return df\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG\n",
      "GTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(G\n",
      "TTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GG\n"
     ]
    }
   ],
   "source": [
    "substitution_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GGC/TTG)AGAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "ins_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGGGA(/GTAA)GAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "del_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGGGAG(AGAC/)CCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "\n",
    "pridict_format = [substitution_example, ins_example, del_example]\n",
    "\n",
    "df = mut_formatter(pridict_format)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GGAGACGAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAAC\n",
      "TTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GGGAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAAC\n"
     ]
    }
   ],
   "source": [
    "print(df.iloc[2]['WT_context'])\n",
    "print(df.iloc[2]['Mutant_context'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PAM_finder(WT_seq, mut_start, mut_end,  PAM, RTT_length_max):\n",
    "    \"\"\"Identifies the location of PAM sequences on the + and - strand.\n",
    "    Returns a 2-d array containing marked locations of PAM sequence start locations on + and - strand.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    WT_seq: WT sequence...\n",
    "\n",
    "    PAM_seq: PAM sequence to search for...\n",
    "\n",
    "    RTT_length_max: maximum size of RTT length for searching...\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    distance_PAM_to_nick = 3\n",
    "    PAM_size = len(PAM)\n",
    "    mut_size = mut_start-mut_end\n",
    "\n",
    "    search_size = RTT_length_max - distance_PAM_to_nick\n",
    "\n",
    "    #deal with extra PAM sequences later on...\n",
    "    #and other potential issues\n",
    "    plus_search = WT_seq[mut_end-search_size-5 : mut_start+distance_PAM_to_nick + PAM_size+5]\n",
    "\n",
    "\n",
    "    #---------------Loading in sequences for PAM Searching------------------#\n",
    "    \n",
    "    search_size = RTT_length - len(PAM)-1 #need to modify this for insertion/deletions...\n",
    "    #size mut doesn't capture the size of insertions, only the size of deletions\n",
    "\n",
    "    plus_search = seq1[seq_start-1-search_size : seq_end+search_size].upper()\n",
    "    minus_search = plus_search.complement().upper()\n",
    "    \n",
    "    mut_start_idx = 1+search_size\n",
    "    mut_end_idx = 1+search_size+size_mut #not accurate for insertions; does work for indexing though...\n",
    "\n",
    "    plus_search1 = plus_search[:mut_start_idx+3+len(PAM)-1]\n",
    "    minus_search1 = minus_search[mut_start_idx-3-len(PAM):]\n",
    "\n",
    "    #---------------PAM Searching------------------#\n",
    "\n",
    "    #replacing N with regex symbol\n",
    "    PAM_regex = PAM.replace('N', '/*.')\n",
    "\n",
    "    PAM_search_plus = re.compile('(?=(' + PAM_regex + '))', re.IGNORECASE)\n",
    "\n",
    "    iterator_plus = PAM_search_plus.finditer(str(plus_search1))\n",
    "    PAM_starts_plus = [match.start() for match in iterator_plus]\n",
    "\n",
    "\n",
    "    PAM_minus = PAM[::-1]#reversing it\n",
    "    PAM_regex_minus = PAM_minus.replace('N', '/*.')\n",
    "\n",
    "    PAM_search_minus = re.compile('(?=(' + PAM_regex_minus + '))', re.IGNORECASE)\n",
    "\n",
    "    iterator_minus = PAM_search_minus.finditer(str(minus_search1))\n",
    "    PAM_starts_minus = [match.start() for match in iterator_minus]\n",
    "    #since things are flipped on minus strand, adding len(PAM) to get the true \"start\" to the PAM\n",
    "\n",
    "    PAM_starts_minus = np.asarray(PAM_starts_minus) + len(PAM) + (mut_start_idx-3-len(PAM))#and correct for indexing\n",
    "\n",
    "    return np.asarray([np.asarray(PAM_starts_plus), PAM_starts_minus], dtype='object')-mut_start_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PAM_finder(WT_seq, mut_start, mut_end, PAM, RTT_length_max):\n",
    "    \"\"\"Identifies the location of PAM sequences on the + and - strand.\n",
    "    Returns a 2-d array containing marked locations of PAM sequence start locations on + and - strand.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    WT_seq: WT sequence...\n",
    "\n",
    "    PAM_seq: PAM sequence to search for...\n",
    "\n",
    "    RTT_length_max: maximum size of RTT length for searching...\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    #---------------Loading in sequences for PAM Searching------------------#\n",
    "    \n",
    "    search_size = RTT_length - len(PAM)-1 #need to modify this for insertion/deletions...\n",
    "    #size mut doesn't capture the size of insertions, only the size of deletions\n",
    "\n",
    "    plus_search = seq1[seq_start-1-search_size : seq_end+search_size].upper()\n",
    "    minus_search = plus_search.complement().upper()\n",
    "    \n",
    "    mut_start_idx = 1+search_size\n",
    "    mut_end_idx = 1+search_size+size_mut #not accurate for insertions; does work for indexing though...\n",
    "\n",
    "    plus_search1 = plus_search[:mut_start_idx+3+len(PAM)-1]\n",
    "    minus_search1 = minus_search[mut_start_idx-3-len(PAM):]\n",
    "\n",
    "    #---------------PAM Searching------------------#\n",
    "\n",
    "    #replacing N with regex symbol\n",
    "    PAM_regex = PAM.replace('N', '/*.')\n",
    "\n",
    "    PAM_search_plus = re.compile('(?=(' + PAM_regex + '))', re.IGNORECASE)\n",
    "\n",
    "    iterator_plus = PAM_search_plus.finditer(str(plus_search1))\n",
    "    PAM_starts_plus = [match.start() for match in iterator_plus]\n",
    "\n",
    "\n",
    "    PAM_minus = PAM[::-1]#reversing it\n",
    "    PAM_regex_minus = PAM_minus.replace('N', '/*.')\n",
    "\n",
    "    PAM_search_minus = re.compile('(?=(' + PAM_regex_minus + '))', re.IGNORECASE)\n",
    "\n",
    "    iterator_minus = PAM_search_minus.finditer(str(minus_search1))\n",
    "    PAM_starts_minus = [match.start() for match in iterator_minus]\n",
    "    #since things are flipped on minus strand, adding len(PAM) to get the true \"start\" to the PAM\n",
    "\n",
    "    PAM_starts_minus = np.asarray(PAM_starts_minus) + len(PAM) + (mut_start_idx-3-len(PAM))#and correct for indexing\n",
    "\n",
    "    return np.asarray([np.asarray(PAM_starts_plus), PAM_starts_minus], dtype='object')-mut_start_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PAM = 'NGG'\n",
    "\n",
    "PAM_finder(WT_seq, mut_start, mut_end, PAM, RTT_length_max)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating list of variants for alvin\n",
    "\n",
    "- Starting with diego library\n",
    "    - includes SNPs in IDR regions\n",
    "    - Also intronic SNPs between exons that are an IDR\n",
    "        - These wont be amenable to MMR-evasive silent edits\n",
    "\n",
    "- Need WT sequnece\n",
    "- Mutant sequence\n",
    "- Frame (i.e. codon frame = 0,1,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pegg\n",
    "import gffutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath = '/Users/samgould/Desktop/FSR Lab/reference files/GRCh37/ncbi-genomes-2022-03-17/GCF_000001405.25_GRCh37.p13_genomic.fna.gz'\n",
    "\n",
    "records, index_list = pegg.genome_loader(filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = '/Users/samgould/Desktop/FSR Lab/reference files/gencode_v19.db'\n",
    "db = gffutils.FeatureDB(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['COUNT', 'Hugo_Symbol', 'Entrez_Gene_Id', 'Center', 'NCBI_Build',\n",
       "       'Chromosome', 'Start_Position', 'End_Position', 'Strand', 'Consequence',\n",
       "       'Variant_Classification', 'Variant_Type', 'Reference_Allele',\n",
       "       'Tumor_Seq_Allele1', 'Tumor_Seq_Allele2', 'dbSNP_RS',\n",
       "       'dbSNP_Val_Status', 'Tumor_Sample_Barcode',\n",
       "       'Matched_Norm_Sample_Barcode', 'Match_Norm_Seq_Allele1',\n",
       "       'Match_Norm_Seq_Allele2', 'Tumor_Validation_Allele1',\n",
       "       'Tumor_Validation_Allele2', 'Match_Norm_Validation_Allele1',\n",
       "       'Match_Norm_Validation_Allele2', 'Verification_Status',\n",
       "       'Validation_Status', 'Mutation_Status', 'Sequencing_Phase',\n",
       "       'Sequence_Source', 'Validation_Method', 'Score', 'BAM_File',\n",
       "       'Sequencer', 't_ref_count', 't_alt_count', 'n_ref_count', 'n_alt_count',\n",
       "       'HGVSc', 'HGVSp', 'HGVSp_Short', 'Transcript_ID', 'RefSeq',\n",
       "       'Protein_position', 'Codons', 'Exon_Number', 'gnomAD_AF',\n",
       "       'gnomAD_AFR_AF', 'gnomAD_AMR_AF', 'gnomAD_ASJ_AF', 'gnomAD_EAS_AF',\n",
       "       'gnomAD_FIN_AF', 'gnomAD_NFE_AF', 'gnomAD_OTH_AF', 'gnomAD_SAS_AF',\n",
       "       'FILTER', 'Polyphen_Prediction', 'Polyphen_Score', 'SIFT_Prediction',\n",
       "       'SIFT_Score', 'SWISSPROT', 'n_depth', 't_depth', 'Annotation_Status',\n",
       "       'mutationInCis_Flag'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#diego variants\n",
    "idr = pd.read_csv('filtered_idr_mutations_5count_6nt_indels.csv')\n",
    "idr.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['DEL', 'DNP', 'INS', 'ONP', 'SNP'], dtype=object),\n",
       " array([ 729,   42,  361,   13, 7955]))"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(idr['Variant_Type'], return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "tx_new = []\n",
    "for i, val in idr.iterrows():\n",
    "    h = val['HGVSc']\n",
    "    tx_new.append(h.split(':')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "idr['transcript_id_TRUE'] = tx_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "#check all transcripts are annotated\n",
    "\n",
    "t_ids = np.unique(idr['transcript_id_TRUE'])\n",
    "\n",
    "not_found = []\n",
    "\n",
    "for tx in t_ids:\n",
    "\n",
    "    cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "    start_end_cds = [[i.start, i.end] for i in cds]\n",
    "    if len(start_end_cds)==0:\n",
    "        not_found.append(tx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[20146427, 20146429],\n",
       " [20148634, 20148725],\n",
       " [20150300, 20150381],\n",
       " [20152075, 20152125],\n",
       " [20153856, 20153959],\n",
       " [20156657, 20156740],\n",
       " [20159743, 20159758]]"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tx = 'ENST00000379607.5'\n",
    "\n",
    "cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "start_end_cds = [[i.start, i.end] for i in cds]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'), id='NC_000022.10', name='NC_000022.10', description='NC_000022.10 Homo sapiens chromosome 22, GRCh37.p13 Primary Assembly', dbxrefs=[])"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records[index_list[21]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "metadata": {},
   "outputs": [],
   "source": [
    "ins = idr[idr['Variant_Type']=='INS']\n",
    "dels = idr[idr['Variant_Type']=='DEL']\n",
    "subs = idr[idr['Variant_Type'].isin(['SNP', 'DNP', 'ONP'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 374,
   "metadata": {},
   "outputs": [],
   "source": [
    "def df_formatter(df, context_size = 60):\n",
    "\n",
    "    wt_w_context = []\n",
    "    alt_w_context = []\n",
    "\n",
    "    seq_start = []\n",
    "    seq_end = []\n",
    "\n",
    "    for i, val in df.iterrows():\n",
    "        vt = val['Variant_Type']\n",
    "        s = val['Start_Position']\n",
    "        e = val['End_Position']\n",
    "        ref = val['Reference_Allele']\n",
    "        alt = val['Tumor_Seq_Allele2']\n",
    "        chrom = val['Chromosome']\n",
    "\n",
    "        if chrom == 'X':\n",
    "            chrom = 22\n",
    "        elif chrom=='Y':\n",
    "            chrom = 23\n",
    "        else:\n",
    "            chrom = int(chrom)-1\n",
    "\n",
    "        chr_seq = records[index_list[chrom]].seq.upper()\n",
    "\n",
    "        if vt in ['SNP', 'ONP', 'DNP']:\n",
    "            ref = ref\n",
    "            alt = alt\n",
    "            #assert ref == chr_seq[s-1:e], print(ref, chr_seq[s-1:e])\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        elif vt =='INS':\n",
    "            ref = ''\n",
    "            alt = alt\n",
    "            left_context = chr_seq[s-1-context_size:s+1] #need to do this since INS reference alleles are blank\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        elif vt=='DEL':\n",
    "            ref = ref\n",
    "            alt = ''\n",
    "\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        wt_seq = left_context + ref + right_context\n",
    "        alt_seq = left_context + alt + right_context\n",
    "\n",
    "        wt_w_context.append(str(wt_seq))\n",
    "        alt_w_context.append(str(alt_seq))\n",
    "\n",
    "        start = s-context_size\n",
    "        end = e+context_size\n",
    "\n",
    "        seq_start.append(start)\n",
    "        seq_end.append(end)\n",
    "\n",
    "        assert str(chr_seq[start-1:end])==str(wt_seq)\n",
    "\n",
    "    cols_to_save = ['COUNT', 'Hugo_Symbol', 'Chromosome', 'Start_Position', 'End_Position', 'Consequence', 'Variant_Classification', 'Variant_Type', 'Reference_Allele', 'Tumor_Seq_Allele2', 'Tumor_Sample_Barcode', 'RefSeq', 'Protein_position', 'Exon_Number', 'HGVSc', 'HGVSp', 'HGVSp_Short', 'transcript_id_TRUE']\n",
    "    df_new = df[cols_to_save]\n",
    "\n",
    "    df_new['seq_start'] = seq_start\n",
    "    df_new['seq_end'] = seq_end\n",
    "    df_new['wt_w_context'] = wt_w_context\n",
    "    df_new['alt_w_context'] = alt_w_context\n",
    "    df_new = df_new.reset_index()\n",
    "    \n",
    "    return df_new\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "metadata": {},
   "outputs": [],
   "source": [
    "subs_new = df_formatter(subs[:400])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 379,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1512.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-JHU-00113-00335</td>\n",
       "      <td>NM_001126112.2</td>\n",
       "      <td>282.0</td>\n",
       "      <td>8/11</td>\n",
       "      <td>ENST00000269305.4:c.844C&gt;T</td>\n",
       "      <td>p.Arg282Trp</td>\n",
       "      <td>p.R282W</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577034</td>\n",
       "      <td>7577154</td>\n",
       "      <td>CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...</td>\n",
       "      <td>CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>777.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175639</td>\n",
       "      <td>112175639</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00223-00468</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4348C&gt;T</td>\n",
       "      <td>p.Arg1450Ter</td>\n",
       "      <td>p.R1450*</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175579</td>\n",
       "      <td>112175699</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>614.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112173917</td>\n",
       "      <td>112173917</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00123-00365</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>876.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.2626C&gt;T</td>\n",
       "      <td>p.Arg876Ter</td>\n",
       "      <td>p.R876*</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112173857</td>\n",
       "      <td>112173977</td>\n",
       "      <td>CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...</td>\n",
       "      <td>CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>538.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577085</td>\n",
       "      <td>7577085</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00694-00783</td>\n",
       "      <td>NM_001126112.2</td>\n",
       "      <td>285.0</td>\n",
       "      <td>8/11</td>\n",
       "      <td>ENST00000269305.4:c.853G&gt;A</td>\n",
       "      <td>p.Glu285Lys</td>\n",
       "      <td>p.E285K</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577025</td>\n",
       "      <td>7577145</td>\n",
       "      <td>TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...</td>\n",
       "      <td>TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>527.0</td>\n",
       "      <td>SRSF2</td>\n",
       "      <td>17</td>\n",
       "      <td>74732959</td>\n",
       "      <td>74732959</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-DFCI-001379-11338</td>\n",
       "      <td>NM_001195427.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>1/3</td>\n",
       "      <td>ENST00000359995.5:c.284C&gt;A</td>\n",
       "      <td>p.Pro95His</td>\n",
       "      <td>p.P95H</td>\n",
       "      <td>ENST00000359995.5</td>\n",
       "      <td>74732899</td>\n",
       "      <td>74733019</td>\n",
       "      <td>TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...</td>\n",
       "      <td>TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>603</td>\n",
       "      <td>19.0</td>\n",
       "      <td>ETV5</td>\n",
       "      <td>3</td>\n",
       "      <td>185797801</td>\n",
       "      <td>185797801</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-DFCI-025211-61496</td>\n",
       "      <td>NM_004454.2</td>\n",
       "      <td>152.0</td>\n",
       "      <td>7/13</td>\n",
       "      <td>ENST00000306376.5:c.455C&gt;A</td>\n",
       "      <td>p.Pro152Gln</td>\n",
       "      <td>p.P152Q</td>\n",
       "      <td>ENST00000306376.5</td>\n",
       "      <td>185797741</td>\n",
       "      <td>185797861</td>\n",
       "      <td>ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...</td>\n",
       "      <td>ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>604</td>\n",
       "      <td>19.0</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>1</td>\n",
       "      <td>11190747</td>\n",
       "      <td>11190747</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-DFCI-007654-6083</td>\n",
       "      <td>NM_004958.3</td>\n",
       "      <td>1818.0</td>\n",
       "      <td>39/58</td>\n",
       "      <td>ENST00000361445.4:c.5452C&gt;T</td>\n",
       "      <td>p.Arg1818Cys</td>\n",
       "      <td>p.R1818C</td>\n",
       "      <td>ENST00000361445.4</td>\n",
       "      <td>11190687</td>\n",
       "      <td>11190807</td>\n",
       "      <td>CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...</td>\n",
       "      <td>CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>605</td>\n",
       "      <td>19.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579591</td>\n",
       "      <td>7579591</td>\n",
       "      <td>splice_acceptor_variant</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-DFCI-171465-2394090</td>\n",
       "      <td>NM_001126112.2</td>\n",
       "      <td>33.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000269305.4:c.97-1G&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>p.X33_splice</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7579531</td>\n",
       "      <td>7579651</td>\n",
       "      <td>TTGTTCAATATCGTCCGGGGACAGCATCAAATCATCCATTGCTTGG...</td>\n",
       "      <td>TTGTTCAATATCGTCCGGGGACAGCATCAAATCATCCATTGCTTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>607</td>\n",
       "      <td>19.0</td>\n",
       "      <td>EIF1AX</td>\n",
       "      <td>X</td>\n",
       "      <td>20156720</td>\n",
       "      <td>20156720</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-MSK-P-0044987-T01-IM6</td>\n",
       "      <td>NM_001412.3</td>\n",
       "      <td>13.0</td>\n",
       "      <td>2/7</td>\n",
       "      <td>ENST00000379607.5:c.37C&gt;T</td>\n",
       "      <td>p.Arg13Cys</td>\n",
       "      <td>p.R13C</td>\n",
       "      <td>ENST00000379607.5</td>\n",
       "      <td>20156660</td>\n",
       "      <td>20156780</td>\n",
       "      <td>GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...</td>\n",
       "      <td>GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>608</td>\n",
       "      <td>19.0</td>\n",
       "      <td>PLEKHG5</td>\n",
       "      <td>1</td>\n",
       "      <td>6528337</td>\n",
       "      <td>6528337</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-UHN-394602-ARC2</td>\n",
       "      <td>NM_001042663.1</td>\n",
       "      <td>909.0</td>\n",
       "      <td>21/22</td>\n",
       "      <td>ENST00000400915.3:c.2727A&gt;C</td>\n",
       "      <td>p.Pro909=</td>\n",
       "      <td>p.P909=</td>\n",
       "      <td>ENST00000400915.3</td>\n",
       "      <td>6528277</td>\n",
       "      <td>6528397</td>\n",
       "      <td>CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...</td>\n",
       "      <td>CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>380 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     index   COUNT Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0        1  1512.0        TP53         17         7577094       7577094   \n",
       "1        3   777.0         APC          5       112175639     112175639   \n",
       "2        6   614.0         APC          5       112173917     112173917   \n",
       "3        7   538.0        TP53         17         7577085       7577085   \n",
       "4        8   527.0       SRSF2         17        74732959      74732959   \n",
       "..     ...     ...         ...        ...             ...           ...   \n",
       "395    603    19.0        ETV5          3       185797801     185797801   \n",
       "396    604    19.0        MTOR          1        11190747      11190747   \n",
       "397    605    19.0        TP53         17         7579591       7579591   \n",
       "398    607    19.0      EIF1AX          X        20156720      20156720   \n",
       "399    608    19.0     PLEKHG5          1         6528337       6528337   \n",
       "\n",
       "                 Consequence Variant_Classification Variant_Type  \\\n",
       "0           missense_variant      Missense_Mutation          SNP   \n",
       "1                stop_gained      Nonsense_Mutation          SNP   \n",
       "2                stop_gained      Nonsense_Mutation          SNP   \n",
       "3           missense_variant      Missense_Mutation          SNP   \n",
       "4           missense_variant      Missense_Mutation          SNP   \n",
       "..                       ...                    ...          ...   \n",
       "395         missense_variant      Missense_Mutation          SNP   \n",
       "396         missense_variant      Missense_Mutation          SNP   \n",
       "397  splice_acceptor_variant            Splice_Site          SNP   \n",
       "398         missense_variant      Missense_Mutation          SNP   \n",
       "399       synonymous_variant                 Silent          SNP   \n",
       "\n",
       "    Reference_Allele Tumor_Seq_Allele2         Tumor_Sample_Barcode  \\\n",
       "0                  G                 A        GENIE-JHU-00113-00335   \n",
       "1                  C                 T        GENIE-JHU-00223-00468   \n",
       "2                  C                 T        GENIE-JHU-00123-00365   \n",
       "3                  C                 T        GENIE-JHU-00694-00783   \n",
       "4                  G                 T      GENIE-DFCI-001379-11338   \n",
       "..               ...               ...                          ...   \n",
       "395                G                 T      GENIE-DFCI-025211-61496   \n",
       "396                G                 A       GENIE-DFCI-007654-6083   \n",
       "397                C                 A    GENIE-DFCI-171465-2394090   \n",
       "398                G                 A  GENIE-MSK-P-0044987-T01-IM6   \n",
       "399                T                 G        GENIE-UHN-394602-ARC2   \n",
       "\n",
       "             RefSeq  Protein_position Exon_Number  \\\n",
       "0    NM_001126112.2             282.0        8/11   \n",
       "1       NM_000038.5            1450.0       16/16   \n",
       "2       NM_000038.5             876.0       16/16   \n",
       "3    NM_001126112.2             285.0        8/11   \n",
       "4    NM_001195427.1              95.0         1/3   \n",
       "..              ...               ...         ...   \n",
       "395     NM_004454.2             152.0        7/13   \n",
       "396     NM_004958.3            1818.0       39/58   \n",
       "397  NM_001126112.2              33.0         NaN   \n",
       "398     NM_001412.3              13.0         2/7   \n",
       "399  NM_001042663.1             909.0       21/22   \n",
       "\n",
       "                           HGVSc         HGVSp   HGVSp_Short  \\\n",
       "0     ENST00000269305.4:c.844C>T   p.Arg282Trp       p.R282W   \n",
       "1    ENST00000257430.4:c.4348C>T  p.Arg1450Ter      p.R1450*   \n",
       "2    ENST00000257430.4:c.2626C>T   p.Arg876Ter       p.R876*   \n",
       "3     ENST00000269305.4:c.853G>A   p.Glu285Lys       p.E285K   \n",
       "4     ENST00000359995.5:c.284C>A    p.Pro95His        p.P95H   \n",
       "..                           ...           ...           ...   \n",
       "395   ENST00000306376.5:c.455C>A   p.Pro152Gln       p.P152Q   \n",
       "396  ENST00000361445.4:c.5452C>T  p.Arg1818Cys      p.R1818C   \n",
       "397  ENST00000269305.4:c.97-1G>T           NaN  p.X33_splice   \n",
       "398    ENST00000379607.5:c.37C>T    p.Arg13Cys        p.R13C   \n",
       "399  ENST00000400915.3:c.2727A>C     p.Pro909=       p.P909=   \n",
       "\n",
       "    transcript_id_TRUE  seq_start    seq_end  \\\n",
       "0    ENST00000269305.4    7577034    7577154   \n",
       "1    ENST00000257430.4  112175579  112175699   \n",
       "2    ENST00000257430.4  112173857  112173977   \n",
       "3    ENST00000269305.4    7577025    7577145   \n",
       "4    ENST00000359995.5   74732899   74733019   \n",
       "..                 ...        ...        ...   \n",
       "395  ENST00000306376.5  185797741  185797861   \n",
       "396  ENST00000361445.4   11190687   11190807   \n",
       "397  ENST00000269305.4    7579531    7579651   \n",
       "398  ENST00000379607.5   20156660   20156780   \n",
       "399  ENST00000400915.3    6528277    6528397   \n",
       "\n",
       "                                          wt_w_context  \\\n",
       "0    CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...   \n",
       "1    ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...   \n",
       "2    CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...   \n",
       "3    TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...   \n",
       "4    TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...   \n",
       "..                                                 ...   \n",
       "395  ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...   \n",
       "396  CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...   \n",
       "397  TTGTTCAATATCGTCCGGGGACAGCATCAAATCATCCATTGCTTGG...   \n",
       "398  GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...   \n",
       "399  CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...   \n",
       "\n",
       "                                         alt_w_context  \n",
       "0    CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...  \n",
       "1    ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...  \n",
       "2    CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...  \n",
       "3    TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...  \n",
       "4    TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...  \n",
       "..                                                 ...  \n",
       "395  ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...  \n",
       "396  CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...  \n",
       "397  TTGTTCAATATCGTCCGGGGACAGCATCAAATCATCCATTGCTTGG...  \n",
       "398  GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...  \n",
       "399  CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...  \n",
       "\n",
       "[380 rows x 23 columns]"
      ]
     },
     "execution_count": 379,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_subs = subs_new[subs_new['Variant_Classification']!='Intron']\n",
    "exon_subs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>983.0</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>17</td>\n",
       "      <td>56435161</td>\n",
       "      <td>56435161</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-DFCI-449465-4787362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>659.0</td>\n",
       "      <td>9/10</td>\n",
       "      <td>ENST00000407977.2:c.1976del</td>\n",
       "      <td>p.Gly659ValfsTer41</td>\n",
       "      <td>p.G659Vfs*41</td>\n",
       "      <td>ENST00000407977.2</td>\n",
       "      <td>56435101</td>\n",
       "      <td>56435221</td>\n",
       "      <td>TGGCAAGCTGGGTGCACAGTTGCATCCTGGGGCCGAGAGCCAGGGG...</td>\n",
       "      <td>TGGCAAGCTGGGTGCACAGTTGCATCCTGGGGCCGAGAGCCAGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>461.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175212</td>\n",
       "      <td>112175216</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>AAAAG</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-JHU-00184-00370</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1307.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.3927_3931del</td>\n",
       "      <td>p.Glu1309AspfsTer4</td>\n",
       "      <td>p.E1309Dfs*4</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175152</td>\n",
       "      <td>112175276</td>\n",
       "      <td>AGGATGTAATCAGACGACACAGGAAGCAGATTCTGCTAATACCCTG...</td>\n",
       "      <td>AGGATGTAATCAGACGACACAGGAAGCAGATTCTGCTAATACCCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11</td>\n",
       "      <td>417.0</td>\n",
       "      <td>TCF7L2</td>\n",
       "      <td>10</td>\n",
       "      <td>114925317</td>\n",
       "      <td>114925317</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-DFCI-449465-4787362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>482.0</td>\n",
       "      <td>15/15</td>\n",
       "      <td>ENST00000355995.4:c.1454del</td>\n",
       "      <td>p.Lys485SerfsTer23</td>\n",
       "      <td>p.K485Sfs*23</td>\n",
       "      <td>ENST00000355995.4</td>\n",
       "      <td>114925257</td>\n",
       "      <td>114925377</td>\n",
       "      <td>TCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTCC...</td>\n",
       "      <td>TCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12</td>\n",
       "      <td>357.0</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>1</td>\n",
       "      <td>27105931</td>\n",
       "      <td>27105931</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-DFCI-026352-61610</td>\n",
       "      <td>NM_006015.4</td>\n",
       "      <td>1848.0</td>\n",
       "      <td>20/20</td>\n",
       "      <td>ENST00000324856.7:c.5548del</td>\n",
       "      <td>p.Asp1850ThrfsTer33</td>\n",
       "      <td>p.D1850Tfs*33</td>\n",
       "      <td>ENST00000324856.7</td>\n",
       "      <td>27105871</td>\n",
       "      <td>27105991</td>\n",
       "      <td>TCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTGC...</td>\n",
       "      <td>TCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>19</td>\n",
       "      <td>293.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175678</td>\n",
       "      <td>112175679</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>AG</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-JHU-00090-00282</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1463.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4393_4394del</td>\n",
       "      <td>p.Ser1465TrpfsTer3</td>\n",
       "      <td>p.S1465Wfs*3</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175618</td>\n",
       "      <td>112175739</td>\n",
       "      <td>CCTCAAACAGCTCAAACCAAGCGAGAAGTACCTAAAAATAAAGCAC...</td>\n",
       "      <td>CCTCAAACAGCTCAAACCAAGCGAGAAGTACCTAAAAATAAAGCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>9040</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NOTCH1</td>\n",
       "      <td>9</td>\n",
       "      <td>139390945</td>\n",
       "      <td>139390947</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GTG</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-VICC-927057-unk-3</td>\n",
       "      <td>NM_017617.3</td>\n",
       "      <td>2415.0</td>\n",
       "      <td>34/34</td>\n",
       "      <td>ENST00000277541.6:c.7244_7246del</td>\n",
       "      <td>p.Pro2415del</td>\n",
       "      <td>p.P2415del</td>\n",
       "      <td>ENST00000277541.6</td>\n",
       "      <td>139390885</td>\n",
       "      <td>139391007</td>\n",
       "      <td>CACTCAGGAAGCTCCGGCCCAGGTGGCCGCTGGCTGCTGAGCTCAC...</td>\n",
       "      <td>CACTCAGGAAGCTCCGGCCCAGGTGGCCGCTGGCTGCTGAGCTCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>725</th>\n",
       "      <td>9080</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NUTM1</td>\n",
       "      <td>15</td>\n",
       "      <td>34648285</td>\n",
       "      <td>34648286</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>AG</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-UCSF-13974-8069T</td>\n",
       "      <td>NM_175741.1</td>\n",
       "      <td>664.0</td>\n",
       "      <td>7/7</td>\n",
       "      <td>ENST00000333756.4:c.1992_1993del</td>\n",
       "      <td>p.Gly666SerfsTer26</td>\n",
       "      <td>p.G666Sfs*26</td>\n",
       "      <td>ENST00000333756.4</td>\n",
       "      <td>34648225</td>\n",
       "      <td>34648346</td>\n",
       "      <td>TCCTCTGCAAGGACAAGGGTTAGAAAAGCAAGTCCTGGGATTGCAG...</td>\n",
       "      <td>TCCTCTGCAAGGACAAGGGTTAGAAAAGCAAGTCCTGGGATTGCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>726</th>\n",
       "      <td>9087</td>\n",
       "      <td>5.0</td>\n",
       "      <td>SIK2</td>\n",
       "      <td>11</td>\n",
       "      <td>111594604</td>\n",
       "      <td>111594606</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>CTC</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-CRUK-MTST0169-primary</td>\n",
       "      <td>NM_015191.1</td>\n",
       "      <td>844.0</td>\n",
       "      <td>15/15</td>\n",
       "      <td>ENST00000304987.3:c.2534_2536del</td>\n",
       "      <td>p.Ser845del</td>\n",
       "      <td>p.S845del</td>\n",
       "      <td>ENST00000304987.3</td>\n",
       "      <td>111594544</td>\n",
       "      <td>111594666</td>\n",
       "      <td>ACCGCCACCACCCCCTCCACCACCACGACAGCCAGGAGCTGCCCCA...</td>\n",
       "      <td>ACCGCCACCACCCCCTCCACCACCACGACAGCCAGGAGCTGCCCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>727</th>\n",
       "      <td>9089</td>\n",
       "      <td>5.0</td>\n",
       "      <td>CSF1R</td>\n",
       "      <td>5</td>\n",
       "      <td>149433716</td>\n",
       "      <td>149433718</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>CTC</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-SCI-0541906662-U3294645R2</td>\n",
       "      <td>NM_005211.3</td>\n",
       "      <td>945.0</td>\n",
       "      <td>22/22</td>\n",
       "      <td>ENST00000286301.3:c.2833_2835del</td>\n",
       "      <td>p.Glu945del</td>\n",
       "      <td>p.E945del</td>\n",
       "      <td>ENST00000286301.3</td>\n",
       "      <td>149433656</td>\n",
       "      <td>149433778</td>\n",
       "      <td>CTGCAGCAAGGGCTGGGCGATATCCCCTTGCTCGCAGCAGGTCAGG...</td>\n",
       "      <td>CTGCAGCAAGGGCTGGGCGATATCCCCTTGCTCGCAGCAGGTCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>728</th>\n",
       "      <td>9090</td>\n",
       "      <td>5.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112174608</td>\n",
       "      <td>112174608</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-COLU-00476-3</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1106.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.3317del</td>\n",
       "      <td>p.Gly1106GlufsTer20</td>\n",
       "      <td>p.G1106Efs*20</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112174548</td>\n",
       "      <td>112174668</td>\n",
       "      <td>ACCTCAAGTTCCAACCACATTTTGGACAGCAGGAATGTGTTTCTCC...</td>\n",
       "      <td>ACCTCAAGTTCCAACCACATTTTGGACAGCAGGAATGTGTTTCTCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>690 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     index  COUNT Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "1        2  983.0       RNF43         17        56435161      56435161   \n",
       "2        9  461.0         APC          5       112175212     112175216   \n",
       "3       11  417.0      TCF7L2         10       114925317     114925317   \n",
       "4       12  357.0      ARID1A          1        27105931      27105931   \n",
       "5       19  293.0         APC          5       112175678     112175679   \n",
       "..     ...    ...         ...        ...             ...           ...   \n",
       "724   9040    5.0      NOTCH1          9       139390945     139390947   \n",
       "725   9080    5.0       NUTM1         15        34648285      34648286   \n",
       "726   9087    5.0        SIK2         11       111594604     111594606   \n",
       "727   9089    5.0       CSF1R          5       149433716     149433718   \n",
       "728   9090    5.0         APC          5       112174608     112174608   \n",
       "\n",
       "            Consequence Variant_Classification Variant_Type Reference_Allele  \\\n",
       "1    frameshift_variant        Frame_Shift_Del          DEL                C   \n",
       "2    frameshift_variant        Frame_Shift_Del          DEL            AAAAG   \n",
       "3    frameshift_variant        Frame_Shift_Del          DEL                A   \n",
       "4    frameshift_variant        Frame_Shift_Del          DEL                G   \n",
       "5    frameshift_variant        Frame_Shift_Del          DEL               AG   \n",
       "..                  ...                    ...          ...              ...   \n",
       "724    inframe_deletion           In_Frame_Del          DEL              GTG   \n",
       "725  frameshift_variant        Frame_Shift_Del          DEL               AG   \n",
       "726    inframe_deletion           In_Frame_Del          DEL              CTC   \n",
       "727    inframe_deletion           In_Frame_Del          DEL              CTC   \n",
       "728  frameshift_variant        Frame_Shift_Del          DEL                G   \n",
       "\n",
       "    Tumor_Seq_Allele2             Tumor_Sample_Barcode       RefSeq  \\\n",
       "1                   -        GENIE-DFCI-449465-4787362          NaN   \n",
       "2                   -            GENIE-JHU-00184-00370  NM_000038.5   \n",
       "3                   -        GENIE-DFCI-449465-4787362          NaN   \n",
       "4                   -          GENIE-DFCI-026352-61610  NM_006015.4   \n",
       "5                   -            GENIE-JHU-00090-00282  NM_000038.5   \n",
       "..                ...                              ...          ...   \n",
       "724                 -          GENIE-VICC-927057-unk-3  NM_017617.3   \n",
       "725                 -           GENIE-UCSF-13974-8069T  NM_175741.1   \n",
       "726                 -      GENIE-CRUK-MTST0169-primary  NM_015191.1   \n",
       "727                 -  GENIE-SCI-0541906662-U3294645R2  NM_005211.3   \n",
       "728                 -               GENIE-COLU-00476-3  NM_000038.5   \n",
       "\n",
       "     Protein_position Exon_Number                             HGVSc  \\\n",
       "1               659.0        9/10       ENST00000407977.2:c.1976del   \n",
       "2              1307.0       16/16  ENST00000257430.4:c.3927_3931del   \n",
       "3               482.0       15/15       ENST00000355995.4:c.1454del   \n",
       "4              1848.0       20/20       ENST00000324856.7:c.5548del   \n",
       "5              1463.0       16/16  ENST00000257430.4:c.4393_4394del   \n",
       "..                ...         ...                               ...   \n",
       "724            2415.0       34/34  ENST00000277541.6:c.7244_7246del   \n",
       "725             664.0         7/7  ENST00000333756.4:c.1992_1993del   \n",
       "726             844.0       15/15  ENST00000304987.3:c.2534_2536del   \n",
       "727             945.0       22/22  ENST00000286301.3:c.2833_2835del   \n",
       "728            1106.0       16/16       ENST00000257430.4:c.3317del   \n",
       "\n",
       "                   HGVSp    HGVSp_Short transcript_id_TRUE  seq_start  \\\n",
       "1     p.Gly659ValfsTer41   p.G659Vfs*41  ENST00000407977.2   56435101   \n",
       "2     p.Glu1309AspfsTer4   p.E1309Dfs*4  ENST00000257430.4  112175152   \n",
       "3     p.Lys485SerfsTer23   p.K485Sfs*23  ENST00000355995.4  114925257   \n",
       "4    p.Asp1850ThrfsTer33  p.D1850Tfs*33  ENST00000324856.7   27105871   \n",
       "5     p.Ser1465TrpfsTer3   p.S1465Wfs*3  ENST00000257430.4  112175618   \n",
       "..                   ...            ...                ...        ...   \n",
       "724         p.Pro2415del     p.P2415del  ENST00000277541.6  139390885   \n",
       "725   p.Gly666SerfsTer26   p.G666Sfs*26  ENST00000333756.4   34648225   \n",
       "726          p.Ser845del      p.S845del  ENST00000304987.3  111594544   \n",
       "727          p.Glu945del      p.E945del  ENST00000286301.3  149433656   \n",
       "728  p.Gly1106GlufsTer20  p.G1106Efs*20  ENST00000257430.4  112174548   \n",
       "\n",
       "       seq_end                                       wt_w_context  \\\n",
       "1     56435221  TGGCAAGCTGGGTGCACAGTTGCATCCTGGGGCCGAGAGCCAGGGG...   \n",
       "2    112175276  AGGATGTAATCAGACGACACAGGAAGCAGATTCTGCTAATACCCTG...   \n",
       "3    114925377  TCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTCC...   \n",
       "4     27105991  TCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTGC...   \n",
       "5    112175739  CCTCAAACAGCTCAAACCAAGCGAGAAGTACCTAAAAATAAAGCAC...   \n",
       "..         ...                                                ...   \n",
       "724  139391007  CACTCAGGAAGCTCCGGCCCAGGTGGCCGCTGGCTGCTGAGCTCAC...   \n",
       "725   34648346  TCCTCTGCAAGGACAAGGGTTAGAAAAGCAAGTCCTGGGATTGCAG...   \n",
       "726  111594666  ACCGCCACCACCCCCTCCACCACCACGACAGCCAGGAGCTGCCCCA...   \n",
       "727  149433778  CTGCAGCAAGGGCTGGGCGATATCCCCTTGCTCGCAGCAGGTCAGG...   \n",
       "728  112174668  ACCTCAAGTTCCAACCACATTTTGGACAGCAGGAATGTGTTTCTCC...   \n",
       "\n",
       "                                         alt_w_context  \n",
       "1    TGGCAAGCTGGGTGCACAGTTGCATCCTGGGGCCGAGAGCCAGGGG...  \n",
       "2    AGGATGTAATCAGACGACACAGGAAGCAGATTCTGCTAATACCCTG...  \n",
       "3    TCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTCC...  \n",
       "4    TCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTGC...  \n",
       "5    CCTCAAACAGCTCAAACCAAGCGAGAAGTACCTAAAAATAAAGCAC...  \n",
       "..                                                 ...  \n",
       "724  CACTCAGGAAGCTCCGGCCCAGGTGGCCGCTGGCTGCTGAGCTCAC...  \n",
       "725  TCCTCTGCAAGGACAAGGGTTAGAAAAGCAAGTCCTGGGATTGCAG...  \n",
       "726  ACCGCCACCACCCCCTCCACCACCACGACAGCCAGGAGCTGCCCCA...  \n",
       "727  CTGCAGCAAGGGCTGGGCGATATCCCCTTGCTCGCAGCAGGTCAGG...  \n",
       "728  ACCTCAAGTTCCAACCACATTTTGGACAGCAGGAATGTGTTTCTCC...  \n",
       "\n",
       "[690 rows x 23 columns]"
      ]
     },
     "execution_count": 369,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_dels = dels_new[dels_new['Variant_Classification']!='Intron']\n",
    "exon_dels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>762.0</td>\n",
       "      <td>ASXL1</td>\n",
       "      <td>20</td>\n",
       "      <td>31022441</td>\n",
       "      <td>31022442</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-003409-1958</td>\n",
       "      <td>NM_015338.5</td>\n",
       "      <td>642.0</td>\n",
       "      <td>13/13</td>\n",
       "      <td>ENST00000375687.4:c.1934dup</td>\n",
       "      <td>p.Gly646TrpfsTer12</td>\n",
       "      <td>p.G646Wfs*12</td>\n",
       "      <td>ENST00000375687.4</td>\n",
       "      <td>31022381</td>\n",
       "      <td>31022502</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>645.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175952</td>\n",
       "      <td>112175953</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-JHU-00198-00378</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1554.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4666dup</td>\n",
       "      <td>p.Thr1556AsnfsTer3</td>\n",
       "      <td>p.T1556Nfs*3</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175892</td>\n",
       "      <td>112176013</td>\n",
       "      <td>GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...</td>\n",
       "      <td>GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>280.0</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>1</td>\n",
       "      <td>27105930</td>\n",
       "      <td>27105931</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-001750-9089</td>\n",
       "      <td>NM_006015.4</td>\n",
       "      <td>1847.0</td>\n",
       "      <td>20/20</td>\n",
       "      <td>ENST00000324856.7:c.5548dup</td>\n",
       "      <td>p.Asp1850GlyfsTer4</td>\n",
       "      <td>p.D1850Gfs*4</td>\n",
       "      <td>ENST00000324856.7</td>\n",
       "      <td>27105870</td>\n",
       "      <td>27105991</td>\n",
       "      <td>CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...</td>\n",
       "      <td>CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>55</td>\n",
       "      <td>108.0</td>\n",
       "      <td>TCF7L2</td>\n",
       "      <td>10</td>\n",
       "      <td>114925316</td>\n",
       "      <td>114925317</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-DFCI-151674-1525793</td>\n",
       "      <td>NaN</td>\n",
       "      <td>482.0</td>\n",
       "      <td>15/15</td>\n",
       "      <td>ENST00000355995.4:c.1454dup</td>\n",
       "      <td>p.Cys486ValfsTer8</td>\n",
       "      <td>p.C486Vfs*8</td>\n",
       "      <td>ENST00000355995.4</td>\n",
       "      <td>114925256</td>\n",
       "      <td>114925377</td>\n",
       "      <td>CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...</td>\n",
       "      <td>CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>60</td>\n",
       "      <td>99.0</td>\n",
       "      <td>GATA3</td>\n",
       "      <td>10</td>\n",
       "      <td>8115874</td>\n",
       "      <td>8115875</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-002328-1774</td>\n",
       "      <td>NaN</td>\n",
       "      <td>407.0</td>\n",
       "      <td>6/6</td>\n",
       "      <td>ENST00000346208.3:c.1221dup</td>\n",
       "      <td>p.Pro408AlafsTer99</td>\n",
       "      <td>p.P408Afs*99</td>\n",
       "      <td>ENST00000346208.3</td>\n",
       "      <td>8115814</td>\n",
       "      <td>8115935</td>\n",
       "      <td>AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...</td>\n",
       "      <td>AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>8994</td>\n",
       "      <td>5.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175621</td>\n",
       "      <td>112175622</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-MSK-P-0048781-T01-IM6</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1444.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4333dup</td>\n",
       "      <td>p.Thr1445AsnfsTer10</td>\n",
       "      <td>p.T1445Nfs*10</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175561</td>\n",
       "      <td>112175682</td>\n",
       "      <td>CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...</td>\n",
       "      <td>CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>9010</td>\n",
       "      <td>5.0</td>\n",
       "      <td>SRSF2</td>\n",
       "      <td>17</td>\n",
       "      <td>74732960</td>\n",
       "      <td>74732961</td>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>GGC</td>\n",
       "      <td>GENIE-UHN-AGI147550-BM1</td>\n",
       "      <td>NM_001195427.1</td>\n",
       "      <td>94.0</td>\n",
       "      <td>1/3</td>\n",
       "      <td>ENST00000359995.5:c.282_283insGCC</td>\n",
       "      <td>p.Arg94_Pro95insAla</td>\n",
       "      <td>p.R94_P95insA</td>\n",
       "      <td>ENST00000359995.5</td>\n",
       "      <td>74732900</td>\n",
       "      <td>74733021</td>\n",
       "      <td>AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...</td>\n",
       "      <td>AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>9011</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>X</td>\n",
       "      <td>15841255</td>\n",
       "      <td>15841256</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>AGCCGC</td>\n",
       "      <td>GENIE-UHN-AGI974552-BM1</td>\n",
       "      <td>NM_005089.3</td>\n",
       "      <td>447.0</td>\n",
       "      <td>11/11</td>\n",
       "      <td>ENST00000307771.7:c.1339_1340insAGCCGC</td>\n",
       "      <td>p.Ser447delinsLysProArg</td>\n",
       "      <td>p.S447delinsKPR</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15841195</td>\n",
       "      <td>15841316</td>\n",
       "      <td>AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...</td>\n",
       "      <td>AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>9013</td>\n",
       "      <td>5.0</td>\n",
       "      <td>TSC1</td>\n",
       "      <td>9</td>\n",
       "      <td>135771988</td>\n",
       "      <td>135771989</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>GCT</td>\n",
       "      <td>GENIE-UHN-DIVA273992-ARC1</td>\n",
       "      <td>NM_001162426.1</td>\n",
       "      <td>1043.0</td>\n",
       "      <td>23/23</td>\n",
       "      <td>ENST00000298552.3:c.3128_3129insAGC</td>\n",
       "      <td>p.Ser1043delinsArgAla</td>\n",
       "      <td>p.S1043delinsRA</td>\n",
       "      <td>ENST00000298552.3</td>\n",
       "      <td>135771928</td>\n",
       "      <td>135772049</td>\n",
       "      <td>CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...</td>\n",
       "      <td>CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>9094</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ZFHX3</td>\n",
       "      <td>16</td>\n",
       "      <td>72821633</td>\n",
       "      <td>72821634</td>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>GCCACT</td>\n",
       "      <td>GENIE-PROV-2219952519-76fe5c70a7</td>\n",
       "      <td>NM_006885.3</td>\n",
       "      <td>3514.0</td>\n",
       "      <td>10/10</td>\n",
       "      <td>ENST00000268489.5:c.10541_10542insAGTGGC</td>\n",
       "      <td>p.Gly3514_Gly3515insValAla</td>\n",
       "      <td>p.G3514_G3515insVA</td>\n",
       "      <td>ENST00000268489.5</td>\n",
       "      <td>72821573</td>\n",
       "      <td>72821694</td>\n",
       "      <td>GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...</td>\n",
       "      <td>GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>347 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     index  COUNT Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0        4  762.0       ASXL1         20        31022441      31022442   \n",
       "1        5  645.0         APC          5       112175952     112175953   \n",
       "2       21  280.0      ARID1A          1        27105930      27105931   \n",
       "4       55  108.0      TCF7L2         10       114925316     114925317   \n",
       "5       60   99.0       GATA3         10         8115874       8115875   \n",
       "..     ...    ...         ...        ...             ...           ...   \n",
       "354   8994    5.0         APC          5       112175621     112175622   \n",
       "355   9010    5.0       SRSF2         17        74732960      74732961   \n",
       "356   9011    5.0       ZRSR2          X        15841255      15841256   \n",
       "357   9013    5.0        TSC1          9       135771988     135771989   \n",
       "360   9094    5.0       ZFHX3         16        72821633      72821634   \n",
       "\n",
       "                  Consequence Variant_Classification Variant_Type  \\\n",
       "0          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "1          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "2          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "4          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "5          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "..                        ...                    ...          ...   \n",
       "354        frameshift_variant        Frame_Shift_Ins          INS   \n",
       "355         inframe_insertion           In_Frame_Ins          INS   \n",
       "356  protein_altering_variant           In_Frame_Ins          INS   \n",
       "357  protein_altering_variant           In_Frame_Ins          INS   \n",
       "360         inframe_insertion           In_Frame_Ins          INS   \n",
       "\n",
       "    Reference_Allele Tumor_Seq_Allele2              Tumor_Sample_Barcode  \\\n",
       "0                  -                 G            GENIE-DFCI-003409-1958   \n",
       "1                  -                 A             GENIE-JHU-00198-00378   \n",
       "2                  -                 G            GENIE-DFCI-001750-9089   \n",
       "4                  -                 A         GENIE-DFCI-151674-1525793   \n",
       "5                  -                 G            GENIE-DFCI-002328-1774   \n",
       "..               ...               ...                               ...   \n",
       "354                -                 A       GENIE-MSK-P-0048781-T01-IM6   \n",
       "355                -               GGC           GENIE-UHN-AGI147550-BM1   \n",
       "356                -            AGCCGC           GENIE-UHN-AGI974552-BM1   \n",
       "357                -               GCT         GENIE-UHN-DIVA273992-ARC1   \n",
       "360                -            GCCACT  GENIE-PROV-2219952519-76fe5c70a7   \n",
       "\n",
       "             RefSeq  Protein_position Exon_Number  \\\n",
       "0       NM_015338.5             642.0       13/13   \n",
       "1       NM_000038.5            1554.0       16/16   \n",
       "2       NM_006015.4            1847.0       20/20   \n",
       "4               NaN             482.0       15/15   \n",
       "5               NaN             407.0         6/6   \n",
       "..              ...               ...         ...   \n",
       "354     NM_000038.5            1444.0       16/16   \n",
       "355  NM_001195427.1              94.0         1/3   \n",
       "356     NM_005089.3             447.0       11/11   \n",
       "357  NM_001162426.1            1043.0       23/23   \n",
       "360     NM_006885.3            3514.0       10/10   \n",
       "\n",
       "                                        HGVSc                       HGVSp  \\\n",
       "0                 ENST00000375687.4:c.1934dup          p.Gly646TrpfsTer12   \n",
       "1                 ENST00000257430.4:c.4666dup          p.Thr1556AsnfsTer3   \n",
       "2                 ENST00000324856.7:c.5548dup          p.Asp1850GlyfsTer4   \n",
       "4                 ENST00000355995.4:c.1454dup           p.Cys486ValfsTer8   \n",
       "5                 ENST00000346208.3:c.1221dup          p.Pro408AlafsTer99   \n",
       "..                                        ...                         ...   \n",
       "354               ENST00000257430.4:c.4333dup         p.Thr1445AsnfsTer10   \n",
       "355         ENST00000359995.5:c.282_283insGCC         p.Arg94_Pro95insAla   \n",
       "356    ENST00000307771.7:c.1339_1340insAGCCGC     p.Ser447delinsLysProArg   \n",
       "357       ENST00000298552.3:c.3128_3129insAGC       p.Ser1043delinsArgAla   \n",
       "360  ENST00000268489.5:c.10541_10542insAGTGGC  p.Gly3514_Gly3515insValAla   \n",
       "\n",
       "            HGVSp_Short transcript_id_TRUE  seq_start    seq_end  \\\n",
       "0          p.G646Wfs*12  ENST00000375687.4   31022381   31022502   \n",
       "1          p.T1556Nfs*3  ENST00000257430.4  112175892  112176013   \n",
       "2          p.D1850Gfs*4  ENST00000324856.7   27105870   27105991   \n",
       "4           p.C486Vfs*8  ENST00000355995.4  114925256  114925377   \n",
       "5          p.P408Afs*99  ENST00000346208.3    8115814    8115935   \n",
       "..                  ...                ...        ...        ...   \n",
       "354       p.T1445Nfs*10  ENST00000257430.4  112175561  112175682   \n",
       "355       p.R94_P95insA  ENST00000359995.5   74732900   74733021   \n",
       "356     p.S447delinsKPR  ENST00000307771.7   15841195   15841316   \n",
       "357     p.S1043delinsRA  ENST00000298552.3  135771928  135772049   \n",
       "360  p.G3514_G3515insVA  ENST00000268489.5   72821573   72821694   \n",
       "\n",
       "                                          wt_w_context  \\\n",
       "0    GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...   \n",
       "1    GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...   \n",
       "2    CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...   \n",
       "4    CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...   \n",
       "5    AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...   \n",
       "..                                                 ...   \n",
       "354  CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...   \n",
       "355  AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...   \n",
       "356  AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...   \n",
       "357  CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...   \n",
       "360  GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...   \n",
       "\n",
       "                                         alt_w_context  \n",
       "0    GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...  \n",
       "1    GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...  \n",
       "2    CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...  \n",
       "4    CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...  \n",
       "5    AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...  \n",
       "..                                                 ...  \n",
       "354  CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...  \n",
       "355  AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...  \n",
       "356  AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...  \n",
       "357  CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...  \n",
       "360  GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...  \n",
       "\n",
       "[347 rows x 23 columns]"
      ]
     },
     "execution_count": 336,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_ins = ins_new[ins_new['Variant_Classification']!='Intron']\n",
    "exon_ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 381,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R282W R282W\n",
      "R1450* R1450*\n",
      "R876* R876*\n",
      "E285K E285K\n",
      "P95H P95H\n",
      "S37F S37F\n",
      "6\n",
      "N1018S N1018_K1019delinsSE\n",
      "S45F S45F\n",
      "P95L P95L\n",
      "R1114* R1114*\n",
      "T41A T41A\n",
      "E286K E286K\n",
      "S37C S37C\n",
      "Q1367* Q1367*\n",
      "S33C S33C\n",
      "S45P S45P\n",
      "P95R P95R\n",
      "20\n",
      "WT X279_splice\n",
      "Q1429* Q1429*\n",
      "R805* R805*\n",
      "S33F S33F\n",
      "Q1378* Q1378*\n",
      "26\n",
      "WT X423_splice\n",
      "D32Y D32Y\n",
      "T41I T41I\n",
      "E1306* E1306*\n",
      "Q1406* Q1406*\n",
      "S257L S257L\n",
      "R693* R693*\n",
      "D32N D32N\n",
      "E1309* E1309*\n",
      "E1353* E1353*\n",
      "G34V G34V\n",
      "Q58L Q58L\n",
      "Q1338* Q1338*\n",
      "R283P R283P\n",
      "P44L P44L\n",
      "E1379* E1379*\n",
      "E79Q E79Q\n",
      "D281H D281H\n",
      "G34E G34E\n",
      "46\n",
      "WT X307_splice\n",
      "Q1291* Q1291*\n",
      "E1322* E1322*\n",
      "E1397* E1397*\n",
      "50\n",
      "WT X307_splice\n",
      "D281N D281N\n",
      "R282G R282G\n",
      "G34R G34R\n",
      "L78P L78P\n",
      "55\n",
      "WT X307_splice\n",
      "Q1303* Q1303*\n",
      "Q1294* Q1294*\n",
      "Y935* Y935*\n",
      "S24F S24F\n",
      "D1010H D1010H\n",
      "61\n",
      "R299R R299=\n",
      "R488C R488C\n",
      "R787* R787*\n",
      "D1010N D1010N\n",
      "E1286* E1286*\n",
      "S33P S33P\n",
      "D281E D281E\n",
      "D281Y D281Y\n",
      "70\n",
      "E914D E914_P915delinsDS\n",
      "E79K E79K\n",
      "D32G D32G\n",
      "73\n",
      "WT X307_splice\n",
      "R2204* R2204*\n",
      "E1408* E1408*\n",
      "D281G D281G\n",
      "78\n",
      "WT X307_splice\n",
      "79\n",
      "WT X33_splice\n",
      "R177Q R177Q\n",
      "E941* E941*\n",
      "R2237* R2237*\n",
      "S1400* S1400*\n",
      "R2400* R2400*\n",
      "G44D G44D\n",
      "S1356* S1356*\n",
      "E82D E82D\n",
      "R358* R358*\n",
      "S37Y S37Y\n",
      "S33Y S33Y\n",
      "G44S G44S\n",
      "R174Q R174Q\n",
      "A59V A59V\n",
      "94\n",
      "Q1054H Q1054_P1055delinsHS\n",
      "R293* R293*\n",
      "D281E D281E\n",
      "E1295* E1295*\n",
      "I692V I692V\n",
      "E1345* E1345*\n",
      "100\n",
      "WT X1010_splice\n",
      "D23Y D23Y\n",
      "Q1328* Q1328*\n",
      "R497* R497*\n",
      "Q276* Q276*\n",
      "R353* R353*\n",
      "S37A S37A\n",
      "E1374* E1374*\n",
      "S1346* S1346*\n",
      "R371* R371*\n",
      "D32V D32V\n",
      "R290H R290H\n",
      "E304K E304K\n",
      "D1010Y D1010Y\n",
      "R179H R179H\n",
      "116\n",
      "WT X32_splice\n",
      "S367L S367L\n",
      "P287S P287S\n",
      "E286G E286G\n",
      "P47A P47A\n",
      "R420W R420W\n",
      "S33A S33A\n",
      "G34E G34E\n",
      "125\n",
      "WT X1010_splice\n",
      "R13H R13H\n",
      "G1322E G1322E\n",
      "R283C R283C\n",
      "N159S N159S\n",
      "L1091A L1091A\n",
      "G28D G28D\n",
      "P2146L P2146L\n",
      "G579R G579R\n",
      "134\n",
      "WT X11_splice\n",
      "135\n",
      "WT X1010_splice\n",
      "R282P R282P\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/5g/xxjcy2g531n02hnyb6d8qdsr0000gn/T/ipykernel_33329/3125738295.py\u001b[0m in \u001b[0;36m<cell line: 4>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     50\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     51\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 52\u001b[0;31m     \u001b[0minc_start\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     53\u001b[0m     \u001b[0minc_end\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     54\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "#try it with one example\n",
    "#val = ins_new.iloc[0]\n",
    "\n",
    "for index1, val in exon_subs.iterrows():\n",
    "    hg = val['HGVSp_Short']\n",
    "\n",
    "    tx = val['transcript_id_TRUE']\n",
    "    cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "    start_end_cds = [[i.start, i.end] for i in cds]\n",
    "    strand = db[tx].strand\n",
    "    chrom = val['Chromosome']\n",
    "\n",
    "    if chrom == 'X':\n",
    "        chrom = 22\n",
    "    elif chrom=='Y':\n",
    "        chrom = 23\n",
    "    else:\n",
    "        chrom = int(chrom)-1\n",
    "\n",
    "    chr_seq = records[index_list[chrom]].seq.upper()\n",
    "\n",
    "\n",
    "    wt_dna = ''\n",
    "    codon_locs = []\n",
    "    for i in start_end_cds:\n",
    "        wt_dna += chr_seq[i[0]-1:i[1]]\n",
    "\n",
    "        for j in range(i[0], i[1]+1):\n",
    "            codon_locs.append(j)\n",
    "\n",
    "    #and add in the last/stop codon\n",
    "    #if strand == '+':\n",
    "\n",
    "    # wt_dna += chr_seq[start_end_cds[-1][1]:start_end_cds[-1][1]+3]\n",
    "    #elif strand == '-':\n",
    "    #   wt_dna = chr_seq[start_end_cds[0][0]-3:start_end_cds[0][0]] +  wt_dna\n",
    "\n",
    "    #check location in cds\n",
    "    #get information about the wt/alt sequence\n",
    "    start = val['seq_start']\n",
    "    end = val['seq_end']\n",
    "    wt_seq = val['wt_w_context']\n",
    "    alt_seq = val['alt_w_context']\n",
    "\n",
    "    inc = []\n",
    "    dna_dict = dict(zip(list(range(start, end+1)), list(range(len(wt_seq)))))\n",
    "    for i in range(start, end+1):\n",
    "        if i in codon_locs:\n",
    "            inc.append(i)\n",
    "\n",
    "\n",
    "    inc_start = inc[0]\n",
    "    inc_end = inc[-1]\n",
    "\n",
    "    wt_start = dna_dict[inc_start]\n",
    "    wt_end = dna_dict[inc_end] + 1\n",
    "\n",
    "    if wt_end == len(wt_seq):\n",
    "        alt_end = len(alt_seq)\n",
    "    else:\n",
    "        alt_end = wt_end - len(wt_seq)\n",
    "\n",
    "    first_half = codon_locs.index(inc_start)\n",
    "    second_half = codon_locs.index(inc_end)\n",
    "\n",
    "\n",
    "    wt_full_dna = wt_dna[:first_half] + wt_seq[wt_start:wt_end] + wt_dna[second_half+1:]\n",
    "    alt_full_dna = wt_dna[:first_half] + alt_seq[wt_start:alt_end] + wt_dna[second_half+1:]\n",
    "\n",
    "    assert wt_full_dna == wt_dna\n",
    "\n",
    "    if strand == '-':\n",
    "        wt_full_dna = wt_full_dna.reverse_complement()\n",
    "        alt_full_dna = alt_full_dna.reverse_complement()\n",
    "\n",
    "    \n",
    "    wt_dna = str(wt_full_dna)\n",
    "    mut_dna = str(alt_full_dna)\n",
    "    wt_seq = Bio.Seq.Seq(wt_dna).transcribe().translate()\n",
    "    mut_seq = Bio.Seq.Seq(alt_full_dna).transcribe().translate()\n",
    "\n",
    "    hgvsp = get_hgvsp_classification(wt_dna, mut_dna, wt_seq, mut_seq)\n",
    "\n",
    "    if hgvsp != hg[2:]:\n",
    "        print(index1)\n",
    "    print(hgvsp, hg[2:])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CAADEFGH'"
      ]
     },
     "execution_count": 355,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = 'ABCAADEFGH'\n",
    "a[2:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 348,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "4-len(a)-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1915"
      ]
     },
     "execution_count": 341,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(wt_full_dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1857"
      ]
     },
     "execution_count": 342,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(wt_dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>762.0</td>\n",
       "      <td>ASXL1</td>\n",
       "      <td>20</td>\n",
       "      <td>31022441</td>\n",
       "      <td>31022442</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-003409-1958</td>\n",
       "      <td>NM_015338.5</td>\n",
       "      <td>642.0</td>\n",
       "      <td>13/13</td>\n",
       "      <td>ENST00000375687.4:c.1934dup</td>\n",
       "      <td>p.Gly646TrpfsTer12</td>\n",
       "      <td>p.G646Wfs*12</td>\n",
       "      <td>ENST00000375687.4</td>\n",
       "      <td>31022381</td>\n",
       "      <td>31022502</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>645.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175952</td>\n",
       "      <td>112175953</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-JHU-00198-00378</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1554.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4666dup</td>\n",
       "      <td>p.Thr1556AsnfsTer3</td>\n",
       "      <td>p.T1556Nfs*3</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175892</td>\n",
       "      <td>112176013</td>\n",
       "      <td>GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...</td>\n",
       "      <td>GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>280.0</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>1</td>\n",
       "      <td>27105930</td>\n",
       "      <td>27105931</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-001750-9089</td>\n",
       "      <td>NM_006015.4</td>\n",
       "      <td>1847.0</td>\n",
       "      <td>20/20</td>\n",
       "      <td>ENST00000324856.7:c.5548dup</td>\n",
       "      <td>p.Asp1850GlyfsTer4</td>\n",
       "      <td>p.D1850Gfs*4</td>\n",
       "      <td>ENST00000324856.7</td>\n",
       "      <td>27105870</td>\n",
       "      <td>27105991</td>\n",
       "      <td>CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...</td>\n",
       "      <td>CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>55</td>\n",
       "      <td>108.0</td>\n",
       "      <td>TCF7L2</td>\n",
       "      <td>10</td>\n",
       "      <td>114925316</td>\n",
       "      <td>114925317</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-DFCI-151674-1525793</td>\n",
       "      <td>NaN</td>\n",
       "      <td>482.0</td>\n",
       "      <td>15/15</td>\n",
       "      <td>ENST00000355995.4:c.1454dup</td>\n",
       "      <td>p.Cys486ValfsTer8</td>\n",
       "      <td>p.C486Vfs*8</td>\n",
       "      <td>ENST00000355995.4</td>\n",
       "      <td>114925256</td>\n",
       "      <td>114925377</td>\n",
       "      <td>CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...</td>\n",
       "      <td>CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>60</td>\n",
       "      <td>99.0</td>\n",
       "      <td>GATA3</td>\n",
       "      <td>10</td>\n",
       "      <td>8115874</td>\n",
       "      <td>8115875</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-002328-1774</td>\n",
       "      <td>NaN</td>\n",
       "      <td>407.0</td>\n",
       "      <td>6/6</td>\n",
       "      <td>ENST00000346208.3:c.1221dup</td>\n",
       "      <td>p.Pro408AlafsTer99</td>\n",
       "      <td>p.P408Afs*99</td>\n",
       "      <td>ENST00000346208.3</td>\n",
       "      <td>8115814</td>\n",
       "      <td>8115935</td>\n",
       "      <td>AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...</td>\n",
       "      <td>AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>8994</td>\n",
       "      <td>5.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175621</td>\n",
       "      <td>112175622</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-MSK-P-0048781-T01-IM6</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1444.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4333dup</td>\n",
       "      <td>p.Thr1445AsnfsTer10</td>\n",
       "      <td>p.T1445Nfs*10</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175561</td>\n",
       "      <td>112175682</td>\n",
       "      <td>CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...</td>\n",
       "      <td>CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>9010</td>\n",
       "      <td>5.0</td>\n",
       "      <td>SRSF2</td>\n",
       "      <td>17</td>\n",
       "      <td>74732960</td>\n",
       "      <td>74732961</td>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>GGC</td>\n",
       "      <td>GENIE-UHN-AGI147550-BM1</td>\n",
       "      <td>NM_001195427.1</td>\n",
       "      <td>94.0</td>\n",
       "      <td>1/3</td>\n",
       "      <td>ENST00000359995.5:c.282_283insGCC</td>\n",
       "      <td>p.Arg94_Pro95insAla</td>\n",
       "      <td>p.R94_P95insA</td>\n",
       "      <td>ENST00000359995.5</td>\n",
       "      <td>74732900</td>\n",
       "      <td>74733021</td>\n",
       "      <td>AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...</td>\n",
       "      <td>AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>9011</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>X</td>\n",
       "      <td>15841255</td>\n",
       "      <td>15841256</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>AGCCGC</td>\n",
       "      <td>GENIE-UHN-AGI974552-BM1</td>\n",
       "      <td>NM_005089.3</td>\n",
       "      <td>447.0</td>\n",
       "      <td>11/11</td>\n",
       "      <td>ENST00000307771.7:c.1339_1340insAGCCGC</td>\n",
       "      <td>p.Ser447delinsLysProArg</td>\n",
       "      <td>p.S447delinsKPR</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15841195</td>\n",
       "      <td>15841316</td>\n",
       "      <td>AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...</td>\n",
       "      <td>AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>9013</td>\n",
       "      <td>5.0</td>\n",
       "      <td>TSC1</td>\n",
       "      <td>9</td>\n",
       "      <td>135771988</td>\n",
       "      <td>135771989</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>GCT</td>\n",
       "      <td>GENIE-UHN-DIVA273992-ARC1</td>\n",
       "      <td>NM_001162426.1</td>\n",
       "      <td>1043.0</td>\n",
       "      <td>23/23</td>\n",
       "      <td>ENST00000298552.3:c.3128_3129insAGC</td>\n",
       "      <td>p.Ser1043delinsArgAla</td>\n",
       "      <td>p.S1043delinsRA</td>\n",
       "      <td>ENST00000298552.3</td>\n",
       "      <td>135771928</td>\n",
       "      <td>135772049</td>\n",
       "      <td>CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...</td>\n",
       "      <td>CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>9094</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ZFHX3</td>\n",
       "      <td>16</td>\n",
       "      <td>72821633</td>\n",
       "      <td>72821634</td>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>GCCACT</td>\n",
       "      <td>GENIE-PROV-2219952519-76fe5c70a7</td>\n",
       "      <td>NM_006885.3</td>\n",
       "      <td>3514.0</td>\n",
       "      <td>10/10</td>\n",
       "      <td>ENST00000268489.5:c.10541_10542insAGTGGC</td>\n",
       "      <td>p.Gly3514_Gly3515insValAla</td>\n",
       "      <td>p.G3514_G3515insVA</td>\n",
       "      <td>ENST00000268489.5</td>\n",
       "      <td>72821573</td>\n",
       "      <td>72821694</td>\n",
       "      <td>GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...</td>\n",
       "      <td>GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>347 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     index  COUNT Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0        4  762.0       ASXL1         20        31022441      31022442   \n",
       "1        5  645.0         APC          5       112175952     112175953   \n",
       "2       21  280.0      ARID1A          1        27105930      27105931   \n",
       "4       55  108.0      TCF7L2         10       114925316     114925317   \n",
       "5       60   99.0       GATA3         10         8115874       8115875   \n",
       "..     ...    ...         ...        ...             ...           ...   \n",
       "354   8994    5.0         APC          5       112175621     112175622   \n",
       "355   9010    5.0       SRSF2         17        74732960      74732961   \n",
       "356   9011    5.0       ZRSR2          X        15841255      15841256   \n",
       "357   9013    5.0        TSC1          9       135771988     135771989   \n",
       "360   9094    5.0       ZFHX3         16        72821633      72821634   \n",
       "\n",
       "                  Consequence Variant_Classification Variant_Type  \\\n",
       "0          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "1          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "2          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "4          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "5          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "..                        ...                    ...          ...   \n",
       "354        frameshift_variant        Frame_Shift_Ins          INS   \n",
       "355         inframe_insertion           In_Frame_Ins          INS   \n",
       "356  protein_altering_variant           In_Frame_Ins          INS   \n",
       "357  protein_altering_variant           In_Frame_Ins          INS   \n",
       "360         inframe_insertion           In_Frame_Ins          INS   \n",
       "\n",
       "    Reference_Allele Tumor_Seq_Allele2              Tumor_Sample_Barcode  \\\n",
       "0                  -                 G            GENIE-DFCI-003409-1958   \n",
       "1                  -                 A             GENIE-JHU-00198-00378   \n",
       "2                  -                 G            GENIE-DFCI-001750-9089   \n",
       "4                  -                 A         GENIE-DFCI-151674-1525793   \n",
       "5                  -                 G            GENIE-DFCI-002328-1774   \n",
       "..               ...               ...                               ...   \n",
       "354                -                 A       GENIE-MSK-P-0048781-T01-IM6   \n",
       "355                -               GGC           GENIE-UHN-AGI147550-BM1   \n",
       "356                -            AGCCGC           GENIE-UHN-AGI974552-BM1   \n",
       "357                -               GCT         GENIE-UHN-DIVA273992-ARC1   \n",
       "360                -            GCCACT  GENIE-PROV-2219952519-76fe5c70a7   \n",
       "\n",
       "             RefSeq  Protein_position Exon_Number  \\\n",
       "0       NM_015338.5             642.0       13/13   \n",
       "1       NM_000038.5            1554.0       16/16   \n",
       "2       NM_006015.4            1847.0       20/20   \n",
       "4               NaN             482.0       15/15   \n",
       "5               NaN             407.0         6/6   \n",
       "..              ...               ...         ...   \n",
       "354     NM_000038.5            1444.0       16/16   \n",
       "355  NM_001195427.1              94.0         1/3   \n",
       "356     NM_005089.3             447.0       11/11   \n",
       "357  NM_001162426.1            1043.0       23/23   \n",
       "360     NM_006885.3            3514.0       10/10   \n",
       "\n",
       "                                        HGVSc                       HGVSp  \\\n",
       "0                 ENST00000375687.4:c.1934dup          p.Gly646TrpfsTer12   \n",
       "1                 ENST00000257430.4:c.4666dup          p.Thr1556AsnfsTer3   \n",
       "2                 ENST00000324856.7:c.5548dup          p.Asp1850GlyfsTer4   \n",
       "4                 ENST00000355995.4:c.1454dup           p.Cys486ValfsTer8   \n",
       "5                 ENST00000346208.3:c.1221dup          p.Pro408AlafsTer99   \n",
       "..                                        ...                         ...   \n",
       "354               ENST00000257430.4:c.4333dup         p.Thr1445AsnfsTer10   \n",
       "355         ENST00000359995.5:c.282_283insGCC         p.Arg94_Pro95insAla   \n",
       "356    ENST00000307771.7:c.1339_1340insAGCCGC     p.Ser447delinsLysProArg   \n",
       "357       ENST00000298552.3:c.3128_3129insAGC       p.Ser1043delinsArgAla   \n",
       "360  ENST00000268489.5:c.10541_10542insAGTGGC  p.Gly3514_Gly3515insValAla   \n",
       "\n",
       "            HGVSp_Short transcript_id_TRUE  seq_start    seq_end  \\\n",
       "0          p.G646Wfs*12  ENST00000375687.4   31022381   31022502   \n",
       "1          p.T1556Nfs*3  ENST00000257430.4  112175892  112176013   \n",
       "2          p.D1850Gfs*4  ENST00000324856.7   27105870   27105991   \n",
       "4           p.C486Vfs*8  ENST00000355995.4  114925256  114925377   \n",
       "5          p.P408Afs*99  ENST00000346208.3    8115814    8115935   \n",
       "..                  ...                ...        ...        ...   \n",
       "354       p.T1445Nfs*10  ENST00000257430.4  112175561  112175682   \n",
       "355       p.R94_P95insA  ENST00000359995.5   74732900   74733021   \n",
       "356     p.S447delinsKPR  ENST00000307771.7   15841195   15841316   \n",
       "357     p.S1043delinsRA  ENST00000298552.3  135771928  135772049   \n",
       "360  p.G3514_G3515insVA  ENST00000268489.5   72821573   72821694   \n",
       "\n",
       "                                          wt_w_context  \\\n",
       "0    GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...   \n",
       "1    GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...   \n",
       "2    CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...   \n",
       "4    CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...   \n",
       "5    AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...   \n",
       "..                                                 ...   \n",
       "354  CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...   \n",
       "355  AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...   \n",
       "356  AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...   \n",
       "357  CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...   \n",
       "360  GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...   \n",
       "\n",
       "                                         alt_w_context  \n",
       "0    GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...  \n",
       "1    GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...  \n",
       "2    CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...  \n",
       "4    CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...  \n",
       "5    AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...  \n",
       "..                                                 ...  \n",
       "354  CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...  \n",
       "355  AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...  \n",
       "356  AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...  \n",
       "357  CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...  \n",
       "360  GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...  \n",
       "\n",
       "[347 rows x 23 columns]"
      ]
     },
     "execution_count": 343,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MKDKQKKKKERTWAEAARLVLENYSDAPMTPKQILQVIEAEGLKEMRSGTSPLA...VVR')"
      ]
     },
     "execution_count": 316,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt_full_dna.transcribe().translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G646Wfs*12\n"
     ]
    }
   ],
   "source": [
    "wt_dna = str(wt_dna)\n",
    "mut_dna = str(alt_full_dna)\n",
    "wt_seq = Bio.Seq.Seq(wt_dna).transcribe().translate()\n",
    "mut_seq = Bio.Seq.Seq(alt_full_dna).transcribe().translate()\n",
    "\n",
    "hgvsp = get_hgvsp_classification(wt_dna, mut_dna, wt_seq, mut_seq)\n",
    "hgvsc = get_hgvsc(wt_dna, mut_dna)\n",
    "\n",
    "print(hgvsp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4624"
      ]
     },
     "execution_count": 333,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(mut_dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/samgould/opt/anaconda3/lib/python3.9/site-packages/Bio/Seq.py:2979: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "mut_seq = Bio.Seq.Seq(alt_full_dna).transcribe().translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CGAAGAAGCGCTTGTGGCCATGCTTCTGCTTGCGCAGGTAGCCGCACTTGCGCACGCTGTGGTTGTTGTTGTTGTTGTTGTTGTTGAGGTTGGGGCCGTCTCCGCTCGCCGGCCCGGGCG'"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CGAAGAAGCGCTTGTGGCCATGCTTCTGCTTGCGCAGGTAGCCGCACTTGCGCACGCTGTGGTTGTTGTTGTTGTTGTTGTTGTTGAGGTTGGGGCCGTCTCCGCTCGCCGGCCCGGGCGGCC\n",
      "CGAAGAAGCGCTTGTGGCCATGCTTCTGCTTGCGCAGGTAGCCGCACTTGCGCACGCTGTTTGTTGTTGTTGTTGTTGTTGTTGAGGTTGGGGCCGTCTCCGCTCGCCGGCCCGGGCGGC\n"
     ]
    }
   ],
   "source": [
    "print(str(chr_seq[start-1:end]))\n",
    "print(val['wt_w_context'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "58"
      ]
     },
     "execution_count": 264,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('GTTGTTGTTGTTGTTGTTGTTGAGGTTGGGGCCGTCTCCGCTCGCCGGCCCGGGCGGC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n"
     ]
    }
   ],
   "source": [
    "for i in range(3,10):\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MASPPRHGPPGPASGDGPNLNNNNNNNNHSVRKCGYLRKQKHGHKRFFVLRGPG...VKE')"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Bio.Seq.Seq(wt_dna.reverse_complement()).transcribe().translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'-'"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val = ins_new.iloc[10]\n",
    "\n",
    "\n",
    "\n",
    "tx = val['transcript_id_TRUE']\n",
    "cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "\n",
    "db[tx].strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Feature CDS (chr13:110408654-110408655[-]) at 0x7fe9d0f4e0d0>,\n",
       " <Feature CDS (chr13:110434389-110438400[-]) at 0x7fe9679b29d0>]"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TAATAAATTATGGCCATGGGAAACATTGTATATTTAGTGTGTGTATTTTGATAA...NNN')"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_seq[31025138:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MKDKQKKKKERTWAEAARLVLENYSDAPMTPKQILQVIEAEGLKEMRSGTSPLA...VR*')"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Bio.Seq.Seq(wt_dna).transcribe().translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31022503"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val['seq_start']\n",
    "val['seq_end']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_hgvsc(mutant_seq, ref_seq):\n",
    "    \"\"\"\n",
    "    Returns the HGVSc notation for a mutant cDNA sequence compared to a reference sequence.\n",
    "\n",
    "    Note: doesn't work for composite mutations. It will only classify the first mutant in the sequence.\n",
    "\n",
    "    Need to make a matching function that converts from hgvsc to mutant_seq given ref_seq and hgvsc\n",
    "    \"\"\"\n",
    "    if mutant_seq == ref_seq:\n",
    "        return 'WT'  # no change, no HGVSc notation\n",
    "    \n",
    "    # find the position(s) of the variant(s) and the type of variation\n",
    "    pos = []\n",
    "    var_type = \"\"\n",
    "    if len(mutant_seq) > len(ref_seq):  # insertion\n",
    "        for i in range(len(ref_seq)):\n",
    "            if mutant_seq[i] != ref_seq[i]:\n",
    "                pos.append(i+1)\n",
    "\n",
    "        var_type = \"ins\" + mutant_seq[pos[0]-1:pos[0] -1 + len(mutant_seq)-len(ref_seq)]\n",
    "        #print(var_type)\n",
    "\n",
    "    elif len(mutant_seq) < len(ref_seq):  # deletion\n",
    "        for i in range(len(mutant_seq)):\n",
    "            if mutant_seq[i] != ref_seq[i]:\n",
    "                pos.append(i+1)\n",
    "\n",
    "        #print(pos)\n",
    "        #print(len(ref_seq)-len(mutant_seq))\n",
    "        var_type = \"del\" + ref_seq[pos[0]-1: pos[0] -1 + len(ref_seq)-len(mutant_seq)]\n",
    "\n",
    "    else:  # substitution; doesn't take into account multiple substitutions...\n",
    "        for i in range(len(mutant_seq)):\n",
    "            if mutant_seq[i] != ref_seq[i]:\n",
    "                pos.append(i+1)\n",
    "        \n",
    "        if len(pos)==1:\n",
    "            var_type = ref_seq[pos[0]-1] + \">\" + mutant_seq[pos[0]-1]\n",
    "        else:\n",
    "            var_type = ref_seq[pos[0]-1:pos[-1]] + \">\" + mutant_seq[pos[0]-1:pos[-1]]\n",
    "    \n",
    "    if var_type[0:3]=='ins':\n",
    "        pos = [i-1 for i in pos]\n",
    "\n",
    "    # generate the HGVSc notation\n",
    "    if len(pos) == 1:\n",
    "        hgvsc = \"c.\" + str(pos[0]) + var_type\n",
    "    else:\n",
    "        if var_type[0:3]=='ins':\n",
    "            hgvsc = \"c.\" + str(pos[0]) + \"_\" + str(pos[0]+1) + var_type\n",
    "        elif var_type[0:3]=='del':\n",
    "            if len(ref_seq)-len(mutant_seq) ==1:\n",
    "                hgvsc = \"c.\" + str(pos[0]) + var_type\n",
    "            else:\n",
    "                hgvsc = \"c.\" + str(pos[0]) + \"_\" + str(pos[0]+len(ref_seq)-len(mutant_seq)-1) + var_type\n",
    "        else:\n",
    "            hgvsc = \"c.\" + str(pos[0]) + \"_\" + str(pos[-1]) + var_type\n",
    "    \n",
    "    return hgvsc\n",
    "\n",
    "def get_hgvsp_classification(wt_dna, mut_dna, wt_seq, mut_seq):\n",
    "    \"\"\"\n",
    "    Given a wildtype protein sequence and a mutant protein sequence, returns the HGVSp classification for the mutant sequence.\n",
    "\n",
    "    Args:\n",
    "        wt_seq (str): The wildtype protein sequence.\n",
    "        mut_seq (str): The mutant protein sequence.\n",
    "\n",
    "    Returns:\n",
    "        str: The HGVSp classification for the mutant protein sequence.\n",
    "\n",
    "\n",
    "    There are a bunch of cases where this fails miserably. The chain of logical statements doesn't necessarily make sense.\n",
    "    \"\"\"\n",
    "    # Find the position of the first amino acid that differs between the wildtype and mutant sequences\n",
    "    pos = None\n",
    "    for i in range(len(wt_seq)):\n",
    "        if i >= len(mut_seq) or wt_seq[i] != mut_seq[i]:\n",
    "            pos = i + 1\n",
    "            break\n",
    "\n",
    "    #WT sequence (no change)\n",
    "    if pos==None:\n",
    "        #check for and identify silent substitution variants\n",
    "\n",
    "        pp = None\n",
    "        for i in range(0,len(mut_dna)-3,3):\n",
    "            mut_cod = mut_dna[i:i+3]\n",
    "            wt_cod = wt_dna[i:i+3]\n",
    "            if mut_cod != wt_cod:\n",
    "                pp = i+1            \n",
    "        \n",
    "        if pp!=None:\n",
    "            pg = pp//3 \n",
    "            return f'{mut_seq[pg]}{pg+1}{mut_seq[pg]}'\n",
    "        else:\n",
    "            return 'WT'\n",
    "\n",
    "    # Check if the mutant sequence contains a stop codon\n",
    "    if '*' in mut_seq and mut_seq.index('*') < pos:\n",
    "        # The mutation is a termination variant\n",
    "        pos = mut_seq.index('*') + 1\n",
    "        return f'{wt_seq[pos-1]}{pos}*'\n",
    "\n",
    "\n",
    "    # Check the difference in sequence length between the wildtype and mutant sequences\n",
    "    diff_len1 = len(mut_dna) - len(wt_dna)\n",
    "\n",
    "    diff_len = len(mut_seq) - len(wt_seq)\n",
    "    #print(diff_len)\n",
    "\n",
    "    if diff_len1 == 0:\n",
    "        # The mutation is a substitution\n",
    "        return f'{wt_seq[pos-1]}{pos}{mut_seq[pos-1]}' \n",
    "        #doesn't take into account potential differences \n",
    "\n",
    "    elif diff_len1 > 0:\n",
    "        # The mutation is an insertion\n",
    "        ins_pos = pos + 1\n",
    "        \n",
    "        #check if it's a frameshift based on mRNA sequence\n",
    "        #not a frameshift if divisible by 3\n",
    "        if diff_len1 % 3 ==0:\n",
    "            \n",
    "            return f'{wt_seq[pos-2]}{pos-1}_{wt_seq[pos-1]}{ins_pos-1}ins{mut_seq[pos-1: pos-1+diff_len]}'\n",
    "        \n",
    "        #otherwise it will be a frameshift\n",
    "        else:\n",
    "            stop_pos = len(mut_seq)\n",
    "            for i in range(pos, len(mut_seq)):\n",
    "                if mut_seq[i]==\"*\":\n",
    "                    stop_pos = i+2-pos\n",
    "                    break\n",
    "\n",
    "            return f'{wt_seq[pos-1]}{pos}{mut_seq[pos-1]}fs*{stop_pos}'\n",
    "\n",
    "    \n",
    "    else:\n",
    "        # The mutation is a deletion or a more complex variant\n",
    "        # Check if the mutation is a pure deletion or a more complex variant\n",
    "        deleted = True\n",
    "        for i in range(pos, pos - diff_len):\n",
    "            if i >= len(mut_seq) or mut_seq[i] != wt_seq[i-1]:\n",
    "                deleted = False\n",
    "                break\n",
    "\n",
    "        if deleted:\n",
    "            # The mutation is a deletion\n",
    "            del_len = -diff_len\n",
    "            return f'{wt_seq[pos-1]}{pos}_{wt_seq[pos+del_len-1]}{pos+del_len-1}del'\n",
    "        \n",
    "        else:\n",
    "            # The mutation is a complex variant\n",
    "            # Find the first stop codon downstream of the mutation\n",
    "            stop_pos = len(mut_seq)\n",
    "            for i in range(pos, len(mut_seq)):\n",
    "                if mut_seq[i]==\"*\":\n",
    "                    stop_pos = i+2-pos\n",
    "                    break\n",
    "\n",
    "            return f'{wt_seq[pos-1]}{pos}{mut_seq[pos-1]}fs*{stop_pos}'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
