{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PEGG 2.0\n",
    "\n",
    "- Updates to PEGG:\n",
    "    - more flexible data input (following PRIDICT/PrimeDesign format)\n",
    "    - more flexible PAM sequence searching\n",
    "    - inclusion of base editing sensor module\n",
    "    - fixing INS/DEL error\n",
    "    - G+19 instead of G+20\n",
    "    - Better fetching of MIT specificity and Rule Set 2/3 information...\n",
    "    - Automated generation of WT/edited sequence generation\n",
    "        - Use this to determine errors in pegRNAs...\n",
    "    - Improved distance to nick and homology overhang sizing\n",
    "    - Also automate finding of silent variants for MMR-evasion...\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import regex as re\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## improved PAM finder for various PAM sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "substitution_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GGC/TTG)AGAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "ins_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGGGA(/GTAA)GAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "del_eaxmple = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGGGAG(AGAC/)CCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#introduce error messages for each of these\n",
    "\n",
    "start = substitution_example.find(\"(\")\n",
    "\n",
    "end = substitution_example.find(\")\")\n",
    "\n",
    "replace_seq = substitution_example[start:end+1]\n",
    "#if '/' in replace_seq:\n",
    "    #if '+' in replace_seq:\n",
    "        #throw an error\n",
    "\n",
    "loc_replace = replace_seq.find('/')\n",
    "wt_replace = replace_seq[1:loc_replace]\n",
    "mut_replace = replace_seq[loc_replace+1:-1]\n",
    "\n",
    "wt_seq = substitution_example[:start] + wt_replace + substitution_example[end+1:]\n",
    "\n",
    "wt_start = start\n",
    "wt_end = start+len(wt_replace)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "replace_seq = '(/AGGG)'\n",
    "loc_replace = replace_seq.find('/')\n",
    "wt_replace = replace_seq[1:loc_replace]\n",
    "mut_replace = replace_seq[loc_replace+1:-1]\n",
    "wt_replace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "103"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "context_size = 50\n",
    "\n",
    "wt_w_context = wt_seq[wt_start-context_size:wt_start] + wt_seq[wt_start:wt_end] + wt_seq[wt_end:wt_end+context_size]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mut_formatter(pridict_format, context_size = 60):\n",
    "    \"\"\" \n",
    "    Takes in mutations in format of AAA(AA/GC)ATAGC\n",
    "    and converts it into a format that allows pegRNAs to be generated\n",
    "\n",
    "    Parameters\n",
    "    _____\n",
    "    pridict_format = list of sequences in e.g. AAA(AA/GC)ATAGC format\n",
    "    context_size = amount of nt on either side of mutation to select\n",
    "    \"\"\"\n",
    "\n",
    "    original_start = []\n",
    "    original_end = []\n",
    "\n",
    "    wt = []\n",
    "    mut = []\n",
    "    \n",
    "    wt_context = []\n",
    "    mut_context = []\n",
    "\n",
    "    left_context_length = []\n",
    "    right_context_length = []\n",
    "\n",
    "    replace_start = []\n",
    "    replace_end = []\n",
    "\n",
    "    for k in pridict_format:\n",
    "        #find mutation\n",
    "        start = k.find(\"(\")\n",
    "        end = k.find(\")\")\n",
    "\n",
    "        #get out the mutant and WT allele\n",
    "        replace_seq = k[start:end+1]\n",
    "        loc_replace = replace_seq.find('/')\n",
    "        wt_replace = replace_seq[1:loc_replace]\n",
    "        mut_replace = replace_seq[loc_replace+1:-1]\n",
    "\n",
    "        wt.append(wt_replace)\n",
    "        mut.append(mut_replace)\n",
    "\n",
    "        #generate full WT sequence\n",
    "        wt_seq = substitution_example[:start] + wt_replace + substitution_example[end+1:]\n",
    "        wt_start = start\n",
    "        wt_end = start+len(wt_replace)\n",
    "        original_start.append(wt_start)\n",
    "        original_end.append(wt_end)\n",
    "\n",
    "        #and just the subset\n",
    "        left_context = wt_seq[wt_start-context_size:wt_start]\n",
    "        print(left_context)\n",
    "        right_context = wt_seq[wt_end:wt_end+context_size]\n",
    "\n",
    "        wt_w_context = left_context + wt_replace + right_context\n",
    "        mut_w_context = left_context + mut_replace + right_context\n",
    "\n",
    "        wt_context.append(wt_w_context)\n",
    "        mut_context.append(mut_w_context)\n",
    "\n",
    "        left_context_length.append(len(left_context))\n",
    "        right_context_length.append(len(right_context))\n",
    "\n",
    "        replace_start.append(len(left_context))\n",
    "        replace_end.append(len(left_context)+len(wt_replace))\n",
    "\n",
    "    col_labels = ['Original_start', 'Original_end', 'WT', 'Mutant', 'WT_context', 'Mutant_context', 'Replace_start', 'Replace_end', 'Left_context_length', 'Right_context_length']\n",
    "    cols = [original_start, original_end, wt, mut, wt_context, mut_context, replace_start, replace_end, left_context_length, right_context_length]\n",
    "\n",
    "    df = pd.DataFrame(dict(zip(col_labels, cols)))\n",
    "\n",
    "    return df\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG\n",
      "GTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(G\n",
      "TTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GG\n"
     ]
    }
   ],
   "source": [
    "substitution_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GGC/TTG)AGAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "ins_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGGGA(/GTAA)GAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "del_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGGGAG(AGAC/)CCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "\n",
    "pridict_format = [substitution_example, ins_example, del_example]\n",
    "\n",
    "df = mut_formatter(pridict_format)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GGAGACGAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAAC\n",
      "TTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GGGAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAAC\n"
     ]
    }
   ],
   "source": [
    "print(df.iloc[2]['WT_context'])\n",
    "print(df.iloc[2]['Mutant_context'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PAM_finder(WT_seq, mut_start, mut_end,  PAM, RTT_length_max):\n",
    "    \"\"\"Identifies the location of PAM sequences on the + and - strand.\n",
    "    Returns a 2-d array containing marked locations of PAM sequence start locations on + and - strand.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    WT_seq: WT sequence...\n",
    "\n",
    "    PAM_seq: PAM sequence to search for...\n",
    "\n",
    "    RTT_length_max: maximum size of RTT length for searching...\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    distance_PAM_to_nick = 3\n",
    "    PAM_size = len(PAM)\n",
    "    mut_size = mut_start-mut_end\n",
    "\n",
    "    search_size = RTT_length_max - distance_PAM_to_nick\n",
    "\n",
    "    #deal with extra PAM sequences later on...\n",
    "    #and other potential issues\n",
    "    plus_search = WT_seq[mut_end-search_size-5 : mut_start+distance_PAM_to_nick + PAM_size+5]\n",
    "\n",
    "\n",
    "    #---------------Loading in sequences for PAM Searching------------------#\n",
    "    \n",
    "    search_size = RTT_length - len(PAM)-1 #need to modify this for insertion/deletions...\n",
    "    #size mut doesn't capture the size of insertions, only the size of deletions\n",
    "\n",
    "    plus_search = seq1[seq_start-1-search_size : seq_end+search_size].upper()\n",
    "    minus_search = plus_search.complement().upper()\n",
    "    \n",
    "    mut_start_idx = 1+search_size\n",
    "    mut_end_idx = 1+search_size+size_mut #not accurate for insertions; does work for indexing though...\n",
    "\n",
    "    plus_search1 = plus_search[:mut_start_idx+3+len(PAM)-1]\n",
    "    minus_search1 = minus_search[mut_start_idx-3-len(PAM):]\n",
    "\n",
    "    #---------------PAM Searching------------------#\n",
    "\n",
    "    #replacing N with regex symbol\n",
    "    PAM_regex = PAM.replace('N', '/*.')\n",
    "\n",
    "    PAM_search_plus = re.compile('(?=(' + PAM_regex + '))', re.IGNORECASE)\n",
    "\n",
    "    iterator_plus = PAM_search_plus.finditer(str(plus_search1))\n",
    "    PAM_starts_plus = [match.start() for match in iterator_plus]\n",
    "\n",
    "\n",
    "    PAM_minus = PAM[::-1]#reversing it\n",
    "    PAM_regex_minus = PAM_minus.replace('N', '/*.')\n",
    "\n",
    "    PAM_search_minus = re.compile('(?=(' + PAM_regex_minus + '))', re.IGNORECASE)\n",
    "\n",
    "    iterator_minus = PAM_search_minus.finditer(str(minus_search1))\n",
    "    PAM_starts_minus = [match.start() for match in iterator_minus]\n",
    "    #since things are flipped on minus strand, adding len(PAM) to get the true \"start\" to the PAM\n",
    "\n",
    "    PAM_starts_minus = np.asarray(PAM_starts_minus) + len(PAM) + (mut_start_idx-3-len(PAM))#and correct for indexing\n",
    "\n",
    "    return np.asarray([np.asarray(PAM_starts_plus), PAM_starts_minus], dtype='object')-mut_start_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PAM_finder(WT_seq, mut_start, mut_end, PAM, RTT_length_max):\n",
    "    \"\"\"Identifies the location of PAM sequences on the + and - strand.\n",
    "    Returns a 2-d array containing marked locations of PAM sequence start locations on + and - strand.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    WT_seq: WT sequence...\n",
    "\n",
    "    PAM_seq: PAM sequence to search for...\n",
    "\n",
    "    RTT_length_max: maximum size of RTT length for searching...\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    #---------------Loading in sequences for PAM Searching------------------#\n",
    "    \n",
    "    search_size = RTT_length - len(PAM)-1 #need to modify this for insertion/deletions...\n",
    "    #size mut doesn't capture the size of insertions, only the size of deletions\n",
    "\n",
    "    plus_search = seq1[seq_start-1-search_size : seq_end+search_size].upper()\n",
    "    minus_search = plus_search.complement().upper()\n",
    "    \n",
    "    mut_start_idx = 1+search_size\n",
    "    mut_end_idx = 1+search_size+size_mut #not accurate for insertions; does work for indexing though...\n",
    "\n",
    "    plus_search1 = plus_search[:mut_start_idx+3+len(PAM)-1]\n",
    "    minus_search1 = minus_search[mut_start_idx-3-len(PAM):]\n",
    "\n",
    "    #---------------PAM Searching------------------#\n",
    "\n",
    "    #replacing N with regex symbol\n",
    "    PAM_regex = PAM.replace('N', '/*.')\n",
    "\n",
    "    PAM_search_plus = re.compile('(?=(' + PAM_regex + '))', re.IGNORECASE)\n",
    "\n",
    "    iterator_plus = PAM_search_plus.finditer(str(plus_search1))\n",
    "    PAM_starts_plus = [match.start() for match in iterator_plus]\n",
    "\n",
    "\n",
    "    PAM_minus = PAM[::-1]#reversing it\n",
    "    PAM_regex_minus = PAM_minus.replace('N', '/*.')\n",
    "\n",
    "    PAM_search_minus = re.compile('(?=(' + PAM_regex_minus + '))', re.IGNORECASE)\n",
    "\n",
    "    iterator_minus = PAM_search_minus.finditer(str(minus_search1))\n",
    "    PAM_starts_minus = [match.start() for match in iterator_minus]\n",
    "    #since things are flipped on minus strand, adding len(PAM) to get the true \"start\" to the PAM\n",
    "\n",
    "    PAM_starts_minus = np.asarray(PAM_starts_minus) + len(PAM) + (mut_start_idx-3-len(PAM))#and correct for indexing\n",
    "\n",
    "    return np.asarray([np.asarray(PAM_starts_plus), PAM_starts_minus], dtype='object')-mut_start_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PAM = 'NGG'\n",
    "\n",
    "PAM_finder(WT_seq, mut_start, mut_end, PAM, RTT_length_max)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating list of variants for alvin\n",
    "\n",
    "- Starting with diego library\n",
    "    - includes SNPs in IDR regions\n",
    "    - Also intronic SNPs between exons that are an IDR\n",
    "        - These wont be amenable to MMR-evasive silent edits\n",
    "\n",
    "- Need WT sequnece\n",
    "- Mutant sequence\n",
    "- Frame (i.e. codon frame = 0,1,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pegg\n",
    "import gffutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath = '/Users/samgould/Desktop/FSR Lab/reference files/GRCh37/ncbi-genomes-2022-03-17/GCF_000001405.25_GRCh37.p13_genomic.fna.gz'\n",
    "\n",
    "records, index_list = pegg.genome_loader(filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = '/Users/samgould/Desktop/FSR Lab/reference files/gencode_v19.db'\n",
    "db = gffutils.FeatureDB(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['COUNT', 'Hugo_Symbol', 'Entrez_Gene_Id', 'Center', 'NCBI_Build',\n",
       "       'Chromosome', 'Start_Position', 'End_Position', 'Strand', 'Consequence',\n",
       "       'Variant_Classification', 'Variant_Type', 'Reference_Allele',\n",
       "       'Tumor_Seq_Allele1', 'Tumor_Seq_Allele2', 'dbSNP_RS',\n",
       "       'dbSNP_Val_Status', 'Tumor_Sample_Barcode',\n",
       "       'Matched_Norm_Sample_Barcode', 'Match_Norm_Seq_Allele1',\n",
       "       'Match_Norm_Seq_Allele2', 'Tumor_Validation_Allele1',\n",
       "       'Tumor_Validation_Allele2', 'Match_Norm_Validation_Allele1',\n",
       "       'Match_Norm_Validation_Allele2', 'Verification_Status',\n",
       "       'Validation_Status', 'Mutation_Status', 'Sequencing_Phase',\n",
       "       'Sequence_Source', 'Validation_Method', 'Score', 'BAM_File',\n",
       "       'Sequencer', 't_ref_count', 't_alt_count', 'n_ref_count', 'n_alt_count',\n",
       "       'HGVSc', 'HGVSp', 'HGVSp_Short', 'Transcript_ID', 'RefSeq',\n",
       "       'Protein_position', 'Codons', 'Exon_Number', 'gnomAD_AF',\n",
       "       'gnomAD_AFR_AF', 'gnomAD_AMR_AF', 'gnomAD_ASJ_AF', 'gnomAD_EAS_AF',\n",
       "       'gnomAD_FIN_AF', 'gnomAD_NFE_AF', 'gnomAD_OTH_AF', 'gnomAD_SAS_AF',\n",
       "       'FILTER', 'Polyphen_Prediction', 'Polyphen_Score', 'SIFT_Prediction',\n",
       "       'SIFT_Score', 'SWISSPROT', 'n_depth', 't_depth', 'Annotation_Status',\n",
       "       'mutationInCis_Flag'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#diego variants\n",
    "idr = pd.read_csv('filtered_idr_mutations_5count_6nt_indels.csv')\n",
    "idr.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['DEL', 'DNP', 'INS', 'ONP', 'SNP'], dtype=object),\n",
       " array([ 729,   42,  361,   13, 7955]))"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(idr['Variant_Type'], return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "tx_new = []\n",
    "for i, val in idr.iterrows():\n",
    "    h = val['HGVSc']\n",
    "    tx_new.append(h.split(':')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "idr['transcript_id_TRUE'] = tx_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "#check all transcripts are annotated\n",
    "\n",
    "t_ids = np.unique(idr['transcript_id_TRUE'])\n",
    "\n",
    "not_found = []\n",
    "\n",
    "for tx in t_ids:\n",
    "\n",
    "    cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "    start_end_cds = [[i.start, i.end] for i in cds]\n",
    "    if len(start_end_cds)==0:\n",
    "        not_found.append(tx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[20146427, 20146429],\n",
       " [20148634, 20148725],\n",
       " [20150300, 20150381],\n",
       " [20152075, 20152125],\n",
       " [20153856, 20153959],\n",
       " [20156657, 20156740],\n",
       " [20159743, 20159758]]"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tx = 'ENST00000379607.5'\n",
    "\n",
    "cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "start_end_cds = [[i.start, i.end] for i in cds]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'), id='NC_000022.10', name='NC_000022.10', description='NC_000022.10 Homo sapiens chromosome 22, GRCh37.p13 Primary Assembly', dbxrefs=[])"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records[index_list[21]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/5g/xxjcy2g531n02hnyb6d8qdsr0000gn/T/ipykernel_33329/3196732702.py\u001b[0m in \u001b[0;36m<cell line: 10>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     23\u001b[0m         \u001b[0mchrom\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchrom\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m     \u001b[0mchr_seq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrecords\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mchrom\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     26\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mvt\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'SNP'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ONP'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DNP'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/Bio/Seq.py\u001b[0m in \u001b[0;36mupper\u001b[0;34m(self, inplace)\u001b[0m\n\u001b[1;32m   1209\u001b[0m         \u001b[0mSee\u001b[0m \u001b[0malso\u001b[0m \u001b[0mthe\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m`\u001b[0m\u001b[0;31m`\u001b[0m\u001b[0mlower\u001b[0m\u001b[0;31m`\u001b[0m\u001b[0;31m`\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1210\u001b[0m         \"\"\"\n\u001b[0;32m-> 1211\u001b[0;31m         \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1212\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1213\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbytearray\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "context_size = 60\n",
    "\n",
    "\n",
    "wt_w_context = []\n",
    "alt_w_context = []\n",
    "\n",
    "seq_start = []\n",
    "seq_end = []\n",
    "\n",
    "for i, val in idr.iterrows():\n",
    "    vt = val['Variant_Type']\n",
    "    s = val['Start_Position']\n",
    "    e = val['End_Position']\n",
    "    ref = val['Reference_Allele']\n",
    "    alt = val['Tumor_Seq_Allele2']\n",
    "    chrom = val['Chromosome']\n",
    "\n",
    "    if chrom == 'X':\n",
    "        chrom = 22\n",
    "    elif chrom=='Y':\n",
    "        chrom = 23\n",
    "    else:\n",
    "        chrom = int(chrom)-1\n",
    "\n",
    "    chr_seq = records[index_list[chrom]].seq.upper()\n",
    "\n",
    "    if vt in ['SNP', 'ONP', 'DNP']:\n",
    "        ref = ref\n",
    "        alt = alt\n",
    "        #assert ref == chr_seq[s-1:e], print(ref, chr_seq[s-1:e])\n",
    "\n",
    "    elif vt =='INS':\n",
    "        ref = ''\n",
    "        alt = alt\n",
    "\n",
    "    elif vt=='DEL':\n",
    "        ref = ref\n",
    "        alt = ''\n",
    "\n",
    "    left_context = chr_seq[s-1-context_size:s-1]\n",
    "    right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "    wt_seq = left_context + ref + right_context\n",
    "    alt_seq = left_context + alt + right_context\n",
    "\n",
    "    wt_w_context.append(wt_seq)\n",
    "    alt_w_context.append(alt_seq)\n",
    "    seq_start.append(s-context_size)\n",
    "    seq_end.append(e+context_size+1)\n",
    "\n",
    "#based on these, get the frame, calculate mutation consequence, etc..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "992"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#ba"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "ins = idr[idr['Variant_Type']=='INS']\n",
    "dels = idr[idr['Variant_Type']=='DEL']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele1</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>dbSNP_RS</th>\n",
       "      <th>dbSNP_Val_Status</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Matched_Norm_Sample_Barcode</th>\n",
       "      <th>Match_Norm_Seq_Allele1</th>\n",
       "      <th>Match_Norm_Seq_Allele2</th>\n",
       "      <th>Tumor_Validation_Allele1</th>\n",
       "      <th>Tumor_Validation_Allele2</th>\n",
       "      <th>Match_Norm_Validation_Allele1</th>\n",
       "      <th>Match_Norm_Validation_Allele2</th>\n",
       "      <th>...</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>gnomAD_AF</th>\n",
       "      <th>gnomAD_AFR_AF</th>\n",
       "      <th>gnomAD_AMR_AF</th>\n",
       "      <th>gnomAD_ASJ_AF</th>\n",
       "      <th>gnomAD_EAS_AF</th>\n",
       "      <th>gnomAD_FIN_AF</th>\n",
       "      <th>gnomAD_NFE_AF</th>\n",
       "      <th>gnomAD_OTH_AF</th>\n",
       "      <th>gnomAD_SAS_AF</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>Polyphen_Prediction</th>\n",
       "      <th>Polyphen_Score</th>\n",
       "      <th>SIFT_Prediction</th>\n",
       "      <th>SIFT_Score</th>\n",
       "      <th>SWISSPROT</th>\n",
       "      <th>n_depth</th>\n",
       "      <th>t_depth</th>\n",
       "      <th>Annotation_Status</th>\n",
       "      <th>mutationInCis_Flag</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>762.0</td>\n",
       "      <td>ASXL1</td>\n",
       "      <td>171023.0</td>\n",
       "      <td>DFCI</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>20</td>\n",
       "      <td>31022441</td>\n",
       "      <td>31022442</td>\n",
       "      <td>+</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>rs756958159</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-DFCI-003409-1958</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000375687</td>\n",
       "      <td>NM_015338.5</td>\n",
       "      <td>642.0</td>\n",
       "      <td>-/G</td>\n",
       "      <td>13/13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>83.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000375687.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>645.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>324.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>5</td>\n",
       "      <td>112175952</td>\n",
       "      <td>112175953</td>\n",
       "      <td>+</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-JHU-00198-00378</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000257430</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1554.0</td>\n",
       "      <td>gaa/gaAa</td>\n",
       "      <td>16/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1268.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>280.0</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>8289.0</td>\n",
       "      <td>DFCI</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>27105930</td>\n",
       "      <td>27105931</td>\n",
       "      <td>+</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>rs758608743</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-DFCI-001750-9089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000324856</td>\n",
       "      <td>NM_006015.4</td>\n",
       "      <td>1847.0</td>\n",
       "      <td>-/G</td>\n",
       "      <td>20/20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>582.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000324856.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>234.0</td>\n",
       "      <td>SETBP1</td>\n",
       "      <td>26040.0</td>\n",
       "      <td>VICC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>18</td>\n",
       "      <td>42456671</td>\n",
       "      <td>42456672</td>\n",
       "      <td>+</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>Intron</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>TCTT</td>\n",
       "      <td>rs33928380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-VICC-330134-unk-4</td>\n",
       "      <td>GENIE-VICC-330134-unk-4-Normal</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000282030</td>\n",
       "      <td>NM_015559.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>294.0</td>\n",
       "      <td>534.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000282030.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>108.0</td>\n",
       "      <td>TCF7L2</td>\n",
       "      <td>6934.0</td>\n",
       "      <td>DFCI</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>10</td>\n",
       "      <td>114925316</td>\n",
       "      <td>114925317</td>\n",
       "      <td>+</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>rs769592153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-DFCI-151674-1525793</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000355995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>482.0</td>\n",
       "      <td>aga/agAa</td>\n",
       "      <td>15/15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>170.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000355995.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9011</th>\n",
       "      <td>5.0</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>8233.0</td>\n",
       "      <td>UHN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>X</td>\n",
       "      <td>15841255</td>\n",
       "      <td>15841256</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>AGCCGC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-UHN-AGI974552-BM1</td>\n",
       "      <td>NORMAL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000307771</td>\n",
       "      <td>NM_005089.3</td>\n",
       "      <td>447.0</td>\n",
       "      <td>agc/aAGCCGCgc</td>\n",
       "      <td>11/11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>819.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9013</th>\n",
       "      <td>5.0</td>\n",
       "      <td>TSC1</td>\n",
       "      <td>7248.0</td>\n",
       "      <td>UHN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>9</td>\n",
       "      <td>135771988</td>\n",
       "      <td>135771989</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>GCT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-UHN-DIVA273992-ARC1</td>\n",
       "      <td>NORMAL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000298552</td>\n",
       "      <td>NM_001162426.1</td>\n",
       "      <td>1043.0</td>\n",
       "      <td>agc/agAGCc</td>\n",
       "      <td>23/23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>326.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000298552.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9033</th>\n",
       "      <td>5.0</td>\n",
       "      <td>SMARCA4</td>\n",
       "      <td>6597.0</td>\n",
       "      <td>VICC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>19</td>\n",
       "      <td>11170571</td>\n",
       "      <td>11170572</td>\n",
       "      <td>+</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>Intron</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-VICC-369751-unk-2</td>\n",
       "      <td>GENIE-VICC-369751-unk-2-Normal</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000344626</td>\n",
       "      <td>NM_003072.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>598.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000344626.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9092</th>\n",
       "      <td>5.0</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>5156.0</td>\n",
       "      <td>UCHI</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>4</td>\n",
       "      <td>55161224</td>\n",
       "      <td>55161225</td>\n",
       "      <td>+</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>Intron</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-UCHI-Patient67-T1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000257290</td>\n",
       "      <td>NM_006206.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3315.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000257290.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9094</th>\n",
       "      <td>5.0</td>\n",
       "      <td>ZFHX3</td>\n",
       "      <td>463.0</td>\n",
       "      <td>PROV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>16</td>\n",
       "      <td>72821633</td>\n",
       "      <td>72821634</td>\n",
       "      <td>+</td>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCCACT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-PROV-2219952519-76fe5c70a7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000268489</td>\n",
       "      <td>NM_006885.3</td>\n",
       "      <td>3514.0</td>\n",
       "      <td>ggc/ggAGTGGCc</td>\n",
       "      <td>10/10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>174.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000268489.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>361 rows Ã— 66 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      COUNT Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  \\\n",
       "4     762.0       ASXL1        171023.0   DFCI     GRCh37         20   \n",
       "5     645.0         APC           324.0    JHU     GRCh37          5   \n",
       "21    280.0      ARID1A          8289.0   DFCI     GRCh37          1   \n",
       "23    234.0      SETBP1         26040.0   VICC     GRCh37         18   \n",
       "55    108.0      TCF7L2          6934.0   DFCI     GRCh37         10   \n",
       "...     ...         ...             ...    ...        ...        ...   \n",
       "9011    5.0       ZRSR2          8233.0    UHN     GRCh37          X   \n",
       "9013    5.0        TSC1          7248.0    UHN     GRCh37          9   \n",
       "9033    5.0     SMARCA4          6597.0   VICC     GRCh37         19   \n",
       "9092    5.0      PDGFRA          5156.0   UCHI     GRCh37          4   \n",
       "9094    5.0       ZFHX3           463.0   PROV     GRCh37         16   \n",
       "\n",
       "      Start_Position  End_Position Strand               Consequence  \\\n",
       "4           31022441      31022442      +        frameshift_variant   \n",
       "5          112175952     112175953      +        frameshift_variant   \n",
       "21          27105930      27105931      +        frameshift_variant   \n",
       "23          42456671      42456672      +            intron_variant   \n",
       "55         114925316     114925317      +        frameshift_variant   \n",
       "...              ...           ...    ...                       ...   \n",
       "9011        15841255      15841256      +  protein_altering_variant   \n",
       "9013       135771988     135771989      +  protein_altering_variant   \n",
       "9033        11170571      11170572      +            intron_variant   \n",
       "9092        55161224      55161225      +            intron_variant   \n",
       "9094        72821633      72821634      +         inframe_insertion   \n",
       "\n",
       "     Variant_Classification Variant_Type Reference_Allele Tumor_Seq_Allele1  \\\n",
       "4           Frame_Shift_Ins          INS                -                 -   \n",
       "5           Frame_Shift_Ins          INS                -               NaN   \n",
       "21          Frame_Shift_Ins          INS                -                 -   \n",
       "23                   Intron          INS                -                 -   \n",
       "55          Frame_Shift_Ins          INS                -                 -   \n",
       "...                     ...          ...              ...               ...   \n",
       "9011           In_Frame_Ins          INS                -                 -   \n",
       "9013           In_Frame_Ins          INS                -                 -   \n",
       "9033                 Intron          INS                -                 -   \n",
       "9092                 Intron          INS                -               NaN   \n",
       "9094           In_Frame_Ins          INS                -               NaN   \n",
       "\n",
       "     Tumor_Seq_Allele2     dbSNP_RS  dbSNP_Val_Status  \\\n",
       "4                    G  rs756958159               NaN   \n",
       "5                    A          NaN               NaN   \n",
       "21                   G  rs758608743               NaN   \n",
       "23                TCTT   rs33928380               NaN   \n",
       "55                   A  rs769592153               NaN   \n",
       "...                ...          ...               ...   \n",
       "9011            AGCCGC          NaN               NaN   \n",
       "9013               GCT          NaN               NaN   \n",
       "9033                 G          NaN               NaN   \n",
       "9092                 G          NaN               NaN   \n",
       "9094            GCCACT          NaN               NaN   \n",
       "\n",
       "                  Tumor_Sample_Barcode     Matched_Norm_Sample_Barcode  \\\n",
       "4               GENIE-DFCI-003409-1958                             NaN   \n",
       "5                GENIE-JHU-00198-00378                             NaN   \n",
       "21              GENIE-DFCI-001750-9089                             NaN   \n",
       "23             GENIE-VICC-330134-unk-4  GENIE-VICC-330134-unk-4-Normal   \n",
       "55           GENIE-DFCI-151674-1525793                             NaN   \n",
       "...                                ...                             ...   \n",
       "9011           GENIE-UHN-AGI974552-BM1                          NORMAL   \n",
       "9013         GENIE-UHN-DIVA273992-ARC1                          NORMAL   \n",
       "9033           GENIE-VICC-369751-unk-2  GENIE-VICC-369751-unk-2-Normal   \n",
       "9092           GENIE-UCHI-Patient67-T1                             NaN   \n",
       "9094  GENIE-PROV-2219952519-76fe5c70a7                             NaN   \n",
       "\n",
       "     Match_Norm_Seq_Allele1 Match_Norm_Seq_Allele2  Tumor_Validation_Allele1  \\\n",
       "4                       NaN                    NaN                       NaN   \n",
       "5                       NaN                    NaN                       NaN   \n",
       "21                      NaN                    NaN                       NaN   \n",
       "23                        C                      C                       NaN   \n",
       "55                      NaN                    NaN                       NaN   \n",
       "...                     ...                    ...                       ...   \n",
       "9011                    NaN                    NaN                       NaN   \n",
       "9013                    NaN                    NaN                       NaN   \n",
       "9033                      C                      C                       NaN   \n",
       "9092                    NaN                    NaN                       NaN   \n",
       "9094                    NaN                    NaN                       NaN   \n",
       "\n",
       "      Tumor_Validation_Allele2  Match_Norm_Validation_Allele1  \\\n",
       "4                          NaN                            NaN   \n",
       "5                          NaN                            NaN   \n",
       "21                         NaN                            NaN   \n",
       "23                         NaN                            NaN   \n",
       "55                         NaN                            NaN   \n",
       "...                        ...                            ...   \n",
       "9011                       NaN                            NaN   \n",
       "9013                       NaN                            NaN   \n",
       "9033                       NaN                            NaN   \n",
       "9092                       NaN                            NaN   \n",
       "9094                       NaN                            NaN   \n",
       "\n",
       "      Match_Norm_Validation_Allele2  ...    Transcript_ID          RefSeq  \\\n",
       "4                               NaN  ...  ENST00000375687     NM_015338.5   \n",
       "5                               NaN  ...  ENST00000257430     NM_000038.5   \n",
       "21                              NaN  ...  ENST00000324856     NM_006015.4   \n",
       "23                              NaN  ...  ENST00000282030     NM_015559.2   \n",
       "55                              NaN  ...  ENST00000355995             NaN   \n",
       "...                             ...  ...              ...             ...   \n",
       "9011                            NaN  ...  ENST00000307771     NM_005089.3   \n",
       "9013                            NaN  ...  ENST00000298552  NM_001162426.1   \n",
       "9033                            NaN  ...  ENST00000344626     NM_003072.3   \n",
       "9092                            NaN  ...  ENST00000257290     NM_006206.4   \n",
       "9094                            NaN  ...  ENST00000268489     NM_006885.3   \n",
       "\n",
       "     Protein_position         Codons  Exon_Number  gnomAD_AF  gnomAD_AFR_AF  \\\n",
       "4               642.0            -/G        13/13        NaN            NaN   \n",
       "5              1554.0       gaa/gaAa        16/16        NaN            NaN   \n",
       "21             1847.0            -/G        20/20        NaN            NaN   \n",
       "23                NaN            NaN          NaN        NaN            NaN   \n",
       "55              482.0       aga/agAa        15/15        NaN            NaN   \n",
       "...               ...            ...          ...        ...            ...   \n",
       "9011            447.0  agc/aAGCCGCgc        11/11        NaN            NaN   \n",
       "9013           1043.0     agc/agAGCc        23/23        NaN            NaN   \n",
       "9033              NaN            NaN          NaN        NaN            NaN   \n",
       "9092              NaN            NaN          NaN        NaN            NaN   \n",
       "9094           3514.0  ggc/ggAGTGGCc        10/10        NaN            NaN   \n",
       "\n",
       "      gnomAD_AMR_AF  gnomAD_ASJ_AF  gnomAD_EAS_AF  gnomAD_FIN_AF  \\\n",
       "4               NaN            NaN            NaN            NaN   \n",
       "5               NaN            NaN            NaN            NaN   \n",
       "21              NaN            NaN            NaN            NaN   \n",
       "23              NaN            NaN            NaN            NaN   \n",
       "55              NaN            NaN            NaN            NaN   \n",
       "...             ...            ...            ...            ...   \n",
       "9011            NaN            NaN            NaN            NaN   \n",
       "9013            NaN            NaN            NaN            NaN   \n",
       "9033            NaN            NaN            NaN            NaN   \n",
       "9092            NaN            NaN            NaN            NaN   \n",
       "9094            NaN            NaN            NaN            NaN   \n",
       "\n",
       "      gnomAD_NFE_AF  gnomAD_OTH_AF gnomAD_SAS_AF FILTER Polyphen_Prediction  \\\n",
       "4               NaN            NaN           NaN   PASS                 NaN   \n",
       "5               NaN            NaN           NaN   PASS                 NaN   \n",
       "21              NaN            NaN           NaN   PASS                 NaN   \n",
       "23              NaN            NaN           NaN   PASS                 NaN   \n",
       "55              NaN            NaN           NaN   PASS                 NaN   \n",
       "...             ...            ...           ...    ...                 ...   \n",
       "9011            NaN            NaN           NaN   PASS                 NaN   \n",
       "9013            NaN            NaN           NaN   PASS                 NaN   \n",
       "9033            NaN            NaN           NaN   PASS                 NaN   \n",
       "9092            NaN            NaN           NaN   PASS                 NaN   \n",
       "9094            NaN            NaN           NaN   PASS                 NaN   \n",
       "\n",
       "     Polyphen_Score SIFT_Prediction  SIFT_Score SWISSPROT n_depth  t_depth  \\\n",
       "4               NaN             NaN         NaN       NaN     NaN     83.0   \n",
       "5               NaN             NaN         NaN       NaN     NaN   1268.0   \n",
       "21              NaN             NaN         NaN       NaN     NaN    582.0   \n",
       "23              NaN             NaN         NaN       NaN   294.0    534.0   \n",
       "55              NaN             NaN         NaN       NaN     NaN    170.0   \n",
       "...             ...             ...         ...       ...     ...      ...   \n",
       "9011            NaN             NaN         NaN       NaN     NaN    819.0   \n",
       "9013            NaN             NaN         NaN       NaN     NaN    326.0   \n",
       "9033            NaN             NaN         NaN       NaN   598.0     70.0   \n",
       "9092            NaN             NaN         NaN       NaN     NaN   3315.0   \n",
       "9094            NaN             NaN         NaN       NaN     NaN    174.0   \n",
       "\n",
       "      Annotation_Status  mutationInCis_Flag  transcript_id_TRUE  \n",
       "4               SUCCESS               False   ENST00000375687.4  \n",
       "5               SUCCESS               False   ENST00000257430.4  \n",
       "21              SUCCESS               False   ENST00000324856.7  \n",
       "23              SUCCESS               False   ENST00000282030.5  \n",
       "55              SUCCESS               False   ENST00000355995.4  \n",
       "...                 ...                 ...                 ...  \n",
       "9011            SUCCESS               False   ENST00000307771.7  \n",
       "9013            SUCCESS               False   ENST00000298552.3  \n",
       "9033            SUCCESS               False   ENST00000344626.4  \n",
       "9092            SUCCESS               False   ENST00000257290.5  \n",
       "9094            SUCCESS               False   ENST00000268489.5  \n",
       "\n",
       "[361 rows x 66 columns]"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
