{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PEGG 2.0\n",
    "\n",
    "- Updates to PEGG:\n",
    "    - more flexible data input (following PRIDICT/PrimeDesign format)\n",
    "    - more flexible PAM sequence searching\n",
    "    - inclusion of base editing sensor module\n",
    "    - fixing INS/DEL error\n",
    "    - G+19 instead of G+20\n",
    "    - Better fetching of MIT specificity and Rule Set 2/3 information...\n",
    "    - Automated generation of WT/edited sequence generation\n",
    "        - Use this to determine errors in pegRNAs...\n",
    "    - Improved distance to nick and homology overhang sizing\n",
    "    - Also automate finding of silent variants for MMR-evasion...\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import regex as re\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import Bio.Seq\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## improved PAM finder for various PAM sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "substitution_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GGC/TTG)AGAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "ins_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGGGA(/GTAA)GAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "del_eaxmple = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGGGAG(AGAC/)CCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#introduce error messages for each of these\n",
    "\n",
    "start = substitution_example.find(\"(\")\n",
    "\n",
    "end = substitution_example.find(\")\")\n",
    "\n",
    "replace_seq = substitution_example[start:end+1]\n",
    "#if '/' in replace_seq:\n",
    "    #if '+' in replace_seq:\n",
    "        #throw an error\n",
    "\n",
    "loc_replace = replace_seq.find('/')\n",
    "wt_replace = replace_seq[1:loc_replace]\n",
    "mut_replace = replace_seq[loc_replace+1:-1]\n",
    "\n",
    "wt_seq = substitution_example[:start] + wt_replace + substitution_example[end+1:]\n",
    "\n",
    "wt_start = start\n",
    "wt_end = start+len(wt_replace)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "replace_seq = '(/AGGG)'\n",
    "loc_replace = replace_seq.find('/')\n",
    "wt_replace = replace_seq[1:loc_replace]\n",
    "mut_replace = replace_seq[loc_replace+1:-1]\n",
    "wt_replace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "103"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "context_size = 50\n",
    "\n",
    "wt_w_context = wt_seq[wt_start-context_size:wt_start] + wt_seq[wt_start:wt_end] + wt_seq[wt_end:wt_end+context_size]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mut_formatter(pridict_format, context_size = 60):\n",
    "    \"\"\" \n",
    "    Takes in mutations in format of AAA(AA/GC)ATAGC\n",
    "    and converts it into a format that allows pegRNAs to be generated\n",
    "\n",
    "    Parameters\n",
    "    _____\n",
    "    pridict_format = list of sequences in e.g. AAA(AA/GC)ATAGC format\n",
    "    context_size = amount of nt on either side of mutation to select\n",
    "    \"\"\"\n",
    "\n",
    "    original_start = []\n",
    "    original_end = []\n",
    "\n",
    "    wt = []\n",
    "    mut = []\n",
    "    \n",
    "    wt_context = []\n",
    "    mut_context = []\n",
    "\n",
    "    left_context_length = []\n",
    "    right_context_length = []\n",
    "\n",
    "    replace_start = []\n",
    "    replace_end = []\n",
    "\n",
    "    for k in pridict_format:\n",
    "        #find mutation\n",
    "        start = k.find(\"(\")\n",
    "        end = k.find(\")\")\n",
    "\n",
    "        #get out the mutant and WT allele\n",
    "        replace_seq = k[start:end+1]\n",
    "        loc_replace = replace_seq.find('/')\n",
    "        wt_replace = replace_seq[1:loc_replace]\n",
    "        mut_replace = replace_seq[loc_replace+1:-1]\n",
    "\n",
    "        wt.append(wt_replace)\n",
    "        mut.append(mut_replace)\n",
    "\n",
    "        #generate full WT sequence\n",
    "        wt_seq = substitution_example[:start] + wt_replace + substitution_example[end+1:]\n",
    "        wt_start = start\n",
    "        wt_end = start+len(wt_replace)\n",
    "        original_start.append(wt_start)\n",
    "        original_end.append(wt_end)\n",
    "\n",
    "        #and just the subset\n",
    "        left_context = wt_seq[wt_start-context_size:wt_start]\n",
    "        print(left_context)\n",
    "        right_context = wt_seq[wt_end:wt_end+context_size]\n",
    "\n",
    "        wt_w_context = left_context + wt_replace + right_context\n",
    "        mut_w_context = left_context + mut_replace + right_context\n",
    "\n",
    "        wt_context.append(wt_w_context)\n",
    "        mut_context.append(mut_w_context)\n",
    "\n",
    "        left_context_length.append(len(left_context))\n",
    "        right_context_length.append(len(right_context))\n",
    "\n",
    "        replace_start.append(len(left_context))\n",
    "        replace_end.append(len(left_context)+len(wt_replace))\n",
    "\n",
    "    col_labels = ['Original_start', 'Original_end', 'WT', 'Mutant', 'WT_context', 'Mutant_context', 'Replace_start', 'Replace_end', 'Left_context_length', 'Right_context_length']\n",
    "    cols = [original_start, original_end, wt, mut, wt_context, mut_context, replace_start, replace_end, left_context_length, right_context_length]\n",
    "\n",
    "    df = pd.DataFrame(dict(zip(col_labels, cols)))\n",
    "\n",
    "    return df\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG\n",
      "GTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(G\n",
      "TTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GG\n"
     ]
    }
   ],
   "source": [
    "substitution_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GGC/TTG)AGAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "ins_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGGGA(/GTAA)GAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "del_example = 'CACACCTACACTGCTCGAAGTAAATATGCGAAGCGCGCGGCCTGGCCGGAGGCGTTCCGCGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGCAATCGTAGTCCGTCAAATTCAGCTCTGTTATCCCGGGCGTTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGGGAG(AGAC/)CCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAACAAGTCGATGCAGGCTCCCGTCTTTGAAAAGGGGTAAACATACAAGTGGATAGATGATGGGTAGGGGCCTCCAATACATCCAACACTCTACGCCCTCTCCAAGAGCTAGAAGGGCACCCTGCAGTTGGAAAGGG'\n",
    "\n",
    "pridict_format = [substitution_example, ins_example, del_example]\n",
    "\n",
    "df = mut_formatter(pridict_format)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GGAGACGAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAAC\n",
      "TTATGTGTCAAATGGCGTAGAACGGGATTGACTGTTTGACGGTAGCTGCTGAGGCGG(GGGAGACCCTCCGTCGGGCTATGTCACTAATACTTTCCAAACGCCCCGTACCGATGCTGAAC\n"
     ]
    }
   ],
   "source": [
    "print(df.iloc[2]['WT_context'])\n",
    "print(df.iloc[2]['Mutant_context'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PAM_finder(WT_seq, mut_start, mut_end,  PAM, RTT_length_max):\n",
    "    \"\"\"Identifies the location of PAM sequences on the + and - strand.\n",
    "    Returns a 2-d array containing marked locations of PAM sequence start locations on + and - strand.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    WT_seq: WT sequence...\n",
    "\n",
    "    PAM_seq: PAM sequence to search for...\n",
    "\n",
    "    RTT_length_max: maximum size of RTT length for searching...\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    distance_PAM_to_nick = 3\n",
    "    PAM_size = len(PAM)\n",
    "    mut_size = mut_start-mut_end\n",
    "\n",
    "    search_size = RTT_length_max - distance_PAM_to_nick\n",
    "\n",
    "    #deal with extra PAM sequences later on...\n",
    "    #and other potential issues\n",
    "    plus_search = WT_seq[mut_end-search_size-5 : mut_start+distance_PAM_to_nick + PAM_size+5]\n",
    "\n",
    "\n",
    "    #---------------Loading in sequences for PAM Searching------------------#\n",
    "    \n",
    "    search_size = RTT_length - len(PAM)-1 #need to modify this for insertion/deletions...\n",
    "    #size mut doesn't capture the size of insertions, only the size of deletions\n",
    "\n",
    "    plus_search = seq1[seq_start-1-search_size : seq_end+search_size].upper()\n",
    "    minus_search = plus_search.complement().upper()\n",
    "    \n",
    "    mut_start_idx = 1+search_size\n",
    "    mut_end_idx = 1+search_size+size_mut #not accurate for insertions; does work for indexing though...\n",
    "\n",
    "    plus_search1 = plus_search[:mut_start_idx+3+len(PAM)-1]\n",
    "    minus_search1 = minus_search[mut_start_idx-3-len(PAM):]\n",
    "\n",
    "    #---------------PAM Searching------------------#\n",
    "\n",
    "    #replacing N with regex symbol\n",
    "    PAM_regex = PAM.replace('N', '/*.')\n",
    "\n",
    "    PAM_search_plus = re.compile('(?=(' + PAM_regex + '))', re.IGNORECASE)\n",
    "\n",
    "    iterator_plus = PAM_search_plus.finditer(str(plus_search1))\n",
    "    PAM_starts_plus = [match.start() for match in iterator_plus]\n",
    "\n",
    "\n",
    "    PAM_minus = PAM[::-1]#reversing it\n",
    "    PAM_regex_minus = PAM_minus.replace('N', '/*.')\n",
    "\n",
    "    PAM_search_minus = re.compile('(?=(' + PAM_regex_minus + '))', re.IGNORECASE)\n",
    "\n",
    "    iterator_minus = PAM_search_minus.finditer(str(minus_search1))\n",
    "    PAM_starts_minus = [match.start() for match in iterator_minus]\n",
    "    #since things are flipped on minus strand, adding len(PAM) to get the true \"start\" to the PAM\n",
    "\n",
    "    PAM_starts_minus = np.asarray(PAM_starts_minus) + len(PAM) + (mut_start_idx-3-len(PAM))#and correct for indexing\n",
    "\n",
    "    return np.asarray([np.asarray(PAM_starts_plus), PAM_starts_minus], dtype='object')-mut_start_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PAM_finder(WT_seq, mut_start, mut_end, PAM, RTT_length_max):\n",
    "    \"\"\"Identifies the location of PAM sequences on the + and - strand.\n",
    "    Returns a 2-d array containing marked locations of PAM sequence start locations on + and - strand.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    WT_seq: WT sequence...\n",
    "\n",
    "    PAM_seq: PAM sequence to search for...\n",
    "\n",
    "    RTT_length_max: maximum size of RTT length for searching...\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    #---------------Loading in sequences for PAM Searching------------------#\n",
    "    \n",
    "    search_size = RTT_length - len(PAM)-1 #need to modify this for insertion/deletions...\n",
    "    #size mut doesn't capture the size of insertions, only the size of deletions\n",
    "\n",
    "    plus_search = seq1[seq_start-1-search_size : seq_end+search_size].upper()\n",
    "    minus_search = plus_search.complement().upper()\n",
    "    \n",
    "    mut_start_idx = 1+search_size\n",
    "    mut_end_idx = 1+search_size+size_mut #not accurate for insertions; does work for indexing though...\n",
    "\n",
    "    plus_search1 = plus_search[:mut_start_idx+3+len(PAM)-1]\n",
    "    minus_search1 = minus_search[mut_start_idx-3-len(PAM):]\n",
    "\n",
    "    #---------------PAM Searching------------------#\n",
    "\n",
    "    #replacing N with regex symbol\n",
    "    PAM_regex = PAM.replace('N', '/*.')\n",
    "\n",
    "    PAM_search_plus = re.compile('(?=(' + PAM_regex + '))', re.IGNORECASE)\n",
    "\n",
    "    iterator_plus = PAM_search_plus.finditer(str(plus_search1))\n",
    "    PAM_starts_plus = [match.start() for match in iterator_plus]\n",
    "\n",
    "\n",
    "    PAM_minus = PAM[::-1]#reversing it\n",
    "    PAM_regex_minus = PAM_minus.replace('N', '/*.')\n",
    "\n",
    "    PAM_search_minus = re.compile('(?=(' + PAM_regex_minus + '))', re.IGNORECASE)\n",
    "\n",
    "    iterator_minus = PAM_search_minus.finditer(str(minus_search1))\n",
    "    PAM_starts_minus = [match.start() for match in iterator_minus]\n",
    "    #since things are flipped on minus strand, adding len(PAM) to get the true \"start\" to the PAM\n",
    "\n",
    "    PAM_starts_minus = np.asarray(PAM_starts_minus) + len(PAM) + (mut_start_idx-3-len(PAM))#and correct for indexing\n",
    "\n",
    "    return np.asarray([np.asarray(PAM_starts_plus), PAM_starts_minus], dtype='object')-mut_start_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PAM = 'NGG'\n",
    "\n",
    "PAM_finder(WT_seq, mut_start, mut_end, PAM, RTT_length_max)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating list of variants for alvin\n",
    "\n",
    "- Starting with diego library\n",
    "    - includes SNPs in IDR regions\n",
    "    - Also intronic SNPs between exons that are an IDR\n",
    "        - These wont be amenable to MMR-evasive silent edits\n",
    "\n",
    "- Need WT sequnece\n",
    "- Mutant sequence\n",
    "- Frame (i.e. codon frame = 0,1,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pegg\n",
    "import gffutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath = '/Users/samgould/Desktop/FSR Lab/reference files/GRCh37/ncbi-genomes-2022-03-17/GCF_000001405.25_GRCh37.p13_genomic.fna.gz'\n",
    "\n",
    "records, index_list = pegg.genome_loader(filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = '/Users/samgould/Desktop/FSR Lab/reference files/gencode_v19.db'\n",
    "db = gffutils.FeatureDB(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['COUNT', 'Hugo_Symbol', 'Entrez_Gene_Id', 'Center', 'NCBI_Build',\n",
       "       'Chromosome', 'Start_Position', 'End_Position', 'Strand', 'Consequence',\n",
       "       'Variant_Classification', 'Variant_Type', 'Reference_Allele',\n",
       "       'Tumor_Seq_Allele1', 'Tumor_Seq_Allele2', 'dbSNP_RS',\n",
       "       'dbSNP_Val_Status', 'Tumor_Sample_Barcode',\n",
       "       'Matched_Norm_Sample_Barcode', 'Match_Norm_Seq_Allele1',\n",
       "       'Match_Norm_Seq_Allele2', 'Tumor_Validation_Allele1',\n",
       "       'Tumor_Validation_Allele2', 'Match_Norm_Validation_Allele1',\n",
       "       'Match_Norm_Validation_Allele2', 'Verification_Status',\n",
       "       'Validation_Status', 'Mutation_Status', 'Sequencing_Phase',\n",
       "       'Sequence_Source', 'Validation_Method', 'Score', 'BAM_File',\n",
       "       'Sequencer', 't_ref_count', 't_alt_count', 'n_ref_count', 'n_alt_count',\n",
       "       'HGVSc', 'HGVSp', 'HGVSp_Short', 'Transcript_ID', 'RefSeq',\n",
       "       'Protein_position', 'Codons', 'Exon_Number', 'gnomAD_AF',\n",
       "       'gnomAD_AFR_AF', 'gnomAD_AMR_AF', 'gnomAD_ASJ_AF', 'gnomAD_EAS_AF',\n",
       "       'gnomAD_FIN_AF', 'gnomAD_NFE_AF', 'gnomAD_OTH_AF', 'gnomAD_SAS_AF',\n",
       "       'FILTER', 'Polyphen_Prediction', 'Polyphen_Score', 'SIFT_Prediction',\n",
       "       'SIFT_Score', 'SWISSPROT', 'n_depth', 't_depth', 'Annotation_Status',\n",
       "       'mutationInCis_Flag'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#diego variants\n",
    "idr = pd.read_csv('filtered_idr_mutations_5count_6nt_indels.csv')\n",
    "idr.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele1</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>dbSNP_RS</th>\n",
       "      <th>dbSNP_Val_Status</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Matched_Norm_Sample_Barcode</th>\n",
       "      <th>Match_Norm_Seq_Allele1</th>\n",
       "      <th>Match_Norm_Seq_Allele2</th>\n",
       "      <th>Tumor_Validation_Allele1</th>\n",
       "      <th>Tumor_Validation_Allele2</th>\n",
       "      <th>Match_Norm_Validation_Allele1</th>\n",
       "      <th>Match_Norm_Validation_Allele2</th>\n",
       "      <th>...</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>gnomAD_AF</th>\n",
       "      <th>gnomAD_AFR_AF</th>\n",
       "      <th>gnomAD_AMR_AF</th>\n",
       "      <th>gnomAD_ASJ_AF</th>\n",
       "      <th>gnomAD_EAS_AF</th>\n",
       "      <th>gnomAD_FIN_AF</th>\n",
       "      <th>gnomAD_NFE_AF</th>\n",
       "      <th>gnomAD_OTH_AF</th>\n",
       "      <th>gnomAD_SAS_AF</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>Polyphen_Prediction</th>\n",
       "      <th>Polyphen_Score</th>\n",
       "      <th>SIFT_Prediction</th>\n",
       "      <th>SIFT_Score</th>\n",
       "      <th>SWISSPROT</th>\n",
       "      <th>n_depth</th>\n",
       "      <th>t_depth</th>\n",
       "      <th>Annotation_Status</th>\n",
       "      <th>mutationInCis_Flag</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1468</th>\n",
       "      <td>12.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>324.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>5</td>\n",
       "      <td>112175760</td>\n",
       "      <td>112175761</td>\n",
       "      <td>+</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>AT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-JHU-01993-02421</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000257430</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1490.0</td>\n",
       "      <td>cAT/cT</td>\n",
       "      <td>16/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2449.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8082</th>\n",
       "      <td>5.0</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>22852.0</td>\n",
       "      <td>PROV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>10</td>\n",
       "      <td>27355404</td>\n",
       "      <td>27355409</td>\n",
       "      <td>+</td>\n",
       "      <td>splice_region_variant,intron_variant</td>\n",
       "      <td>Splice_Region</td>\n",
       "      <td>DEL</td>\n",
       "      <td>CATAGA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>rs386742210</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-PROV-163a364f5a-bb28ef10d7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000376087</td>\n",
       "      <td>NM_014915.2</td>\n",
       "      <td>423.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>384.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000376087.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8571</th>\n",
       "      <td>5.0</td>\n",
       "      <td>MYB</td>\n",
       "      <td>4602.0</td>\n",
       "      <td>DFCI</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>6</td>\n",
       "      <td>135539013</td>\n",
       "      <td>135539016</td>\n",
       "      <td>+</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>TACC</td>\n",
       "      <td>TACC</td>\n",
       "      <td>GA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-DFCI-234520-3238909</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000367814</td>\n",
       "      <td>NM_001161659.1</td>\n",
       "      <td>606.0</td>\n",
       "      <td>agTACC/agGA</td>\n",
       "      <td>15/15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>477.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367814.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8803</th>\n",
       "      <td>5.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>324.0</td>\n",
       "      <td>MSK</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>5</td>\n",
       "      <td>112177842</td>\n",
       "      <td>112177846</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>AAAGT</td>\n",
       "      <td>AAAGT</td>\n",
       "      <td>TA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-MSK-P-0067095-T05-IH3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000257430</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>2184.0</td>\n",
       "      <td>gAAAGT/gTA</td>\n",
       "      <td>16/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>649.0</td>\n",
       "      <td>882.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 66 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      COUNT Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  \\\n",
       "1468   12.0         APC           324.0    JHU     GRCh37          5   \n",
       "8082    5.0     ANKRD26         22852.0   PROV     GRCh37         10   \n",
       "8571    5.0         MYB          4602.0   DFCI     GRCh37          6   \n",
       "8803    5.0         APC           324.0    MSK     GRCh37          5   \n",
       "\n",
       "      Start_Position  End_Position Strand  \\\n",
       "1468       112175760     112175761      +   \n",
       "8082        27355404      27355409      +   \n",
       "8571       135539013     135539016      +   \n",
       "8803       112177842     112177846      +   \n",
       "\n",
       "                               Consequence Variant_Classification  \\\n",
       "1468                    frameshift_variant        Frame_Shift_Del   \n",
       "8082  splice_region_variant,intron_variant          Splice_Region   \n",
       "8571                    frameshift_variant        Frame_Shift_Del   \n",
       "8803              protein_altering_variant           In_Frame_Del   \n",
       "\n",
       "     Variant_Type Reference_Allele Tumor_Seq_Allele1 Tumor_Seq_Allele2  \\\n",
       "1468          DEL               AT               NaN                 T   \n",
       "8082          DEL           CATAGA               NaN                 T   \n",
       "8571          DEL             TACC              TACC                GA   \n",
       "8803          DEL            AAAGT             AAAGT                TA   \n",
       "\n",
       "         dbSNP_RS  dbSNP_Val_Status              Tumor_Sample_Barcode  \\\n",
       "1468          NaN               NaN             GENIE-JHU-01993-02421   \n",
       "8082  rs386742210               NaN  GENIE-PROV-163a364f5a-bb28ef10d7   \n",
       "8571          NaN               NaN         GENIE-DFCI-234520-3238909   \n",
       "8803          NaN               NaN       GENIE-MSK-P-0067095-T05-IH3   \n",
       "\n",
       "     Matched_Norm_Sample_Barcode Match_Norm_Seq_Allele1  \\\n",
       "1468                         NaN                    NaN   \n",
       "8082                         NaN                    NaN   \n",
       "8571                         NaN                    NaN   \n",
       "8803                         NaN                    NaN   \n",
       "\n",
       "     Match_Norm_Seq_Allele2  Tumor_Validation_Allele1  \\\n",
       "1468                    NaN                       NaN   \n",
       "8082                    NaN                       NaN   \n",
       "8571                    NaN                       NaN   \n",
       "8803                    NaN                       NaN   \n",
       "\n",
       "      Tumor_Validation_Allele2  Match_Norm_Validation_Allele1  \\\n",
       "1468                       NaN                            NaN   \n",
       "8082                       NaN                            NaN   \n",
       "8571                       NaN                            NaN   \n",
       "8803                       NaN                            NaN   \n",
       "\n",
       "      Match_Norm_Validation_Allele2  ...    Transcript_ID          RefSeq  \\\n",
       "1468                            NaN  ...  ENST00000257430     NM_000038.5   \n",
       "8082                            NaN  ...  ENST00000376087     NM_014915.2   \n",
       "8571                            NaN  ...  ENST00000367814  NM_001161659.1   \n",
       "8803                            NaN  ...  ENST00000257430     NM_000038.5   \n",
       "\n",
       "     Protein_position       Codons  Exon_Number  gnomAD_AF  gnomAD_AFR_AF  \\\n",
       "1468           1490.0       cAT/cT        16/16        NaN            NaN   \n",
       "8082            423.0          NaN          NaN        NaN            NaN   \n",
       "8571            606.0  agTACC/agGA        15/15        NaN            NaN   \n",
       "8803           2184.0   gAAAGT/gTA        16/16        NaN            NaN   \n",
       "\n",
       "      gnomAD_AMR_AF  gnomAD_ASJ_AF  gnomAD_EAS_AF  gnomAD_FIN_AF  \\\n",
       "1468            NaN            NaN            NaN            NaN   \n",
       "8082            NaN            NaN            NaN            NaN   \n",
       "8571            NaN            NaN            NaN            NaN   \n",
       "8803            NaN            NaN            NaN            NaN   \n",
       "\n",
       "      gnomAD_NFE_AF  gnomAD_OTH_AF gnomAD_SAS_AF FILTER Polyphen_Prediction  \\\n",
       "1468            NaN            NaN           NaN   PASS                 NaN   \n",
       "8082            NaN            NaN           NaN   PASS                 NaN   \n",
       "8571            NaN            NaN           NaN   PASS                 NaN   \n",
       "8803            NaN            NaN           NaN   PASS                 NaN   \n",
       "\n",
       "     Polyphen_Score SIFT_Prediction  SIFT_Score SWISSPROT n_depth  t_depth  \\\n",
       "1468            NaN             NaN         NaN       NaN     NaN   2449.0   \n",
       "8082            NaN             NaN         NaN       NaN     NaN    384.0   \n",
       "8571            NaN             NaN         NaN       NaN     NaN    477.0   \n",
       "8803            NaN             NaN         NaN       NaN   649.0    882.0   \n",
       "\n",
       "      Annotation_Status  mutationInCis_Flag  transcript_id_TRUE  \n",
       "1468            SUCCESS               False   ENST00000257430.4  \n",
       "8082            SUCCESS               False   ENST00000376087.4  \n",
       "8571            SUCCESS               False   ENST00000367814.4  \n",
       "8803            SUCCESS               False   ENST00000257430.4  \n",
       "\n",
       "[4 rows x 66 columns]"
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dels1 = idr[idr['Variant_Type']=='DEL']\n",
    "dels1[dels1['Tumor_Seq_Allele2']!='-']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele1</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>dbSNP_RS</th>\n",
       "      <th>dbSNP_Val_Status</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Matched_Norm_Sample_Barcode</th>\n",
       "      <th>Match_Norm_Seq_Allele1</th>\n",
       "      <th>Match_Norm_Seq_Allele2</th>\n",
       "      <th>Tumor_Validation_Allele1</th>\n",
       "      <th>Tumor_Validation_Allele2</th>\n",
       "      <th>Match_Norm_Validation_Allele1</th>\n",
       "      <th>Match_Norm_Validation_Allele2</th>\n",
       "      <th>...</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>gnomAD_AF</th>\n",
       "      <th>gnomAD_AFR_AF</th>\n",
       "      <th>gnomAD_AMR_AF</th>\n",
       "      <th>gnomAD_ASJ_AF</th>\n",
       "      <th>gnomAD_EAS_AF</th>\n",
       "      <th>gnomAD_FIN_AF</th>\n",
       "      <th>gnomAD_NFE_AF</th>\n",
       "      <th>gnomAD_OTH_AF</th>\n",
       "      <th>gnomAD_SAS_AF</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>Polyphen_Prediction</th>\n",
       "      <th>Polyphen_Score</th>\n",
       "      <th>SIFT_Prediction</th>\n",
       "      <th>SIFT_Score</th>\n",
       "      <th>SWISSPROT</th>\n",
       "      <th>n_depth</th>\n",
       "      <th>t_depth</th>\n",
       "      <th>Annotation_Status</th>\n",
       "      <th>mutationInCis_Flag</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7729</th>\n",
       "      <td>5.0</td>\n",
       "      <td>XBP1</td>\n",
       "      <td>7494.0</td>\n",
       "      <td>MSK</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>22</td>\n",
       "      <td>29192182</td>\n",
       "      <td>29192183</td>\n",
       "      <td>+</td>\n",
       "      <td>splice_acceptor_variant,intron_variant</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>INS</td>\n",
       "      <td>TG</td>\n",
       "      <td>TG</td>\n",
       "      <td>GTGGGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENIE-MSK-P-0015800-T11-IH3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000216037</td>\n",
       "      <td>NM_005080.3</td>\n",
       "      <td>152.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>834.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000216037.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 66 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      COUNT Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  \\\n",
       "7729    5.0        XBP1          7494.0    MSK     GRCh37         22   \n",
       "\n",
       "      Start_Position  End_Position Strand  \\\n",
       "7729        29192182      29192183      +   \n",
       "\n",
       "                                 Consequence Variant_Classification  \\\n",
       "7729  splice_acceptor_variant,intron_variant            Splice_Site   \n",
       "\n",
       "     Variant_Type Reference_Allele Tumor_Seq_Allele1 Tumor_Seq_Allele2  \\\n",
       "7729          INS               TG                TG            GTGGGG   \n",
       "\n",
       "     dbSNP_RS  dbSNP_Val_Status         Tumor_Sample_Barcode  \\\n",
       "7729      NaN               NaN  GENIE-MSK-P-0015800-T11-IH3   \n",
       "\n",
       "     Matched_Norm_Sample_Barcode Match_Norm_Seq_Allele1  \\\n",
       "7729                         NaN                    NaN   \n",
       "\n",
       "     Match_Norm_Seq_Allele2  Tumor_Validation_Allele1  \\\n",
       "7729                    NaN                       NaN   \n",
       "\n",
       "      Tumor_Validation_Allele2  Match_Norm_Validation_Allele1  \\\n",
       "7729                       NaN                            NaN   \n",
       "\n",
       "      Match_Norm_Validation_Allele2  ...    Transcript_ID       RefSeq  \\\n",
       "7729                            NaN  ...  ENST00000216037  NM_005080.3   \n",
       "\n",
       "     Protein_position  Codons  Exon_Number  gnomAD_AF  gnomAD_AFR_AF  \\\n",
       "7729            152.0     NaN          NaN        NaN            NaN   \n",
       "\n",
       "      gnomAD_AMR_AF  gnomAD_ASJ_AF  gnomAD_EAS_AF  gnomAD_FIN_AF  \\\n",
       "7729            NaN            NaN            NaN            NaN   \n",
       "\n",
       "      gnomAD_NFE_AF  gnomAD_OTH_AF gnomAD_SAS_AF FILTER Polyphen_Prediction  \\\n",
       "7729            NaN            NaN           NaN   PASS                 NaN   \n",
       "\n",
       "     Polyphen_Score SIFT_Prediction  SIFT_Score SWISSPROT n_depth  t_depth  \\\n",
       "7729            NaN             NaN         NaN       NaN     0.0    834.0   \n",
       "\n",
       "      Annotation_Status  mutationInCis_Flag  transcript_id_TRUE  \n",
       "7729            SUCCESS               False   ENST00000216037.6  \n",
       "\n",
       "[1 rows x 66 columns]"
      ]
     },
     "execution_count": 264,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ins1 = idr[idr['Variant_Type']=='INS']\n",
    "ins1[ins1['Reference_Allele']!='-']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['DEL', 'DNP', 'INS', 'ONP', 'SNP'], dtype=object),\n",
       " array([ 729,   42,  361,   13, 7955]))"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(idr['Variant_Type'], return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "tx_new = []\n",
    "for i, val in idr.iterrows():\n",
    "    h = val['HGVSc']\n",
    "    tx_new.append(h.split(':')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "idr['transcript_id_TRUE'] = tx_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#check all transcripts are annotated\n",
    "\n",
    "t_ids = np.unique(idr['transcript_id_TRUE'])\n",
    "\n",
    "not_found = []\n",
    "\n",
    "for tx in t_ids:\n",
    "\n",
    "    cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "    start_end_cds = [[i.start, i.end] for i in cds]\n",
    "    if len(start_end_cds)==0:\n",
    "        not_found.append(tx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[20146427, 20146429],\n",
       " [20148634, 20148725],\n",
       " [20150300, 20150381],\n",
       " [20152075, 20152125],\n",
       " [20153856, 20153959],\n",
       " [20156657, 20156740],\n",
       " [20159743, 20159758]]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tx = 'ENST00000379607.5'\n",
    "\n",
    "cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "start_end_cds = [[i.start, i.end] for i in cds]\n",
    "start_end_cds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "ins = idr[idr['Variant_Type']=='INS']\n",
    "dels = idr[idr['Variant_Type']=='DEL']\n",
    "subs = idr[idr['Variant_Type'].isin(['SNP', 'DNP', 'ONP'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def df_formatter(df, context_size = 60):\n",
    "\n",
    "    wt_w_context = []\n",
    "    alt_w_context = []\n",
    "\n",
    "    seq_start = []\n",
    "    seq_end = []\n",
    "\n",
    "    for i, val in df.iterrows():\n",
    "        vt = val['Variant_Type']\n",
    "        s = val['Start_Position']\n",
    "        e = val['End_Position']\n",
    "        ref = val['Reference_Allele']\n",
    "        alt = val['Tumor_Seq_Allele2']\n",
    "        chrom = val['Chromosome']\n",
    "\n",
    "        if chrom == 'X':\n",
    "            chrom = 22\n",
    "        elif chrom=='Y':\n",
    "            chrom = 23\n",
    "        else:\n",
    "            chrom = int(chrom)-1\n",
    "\n",
    "        chr_seq = records[index_list[chrom]].seq.upper()\n",
    "\n",
    "        if vt in ['SNP', 'ONP', 'DNP']:\n",
    "            ref = ref\n",
    "            alt = alt\n",
    "            #assert ref == chr_seq[s-1:e], print(ref, chr_seq[s-1:e])\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        elif vt =='INS':\n",
    "            ref = ''\n",
    "            alt = alt\n",
    "            left_context = chr_seq[s-1-context_size:s+1] #need to do this since INS reference alleles are blank\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        elif vt=='DEL':\n",
    "            ref = ref\n",
    "            alt = ''\n",
    "\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        wt_seq = left_context + ref + right_context\n",
    "        alt_seq = left_context + alt + right_context\n",
    "\n",
    "        wt_w_context.append(str(wt_seq))\n",
    "        alt_w_context.append(str(alt_seq))\n",
    "\n",
    "        start = s-context_size\n",
    "        end = e+context_size\n",
    "\n",
    "        seq_start.append(start)\n",
    "        seq_end.append(end)\n",
    "\n",
    "        assert str(chr_seq[start-1:end])==str(wt_seq)\n",
    "\n",
    "    cols_to_save = ['COUNT', 'Hugo_Symbol', 'Chromosome', 'Start_Position', 'End_Position', 'Consequence', 'Variant_Classification', 'Variant_Type', 'Reference_Allele', 'Tumor_Seq_Allele2', 'Tumor_Sample_Barcode', 'RefSeq', 'Protein_position', 'Exon_Number', 'HGVSc', 'HGVSp', 'HGVSp_Short', 'transcript_id_TRUE']\n",
    "    df_new = df[cols_to_save]\n",
    "\n",
    "    df_new['seq_start'] = seq_start\n",
    "    df_new['seq_end'] = seq_end\n",
    "    df_new['wt_w_context'] = wt_w_context\n",
    "    df_new['alt_w_context'] = alt_w_context\n",
    "    df_new = df_new.reset_index()\n",
    "    \n",
    "    return df_new\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [],
   "source": [
    "def frame_determiner(exon_subs):\n",
    "    \n",
    "    frame_list = []\n",
    "    coding_start_list = []\n",
    "    coding_end_list = []\n",
    "    CDS_WT_correct_orientation = []\n",
    "\n",
    "    for index1, val in exon_subs.iterrows():\n",
    "        hg = val['HGVSp_Short']\n",
    "\n",
    "        tx = val['transcript_id_TRUE']\n",
    "        cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "        start_end_cds = [[i.start, i.end] for i in cds]\n",
    "        strand = db[tx].strand\n",
    "        #print(strand)\n",
    "        chrom = val['Chromosome']\n",
    "\n",
    "        if chrom == 'X':\n",
    "            chrom = 22\n",
    "        elif chrom=='Y':\n",
    "            chrom = 23\n",
    "        else:\n",
    "            chrom = int(chrom)-1\n",
    "\n",
    "        chr_seq = records[index_list[chrom]].seq.upper()\n",
    "\n",
    "\n",
    "        wt_dna = ''\n",
    "        codon_locs = []\n",
    "        for i in start_end_cds:\n",
    "            wt_dna += chr_seq[i[0]-1:i[1]]\n",
    "\n",
    "            for j in range(i[0], i[1]+1):\n",
    "                codon_locs.append(j)\n",
    "\n",
    "        #and add in the last/stop codon\n",
    "        #if strand == '+':\n",
    "\n",
    "        # wt_dna += chr_seq[start_end_cds[-1][1]:start_end_cds[-1][1]+3]\n",
    "        #elif strand == '-':\n",
    "        #   wt_dna = chr_seq[start_end_cds[0][0]-3:start_end_cds[0][0]] +  wt_dna\n",
    "\n",
    "        #check location in cds\n",
    "        #get information about the wt/alt sequence\n",
    "        start = val['seq_start']\n",
    "        end = val['seq_end']\n",
    "        wt_seq = val['wt_w_context']\n",
    "        alt_seq = val['alt_w_context']\n",
    "\n",
    "        inc = []\n",
    "        dna_dict = dict(zip(list(range(start, end+1)), list(range(len(wt_seq)))))\n",
    "        for i in range(start, end+1):\n",
    "            if i in codon_locs:\n",
    "                inc.append(i)\n",
    "\n",
    "        if len(inc)==0:\n",
    "            frame = 'None'\n",
    "            frame_list.append(frame)\n",
    "            coding_start_list.append('None')\n",
    "            coding_end_list.append('None')\n",
    "            CDS_WT_correct_orientation.append('None')\n",
    "            #means it's not in a coding region (could be splice or intron)\n",
    "        \n",
    "        else:\n",
    "            inc_start = inc[0]\n",
    "            inc_end = inc[-1]\n",
    "\n",
    "            #------and then double check that the frame is correct--------------\n",
    "            #by checking if the subsequence is located in the WT sequence\n",
    "\n",
    "            wt_start = dna_dict[inc_start]\n",
    "            wt_end = dna_dict[inc_end] + 1\n",
    "\n",
    "            #record what part of the subsequence is part of a CDS\n",
    "            coding_start_list.append(wt_start)\n",
    "            coding_end_list.append(wt_end)\n",
    "\n",
    "            full_prot = str(Bio.Seq.Seq(wt_dna).transcribe().translate())\n",
    "\n",
    "            subseq = Bio.Seq.Seq(wt_seq[wt_start:wt_end])\n",
    "\n",
    "            if strand =='-':\n",
    "                subseq = subseq.reverse_complement()\n",
    "                inc_start = inc[-1]\n",
    "                inc_end = inc[0]\n",
    "                full_prot = str(Bio.Seq.Seq(wt_dna).reverse_complement().transcribe().translate())\n",
    "                #have to reverse this list as well if the transcript is in the - orientation\n",
    "                codon_locs = codon_locs[::-1]\n",
    "\n",
    "            CDS_WT_correct_orientation.append(str(subseq))\n",
    "\n",
    "            #determine the frame\n",
    "            start_idx = codon_locs.index(inc_start)\n",
    "\n",
    "            frame = (3 - (start_idx % 3)) % 3\n",
    "            frame_list.append(frame)\n",
    "            \n",
    "            #and then confirm whether the subsequence is in the protein\n",
    "            sub1 = str(Bio.Seq.Seq(subseq[frame:]).transcribe().translate())\n",
    "            \n",
    "            #assert sub1 in full_prot, print(sub1)\n",
    "            if sub1 not in full_prot:\n",
    "                print('error')\n",
    "                print(strand)\n",
    "                print(frame)\n",
    "            \n",
    "    \n",
    "    exon_subs['Frame'] = frame_list\n",
    "    exon_subs['CDS_start'] = coding_start_list\n",
    "    exon_subs['CDS_end'] = coding_end_list\n",
    "    exon_subs['CDS_WT_correct_orientation'] = CDS_WT_correct_orientation\n",
    "\n",
    "    return exon_subs, codon_locs, start_idx, wt_seq, subseq\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(3 - (11%3)) %3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level_0</th>\n",
       "      <th>index</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1512.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-JHU-00113-00335</td>\n",
       "      <td>NM_001126112.2</td>\n",
       "      <td>282.0</td>\n",
       "      <td>8/11</td>\n",
       "      <td>ENST00000269305.4:c.844C&gt;T</td>\n",
       "      <td>p.Arg282Trp</td>\n",
       "      <td>p.R282W</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577034</td>\n",
       "      <td>7577154</td>\n",
       "      <td>CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...</td>\n",
       "      <td>CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>777.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175639</td>\n",
       "      <td>112175639</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00223-00468</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4348C&gt;T</td>\n",
       "      <td>p.Arg1450Ter</td>\n",
       "      <td>p.R1450*</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175579</td>\n",
       "      <td>112175699</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>614.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112173917</td>\n",
       "      <td>112173917</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00123-00365</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>876.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.2626C&gt;T</td>\n",
       "      <td>p.Arg876Ter</td>\n",
       "      <td>p.R876*</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112173857</td>\n",
       "      <td>112173977</td>\n",
       "      <td>CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...</td>\n",
       "      <td>CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>538.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577085</td>\n",
       "      <td>7577085</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00694-00783</td>\n",
       "      <td>NM_001126112.2</td>\n",
       "      <td>285.0</td>\n",
       "      <td>8/11</td>\n",
       "      <td>ENST00000269305.4:c.853G&gt;A</td>\n",
       "      <td>p.Glu285Lys</td>\n",
       "      <td>p.E285K</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577025</td>\n",
       "      <td>7577145</td>\n",
       "      <td>TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...</td>\n",
       "      <td>TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>527.0</td>\n",
       "      <td>SRSF2</td>\n",
       "      <td>17</td>\n",
       "      <td>74732959</td>\n",
       "      <td>74732959</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-DFCI-001379-11338</td>\n",
       "      <td>NM_001195427.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>1/3</td>\n",
       "      <td>ENST00000359995.5:c.284C&gt;A</td>\n",
       "      <td>p.Pro95His</td>\n",
       "      <td>p.P95H</td>\n",
       "      <td>ENST00000359995.5</td>\n",
       "      <td>74732899</td>\n",
       "      <td>74733019</td>\n",
       "      <td>TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...</td>\n",
       "      <td>TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>394</td>\n",
       "      <td>602</td>\n",
       "      <td>19.0</td>\n",
       "      <td>SETD2</td>\n",
       "      <td>3</td>\n",
       "      <td>47164832</td>\n",
       "      <td>47164832</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-DFCI-008656-6895</td>\n",
       "      <td>NM_014159.6</td>\n",
       "      <td>432.0</td>\n",
       "      <td>3/21</td>\n",
       "      <td>ENST00000409792.3:c.1294C&gt;T</td>\n",
       "      <td>p.Arg432Cys</td>\n",
       "      <td>p.R432C</td>\n",
       "      <td>ENST00000409792.3</td>\n",
       "      <td>47164772</td>\n",
       "      <td>47164892</td>\n",
       "      <td>CTGTGTATGGCCGAGAATAGCGCGTCCTCTCTCGATAAGGGGAGCT...</td>\n",
       "      <td>CTGTGTATGGCCGAGAATAGCGCGTCCTCTCTCGATAAGGGGAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>395</td>\n",
       "      <td>603</td>\n",
       "      <td>19.0</td>\n",
       "      <td>ETV5</td>\n",
       "      <td>3</td>\n",
       "      <td>185797801</td>\n",
       "      <td>185797801</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-DFCI-025211-61496</td>\n",
       "      <td>NM_004454.2</td>\n",
       "      <td>152.0</td>\n",
       "      <td>7/13</td>\n",
       "      <td>ENST00000306376.5:c.455C&gt;A</td>\n",
       "      <td>p.Pro152Gln</td>\n",
       "      <td>p.P152Q</td>\n",
       "      <td>ENST00000306376.5</td>\n",
       "      <td>185797741</td>\n",
       "      <td>185797861</td>\n",
       "      <td>ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...</td>\n",
       "      <td>ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>396</td>\n",
       "      <td>604</td>\n",
       "      <td>19.0</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>1</td>\n",
       "      <td>11190747</td>\n",
       "      <td>11190747</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-DFCI-007654-6083</td>\n",
       "      <td>NM_004958.3</td>\n",
       "      <td>1818.0</td>\n",
       "      <td>39/58</td>\n",
       "      <td>ENST00000361445.4:c.5452C&gt;T</td>\n",
       "      <td>p.Arg1818Cys</td>\n",
       "      <td>p.R1818C</td>\n",
       "      <td>ENST00000361445.4</td>\n",
       "      <td>11190687</td>\n",
       "      <td>11190807</td>\n",
       "      <td>CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...</td>\n",
       "      <td>CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>398</td>\n",
       "      <td>607</td>\n",
       "      <td>19.0</td>\n",
       "      <td>EIF1AX</td>\n",
       "      <td>X</td>\n",
       "      <td>20156720</td>\n",
       "      <td>20156720</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-MSK-P-0044987-T01-IM6</td>\n",
       "      <td>NM_001412.3</td>\n",
       "      <td>13.0</td>\n",
       "      <td>2/7</td>\n",
       "      <td>ENST00000379607.5:c.37C&gt;T</td>\n",
       "      <td>p.Arg13Cys</td>\n",
       "      <td>p.R13C</td>\n",
       "      <td>ENST00000379607.5</td>\n",
       "      <td>20156660</td>\n",
       "      <td>20156780</td>\n",
       "      <td>GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...</td>\n",
       "      <td>GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>399</td>\n",
       "      <td>608</td>\n",
       "      <td>19.0</td>\n",
       "      <td>PLEKHG5</td>\n",
       "      <td>1</td>\n",
       "      <td>6528337</td>\n",
       "      <td>6528337</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-UHN-394602-ARC2</td>\n",
       "      <td>NM_001042663.1</td>\n",
       "      <td>909.0</td>\n",
       "      <td>21/22</td>\n",
       "      <td>ENST00000400915.3:c.2727A&gt;C</td>\n",
       "      <td>p.Pro909=</td>\n",
       "      <td>p.P909=</td>\n",
       "      <td>ENST00000400915.3</td>\n",
       "      <td>6528277</td>\n",
       "      <td>6528397</td>\n",
       "      <td>CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...</td>\n",
       "      <td>CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>356 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     level_0  index   COUNT Hugo_Symbol Chromosome  Start_Position  \\\n",
       "0          0      1  1512.0        TP53         17         7577094   \n",
       "1          1      3   777.0         APC          5       112175639   \n",
       "2          2      6   614.0         APC          5       112173917   \n",
       "3          3      7   538.0        TP53         17         7577085   \n",
       "4          4      8   527.0       SRSF2         17        74732959   \n",
       "..       ...    ...     ...         ...        ...             ...   \n",
       "351      394    602    19.0       SETD2          3        47164832   \n",
       "352      395    603    19.0        ETV5          3       185797801   \n",
       "353      396    604    19.0        MTOR          1        11190747   \n",
       "354      398    607    19.0      EIF1AX          X        20156720   \n",
       "355      399    608    19.0     PLEKHG5          1         6528337   \n",
       "\n",
       "     End_Position         Consequence Variant_Classification Variant_Type  \\\n",
       "0         7577094    missense_variant      Missense_Mutation          SNP   \n",
       "1       112175639         stop_gained      Nonsense_Mutation          SNP   \n",
       "2       112173917         stop_gained      Nonsense_Mutation          SNP   \n",
       "3         7577085    missense_variant      Missense_Mutation          SNP   \n",
       "4        74732959    missense_variant      Missense_Mutation          SNP   \n",
       "..            ...                 ...                    ...          ...   \n",
       "351      47164832    missense_variant      Missense_Mutation          SNP   \n",
       "352     185797801    missense_variant      Missense_Mutation          SNP   \n",
       "353      11190747    missense_variant      Missense_Mutation          SNP   \n",
       "354      20156720    missense_variant      Missense_Mutation          SNP   \n",
       "355       6528337  synonymous_variant                 Silent          SNP   \n",
       "\n",
       "    Reference_Allele Tumor_Seq_Allele2         Tumor_Sample_Barcode  \\\n",
       "0                  G                 A        GENIE-JHU-00113-00335   \n",
       "1                  C                 T        GENIE-JHU-00223-00468   \n",
       "2                  C                 T        GENIE-JHU-00123-00365   \n",
       "3                  C                 T        GENIE-JHU-00694-00783   \n",
       "4                  G                 T      GENIE-DFCI-001379-11338   \n",
       "..               ...               ...                          ...   \n",
       "351                G                 A       GENIE-DFCI-008656-6895   \n",
       "352                G                 T      GENIE-DFCI-025211-61496   \n",
       "353                G                 A       GENIE-DFCI-007654-6083   \n",
       "354                G                 A  GENIE-MSK-P-0044987-T01-IM6   \n",
       "355                T                 G        GENIE-UHN-394602-ARC2   \n",
       "\n",
       "             RefSeq  Protein_position Exon_Number  \\\n",
       "0    NM_001126112.2             282.0        8/11   \n",
       "1       NM_000038.5            1450.0       16/16   \n",
       "2       NM_000038.5             876.0       16/16   \n",
       "3    NM_001126112.2             285.0        8/11   \n",
       "4    NM_001195427.1              95.0         1/3   \n",
       "..              ...               ...         ...   \n",
       "351     NM_014159.6             432.0        3/21   \n",
       "352     NM_004454.2             152.0        7/13   \n",
       "353     NM_004958.3            1818.0       39/58   \n",
       "354     NM_001412.3              13.0         2/7   \n",
       "355  NM_001042663.1             909.0       21/22   \n",
       "\n",
       "                           HGVSc         HGVSp HGVSp_Short transcript_id_TRUE  \\\n",
       "0     ENST00000269305.4:c.844C>T   p.Arg282Trp     p.R282W  ENST00000269305.4   \n",
       "1    ENST00000257430.4:c.4348C>T  p.Arg1450Ter    p.R1450*  ENST00000257430.4   \n",
       "2    ENST00000257430.4:c.2626C>T   p.Arg876Ter     p.R876*  ENST00000257430.4   \n",
       "3     ENST00000269305.4:c.853G>A   p.Glu285Lys     p.E285K  ENST00000269305.4   \n",
       "4     ENST00000359995.5:c.284C>A    p.Pro95His      p.P95H  ENST00000359995.5   \n",
       "..                           ...           ...         ...                ...   \n",
       "351  ENST00000409792.3:c.1294C>T   p.Arg432Cys     p.R432C  ENST00000409792.3   \n",
       "352   ENST00000306376.5:c.455C>A   p.Pro152Gln     p.P152Q  ENST00000306376.5   \n",
       "353  ENST00000361445.4:c.5452C>T  p.Arg1818Cys    p.R1818C  ENST00000361445.4   \n",
       "354    ENST00000379607.5:c.37C>T    p.Arg13Cys      p.R13C  ENST00000379607.5   \n",
       "355  ENST00000400915.3:c.2727A>C     p.Pro909=     p.P909=  ENST00000400915.3   \n",
       "\n",
       "     seq_start    seq_end                                       wt_w_context  \\\n",
       "0      7577034    7577154  CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...   \n",
       "1    112175579  112175699  ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...   \n",
       "2    112173857  112173977  CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...   \n",
       "3      7577025    7577145  TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...   \n",
       "4     74732899   74733019  TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...   \n",
       "..         ...        ...                                                ...   \n",
       "351   47164772   47164892  CTGTGTATGGCCGAGAATAGCGCGTCCTCTCTCGATAAGGGGAGCT...   \n",
       "352  185797741  185797861  ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...   \n",
       "353   11190687   11190807  CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...   \n",
       "354   20156660   20156780  GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...   \n",
       "355    6528277    6528397  CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...   \n",
       "\n",
       "                                         alt_w_context  \n",
       "0    CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...  \n",
       "1    ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...  \n",
       "2    CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...  \n",
       "3    TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...  \n",
       "4    TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...  \n",
       "..                                                 ...  \n",
       "351  CTGTGTATGGCCGAGAATAGCGCGTCCTCTCTCGATAAGGGGAGCT...  \n",
       "352  ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...  \n",
       "353  CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...  \n",
       "354  GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...  \n",
       "355  CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...  \n",
       "\n",
       "[356 rows x 24 columns]"
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_subs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [],
   "source": [
    "e, codon_locs, start_idx, wt_seq, subseq = frame_determiner(exon_subs[:])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level_0</th>\n",
       "      <th>index</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "      <th>Frame</th>\n",
       "      <th>CDS_start</th>\n",
       "      <th>CDS_end</th>\n",
       "      <th>CDS_WT_correct_orientation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1512.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-JHU-00113-00335</td>\n",
       "      <td>NM_001126112.2</td>\n",
       "      <td>282.0</td>\n",
       "      <td>8/11</td>\n",
       "      <td>ENST00000269305.4:c.844C&gt;T</td>\n",
       "      <td>p.Arg282Trp</td>\n",
       "      <td>p.R282W</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577034</td>\n",
       "      <td>7577154</td>\n",
       "      <td>CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...</td>\n",
       "      <td>CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>GGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>777.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175639</td>\n",
       "      <td>112175639</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00223-00468</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4348C&gt;T</td>\n",
       "      <td>p.Arg1450Ter</td>\n",
       "      <td>p.R1450*</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175579</td>\n",
       "      <td>112175699</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>614.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112173917</td>\n",
       "      <td>112173917</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00123-00365</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>876.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.2626C&gt;T</td>\n",
       "      <td>p.Arg876Ter</td>\n",
       "      <td>p.R876*</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112173857</td>\n",
       "      <td>112173977</td>\n",
       "      <td>CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...</td>\n",
       "      <td>CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>538.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577085</td>\n",
       "      <td>7577085</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00694-00783</td>\n",
       "      <td>NM_001126112.2</td>\n",
       "      <td>285.0</td>\n",
       "      <td>8/11</td>\n",
       "      <td>ENST00000269305.4:c.853G&gt;A</td>\n",
       "      <td>p.Glu285Lys</td>\n",
       "      <td>p.E285K</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577025</td>\n",
       "      <td>7577145</td>\n",
       "      <td>TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...</td>\n",
       "      <td>TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>CTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>527.0</td>\n",
       "      <td>SRSF2</td>\n",
       "      <td>17</td>\n",
       "      <td>74732959</td>\n",
       "      <td>74732959</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-DFCI-001379-11338</td>\n",
       "      <td>NM_001195427.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>1/3</td>\n",
       "      <td>ENST00000359995.5:c.284C&gt;A</td>\n",
       "      <td>p.Pro95His</td>\n",
       "      <td>p.P95H</td>\n",
       "      <td>ENST00000359995.5</td>\n",
       "      <td>74732899</td>\n",
       "      <td>74733019</td>\n",
       "      <td>TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...</td>\n",
       "      <td>TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>TGGACGGGGCCGTGCTGGACGGCCGCGAGCTGCGGGTGCAAATGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>394</td>\n",
       "      <td>602</td>\n",
       "      <td>19.0</td>\n",
       "      <td>SETD2</td>\n",
       "      <td>3</td>\n",
       "      <td>47164832</td>\n",
       "      <td>47164832</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-DFCI-008656-6895</td>\n",
       "      <td>NM_014159.6</td>\n",
       "      <td>432.0</td>\n",
       "      <td>3/21</td>\n",
       "      <td>ENST00000409792.3:c.1294C&gt;T</td>\n",
       "      <td>p.Arg432Cys</td>\n",
       "      <td>p.R432C</td>\n",
       "      <td>ENST00000409792.3</td>\n",
       "      <td>47164772</td>\n",
       "      <td>47164892</td>\n",
       "      <td>CTGTGTATGGCCGAGAATAGCGCGTCCTCTCTCGATAAGGGGAGCT...</td>\n",
       "      <td>CTGTGTATGGCCGAGAATAGCGCGTCCTCTCTCGATAAGGGGAGCT...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>GGCTCTAGAACTAATTTATCCTATTCCAGGTCAGAACGATCTCATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>395</td>\n",
       "      <td>603</td>\n",
       "      <td>19.0</td>\n",
       "      <td>ETV5</td>\n",
       "      <td>3</td>\n",
       "      <td>185797801</td>\n",
       "      <td>185797801</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-DFCI-025211-61496</td>\n",
       "      <td>NM_004454.2</td>\n",
       "      <td>152.0</td>\n",
       "      <td>7/13</td>\n",
       "      <td>ENST00000306376.5:c.455C&gt;A</td>\n",
       "      <td>p.Pro152Gln</td>\n",
       "      <td>p.P152Q</td>\n",
       "      <td>ENST00000306376.5</td>\n",
       "      <td>185797741</td>\n",
       "      <td>185797861</td>\n",
       "      <td>ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...</td>\n",
       "      <td>ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>AGCCATTAACCCCTCCTACAACCCCCCTCTCACCCACCCATCAGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>396</td>\n",
       "      <td>604</td>\n",
       "      <td>19.0</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>1</td>\n",
       "      <td>11190747</td>\n",
       "      <td>11190747</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-DFCI-007654-6083</td>\n",
       "      <td>NM_004958.3</td>\n",
       "      <td>1818.0</td>\n",
       "      <td>39/58</td>\n",
       "      <td>ENST00000361445.4:c.5452C&gt;T</td>\n",
       "      <td>p.Arg1818Cys</td>\n",
       "      <td>p.R1818C</td>\n",
       "      <td>ENST00000361445.4</td>\n",
       "      <td>11190687</td>\n",
       "      <td>11190807</td>\n",
       "      <td>CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...</td>\n",
       "      <td>CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>TTCGAAGCTGTGCTACACTACAAACATCAGAACCAAGCCCGCGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>398</td>\n",
       "      <td>607</td>\n",
       "      <td>19.0</td>\n",
       "      <td>EIF1AX</td>\n",
       "      <td>X</td>\n",
       "      <td>20156720</td>\n",
       "      <td>20156720</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-MSK-P-0044987-T01-IM6</td>\n",
       "      <td>NM_001412.3</td>\n",
       "      <td>13.0</td>\n",
       "      <td>2/7</td>\n",
       "      <td>ENST00000379607.5:c.37C&gt;T</td>\n",
       "      <td>p.Arg13Cys</td>\n",
       "      <td>p.R13C</td>\n",
       "      <td>ENST00000379607.5</td>\n",
       "      <td>20156660</td>\n",
       "      <td>20156780</td>\n",
       "      <td>GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...</td>\n",
       "      <td>GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>GTAAAGGAGGTAAAAACAGACGCAGGGGTAAGAATGAGAATGAATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>399</td>\n",
       "      <td>608</td>\n",
       "      <td>19.0</td>\n",
       "      <td>PLEKHG5</td>\n",
       "      <td>1</td>\n",
       "      <td>6528337</td>\n",
       "      <td>6528337</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-UHN-394602-ARC2</td>\n",
       "      <td>NM_001042663.1</td>\n",
       "      <td>909.0</td>\n",
       "      <td>21/22</td>\n",
       "      <td>ENST00000400915.3:c.2727A&gt;C</td>\n",
       "      <td>p.Pro909=</td>\n",
       "      <td>p.P909=</td>\n",
       "      <td>ENST00000400915.3</td>\n",
       "      <td>6528277</td>\n",
       "      <td>6528397</td>\n",
       "      <td>CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...</td>\n",
       "      <td>CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>AGGCCCAATGGCAGAGCTAGTGCCTCGGGCCCCAGAGTCCCCACGA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>356 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     level_0  index   COUNT Hugo_Symbol Chromosome  Start_Position  \\\n",
       "0          0      1  1512.0        TP53         17         7577094   \n",
       "1          1      3   777.0         APC          5       112175639   \n",
       "2          2      6   614.0         APC          5       112173917   \n",
       "3          3      7   538.0        TP53         17         7577085   \n",
       "4          4      8   527.0       SRSF2         17        74732959   \n",
       "..       ...    ...     ...         ...        ...             ...   \n",
       "351      394    602    19.0       SETD2          3        47164832   \n",
       "352      395    603    19.0        ETV5          3       185797801   \n",
       "353      396    604    19.0        MTOR          1        11190747   \n",
       "354      398    607    19.0      EIF1AX          X        20156720   \n",
       "355      399    608    19.0     PLEKHG5          1         6528337   \n",
       "\n",
       "     End_Position         Consequence Variant_Classification Variant_Type  \\\n",
       "0         7577094    missense_variant      Missense_Mutation          SNP   \n",
       "1       112175639         stop_gained      Nonsense_Mutation          SNP   \n",
       "2       112173917         stop_gained      Nonsense_Mutation          SNP   \n",
       "3         7577085    missense_variant      Missense_Mutation          SNP   \n",
       "4        74732959    missense_variant      Missense_Mutation          SNP   \n",
       "..            ...                 ...                    ...          ...   \n",
       "351      47164832    missense_variant      Missense_Mutation          SNP   \n",
       "352     185797801    missense_variant      Missense_Mutation          SNP   \n",
       "353      11190747    missense_variant      Missense_Mutation          SNP   \n",
       "354      20156720    missense_variant      Missense_Mutation          SNP   \n",
       "355       6528337  synonymous_variant                 Silent          SNP   \n",
       "\n",
       "    Reference_Allele Tumor_Seq_Allele2         Tumor_Sample_Barcode  \\\n",
       "0                  G                 A        GENIE-JHU-00113-00335   \n",
       "1                  C                 T        GENIE-JHU-00223-00468   \n",
       "2                  C                 T        GENIE-JHU-00123-00365   \n",
       "3                  C                 T        GENIE-JHU-00694-00783   \n",
       "4                  G                 T      GENIE-DFCI-001379-11338   \n",
       "..               ...               ...                          ...   \n",
       "351                G                 A       GENIE-DFCI-008656-6895   \n",
       "352                G                 T      GENIE-DFCI-025211-61496   \n",
       "353                G                 A       GENIE-DFCI-007654-6083   \n",
       "354                G                 A  GENIE-MSK-P-0044987-T01-IM6   \n",
       "355                T                 G        GENIE-UHN-394602-ARC2   \n",
       "\n",
       "             RefSeq  Protein_position Exon_Number  \\\n",
       "0    NM_001126112.2             282.0        8/11   \n",
       "1       NM_000038.5            1450.0       16/16   \n",
       "2       NM_000038.5             876.0       16/16   \n",
       "3    NM_001126112.2             285.0        8/11   \n",
       "4    NM_001195427.1              95.0         1/3   \n",
       "..              ...               ...         ...   \n",
       "351     NM_014159.6             432.0        3/21   \n",
       "352     NM_004454.2             152.0        7/13   \n",
       "353     NM_004958.3            1818.0       39/58   \n",
       "354     NM_001412.3              13.0         2/7   \n",
       "355  NM_001042663.1             909.0       21/22   \n",
       "\n",
       "                           HGVSc         HGVSp HGVSp_Short transcript_id_TRUE  \\\n",
       "0     ENST00000269305.4:c.844C>T   p.Arg282Trp     p.R282W  ENST00000269305.4   \n",
       "1    ENST00000257430.4:c.4348C>T  p.Arg1450Ter    p.R1450*  ENST00000257430.4   \n",
       "2    ENST00000257430.4:c.2626C>T   p.Arg876Ter     p.R876*  ENST00000257430.4   \n",
       "3     ENST00000269305.4:c.853G>A   p.Glu285Lys     p.E285K  ENST00000269305.4   \n",
       "4     ENST00000359995.5:c.284C>A    p.Pro95His      p.P95H  ENST00000359995.5   \n",
       "..                           ...           ...         ...                ...   \n",
       "351  ENST00000409792.3:c.1294C>T   p.Arg432Cys     p.R432C  ENST00000409792.3   \n",
       "352   ENST00000306376.5:c.455C>A   p.Pro152Gln     p.P152Q  ENST00000306376.5   \n",
       "353  ENST00000361445.4:c.5452C>T  p.Arg1818Cys    p.R1818C  ENST00000361445.4   \n",
       "354    ENST00000379607.5:c.37C>T    p.Arg13Cys      p.R13C  ENST00000379607.5   \n",
       "355  ENST00000400915.3:c.2727A>C     p.Pro909=     p.P909=  ENST00000400915.3   \n",
       "\n",
       "     seq_start    seq_end                                       wt_w_context  \\\n",
       "0      7577034    7577154  CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...   \n",
       "1    112175579  112175699  ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...   \n",
       "2    112173857  112173977  CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...   \n",
       "3      7577025    7577145  TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...   \n",
       "4     74732899   74733019  TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...   \n",
       "..         ...        ...                                                ...   \n",
       "351   47164772   47164892  CTGTGTATGGCCGAGAATAGCGCGTCCTCTCTCGATAAGGGGAGCT...   \n",
       "352  185797741  185797861  ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...   \n",
       "353   11190687   11190807  CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...   \n",
       "354   20156660   20156780  GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...   \n",
       "355    6528277    6528397  CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...   \n",
       "\n",
       "                                         alt_w_context  Frame  CDS_start  \\\n",
       "0    CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...      0          0   \n",
       "1    ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...      0          0   \n",
       "2    CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...      0          0   \n",
       "3    TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...      0          0   \n",
       "4    TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...      2          0   \n",
       "..                                                 ...    ...        ...   \n",
       "351  CTGTGTATGGCCGAGAATAGCGCGTCCTCTCTCGATAAGGGGAGCT...      0          0   \n",
       "352  ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...      2          0   \n",
       "353  CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...      0          0   \n",
       "354  GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...      2          0   \n",
       "355  CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...      1          0   \n",
       "\n",
       "     CDS_end                         CDS_WT_correct_orientation  \n",
       "0        121  GGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCT...  \n",
       "1        121  ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...  \n",
       "2        121  CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...  \n",
       "3        121  CTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGA...  \n",
       "4        121  TGGACGGGGCCGTGCTGGACGGCCGCGAGCTGCGGGTGCAAATGGC...  \n",
       "..       ...                                                ...  \n",
       "351      121  GGCTCTAGAACTAATTTATCCTATTCCAGGTCAGAACGATCTCATT...  \n",
       "352      121  AGCCATTAACCCCTCCTACAACCCCCCTCTCACCCACCCATCAGAA...  \n",
       "353      121  TTCGAAGCTGTGCTACACTACAAACATCAGAACCAAGCCCGCGATG...  \n",
       "354       81  GTAAAGGAGGTAAAAACAGACGCAGGGGTAAGAATGAGAATGAATC...  \n",
       "355      121  AGGCCCAATGGCAGAGCTAGTGCCTCGGGCCCCAGAGTCCCCACGA...  \n",
       "\n",
       "[356 rows x 28 columns]"
      ]
     },
     "execution_count": 257,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-1"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fp = 'MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD'\n",
    "\n",
    "fp.find('WDGTALRCVFVPVLGETGAQRKRISARKGSLTTSCP')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TGGAACCAGACAGAAAAGCGGCTGTTAGTCACTGGCAGCAACAGTCTTACCTGGACTCTGGAATCCATTCTGGTGCCACTACCACAGCTCCTTCTCTGAGTGGTAAAGGCAATCCTGAGGA'"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGGCCATGGAACCAGACAGAAAAGCGGCTGTTAGTCACTGGCAGCAACAGTCTTACCTGGACTCTGGAATCCATTCTGGTGCCACTACCACAGCTCCTTCTCTGAGTGGTAAAGGCAATC'"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('NPADCRVLVYLQNQPGTKLLNFLQERNLPPKVVLRHPKVH')"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Bio.Seq.Seq(wt_seq[1:]).transcribe().translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[41265560,\n",
       " 41265561,\n",
       " 41265562,\n",
       " 41265563,\n",
       " 41265564,\n",
       " 41265565,\n",
       " 41265566,\n",
       " 41265567,\n",
       " 41265568,\n",
       " 41265569,\n",
       " 41265570,\n",
       " 41265571,\n",
       " 41265572,\n",
       " 41266017,\n",
       " 41266018,\n",
       " 41266019,\n",
       " 41266020,\n",
       " 41266021,\n",
       " 41266022,\n",
       " 41266023,\n",
       " 41266024,\n",
       " 41266025,\n",
       " 41266026,\n",
       " 41266027,\n",
       " 41266028,\n",
       " 41266029,\n",
       " 41266030,\n",
       " 41266031,\n",
       " 41266032,\n",
       " 41266033,\n",
       " 41266034,\n",
       " 41266035,\n",
       " 41266036,\n",
       " 41266037,\n",
       " 41266038,\n",
       " 41266039,\n",
       " 41266040,\n",
       " 41266041,\n",
       " 41266042,\n",
       " 41266043,\n",
       " 41266044,\n",
       " 41266045,\n",
       " 41266046,\n",
       " 41266047,\n",
       " 41266048,\n",
       " 41266049,\n",
       " 41266050,\n",
       " 41266051,\n",
       " 41266052,\n",
       " 41266053,\n",
       " 41266054,\n",
       " 41266055,\n",
       " 41266056,\n",
       " 41266057,\n",
       " 41266058,\n",
       " 41266059,\n",
       " 41266060,\n",
       " 41266061,\n",
       " 41266062,\n",
       " 41266063,\n",
       " 41266064,\n",
       " 41266065,\n",
       " 41266066,\n",
       " 41266067,\n",
       " 41266068,\n",
       " 41266069,\n",
       " 41266070,\n",
       " 41266071,\n",
       " 41266072,\n",
       " 41266073,\n",
       " 41266074,\n",
       " 41266075,\n",
       " 41266076,\n",
       " 41266077,\n",
       " 41266078,\n",
       " 41266079,\n",
       " 41266080,\n",
       " 41266081,\n",
       " 41266082,\n",
       " 41266083,\n",
       " 41266084,\n",
       " 41266085,\n",
       " 41266086,\n",
       " 41266087,\n",
       " 41266088,\n",
       " 41266089,\n",
       " 41266090,\n",
       " 41266091,\n",
       " 41266092,\n",
       " 41266093,\n",
       " 41266094,\n",
       " 41266095,\n",
       " 41266096,\n",
       " 41266097,\n",
       " 41266098,\n",
       " 41266099,\n",
       " 41266100,\n",
       " 41266101,\n",
       " 41266102,\n",
       " 41266103,\n",
       " 41266104,\n",
       " 41266105,\n",
       " 41266106,\n",
       " 41266107,\n",
       " 41266108,\n",
       " 41266109,\n",
       " 41266110,\n",
       " 41266111,\n",
       " 41266112,\n",
       " 41266113,\n",
       " 41266114,\n",
       " 41266115,\n",
       " 41266116,\n",
       " 41266117,\n",
       " 41266118,\n",
       " 41266119,\n",
       " 41266120,\n",
       " 41266121,\n",
       " 41266122,\n",
       " 41266123,\n",
       " 41266124,\n",
       " 41266125,\n",
       " 41266126,\n",
       " 41266127,\n",
       " 41266128,\n",
       " 41266129,\n",
       " 41266130,\n",
       " 41266131,\n",
       " 41266132,\n",
       " 41266133,\n",
       " 41266134,\n",
       " 41266135,\n",
       " 41266136,\n",
       " 41266137,\n",
       " 41266138,\n",
       " 41266139,\n",
       " 41266140,\n",
       " 41266141,\n",
       " 41266142,\n",
       " 41266143,\n",
       " 41266144,\n",
       " 41266145,\n",
       " 41266146,\n",
       " 41266147,\n",
       " 41266148,\n",
       " 41266149,\n",
       " 41266150,\n",
       " 41266151,\n",
       " 41266152,\n",
       " 41266153,\n",
       " 41266154,\n",
       " 41266155,\n",
       " 41266156,\n",
       " 41266157,\n",
       " 41266158,\n",
       " 41266159,\n",
       " 41266160,\n",
       " 41266161,\n",
       " 41266162,\n",
       " 41266163,\n",
       " 41266164,\n",
       " 41266165,\n",
       " 41266166,\n",
       " 41266167,\n",
       " 41266168,\n",
       " 41266169,\n",
       " 41266170,\n",
       " 41266171,\n",
       " 41266172,\n",
       " 41266173,\n",
       " 41266174,\n",
       " 41266175,\n",
       " 41266176,\n",
       " 41266177,\n",
       " 41266178,\n",
       " 41266179,\n",
       " 41266180,\n",
       " 41266181,\n",
       " 41266182,\n",
       " 41266183,\n",
       " 41266184,\n",
       " 41266185,\n",
       " 41266186,\n",
       " 41266187,\n",
       " 41266188,\n",
       " 41266189,\n",
       " 41266190,\n",
       " 41266191,\n",
       " 41266192,\n",
       " 41266193,\n",
       " 41266194,\n",
       " 41266195,\n",
       " 41266196,\n",
       " 41266197,\n",
       " 41266198,\n",
       " 41266199,\n",
       " 41266200,\n",
       " 41266201,\n",
       " 41266202,\n",
       " 41266203,\n",
       " 41266204,\n",
       " 41266205,\n",
       " 41266206,\n",
       " 41266207,\n",
       " 41266208,\n",
       " 41266209,\n",
       " 41266210,\n",
       " 41266211,\n",
       " 41266212,\n",
       " 41266213,\n",
       " 41266214,\n",
       " 41266215,\n",
       " 41266216,\n",
       " 41266217,\n",
       " 41266218,\n",
       " 41266219,\n",
       " 41266220,\n",
       " 41266221,\n",
       " 41266222,\n",
       " 41266223,\n",
       " 41266224,\n",
       " 41266225,\n",
       " 41266226,\n",
       " 41266227,\n",
       " 41266228,\n",
       " 41266229,\n",
       " 41266230,\n",
       " 41266231,\n",
       " 41266232,\n",
       " 41266233,\n",
       " 41266234,\n",
       " 41266235,\n",
       " 41266236,\n",
       " 41266237,\n",
       " 41266238,\n",
       " 41266239,\n",
       " 41266240,\n",
       " 41266241,\n",
       " 41266242,\n",
       " 41266243,\n",
       " 41266244,\n",
       " 41266445,\n",
       " 41266446,\n",
       " 41266447,\n",
       " 41266448,\n",
       " 41266449,\n",
       " 41266450,\n",
       " 41266451,\n",
       " 41266452,\n",
       " 41266453,\n",
       " 41266454,\n",
       " 41266455,\n",
       " 41266456,\n",
       " 41266457,\n",
       " 41266458,\n",
       " 41266459,\n",
       " 41266460,\n",
       " 41266461,\n",
       " 41266462,\n",
       " 41266463,\n",
       " 41266464,\n",
       " 41266465,\n",
       " 41266466,\n",
       " 41266467,\n",
       " 41266468,\n",
       " 41266469,\n",
       " 41266470,\n",
       " 41266471,\n",
       " 41266472,\n",
       " 41266473,\n",
       " 41266474,\n",
       " 41266475,\n",
       " 41266476,\n",
       " 41266477,\n",
       " 41266478,\n",
       " 41266479,\n",
       " 41266480,\n",
       " 41266481,\n",
       " 41266482,\n",
       " 41266483,\n",
       " 41266484,\n",
       " 41266485,\n",
       " 41266486,\n",
       " 41266487,\n",
       " 41266488,\n",
       " 41266489,\n",
       " 41266490,\n",
       " 41266491,\n",
       " 41266492,\n",
       " 41266493,\n",
       " 41266494,\n",
       " 41266495,\n",
       " 41266496,\n",
       " 41266497,\n",
       " 41266498,\n",
       " 41266499,\n",
       " 41266500,\n",
       " 41266501,\n",
       " 41266502,\n",
       " 41266503,\n",
       " 41266504,\n",
       " 41266505,\n",
       " 41266506,\n",
       " 41266507,\n",
       " 41266508,\n",
       " 41266509,\n",
       " 41266510,\n",
       " 41266511,\n",
       " 41266512,\n",
       " 41266513,\n",
       " 41266514,\n",
       " 41266515,\n",
       " 41266516,\n",
       " 41266517,\n",
       " 41266518,\n",
       " 41266519,\n",
       " 41266520,\n",
       " 41266521,\n",
       " 41266522,\n",
       " 41266523,\n",
       " 41266524,\n",
       " 41266525,\n",
       " 41266526,\n",
       " 41266527,\n",
       " 41266528,\n",
       " 41266529,\n",
       " 41266530,\n",
       " 41266531,\n",
       " 41266532,\n",
       " 41266533,\n",
       " 41266534,\n",
       " 41266535,\n",
       " 41266536,\n",
       " 41266537,\n",
       " 41266538,\n",
       " 41266539,\n",
       " 41266540,\n",
       " 41266541,\n",
       " 41266542,\n",
       " 41266543,\n",
       " 41266544,\n",
       " 41266545,\n",
       " 41266546,\n",
       " 41266547,\n",
       " 41266548,\n",
       " 41266549,\n",
       " 41266550,\n",
       " 41266551,\n",
       " 41266552,\n",
       " 41266553,\n",
       " 41266554,\n",
       " 41266555,\n",
       " 41266556,\n",
       " 41266557,\n",
       " 41266558,\n",
       " 41266559,\n",
       " 41266560,\n",
       " 41266561,\n",
       " 41266562,\n",
       " 41266563,\n",
       " 41266564,\n",
       " 41266565,\n",
       " 41266566,\n",
       " 41266567,\n",
       " 41266568,\n",
       " 41266569,\n",
       " 41266570,\n",
       " 41266571,\n",
       " 41266572,\n",
       " 41266573,\n",
       " 41266574,\n",
       " 41266575,\n",
       " 41266576,\n",
       " 41266577,\n",
       " 41266578,\n",
       " 41266579,\n",
       " 41266580,\n",
       " 41266581,\n",
       " 41266582,\n",
       " 41266583,\n",
       " 41266584,\n",
       " 41266585,\n",
       " 41266586,\n",
       " 41266587,\n",
       " 41266588,\n",
       " 41266589,\n",
       " 41266590,\n",
       " 41266591,\n",
       " 41266592,\n",
       " 41266593,\n",
       " 41266594,\n",
       " 41266595,\n",
       " 41266596,\n",
       " 41266597,\n",
       " 41266598,\n",
       " 41266599,\n",
       " 41266600,\n",
       " 41266601,\n",
       " 41266602,\n",
       " 41266603,\n",
       " 41266604,\n",
       " 41266605,\n",
       " 41266606,\n",
       " 41266607,\n",
       " 41266608,\n",
       " 41266609,\n",
       " 41266610,\n",
       " 41266611,\n",
       " 41266612,\n",
       " 41266613,\n",
       " 41266614,\n",
       " 41266615,\n",
       " 41266616,\n",
       " 41266617,\n",
       " 41266618,\n",
       " 41266619,\n",
       " 41266620,\n",
       " 41266621,\n",
       " 41266622,\n",
       " 41266623,\n",
       " 41266624,\n",
       " 41266625,\n",
       " 41266626,\n",
       " 41266627,\n",
       " 41266628,\n",
       " 41266629,\n",
       " 41266630,\n",
       " 41266631,\n",
       " 41266632,\n",
       " 41266633,\n",
       " 41266634,\n",
       " 41266635,\n",
       " 41266636,\n",
       " 41266637,\n",
       " 41266638,\n",
       " 41266639,\n",
       " 41266640,\n",
       " 41266641,\n",
       " 41266642,\n",
       " 41266643,\n",
       " 41266644,\n",
       " 41266645,\n",
       " 41266646,\n",
       " 41266647,\n",
       " 41266648,\n",
       " 41266649,\n",
       " 41266650,\n",
       " 41266651,\n",
       " 41266652,\n",
       " 41266653,\n",
       " 41266654,\n",
       " 41266655,\n",
       " 41266656,\n",
       " 41266657,\n",
       " 41266658,\n",
       " 41266659,\n",
       " 41266660,\n",
       " 41266661,\n",
       " 41266662,\n",
       " 41266663,\n",
       " 41266664,\n",
       " 41266665,\n",
       " 41266666,\n",
       " 41266667,\n",
       " 41266668,\n",
       " 41266669,\n",
       " 41266670,\n",
       " 41266671,\n",
       " 41266672,\n",
       " 41266673,\n",
       " 41266674,\n",
       " 41266675,\n",
       " 41266676,\n",
       " 41266677,\n",
       " 41266678,\n",
       " 41266679,\n",
       " 41266680,\n",
       " 41266681,\n",
       " 41266682,\n",
       " 41266683,\n",
       " 41266684,\n",
       " 41266685,\n",
       " 41266686,\n",
       " 41266687,\n",
       " 41266688,\n",
       " 41266689,\n",
       " 41266690,\n",
       " 41266691,\n",
       " 41266692,\n",
       " 41266693,\n",
       " 41266694,\n",
       " 41266695,\n",
       " 41266696,\n",
       " 41266697,\n",
       " 41266698,\n",
       " 41266825,\n",
       " 41266826,\n",
       " 41266827,\n",
       " 41266828,\n",
       " 41266829,\n",
       " 41266830,\n",
       " 41266831,\n",
       " 41266832,\n",
       " 41266833,\n",
       " 41266834,\n",
       " 41266835,\n",
       " 41266836,\n",
       " 41266837,\n",
       " 41266838,\n",
       " 41266839,\n",
       " 41266840,\n",
       " 41266841,\n",
       " 41266842,\n",
       " 41266843,\n",
       " 41266844,\n",
       " 41266845,\n",
       " 41266846,\n",
       " 41266847,\n",
       " 41266848,\n",
       " 41266849,\n",
       " 41266850,\n",
       " 41266851,\n",
       " 41266852,\n",
       " 41266853,\n",
       " 41266854,\n",
       " 41266855,\n",
       " 41266856,\n",
       " 41266857,\n",
       " 41266858,\n",
       " 41266859,\n",
       " 41266860,\n",
       " 41266861,\n",
       " 41266862,\n",
       " 41266863,\n",
       " 41266864,\n",
       " 41266865,\n",
       " 41266866,\n",
       " 41266867,\n",
       " 41266868,\n",
       " 41266869,\n",
       " 41266870,\n",
       " 41266871,\n",
       " 41266872,\n",
       " 41266873,\n",
       " 41266874,\n",
       " 41266875,\n",
       " 41266876,\n",
       " 41266877,\n",
       " 41266878,\n",
       " 41266879,\n",
       " 41266880,\n",
       " 41266881,\n",
       " 41266882,\n",
       " 41266883,\n",
       " 41266884,\n",
       " 41266885,\n",
       " 41266886,\n",
       " 41266887,\n",
       " 41266888,\n",
       " 41266889,\n",
       " 41266890,\n",
       " 41266891,\n",
       " 41266892,\n",
       " 41266893,\n",
       " 41266894,\n",
       " 41266895,\n",
       " 41266896,\n",
       " 41266897,\n",
       " 41266898,\n",
       " 41266899,\n",
       " 41266900,\n",
       " 41266901,\n",
       " 41266902,\n",
       " 41266903,\n",
       " 41266904,\n",
       " 41266905,\n",
       " 41266906,\n",
       " 41266907,\n",
       " 41266908,\n",
       " 41266909,\n",
       " 41266910,\n",
       " 41266911,\n",
       " 41266912,\n",
       " 41266913,\n",
       " 41266914,\n",
       " 41266915,\n",
       " 41266916,\n",
       " 41266917,\n",
       " 41266918,\n",
       " 41266919,\n",
       " 41266920,\n",
       " 41266921,\n",
       " 41266922,\n",
       " 41266923,\n",
       " 41266924,\n",
       " 41266925,\n",
       " 41266926,\n",
       " 41266927,\n",
       " 41266928,\n",
       " 41266929,\n",
       " 41266930,\n",
       " 41266931,\n",
       " 41266932,\n",
       " 41266933,\n",
       " 41266934,\n",
       " 41266935,\n",
       " 41266936,\n",
       " 41266937,\n",
       " 41266938,\n",
       " 41266939,\n",
       " 41266940,\n",
       " 41266941,\n",
       " 41266942,\n",
       " 41266943,\n",
       " 41266944,\n",
       " 41266945,\n",
       " 41266946,\n",
       " 41266947,\n",
       " 41266948,\n",
       " 41266949,\n",
       " 41266950,\n",
       " 41266951,\n",
       " 41266952,\n",
       " 41266953,\n",
       " 41266954,\n",
       " 41266955,\n",
       " 41266956,\n",
       " 41266957,\n",
       " 41266958,\n",
       " 41266959,\n",
       " 41266960,\n",
       " 41266961,\n",
       " 41266962,\n",
       " 41266963,\n",
       " 41266964,\n",
       " 41266965,\n",
       " 41266966,\n",
       " 41266967,\n",
       " 41266968,\n",
       " 41266969,\n",
       " 41266970,\n",
       " 41266971,\n",
       " 41266972,\n",
       " 41266973,\n",
       " 41266974,\n",
       " 41266975,\n",
       " 41266976,\n",
       " 41266977,\n",
       " 41266978,\n",
       " 41266979,\n",
       " 41266980,\n",
       " 41266981,\n",
       " 41266982,\n",
       " 41266983,\n",
       " 41266984,\n",
       " 41266985,\n",
       " 41266986,\n",
       " 41266987,\n",
       " 41266988,\n",
       " 41266989,\n",
       " 41266990,\n",
       " 41266991,\n",
       " 41266992,\n",
       " 41266993,\n",
       " 41266994,\n",
       " 41266995,\n",
       " 41266996,\n",
       " 41266997,\n",
       " 41266998,\n",
       " 41266999,\n",
       " 41267000,\n",
       " 41267001,\n",
       " 41267002,\n",
       " 41267003,\n",
       " 41267004,\n",
       " 41267005,\n",
       " 41267006,\n",
       " 41267007,\n",
       " 41267008,\n",
       " 41267009,\n",
       " 41267010,\n",
       " 41267011,\n",
       " 41267012,\n",
       " 41267013,\n",
       " 41267014,\n",
       " 41267015,\n",
       " 41267016,\n",
       " 41267017,\n",
       " 41267018,\n",
       " 41267019,\n",
       " 41267020,\n",
       " 41267021,\n",
       " 41267022,\n",
       " 41267023,\n",
       " 41267024,\n",
       " 41267025,\n",
       " 41267026,\n",
       " 41267027,\n",
       " 41267028,\n",
       " 41267029,\n",
       " 41267030,\n",
       " 41267031,\n",
       " 41267032,\n",
       " 41267033,\n",
       " 41267034,\n",
       " 41267035,\n",
       " 41267036,\n",
       " 41267037,\n",
       " 41267038,\n",
       " 41267039,\n",
       " 41267040,\n",
       " 41267041,\n",
       " 41267042,\n",
       " 41267043,\n",
       " 41267044,\n",
       " 41267045,\n",
       " 41267046,\n",
       " 41267047,\n",
       " 41267048,\n",
       " 41267049,\n",
       " 41267050,\n",
       " 41267051,\n",
       " 41267052,\n",
       " 41267053,\n",
       " 41267054,\n",
       " 41267055,\n",
       " 41267056,\n",
       " 41267057,\n",
       " 41267058,\n",
       " 41267059,\n",
       " 41267060,\n",
       " 41267061,\n",
       " 41267062,\n",
       " 41267063,\n",
       " 41267151,\n",
       " 41267152,\n",
       " 41267153,\n",
       " 41267154,\n",
       " 41267155,\n",
       " 41267156,\n",
       " 41267157,\n",
       " 41267158,\n",
       " 41267159,\n",
       " 41267160,\n",
       " 41267161,\n",
       " 41267162,\n",
       " 41267163,\n",
       " 41267164,\n",
       " 41267165,\n",
       " 41267166,\n",
       " 41267167,\n",
       " 41267168,\n",
       " 41267169,\n",
       " 41267170,\n",
       " 41267171,\n",
       " 41267172,\n",
       " 41267173,\n",
       " 41267174,\n",
       " 41267175,\n",
       " 41267176,\n",
       " 41267177,\n",
       " 41267178,\n",
       " 41267179,\n",
       " 41267180,\n",
       " 41267181,\n",
       " 41267182,\n",
       " 41267183,\n",
       " 41267184,\n",
       " 41267185,\n",
       " 41267186,\n",
       " 41267187,\n",
       " 41267188,\n",
       " 41267189,\n",
       " 41267190,\n",
       " 41267191,\n",
       " 41267192,\n",
       " 41267193,\n",
       " 41267194,\n",
       " 41267195,\n",
       " 41267196,\n",
       " 41267197,\n",
       " 41267198,\n",
       " 41267199,\n",
       " 41267200,\n",
       " 41267201,\n",
       " 41267202,\n",
       " 41267203,\n",
       " 41267204,\n",
       " 41267205,\n",
       " 41267206,\n",
       " 41267207,\n",
       " 41267208,\n",
       " 41267209,\n",
       " 41267210,\n",
       " 41267211,\n",
       " 41267212,\n",
       " 41267213,\n",
       " 41267214,\n",
       " 41267215,\n",
       " 41267216,\n",
       " 41267217,\n",
       " 41267218,\n",
       " 41267219,\n",
       " 41267220,\n",
       " 41267221,\n",
       " 41267222,\n",
       " 41267223,\n",
       " 41267224,\n",
       " 41267225,\n",
       " 41267226,\n",
       " 41267227,\n",
       " 41267228,\n",
       " 41267229,\n",
       " 41267230,\n",
       " 41267231,\n",
       " 41267232,\n",
       " 41267233,\n",
       " 41267234,\n",
       " 41267235,\n",
       " 41267236,\n",
       " 41267237,\n",
       " 41267238,\n",
       " 41267239,\n",
       " 41267240,\n",
       " 41267241,\n",
       " 41267242,\n",
       " 41267243,\n",
       " 41267244,\n",
       " 41267245,\n",
       " 41267246,\n",
       " 41267247,\n",
       " 41267248,\n",
       " 41267249,\n",
       " 41267250,\n",
       " 41267251,\n",
       " 41267252,\n",
       " 41267253,\n",
       " 41267254,\n",
       " 41267255,\n",
       " 41267256,\n",
       " 41267257,\n",
       " 41267258,\n",
       " 41267259,\n",
       " 41267260,\n",
       " 41267261,\n",
       " 41267262,\n",
       " 41267263,\n",
       " 41267264,\n",
       " 41267265,\n",
       " 41267266,\n",
       " 41267267,\n",
       " 41267268,\n",
       " 41267269,\n",
       " 41267270,\n",
       " 41267271,\n",
       " 41267272,\n",
       " 41267273,\n",
       " 41267274,\n",
       " 41267275,\n",
       " 41267276,\n",
       " 41267277,\n",
       " 41267278,\n",
       " 41267279,\n",
       " 41267280,\n",
       " 41267281,\n",
       " 41267282,\n",
       " 41267283,\n",
       " 41267284,\n",
       " 41267285,\n",
       " 41267286,\n",
       " 41267287,\n",
       " 41267288,\n",
       " 41267289,\n",
       " 41267290,\n",
       " 41267291,\n",
       " 41267292,\n",
       " 41267293,\n",
       " 41267294,\n",
       " 41267295,\n",
       " 41267296,\n",
       " 41267297,\n",
       " 41267298,\n",
       " 41267299,\n",
       " 41267300,\n",
       " 41267301,\n",
       " 41267302,\n",
       " 41267303,\n",
       " 41267304,\n",
       " 41267305,\n",
       " 41267306,\n",
       " 41267307,\n",
       " 41267308,\n",
       " 41267309,\n",
       " 41267310,\n",
       " 41267311,\n",
       " 41267312,\n",
       " 41267313,\n",
       " 41267314,\n",
       " 41267315,\n",
       " 41267316,\n",
       " 41267317,\n",
       " 41267318,\n",
       " 41267319,\n",
       " 41267320,\n",
       " 41267321,\n",
       " 41267322,\n",
       " 41267323,\n",
       " 41267324,\n",
       " 41267325,\n",
       " 41267326,\n",
       " 41267327,\n",
       " 41267328,\n",
       " 41267329,\n",
       " 41267330,\n",
       " 41267331,\n",
       " 41267332,\n",
       " 41267333,\n",
       " 41267334,\n",
       " 41267335,\n",
       " 41267336,\n",
       " 41267337,\n",
       " 41267338,\n",
       " 41267339,\n",
       " 41267340,\n",
       " 41267341,\n",
       " 41267342,\n",
       " 41267343,\n",
       " 41267344,\n",
       " 41267345,\n",
       " 41267346,\n",
       " 41267347,\n",
       " 41267348,\n",
       " 41267349,\n",
       " 41267350,\n",
       " 41267351,\n",
       " 41267352,\n",
       " 41268699,\n",
       " 41268700,\n",
       " 41268701,\n",
       " 41268702,\n",
       " 41268703,\n",
       " 41268704,\n",
       " 41268705,\n",
       " 41268706,\n",
       " 41268707,\n",
       " 41268708,\n",
       " 41268709,\n",
       " 41268710,\n",
       " 41268711,\n",
       " 41268712,\n",
       " 41268713,\n",
       " 41268714,\n",
       " 41268715,\n",
       " 41268716,\n",
       " 41268717,\n",
       " 41268718,\n",
       " 41268719,\n",
       " 41268720,\n",
       " 41268721,\n",
       " 41268722,\n",
       " 41268723,\n",
       " 41268724,\n",
       " 41268725,\n",
       " 41268726,\n",
       " 41268727,\n",
       " 41268728,\n",
       " 41268729,\n",
       " 41268730,\n",
       " 41268731,\n",
       " 41268732,\n",
       " 41268733,\n",
       " 41268734,\n",
       " 41268735,\n",
       " 41268736,\n",
       " 41268737,\n",
       " 41268738,\n",
       " 41268739,\n",
       " 41268740,\n",
       " 41268741,\n",
       " 41268742,\n",
       " 41268743,\n",
       " 41268744,\n",
       " 41268745,\n",
       " 41268746,\n",
       " 41268747,\n",
       " 41268748,\n",
       " 41268749,\n",
       " 41268750,\n",
       " 41268751,\n",
       " 41268752,\n",
       " 41268753,\n",
       " 41268754,\n",
       " 41268755,\n",
       " 41268756,\n",
       " 41268757,\n",
       " 41268758,\n",
       " 41268759,\n",
       " 41268760,\n",
       " 41268761,\n",
       " 41268762,\n",
       " ...]"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_locs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "subs_new = df_formatter(subs[:400])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Intron', 'Missense_Mutation', 'Nonsense_Mutation', 'Silent',\n",
       "       'Splice_Region', 'Splice_Site'], dtype=object)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(subs_new['Variant_Classification'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level_0</th>\n",
       "      <th>index</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1512.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-JHU-00113-00335</td>\n",
       "      <td>NM_001126112.2</td>\n",
       "      <td>282.0</td>\n",
       "      <td>8/11</td>\n",
       "      <td>ENST00000269305.4:c.844C&gt;T</td>\n",
       "      <td>p.Arg282Trp</td>\n",
       "      <td>p.R282W</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577034</td>\n",
       "      <td>7577154</td>\n",
       "      <td>CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...</td>\n",
       "      <td>CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>777.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175639</td>\n",
       "      <td>112175639</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00223-00468</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4348C&gt;T</td>\n",
       "      <td>p.Arg1450Ter</td>\n",
       "      <td>p.R1450*</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175579</td>\n",
       "      <td>112175699</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "      <td>ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>614.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112173917</td>\n",
       "      <td>112173917</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00123-00365</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>876.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.2626C&gt;T</td>\n",
       "      <td>p.Arg876Ter</td>\n",
       "      <td>p.R876*</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112173857</td>\n",
       "      <td>112173977</td>\n",
       "      <td>CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...</td>\n",
       "      <td>CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>538.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577085</td>\n",
       "      <td>7577085</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-JHU-00694-00783</td>\n",
       "      <td>NM_001126112.2</td>\n",
       "      <td>285.0</td>\n",
       "      <td>8/11</td>\n",
       "      <td>ENST00000269305.4:c.853G&gt;A</td>\n",
       "      <td>p.Glu285Lys</td>\n",
       "      <td>p.E285K</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577025</td>\n",
       "      <td>7577145</td>\n",
       "      <td>TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...</td>\n",
       "      <td>TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>527.0</td>\n",
       "      <td>SRSF2</td>\n",
       "      <td>17</td>\n",
       "      <td>74732959</td>\n",
       "      <td>74732959</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-DFCI-001379-11338</td>\n",
       "      <td>NM_001195427.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>1/3</td>\n",
       "      <td>ENST00000359995.5:c.284C&gt;A</td>\n",
       "      <td>p.Pro95His</td>\n",
       "      <td>p.P95H</td>\n",
       "      <td>ENST00000359995.5</td>\n",
       "      <td>74732899</td>\n",
       "      <td>74733019</td>\n",
       "      <td>TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...</td>\n",
       "      <td>TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>394</td>\n",
       "      <td>602</td>\n",
       "      <td>19.0</td>\n",
       "      <td>SETD2</td>\n",
       "      <td>3</td>\n",
       "      <td>47164832</td>\n",
       "      <td>47164832</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-DFCI-008656-6895</td>\n",
       "      <td>NM_014159.6</td>\n",
       "      <td>432.0</td>\n",
       "      <td>3/21</td>\n",
       "      <td>ENST00000409792.3:c.1294C&gt;T</td>\n",
       "      <td>p.Arg432Cys</td>\n",
       "      <td>p.R432C</td>\n",
       "      <td>ENST00000409792.3</td>\n",
       "      <td>47164772</td>\n",
       "      <td>47164892</td>\n",
       "      <td>CTGTGTATGGCCGAGAATAGCGCGTCCTCTCTCGATAAGGGGAGCT...</td>\n",
       "      <td>CTGTGTATGGCCGAGAATAGCGCGTCCTCTCTCGATAAGGGGAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>395</td>\n",
       "      <td>603</td>\n",
       "      <td>19.0</td>\n",
       "      <td>ETV5</td>\n",
       "      <td>3</td>\n",
       "      <td>185797801</td>\n",
       "      <td>185797801</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GENIE-DFCI-025211-61496</td>\n",
       "      <td>NM_004454.2</td>\n",
       "      <td>152.0</td>\n",
       "      <td>7/13</td>\n",
       "      <td>ENST00000306376.5:c.455C&gt;A</td>\n",
       "      <td>p.Pro152Gln</td>\n",
       "      <td>p.P152Q</td>\n",
       "      <td>ENST00000306376.5</td>\n",
       "      <td>185797741</td>\n",
       "      <td>185797861</td>\n",
       "      <td>ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...</td>\n",
       "      <td>ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>396</td>\n",
       "      <td>604</td>\n",
       "      <td>19.0</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>1</td>\n",
       "      <td>11190747</td>\n",
       "      <td>11190747</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-DFCI-007654-6083</td>\n",
       "      <td>NM_004958.3</td>\n",
       "      <td>1818.0</td>\n",
       "      <td>39/58</td>\n",
       "      <td>ENST00000361445.4:c.5452C&gt;T</td>\n",
       "      <td>p.Arg1818Cys</td>\n",
       "      <td>p.R1818C</td>\n",
       "      <td>ENST00000361445.4</td>\n",
       "      <td>11190687</td>\n",
       "      <td>11190807</td>\n",
       "      <td>CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...</td>\n",
       "      <td>CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>398</td>\n",
       "      <td>607</td>\n",
       "      <td>19.0</td>\n",
       "      <td>EIF1AX</td>\n",
       "      <td>X</td>\n",
       "      <td>20156720</td>\n",
       "      <td>20156720</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-MSK-P-0044987-T01-IM6</td>\n",
       "      <td>NM_001412.3</td>\n",
       "      <td>13.0</td>\n",
       "      <td>2/7</td>\n",
       "      <td>ENST00000379607.5:c.37C&gt;T</td>\n",
       "      <td>p.Arg13Cys</td>\n",
       "      <td>p.R13C</td>\n",
       "      <td>ENST00000379607.5</td>\n",
       "      <td>20156660</td>\n",
       "      <td>20156780</td>\n",
       "      <td>GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...</td>\n",
       "      <td>GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>399</td>\n",
       "      <td>608</td>\n",
       "      <td>19.0</td>\n",
       "      <td>PLEKHG5</td>\n",
       "      <td>1</td>\n",
       "      <td>6528337</td>\n",
       "      <td>6528337</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-UHN-394602-ARC2</td>\n",
       "      <td>NM_001042663.1</td>\n",
       "      <td>909.0</td>\n",
       "      <td>21/22</td>\n",
       "      <td>ENST00000400915.3:c.2727A&gt;C</td>\n",
       "      <td>p.Pro909=</td>\n",
       "      <td>p.P909=</td>\n",
       "      <td>ENST00000400915.3</td>\n",
       "      <td>6528277</td>\n",
       "      <td>6528397</td>\n",
       "      <td>CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...</td>\n",
       "      <td>CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>356 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     level_0  index   COUNT Hugo_Symbol Chromosome  Start_Position  \\\n",
       "0          0      1  1512.0        TP53         17         7577094   \n",
       "1          1      3   777.0         APC          5       112175639   \n",
       "2          2      6   614.0         APC          5       112173917   \n",
       "3          3      7   538.0        TP53         17         7577085   \n",
       "4          4      8   527.0       SRSF2         17        74732959   \n",
       "..       ...    ...     ...         ...        ...             ...   \n",
       "351      394    602    19.0       SETD2          3        47164832   \n",
       "352      395    603    19.0        ETV5          3       185797801   \n",
       "353      396    604    19.0        MTOR          1        11190747   \n",
       "354      398    607    19.0      EIF1AX          X        20156720   \n",
       "355      399    608    19.0     PLEKHG5          1         6528337   \n",
       "\n",
       "     End_Position         Consequence Variant_Classification Variant_Type  \\\n",
       "0         7577094    missense_variant      Missense_Mutation          SNP   \n",
       "1       112175639         stop_gained      Nonsense_Mutation          SNP   \n",
       "2       112173917         stop_gained      Nonsense_Mutation          SNP   \n",
       "3         7577085    missense_variant      Missense_Mutation          SNP   \n",
       "4        74732959    missense_variant      Missense_Mutation          SNP   \n",
       "..            ...                 ...                    ...          ...   \n",
       "351      47164832    missense_variant      Missense_Mutation          SNP   \n",
       "352     185797801    missense_variant      Missense_Mutation          SNP   \n",
       "353      11190747    missense_variant      Missense_Mutation          SNP   \n",
       "354      20156720    missense_variant      Missense_Mutation          SNP   \n",
       "355       6528337  synonymous_variant                 Silent          SNP   \n",
       "\n",
       "    Reference_Allele Tumor_Seq_Allele2         Tumor_Sample_Barcode  \\\n",
       "0                  G                 A        GENIE-JHU-00113-00335   \n",
       "1                  C                 T        GENIE-JHU-00223-00468   \n",
       "2                  C                 T        GENIE-JHU-00123-00365   \n",
       "3                  C                 T        GENIE-JHU-00694-00783   \n",
       "4                  G                 T      GENIE-DFCI-001379-11338   \n",
       "..               ...               ...                          ...   \n",
       "351                G                 A       GENIE-DFCI-008656-6895   \n",
       "352                G                 T      GENIE-DFCI-025211-61496   \n",
       "353                G                 A       GENIE-DFCI-007654-6083   \n",
       "354                G                 A  GENIE-MSK-P-0044987-T01-IM6   \n",
       "355                T                 G        GENIE-UHN-394602-ARC2   \n",
       "\n",
       "             RefSeq  Protein_position Exon_Number  \\\n",
       "0    NM_001126112.2             282.0        8/11   \n",
       "1       NM_000038.5            1450.0       16/16   \n",
       "2       NM_000038.5             876.0       16/16   \n",
       "3    NM_001126112.2             285.0        8/11   \n",
       "4    NM_001195427.1              95.0         1/3   \n",
       "..              ...               ...         ...   \n",
       "351     NM_014159.6             432.0        3/21   \n",
       "352     NM_004454.2             152.0        7/13   \n",
       "353     NM_004958.3            1818.0       39/58   \n",
       "354     NM_001412.3              13.0         2/7   \n",
       "355  NM_001042663.1             909.0       21/22   \n",
       "\n",
       "                           HGVSc         HGVSp HGVSp_Short transcript_id_TRUE  \\\n",
       "0     ENST00000269305.4:c.844C>T   p.Arg282Trp     p.R282W  ENST00000269305.4   \n",
       "1    ENST00000257430.4:c.4348C>T  p.Arg1450Ter    p.R1450*  ENST00000257430.4   \n",
       "2    ENST00000257430.4:c.2626C>T   p.Arg876Ter     p.R876*  ENST00000257430.4   \n",
       "3     ENST00000269305.4:c.853G>A   p.Glu285Lys     p.E285K  ENST00000269305.4   \n",
       "4     ENST00000359995.5:c.284C>A    p.Pro95His      p.P95H  ENST00000359995.5   \n",
       "..                           ...           ...         ...                ...   \n",
       "351  ENST00000409792.3:c.1294C>T   p.Arg432Cys     p.R432C  ENST00000409792.3   \n",
       "352   ENST00000306376.5:c.455C>A   p.Pro152Gln     p.P152Q  ENST00000306376.5   \n",
       "353  ENST00000361445.4:c.5452C>T  p.Arg1818Cys    p.R1818C  ENST00000361445.4   \n",
       "354    ENST00000379607.5:c.37C>T    p.Arg13Cys      p.R13C  ENST00000379607.5   \n",
       "355  ENST00000400915.3:c.2727A>C     p.Pro909=     p.P909=  ENST00000400915.3   \n",
       "\n",
       "     seq_start    seq_end                                       wt_w_context  \\\n",
       "0      7577034    7577154  CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...   \n",
       "1    112175579  112175699  ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...   \n",
       "2    112173857  112173977  CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...   \n",
       "3      7577025    7577145  TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...   \n",
       "4     74732899   74733019  TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...   \n",
       "..         ...        ...                                                ...   \n",
       "351   47164772   47164892  CTGTGTATGGCCGAGAATAGCGCGTCCTCTCTCGATAAGGGGAGCT...   \n",
       "352  185797741  185797861  ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...   \n",
       "353   11190687   11190807  CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...   \n",
       "354   20156660   20156780  GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...   \n",
       "355    6528277    6528397  CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...   \n",
       "\n",
       "                                         alt_w_context  \n",
       "0    CTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCGGAGATTCTC...  \n",
       "1    ACCATGCCACCAAGCAGAAGTAAAACACCTCCACCACCTCCTCAAA...  \n",
       "2    CGCGGAATTGGTCTAGGCAACTACCATCCAGCAACAGAAAATCCAG...  \n",
       "3    TAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCTTGCG...  \n",
       "4    TAGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGT...  \n",
       "..                                                 ...  \n",
       "351  CTGTGTATGGCCGAGAATAGCGCGTCCTCTCTCGATAAGGGGAGCT...  \n",
       "352  ACACCTTGAACTGGGCCAGCTGCAGGGGCATGCCCTGAGGTGGGCA...  \n",
       "353  CAGTGGCGGCCGTGGTGGCGGCAGTGGTGGCGTTGGTGATGTTGGC...  \n",
       "354  GACCATCCTCTTTGAATACCAGTTCTCTTTTTTCAGATTCATTCTC...  \n",
       "355  CAGGTGGGGCGGGCAGCTCAACAGCTGGACAGGGGTGCGGCGGCGG...  \n",
       "\n",
       "[356 rows x 24 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coding_muts = ['Missense_Mutation', 'Nonsense_Mutation', 'Silent']\n",
    "exon_subs = subs_new[subs_new['Variant_Classification'].isin(coding_muts)].reset_index()\n",
    "exon_subs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>762.0</td>\n",
       "      <td>ASXL1</td>\n",
       "      <td>20</td>\n",
       "      <td>31022441</td>\n",
       "      <td>31022442</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-003409-1958</td>\n",
       "      <td>NM_015338.5</td>\n",
       "      <td>642.0</td>\n",
       "      <td>13/13</td>\n",
       "      <td>ENST00000375687.4:c.1934dup</td>\n",
       "      <td>p.Gly646TrpfsTer12</td>\n",
       "      <td>p.G646Wfs*12</td>\n",
       "      <td>ENST00000375687.4</td>\n",
       "      <td>31022381</td>\n",
       "      <td>31022502</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>645.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175952</td>\n",
       "      <td>112175953</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-JHU-00198-00378</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1554.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4666dup</td>\n",
       "      <td>p.Thr1556AsnfsTer3</td>\n",
       "      <td>p.T1556Nfs*3</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175892</td>\n",
       "      <td>112176013</td>\n",
       "      <td>GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...</td>\n",
       "      <td>GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>280.0</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>1</td>\n",
       "      <td>27105930</td>\n",
       "      <td>27105931</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-001750-9089</td>\n",
       "      <td>NM_006015.4</td>\n",
       "      <td>1847.0</td>\n",
       "      <td>20/20</td>\n",
       "      <td>ENST00000324856.7:c.5548dup</td>\n",
       "      <td>p.Asp1850GlyfsTer4</td>\n",
       "      <td>p.D1850Gfs*4</td>\n",
       "      <td>ENST00000324856.7</td>\n",
       "      <td>27105870</td>\n",
       "      <td>27105991</td>\n",
       "      <td>CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...</td>\n",
       "      <td>CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>55</td>\n",
       "      <td>108.0</td>\n",
       "      <td>TCF7L2</td>\n",
       "      <td>10</td>\n",
       "      <td>114925316</td>\n",
       "      <td>114925317</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-DFCI-151674-1525793</td>\n",
       "      <td>NaN</td>\n",
       "      <td>482.0</td>\n",
       "      <td>15/15</td>\n",
       "      <td>ENST00000355995.4:c.1454dup</td>\n",
       "      <td>p.Cys486ValfsTer8</td>\n",
       "      <td>p.C486Vfs*8</td>\n",
       "      <td>ENST00000355995.4</td>\n",
       "      <td>114925256</td>\n",
       "      <td>114925377</td>\n",
       "      <td>CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...</td>\n",
       "      <td>CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>60</td>\n",
       "      <td>99.0</td>\n",
       "      <td>GATA3</td>\n",
       "      <td>10</td>\n",
       "      <td>8115874</td>\n",
       "      <td>8115875</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-002328-1774</td>\n",
       "      <td>NaN</td>\n",
       "      <td>407.0</td>\n",
       "      <td>6/6</td>\n",
       "      <td>ENST00000346208.3:c.1221dup</td>\n",
       "      <td>p.Pro408AlafsTer99</td>\n",
       "      <td>p.P408Afs*99</td>\n",
       "      <td>ENST00000346208.3</td>\n",
       "      <td>8115814</td>\n",
       "      <td>8115935</td>\n",
       "      <td>AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...</td>\n",
       "      <td>AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>8994</td>\n",
       "      <td>5.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175621</td>\n",
       "      <td>112175622</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-MSK-P-0048781-T01-IM6</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1444.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4333dup</td>\n",
       "      <td>p.Thr1445AsnfsTer10</td>\n",
       "      <td>p.T1445Nfs*10</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175561</td>\n",
       "      <td>112175682</td>\n",
       "      <td>CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...</td>\n",
       "      <td>CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>9010</td>\n",
       "      <td>5.0</td>\n",
       "      <td>SRSF2</td>\n",
       "      <td>17</td>\n",
       "      <td>74732960</td>\n",
       "      <td>74732961</td>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>GGC</td>\n",
       "      <td>GENIE-UHN-AGI147550-BM1</td>\n",
       "      <td>NM_001195427.1</td>\n",
       "      <td>94.0</td>\n",
       "      <td>1/3</td>\n",
       "      <td>ENST00000359995.5:c.282_283insGCC</td>\n",
       "      <td>p.Arg94_Pro95insAla</td>\n",
       "      <td>p.R94_P95insA</td>\n",
       "      <td>ENST00000359995.5</td>\n",
       "      <td>74732900</td>\n",
       "      <td>74733021</td>\n",
       "      <td>AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...</td>\n",
       "      <td>AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>9011</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>X</td>\n",
       "      <td>15841255</td>\n",
       "      <td>15841256</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>AGCCGC</td>\n",
       "      <td>GENIE-UHN-AGI974552-BM1</td>\n",
       "      <td>NM_005089.3</td>\n",
       "      <td>447.0</td>\n",
       "      <td>11/11</td>\n",
       "      <td>ENST00000307771.7:c.1339_1340insAGCCGC</td>\n",
       "      <td>p.Ser447delinsLysProArg</td>\n",
       "      <td>p.S447delinsKPR</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15841195</td>\n",
       "      <td>15841316</td>\n",
       "      <td>AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...</td>\n",
       "      <td>AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>9013</td>\n",
       "      <td>5.0</td>\n",
       "      <td>TSC1</td>\n",
       "      <td>9</td>\n",
       "      <td>135771988</td>\n",
       "      <td>135771989</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>GCT</td>\n",
       "      <td>GENIE-UHN-DIVA273992-ARC1</td>\n",
       "      <td>NM_001162426.1</td>\n",
       "      <td>1043.0</td>\n",
       "      <td>23/23</td>\n",
       "      <td>ENST00000298552.3:c.3128_3129insAGC</td>\n",
       "      <td>p.Ser1043delinsArgAla</td>\n",
       "      <td>p.S1043delinsRA</td>\n",
       "      <td>ENST00000298552.3</td>\n",
       "      <td>135771928</td>\n",
       "      <td>135772049</td>\n",
       "      <td>CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...</td>\n",
       "      <td>CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>9094</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ZFHX3</td>\n",
       "      <td>16</td>\n",
       "      <td>72821633</td>\n",
       "      <td>72821634</td>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>GCCACT</td>\n",
       "      <td>GENIE-PROV-2219952519-76fe5c70a7</td>\n",
       "      <td>NM_006885.3</td>\n",
       "      <td>3514.0</td>\n",
       "      <td>10/10</td>\n",
       "      <td>ENST00000268489.5:c.10541_10542insAGTGGC</td>\n",
       "      <td>p.Gly3514_Gly3515insValAla</td>\n",
       "      <td>p.G3514_G3515insVA</td>\n",
       "      <td>ENST00000268489.5</td>\n",
       "      <td>72821573</td>\n",
       "      <td>72821694</td>\n",
       "      <td>GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...</td>\n",
       "      <td>GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>347 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     index  COUNT Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0        4  762.0       ASXL1         20        31022441      31022442   \n",
       "1        5  645.0         APC          5       112175952     112175953   \n",
       "2       21  280.0      ARID1A          1        27105930      27105931   \n",
       "4       55  108.0      TCF7L2         10       114925316     114925317   \n",
       "5       60   99.0       GATA3         10         8115874       8115875   \n",
       "..     ...    ...         ...        ...             ...           ...   \n",
       "354   8994    5.0         APC          5       112175621     112175622   \n",
       "355   9010    5.0       SRSF2         17        74732960      74732961   \n",
       "356   9011    5.0       ZRSR2          X        15841255      15841256   \n",
       "357   9013    5.0        TSC1          9       135771988     135771989   \n",
       "360   9094    5.0       ZFHX3         16        72821633      72821634   \n",
       "\n",
       "                  Consequence Variant_Classification Variant_Type  \\\n",
       "0          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "1          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "2          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "4          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "5          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "..                        ...                    ...          ...   \n",
       "354        frameshift_variant        Frame_Shift_Ins          INS   \n",
       "355         inframe_insertion           In_Frame_Ins          INS   \n",
       "356  protein_altering_variant           In_Frame_Ins          INS   \n",
       "357  protein_altering_variant           In_Frame_Ins          INS   \n",
       "360         inframe_insertion           In_Frame_Ins          INS   \n",
       "\n",
       "    Reference_Allele Tumor_Seq_Allele2              Tumor_Sample_Barcode  \\\n",
       "0                  -                 G            GENIE-DFCI-003409-1958   \n",
       "1                  -                 A             GENIE-JHU-00198-00378   \n",
       "2                  -                 G            GENIE-DFCI-001750-9089   \n",
       "4                  -                 A         GENIE-DFCI-151674-1525793   \n",
       "5                  -                 G            GENIE-DFCI-002328-1774   \n",
       "..               ...               ...                               ...   \n",
       "354                -                 A       GENIE-MSK-P-0048781-T01-IM6   \n",
       "355                -               GGC           GENIE-UHN-AGI147550-BM1   \n",
       "356                -            AGCCGC           GENIE-UHN-AGI974552-BM1   \n",
       "357                -               GCT         GENIE-UHN-DIVA273992-ARC1   \n",
       "360                -            GCCACT  GENIE-PROV-2219952519-76fe5c70a7   \n",
       "\n",
       "             RefSeq  Protein_position Exon_Number  \\\n",
       "0       NM_015338.5             642.0       13/13   \n",
       "1       NM_000038.5            1554.0       16/16   \n",
       "2       NM_006015.4            1847.0       20/20   \n",
       "4               NaN             482.0       15/15   \n",
       "5               NaN             407.0         6/6   \n",
       "..              ...               ...         ...   \n",
       "354     NM_000038.5            1444.0       16/16   \n",
       "355  NM_001195427.1              94.0         1/3   \n",
       "356     NM_005089.3             447.0       11/11   \n",
       "357  NM_001162426.1            1043.0       23/23   \n",
       "360     NM_006885.3            3514.0       10/10   \n",
       "\n",
       "                                        HGVSc                       HGVSp  \\\n",
       "0                 ENST00000375687.4:c.1934dup          p.Gly646TrpfsTer12   \n",
       "1                 ENST00000257430.4:c.4666dup          p.Thr1556AsnfsTer3   \n",
       "2                 ENST00000324856.7:c.5548dup          p.Asp1850GlyfsTer4   \n",
       "4                 ENST00000355995.4:c.1454dup           p.Cys486ValfsTer8   \n",
       "5                 ENST00000346208.3:c.1221dup          p.Pro408AlafsTer99   \n",
       "..                                        ...                         ...   \n",
       "354               ENST00000257430.4:c.4333dup         p.Thr1445AsnfsTer10   \n",
       "355         ENST00000359995.5:c.282_283insGCC         p.Arg94_Pro95insAla   \n",
       "356    ENST00000307771.7:c.1339_1340insAGCCGC     p.Ser447delinsLysProArg   \n",
       "357       ENST00000298552.3:c.3128_3129insAGC       p.Ser1043delinsArgAla   \n",
       "360  ENST00000268489.5:c.10541_10542insAGTGGC  p.Gly3514_Gly3515insValAla   \n",
       "\n",
       "            HGVSp_Short transcript_id_TRUE  seq_start    seq_end  \\\n",
       "0          p.G646Wfs*12  ENST00000375687.4   31022381   31022502   \n",
       "1          p.T1556Nfs*3  ENST00000257430.4  112175892  112176013   \n",
       "2          p.D1850Gfs*4  ENST00000324856.7   27105870   27105991   \n",
       "4           p.C486Vfs*8  ENST00000355995.4  114925256  114925377   \n",
       "5          p.P408Afs*99  ENST00000346208.3    8115814    8115935   \n",
       "..                  ...                ...        ...        ...   \n",
       "354       p.T1445Nfs*10  ENST00000257430.4  112175561  112175682   \n",
       "355       p.R94_P95insA  ENST00000359995.5   74732900   74733021   \n",
       "356     p.S447delinsKPR  ENST00000307771.7   15841195   15841316   \n",
       "357     p.S1043delinsRA  ENST00000298552.3  135771928  135772049   \n",
       "360  p.G3514_G3515insVA  ENST00000268489.5   72821573   72821694   \n",
       "\n",
       "                                          wt_w_context  \\\n",
       "0    GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...   \n",
       "1    GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...   \n",
       "2    CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...   \n",
       "4    CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...   \n",
       "5    AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...   \n",
       "..                                                 ...   \n",
       "354  CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...   \n",
       "355  AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...   \n",
       "356  AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...   \n",
       "357  CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...   \n",
       "360  GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...   \n",
       "\n",
       "                                         alt_w_context  \n",
       "0    GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...  \n",
       "1    GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...  \n",
       "2    CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...  \n",
       "4    CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...  \n",
       "5    AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...  \n",
       "..                                                 ...  \n",
       "354  CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...  \n",
       "355  AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...  \n",
       "356  AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...  \n",
       "357  CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...  \n",
       "360  GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...  \n",
       "\n",
       "[347 rows x 23 columns]"
      ]
     },
     "execution_count": 336,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_ins = ins_new[ins_new['Variant_Classification']!='Intron']\n",
    "exon_ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "#try it with one example\n",
    "#val = ins_new.iloc[0]d\n",
    "\n",
    "def variant_checker(exon_subs):\n",
    "    \n",
    "    for index1, val in exon_subs.iterrows():\n",
    "        hg = val['HGVSp_Short']\n",
    "\n",
    "        tx = val['transcript_id_TRUE']\n",
    "        cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "        start_end_cds = [[i.start, i.end] for i in cds]\n",
    "        strand = db[tx].strand\n",
    "        chrom = val['Chromosome']\n",
    "\n",
    "        if chrom == 'X':\n",
    "            chrom = 22\n",
    "        elif chrom=='Y':\n",
    "            chrom = 23\n",
    "        else:\n",
    "            chrom = int(chrom)-1\n",
    "\n",
    "        chr_seq = records[index_list[chrom]].seq.upper()\n",
    "\n",
    "\n",
    "        wt_dna = ''\n",
    "        codon_locs = []\n",
    "        for i in start_end_cds:\n",
    "            wt_dna += chr_seq[i[0]-1:i[1]]\n",
    "\n",
    "            for j in range(i[0], i[1]+1):\n",
    "                codon_locs.append(j)\n",
    "\n",
    "        #and add in the last/stop codon\n",
    "        #if strand == '+':\n",
    "\n",
    "        # wt_dna += chr_seq[start_end_cds[-1][1]:start_end_cds[-1][1]+3]\n",
    "        #elif strand == '-':\n",
    "        #   wt_dna = chr_seq[start_end_cds[0][0]-3:start_end_cds[0][0]] +  wt_dna\n",
    "\n",
    "        #check location in cds\n",
    "        #get information about the wt/alt sequence\n",
    "        start = val['seq_start']\n",
    "        end = val['seq_end']\n",
    "        wt_seq = val['wt_w_context']\n",
    "        alt_seq = val['alt_w_context']\n",
    "\n",
    "        inc = []\n",
    "        dna_dict = dict(zip(list(range(start, end+1)), list(range(len(wt_seq)))))\n",
    "        for i in range(start, end+1):\n",
    "            if i in codon_locs:\n",
    "                inc.append(i)\n",
    "\n",
    "        if len(inc)==0:\n",
    "            print(index1)\n",
    "            \n",
    "        inc_start = inc[0]\n",
    "        inc_end = inc[-1]\n",
    "\n",
    "        wt_start = dna_dict[inc_start]\n",
    "        wt_end = dna_dict[inc_end] + 1\n",
    "\n",
    "        if wt_end == len(wt_seq):\n",
    "            alt_end = len(alt_seq)\n",
    "        else:\n",
    "            alt_end = wt_end - len(wt_seq)\n",
    "\n",
    "        first_half = codon_locs.index(inc_start)\n",
    "        second_half = codon_locs.index(inc_end)\n",
    "\n",
    "\n",
    "        wt_full_dna = wt_dna[:first_half] + wt_seq[wt_start:wt_end] + wt_dna[second_half+1:]\n",
    "        alt_full_dna = wt_dna[:first_half] + alt_seq[wt_start:alt_end] + wt_dna[second_half+1:]\n",
    "\n",
    "        assert wt_full_dna == wt_dna\n",
    "\n",
    "        if strand == '-':\n",
    "            wt_full_dna = wt_full_dna.reverse_complement()\n",
    "            alt_full_dna = alt_full_dna.reverse_complement()\n",
    "\n",
    "        \n",
    "        wt_dna = str(wt_full_dna)\n",
    "        mut_dna = str(alt_full_dna)\n",
    "\n",
    "        print(len(wt_dna), len(mut_dna))\n",
    "        wt_seq = Bio.Seq.Seq(wt_dna).transcribe().translate()\n",
    "        mut_seq = Bio.Seq.Seq(alt_full_dna).transcribe().translate()\n",
    "\n",
    "        try:\n",
    "            hgvsp = get_hgvsp_classification(wt_dna, mut_dna, wt_seq, mut_seq)\n",
    "        except:\n",
    "            hgvsp = 'undefined'\n",
    "\n",
    "        #if hgvsp != hg[2:]:\n",
    "        print(index1)\n",
    "        print(hgvsp, hg[2:])\n",
    "\n",
    "        return wt_dna, mut_dna\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "dels_new = df_formatter(dels[:400])\n",
    "#coding_muts = ['Frame_Shift_Del'] #'In_Frame_Del',\n",
    "coding_muts = ['In_Frame_Del'] #'In_Frame_Del',\n",
    "\n",
    "exon_dels = dels_new[dels_new['Variant_Classification'].isin(coding_muts)].reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "ins_new = df_formatter(ins[:400])\n",
    "#coding_muts = ['Frame_Shift_Del'] #'In_Frame_Del',\n",
    "coding_muts = ['In_Frame_Ins', 'Frame_Shift_Ins'] #'In_Frame_Del',\n",
    "\n",
    "exon_ins = ins_new[ins_new['Variant_Classification'].isin(coding_muts)].reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Frame_Shift_Ins', 'In_Frame_Ins', 'Intron', 'Nonsense_Mutation',\n",
       "       'Splice_Region', 'Splice_Site'], dtype=object)"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(ins_new['Variant_Classification'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Frame_Shift_Ins', 'In_Frame_Ins')"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'Frame_Shift_Ins', 'In_Frame_Ins'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4623 4624\n",
      "0\n",
      "G646Wfs*12 G646Wfs*12\n",
      "8529 8530\n",
      "1\n",
      "T1556Nfs*3 T1556Nfs*3\n",
      "6855 6856\n",
      "2\n",
      "D1850Gfs*4 D1850Gfs*4\n",
      "1857 1858\n",
      "3\n",
      "C486Vfs*8 C486Vfs*8\n",
      "1329 1330\n",
      "4\n",
      "P408Afs*443 P408Afs*99\n",
      "1329 1330\n",
      "5\n",
      "P408Sfs*443 P408Afs*99\n",
      "6855 6858\n",
      "6\n",
      "E38_A39insG A45dup\n",
      "7242 7243\n",
      "7\n",
      "H2324Pfs*55 H2324Pfs*55\n",
      "6708 6714\n",
      "8\n",
      "Q121_Q122insPA Q130_Q131dup\n",
      "4014 4017\n",
      "9\n",
      "N27_N28insK N28dup\n"
     ]
    }
   ],
   "source": [
    "variant_checker(exon_ins.iloc[0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4014 4017\n",
      "9\n",
      "N27_N28insK N28dup\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'N27_N28insK'"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "w, m = variant_checker(exon_ins.iloc[9:10])\n",
    "\n",
    "wt_dna = w\n",
    "mut_dna = m\n",
    "wt_seq = str(Bio.Seq.Seq(w).transcribe().translate())\n",
    "mut_seq = str(Bio.Seq.Seq(m).transcribe().translate())\n",
    "get_hgvsp_classification(wt_dna, mut_dna, wt_seq, mut_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MASPPRHGPPGPASGDGPNLNNNNNNNNHSVRKCGYLRKQKHGHKRFFVLRGPGAGGDEATAGGGSAPQPPRLEYYESEKKWRSKAGAPKRVIALDCCLNINKRADAKHKYLIALYTKDEYFAVAAENEQEQEGWYRALTDLVSEGRAAAGDAPPAAAPAASCSASLPGALGGSAGAAGAEDSYGLVAPATAAYREVWQVNLKPKGLGQSKNLTGVYRLCLSARTIGFVKLNCEQPSVTLQLMNIRRCGHSDSFFFIEVGRSAVTGPGELWMQADDSVVAQNIHETILEAMKALKELFEFRPRSKSQSSGSSATHPISVPGARRHHHLVNLPPSQTGLVRRSRTDSLAATPPAAKCSSCRVRTASEGDGGAAAGAAAAGARPVSVAGSPLSPGPVRAPLSRSHTLSGGCGGRGSKVALLPAGGALQHSRSMSMPVAHSPPAATSPGSLSSSSGHGSGSYPPPPGPHPPLPHPLHHGPGQRPSSGSASASGSPSDPGFMSLDEYGSSPGDLRAFCSHRSNTPESIAETPPARDGGGGGEFYGYMTMDRPLSHCGRSYRRVSGDAAQDLDRGLRKRTYSLTTPARQRPVPQPSSASLDEYTLMRATFSGSAGRLCPSCPASSPKVAYHPYPEDYGDIEIGSHRSSSSNLGADDGYMPMTPGAALAGSGSGSCRSDDYMPMSPASVSAPKQILQPRAAAAAAAAVPSAGPAGPAPTSAAGRTFPASGGGYKASSPAESSPEDSGYMRMWCGSKLSMEHADGKLLPNGDYLNVSPSDAVTTGTPPDFFSAALHPGGEPLRGVPGCCYSSLPRSYKAPYTCGGDSDQYVLMSSPVGRILEEERLEPQATPGPSQAASAFGAGPTQPPHPVVPSPVRPSGGRPEGFLGQRGRAVRPTRLSLEGLPSLPSMHEYPLPPEPKSPGEYINIDFGEPGARLSPPAPPLLASAASSSSLLSASSPASSLGSGTPGTSSDSRQRSPLSDYMNLDFSSPKSPKPGAPSGHPVGSLDGLLSPEASSPYPPLPPRPSASPSSSLQPPPPPPAPGELYRLPPASAVATAQGPGAASSLSSDTGDNGDYTEMAFGVAATPPQPIAAPPKPEAARVASPTSGVKRLSLMEQVSGVEAFLQASQPPDPHRGAKVIRADPQGGRRRHSSETFSSTTTVTPVSPSFAHNPKRHNSASVENVSLRKSSEGGVGVGPGGGDEPPTSPRQLQPAPPLAPQGRPWTPGQPGGLVGCPGSGGSPMRRETSAGFQNGLNYIAIDVREEPGLPPQPQPPPPPLPQPGDKSSWGRTRSLGGLISAVGVGSTGGGCGGPGPGALPPANTYASIDFLSHHLKEATIVKE\n",
      "MASPPRHGPPGPASGDGPNLNNNNNNNKTHSVRKCGYLRKQKHGHKRFFVLRGPGAGGDEATAGGGSAPQPPRLEYYESEKKWRSKAGAPKRVIALDCCLNINKRADAKHKYLIALYTKDEYFAVAAENEQEQEGWYRALTDLVSEGRAAAGDAPPAAAPAASCSASLPGALGGSAGAAGAEDSYGLVAPATAAYREVWQVNLKPKGLGQSKNLTGVYRLCLSARTIGFVKLNCEQPSVTLQLMNIRRCGHSDSFFFIEVGRSAVTGPGELWMQADDSVVAQNIHETILEAMKALKELFEFRPRSKSQSSGSSATHPISVPGARRHHHLVNLPPSQTGLVRRSRTDSLAATPPAAKCSSCRVRTASEGDGGAAAGAAAAGARPVSVAGSPLSPGPVRAPLSRSHTLSGGCGGRGSKVALLPAGGALQHSRSMSMPVAHSPPAATSPGSLSSSSGHGSGSYPPPPGPHPPLPHPLHHGPGQRPSSGSASASGSPSDPGFMSLDEYGSSPGDLRAFCSHRSNTPESIAETPPARDGGGGGEFYGYMTMDRPLSHCGRSYRRVSGDAAQDLDRGLRKRTYSLTTPARQRPVPQPSSASLDEYTLMRATFSGSAGRLCPSCPASSPKVAYHPYPEDYGDIEIGSHRSSSSNLGADDGYMPMTPGAALAGSGSGSCRSDDYMPMSPASVSAPKQILQPRAAAAAAAAVPSAGPAGPAPTSAAGRTFPASGGGYKASSPAESSPEDSGYMRMWCGSKLSMEHADGKLLPNGDYLNVSPSDAVTTGTPPDFFSAALHPGGEPLRGVPGCCYSSLPRSYKAPYTCGGDSDQYVLMSSPVGRILEEERLEPQATPGPSQAASAFGAGPTQPPHPVVPSPVRPSGGRPEGFLGQRGRAVRPTRLSLEGLPSLPSMHEYPLPPEPKSPGEYINIDFGEPGARLSPPAPPLLASAASSSSLLSASSPASSLGSGTPGTSSDSRQRSPLSDYMNLDFSSPKSPKPGAPSGHPVGSLDGLLSPEASSPYPPLPPRPSASPSSSLQPPPPPPAPGELYRLPPASAVATAQGPGAASSLSSDTGDNGDYTEMAFGVAATPPQPIAAPPKPEAARVASPTSGVKRLSLMEQVSGVEAFLQASQPPDPHRGAKVIRADPQGGRRRHSSETFSSTTTVTPVSPSFAHNPKRHNSASVENVSLRKSSEGGVGVGPGGGDEPPTSPRQLQPAPPLAPQGRPWTPGQPGGLVGCPGSGGSPMRRETSAGFQNGLNYIAIDVREEPGLPPQPQPPPPPLPQPGDKSSWGRTRSLGGLISAVGVGSTGGGCGGPGPGALPPANTYASIDFLSHHLKEATIVKE\n"
     ]
    }
   ],
   "source": [
    "print(str(Bio.Seq.Seq(w).transcribe().translate()))\n",
    "print(str(Bio.Seq.Seq(m).transcribe().translate()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MQPPSLLLLLLLLLLCVSVVRPRGLLCGSFPEPCANGGTCLSLSLGQGTCQCAPGFLGETCQFPDPCQNAQLCQNGGSCQALLPAPLGLPSSPSPLTPSFLCTCLPGFTGERCQAKLEDPCPPSFCSKRGRCHIQASGRPQCSCMPGWTGEQCQLRDFCSANPCVNGGVCLATYPQIQCHCPPGFEGHACERDVNECFQDPGPCPKGTSCHNTLGSFQCLCPVGQEGPRCELRAGPCPPRGCSNGGTCQLMPEKDSTFHLCLCPPGFIGPDCEVNPDNCVSHQCQNGGTCQDGLDTYTCLCPETWTGWDCSEDVDECETQGPPHCRNGGTCQNSAGSFHCVCVSGWGGTSCEENLDDCIAATCAPGSTCIDRVGSFSCLCPPGRTGLLCHLEDMCLSQPCHGDAQCSTNPLTGSTLCLCQPGYSGPTCHQDLDECLMAQQGPSPCEHGGSCLNTPGSFNCLCPPGYTGSRCEADHNECLSQPCHPGSTCLDLLATFHCLCPPGLEGQLCEVETNECASAPCLNHADCHDLLNGFQCICLPGFSGTRCEEDIDECRSSPCANGGQCQDQPGAFHCKCLPGFEGPRCQTEVDECLSDPCPVGASCLDLPGAFFCLCPSGFTGQLCEVPLCAPNLCQPKQICKDQKDKANCLCPDGSPGCAPPEDNCTCHHGHCQRSSCVCDVGWTGPECEAELGGCISAPCAHGGTCYPQPSGYNCTCPTGYTGPTCSEEMTACHSGPCLNGGSCNPSPGGYYCTCPPSHTGPQCQTSTDYCVSAPCFNGGTCVNRPGTFSCLCAMGFQGPRCEGKLRPSCADSPCRNRATCQDSPQGPRCLCPTGYTGGSCQTLMDLCAQKPCPRNSHCLQTGPSFHCLCLQGWTGPLCNLPLSSCQKAALSQGIDVSSLCHNGGLCVDSGPSYFCHCPPGFQGSLCQDHVNPCESRPCQNGATCMAQPSGYLCQCAPGYDGQNCSKELDACQSQPCHNHGTCTPKPGGFHCACPPGFVGLRCEGDVDECLDQPCHPTGTAACHSLANAFYCQCLPGHTGQWCEVEIDPCHSQPCFHGGTCEATAGSPLGFICHCPKGFEGPTCSHRAPSCGFHHCHHGGLCLPSPKPGFPPRCACLSGYGGPDCLTPPAPKGCGPPSPCLYNGSCSETTGLGGPGFRCSCPHSSPGPRCQKPGAKGCEGRSGDGACDAGCSGPGGNWDGGDCSLGVPDPWKGCPSHSRCWLLFRDGQCHPQCDSEECLFDGYDCETPPACTPAYDQYCHDHFHNGHCEKGCNTAECGWDGGDCRPEDGDPEWGPSLALLVVLSPPALDQQLFALARVLSLTLRVGLWVRKDRDGRDMVYPYPGARAEEKLGGTRDPTYQERAAPQTQPLGKETDSLSAGFVVVMGVDLSRCGPDHPASRCPWDPGLLLRFLAAMAAVGALEPLLPGPLLAVHPHAGTAPPANQLPWPVLCSPVAGVILLALGALLVLQLIRRRRREHGALWLPPGFTRRPRTQSAPHRRRPPLGEDSIGLKALKPKAEVDEDGVVMCSGPEEGEEVGQAEETGPPSTCQLWSLSGGCGALPQAAMLTPPQESEMEAPDLDTRGPDGVTPLMSAVCCGEVQSGTFQGAWLGCPEPWEPLLDGGACPQAHTVGTGETPLHLAARFSRPTAARRLLEAGANPNQPDRAGRTPLHAAVAADAREVCQLLLRSRQTAVDARTEDGTTPLMLAARLAVEDLVEELIAAQADVGARDKWGKTALHWAAAVNNARAARSLLQAGADKDAQDNREQTPLFLAAREGAVEVAQLLLGLGAARELRDQAGLAPADVAHQRNHWDLLTLLEGAGPPEARHKATPGREAGPFPRARTVSVSVPPHGGGALPRCRTLSAGAGPRGGGACLQARTWSVDLAARGGGAYSHCRSLSGVGAGGGPTPRGRRFSAGMRGPRPNPAIMRGRYGVAAGRGGRVSTDDWPCDWVALGACGSASNIPIPPPCLTPSPERGSPQLDCGPPALQEMPINQGGEGKK\n"
     ]
    }
   ],
   "source": [
    "print(str(Bio.Seq.Seq(m).transcribe().translate()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGCAGCCCCCTTCACTGCTGCTGCTGCTGCTGCTGCTGCTGCTATGTGTCTCAGTGGTCAGACCCAGAGGGCTGCTGTGTGGGAGTTTCCCAGAACCCTGTGCCAATGGAGGCACCTGCCTGAGCCTGTCTCTGGGACAAGGGACCTGCCAGTGTGCCCCTGGCTTCCTGGGTGAGACGTGCCAGTTTCCTGACCCCTGCCAGAACGCCCAGCTCTGCCAAAATGGAGGCAGCTGCCAAGCCCTGCTTCCCGCTCCCCTAGGGCTCCCCAGCTCTCCCTCTCCATTGACACCCAGCTTCTTGTGCACTTGCCTCCCTGGCTTCACTGGTGAGAGATGCCAGGCCAAGCTTGAAGACCCTTGTCCTCCCTCCTTCTGTTCCAAAAGGGGCCGCTGCCACATCCAGGCCTCGGGCCGCCCACAGTGCTCCTGCATGCCTGGATGGACAGGTGAGCAGTGCCAGCTTCGGGACTTCTGTTCAGCCAACCCATGTGTTAATGGAGGGGTGTGTCTGGCCACATACCCCCAGATCCAGTGCCACTGCCCACCGGGCTTCGAGGGCCATGCCTGTGAACGTGATGTCAACGAGTGCTTCCAGGACCCAGGACCCTGCCCCAAAGGCACCTCCTGCCATAACACCCTGGGCTCCTTCCAGTGCCTCTGCCCTGTGGGGCAGGAGGGTCCACGTTGTGAGCTGCGGGCAGGACCCTGCCCTCCTAGGGGCTGTTCGAATGGGGGCACCTGCCAGCTGATGCCAGAGAAAGACTCCACCTTTCACCTCTGCCTCTGTCCCCCAGGTTTCATAGGCCCAGACTGTGAGGTGAATCCAGACAACTGTGTCAGCCACCAGTGTCAGAATGGGGGCACTTGCCAGGATGGGCTGGACACCTACACCTGCCTCTGCCCAGAAACCTGGACAGGCTGGGACTGCTCCGAAGATGTGGATGAGTGTGAGACCCAGGGTCCCCCTCACTGCAGAAACGGGGGCACCTGCCAGAACTCTGCTGGTAGCTTTCACTGCGTGTGTGTGAGTGGCTGGGGCGGCACAAGCTGTGAGGAGAACCTGGATGACTGTATTGCTGCCACCTGTGCCCCGGGATCCACCTGCATTGACCGGGTGGGCTCTTTCTCCTGCCTCTGCCCACCTGGACGCACAGGACTCCTGTGCCACTTGGAAGACATGTGTCTGAGCCAGCCGTGCCATGGGGATGCCCAATGCAGCACCAACCCCCTCACAGGCTCCACACTCTGCCTGTGTCAGCCTGGCTATTCGGGGCCCACCTGCCACCAGGACCTGGACGAGTGTCTGATGGCCCAGCAAGGCCCAAGTCCCTGTGAACATGGCGGTTCCTGCCTCAACACTCCTGGCTCCTTCAACTGCCTCTGTCCACCTGGCTACACAGGCTCCCGTTGTGAGGCTGATCACAATGAGTGCCTCTCCCAGCCCTGCCACCCAGGAAGCACCTGTCTGGACCTACTTGCCACCTTCCACTGCCTCTGCCCGCCAGGCTTAGAAGGGCAGCTCTGTGAGGTGGAGACCAACGAGTGTGCCTCAGCTCCCTGCCTGAACCACGCGGATTGCCATGACCTGCTCAACGGCTTCCAGTGCATCTGCCTGCCTGGATTCTCCGGCACCCGATGTGAGGAGGATATCGATGAGTGCAGAAGCTCTCCCTGTGCCAATGGTGGGCAGTGCCAGGACCAGCCTGGAGCCTTCCACTGCAAGTGTCTCCCAGGCTTTGAAGGGCCACGCTGTCAAACAGAGGTGGATGAGTGCCTGAGTGACCCATGTCCCGTTGGAGCCAGCTGCCTTGATCTTCCAGGAGCCTTCTTTTGCCTCTGCCCCTCTGGTTTCACAGGCCAGCTCTGTGAGGTTCCCCTGTGTGCTCCCAACCTGTGCCAGCCCAAGCAGATATGTAAGGACCAGAAAGACAAGGCCAACTGCCTCTGTCCTGATGGAAGCCCTGGCTGTGCCCCACCTGAGGACAACTGCACCTGCCACCACGGGCACTGCCAGAGATCCTCATGTGTGTGTGACGTGGGTTGGACGGGGCCAGAGTGTGAGGCAGAGCTAGGGGGCTGCATCTCTGCACCCTGTGCCCATGGGGGGACCTGCTACCCCCAGCCCTCTGGCTACAACTGCACCTGCCCTACAGGCTACACAGGACCCACCTGTAGTGAGGAGATGACAGCTTGTCACTCAGGGCCATGTCTCAATGGCGGCTCCTGCAACCCTAGCCCTGGAGGCTACTACTGCACCTGCCCTCCAAGCCACACAGGGCCCCAGTGCCAAACCAGCACTGACTACTGTGTGTCTGCCCCGTGCTTCAATGGGGGTACCTGTGTGAACAGGCCTGGCACCTTCTCCTGCCTCTGTGCCATGGGCTTCCAGGGCCCGCGCTGTGAGGGAAAGCTCCGCCCCAGCTGTGCAGACAGCCCCTGTAGGAATAGGGCAACCTGCCAGGACAGCCCTCAGGGTCCCCGCTGCCTCTGCCCCACTGGCTACACCGGAGGCAGCTGCCAGACTCTGATGGACTTATGTGCCCAGAAGCCCTGCCCACGCAATTCCCACTGCCTCCAGACTGGGCCCTCCTTCCACTGCTTGTGCCTCCAGGGATGGACCGGGCCTCTCTGCAACCTTCCACTGTCCTCCTGCCAGAAGGCTGCACTGAGCCAAGGCATAGACGTCTCTTCCCTTTGCCACAATGGAGGCCTCTGTGTCGACAGCGGCCCCTCCTATTTCTGCCACTGCCCCCCTGGATTCCAAGGCAGCCTGTGCCAGGATCACGTGAACCCATGTGAGTCCAGGCCTTGCCAGAACGGGGCCACCTGCATGGCCCAGCCCAGTGGGTATCTCTGCCAGTGTGCCCCAGGCTACGATGGACAGAACTGCTCAAAGGAACTCGATGCTTGTCAGTCCCAACCCTGTCACAACCATGGAACCTGTACTCCCAAACCTGGAGGATTCCACTGTGCCTGCCCTCCAGGCTTTGTGGGGCTACGCTGTGAGGGAGACGTGGACGAGTGTCTGGACCAGCCCTGCCACCCCACAGGCACTGCAGCCTGCCACTCTCTGGCCAATGCCTTCTACTGCCAGTGTCTGCCTGGACACACAGGCCAGTGGTGTGAGGTGGAGATAGACCCCTGCCACAGCCAACCCTGCTTTCATGGAGGGACCTGTGAGGCCACAGCAGGATCACCCCTGGGTTTCATCTGCCACTGCCCCAAGGGTTTTGAAGGCCCCACCTGCAGCCACAGGGCCCCTTCCTGCGGCTTCCATCACTGCCACCACGGAGGCCTGTGTCTGCCCTCCCCTAAGCCAGGCTTCCCACCACGCTGTGCCTGCCTCAGTGGCTATGGGGGTCCTGACTGCCTGACCCCACCAGCTCCTAAAGGCTGTGGCCCTCCCTCCCCATGCCTATACAATGGCAGCTGCTCAGAGACCACGGGCTTGGGGGGCCCAGGCTTTCGATGCTCCTGCCCTCACAGCTCTCCAGGGCCCCGGTGTCAGAAACCCGGAGCCAAGGGGTGTGAGGGCAGAAGTGGAGATGGGGCCTGCGATGCTGGCTGCAGTGGCCCGGGAGGAAACTGGGATGGAGGGGACTGCTCTCTGGGAGTCCCAGACCCCTGGAAGGGCTGCCCCTCCCACTCTCGGTGCTGGCTTCTCTTCCGGGACGGGCAGTGCCACCCACAGTGTGACTCTGAAGAGTGTCTGTTTGATGGCTACGACTGTGAGACCCCTCCAGCCTGCACTCCAGCCTATGACCAGTACTGCCATGATCACTTCCACAACGGGCACTGTGAGAAAGGCTGCAACACTGCAGAGTGTGGCTGGGATGGAGGTGACTGCAGGCCTGAAGATGGGGACCCAGAGTGGGGGCCCTCCCTGGCCCTGCTGGTGGTACTGAGCCCCCCAGCCCTAGACCAGCAGCTGTTTGCCCTGGCCCGGGTGCTGTCCCTGACTCTGAGGGTAGGACTCTGGGTAAGGAAGGATCGTGATGGCAGGGACATGGTGTACCCCTATCCTGGGGCCCGGGCTGAAGAAAAGCTAGGAGGAACTCGGGACCCCACCTATCAGGAGAGAGCAGCCCCTCAAACGCAGCCCCTGGGCAAGGAGACCGACTCCCTCAGTGCTGGGTTTGTGGTGGTCATGGGTGTGGATTTGTCCCGCTGTGGCCCTGACCACCCGGCATCCCGCTGTCCCTGGGACCCTGGGCTTCTACTCCGCTTCCTTGCTGCGATGGCTGCAGTGGGAGCCCTGGAGCCCCTGCTGCCTGGACCACTGCTGGCTGTCCACCCTCATGCAGGGACCGCACCCCCTGCCAACCAGCTTCCCTGGCCTGTGCTGTGCTCCCCAGTGGCCGGGGTGATTCTCCTGGCCCTAGGGGCTCTTCTCGTCCTCCAGCTCATCCGGCGTCGACGCCGAGAGCATGGAGCTCTCTGGCTGCCCCCTGGTTTCACTCGACGGCCTCGGACTCAGTCAGCTCCCCACCGACGCCGGCCCCCACTAGGCGAGGACAGCATTGGTCTCAAGGCACTGAAGCCAAAGGCAGAAGTTGATGAGGATGGAGTTGTGATGTGCTCAGGCCCTGAGGAGGGAGAGGAGGTGGGCCAGGCTGAAGAAACAGGCCCACCCTCCACGTGCCAGCTCTGGTCTCTGAGTGGTGGCTGTGGGGCGCTCCCTCAGGCAGCCATGCTAACTCCTCCCCAGGAATCTGAGATGGAAGCCCCTGACCTGGACACCCGTGGACCTGATGGGGTGACACCCCTGATGTCAGCAGTTTGCTGTGGGGAAGTACAGTCCGGGACCTTCCAAGGGGCATGGTTGGGATGTCCTGAGCCCTGGGAACCTCTGCTGGATGGAGGGGCCTGTCCCCAGGCTCACACCGTGGGCACTGGGGAGACCCCCCTGCACCTGGCTGCCCGATTCTCCCGGCCAACCGCTGCCCGCCGCCTCCTTGAGGCTGGAGCCAACCCCAACCAGCCAGACCGGGCAGGGCGCACACCCCTTCATGCTGCTGTGGCTGCTGATGCTCGGGAGGTCTGCCAGCTTCTGCTCCGTAGCAGACAAACTGCAGTGGACGCTCGCACAGAGGACGGGACCACACCCTTGATGCTGGCTGCCAGGCTGGCGGTGGAAGACCTGGTTGAAGAACTGATTGCAGCCCAAGCAGACGTGGGGGCCAGAGATAAATGGGGGAAAACTGCGCTGCACTGGGCTGCTGCCGTGAACAACGCCCGAGCCGCCCGCTCGCTTCTCCAGGCCGGAGCCGATAAAGATGCCCAGGACAACAGGGAGCAGACGCCGCTATTCCTGGCGGCGCGGGAAGGAGCGGTGGAAGTAGCCCAGCTACTGCTGGGGCTGGGGGCAGCCCGAGAGCTGCGGGACCAGGCTGGGCTAGCGCCGGCGGACGTCGCTCACCAACGTAACCACTGGGATCTGCTGACGCTGCTGGAAGGGGCTGGGCCACCAGAGGCCCGTCACAAAGCCACGCCGGGCCGCGAGGCTGGGCCCTTCCCGCGCGCACGGACGGTGTCAGTAAGCGTGCCCCCGCATGGGGGCGGGGCTCTGCCGCGCTGCCGGACGCTGTCAGCCGGAGCAGGCCCTCGTGGGGGCGGAGCTTGTCTGCAGGCTCGGACTTGGTCCGTAGACTTGGCTGCGCGGGGGGGCGGGGCCTATTCTCATTGCCGGAGCCTCTCGGGAGTAGGAGCAGGAGGAGGCCCGACCCCTCGCGGCCGTAGGTTTTCTGCAGGCATGCGCGGGCCTCGGCCCAACCCTGCGATAATGCGAGGAAGATACGGAGTGGCTGCCGGGCGCGGAGGCAGGGTCTCAACGGATGACTGGCCCTGTGATTGGGTGGCCCTGGGAGCTTGCGGTTCTGCCTCCAACATTCCGATCCCGCCTCCTTGCCTTACTCCGTCCCCGGAGCGGGGATCACCTCAACTTGACTGTGGTCCCCCAGCCCTCCAAGAAATGCCCATAAACCAAGGAGGAGAGGGTAAAAAA'"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level_0</th>\n",
       "      <th>index</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>396</td>\n",
       "      <td>3349</td>\n",
       "      <td>8.0</td>\n",
       "      <td>SLC34A2</td>\n",
       "      <td>4</td>\n",
       "      <td>25678148</td>\n",
       "      <td>25678150</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GCT</td>\n",
       "      <td>-</td>\n",
       "      <td>GENIE-DFCI-271759-3854265</td>\n",
       "      <td>NM_006424.2</td>\n",
       "      <td>617.0</td>\n",
       "      <td>13/13</td>\n",
       "      <td>ENST00000382051.3:c.1864_1866del</td>\n",
       "      <td>p.Cys622del</td>\n",
       "      <td>p.C622del</td>\n",
       "      <td>ENST00000382051.3</td>\n",
       "      <td>25678088</td>\n",
       "      <td>25678210</td>\n",
       "      <td>TGCGCTCGCTGAAGCCCTGGGATGCCGTCGTCTCCAAGTTCACCGG...</td>\n",
       "      <td>TGCGCTCGCTGAAGCCCTGGGATGCCGTCGTCTCCAAGTTCACCGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     level_0  index  COUNT Hugo_Symbol Chromosome  Start_Position  \\\n",
       "356      396   3349    8.0     SLC34A2          4        25678148   \n",
       "\n",
       "     End_Position       Consequence Variant_Classification Variant_Type  \\\n",
       "356      25678150  inframe_deletion           In_Frame_Del          DEL   \n",
       "\n",
       "    Reference_Allele Tumor_Seq_Allele2       Tumor_Sample_Barcode  \\\n",
       "356              GCT                 -  GENIE-DFCI-271759-3854265   \n",
       "\n",
       "          RefSeq  Protein_position Exon_Number  \\\n",
       "356  NM_006424.2             617.0       13/13   \n",
       "\n",
       "                                HGVSc        HGVSp HGVSp_Short  \\\n",
       "356  ENST00000382051.3:c.1864_1866del  p.Cys622del   p.C622del   \n",
       "\n",
       "    transcript_id_TRUE  seq_start   seq_end  \\\n",
       "356  ENST00000382051.3   25678088  25678210   \n",
       "\n",
       "                                          wt_w_context  \\\n",
       "356  TGCGCTCGCTGAAGCCCTGGGATGCCGTCGTCTCCAAGTTCACCGG...   \n",
       "\n",
       "                                         alt_w_context  \n",
       "356  TGCGCTCGCTGAAGCCCTGGGATGCCGTCGTCTCCAAGTTCACCGG...  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_dels.iloc[[356]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Frame_Shift_Del', 'In_Frame_Del', 'Intron', 'Splice_Region',\n",
       "       'Splice_Site', 'Translation_Start_Site'], dtype=object)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(dels_new['Variant_Classification'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1915"
      ]
     },
     "execution_count": 341,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(wt_full_dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1857"
      ]
     },
     "execution_count": 342,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(wt_dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>transcript_id_TRUE</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>762.0</td>\n",
       "      <td>ASXL1</td>\n",
       "      <td>20</td>\n",
       "      <td>31022441</td>\n",
       "      <td>31022442</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-003409-1958</td>\n",
       "      <td>NM_015338.5</td>\n",
       "      <td>642.0</td>\n",
       "      <td>13/13</td>\n",
       "      <td>ENST00000375687.4:c.1934dup</td>\n",
       "      <td>p.Gly646TrpfsTer12</td>\n",
       "      <td>p.G646Wfs*12</td>\n",
       "      <td>ENST00000375687.4</td>\n",
       "      <td>31022381</td>\n",
       "      <td>31022502</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "      <td>GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>645.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175952</td>\n",
       "      <td>112175953</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-JHU-00198-00378</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1554.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4666dup</td>\n",
       "      <td>p.Thr1556AsnfsTer3</td>\n",
       "      <td>p.T1556Nfs*3</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175892</td>\n",
       "      <td>112176013</td>\n",
       "      <td>GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...</td>\n",
       "      <td>GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>280.0</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>1</td>\n",
       "      <td>27105930</td>\n",
       "      <td>27105931</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-001750-9089</td>\n",
       "      <td>NM_006015.4</td>\n",
       "      <td>1847.0</td>\n",
       "      <td>20/20</td>\n",
       "      <td>ENST00000324856.7:c.5548dup</td>\n",
       "      <td>p.Asp1850GlyfsTer4</td>\n",
       "      <td>p.D1850Gfs*4</td>\n",
       "      <td>ENST00000324856.7</td>\n",
       "      <td>27105870</td>\n",
       "      <td>27105991</td>\n",
       "      <td>CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...</td>\n",
       "      <td>CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>55</td>\n",
       "      <td>108.0</td>\n",
       "      <td>TCF7L2</td>\n",
       "      <td>10</td>\n",
       "      <td>114925316</td>\n",
       "      <td>114925317</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-DFCI-151674-1525793</td>\n",
       "      <td>NaN</td>\n",
       "      <td>482.0</td>\n",
       "      <td>15/15</td>\n",
       "      <td>ENST00000355995.4:c.1454dup</td>\n",
       "      <td>p.Cys486ValfsTer8</td>\n",
       "      <td>p.C486Vfs*8</td>\n",
       "      <td>ENST00000355995.4</td>\n",
       "      <td>114925256</td>\n",
       "      <td>114925377</td>\n",
       "      <td>CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...</td>\n",
       "      <td>CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>60</td>\n",
       "      <td>99.0</td>\n",
       "      <td>GATA3</td>\n",
       "      <td>10</td>\n",
       "      <td>8115874</td>\n",
       "      <td>8115875</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>GENIE-DFCI-002328-1774</td>\n",
       "      <td>NaN</td>\n",
       "      <td>407.0</td>\n",
       "      <td>6/6</td>\n",
       "      <td>ENST00000346208.3:c.1221dup</td>\n",
       "      <td>p.Pro408AlafsTer99</td>\n",
       "      <td>p.P408Afs*99</td>\n",
       "      <td>ENST00000346208.3</td>\n",
       "      <td>8115814</td>\n",
       "      <td>8115935</td>\n",
       "      <td>AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...</td>\n",
       "      <td>AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>8994</td>\n",
       "      <td>5.0</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175621</td>\n",
       "      <td>112175622</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>GENIE-MSK-P-0048781-T01-IM6</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>1444.0</td>\n",
       "      <td>16/16</td>\n",
       "      <td>ENST00000257430.4:c.4333dup</td>\n",
       "      <td>p.Thr1445AsnfsTer10</td>\n",
       "      <td>p.T1445Nfs*10</td>\n",
       "      <td>ENST00000257430.4</td>\n",
       "      <td>112175561</td>\n",
       "      <td>112175682</td>\n",
       "      <td>CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...</td>\n",
       "      <td>CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>9010</td>\n",
       "      <td>5.0</td>\n",
       "      <td>SRSF2</td>\n",
       "      <td>17</td>\n",
       "      <td>74732960</td>\n",
       "      <td>74732961</td>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>GGC</td>\n",
       "      <td>GENIE-UHN-AGI147550-BM1</td>\n",
       "      <td>NM_001195427.1</td>\n",
       "      <td>94.0</td>\n",
       "      <td>1/3</td>\n",
       "      <td>ENST00000359995.5:c.282_283insGCC</td>\n",
       "      <td>p.Arg94_Pro95insAla</td>\n",
       "      <td>p.R94_P95insA</td>\n",
       "      <td>ENST00000359995.5</td>\n",
       "      <td>74732900</td>\n",
       "      <td>74733021</td>\n",
       "      <td>AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...</td>\n",
       "      <td>AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>9011</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>X</td>\n",
       "      <td>15841255</td>\n",
       "      <td>15841256</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>AGCCGC</td>\n",
       "      <td>GENIE-UHN-AGI974552-BM1</td>\n",
       "      <td>NM_005089.3</td>\n",
       "      <td>447.0</td>\n",
       "      <td>11/11</td>\n",
       "      <td>ENST00000307771.7:c.1339_1340insAGCCGC</td>\n",
       "      <td>p.Ser447delinsLysProArg</td>\n",
       "      <td>p.S447delinsKPR</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15841195</td>\n",
       "      <td>15841316</td>\n",
       "      <td>AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...</td>\n",
       "      <td>AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>9013</td>\n",
       "      <td>5.0</td>\n",
       "      <td>TSC1</td>\n",
       "      <td>9</td>\n",
       "      <td>135771988</td>\n",
       "      <td>135771989</td>\n",
       "      <td>protein_altering_variant</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>GCT</td>\n",
       "      <td>GENIE-UHN-DIVA273992-ARC1</td>\n",
       "      <td>NM_001162426.1</td>\n",
       "      <td>1043.0</td>\n",
       "      <td>23/23</td>\n",
       "      <td>ENST00000298552.3:c.3128_3129insAGC</td>\n",
       "      <td>p.Ser1043delinsArgAla</td>\n",
       "      <td>p.S1043delinsRA</td>\n",
       "      <td>ENST00000298552.3</td>\n",
       "      <td>135771928</td>\n",
       "      <td>135772049</td>\n",
       "      <td>CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...</td>\n",
       "      <td>CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>9094</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ZFHX3</td>\n",
       "      <td>16</td>\n",
       "      <td>72821633</td>\n",
       "      <td>72821634</td>\n",
       "      <td>inframe_insertion</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>GCCACT</td>\n",
       "      <td>GENIE-PROV-2219952519-76fe5c70a7</td>\n",
       "      <td>NM_006885.3</td>\n",
       "      <td>3514.0</td>\n",
       "      <td>10/10</td>\n",
       "      <td>ENST00000268489.5:c.10541_10542insAGTGGC</td>\n",
       "      <td>p.Gly3514_Gly3515insValAla</td>\n",
       "      <td>p.G3514_G3515insVA</td>\n",
       "      <td>ENST00000268489.5</td>\n",
       "      <td>72821573</td>\n",
       "      <td>72821694</td>\n",
       "      <td>GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...</td>\n",
       "      <td>GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>347 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     index  COUNT Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0        4  762.0       ASXL1         20        31022441      31022442   \n",
       "1        5  645.0         APC          5       112175952     112175953   \n",
       "2       21  280.0      ARID1A          1        27105930      27105931   \n",
       "4       55  108.0      TCF7L2         10       114925316     114925317   \n",
       "5       60   99.0       GATA3         10         8115874       8115875   \n",
       "..     ...    ...         ...        ...             ...           ...   \n",
       "354   8994    5.0         APC          5       112175621     112175622   \n",
       "355   9010    5.0       SRSF2         17        74732960      74732961   \n",
       "356   9011    5.0       ZRSR2          X        15841255      15841256   \n",
       "357   9013    5.0        TSC1          9       135771988     135771989   \n",
       "360   9094    5.0       ZFHX3         16        72821633      72821634   \n",
       "\n",
       "                  Consequence Variant_Classification Variant_Type  \\\n",
       "0          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "1          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "2          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "4          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "5          frameshift_variant        Frame_Shift_Ins          INS   \n",
       "..                        ...                    ...          ...   \n",
       "354        frameshift_variant        Frame_Shift_Ins          INS   \n",
       "355         inframe_insertion           In_Frame_Ins          INS   \n",
       "356  protein_altering_variant           In_Frame_Ins          INS   \n",
       "357  protein_altering_variant           In_Frame_Ins          INS   \n",
       "360         inframe_insertion           In_Frame_Ins          INS   \n",
       "\n",
       "    Reference_Allele Tumor_Seq_Allele2              Tumor_Sample_Barcode  \\\n",
       "0                  -                 G            GENIE-DFCI-003409-1958   \n",
       "1                  -                 A             GENIE-JHU-00198-00378   \n",
       "2                  -                 G            GENIE-DFCI-001750-9089   \n",
       "4                  -                 A         GENIE-DFCI-151674-1525793   \n",
       "5                  -                 G            GENIE-DFCI-002328-1774   \n",
       "..               ...               ...                               ...   \n",
       "354                -                 A       GENIE-MSK-P-0048781-T01-IM6   \n",
       "355                -               GGC           GENIE-UHN-AGI147550-BM1   \n",
       "356                -            AGCCGC           GENIE-UHN-AGI974552-BM1   \n",
       "357                -               GCT         GENIE-UHN-DIVA273992-ARC1   \n",
       "360                -            GCCACT  GENIE-PROV-2219952519-76fe5c70a7   \n",
       "\n",
       "             RefSeq  Protein_position Exon_Number  \\\n",
       "0       NM_015338.5             642.0       13/13   \n",
       "1       NM_000038.5            1554.0       16/16   \n",
       "2       NM_006015.4            1847.0       20/20   \n",
       "4               NaN             482.0       15/15   \n",
       "5               NaN             407.0         6/6   \n",
       "..              ...               ...         ...   \n",
       "354     NM_000038.5            1444.0       16/16   \n",
       "355  NM_001195427.1              94.0         1/3   \n",
       "356     NM_005089.3             447.0       11/11   \n",
       "357  NM_001162426.1            1043.0       23/23   \n",
       "360     NM_006885.3            3514.0       10/10   \n",
       "\n",
       "                                        HGVSc                       HGVSp  \\\n",
       "0                 ENST00000375687.4:c.1934dup          p.Gly646TrpfsTer12   \n",
       "1                 ENST00000257430.4:c.4666dup          p.Thr1556AsnfsTer3   \n",
       "2                 ENST00000324856.7:c.5548dup          p.Asp1850GlyfsTer4   \n",
       "4                 ENST00000355995.4:c.1454dup           p.Cys486ValfsTer8   \n",
       "5                 ENST00000346208.3:c.1221dup          p.Pro408AlafsTer99   \n",
       "..                                        ...                         ...   \n",
       "354               ENST00000257430.4:c.4333dup         p.Thr1445AsnfsTer10   \n",
       "355         ENST00000359995.5:c.282_283insGCC         p.Arg94_Pro95insAla   \n",
       "356    ENST00000307771.7:c.1339_1340insAGCCGC     p.Ser447delinsLysProArg   \n",
       "357       ENST00000298552.3:c.3128_3129insAGC       p.Ser1043delinsArgAla   \n",
       "360  ENST00000268489.5:c.10541_10542insAGTGGC  p.Gly3514_Gly3515insValAla   \n",
       "\n",
       "            HGVSp_Short transcript_id_TRUE  seq_start    seq_end  \\\n",
       "0          p.G646Wfs*12  ENST00000375687.4   31022381   31022502   \n",
       "1          p.T1556Nfs*3  ENST00000257430.4  112175892  112176013   \n",
       "2          p.D1850Gfs*4  ENST00000324856.7   27105870   27105991   \n",
       "4           p.C486Vfs*8  ENST00000355995.4  114925256  114925377   \n",
       "5          p.P408Afs*99  ENST00000346208.3    8115814    8115935   \n",
       "..                  ...                ...        ...        ...   \n",
       "354       p.T1445Nfs*10  ENST00000257430.4  112175561  112175682   \n",
       "355       p.R94_P95insA  ENST00000359995.5   74732900   74733021   \n",
       "356     p.S447delinsKPR  ENST00000307771.7   15841195   15841316   \n",
       "357     p.S1043delinsRA  ENST00000298552.3  135771928  135772049   \n",
       "360  p.G3514_G3515insVA  ENST00000268489.5   72821573   72821694   \n",
       "\n",
       "                                          wt_w_context  \\\n",
       "0    GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...   \n",
       "1    GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...   \n",
       "2    CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...   \n",
       "4    CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...   \n",
       "5    AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...   \n",
       "..                                                 ...   \n",
       "354  CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...   \n",
       "355  AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...   \n",
       "356  AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...   \n",
       "357  CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...   \n",
       "360  GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...   \n",
       "\n",
       "                                         alt_w_context  \n",
       "0    GCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCC...  \n",
       "1    GGAATGAAACAGAATCAGAGCAGCCTAAAGAATCAAATGAAAACCA...  \n",
       "2    CTCAGATAAGCTTGGGCGTGTGCAGGAGTTTGACAGTGGCCTGCTG...  \n",
       "4    CTCTCTCCCTTGGCATCTGTGCCCTCTATTCACAGATAACTCTCTC...  \n",
       "5    AGAACAGCTCGTTTAACCCGGCCGCCCTCTCCAGACACATGTCCTC...  \n",
       "..                                                 ...  \n",
       "354  CCAGATAGCCCTGGACAAACCATGCCACCAAGCAGAAGTAAAACAC...  \n",
       "355  AGCCACCGCCCCCGTACCTGCGGGGTGGCGGTCCCCGGCGGCTGTG...  \n",
       "356  AGGGACCGCAGCAGGGACCGCAGCCGGGGCCGGGGCAGCCGGAGCC...  \n",
       "357  CCACCGACTGCTGAATGGGCCTGCCCTCTGGTGTGGGGGTTTCTCT...  \n",
       "360  GCACGCCAGGCAGTGGTACGAGCCGCCGCCGCCGCCGCCGCCGCCA...  \n",
       "\n",
       "[347 rows x 23 columns]"
      ]
     },
     "execution_count": 343,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MKDKQKKKKERTWAEAARLVLENYSDAPMTPKQILQVIEAEGLKEMRSGTSPLA...VVR')"
      ]
     },
     "execution_count": 316,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt_full_dna.transcribe().translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G646Wfs*12\n"
     ]
    }
   ],
   "source": [
    "wt_dna = str(wt_dna)\n",
    "mut_dna = str(alt_full_dna)\n",
    "wt_seq = Bio.Seq.Seq(wt_dna).transcribe().translate()\n",
    "mut_seq = Bio.Seq.Seq(alt_full_dna).transcribe().translate()\n",
    "\n",
    "hgvsp = get_hgvsp_classification(wt_dna, mut_dna, wt_seq, mut_seq)\n",
    "hgvsc = get_hgvsc(wt_dna, mut_dna)\n",
    "\n",
    "print(hgvsp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4624"
      ]
     },
     "execution_count": 333,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(mut_dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/samgould/opt/anaconda3/lib/python3.9/site-packages/Bio/Seq.py:2979: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "mut_seq = Bio.Seq.Seq(alt_full_dna).transcribe().translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CGAAGAAGCGCTTGTGGCCATGCTTCTGCTTGCGCAGGTAGCCGCACTTGCGCACGCTGTGGTTGTTGTTGTTGTTGTTGTTGTTGAGGTTGGGGCCGTCTCCGCTCGCCGGCCCGGGCG'"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CGAAGAAGCGCTTGTGGCCATGCTTCTGCTTGCGCAGGTAGCCGCACTTGCGCACGCTGTGGTTGTTGTTGTTGTTGTTGTTGTTGAGGTTGGGGCCGTCTCCGCTCGCCGGCCCGGGCGGCC\n",
      "CGAAGAAGCGCTTGTGGCCATGCTTCTGCTTGCGCAGGTAGCCGCACTTGCGCACGCTGTTTGTTGTTGTTGTTGTTGTTGTTGAGGTTGGGGCCGTCTCCGCTCGCCGGCCCGGGCGGC\n"
     ]
    }
   ],
   "source": [
    "print(str(chr_seq[start-1:end]))\n",
    "print(val['wt_w_context'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "58"
      ]
     },
     "execution_count": 264,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('GTTGTTGTTGTTGTTGTTGTTGAGGTTGGGGCCGTCTCCGCTCGCCGGCCCGGGCGGC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n"
     ]
    }
   ],
   "source": [
    "for i in range(3,10):\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MASPPRHGPPGPASGDGPNLNNNNNNNNHSVRKCGYLRKQKHGHKRFFVLRGPG...VKE')"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Bio.Seq.Seq(wt_dna.reverse_complement()).transcribe().translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'-'"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val = ins_new.iloc[10]\n",
    "\n",
    "\n",
    "\n",
    "tx = val['transcript_id_TRUE']\n",
    "cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "\n",
    "db[tx].strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Feature CDS (chr13:110408654-110408655[-]) at 0x7fe9d0f4e0d0>,\n",
       " <Feature CDS (chr13:110434389-110438400[-]) at 0x7fe9679b29d0>]"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TAATAAATTATGGCCATGGGAAACATTGTATATTTAGTGTGTGTATTTTGATAA...NNN')"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_seq[31025138:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MKDKQKKKKERTWAEAARLVLENYSDAPMTPKQILQVIEAEGLKEMRSGTSPLA...VR*')"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Bio.Seq.Seq(wt_dna).transcribe().translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31022503"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val['seq_start']\n",
    "val['seq_end']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_hgvsc(mutant_seq, ref_seq):\n",
    "    \"\"\"\n",
    "    Returns the HGVSc notation for a mutant cDNA sequence compared to a reference sequence.\n",
    "\n",
    "    Note: doesn't work for composite mutations. It will only classify the first mutant in the sequence.\n",
    "\n",
    "    Need to make a matching function that converts from hgvsc to mutant_seq given ref_seq and hgvsc\n",
    "    \"\"\"\n",
    "    if mutant_seq == ref_seq:\n",
    "        return 'WT'  # no change, no HGVSc notation\n",
    "    \n",
    "    # find the position(s) of the variant(s) and the type of variation\n",
    "    pos = []\n",
    "    var_type = \"\"\n",
    "    if len(mutant_seq) > len(ref_seq):  # insertion\n",
    "        for i in range(len(ref_seq)):\n",
    "            if mutant_seq[i] != ref_seq[i]:\n",
    "                pos.append(i+1)\n",
    "\n",
    "        var_type = \"ins\" + mutant_seq[pos[0]-1:pos[0] -1 + len(mutant_seq)-len(ref_seq)]\n",
    "        #print(var_type)\n",
    "\n",
    "    elif len(mutant_seq) < len(ref_seq):  # deletion\n",
    "        for i in range(len(mutant_seq)):\n",
    "            if mutant_seq[i] != ref_seq[i]:\n",
    "                pos.append(i+1)\n",
    "\n",
    "        #print(pos)\n",
    "        #print(len(ref_seq)-len(mutant_seq))\n",
    "        var_type = \"del\" + ref_seq[pos[0]-1: pos[0] -1 + len(ref_seq)-len(mutant_seq)]\n",
    "\n",
    "    else:  # substitution; doesn't take into account multiple substitutions...\n",
    "        for i in range(len(mutant_seq)):\n",
    "            if mutant_seq[i] != ref_seq[i]:\n",
    "                pos.append(i+1)\n",
    "        \n",
    "        if len(pos)==1:\n",
    "            var_type = ref_seq[pos[0]-1] + \">\" + mutant_seq[pos[0]-1]\n",
    "        else:\n",
    "            var_type = ref_seq[pos[0]-1:pos[-1]] + \">\" + mutant_seq[pos[0]-1:pos[-1]]\n",
    "    \n",
    "    if var_type[0:3]=='ins':\n",
    "        pos = [i-1 for i in pos]\n",
    "\n",
    "    # generate the HGVSc notation\n",
    "    if len(pos) == 1:\n",
    "        hgvsc = \"c.\" + str(pos[0]) + var_type\n",
    "    else:\n",
    "        if var_type[0:3]=='ins':\n",
    "            hgvsc = \"c.\" + str(pos[0]) + \"_\" + str(pos[0]+1) + var_type\n",
    "        elif var_type[0:3]=='del':\n",
    "            if len(ref_seq)-len(mutant_seq) ==1:\n",
    "                hgvsc = \"c.\" + str(pos[0]) + var_type\n",
    "            else:\n",
    "                hgvsc = \"c.\" + str(pos[0]) + \"_\" + str(pos[0]+len(ref_seq)-len(mutant_seq)-1) + var_type\n",
    "        else:\n",
    "            hgvsc = \"c.\" + str(pos[0]) + \"_\" + str(pos[-1]) + var_type\n",
    "    \n",
    "    return hgvsc\n",
    "\n",
    "def get_hgvsp_classification(wt_dna, mut_dna, wt_seq, mut_seq):\n",
    "    \"\"\"\n",
    "    Given a wildtype protein sequence and a mutant protein sequence, returns the HGVSp classification for the mutant sequence.\n",
    "\n",
    "    Args:\n",
    "        wt_seq (str): The wildtype protein sequence.\n",
    "        mut_seq (str): The mutant protein sequence.\n",
    "\n",
    "    Returns:\n",
    "        str: The HGVSp classification for the mutant protein sequence.\n",
    "\n",
    "\n",
    "    There are a bunch of cases where this fails miserably. The chain of logical statements doesn't necessarily make sense.\n",
    "    \"\"\"\n",
    "    # Find the position of the first amino acid that differs between the wildtype and mutant sequences\n",
    "    pos = None\n",
    "    for i in range(len(wt_seq)):\n",
    "        if i >= len(mut_seq) or wt_seq[i] != mut_seq[i]:\n",
    "            pos = i + 1\n",
    "            break\n",
    "\n",
    "    #WT sequence (no change)\n",
    "    if pos==None:\n",
    "        #check for and identify silent substitution variants\n",
    "\n",
    "        pp = None\n",
    "        for i in range(0,len(mut_dna)-3,3):\n",
    "            mut_cod = mut_dna[i:i+3]\n",
    "            wt_cod = wt_dna[i:i+3]\n",
    "            if mut_cod != wt_cod:\n",
    "                pp = i+1            \n",
    "        \n",
    "        if pp!=None:\n",
    "            pg = pp//3 \n",
    "            return f'{mut_seq[pg]}{pg+1}{mut_seq[pg]}'\n",
    "        else:\n",
    "            return 'WT'\n",
    "\n",
    "    # Check if the mutant sequence contains a stop codon\n",
    "    if '*' in mut_seq and mut_seq.index('*') < pos:\n",
    "        # The mutation is a termination variant\n",
    "        pos = mut_seq.index('*') + 1\n",
    "        return f'{wt_seq[pos-1]}{pos}*'\n",
    "\n",
    "\n",
    "    # Check the difference in sequence length between the wildtype and mutant sequences\n",
    "    diff_len1 = len(mut_dna) - len(wt_dna)\n",
    "\n",
    "    diff_len = len(mut_seq) - len(wt_seq)\n",
    "    #print(diff_len)\n",
    "\n",
    "    if diff_len1 == 0:\n",
    "        # The mutation is a substitution\n",
    "        return f'{wt_seq[pos-1]}{pos}{mut_seq[pos-1]}' \n",
    "        #doesn't take into account potential differences \n",
    "\n",
    "    elif diff_len1 > 0:\n",
    "        # The mutation is an insertion\n",
    "        ins_pos = pos + 1\n",
    "        \n",
    "        #check if it's a frameshift based on mRNA sequence\n",
    "        #not a frameshift if divisible by 3\n",
    "        if diff_len1 % 3 ==0:\n",
    "            \n",
    "            return f'{wt_seq[pos-2]}{pos-1}_{wt_seq[pos-1]}{ins_pos-1}ins{mut_seq[pos-1: pos-1+diff_len]}'\n",
    "        \n",
    "        #otherwise it will be a frameshift\n",
    "        else:\n",
    "            stop_pos = len(mut_seq)\n",
    "            for i in range(pos, len(mut_seq)):\n",
    "                if mut_seq[i]==\"*\":\n",
    "                    stop_pos = i+2-pos\n",
    "                    break\n",
    "\n",
    "            return f'{wt_seq[pos-1]}{pos}{mut_seq[pos-1]}fs*{stop_pos}'\n",
    "\n",
    "    \n",
    "    else:\n",
    "        # The mutation is a deletion or a more complex variant\n",
    "        # Check if the mutation is a pure deletion or a more complex variant\n",
    "        deleted = True\n",
    "        for i in range(pos, pos - diff_len):\n",
    "            if i >= len(mut_seq) or mut_seq[i] != wt_seq[i-1]:\n",
    "                deleted = False\n",
    "                break\n",
    "\n",
    "        if deleted:\n",
    "            # The mutation is a deletion\n",
    "            del_len = -diff_len\n",
    "            return f'{wt_seq[pos-1]}{pos}_{wt_seq[pos+del_len-1]}{pos+del_len-1}del'\n",
    "        \n",
    "        else:\n",
    "            # The mutation is a complex variant\n",
    "            # Find the first stop codon downstream of the mutation\n",
    "            stop_pos = len(mut_seq)\n",
    "            for i in range(pos, len(mut_seq)):\n",
    "                if mut_seq[i]==\"*\":\n",
    "                    stop_pos = i+2-pos\n",
    "                    break\n",
    "\n",
    "            return f'{wt_seq[pos-1]}{pos}{mut_seq[pos-1]}fs*{stop_pos}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
