{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PEGG 2.0\n",
    "\n",
    "This is the real notebook for making the next generation of PEGG...\n",
    "\n",
    "- Start with the flexible input + getting it to work WITH error checking, etc...\n",
    "- I should translate everything into the easiest possible format to work with that is the most flexible...\n",
    "\n",
    "\n",
    "- To do:\n",
    "    - Get the flexible input format in PrimeDesign format working as well!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt \n",
    "import Bio.Seq\n",
    "import gzip\n",
    "from Bio import SeqIO\n",
    "from Bio.Align import PairwiseAligner\n",
    "from Bio.pairwise2 import format_alignment\n",
    "import warnings\n",
    "import re\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "MATCH_SCORE = 1\n",
    "MISMATCH_SCORE = -0.5\n",
    "OPEN_GAP_SCORE = -5\n",
    "EXTEND_GAP_SCORE = -0.1\n",
    "TARGET_END_GAP_SCORE = 0\n",
    "QUERY_END_GAP_SCORE = 0\n",
    "\n",
    "def make_aligner():\n",
    "    aligner = PairwiseAligner()\n",
    "    aligner.mode = 'global'\n",
    "    aligner.match_score = MATCH_SCORE\n",
    "    aligner.mismatch_score = MISMATCH_SCORE\n",
    "    aligner.open_gap_score = OPEN_GAP_SCORE\n",
    "    aligner.extend_gap_score = EXTEND_GAP_SCORE\n",
    "    aligner.target_end_gap_score = TARGET_END_GAP_SCORE\n",
    "    aligner.query_end_gap_score = QUERY_END_GAP_SCORE\n",
    "    return aligner\n",
    "\n",
    "aligner = make_aligner()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class mutation:\n",
    "    def __init__(self, forward_sequence, mut_forward_sequence, chrom=None, genome=None):\n",
    "        \"\"\" \n",
    "        change seq_start and seq_end to correspond with chromosome coordinates\n",
    "        or some other organizational method...\n",
    "        \"\"\"\n",
    "        \n",
    "        #WT sequence\n",
    "        #enforce uppercase\n",
    "        self.forward_sequence = forward_sequence.upper()\n",
    "        rc = str(Bio.Seq.Seq(self.forward_sequence).reverse_complement())\n",
    "        self.rev_comp_sequence = rc\n",
    "\n",
    "        #seq start refers to the WIDE sequence start (i.e. the WT sequence with e.g. 60 nt of context)\n",
    "        self.seq_start = 0\n",
    "        self.seq_end = len(self.forward_sequence)\n",
    "        self.chrom = chrom\n",
    "        self.genome_build = genome\n",
    "\n",
    "        #mutation information\n",
    "        self.mut_forward_sequence = mut_forward_sequence.upper()\n",
    "        rc_mut = str(Bio.Seq.Seq(self.mut_forward_sequence).reverse_complement())\n",
    "        self.mut_rev_comp_sequence = rc_mut\n",
    "\n",
    "        #self.variant_type = \n",
    "        #self.ref = \n",
    "        #self.alt = \n",
    "        #self.alt_size = (positive or negative integer)\n",
    "        \n",
    "\n",
    "        #information about the PAM sequence location\n",
    "        self.PAM_idx_forward = None\n",
    "        self.PAM_idx_rev_comp = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------functions------------\n",
    "def genome_loader(filepath_gz):\n",
    "    \"\"\"\n",
    "    Takes in filepath of human genome (GrCH37 or GrCh38) and returns records and index_list for PEGG parsing.\n",
    "    \n",
    "    Parameters\n",
    "    -----------\n",
    "    filepath_gz\n",
    "        *type = str*\n",
    "        \n",
    "        The filepath to the .gz file holding the reference genome file.\n",
    "    \n",
    "    \"\"\"\n",
    "    #------loading in reference genome and organizing it into a 2-d list by chromosome---------------------\n",
    "    wrong = [\"alternate\", \"unplaced\", \"unlocalized\", \"patch\", \"mitochondrion\"]\n",
    "    filtered = []\n",
    "    chrom_list = []\n",
    "    seq_list = []\n",
    "\n",
    "    with gzip.open(filepath_gz, \"rt\") as handle:\n",
    "        #records = list(SeqIO.parse(handle, \"fasta\")) #about 4 Gb in  memory\n",
    "\n",
    "        for i in SeqIO.parse(handle, \"fasta\"):\n",
    "            ii = i.description\n",
    "            ignore=False \n",
    "\n",
    "            for key in wrong:\n",
    "                if key in ii:\n",
    "                    ignore = True\n",
    "            \n",
    "            if ignore==False:\n",
    "\n",
    "                x = ii.find('chromosome')\n",
    "                chrom = ii[x:].split(' ')[1]\n",
    "                chrom = chrom.split(',')[0]\n",
    "                if chrom not in ['X', 'Y']:\n",
    "                    chrom = int(chrom)\n",
    "                chrom_list.append(chrom)\n",
    "                seq_list.append(i.seq)\n",
    "\n",
    "                filtered.append(ii)\n",
    "\n",
    "  \n",
    "    return dict(zip(chrom_list, seq_list)), filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath_37 = '/Users/samgould/Desktop/FSR Lab/reference files/GRCh37/ncbi-genomes-2022-03-17/GCF_000001405.25_GRCh37.p13_genomic.fna.gz'\n",
    "filepath_38 = '/Users/samgould/Desktop/FSR Lab/reference files/GRCh38/ncbi-genomes-2022-02-23/GCF_000001405.26_GRCh38_genomic.fna.gz'\n",
    "mouse = '/Users/samgould/Desktop/FSR Lab/reference files/GRCm38.p6 (mouse)/ncbi-genomes-2022-04-04/GCF_000001635.26_GRCm38.p6_genomic.fna.gz'\n",
    "chrom_dict, i = genome_loader(filepath_37)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "#starting by bringing in the code for extracting sequence\n",
    "\n",
    "#need to modify this so that it works more generally (the chromosome shit in particular...)\n",
    "\n",
    "def df_formatter(df, chrom_dict, context_size = 60, \n",
    "                 cols_to_save = ['Hugo_Symbol', 'Chromosome', 'Start_Position', 'End_Position', 'Consequence', 'Variant_Classification', 'Variant_Type', 'Reference_Allele', 'Tumor_Seq_Allele2',  'HGVSc', 'HGVSp_Short']):\n",
    "\n",
    "    \"\"\" \n",
    "    Takes in variants (in cBioPortal format!)\n",
    "    and outputs dataframe with WT and ALT oligos with designated context_size\n",
    "    context_size = the amount of nt on either side of the variant e.g. AAA(A/G)AAA = context_size of 3\n",
    "    cols_to_save = information to save about the mutations...(can modify beyond defaults...)\n",
    "    \"\"\"\n",
    "\n",
    "    wt_w_context = []\n",
    "    alt_w_context = []\n",
    "\n",
    "    seq_start = []\n",
    "    seq_end = []\n",
    "\n",
    "    for i, val in df.iterrows():\n",
    "        vt = val['Variant_Type']\n",
    "        s = val['Start_Position']\n",
    "        e = val['End_Position']\n",
    "        ref = val['Reference_Allele']\n",
    "        alt = val['Tumor_Seq_Allele2']\n",
    "        chrom = val['Chromosome']\n",
    "\n",
    "        if chrom not in ['X','Y']:\n",
    "            chrom = int(chrom)\n",
    "       \n",
    "        chr_seq = chrom_dict[chrom].upper()\n",
    "\n",
    "        if vt in ['SNP', 'ONP', 'DNP']:\n",
    "            ref = ref\n",
    "            alt = alt\n",
    "            #assert ref == chr_seq[s-1:e], print(ref, chr_seq[s-1:e])\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        elif vt =='INS':\n",
    "            ref = ''\n",
    "            alt = alt\n",
    "            #left_context = chr_seq[s-1-context_size:s+1] #need to do this since INS reference alleles are blank\n",
    "            left_context = chr_seq[s-1-context_size:s]\n",
    "            right_context = chr_seq[e-1:e+context_size]\n",
    "\n",
    "        elif vt=='DEL':\n",
    "            ref = ref\n",
    "            alt = ''\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        wt_seq = left_context + ref + right_context\n",
    "        alt_seq = left_context + alt + right_context\n",
    "\n",
    "        wt_w_context.append(str(wt_seq))\n",
    "        alt_w_context.append(str(alt_seq))\n",
    "\n",
    "        start = s-context_size\n",
    "        end = e+context_size\n",
    "\n",
    "        seq_start.append(start)\n",
    "        seq_end.append(end)\n",
    "\n",
    "        assert str(chr_seq[start-1:end])==str(wt_seq), print(chr_seq[start-1:end] + '\\n' + str(wt_seq))\n",
    "                                                            \n",
    "\n",
    "    df_new = df[cols_to_save]\n",
    "\n",
    "    df_new['seq_start'] = seq_start\n",
    "    df_new['seq_end'] = seq_end\n",
    "    df_new['wt_w_context'] = wt_w_context\n",
    "    df_new['alt_w_context'] = alt_w_context\n",
    "    df_new = df_new.reset_index()\n",
    "    \n",
    "    return df_new\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/5g/xxjcy2g531n02hnyb6d8qdsr0000gn/T/ipykernel_64096/4206583788.py:1: DtypeWarning: Columns (45,48,88) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  impact = pd.read_csv('/Users/samgould/Desktop/FSR Lab/reference files/2020-06-16-MSK-IMPACT_EDITED.txt', sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "impact = pd.read_csv('/Users/samgould/Desktop/FSR Lab/reference files/2020-06-16-MSK-IMPACT_EDITED.txt', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "p53 = impact[impact['Hugo_Symbol']=='TP53'].drop_duplicates(subset='HGVSc')\n",
    "\n",
    "dels1 = p53[p53['Variant_Type']=='DEL']\n",
    "bad_del_idx = list(dels1[dels1['Tumor_Seq_Allele2']!='-'].index)\n",
    "\n",
    "\n",
    "ins1 = p53[p53['Variant_Type']=='INS']\n",
    "bad_ins_idx = list(ins1[ins1['Reference_Allele']!='-'].index)\n",
    "\n",
    "bad_idxs = bad_del_idx + bad_ins_idx\n",
    "\n",
    "p53_filtered = p53.drop(index=bad_idxs).reset_index().drop(columns=['index'])\n",
    "\n",
    "p53_formatted = df_formatter(p53_filtered, chrom_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>257</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578398</td>\n",
       "      <td>7578398</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.532delC</td>\n",
       "      <td>p.H178Tfs*69</td>\n",
       "      <td>7578338</td>\n",
       "      <td>7578458</td>\n",
       "      <td>AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...</td>\n",
       "      <td>AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>258</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578217</td>\n",
       "      <td>7578240</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GTGTTTCTGTCATCCAAATACTCC</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.609_632del</td>\n",
       "      <td>p.E204_T211del</td>\n",
       "      <td>7578157</td>\n",
       "      <td>7578300</td>\n",
       "      <td>ACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATAGGGCACCACCA...</td>\n",
       "      <td>ACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATAGGGCACCACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>270</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578206</td>\n",
       "      <td>7578207</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>TA</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.642_643delTA</td>\n",
       "      <td>p.H214Qfs*7</td>\n",
       "      <td>7578146</td>\n",
       "      <td>7578267</td>\n",
       "      <td>TCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATA...</td>\n",
       "      <td>TCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>274</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578237</td>\n",
       "      <td>7578253</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>CTCCACACGCAAATTTC</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.596_612del</td>\n",
       "      <td>p.G199Vfs*4</td>\n",
       "      <td>7578177</td>\n",
       "      <td>7578313</td>\n",
       "      <td>CTCAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTT...</td>\n",
       "      <td>CTCAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>276</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577525</td>\n",
       "      <td>7577527</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GAG</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.754_756del</td>\n",
       "      <td>p.L252del</td>\n",
       "      <td>7577465</td>\n",
       "      <td>7577587</td>\n",
       "      <td>AGGCCAGTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCC...</td>\n",
       "      <td>AGGCCAGTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2996</th>\n",
       "      <td>2996</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579319</td>\n",
       "      <td>7579320</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.367_368del</td>\n",
       "      <td>p.T123Lfs*25</td>\n",
       "      <td>7579259</td>\n",
       "      <td>7579380</td>\n",
       "      <td>GATACGGCCAGGCATTGAAGTCTCATGGAAGCCAGCCCCTCAGGGC...</td>\n",
       "      <td>GATACGGCCAGGCATTGAAGTCTCATGGAAGCCAGCCCCTCAGGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2999</th>\n",
       "      <td>2999</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579535</td>\n",
       "      <td>7579556</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>TCAATATCGTCCGGGGACAGCA</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.131_152del</td>\n",
       "      <td>p.M44Nfs*72</td>\n",
       "      <td>7579475</td>\n",
       "      <td>7579616</td>\n",
       "      <td>GGAGCAGCCTCTGGCATTCTGGGAGCTTCATCTGGACCTGGGTCTT...</td>\n",
       "      <td>GGAGCAGCCTCTGGCATTCTGGGAGCTTCATCTGGACCTGGGTCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3000</th>\n",
       "      <td>3000</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577539</td>\n",
       "      <td>7577541</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GGT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.740_742del</td>\n",
       "      <td>p.N247del</td>\n",
       "      <td>7577479</td>\n",
       "      <td>7577601</td>\n",
       "      <td>GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...</td>\n",
       "      <td>GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3002</th>\n",
       "      <td>3002</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577531</td>\n",
       "      <td>7577532</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GG</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.749_750delCC</td>\n",
       "      <td>p.P250Hfs*13</td>\n",
       "      <td>7577471</td>\n",
       "      <td>7577592</td>\n",
       "      <td>GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...</td>\n",
       "      <td>GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3003</th>\n",
       "      <td>3003</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578421</td>\n",
       "      <td>7578440</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GTCATGTGCTGTGACTGCTT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.490_509del</td>\n",
       "      <td>p.K164Gfs*10</td>\n",
       "      <td>7578361</td>\n",
       "      <td>7578500</td>\n",
       "      <td>AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...</td>\n",
       "      <td>AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1327 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "257     257        TP53         17         7578398       7578398   \n",
       "258     258        TP53         17         7578217       7578240   \n",
       "270     270        TP53         17         7578206       7578207   \n",
       "274     274        TP53         17         7578237       7578253   \n",
       "276     276        TP53         17         7577525       7577527   \n",
       "...     ...         ...        ...             ...           ...   \n",
       "2996   2996        TP53         17         7579319       7579320   \n",
       "2999   2999        TP53         17         7579535       7579556   \n",
       "3000   3000        TP53         17         7577539       7577541   \n",
       "3002   3002        TP53         17         7577531       7577532   \n",
       "3003   3003        TP53         17         7578421       7578440   \n",
       "\n",
       "             Consequence Variant_Classification Variant_Type  \\\n",
       "257   frameshift_variant        Frame_Shift_Del          DEL   \n",
       "258     inframe_deletion           In_Frame_Del          DEL   \n",
       "270   frameshift_variant        Frame_Shift_Del          DEL   \n",
       "274   frameshift_variant        Frame_Shift_Del          DEL   \n",
       "276     inframe_deletion           In_Frame_Del          DEL   \n",
       "...                  ...                    ...          ...   \n",
       "2996  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "2999  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "3000    inframe_deletion           In_Frame_Del          DEL   \n",
       "3002  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "3003  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "\n",
       "              Reference_Allele Tumor_Seq_Allele2  \\\n",
       "257                          G                 -   \n",
       "258   GTGTTTCTGTCATCCAAATACTCC                 -   \n",
       "270                         TA                 -   \n",
       "274          CTCCACACGCAAATTTC                 -   \n",
       "276                        GAG                 -   \n",
       "...                        ...               ...   \n",
       "2996                        GT                 -   \n",
       "2999    TCAATATCGTCCGGGGACAGCA                 -   \n",
       "3000                       GGT                 -   \n",
       "3002                        GG                 -   \n",
       "3003      GTCATGTGCTGTGACTGCTT                 -   \n",
       "\n",
       "                                 HGVSc     HGVSp_Short  seq_start  seq_end  \\\n",
       "257        ENST00000269305.4:c.532delC    p.H178Tfs*69    7578338  7578458   \n",
       "258     ENST00000269305.4:c.609_632del  p.E204_T211del    7578157  7578300   \n",
       "270   ENST00000269305.4:c.642_643delTA     p.H214Qfs*7    7578146  7578267   \n",
       "274     ENST00000269305.4:c.596_612del     p.G199Vfs*4    7578177  7578313   \n",
       "276     ENST00000269305.4:c.754_756del       p.L252del    7577465  7577587   \n",
       "...                                ...             ...        ...      ...   \n",
       "2996    ENST00000269305.4:c.367_368del    p.T123Lfs*25    7579259  7579380   \n",
       "2999    ENST00000269305.4:c.131_152del     p.M44Nfs*72    7579475  7579616   \n",
       "3000    ENST00000269305.4:c.740_742del       p.N247del    7577479  7577601   \n",
       "3002  ENST00000269305.4:c.749_750delCC    p.P250Hfs*13    7577471  7577592   \n",
       "3003    ENST00000269305.4:c.490_509del    p.K164Gfs*10    7578361  7578500   \n",
       "\n",
       "                                           wt_w_context  \\\n",
       "257   AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...   \n",
       "258   ACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATAGGGCACCACCA...   \n",
       "270   TCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATA...   \n",
       "274   CTCAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTT...   \n",
       "276   AGGCCAGTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCC...   \n",
       "...                                                 ...   \n",
       "2996  GATACGGCCAGGCATTGAAGTCTCATGGAAGCCAGCCCCTCAGGGC...   \n",
       "2999  GGAGCAGCCTCTGGCATTCTGGGAGCTTCATCTGGACCTGGGTCTT...   \n",
       "3000  GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...   \n",
       "3002  GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...   \n",
       "3003  AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...   \n",
       "\n",
       "                                          alt_w_context  \n",
       "257   AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...  \n",
       "258   ACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATAGGGCACCACCA...  \n",
       "270   TCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATA...  \n",
       "274   CTCAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTT...  \n",
       "276   AGGCCAGTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCC...  \n",
       "...                                                 ...  \n",
       "2996  GATACGGCCAGGCATTGAAGTCTCATGGAAGCCAGCCCCTCAGGGC...  \n",
       "2999  GGAGCAGCCTCTGGCATTCTGGGAGCTTCATCTGGACCTGGGTCTT...  \n",
       "3000  GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...  \n",
       "3002  GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...  \n",
       "3003  AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...  \n",
       "\n",
       "[1327 rows x 16 columns]"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p53_formatted[p53_formatted['Variant_Type']=='DEL']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGCGCCTCACAACCTCCGTCATGTGCTGTGACTGCTTGTAGATGGCCATGGCGCG\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGT-GGGGCAGCGCCTCACAACCTCCGTCATGTGCTGTGACTGCTTGTAGATGGCCATGGCGCG\n",
      "\n"
     ]
    }
   ],
   "source": [
    "i=257\n",
    "alignments = aligner.align(p53_formatted.iloc[i]['wt_w_context'], p53_formatted.iloc[i]['alt_w_context'])\n",
    "print(alignments[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "((0, 60), (61, 121))\n",
      "((0, 60), (60, 120))\n"
     ]
    }
   ],
   "source": [
    "WT_align = alignments[0].aligned[0]\n",
    "alt_align = alignments[0].aligned[1]\n",
    "\n",
    "print(WT_align)\n",
    "print(alt_align)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((0, 60), (61, 121))"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "WT_align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGT\n",
      "GGGGCAGCGCCTCACAACCTCCGTCATGTGCTGTGACTGCTTGTAGATGGCCATGGCGCG\n"
     ]
    }
   ],
   "source": [
    "s1 = p53_formatted.iloc[i]['wt_w_context']\n",
    "\n",
    "print(s1[WT_align[0][0]:WT_align[0][1]])\n",
    "print(s1[WT_align[1][0]:WT_align[1][1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#figure out the best way to encode this information..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PAM searching\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PAM_finder(seq, PAM):\n",
    "    \"\"\" \n",
    "    Finds indeces of PAM sequences\n",
    "    Formatting: e.g. \"NGG\"\n",
    "    N = [A|T|C|G]\n",
    "    R = [A|G]\n",
    "    Y = [C|T]\n",
    "    S = [G|C]\n",
    "    W = [A|T]\n",
    "    K = [G|T]\n",
    "    M = [A|C]\n",
    "    B = [C|G|T]\n",
    "    D = [A|G|T]\n",
    "    H = [A|C|T]\n",
    "    V = [A|C|G]\n",
    "    \"\"\"\n",
    "    PAM_dict = {\"A\":\"A\",\n",
    "                \"G\":\"G\",\n",
    "                \"C\":\"C\",\n",
    "                \"T\":\"T\",\n",
    "                \"N\":\"[A|T|C|G]\",\n",
    "                \"R\":\"[A|G]\",\n",
    "                \"Y\":\"[C|T]\",\n",
    "                \"S\":\"[G|C]\",\n",
    "                \"W\":\"[A|T]\",\n",
    "                \"K\":\"[G|T]\",\n",
    "                \"M\":\"[A|C]\",\n",
    "                \"B\":\"[C|G|T]\",\n",
    "                \"D\":\"[A|G|T]\",\n",
    "                \"H\":\"[A|C|T]\",\n",
    "                \"V\":\"[A|C|G]\"}\n",
    "\n",
    "    PAM_new = \"\"\n",
    "    for i in PAM:\n",
    "        PAM_new += PAM_dict[i]\n",
    "\n",
    "    p = re.compile(PAM_new)\n",
    "\n",
    "    iterator = p.finditer(seq)\n",
    "    PAM_holder = []\n",
    "    for match in iterator:\n",
    "        PAM_holder.append(match.span())\n",
    "\n",
    "    return PAM_holder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGGACAGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATAGGAAAAGAGAAGC\n",
      "CGCC\n",
      "GACA\n",
      "GGCA\n",
      "AACA\n",
      "CGCA\n",
      "AGCT\n",
      "TACC\n",
      "TACT\n"
     ]
    }
   ],
   "source": [
    "seq = p53_formatted.iloc[0]['wt_w_context']\n",
    "print(seq)\n",
    "PAM = \"NRCH\"\n",
    "PAM_holder = PAM_finder(seq, PAM)\n",
    "\n",
    "for i in PAM_holder:\n",
    "    print(seq[i[0]:i[1]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Error checking\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Align import PairwiseAligner\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sensor_outcome_corrected(p53_lib, i):\n",
    "    #p53_lib = df containing pegRNA + target sequence + other info\n",
    "    #i = index\n",
    "\n",
    "    sensor_true = p53_lib.iloc[i]['target sequence']\n",
    "\n",
    "    #let's quantify the homology overhang as well...\n",
    "    PAM_strand = p53_lib.iloc[i]['PAM strand']\n",
    "    var_type = p53_lib.iloc[i]['variant_type']\n",
    "\n",
    "\n",
    "    if PAM_strand=='+':\n",
    "        sensor = Bio.Seq.Seq(p53_lib.iloc[i]['target sequence']).reverse_complement()\n",
    "\n",
    "    elif PAM_strand=='-':\n",
    "        sensor = p53_lib.iloc[i]['target sequence']\n",
    "\n",
    "    pbs_length = p53_lib.iloc[i]['PBS length']\n",
    "    pbs = p53_lib.iloc[i]['PBS_RTT_5to3'][-pbs_length:]\n",
    "    rtt = p53_lib.iloc[i]['PBS_RTT_5to3'][:-(pbs_length-1)] #FIXING [:-0] indexing issue...\n",
    "\n",
    "    #pbs_rtt = p53_lib.iloc[i]['PBS_RTT_5to3']\n",
    "\n",
    "    pbs_find = sensor.find(pbs)\n",
    "\n",
    "    #first compute the portion before the mutation for computing the right flank\n",
    "    for kk, val in enumerate(reversed(rtt)):\n",
    "        idx = pbs_find-kk\n",
    "        if val!= sensor[idx]: #find first mismatch\n",
    "            mut_start=kk\n",
    "            break\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "\n",
    "    try: mut_start\n",
    "    except: #case where's theres no identifiable change (because of homology overhang being too short/matching coincidentally...)\n",
    "        return sensor_true, sensor_true, 0, \"no homology overhang\"\n",
    "\n",
    "    #alignments = aligner.align(pbs, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(rtt, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #now we have the right flank...\n",
    "    right_flank = rtt[-mut_start:-1] + sensor[pbs_find:]\n",
    "\n",
    "    #alignments = aligner.align(right_flank, sensor)\n",
    "    #print(alignments[0])\n",
    "    \n",
    "    if var_type in ['SNP', 'ONP']:\n",
    "\n",
    "        matching_arr = np.zeros(len(rtt[:-mut_start]))\n",
    "\n",
    "        for kk, val in enumerate(reversed(rtt[:-mut_start])):\n",
    "            idx = pbs_find-mut_start-kk\n",
    "\n",
    "            if val== sensor[idx]: #find regions of homology\n",
    "                matching_arr[kk]+=1\n",
    "                \n",
    "            else:\n",
    "                continue\n",
    "        \n",
    "        #based on values in match array, determine where mutation vs. homology overhang is...\n",
    "        end_mut = np.where(matching_arr==0)[0][-1]\n",
    "\n",
    "        mut_allele = rtt[-(mut_start+end_mut+1):-mut_start]\n",
    "        homology_overhang = rtt[:-(mut_start+end_mut+1)]\n",
    "        #this formulation will NOT catch situations where the homology overhang doesn't match???????\n",
    "        #(though this shouldn't be an issue for SNPs/ONPs)\n",
    "\n",
    "        left_flank = sensor[:(pbs_find-(len(rtt)-1))]\n",
    "\n",
    "    elif var_type=='DEL':\n",
    "\n",
    "        #find the size of the deletion by shifting and checking for matches...\n",
    "        #will allow for detection of insufficient homology overhang...\n",
    "        length_remaining = len(rtt[:-mut_start])\n",
    "        homology_overhang = rtt[:-mut_start]\n",
    "        mut_allele = ''\n",
    "\n",
    "        start_idx = pbs_find-mut_start\n",
    "        for i in range(len(sensor[:start_idx])): #check rest of the sensor...\n",
    "        \n",
    "            #sliding window to find size of deletion/perfect match...\n",
    "            sensor_a = sensor[start_idx-i-length_remaining:start_idx-i]\n",
    "\n",
    "            if sensor_a==homology_overhang: #once match is found, record and break...\n",
    "                size_del = i\n",
    "                break\n",
    "        \n",
    "            else:\n",
    "                continue\n",
    "\n",
    "        #error where the full homology overhang region not included in sensor...\n",
    "        try: size_del\n",
    "        except: \n",
    "            return sensor_true, sensor_true, len(homology_overhang), 'sensor too short' #no excepted editing, return WT sensor\n",
    "\n",
    "        left_flank = sensor[:start_idx-size_del-length_remaining]\n",
    "\n",
    "    elif var_type=='INS':\n",
    "\n",
    "        length_remaining = len(rtt[:-mut_start])\n",
    "        start_idx = pbs_find-mut_start+1\n",
    "\n",
    "        for i in range(length_remaining):\n",
    "            rtt_region = rtt[:-(mut_start+i)]\n",
    "            sensor_a = sensor[start_idx-len(rtt_region):start_idx]\n",
    "\n",
    "            #print('rt: ' + rtt_region)\n",
    "            #print('s: ' + sensor_a)\n",
    "\n",
    "            if sensor_a==rtt_region:\n",
    "                mut_allele = rtt[-(mut_start+i):-mut_start]\n",
    "                break\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "        #in case of no homology overhang...\n",
    "        try: mut_allele\n",
    "        except: \n",
    "            return sensor_true, sensor_true, 0, 'no homology overhang' #no excepted editing, return WT sensor\n",
    "\n",
    "        homology_overhang = sensor_a\n",
    "        left_flank = sensor[:start_idx-len(rtt_region)]\n",
    "\n",
    "    edited_sensor = left_flank + homology_overhang + mut_allele + right_flank      \n",
    "\n",
    "    if PAM_strand=='+':\n",
    "        edited_sensor_true = str(Bio.Seq.Seq(edited_sensor).reverse_complement())\n",
    "\n",
    "    elif PAM_strand=='-':\n",
    "        edited_sensor_true=edited_sensor\n",
    "\n",
    "\n",
    "\n",
    "    #alignments = aligner.align(rtt[-mut_start:], sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(right_flank, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(homology_overhang, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(edited_sensor, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(edited_sensor_true, sensor_true)\n",
    "    #print(alignments[0])\n",
    "\n",
    "\n",
    "    return sensor_true, edited_sensor_true, len(homology_overhang), 'no error'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "p53_lib = ....\n",
    "\n",
    "#running it for every mutation in the p53 library\n",
    "edit_sens = []\n",
    "homo = []\n",
    "idx_holder = []\n",
    "classif = []\n",
    "#del_idx = p53_del.index\n",
    "\n",
    "#ins_idx = p53_lib[p53_lib['variant_type']=='SNP'].index\n",
    "\n",
    "error = []\n",
    "for i in range(len(p53_lib)):\n",
    "    try:\n",
    "        s, s_e, homo2, m = sensor_outcome_corrected(p53_lib, i)\n",
    "        idx_holder.append(i)\n",
    "        edit_sens.append(s_e)\n",
    "        homo.append(homo2)\n",
    "        classif.append(m)\n",
    "    \n",
    "    except:\n",
    "        error.append(i)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
