{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PEGG 2.0\n",
    "\n",
    "This is the real notebook for making the next generation of PEGG...\n",
    "\n",
    "- Start with the flexible input + getting it to work WITH error checking, etc...\n",
    "- I should translate everything into the easiest possible format to work with that is the most flexible...\n",
    "\n",
    "\n",
    "- To do:\n",
    "    - Get the flexible input format in PrimeDesign format working as well!\n",
    "    - Have it detect what the input format is??\n",
    "    - Or not?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt \n",
    "import Bio.Seq\n",
    "import gzip\n",
    "from Bio import SeqIO\n",
    "from Bio.Align import PairwiseAligner\n",
    "from Bio.pairwise2 import format_alignment\n",
    "import warnings\n",
    "import regex as re\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "MATCH_SCORE = 1\n",
    "MISMATCH_SCORE = -0.5\n",
    "OPEN_GAP_SCORE = -5\n",
    "EXTEND_GAP_SCORE = -0.1\n",
    "TARGET_END_GAP_SCORE = 0\n",
    "QUERY_END_GAP_SCORE = 0\n",
    "\n",
    "def make_aligner():\n",
    "    aligner = PairwiseAligner()\n",
    "    aligner.mode = 'global'\n",
    "    aligner.match_score = MATCH_SCORE\n",
    "    aligner.mismatch_score = MISMATCH_SCORE\n",
    "    aligner.open_gap_score = OPEN_GAP_SCORE\n",
    "    aligner.extend_gap_score = EXTEND_GAP_SCORE\n",
    "    aligner.target_end_gap_score = TARGET_END_GAP_SCORE\n",
    "    aligner.query_end_gap_score = QUERY_END_GAP_SCORE\n",
    "    return aligner\n",
    "\n",
    "aligner = make_aligner()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "class mutation:\n",
    "    def __init__(self, wt_w_context, alt_w_context, left_seq, right_seq, var_type, ref_seq, alt_seq, chrom=None, genome=None):\n",
    "        \"\"\" \n",
    "        change seq_start and seq_end to correspond with chromosome coordinates\n",
    "        or some other organizational method...\n",
    "        \"\"\"\n",
    "        \n",
    "        #WT sequence\n",
    "        self.wt_forward = wt_w_context.upper()\n",
    "        rc = str(Bio.Seq.Seq(self.wt_forward).reverse_complement())\n",
    "        self.wt_rc = rc\n",
    "\n",
    "        #ALT sequence\n",
    "        self.mut_forward = alt_w_context.upper()\n",
    "        rc_mut = str(Bio.Seq.Seq(self.mut_forward).reverse_complement())\n",
    "        self.mut_rc = rc_mut\n",
    "\n",
    "        #seq start refers to the WIDE sequence start (i.e. the WT sequence with e.g. 60 nt of context)\n",
    "        self.seq_start = 0\n",
    "        self.seq_end = len(self.wt_forward)\n",
    "        self.chrom = chrom\n",
    "        self.genome_build = genome\n",
    "\n",
    "        #and splitting it up into the constituent components\n",
    "        self.left_seq = left_seq.upper()\n",
    "        self.right_seq = right_seq.upper()\n",
    "        self.ref_seq = ref_seq.upper()\n",
    "        self.alt_seq = alt_seq.upper()\n",
    "\n",
    "        #and do the same to get it into the RC format\n",
    "\n",
    "        #RIGHT/LEFT ARE SWAPPED ON REVERSE COMPLEMENT\n",
    "        self.left_seq_rc = str(Bio.Seq.Seq(right_seq).reverse_complement())\n",
    "        self.right_seq_rc = str(Bio.Seq.Seq(left_seq).reverse_complement())\n",
    "        self.ref_seq_rc = str(Bio.Seq.Seq(ref_seq).reverse_complement())\n",
    "        self.alt_seq_rc = str(Bio.Seq.Seq(alt_seq).reverse_complement())\n",
    "\n",
    "        self.variant_type = var_type\n",
    "        #self.alt_size = (positive or negative integer)\n",
    "        \n",
    "        #information about the PAM sequence location\n",
    "        self.PAM_idx_forward = None\n",
    "        self.PAM_idx_rc = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAAACCCTCTCTC\n",
      "GAGAGAGGGTTTT\n",
      "GAGAGAGGGTTTT\n"
     ]
    }
   ],
   "source": [
    "left_seq = 'AAAA'\n",
    "ref_seq = 'TGT'\n",
    "alt_seq = 'CCC'\n",
    "right_seq = 'TCTCTC'\n",
    "\n",
    "wt_w_context = left_seq + ref_seq + right_seq\n",
    "alt_w_context = left_seq + alt_seq + right_seq\n",
    "print(alt_w_context)\n",
    "print(Bio.Seq.Seq(alt_w_context).reverse_complement())\n",
    "\n",
    "print(Bio.Seq.Seq(right_seq).reverse_complement() + Bio.Seq.Seq(alt_seq).reverse_complement() + Bio.Seq.Seq(left_seq).reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------functions------------\n",
    "def genome_loader(filepath_gz):\n",
    "    \"\"\"\n",
    "    Takes in filepath of human genome (GrCH37 or GrCh38) and returns records and index_list for PEGG parsing.\n",
    "    \n",
    "    Parameters\n",
    "    -----------\n",
    "    filepath_gz\n",
    "        *type = str*\n",
    "        \n",
    "        The filepath to the .gz file holding the reference genome file.\n",
    "    \n",
    "    \"\"\"\n",
    "    #------loading in reference genome and organizing it into a 2-d list by chromosome---------------------\n",
    "    wrong = [\"alternate\", \"unplaced\", \"unlocalized\", \"patch\", \"mitochondrion\"]\n",
    "    filtered = []\n",
    "    chrom_list = []\n",
    "    seq_list = []\n",
    "\n",
    "    with gzip.open(filepath_gz, \"rt\") as handle:\n",
    "        #records = list(SeqIO.parse(handle, \"fasta\")) #about 4 Gb in  memory\n",
    "\n",
    "        for i in SeqIO.parse(handle, \"fasta\"):\n",
    "            ii = i.description\n",
    "            ignore=False \n",
    "\n",
    "            for key in wrong:\n",
    "                if key in ii:\n",
    "                    ignore = True\n",
    "            \n",
    "            if ignore==False:\n",
    "\n",
    "                x = ii.find('chromosome')\n",
    "                chrom = ii[x:].split(' ')[1]\n",
    "                chrom = chrom.split(',')[0]\n",
    "                if chrom not in ['X', 'Y']:\n",
    "                    chrom = int(chrom)\n",
    "                chrom_list.append(chrom)\n",
    "                seq_list.append(i.seq)\n",
    "\n",
    "                filtered.append(ii)\n",
    "\n",
    "  \n",
    "    return dict(zip(chrom_list, seq_list)), filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath_37 = '/Users/samgould/Desktop/FSR Lab/reference files/GRCh37/ncbi-genomes-2022-03-17/GCF_000001405.25_GRCh37.p13_genomic.fna.gz'\n",
    "filepath_38 = '/Users/samgould/Desktop/FSR Lab/reference files/GRCh38/ncbi-genomes-2022-02-23/GCF_000001405.26_GRCh38_genomic.fna.gz'\n",
    "mouse = '/Users/samgould/Desktop/FSR Lab/reference files/GRCm38.p6 (mouse)/ncbi-genomes-2022-04-04/GCF_000001635.26_GRCm38.p6_genomic.fna.gz'\n",
    "chrom_dict, i = genome_loader(filepath_37)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#starting by bringing in the code for extracting sequence\n",
    "\n",
    "#need to modify this so that it works more generally (the chromosome shit in particular...)\n",
    "\n",
    "def df_formatter(df, chrom_dict, context_size = 120, \n",
    "                 cols_to_save = ['Hugo_Symbol', 'Chromosome', 'Start_Position', 'End_Position', 'Consequence', 'Variant_Classification', 'Variant_Type', 'Reference_Allele', 'Tumor_Seq_Allele2',  'HGVSc', 'HGVSp_Short']):\n",
    "\n",
    "    \"\"\" \n",
    "    Takes in variants (in cBioPortal format!)\n",
    "    and outputs dataframe with WT and ALT oligos with designated context_size\n",
    "    context_size = the amount of nt on either side of the variant e.g. AAA(A/G)AAA = context_size of 3\n",
    "    cols_to_save = information to save about the mutations...(can modify beyond defaults...)\n",
    "    \"\"\"\n",
    "\n",
    "    wt_w_context = []\n",
    "    alt_w_context = []\n",
    "    left_context_list = []\n",
    "    right_context_list = []\n",
    "    ref_allele = []\n",
    "    alt_allele = []\n",
    "\n",
    "    seq_start = []\n",
    "    seq_end = []\n",
    "\n",
    "    for i, val in df.iterrows():\n",
    "        vt = val['Variant_Type']\n",
    "        s = val['Start_Position']\n",
    "        e = val['End_Position']\n",
    "        ref = val['Reference_Allele']\n",
    "        alt = val['Tumor_Seq_Allele2']\n",
    "        chrom = val['Chromosome']\n",
    "\n",
    "        if chrom not in ['X','Y']:\n",
    "            chrom = int(chrom)\n",
    "       \n",
    "        chr_seq = chrom_dict[chrom].upper()\n",
    "\n",
    "        if vt in ['SNP', 'ONP', 'DNP']:\n",
    "            ref = ref\n",
    "            alt = alt\n",
    "            #assert ref == chr_seq[s-1:e], print(ref, chr_seq[s-1:e])\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        elif vt =='INS':\n",
    "            ref = ''\n",
    "            alt = alt\n",
    "            #left_context = chr_seq[s-1-context_size:s+1] #need to do this since INS reference alleles are blank\n",
    "            left_context = chr_seq[s-1-context_size:s]\n",
    "            right_context = chr_seq[e-1:e+context_size]\n",
    "\n",
    "        elif vt=='DEL':\n",
    "            ref = ref\n",
    "            alt = ''\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        wt_seq = left_context + ref + right_context\n",
    "        alt_seq = left_context + alt + right_context\n",
    "\n",
    "        wt_w_context.append(str(wt_seq))\n",
    "        alt_w_context.append(str(alt_seq))\n",
    "        left_context_list.append(str(left_context))\n",
    "        right_context_list.append(str(right_context))\n",
    "        ref_allele.append(str(ref))\n",
    "        alt_allele.append(str(alt))\n",
    "\n",
    "        start = s-context_size\n",
    "        end = e+context_size\n",
    "\n",
    "        seq_start.append(start)\n",
    "        seq_end.append(end)\n",
    "\n",
    "        assert str(chr_seq[start-1:end])==str(wt_seq), print(chr_seq[start-1:end] + '\\n' + str(wt_seq))\n",
    "                                                            \n",
    "\n",
    "    df_new = df[cols_to_save]\n",
    "\n",
    "    df_new['seq_start'] = seq_start\n",
    "    df_new['seq_end'] = seq_end\n",
    "    df_new['wt_w_context'] = wt_w_context\n",
    "    df_new['alt_w_context'] = alt_w_context\n",
    "    df_new['left_context'] = left_context_list\n",
    "    df_new['right_context'] = right_context_list\n",
    "    df_new['REF'] = ref_allele\n",
    "    df_new['ALT'] = alt_allele\n",
    "    df_new = df_new.reset_index()\n",
    "    \n",
    "    return df_new\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "impact = pd.read_csv('/Users/samgould/Desktop/FSR Lab/reference files/2020-06-16-MSK-IMPACT_EDITED.txt', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [],
   "source": [
    "p53 = impact[impact['Hugo_Symbol']=='TP53'].drop_duplicates(subset='HGVSc')\n",
    "\n",
    "#or change these to INDELs\n",
    "dels1 = p53[p53['Variant_Type']=='DEL']\n",
    "bad_del_idx = list(dels1[dels1['Tumor_Seq_Allele2']!='-'].index)\n",
    "\n",
    "\n",
    "ins1 = p53[p53['Variant_Type']=='INS']\n",
    "bad_ins_idx = list(ins1[ins1['Reference_Allele']!='-'].index)\n",
    "\n",
    "bad_idxs = bad_del_idx + bad_ins_idx\n",
    "\n",
    "p53_filtered = p53.drop(index=bad_idxs).reset_index().drop(columns=['index'])\n",
    "\n",
    "p53_formatted = df_formatter(p53_filtered, chrom_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "      <th>left_context</th>\n",
       "      <th>right_context</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.818G&gt;A</td>\n",
       "      <td>p.R273H</td>\n",
       "      <td>7577060</td>\n",
       "      <td>7577180</td>\n",
       "      <td>CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAG...</td>\n",
       "      <td>CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAG...</td>\n",
       "      <td>CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAG...</td>\n",
       "      <td>GCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGAT...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577108</td>\n",
       "      <td>7577108</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.830G&gt;T</td>\n",
       "      <td>p.C277F</td>\n",
       "      <td>7577048</td>\n",
       "      <td>7577168</td>\n",
       "      <td>TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCC...</td>\n",
       "      <td>TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCC...</td>\n",
       "      <td>TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCC...</td>\n",
       "      <td>AGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578239</td>\n",
       "      <td>7578239</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.610G&gt;T</td>\n",
       "      <td>p.E204*</td>\n",
       "      <td>7578179</td>\n",
       "      <td>7578299</td>\n",
       "      <td>CAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTTCT...</td>\n",
       "      <td>CAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTTCT...</td>\n",
       "      <td>CAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTTCT...</td>\n",
       "      <td>CACACGCAAATTTCCTTCCACTCGGATAAGATGCTGAGGAGGGGCC...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577121</td>\n",
       "      <td>7577121</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.817C&gt;T</td>\n",
       "      <td>p.R273C</td>\n",
       "      <td>7577061</td>\n",
       "      <td>7577181</td>\n",
       "      <td>CTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGG...</td>\n",
       "      <td>CTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGG...</td>\n",
       "      <td>CTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGG...</td>\n",
       "      <td>CACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATA...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577506</td>\n",
       "      <td>7577506</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.775G&gt;T</td>\n",
       "      <td>p.D259Y</td>\n",
       "      <td>7577446</td>\n",
       "      <td>7577566</td>\n",
       "      <td>GCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGCAAGTGGC...</td>\n",
       "      <td>GCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGCAAGTGGC...</td>\n",
       "      <td>GCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGCAAGTGGC...</td>\n",
       "      <td>TTCCAGTGTGATGATGGTGAGGATGGGCCTCCGGTTCATGCCGCCC...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3000</th>\n",
       "      <td>3000</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577539</td>\n",
       "      <td>7577541</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GGT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.740_742del</td>\n",
       "      <td>p.N247del</td>\n",
       "      <td>7577479</td>\n",
       "      <td>7577601</td>\n",
       "      <td>GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...</td>\n",
       "      <td>GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...</td>\n",
       "      <td>GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...</td>\n",
       "      <td>TCATGCCGCCCATGCAGGAACTGTTACACATGTAGTTGTAGTGGAT...</td>\n",
       "      <td>GGT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3001</th>\n",
       "      <td>3001</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578438</td>\n",
       "      <td>7578438</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>ENST00000269305.4:c.492G&gt;C</td>\n",
       "      <td>p.K164N</td>\n",
       "      <td>7578378</td>\n",
       "      <td>7578498</td>\n",
       "      <td>ATCTGAGCAGCGCTCATGGTGGGGGCAGCGCCTCACAACCTCCGTC...</td>\n",
       "      <td>ATCTGAGCAGCGCTCATGGTGGGGGCAGCGCCTCACAACCTCCGTC...</td>\n",
       "      <td>ATCTGAGCAGCGCTCATGGTGGGGGCAGCGCCTCACAACCTCCGTC...</td>\n",
       "      <td>TTGTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGG...</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3002</th>\n",
       "      <td>3002</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577531</td>\n",
       "      <td>7577532</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GG</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.749_750delCC</td>\n",
       "      <td>p.P250Hfs*13</td>\n",
       "      <td>7577471</td>\n",
       "      <td>7577592</td>\n",
       "      <td>GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...</td>\n",
       "      <td>GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...</td>\n",
       "      <td>GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...</td>\n",
       "      <td>GCCTCCGGTTCATGCCGCCCATGCAGGAACTGTTACACATGTAGTT...</td>\n",
       "      <td>GG</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3003</th>\n",
       "      <td>3003</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578421</td>\n",
       "      <td>7578440</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GTCATGTGCTGTGACTGCTT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.490_509del</td>\n",
       "      <td>p.K164Gfs*10</td>\n",
       "      <td>7578361</td>\n",
       "      <td>7578500</td>\n",
       "      <td>AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...</td>\n",
       "      <td>AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...</td>\n",
       "      <td>AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...</td>\n",
       "      <td>GTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGGAA...</td>\n",
       "      <td>GTCATGTGCTGTGACTGCTT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3004</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.339C&gt;A</td>\n",
       "      <td>p.F113L</td>\n",
       "      <td>7579288</td>\n",
       "      <td>7579408</td>\n",
       "      <td>AGCCAGCCCCTCAGGGCAACTGACCGTGCAAGTCACAGACTTGGCT...</td>\n",
       "      <td>AGCCAGCCCCTCAGGGCAACTGACCGTGCAAGTCACAGACTTGGCT...</td>\n",
       "      <td>AGCCAGCCCCTCAGGGCAACTGACCGTGCAAGTCACAGACTTGGCT...</td>\n",
       "      <td>AAGCCCAGACGGAAACCGTAGCTGCCCTGGTAGGTTTTCTGGGAAG...</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3005 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0         0        TP53         17         7577120       7577120   \n",
       "1         1        TP53         17         7577108       7577108   \n",
       "2         2        TP53         17         7578239       7578239   \n",
       "3         3        TP53         17         7577121       7577121   \n",
       "4         4        TP53         17         7577506       7577506   \n",
       "...     ...         ...        ...             ...           ...   \n",
       "3000   3000        TP53         17         7577539       7577541   \n",
       "3001   3001        TP53         17         7578438       7578438   \n",
       "3002   3002        TP53         17         7577531       7577532   \n",
       "3003   3003        TP53         17         7578421       7578440   \n",
       "3004   3004        TP53         17         7579348       7579348   \n",
       "\n",
       "             Consequence Variant_Classification Variant_Type  \\\n",
       "0       missense_variant      Missense_Mutation          SNP   \n",
       "1       missense_variant      Missense_Mutation          SNP   \n",
       "2            stop_gained      Nonsense_Mutation          SNP   \n",
       "3       missense_variant      Missense_Mutation          SNP   \n",
       "4       missense_variant      Missense_Mutation          SNP   \n",
       "...                  ...                    ...          ...   \n",
       "3000    inframe_deletion           In_Frame_Del          DEL   \n",
       "3001    missense_variant      Missense_Mutation          SNP   \n",
       "3002  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "3003  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "3004    missense_variant      Missense_Mutation          SNP   \n",
       "\n",
       "          Reference_Allele Tumor_Seq_Allele2  \\\n",
       "0                        C                 T   \n",
       "1                        C                 A   \n",
       "2                        C                 A   \n",
       "3                        G                 A   \n",
       "4                        C                 A   \n",
       "...                    ...               ...   \n",
       "3000                   GGT                 -   \n",
       "3001                     C                 G   \n",
       "3002                    GG                 -   \n",
       "3003  GTCATGTGCTGTGACTGCTT                 -   \n",
       "3004                     G                 T   \n",
       "\n",
       "                                 HGVSc   HGVSp_Short  seq_start  seq_end  \\\n",
       "0           ENST00000269305.4:c.818G>A       p.R273H    7577060  7577180   \n",
       "1           ENST00000269305.4:c.830G>T       p.C277F    7577048  7577168   \n",
       "2           ENST00000269305.4:c.610G>T       p.E204*    7578179  7578299   \n",
       "3           ENST00000269305.4:c.817C>T       p.R273C    7577061  7577181   \n",
       "4           ENST00000269305.4:c.775G>T       p.D259Y    7577446  7577566   \n",
       "...                                ...           ...        ...      ...   \n",
       "3000    ENST00000269305.4:c.740_742del     p.N247del    7577479  7577601   \n",
       "3001        ENST00000269305.4:c.492G>C       p.K164N    7578378  7578498   \n",
       "3002  ENST00000269305.4:c.749_750delCC  p.P250Hfs*13    7577471  7577592   \n",
       "3003    ENST00000269305.4:c.490_509del  p.K164Gfs*10    7578361  7578500   \n",
       "3004        ENST00000269305.4:c.339C>A       p.F113L    7579288  7579408   \n",
       "\n",
       "                                           wt_w_context  \\\n",
       "0     CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAG...   \n",
       "1     TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCC...   \n",
       "2     CAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTTCT...   \n",
       "3     CTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGG...   \n",
       "4     GCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGCAAGTGGC...   \n",
       "...                                                 ...   \n",
       "3000  GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...   \n",
       "3001  ATCTGAGCAGCGCTCATGGTGGGGGCAGCGCCTCACAACCTCCGTC...   \n",
       "3002  GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...   \n",
       "3003  AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...   \n",
       "3004  AGCCAGCCCCTCAGGGCAACTGACCGTGCAAGTCACAGACTTGGCT...   \n",
       "\n",
       "                                          alt_w_context  \\\n",
       "0     CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAG...   \n",
       "1     TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCC...   \n",
       "2     CAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTTCT...   \n",
       "3     CTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGG...   \n",
       "4     GCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGCAAGTGGC...   \n",
       "...                                                 ...   \n",
       "3000  GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...   \n",
       "3001  ATCTGAGCAGCGCTCATGGTGGGGGCAGCGCCTCACAACCTCCGTC...   \n",
       "3002  GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...   \n",
       "3003  AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...   \n",
       "3004  AGCCAGCCCCTCAGGGCAACTGACCGTGCAAGTCACAGACTTGGCT...   \n",
       "\n",
       "                                           left_context  \\\n",
       "0     CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAG...   \n",
       "1     TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCC...   \n",
       "2     CAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTTCT...   \n",
       "3     CTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGG...   \n",
       "4     GCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGCAAGTGGC...   \n",
       "...                                                 ...   \n",
       "3000  GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...   \n",
       "3001  ATCTGAGCAGCGCTCATGGTGGGGGCAGCGCCTCACAACCTCCGTC...   \n",
       "3002  GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...   \n",
       "3003  AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...   \n",
       "3004  AGCCAGCCCCTCAGGGCAACTGACCGTGCAAGTCACAGACTTGGCT...   \n",
       "\n",
       "                                          right_context                   REF  \\\n",
       "0     GCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGAT...                     C   \n",
       "1     AGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACC...                     C   \n",
       "2     CACACGCAAATTTCCTTCCACTCGGATAAGATGCTGAGGAGGGGCC...                     C   \n",
       "3     CACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATA...                     G   \n",
       "4     TTCCAGTGTGATGATGGTGAGGATGGGCCTCCGGTTCATGCCGCCC...                     C   \n",
       "...                                                 ...                   ...   \n",
       "3000  TCATGCCGCCCATGCAGGAACTGTTACACATGTAGTTGTAGTGGAT...                   GGT   \n",
       "3001  TTGTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGG...                     C   \n",
       "3002  GCCTCCGGTTCATGCCGCCCATGCAGGAACTGTTACACATGTAGTT...                    GG   \n",
       "3003  GTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGGAA...  GTCATGTGCTGTGACTGCTT   \n",
       "3004  AAGCCCAGACGGAAACCGTAGCTGCCCTGGTAGGTTTTCTGGGAAG...                     G   \n",
       "\n",
       "     ALT  \n",
       "0      T  \n",
       "1      A  \n",
       "2      A  \n",
       "3      A  \n",
       "4      A  \n",
       "...   ..  \n",
       "3000      \n",
       "3001   G  \n",
       "3002      \n",
       "3003      \n",
       "3004   T  \n",
       "\n",
       "[3005 rows x 20 columns]"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p53_formatted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Flexible input format\n",
    "\n",
    "- Get this working so that you can also put in sequences in PrimeDesign format AND in for just inputting the ref and mutant seq\n",
    "\n",
    "- Need to get out the indeces as well? or can just count the size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mut_formatter(wt, alt):\n",
    "\n",
    "    alignments = aligner.align(wt, alt)\n",
    "\n",
    "    WT_align = alignments[0].aligned[0]\n",
    "    alt_align = alignments[0].aligned[1]\n",
    "\n",
    "    if len(WT_align) == 1:\n",
    "        #Substitution (SNP or ONP)\n",
    "\n",
    "        #find indexes of non-matching values to get right/left context sequences and mutant sequence\n",
    "        non_matching_idx = []\n",
    "\n",
    "        for i, val in enumerate(list(zip(wt, alt))):\n",
    "            if val[0] != val[1]:\n",
    "                non_matching_idx.append(i)\n",
    "        \n",
    "        assert len(non_matching_idx) !=0, f'No mutation detected in alt sequence provided | row index = {i}'\n",
    "\n",
    "        if len(non_matching_idx)==1:\n",
    "            var_type = 'SNP'\n",
    "        else:\n",
    "            var_type = 'ONP'\n",
    "\n",
    "        left_seq = wt[:non_matching_idx[0]]\n",
    "        ref_seq = wt[non_matching_idx[0]:non_matching_idx[-1]+1]\n",
    "        alt_seq = alt[non_matching_idx[0]:non_matching_idx[-1]+1]\n",
    "        right_seq = wt[non_matching_idx[-1]+1:]\n",
    "\n",
    "\n",
    "    elif len(WT_align)>1:\n",
    "        if len(alt)>len(wt):\n",
    "            var_type = 'INS'\n",
    "        elif len(alt)<len(wt):\n",
    "            var_type='DEL'\n",
    "\n",
    "        left_seq = wt[WT_align[0][0]:WT_align[0][1]]\n",
    "        right_seq = wt[WT_align[1][0]:WT_align[1][1]]\n",
    "\n",
    "        if var_type=='INS':\n",
    "        #INS or DEL\n",
    "            ref_seq = ''\n",
    "            alt_seq = alt[alt_align[0][1]:alt_align[1][0]]\n",
    "            \n",
    "        elif var_type =='DEL':\n",
    "            ref_seq = wt[WT_align[0][1]:WT_align[1][0]]\n",
    "            alt_seq = ''\n",
    "            \n",
    "\n",
    "    return var_type, left_seq, ref_seq, alt_seq, right_seq\n",
    "    print(var_type)\n",
    "    print(left_seq)\n",
    "    print(f'REF: {ref_seq}')\n",
    "    print(f'ALT: {alt_seq}')\n",
    "    print(right_seq)\n",
    "\n",
    "    #print(alignments[0])\n",
    "    #this program will break for complex substitutions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def primedesign_formatter(seq):\n",
    "    \"\"\" \n",
    "    takes as input sequence in prime design format \n",
    "    e.g. AATTCCG(G/C)AATTCGCT\n",
    "    \"\"\"\n",
    "\n",
    "    start = seq.find(\"(\")\n",
    "\n",
    "    end = seq.find(\")\")\n",
    "\n",
    "    replace_seq = seq[start:end+1]\n",
    "    #if '/' in replace_seq:\n",
    "        #if '+' in replace_seq:\n",
    "            #throw an error\n",
    "\n",
    "    loc_replace = replace_seq.find('/')\n",
    "    ref_seq = replace_seq[1:loc_replace]\n",
    "    alt_seq = replace_seq[loc_replace+1:-1]\n",
    "\n",
    "    left_seq = seq[:start]\n",
    "    right_seq = seq[end+1:]\n",
    "\n",
    "    wt_w_context = left_seq + ref_seq + right_seq\n",
    "    alt_w_context = left_seq + alt_seq + right_seq\n",
    "\n",
    "    if len(ref_seq)==0:\n",
    "        var_type='INS'\n",
    "\n",
    "    else:\n",
    "        if len(ref_seq)==len(alt_seq):\n",
    "            if len(ref_seq)==1:\n",
    "                var_type='SNP'\n",
    "            else:\n",
    "                var_type='ONP'\n",
    "\n",
    "        else:\n",
    "            if len(alt_seq) == 0:\n",
    "                var_type = 'DEL'\n",
    "            else:\n",
    "                var_type = 'INDEL'\n",
    "\n",
    "\n",
    "    \n",
    "    #print(var_type)\n",
    "\n",
    "    #print(wt_w_context)\n",
    "    #print(alt_w_context)\n",
    "    #print(f'LEFT: {left_seq}')\n",
    "    #print(f'REF: {ref_seq}')\n",
    "    #print(f'ALT: {alt_seq}')\n",
    "    #print(f'RIGHT: {right_seq}')\n",
    "\n",
    "    return var_type, wt_w_context, alt_w_context, left_seq, right_seq, ref_seq, alt_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = 'AATTCGTCGTAGC(G/)TCGGCTAGCTAGCTGGCA'\n",
    "var_type, wt_w_context, alt_w_context, left_seq, right_seq, ref_seq, alt_seq = primedesign_formatter(seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PAM searching\n",
    "\n",
    "- Idea is to only do everything with respect to the forward strand\n",
    "- Not code a bespoke solution for the reverse complement strand\n",
    "- And just fix the coordinates afterward (translate them to the forward strand coordinates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PAM_finder(seq, PAM):\n",
    "    \"\"\" \n",
    "    Finds indeces of PAM sequences\n",
    "    Formatting: e.g. \"NGG\"\n",
    "    N = [A|T|C|G]\n",
    "    R = [A|G]\n",
    "    Y = [C|T]\n",
    "    S = [G|C]\n",
    "    W = [A|T]\n",
    "    K = [G|T]\n",
    "    M = [A|C]\n",
    "    B = [C|G|T]\n",
    "    D = [A|G|T]\n",
    "    H = [A|C|T]\n",
    "    V = [A|C|G]\n",
    "    \"\"\"\n",
    "    PAM_dict = {\"A\":\"A\",\n",
    "                \"G\":\"G\",\n",
    "                \"C\":\"C\",\n",
    "                \"T\":\"T\",\n",
    "                \"N\":\"[A|T|C|G]\",\n",
    "                \"R\":\"[A|G]\",\n",
    "                \"Y\":\"[C|T]\",\n",
    "                \"S\":\"[G|C]\",\n",
    "                \"W\":\"[A|T]\",\n",
    "                \"K\":\"[G|T]\",\n",
    "                \"M\":\"[A|C]\",\n",
    "                \"B\":\"[C|G|T]\",\n",
    "                \"D\":\"[A|G|T]\",\n",
    "                \"H\":\"[A|C|T]\",\n",
    "                \"V\":\"[A|C|G]\"}\n",
    "\n",
    "    PAM_new = \"\"\n",
    "    for i in PAM:\n",
    "        PAM_new += PAM_dict[i]\n",
    "\n",
    "    p = re.compile(PAM_new)\n",
    "\n",
    "    iterator = p.finditer(seq, overlapped=True)\n",
    "    PAM_holder = []\n",
    "    for match in iterator:\n",
    "        PAM_holder.append(match.span())\n",
    "\n",
    "    return PAM_holder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 3 GGG\n",
      "1 4 GGG\n",
      "4 7 TGG\n",
      "5 8 GGG\n"
     ]
    }
   ],
   "source": [
    "PAM = 'NGG'\n",
    "seq = 'GGGGTGGG'\n",
    "PAM_holder = PAM_finder(seq, PAM)\n",
    "\n",
    "for i in PAM_holder:\n",
    "    print(i[0], i[1], seq[i[0]:i[1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(18, 21), (30, 33)]"
      ]
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PAM_holder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ONP\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTACGTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTGCATGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "LEFT: GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTT\n",
      "REF: ACG\n",
      "ALT: GCA\n",
      "RIGHT: TGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "TGG\n",
      "TGG\n",
      "TGG\n",
      "TGG\n",
      "TGG\n",
      "TGG\n",
      "TGG\n"
     ]
    }
   ],
   "source": [
    "PAM = 'NGG'\n",
    "\n",
    "#seq = 'ATGATGCGCGGCTAGAATTCGTTGGCGTAGCGTAGCGT(G/)TCTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA'\n",
    "seq = 'GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTT(ACG/GCA)TGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA'\n",
    "\n",
    "var_type, wt_w_context, alt_w_context, left_seq, right_seq, ref_seq, alt_seq = primedesign_formatter(seq)\n",
    "\n",
    "\n",
    "PAM_holder = PAM_finder(wt_w_context, PAM)\n",
    "\n",
    "for i in PAM_holder:\n",
    "    print(wt_w_context[i[0]:i[1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 502,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 502,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('CGTAAGCGT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 505,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(26, 29), (43, 46)]"
      ]
     },
     "execution_count": 505,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#searching for PAM sequences on the FORWARD STRAND\n",
    "\n",
    "#search on max_RTT_length - \n",
    "\n",
    "left_len = len(left_seq)\n",
    "right_len = len(right_seq)\n",
    "max_RTT_length = 20\n",
    "\n",
    "if var_type =='INS':\n",
    "    alt_size = len(alt_seq)\n",
    "    m_start = left_len - (max_RTT_length - alt_size - 3)\n",
    "    m_end = left_len + 4\n",
    "\n",
    "else:\n",
    "    ref_size = len(ref_seq)\n",
    "    alt_size = len(alt_seq)\n",
    "    m_start = left_len - (max_RTT_length - alt_size - 3)\n",
    "    m_end = left_len + 3\n",
    "\n",
    "proto_size = 19\n",
    "\n",
    "eligible_PAM_start = max(m_start, proto_size + 3) #\n",
    "eligible_PAM_end = m_end\n",
    "\n",
    "eligible_PAMS = []\n",
    "for i in PAM_holder:\n",
    "    #check start of PAM site\n",
    "    s = i[0]\n",
    "    if (s>=eligible_PAM_start) and (s<=eligible_PAM_end):\n",
    "        eligible_PAMS.append(i)\n",
    "\n",
    "eligible_PAMS\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 506,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------GGGCGTAAGCGTTTACGTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "--------------------------|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTACGTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTACGTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "||||||||||||||||||||||||||||||||||||||||.|.||||||||||||||||||||||||||||||||||||||||||\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTGCATGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "\n",
      "-------------------------------------------TGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "-------------------------------------------||||||||||||||||||||||||||||||||||||||||||\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTACGTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTACGTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "||||||||||||||||||||||||||||||||||||||||.|.||||||||||||||||||||||||||||||||||||||||||\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTGCATGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in eligible_PAMS:\n",
    "    s = i[0]\n",
    "    alignments = aligner.align(wt_w_context[s:], wt_w_context)\n",
    "\n",
    "    print(alignments[0])\n",
    "\n",
    "    alignments2 = aligner.align(wt_w_context, alt_w_context)\n",
    "    print(alignments2[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 492,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 492,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('GGGCGTAAGCGTTTACG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGGACAGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATAGGAAAAGAGAAGC\n",
      "CGCC\n",
      "GACA\n",
      "GGCA\n",
      "AACA\n",
      "CGCA\n",
      "AGCT\n",
      "TACC\n",
      "TACT\n"
     ]
    }
   ],
   "source": [
    "seq = p53_formatted.iloc[0]['wt_w_context']\n",
    "print(seq)\n",
    "PAM = \"NRCH\"\n",
    "PAM_holder = PAM_finder(seq, PAM)\n",
    "\n",
    "for i in PAM_holder:\n",
    "    print(seq[i[0]:i[1]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Putting some things together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEL\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTATGGTAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "LEFT: GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTT\n",
      "REF: ATGG\n",
      "ALT: \n",
      "RIGHT: TAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n"
     ]
    }
   ],
   "source": [
    "seq = 'GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTT(ATGG/)TAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA'\n",
    "\n",
    "var_type, wt_w_context, alt_w_context, left_seq, right_seq, ref_seq, alt_seq = primedesign_formatter(seq)\n",
    "mut = mutation(wt_w_context, alt_w_context, left_seq, right_seq, var_type, ref_seq, alt_seq, chrom=None, genome=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def eligible_PAM_finder(mut, PAM, max_RTT_length, proto_size=19):\n",
    "    \"\"\" \n",
    "    Determines eligible PAM sequences for creating pegRNAs...\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    forward = PAM_finder(mut.wt_forward, PAM)\n",
    "    reverse = PAM_finder(mut.wt_rc, PAM)\n",
    "\n",
    "\n",
    "    #first do the forward sequence\n",
    "    left_len_F = len(mut.left_seq)\n",
    "    left_len_R = len(mut.left_seq_rc)\n",
    "\n",
    "    if var_type =='INS':\n",
    "        alt_size = len(mut.alt_seq)\n",
    "        m_start_F = left_len_F - (max_RTT_length - alt_size - 3)\n",
    "        m_end_F = left_len_F + 4\n",
    "        \n",
    "        m_start_R = left_len_R - (max_RTT_length - alt_size - 3)\n",
    "        m_end_R = left_len_R + 4\n",
    "\n",
    "    else:\n",
    "        alt_size = len(mut.alt_seq)\n",
    "        m_start_F = left_len_F - (max_RTT_length - alt_size - 3)\n",
    "        m_end_F = left_len_F + 3\n",
    "\n",
    "        m_start_R = left_len_R - (max_RTT_length - alt_size - 3)\n",
    "        m_end_R = left_len_R + 3\n",
    "\n",
    "\n",
    "    eligible_PAM_start_F = max(m_start_F, proto_size + 3) #\n",
    "    eligible_PAM_end_F = m_end_F\n",
    "    \n",
    "    eligible_PAM_start_R = max(m_start_R, proto_size + 3) #\n",
    "    eligible_PAM_end_R = m_end_R\n",
    "\n",
    "    eligible_PAMS_F = []\n",
    "    eligible_PAMS_R = []\n",
    "\n",
    "    for i in forward:\n",
    "        #check start of PAM site\n",
    "        s = i[0]\n",
    "        if (s>=eligible_PAM_start_F) and (s<=eligible_PAM_end_F):\n",
    "            eligible_PAMS_F.append(i)\n",
    "\n",
    "    for i in reverse:\n",
    "        #check start of PAM site\n",
    "        s = i[0]\n",
    "        if (s>=eligible_PAM_start_R) and (s<=eligible_PAM_end_R):\n",
    "            eligible_PAMS_R.append(i)\n",
    "\n",
    "    return eligible_PAMS_F, eligible_PAMS_R\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#eligible PAM finding\n",
    "PAM = \"NGG\"\n",
    "max_RTT_length = 20\n",
    "proto_size = 10\n",
    "\n",
    "mut.PAM_idx_forward = PAM_finder(mut.wt_forward, PAM)\n",
    "mut.PAM_idx_rc = PAM_finder(mut.wt_rc, PAM)\n",
    "\n",
    "mut.PAM_idx_forward, mut.PAM_idx_rc = eligible_PAM_finder(mut, PAM, max_RTT_length, proto_size)\n",
    "#mut.PAM_idx_forward\n",
    "#mut.PAM_idx_rc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(23, 26), (24, 27), (25, 28), (26, 29), (41, 44)]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut.PAM_idx_forward"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(33, 36)]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut.PAM_idx_rc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# pegRNA design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pegRNA_generator(mut, PAM, orientation, proto_size, RTT_lengths, PBS_lengths):\n",
    "    #------function----\n",
    "    protospacer_seqs = []\n",
    "    PAM_start_list = []\n",
    "    PAM_end_list = []\n",
    "    PAM_list = []\n",
    "\n",
    "    RTT_seqs = []\n",
    "    RTT_size = []\n",
    "    PBS_seqs = []\n",
    "    PBS_size = []\n",
    "    RTT_PBS_seqs = []\n",
    "\n",
    "    #descriptors of pegRNAs...\n",
    "    distance_to_nick = [] #distance from the start of the mutation to the nick introduced by nCas9 (-3 from PAM)\n",
    "    RHA = [] #right homology arm (distance from end of mutation to end of RTT)\n",
    "    PAM_disrupted_list = []\n",
    "    proto_disrupted_list = []\n",
    "\n",
    "    #Get mutation info, depending on whether we're looking on the + or the - strand\n",
    "    if orientation == '+':\n",
    "        left_len = len(mut.left_seq)\n",
    "        ref_len = len(mut.ref_seq)\n",
    "        alt_len = len(mut.alt_seq)\n",
    "        alt_seq = mut.alt_seq\n",
    "        seq_F = mut.wt_forward\n",
    "\n",
    "        PAM_seqs = mut.PAM_idx_forward\n",
    "\n",
    "    elif orientation == '-':\n",
    "        left_len = len(mut.left_seq_rc)\n",
    "        ref_len = len(mut.ref_seq_rc)\n",
    "        alt_len = len(mut.alt_seq_rc)\n",
    "        alt_seq = mut.alt_seq_rc\n",
    "        seq_F = mut.wt_rc\n",
    "\n",
    "        PAM_seqs = mut.PAM_idx_rc\n",
    "\n",
    "    #iterate over the PAM sequences:\n",
    "    for i in PAM_seqs:\n",
    "        PAM_start = i[0]\n",
    "        PAM_end = i[1]\n",
    "\n",
    "        RTT_start = PAM_start - 3\n",
    "\n",
    "        left_RTT = seq_F[RTT_start:left_len]\n",
    "\n",
    "        for RTT_length in RTT_lengths:\n",
    "\n",
    "            remaining_length = RTT_length - (len(left_RTT) + alt_len)\n",
    "\n",
    "            #make sure RTT is long enough\n",
    "            #and make sure that there's sufficient sequence to actually pull from...\n",
    "            if (remaining_length <0) or ((left_len+ref_len + remaining_length) > len(seq_F)):\n",
    "                RTT_seqs.append(None)\n",
    "                distance_to_nick.append(None)\n",
    "                RHA.append(None)\n",
    "                PAM_disrupted_list.append(None)\n",
    "                proto_disrupted_list.append(None)\n",
    "                PAM_start_list.append(None)\n",
    "                PAM_end_list.append(None)\n",
    "                PAM_list.append(None)\n",
    "                protospacer_seqs.append(None)\n",
    "                PBS_seqs.append(None)\n",
    "                RTT_PBS_seqs.append(None)\n",
    "                RTT_size.append(None)\n",
    "                PBS_size.append(None)\n",
    "\n",
    "            else: #able to design the pegRNA\n",
    "\n",
    "                right_RTT = seq_F[left_len+ref_len:left_len+ref_len + remaining_length]\n",
    "                RTT = left_RTT + alt_seq + right_RTT\n",
    "\n",
    "                #and then determine if the PAM sequence is disrupted\n",
    "                #protospacer disrupted would just be if the distance to nick is < 3\n",
    "                #fish out the PAM sequence from RTT\n",
    "                PAM_new = RTT[3:3+len(PAM)]\n",
    "                pam_list = PAM_finder(PAM_new, PAM)\n",
    "\n",
    "                PAM_disrupted = False\n",
    "                if len(pam_list)==0:\n",
    "                    PAM_disrupted = True\n",
    "\n",
    "                proto_disrupted = False\n",
    "                if len(left_RTT) < 3:\n",
    "                    proto_disrupted = True\n",
    "\n",
    "                #and finally reverse complement the RTT\n",
    "                RTT = str(Bio.Seq.Seq(RTT).reverse_complement())\n",
    "\n",
    "                #pull out the PAM sequence and protospacer as well\n",
    "                PAM_sequence = seq_F[PAM_start:PAM_end]\n",
    "                protospacer =  'G' + seq_F[PAM_start - proto_size:PAM_start]\n",
    "\n",
    "\n",
    "                #iterate over PBS lengths for adding things to list\n",
    "                for PBS_length in PBS_lengths:\n",
    "                    RTT_seqs.append(RTT)\n",
    "                    RTT_size.append(RTT_length)\n",
    "                    PBS_size.append(PBS_length)\n",
    "                    distance_to_nick.append(len(left_RTT))\n",
    "                    RHA.append(len(right_RTT))\n",
    "                    PAM_disrupted_list.append(PAM_disrupted)\n",
    "                    proto_disrupted_list.append(proto_disrupted)\n",
    "\n",
    "                    #------add in PAM info, protospacer, + PBS/RTT_PBS\n",
    "                    #record info about PAM sequences\n",
    "                    if orientation == '+':\n",
    "                        PAM_start_list.append(PAM_start)\n",
    "                        PAM_end_list.append(PAM_end)\n",
    "                    elif orientation == '-':\n",
    "                        PAM_start_list.append(len(seq_F)-PAM_start)\n",
    "                        PAM_end_list.append(len(seq_F)-PAM_end)\n",
    "\n",
    "                    PAM_list.append(PAM_sequence)\n",
    "                    protospacer_seqs.append(protospacer)\n",
    "\n",
    "                    #----and finally design the PBS sequences----\n",
    "                    #NEED TO HAVE AN ERROR MESSAGE FOR PBS SEQUENCES LARGER THAN 17\n",
    "                    PBS = seq_F[RTT_start-PBS_length:RTT_start]\n",
    "                    PBS = str(Bio.Seq.Seq(PBS).reverse_complement())\n",
    "                    PBS_seqs.append(PBS)\n",
    "\n",
    "                    #---and then add it to the RTT to make the full 3' extension\n",
    "                    RTT_PBS_seqs.append(RTT+PBS)\n",
    "\n",
    "    cols = [PAM_start_list, PAM_end_list, PAM_list, orientation, protospacer_seqs, RTT_seqs, RTT_size, PBS_seqs, PBS_size, RTT_PBS_seqs, distance_to_nick, RHA, PAM_disrupted_list, proto_disrupted_list]\n",
    "    col_labels = [\"PAM_start\", \"PAM_end\", \"PAM\", \"PAM_strand\",\"Protospacer\", \"RTT\", \"RTT_length\", \"PBS\", \"PBS_length\", \"RTT_PBS\", \"Distance_to_nick\", \"RHA_size\", \"PAM_disrupted\", \"Proto_disrupted\"]\n",
    "    dtypes = ['int', 'int', 'str', 'str', 'str', 'str', 'int', 'str', 'int', 'str', 'int', 'int', 'bool', 'bool']\n",
    "\n",
    "    dtype_dict = dict(zip(col_labels, dtypes))\n",
    "\n",
    "    peg_df = pd.DataFrame(dict(zip(col_labels, cols))).dropna().reset_index().drop(columns='index').astype(dtype_dict)\n",
    "\n",
    "    return peg_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEL\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTATGGTAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "LEFT: GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTT\n",
      "REF: ATGG\n",
      "ALT: \n",
      "RIGHT: TAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PAM_start</th>\n",
       "      <th>PAM_end</th>\n",
       "      <th>PAM</th>\n",
       "      <th>PAM_strand</th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>RTT</th>\n",
       "      <th>RTT_length</th>\n",
       "      <th>PBS</th>\n",
       "      <th>PBS_length</th>\n",
       "      <th>RTT_PBS</th>\n",
       "      <th>Distance_to_nick</th>\n",
       "      <th>RHA_size</th>\n",
       "      <th>PAM_disrupted</th>\n",
       "      <th>Proto_disrupted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>25</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>25</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>25</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>25</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>AGCTACTAAA</td>\n",
       "      <td>10</td>\n",
       "      <td>ACGCTTACGC</td>\n",
       "      <td>10</td>\n",
       "      <td>AGCTACTAAAACGCTTACGC</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>AGCTACTAAA</td>\n",
       "      <td>10</td>\n",
       "      <td>ACGCTTACGCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>AGCTACTAAAACGCTTACGCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>AGCTACTAAA</td>\n",
       "      <td>10</td>\n",
       "      <td>ACGCTTACGCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>AGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>AGCTACTAAA</td>\n",
       "      <td>10</td>\n",
       "      <td>ACGCTTACGCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>AGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>20</td>\n",
       "      <td>ACGCTTACGC</td>\n",
       "      <td>10</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAAACGCTTACGC</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>20</td>\n",
       "      <td>ACGCTTACGCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAAACGCTTACGCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>20</td>\n",
       "      <td>ACGCTTACGCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>20</td>\n",
       "      <td>ACGCTTACGCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>30</td>\n",
       "      <td>ACGCTTACGC</td>\n",
       "      <td>10</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGC</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>30</td>\n",
       "      <td>ACGCTTACGCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>30</td>\n",
       "      <td>ACGCTTACGCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>30</td>\n",
       "      <td>ACGCTTACGCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>20</td>\n",
       "      <td>TGGCATGGCA</td>\n",
       "      <td>10</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGCTGGCATGGCA</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>20</td>\n",
       "      <td>TGGCATGGCATGG</td>\n",
       "      <td>13</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGCTGGCATGGCATGG</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>20</td>\n",
       "      <td>TGGCATGGCATGGCA</td>\n",
       "      <td>15</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCA</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>20</td>\n",
       "      <td>TGGCATGGCATGGCATG</td>\n",
       "      <td>17</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCATG</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>30</td>\n",
       "      <td>TGGCATGGCA</td>\n",
       "      <td>10</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCA</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>30</td>\n",
       "      <td>TGGCATGGCATGG</td>\n",
       "      <td>13</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGG</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>30</td>\n",
       "      <td>TGGCATGGCATGGCA</td>\n",
       "      <td>15</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCA</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>30</td>\n",
       "      <td>TGGCATGGCATGGCATG</td>\n",
       "      <td>17</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCATG</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    PAM_start  PAM_end  PAM PAM_strand           Protospacer  \\\n",
       "0          22       25  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "1          22       25  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "2          22       25  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "3          22       25  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "4          23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "5          23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "6          23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "7          23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "8          23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "9          23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "10         23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "11         23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "12         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "13         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "14         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "15         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "16         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "17         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "18         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "19         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "20         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "21         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "22         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "23         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "24         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "25         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "26         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "27         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "28         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "29         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "30         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "31         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "32         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "33         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "34         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "35         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "36         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "37         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "38         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "39         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "40         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "41         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "42         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "43         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "44         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "45         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "46         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "47         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "48         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "49         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "50         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "51         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "52         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "53         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "54         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "55         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "\n",
       "                               RTT  RTT_length                PBS  PBS_length  \\\n",
       "0   TAGCTACTAAAACGCTTACGCCCCCCCCCC          30         CCCCCCCCCC          10   \n",
       "1   TAGCTACTAAAACGCTTACGCCCCCCCCCC          30      CCCCCCCCCCCCC          13   \n",
       "2   TAGCTACTAAAACGCTTACGCCCCCCCCCC          30    CCCCCCCCCCCCCCC          15   \n",
       "3   TAGCTACTAAAACGCTTACGCCCCCCCCCC          30  CCCCCCCCCCCCCCCCC          17   \n",
       "4             AAACGCTTACGCCCCCCCCC          20         CCCCCCCCCC          10   \n",
       "5             AAACGCTTACGCCCCCCCCC          20      CCCCCCCCCCCCC          13   \n",
       "6             AAACGCTTACGCCCCCCCCC          20    CCCCCCCCCCCCCCC          15   \n",
       "7             AAACGCTTACGCCCCCCCCC          20  CCCCCCCCCCCCCCCCC          17   \n",
       "8   CTAGCTACTAAAACGCTTACGCCCCCCCCC          30         CCCCCCCCCC          10   \n",
       "9   CTAGCTACTAAAACGCTTACGCCCCCCCCC          30      CCCCCCCCCCCCC          13   \n",
       "10  CTAGCTACTAAAACGCTTACGCCCCCCCCC          30    CCCCCCCCCCCCCCC          15   \n",
       "11  CTAGCTACTAAAACGCTTACGCCCCCCCCC          30  CCCCCCCCCCCCCCCCC          17   \n",
       "12            AAAACGCTTACGCCCCCCCC          20         CCCCCCCCCC          10   \n",
       "13            AAAACGCTTACGCCCCCCCC          20      CCCCCCCCCCCCC          13   \n",
       "14            AAAACGCTTACGCCCCCCCC          20    CCCCCCCCCCCCCCC          15   \n",
       "15            AAAACGCTTACGCCCCCCCC          20  CCCCCCCCCCCCCCCCC          17   \n",
       "16  GCTAGCTACTAAAACGCTTACGCCCCCCCC          30         CCCCCCCCCC          10   \n",
       "17  GCTAGCTACTAAAACGCTTACGCCCCCCCC          30      CCCCCCCCCCCCC          13   \n",
       "18  GCTAGCTACTAAAACGCTTACGCCCCCCCC          30    CCCCCCCCCCCCCCC          15   \n",
       "19  GCTAGCTACTAAAACGCTTACGCCCCCCCC          30  CCCCCCCCCCCCCCCCC          17   \n",
       "20            TAAAACGCTTACGCCCCCCC          20         CCCCCCCCCC          10   \n",
       "21            TAAAACGCTTACGCCCCCCC          20      CCCCCCCCCCCCC          13   \n",
       "22            TAAAACGCTTACGCCCCCCC          20    CCCCCCCCCCCCCCC          15   \n",
       "23            TAAAACGCTTACGCCCCCCC          20  CCCCCCCCCCCCCCCCC          17   \n",
       "24  GGCTAGCTACTAAAACGCTTACGCCCCCCC          30         CCCCCCCCCC          10   \n",
       "25  GGCTAGCTACTAAAACGCTTACGCCCCCCC          30      CCCCCCCCCCCCC          13   \n",
       "26  GGCTAGCTACTAAAACGCTTACGCCCCCCC          30    CCCCCCCCCCCCCCC          15   \n",
       "27  GGCTAGCTACTAAAACGCTTACGCCCCCCC          30  CCCCCCCCCCCCCCCCC          17   \n",
       "28            CTAAAACGCTTACGCCCCCC          20         CCCCCCCCCC          10   \n",
       "29            CTAAAACGCTTACGCCCCCC          20      CCCCCCCCCCCCC          13   \n",
       "30            CTAAAACGCTTACGCCCCCC          20    CCCCCCCCCCCCCCC          15   \n",
       "31            CTAAAACGCTTACGCCCCCC          20  CCCCCCCCCCCCCCCCC          17   \n",
       "32  AGGCTAGCTACTAAAACGCTTACGCCCCCC          30         CCCCCCCCCC          10   \n",
       "33  AGGCTAGCTACTAAAACGCTTACGCCCCCC          30      CCCCCCCCCCCCC          13   \n",
       "34  AGGCTAGCTACTAAAACGCTTACGCCCCCC          30    CCCCCCCCCCCCCCC          15   \n",
       "35  AGGCTAGCTACTAAAACGCTTACGCCCCCC          30  CCCCCCCCCCCCCCCCC          17   \n",
       "36                      AGCTACTAAA          10         ACGCTTACGC          10   \n",
       "37                      AGCTACTAAA          10      ACGCTTACGCCCC          13   \n",
       "38                      AGCTACTAAA          10    ACGCTTACGCCCCCC          15   \n",
       "39                      AGCTACTAAA          10  ACGCTTACGCCCCCCCC          17   \n",
       "40            CAGCTAGGCTAGCTACTAAA          20         ACGCTTACGC          10   \n",
       "41            CAGCTAGGCTAGCTACTAAA          20      ACGCTTACGCCCC          13   \n",
       "42            CAGCTAGGCTAGCTACTAAA          20    ACGCTTACGCCCCCC          15   \n",
       "43            CAGCTAGGCTAGCTACTAAA          20  ACGCTTACGCCCCCCCC          17   \n",
       "44  CATGCCATGCCAGCTAGGCTAGCTACTAAA          30         ACGCTTACGC          10   \n",
       "45  CATGCCATGCCAGCTAGGCTAGCTACTAAA          30      ACGCTTACGCCCC          13   \n",
       "46  CATGCCATGCCAGCTAGGCTAGCTACTAAA          30    ACGCTTACGCCCCCC          15   \n",
       "47  CATGCCATGCCAGCTAGGCTAGCTACTAAA          30  ACGCTTACGCCCCCCCC          17   \n",
       "48            GTTTTAGTAGCTAGCCTAGC          20         TGGCATGGCA          10   \n",
       "49            GTTTTAGTAGCTAGCCTAGC          20      TGGCATGGCATGG          13   \n",
       "50            GTTTTAGTAGCTAGCCTAGC          20    TGGCATGGCATGGCA          15   \n",
       "51            GTTTTAGTAGCTAGCCTAGC          20  TGGCATGGCATGGCATG          17   \n",
       "52  GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC          30         TGGCATGGCA          10   \n",
       "53  GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC          30      TGGCATGGCATGG          13   \n",
       "54  GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC          30    TGGCATGGCATGGCA          15   \n",
       "55  GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC          30  TGGCATGGCATGGCATG          17   \n",
       "\n",
       "                                            RTT_PBS  Distance_to_nick  \\\n",
       "0          TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCC                21   \n",
       "1       TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC                21   \n",
       "2     TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCC                21   \n",
       "3   TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCCC                21   \n",
       "4                    AAACGCTTACGCCCCCCCCCCCCCCCCCCC                20   \n",
       "5                 AAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC                20   \n",
       "6               AAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC                20   \n",
       "7             AAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCC                20   \n",
       "8          CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCC                20   \n",
       "9       CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC                20   \n",
       "10    CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC                20   \n",
       "11  CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCC                20   \n",
       "12                   AAAACGCTTACGCCCCCCCCCCCCCCCCCC                19   \n",
       "13                AAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC                19   \n",
       "14              AAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC                19   \n",
       "15            AAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCC                19   \n",
       "16         GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCC                19   \n",
       "17      GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC                19   \n",
       "18    GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC                19   \n",
       "19  GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCC                19   \n",
       "20                   TAAAACGCTTACGCCCCCCCCCCCCCCCCC                18   \n",
       "21                TAAAACGCTTACGCCCCCCCCCCCCCCCCCCCC                18   \n",
       "22              TAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC                18   \n",
       "23            TAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC                18   \n",
       "24         GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCC                18   \n",
       "25      GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCC                18   \n",
       "26    GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC                18   \n",
       "27  GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC                18   \n",
       "28                   CTAAAACGCTTACGCCCCCCCCCCCCCCCC                17   \n",
       "29                CTAAAACGCTTACGCCCCCCCCCCCCCCCCCCC                17   \n",
       "30              CTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC                17   \n",
       "31            CTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC                17   \n",
       "32         AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCC                17   \n",
       "33      AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCC                17   \n",
       "34    AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC                17   \n",
       "35  AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC                17   \n",
       "36                             AGCTACTAAAACGCTTACGC                 2   \n",
       "37                          AGCTACTAAAACGCTTACGCCCC                 2   \n",
       "38                        AGCTACTAAAACGCTTACGCCCCCC                 2   \n",
       "39                      AGCTACTAAAACGCTTACGCCCCCCCC                 2   \n",
       "40                   CAGCTAGGCTAGCTACTAAAACGCTTACGC                 2   \n",
       "41                CAGCTAGGCTAGCTACTAAAACGCTTACGCCCC                 2   \n",
       "42              CAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCC                 2   \n",
       "43            CAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCCCC                 2   \n",
       "44         CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGC                 2   \n",
       "45      CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGCCCC                 2   \n",
       "46    CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCC                 2   \n",
       "47  CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCCCC                 2   \n",
       "48                   GTTTTAGTAGCTAGCCTAGCTGGCATGGCA                16   \n",
       "49                GTTTTAGTAGCTAGCCTAGCTGGCATGGCATGG                16   \n",
       "50              GTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCA                16   \n",
       "51            GTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCATG                16   \n",
       "52         GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCA                16   \n",
       "53      GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGG                16   \n",
       "54    GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCA                16   \n",
       "55  GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCATG                16   \n",
       "\n",
       "    RHA_size  PAM_disrupted  Proto_disrupted  \n",
       "0          9          False            False  \n",
       "1          9          False            False  \n",
       "2          9          False            False  \n",
       "3          9          False            False  \n",
       "4          0          False            False  \n",
       "5          0          False            False  \n",
       "6          0          False            False  \n",
       "7          0          False            False  \n",
       "8         10          False            False  \n",
       "9         10          False            False  \n",
       "10        10          False            False  \n",
       "11        10          False            False  \n",
       "12         1          False            False  \n",
       "13         1          False            False  \n",
       "14         1          False            False  \n",
       "15         1          False            False  \n",
       "16        11          False            False  \n",
       "17        11          False            False  \n",
       "18        11          False            False  \n",
       "19        11          False            False  \n",
       "20         2          False            False  \n",
       "21         2          False            False  \n",
       "22         2          False            False  \n",
       "23         2          False            False  \n",
       "24        12          False            False  \n",
       "25        12          False            False  \n",
       "26        12          False            False  \n",
       "27        12          False            False  \n",
       "28         3          False            False  \n",
       "29         3          False            False  \n",
       "30         3          False            False  \n",
       "31         3          False            False  \n",
       "32        13          False            False  \n",
       "33        13          False            False  \n",
       "34        13          False            False  \n",
       "35        13          False            False  \n",
       "36         8           True             True  \n",
       "37         8           True             True  \n",
       "38         8           True             True  \n",
       "39         8           True             True  \n",
       "40        18           True             True  \n",
       "41        18           True             True  \n",
       "42        18           True             True  \n",
       "43        18           True             True  \n",
       "44        28           True             True  \n",
       "45        28           True             True  \n",
       "46        28           True             True  \n",
       "47        28           True             True  \n",
       "48         4          False            False  \n",
       "49         4          False            False  \n",
       "50         4          False            False  \n",
       "51         4          False            False  \n",
       "52        14          False            False  \n",
       "53        14          False            False  \n",
       "54        14          False            False  \n",
       "55        14          False            False  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#mutation info\n",
    "seq = 'GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTT(ATGG/)TAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA'\n",
    "PAM = \"NGG\"\n",
    "#--params\n",
    "RTT_lengths = [10,20,30]\n",
    "PBS_lengths = [10,13, 15,17]\n",
    "orientation = '-'\n",
    "proto_size = 19\n",
    "\n",
    "max_RTT_length = max(RTT_lengths)\n",
    "\n",
    "var_type, wt_w_context, alt_w_context, left_seq, right_seq, ref_seq, alt_seq = primedesign_formatter(seq)\n",
    "mut = mutation(wt_w_context, alt_w_context, left_seq, right_seq, var_type, ref_seq, alt_seq, chrom=None, genome=None)\n",
    "mut.PAM_idx_forward, mut.PAM_idx_rc = eligible_PAM_finder(mut, PAM, max_RTT_length, proto_size)\n",
    "\n",
    "\n",
    "\n",
    "peg_df_plus = pegRNA_generator(mut, PAM, '+', proto_size, RTT_lengths, PBS_lengths)\n",
    "peg_df_minus = pegRNA_generator(mut, PAM, '-', proto_size, RTT_lengths, PBS_lengths)\n",
    "\n",
    "#need to translate PAM indexing for the - strand!!!!\n",
    "#convert to genome coordinates as well...\n",
    "pd.concat((peg_df_plus, peg_df_minus)).reset_index().drop(columns='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 738,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CCT'"
      ]
     },
     "execution_count": 738,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut.wt_forward[54:57]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGCCATGCCATGCCATGCCATGCCATGCCAGCTAGGCTAGCTACTACCATAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCCCCC\n",
      "-------------.|||||||||||||||||||---------------------------------------------------------\n",
      "-------------GATGCCATGCCATGCCAGCT---------------------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "alignments2 = aligner.align(mut.wt_rc, \n",
    "                            'GATGCCATGCCATGCCAGCT')\n",
    "print(alignments2[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing on larger set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "p53 = impact[impact['Hugo_Symbol']=='TP53'].drop_duplicates(subset='HGVSc')\n",
    "\n",
    "#or change these to INDELs\n",
    "dels1 = p53[p53['Variant_Type']=='DEL']\n",
    "bad_del_idx = list(dels1[dels1['Tumor_Seq_Allele2']!='-'].index)\n",
    "\n",
    "\n",
    "ins1 = p53[p53['Variant_Type']=='INS']\n",
    "bad_ins_idx = list(ins1[ins1['Reference_Allele']!='-'].index)\n",
    "\n",
    "bad_idxs = bad_del_idx + bad_ins_idx\n",
    "\n",
    "p53_filtered = p53.drop(index=bad_idxs).reset_index().drop(columns=['index'])\n",
    "\n",
    "p53_formatted = df_formatter(p53_filtered, chrom_dict, context_size=100)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_df = p53_formatted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "      <th>left_context</th>\n",
       "      <th>right_context</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.818G&gt;A</td>\n",
       "      <td>p.R273H</td>\n",
       "      <td>7577020</td>\n",
       "      <td>7577220</td>\n",
       "      <td>TCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTC...</td>\n",
       "      <td>TCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTC...</td>\n",
       "      <td>TCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTC...</td>\n",
       "      <td>GCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGAT...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577108</td>\n",
       "      <td>7577108</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.830G&gt;T</td>\n",
       "      <td>p.C277F</td>\n",
       "      <td>7577008</td>\n",
       "      <td>7577208</td>\n",
       "      <td>TGCTTGCTTACCTCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGA...</td>\n",
       "      <td>TGCTTGCTTACCTCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGA...</td>\n",
       "      <td>TGCTTGCTTACCTCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGA...</td>\n",
       "      <td>AGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578239</td>\n",
       "      <td>7578239</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.610G&gt;T</td>\n",
       "      <td>p.E204*</td>\n",
       "      <td>7578139</td>\n",
       "      <td>7578339</td>\n",
       "      <td>TAACCCCTCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCG...</td>\n",
       "      <td>TAACCCCTCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCG...</td>\n",
       "      <td>TAACCCCTCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCG...</td>\n",
       "      <td>CACACGCAAATTTCCTTCCACTCGGATAAGATGCTGAGGAGGGGCC...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577121</td>\n",
       "      <td>7577121</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.817C&gt;T</td>\n",
       "      <td>p.R273C</td>\n",
       "      <td>7577021</td>\n",
       "      <td>7577221</td>\n",
       "      <td>CGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCT...</td>\n",
       "      <td>CGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCT...</td>\n",
       "      <td>CGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCT...</td>\n",
       "      <td>CACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATA...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577506</td>\n",
       "      <td>7577506</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.775G&gt;T</td>\n",
       "      <td>p.D259Y</td>\n",
       "      <td>7577406</td>\n",
       "      <td>7577606</td>\n",
       "      <td>GAAGAAATCGGTAAGAGGTGGGCCCAGGGGTCAGAGGCAAGCAGAG...</td>\n",
       "      <td>GAAGAAATCGGTAAGAGGTGGGCCCAGGGGTCAGAGGCAAGCAGAG...</td>\n",
       "      <td>GAAGAAATCGGTAAGAGGTGGGCCCAGGGGTCAGAGGCAAGCAGAG...</td>\n",
       "      <td>TTCCAGTGTGATGATGGTGAGGATGGGCCTCCGGTTCATGCCGCCC...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3000</th>\n",
       "      <td>3000</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577539</td>\n",
       "      <td>7577541</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GGT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.740_742del</td>\n",
       "      <td>p.N247del</td>\n",
       "      <td>7577439</td>\n",
       "      <td>7577641</td>\n",
       "      <td>GAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGC...</td>\n",
       "      <td>GAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGC...</td>\n",
       "      <td>GAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGC...</td>\n",
       "      <td>TCATGCCGCCCATGCAGGAACTGTTACACATGTAGTTGTAGTGGAT...</td>\n",
       "      <td>GGT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3001</th>\n",
       "      <td>3001</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578438</td>\n",
       "      <td>7578438</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>ENST00000269305.4:c.492G&gt;C</td>\n",
       "      <td>p.K164N</td>\n",
       "      <td>7578338</td>\n",
       "      <td>7578538</td>\n",
       "      <td>AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...</td>\n",
       "      <td>AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...</td>\n",
       "      <td>AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...</td>\n",
       "      <td>TTGTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGG...</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3002</th>\n",
       "      <td>3002</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577531</td>\n",
       "      <td>7577532</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GG</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.749_750delCC</td>\n",
       "      <td>p.P250Hfs*13</td>\n",
       "      <td>7577431</td>\n",
       "      <td>7577632</td>\n",
       "      <td>AGGGGTCAGAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGC...</td>\n",
       "      <td>AGGGGTCAGAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGC...</td>\n",
       "      <td>AGGGGTCAGAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGC...</td>\n",
       "      <td>GCCTCCGGTTCATGCCGCCCATGCAGGAACTGTTACACATGTAGTT...</td>\n",
       "      <td>GG</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3003</th>\n",
       "      <td>3003</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578421</td>\n",
       "      <td>7578440</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GTCATGTGCTGTGACTGCTT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.490_509del</td>\n",
       "      <td>p.K164Gfs*10</td>\n",
       "      <td>7578321</td>\n",
       "      <td>7578540</td>\n",
       "      <td>GGGGACCCTGGGCAACCAGCCCTGTCGTCTCTCCAGCCCCAGCTGC...</td>\n",
       "      <td>GGGGACCCTGGGCAACCAGCCCTGTCGTCTCTCCAGCCCCAGCTGC...</td>\n",
       "      <td>GGGGACCCTGGGCAACCAGCCCTGTCGTCTCTCCAGCCCCAGCTGC...</td>\n",
       "      <td>GTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGGAA...</td>\n",
       "      <td>GTCATGTGCTGTGACTGCTT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3004</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.339C&gt;A</td>\n",
       "      <td>p.F113L</td>\n",
       "      <td>7579248</td>\n",
       "      <td>7579448</td>\n",
       "      <td>AAATGCAGGGGGATACGGCCAGGCATTGAAGTCTCATGGAAGCCAG...</td>\n",
       "      <td>AAATGCAGGGGGATACGGCCAGGCATTGAAGTCTCATGGAAGCCAG...</td>\n",
       "      <td>AAATGCAGGGGGATACGGCCAGGCATTGAAGTCTCATGGAAGCCAG...</td>\n",
       "      <td>AAGCCCAGACGGAAACCGTAGCTGCCCTGGTAGGTTTTCTGGGAAG...</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3005 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0         0        TP53         17         7577120       7577120   \n",
       "1         1        TP53         17         7577108       7577108   \n",
       "2         2        TP53         17         7578239       7578239   \n",
       "3         3        TP53         17         7577121       7577121   \n",
       "4         4        TP53         17         7577506       7577506   \n",
       "...     ...         ...        ...             ...           ...   \n",
       "3000   3000        TP53         17         7577539       7577541   \n",
       "3001   3001        TP53         17         7578438       7578438   \n",
       "3002   3002        TP53         17         7577531       7577532   \n",
       "3003   3003        TP53         17         7578421       7578440   \n",
       "3004   3004        TP53         17         7579348       7579348   \n",
       "\n",
       "             Consequence Variant_Classification Variant_Type  \\\n",
       "0       missense_variant      Missense_Mutation          SNP   \n",
       "1       missense_variant      Missense_Mutation          SNP   \n",
       "2            stop_gained      Nonsense_Mutation          SNP   \n",
       "3       missense_variant      Missense_Mutation          SNP   \n",
       "4       missense_variant      Missense_Mutation          SNP   \n",
       "...                  ...                    ...          ...   \n",
       "3000    inframe_deletion           In_Frame_Del          DEL   \n",
       "3001    missense_variant      Missense_Mutation          SNP   \n",
       "3002  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "3003  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "3004    missense_variant      Missense_Mutation          SNP   \n",
       "\n",
       "          Reference_Allele Tumor_Seq_Allele2  \\\n",
       "0                        C                 T   \n",
       "1                        C                 A   \n",
       "2                        C                 A   \n",
       "3                        G                 A   \n",
       "4                        C                 A   \n",
       "...                    ...               ...   \n",
       "3000                   GGT                 -   \n",
       "3001                     C                 G   \n",
       "3002                    GG                 -   \n",
       "3003  GTCATGTGCTGTGACTGCTT                 -   \n",
       "3004                     G                 T   \n",
       "\n",
       "                                 HGVSc   HGVSp_Short  seq_start  seq_end  \\\n",
       "0           ENST00000269305.4:c.818G>A       p.R273H    7577020  7577220   \n",
       "1           ENST00000269305.4:c.830G>T       p.C277F    7577008  7577208   \n",
       "2           ENST00000269305.4:c.610G>T       p.E204*    7578139  7578339   \n",
       "3           ENST00000269305.4:c.817C>T       p.R273C    7577021  7577221   \n",
       "4           ENST00000269305.4:c.775G>T       p.D259Y    7577406  7577606   \n",
       "...                                ...           ...        ...      ...   \n",
       "3000    ENST00000269305.4:c.740_742del     p.N247del    7577439  7577641   \n",
       "3001        ENST00000269305.4:c.492G>C       p.K164N    7578338  7578538   \n",
       "3002  ENST00000269305.4:c.749_750delCC  p.P250Hfs*13    7577431  7577632   \n",
       "3003    ENST00000269305.4:c.490_509del  p.K164Gfs*10    7578321  7578540   \n",
       "3004        ENST00000269305.4:c.339C>A       p.F113L    7579248  7579448   \n",
       "\n",
       "                                           wt_w_context  \\\n",
       "0     TCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTC...   \n",
       "1     TGCTTGCTTACCTCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGA...   \n",
       "2     TAACCCCTCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCG...   \n",
       "3     CGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCT...   \n",
       "4     GAAGAAATCGGTAAGAGGTGGGCCCAGGGGTCAGAGGCAAGCAGAG...   \n",
       "...                                                 ...   \n",
       "3000  GAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGC...   \n",
       "3001  AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...   \n",
       "3002  AGGGGTCAGAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGC...   \n",
       "3003  GGGGACCCTGGGCAACCAGCCCTGTCGTCTCTCCAGCCCCAGCTGC...   \n",
       "3004  AAATGCAGGGGGATACGGCCAGGCATTGAAGTCTCATGGAAGCCAG...   \n",
       "\n",
       "                                          alt_w_context  \\\n",
       "0     TCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTC...   \n",
       "1     TGCTTGCTTACCTCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGA...   \n",
       "2     TAACCCCTCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCG...   \n",
       "3     CGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCT...   \n",
       "4     GAAGAAATCGGTAAGAGGTGGGCCCAGGGGTCAGAGGCAAGCAGAG...   \n",
       "...                                                 ...   \n",
       "3000  GAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGC...   \n",
       "3001  AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...   \n",
       "3002  AGGGGTCAGAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGC...   \n",
       "3003  GGGGACCCTGGGCAACCAGCCCTGTCGTCTCTCCAGCCCCAGCTGC...   \n",
       "3004  AAATGCAGGGGGATACGGCCAGGCATTGAAGTCTCATGGAAGCCAG...   \n",
       "\n",
       "                                           left_context  \\\n",
       "0     TCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTC...   \n",
       "1     TGCTTGCTTACCTCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGA...   \n",
       "2     TAACCCCTCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCG...   \n",
       "3     CGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCT...   \n",
       "4     GAAGAAATCGGTAAGAGGTGGGCCCAGGGGTCAGAGGCAAGCAGAG...   \n",
       "...                                                 ...   \n",
       "3000  GAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGC...   \n",
       "3001  AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...   \n",
       "3002  AGGGGTCAGAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGC...   \n",
       "3003  GGGGACCCTGGGCAACCAGCCCTGTCGTCTCTCCAGCCCCAGCTGC...   \n",
       "3004  AAATGCAGGGGGATACGGCCAGGCATTGAAGTCTCATGGAAGCCAG...   \n",
       "\n",
       "                                          right_context                   REF  \\\n",
       "0     GCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGAT...                     C   \n",
       "1     AGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACC...                     C   \n",
       "2     CACACGCAAATTTCCTTCCACTCGGATAAGATGCTGAGGAGGGGCC...                     C   \n",
       "3     CACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATA...                     G   \n",
       "4     TTCCAGTGTGATGATGGTGAGGATGGGCCTCCGGTTCATGCCGCCC...                     C   \n",
       "...                                                 ...                   ...   \n",
       "3000  TCATGCCGCCCATGCAGGAACTGTTACACATGTAGTTGTAGTGGAT...                   GGT   \n",
       "3001  TTGTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGG...                     C   \n",
       "3002  GCCTCCGGTTCATGCCGCCCATGCAGGAACTGTTACACATGTAGTT...                    GG   \n",
       "3003  GTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGGAA...  GTCATGTGCTGTGACTGCTT   \n",
       "3004  AAGCCCAGACGGAAACCGTAGCTGCCCTGGTAGGTTTTCTGGGAAG...                     G   \n",
       "\n",
       "     ALT  \n",
       "0      T  \n",
       "1      A  \n",
       "2      A  \n",
       "3      A  \n",
       "4      A  \n",
       "...   ..  \n",
       "3000      \n",
       "3001   G  \n",
       "3002      \n",
       "3003      \n",
       "3004   T  \n",
       "\n",
       "[3005 rows x 20 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--params\n",
    "RTT_lengths = [10,20,30]\n",
    "PBS_lengths = [10,13, 15,17]\n",
    "orientation = '-'\n",
    "proto_size = 19\n",
    "PAM = \"NGG\"\n",
    "\n",
    "combined_peg_dfs = []\n",
    "\n",
    "for i, val in input_df.iterrows():\n",
    "\n",
    "\n",
    "    mut = mutation(val['wt_w_context'], val['alt_w_context'], val['left_context'], val['right_context'], val['Variant_Type'], val['REF'], val['ALT'], chrom=None, genome=None)\n",
    "    mut.PAM_idx_forward, mut.PAM_idx_rc = eligible_PAM_finder(mut, PAM, max(RTT_lengths), proto_size)\n",
    "\n",
    "    peg_df_plus = pegRNA_generator(mut, PAM, '+', proto_size, RTT_lengths, PBS_lengths)\n",
    "    peg_df_minus = pegRNA_generator(mut, PAM, '-', proto_size, RTT_lengths, PBS_lengths)\n",
    "\n",
    "    #need to translate PAM indexing for the - strand!!!!\n",
    "    #convert to genome coordinates as well...\n",
    "    peg_df = pd.concat((peg_df_plus, peg_df_minus))\n",
    "    peg_df['mutation_idx'] = i\n",
    "\n",
    "    combined_peg_dfs.append(peg_df)\n",
    "\n",
    "peg_df = pd.concat(combined_peg_dfs).set_index('mutation_idx').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sensor design\n",
    "\n",
    "- Put it in reverse complement orientation\n",
    "- 5 nt of context prior to protospacer (adjustable)\n",
    "- Also output the edited sensor\n",
    "- Include whether the sensor is large enough to accomodate the desired edit..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>...</th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>RTT</th>\n",
       "      <th>RTT_length</th>\n",
       "      <th>PBS</th>\n",
       "      <th>PBS_length</th>\n",
       "      <th>RTT_PBS</th>\n",
       "      <th>Distance_to_nick</th>\n",
       "      <th>RHA_size</th>\n",
       "      <th>PAM_disrupted</th>\n",
       "      <th>Proto_disrupted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGC</td>\n",
       "      <td>10</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGC</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGCGCA</td>\n",
       "      <td>13</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCA</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGCGCACA</td>\n",
       "      <td>15</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACA</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGCGCACAGA</td>\n",
       "      <td>17</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGA</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>TTTGAGGTGCATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>AGAGACCGGC</td>\n",
       "      <td>10</td>\n",
       "      <td>TTTGAGGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGC</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122567</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>CAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>20</td>\n",
       "      <td>CGGAAACCGTAGCTGCC</td>\n",
       "      <td>17</td>\n",
       "      <td>CAGAATGCAATAAGCCCAGACGGAAACCGTAGCTGCC</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122568</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>30</td>\n",
       "      <td>CGGAAACCGT</td>\n",
       "      <td>10</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGT</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122569</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>30</td>\n",
       "      <td>CGGAAACCGTAGC</td>\n",
       "      <td>13</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGC</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122570</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>30</td>\n",
       "      <td>CGGAAACCGTAGCTG</td>\n",
       "      <td>15</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGCTG</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122571</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>30</td>\n",
       "      <td>CGGAAACCGTAGCTGCC</td>\n",
       "      <td>17</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGCTGCC</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122572 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mutation_idx Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0                  0        TP53         17         7577120       7577120   \n",
       "1                  0        TP53         17         7577120       7577120   \n",
       "2                  0        TP53         17         7577120       7577120   \n",
       "3                  0        TP53         17         7577120       7577120   \n",
       "4                  0        TP53         17         7577120       7577120   \n",
       "...              ...         ...        ...             ...           ...   \n",
       "122567          3004        TP53         17         7579348       7579348   \n",
       "122568          3004        TP53         17         7579348       7579348   \n",
       "122569          3004        TP53         17         7579348       7579348   \n",
       "122570          3004        TP53         17         7579348       7579348   \n",
       "122571          3004        TP53         17         7579348       7579348   \n",
       "\n",
       "             Consequence Variant_Classification Variant_Type Reference_Allele  \\\n",
       "0       missense_variant      Missense_Mutation          SNP                C   \n",
       "1       missense_variant      Missense_Mutation          SNP                C   \n",
       "2       missense_variant      Missense_Mutation          SNP                C   \n",
       "3       missense_variant      Missense_Mutation          SNP                C   \n",
       "4       missense_variant      Missense_Mutation          SNP                C   \n",
       "...                  ...                    ...          ...              ...   \n",
       "122567  missense_variant      Missense_Mutation          SNP                G   \n",
       "122568  missense_variant      Missense_Mutation          SNP                G   \n",
       "122569  missense_variant      Missense_Mutation          SNP                G   \n",
       "122570  missense_variant      Missense_Mutation          SNP                G   \n",
       "122571  missense_variant      Missense_Mutation          SNP                G   \n",
       "\n",
       "       Tumor_Seq_Allele2  ...           Protospacer  \\\n",
       "0                      T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "1                      T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "2                      T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "3                      T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "4                      T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "...                  ...  ...                   ...   \n",
       "122567                 T  ...  GGCAGCTACGGTTTCCGTCT   \n",
       "122568                 T  ...  GGCAGCTACGGTTTCCGTCT   \n",
       "122569                 T  ...  GGCAGCTACGGTTTCCGTCT   \n",
       "122570                 T  ...  GGCAGCTACGGTTTCCGTCT   \n",
       "122571                 T  ...  GGCAGCTACGGTTTCCGTCT   \n",
       "\n",
       "                                   RTT  RTT_length                PBS  \\\n",
       "0                 ATGTTTGTGCCTGTCCTGGG          20         AGAGACCGGC   \n",
       "1                 ATGTTTGTGCCTGTCCTGGG          20      AGAGACCGGCGCA   \n",
       "2                 ATGTTTGTGCCTGTCCTGGG          20    AGAGACCGGCGCACA   \n",
       "3                 ATGTTTGTGCCTGTCCTGGG          20  AGAGACCGGCGCACAGA   \n",
       "4       TTTGAGGTGCATGTTTGTGCCTGTCCTGGG          30         AGAGACCGGC   \n",
       "...                                ...         ...                ...   \n",
       "122567            CAGAATGCAATAAGCCCAGA          20  CGGAAACCGTAGCTGCC   \n",
       "122568  TTGGCTGTCCCAGAATGCAATAAGCCCAGA          30         CGGAAACCGT   \n",
       "122569  TTGGCTGTCCCAGAATGCAATAAGCCCAGA          30      CGGAAACCGTAGC   \n",
       "122570  TTGGCTGTCCCAGAATGCAATAAGCCCAGA          30    CGGAAACCGTAGCTG   \n",
       "122571  TTGGCTGTCCCAGAATGCAATAAGCCCAGA          30  CGGAAACCGTAGCTGCC   \n",
       "\n",
       "       PBS_length                                          RTT_PBS  \\\n",
       "0              10                   ATGTTTGTGCCTGTCCTGGGAGAGACCGGC   \n",
       "1              13                ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCA   \n",
       "2              15              ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACA   \n",
       "3              17            ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGA   \n",
       "4              10         TTTGAGGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGC   \n",
       "...           ...                                              ...   \n",
       "122567         17            CAGAATGCAATAAGCCCAGACGGAAACCGTAGCTGCC   \n",
       "122568         10         TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGT   \n",
       "122569         13      TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGC   \n",
       "122570         15    TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGCTG   \n",
       "122571         17  TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGCTGCC   \n",
       "\n",
       "       Distance_to_nick RHA_size PAM_disrupted Proto_disrupted  \n",
       "0                    19        0         False           False  \n",
       "1                    19        0         False           False  \n",
       "2                    19        0         False           False  \n",
       "3                    19        0         False           False  \n",
       "4                    19       10         False           False  \n",
       "...                 ...      ...           ...             ...  \n",
       "122567                9       10         False           False  \n",
       "122568                9       20         False           False  \n",
       "122569                9       20         False           False  \n",
       "122570                9       20         False           False  \n",
       "122571                9       20         False           False  \n",
       "\n",
       "[122572 rows x 34 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df = pd.merge(input_df, peg_df, on='mutation_idx')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "i=10176\n",
    "#for i in range(100000):\n",
    "wt = df.iloc[i]['wt_w_context']\n",
    "alt = df.iloc[i]['alt_w_context']\n",
    "\n",
    "strand = df.iloc[i]['PAM_strand']\n",
    "ref_minus_alt = len(df.iloc[i]['REF']) - len(df.iloc[i]['ALT'])\n",
    "strand\n",
    "\n",
    "pam_start = df.iloc[i]['PAM_start']\n",
    "pam_end = df.iloc[i]['PAM_end']\n",
    "\n",
    "before_proto_context = 5\n",
    "sensor_length = 60\n",
    "\n",
    "s1 = pam_start-20-before_proto_context\n",
    "if s1 <0:\n",
    "    print('insufficient context sequence provided; try increasing context size')\n",
    "\n",
    "proto_pam = wt[s1:pam_end]\n",
    "proto_pam_alt = alt[s1:pam_end]\n",
    "\n",
    "remain_len = sensor_length - len(proto_pam)\n",
    "if remain_len < 0:\n",
    "    print('sensor too small; increase sensor size in parameters; or decrease before_proto_context')\n",
    "remain_len_alt = sensor_length - len(proto_pam) - ref_minus_alt\n",
    "\n",
    "e1 = pam_end + remain_len\n",
    "e2 = pam_end + remain_len_alt\n",
    "if e1>len(wt):\n",
    "    print('insufficient context sequence provided; try increasing context size')\n",
    "\n",
    "rtt_region = wt[pam_end: e1]\n",
    "rtt_region_alt = alt[pam_end:e2]\n",
    "\n",
    "\n",
    "sensor_wt = proto_pam + rtt_region\n",
    "sensor_alt = proto_pam_alt + rtt_region_alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGGGCATCCTTGAGTTCCAAGGCCTC-ATTCAGCTCTCGGAACATCTCGAAGCGCTCACGC\n",
      "||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||\n",
      "TGGGCATCCTTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAACATCTCGAAGCGCTCACGC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "alignments = aligner.align(sensor_wt, sensor_alt)\n",
    "print(alignments[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CTTCCAGTGTGATGATGGTGAGGATGCC'"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proto_pam_alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>...</th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>RTT</th>\n",
       "      <th>RTT_length</th>\n",
       "      <th>PBS</th>\n",
       "      <th>PBS_length</th>\n",
       "      <th>RTT_PBS</th>\n",
       "      <th>Distance_to_nick</th>\n",
       "      <th>RHA_size</th>\n",
       "      <th>PAM_disrupted</th>\n",
       "      <th>Proto_disrupted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10176</th>\n",
       "      <td>259</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7573991</td>\n",
       "      <td>7573992</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GATGTTCCGAGAGCTGAATG</td>\n",
       "      <td>AGGCCTCAAT</td>\n",
       "      <td>10</td>\n",
       "      <td>TCAGCTCTCG</td>\n",
       "      <td>10</td>\n",
       "      <td>AGGCCTCAATTCAGCTCTCG</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10177</th>\n",
       "      <td>259</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7573991</td>\n",
       "      <td>7573992</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GATGTTCCGAGAGCTGAATG</td>\n",
       "      <td>AGGCCTCAAT</td>\n",
       "      <td>10</td>\n",
       "      <td>TCAGCTCTCGGAA</td>\n",
       "      <td>13</td>\n",
       "      <td>AGGCCTCAATTCAGCTCTCGGAA</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10178</th>\n",
       "      <td>259</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7573991</td>\n",
       "      <td>7573992</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GATGTTCCGAGAGCTGAATG</td>\n",
       "      <td>AGGCCTCAAT</td>\n",
       "      <td>10</td>\n",
       "      <td>TCAGCTCTCGGAACA</td>\n",
       "      <td>15</td>\n",
       "      <td>AGGCCTCAATTCAGCTCTCGGAACA</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10179</th>\n",
       "      <td>259</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7573991</td>\n",
       "      <td>7573992</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GATGTTCCGAGAGCTGAATG</td>\n",
       "      <td>AGGCCTCAAT</td>\n",
       "      <td>10</td>\n",
       "      <td>TCAGCTCTCGGAACATC</td>\n",
       "      <td>17</td>\n",
       "      <td>AGGCCTCAATTCAGCTCTCGGAACATC</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10180</th>\n",
       "      <td>259</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7573991</td>\n",
       "      <td>7573992</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GATGTTCCGAGAGCTGAATG</td>\n",
       "      <td>TTGAGTTCCAAGGCCTCAAT</td>\n",
       "      <td>20</td>\n",
       "      <td>TCAGCTCTCG</td>\n",
       "      <td>10</td>\n",
       "      <td>TTGAGTTCCAAGGCCTCAATTCAGCTCTCG</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10181</th>\n",
       "      <td>259</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7573991</td>\n",
       "      <td>7573992</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GATGTTCCGAGAGCTGAATG</td>\n",
       "      <td>TTGAGTTCCAAGGCCTCAAT</td>\n",
       "      <td>20</td>\n",
       "      <td>TCAGCTCTCGGAA</td>\n",
       "      <td>13</td>\n",
       "      <td>TTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAA</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10182</th>\n",
       "      <td>259</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7573991</td>\n",
       "      <td>7573992</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GATGTTCCGAGAGCTGAATG</td>\n",
       "      <td>TTGAGTTCCAAGGCCTCAAT</td>\n",
       "      <td>20</td>\n",
       "      <td>TCAGCTCTCGGAACA</td>\n",
       "      <td>15</td>\n",
       "      <td>TTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAACA</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10183</th>\n",
       "      <td>259</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7573991</td>\n",
       "      <td>7573992</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GATGTTCCGAGAGCTGAATG</td>\n",
       "      <td>TTGAGTTCCAAGGCCTCAAT</td>\n",
       "      <td>20</td>\n",
       "      <td>TCAGCTCTCGGAACATC</td>\n",
       "      <td>17</td>\n",
       "      <td>TTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAACATC</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10184</th>\n",
       "      <td>259</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7573991</td>\n",
       "      <td>7573992</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GATGTTCCGAGAGCTGAATG</td>\n",
       "      <td>CTGGGCATCCTTGAGTTCCAAGGCCTCAAT</td>\n",
       "      <td>30</td>\n",
       "      <td>TCAGCTCTCG</td>\n",
       "      <td>10</td>\n",
       "      <td>CTGGGCATCCTTGAGTTCCAAGGCCTCAATTCAGCTCTCG</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10185</th>\n",
       "      <td>259</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7573991</td>\n",
       "      <td>7573992</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GATGTTCCGAGAGCTGAATG</td>\n",
       "      <td>CTGGGCATCCTTGAGTTCCAAGGCCTCAAT</td>\n",
       "      <td>30</td>\n",
       "      <td>TCAGCTCTCGGAA</td>\n",
       "      <td>13</td>\n",
       "      <td>CTGGGCATCCTTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAA</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10186</th>\n",
       "      <td>259</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7573991</td>\n",
       "      <td>7573992</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GATGTTCCGAGAGCTGAATG</td>\n",
       "      <td>CTGGGCATCCTTGAGTTCCAAGGCCTCAAT</td>\n",
       "      <td>30</td>\n",
       "      <td>TCAGCTCTCGGAACA</td>\n",
       "      <td>15</td>\n",
       "      <td>CTGGGCATCCTTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAACA</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10187</th>\n",
       "      <td>259</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7573991</td>\n",
       "      <td>7573992</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GATGTTCCGAGAGCTGAATG</td>\n",
       "      <td>CTGGGCATCCTTGAGTTCCAAGGCCTCAAT</td>\n",
       "      <td>30</td>\n",
       "      <td>TCAGCTCTCGGAACATC</td>\n",
       "      <td>17</td>\n",
       "      <td>CTGGGCATCCTTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAACATC</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11548</th>\n",
       "      <td>292</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578508</td>\n",
       "      <td>7578509</td>\n",
       "      <td>stop_gained,frameshift_variant</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCCACAGCTGCACAGGGC</td>\n",
       "      <td>AGACCTAGCC</td>\n",
       "      <td>10</td>\n",
       "      <td>CTGTGCAGCT</td>\n",
       "      <td>10</td>\n",
       "      <td>AGACCTAGCCCTGTGCAGCT</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11549</th>\n",
       "      <td>292</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578508</td>\n",
       "      <td>7578509</td>\n",
       "      <td>stop_gained,frameshift_variant</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCCACAGCTGCACAGGGC</td>\n",
       "      <td>AGACCTAGCC</td>\n",
       "      <td>10</td>\n",
       "      <td>CTGTGCAGCTGTG</td>\n",
       "      <td>13</td>\n",
       "      <td>AGACCTAGCCCTGTGCAGCTGTG</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11550</th>\n",
       "      <td>292</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578508</td>\n",
       "      <td>7578509</td>\n",
       "      <td>stop_gained,frameshift_variant</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCCACAGCTGCACAGGGC</td>\n",
       "      <td>AGACCTAGCC</td>\n",
       "      <td>10</td>\n",
       "      <td>CTGTGCAGCTGTGGG</td>\n",
       "      <td>15</td>\n",
       "      <td>AGACCTAGCCCTGTGCAGCTGTGGG</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11551</th>\n",
       "      <td>292</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578508</td>\n",
       "      <td>7578509</td>\n",
       "      <td>stop_gained,frameshift_variant</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCCACAGCTGCACAGGGC</td>\n",
       "      <td>AGACCTAGCC</td>\n",
       "      <td>10</td>\n",
       "      <td>CTGTGCAGCTGTGGGTT</td>\n",
       "      <td>17</td>\n",
       "      <td>AGACCTAGCCCTGTGCAGCTGTGGGTT</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11552</th>\n",
       "      <td>292</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578508</td>\n",
       "      <td>7578509</td>\n",
       "      <td>stop_gained,frameshift_variant</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCCACAGCTGCACAGGGC</td>\n",
       "      <td>CAACTGGCCAAGACCTAGCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CTGTGCAGCT</td>\n",
       "      <td>10</td>\n",
       "      <td>CAACTGGCCAAGACCTAGCCCTGTGCAGCT</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11553</th>\n",
       "      <td>292</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578508</td>\n",
       "      <td>7578509</td>\n",
       "      <td>stop_gained,frameshift_variant</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCCACAGCTGCACAGGGC</td>\n",
       "      <td>CAACTGGCCAAGACCTAGCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CTGTGCAGCTGTG</td>\n",
       "      <td>13</td>\n",
       "      <td>CAACTGGCCAAGACCTAGCCCTGTGCAGCTGTG</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11554</th>\n",
       "      <td>292</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578508</td>\n",
       "      <td>7578509</td>\n",
       "      <td>stop_gained,frameshift_variant</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCCACAGCTGCACAGGGC</td>\n",
       "      <td>CAACTGGCCAAGACCTAGCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CTGTGCAGCTGTGGG</td>\n",
       "      <td>15</td>\n",
       "      <td>CAACTGGCCAAGACCTAGCCCTGTGCAGCTGTGGG</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11555</th>\n",
       "      <td>292</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578508</td>\n",
       "      <td>7578509</td>\n",
       "      <td>stop_gained,frameshift_variant</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCCACAGCTGCACAGGGC</td>\n",
       "      <td>CAACTGGCCAAGACCTAGCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CTGTGCAGCTGTGGGTT</td>\n",
       "      <td>17</td>\n",
       "      <td>CAACTGGCCAAGACCTAGCCCTGTGCAGCTGTGGGTT</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11556</th>\n",
       "      <td>292</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578508</td>\n",
       "      <td>7578509</td>\n",
       "      <td>stop_gained,frameshift_variant</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCCACAGCTGCACAGGGC</td>\n",
       "      <td>GATGTTTTGCCAACTGGCCAAGACCTAGCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CTGTGCAGCT</td>\n",
       "      <td>10</td>\n",
       "      <td>GATGTTTTGCCAACTGGCCAAGACCTAGCCCTGTGCAGCT</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11557</th>\n",
       "      <td>292</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578508</td>\n",
       "      <td>7578509</td>\n",
       "      <td>stop_gained,frameshift_variant</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCCACAGCTGCACAGGGC</td>\n",
       "      <td>GATGTTTTGCCAACTGGCCAAGACCTAGCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CTGTGCAGCTGTG</td>\n",
       "      <td>13</td>\n",
       "      <td>GATGTTTTGCCAACTGGCCAAGACCTAGCCCTGTGCAGCTGTG</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11558</th>\n",
       "      <td>292</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578508</td>\n",
       "      <td>7578509</td>\n",
       "      <td>stop_gained,frameshift_variant</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCCACAGCTGCACAGGGC</td>\n",
       "      <td>GATGTTTTGCCAACTGGCCAAGACCTAGCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CTGTGCAGCTGTGGG</td>\n",
       "      <td>15</td>\n",
       "      <td>GATGTTTTGCCAACTGGCCAAGACCTAGCCCTGTGCAGCTGTGGG</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11559</th>\n",
       "      <td>292</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578508</td>\n",
       "      <td>7578509</td>\n",
       "      <td>stop_gained,frameshift_variant</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCCACAGCTGCACAGGGC</td>\n",
       "      <td>GATGTTTTGCCAACTGGCCAAGACCTAGCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CTGTGCAGCTGTGGGTT</td>\n",
       "      <td>17</td>\n",
       "      <td>GATGTTTTGCCAACTGGCCAAGACCTAGCCCTGTGCAGCTGTGGGTT</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11608</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACAGAAGATGACAGGGGCC</td>\n",
       "      <td>CCTCCTTGGC</td>\n",
       "      <td>10</td>\n",
       "      <td>CCCTGTCATC</td>\n",
       "      <td>10</td>\n",
       "      <td>CCTCCTTGGCCCCTGTCATC</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11609</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACAGAAGATGACAGGGGCC</td>\n",
       "      <td>CCTCCTTGGC</td>\n",
       "      <td>10</td>\n",
       "      <td>CCCTGTCATCTTC</td>\n",
       "      <td>13</td>\n",
       "      <td>CCTCCTTGGCCCCTGTCATCTTC</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11610</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACAGAAGATGACAGGGGCC</td>\n",
       "      <td>CCTCCTTGGC</td>\n",
       "      <td>10</td>\n",
       "      <td>CCCTGTCATCTTCTG</td>\n",
       "      <td>15</td>\n",
       "      <td>CCTCCTTGGCCCCTGTCATCTTCTG</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11611</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACAGAAGATGACAGGGGCC</td>\n",
       "      <td>CCTCCTTGGC</td>\n",
       "      <td>10</td>\n",
       "      <td>CCCTGTCATCTTCTGTC</td>\n",
       "      <td>17</td>\n",
       "      <td>CCTCCTTGGCCCCTGTCATCTTCTGTC</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11612</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACAGAAGATGACAGGGGCC</td>\n",
       "      <td>GCACCAGCCCCCTCCTTGGC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCTGTCATC</td>\n",
       "      <td>10</td>\n",
       "      <td>GCACCAGCCCCCTCCTTGGCCCCTGTCATC</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11613</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACAGAAGATGACAGGGGCC</td>\n",
       "      <td>GCACCAGCCCCCTCCTTGGC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCTGTCATCTTC</td>\n",
       "      <td>13</td>\n",
       "      <td>GCACCAGCCCCCTCCTTGGCCCCTGTCATCTTC</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11614</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACAGAAGATGACAGGGGCC</td>\n",
       "      <td>GCACCAGCCCCCTCCTTGGC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCTGTCATCTTCTG</td>\n",
       "      <td>15</td>\n",
       "      <td>GCACCAGCCCCCTCCTTGGCCCCTGTCATCTTCTG</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11615</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACAGAAGATGACAGGGGCC</td>\n",
       "      <td>GCACCAGCCCCCTCCTTGGC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCTGTCATCTTCTGTC</td>\n",
       "      <td>17</td>\n",
       "      <td>GCACCAGCCCCCTCCTTGGCCCCTGTCATCTTCTGTC</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11616</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACAGAAGATGACAGGGGCC</td>\n",
       "      <td>GGCGGCCCCTGCACCAGCCCCCTCCTTGGC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCTGTCATC</td>\n",
       "      <td>10</td>\n",
       "      <td>GGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATC</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11617</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACAGAAGATGACAGGGGCC</td>\n",
       "      <td>GGCGGCCCCTGCACCAGCCCCCTCCTTGGC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCTGTCATCTTC</td>\n",
       "      <td>13</td>\n",
       "      <td>GGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTTC</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11618</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACAGAAGATGACAGGGGCC</td>\n",
       "      <td>GGCGGCCCCTGCACCAGCCCCCTCCTTGGC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCTGTCATCTTCTG</td>\n",
       "      <td>15</td>\n",
       "      <td>GGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTTCTG</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11619</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACAGAAGATGACAGGGGCC</td>\n",
       "      <td>GGCGGCCCCTGCACCAGCCCCCTCCTTGGC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCTGTCATCTTCTGTC</td>\n",
       "      <td>17</td>\n",
       "      <td>GGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTTCTGTC</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11620</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAAGATGACAGGGGCCAGG</td>\n",
       "      <td>CCCCCTCCTT</td>\n",
       "      <td>10</td>\n",
       "      <td>GGCCCCTGTC</td>\n",
       "      <td>10</td>\n",
       "      <td>CCCCCTCCTTGGCCCCTGTC</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11621</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAAGATGACAGGGGCCAGG</td>\n",
       "      <td>CCCCCTCCTT</td>\n",
       "      <td>10</td>\n",
       "      <td>GGCCCCTGTCATC</td>\n",
       "      <td>13</td>\n",
       "      <td>CCCCCTCCTTGGCCCCTGTCATC</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11622</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAAGATGACAGGGGCCAGG</td>\n",
       "      <td>CCCCCTCCTT</td>\n",
       "      <td>10</td>\n",
       "      <td>GGCCCCTGTCATCTT</td>\n",
       "      <td>15</td>\n",
       "      <td>CCCCCTCCTTGGCCCCTGTCATCTT</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11623</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAAGATGACAGGGGCCAGG</td>\n",
       "      <td>CCCCCTCCTT</td>\n",
       "      <td>10</td>\n",
       "      <td>GGCCCCTGTCATCTTCT</td>\n",
       "      <td>17</td>\n",
       "      <td>CCCCCTCCTTGGCCCCTGTCATCTTCT</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11624</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAAGATGACAGGGGCCAGG</td>\n",
       "      <td>CCTGCACCAGCCCCCTCCTT</td>\n",
       "      <td>20</td>\n",
       "      <td>GGCCCCTGTC</td>\n",
       "      <td>10</td>\n",
       "      <td>CCTGCACCAGCCCCCTCCTTGGCCCCTGTC</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11625</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAAGATGACAGGGGCCAGG</td>\n",
       "      <td>CCTGCACCAGCCCCCTCCTT</td>\n",
       "      <td>20</td>\n",
       "      <td>GGCCCCTGTCATC</td>\n",
       "      <td>13</td>\n",
       "      <td>CCTGCACCAGCCCCCTCCTTGGCCCCTGTCATC</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11626</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAAGATGACAGGGGCCAGG</td>\n",
       "      <td>CCTGCACCAGCCCCCTCCTT</td>\n",
       "      <td>20</td>\n",
       "      <td>GGCCCCTGTCATCTT</td>\n",
       "      <td>15</td>\n",
       "      <td>CCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTT</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11627</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAAGATGACAGGGGCCAGG</td>\n",
       "      <td>CCTGCACCAGCCCCCTCCTT</td>\n",
       "      <td>20</td>\n",
       "      <td>GGCCCCTGTCATCTTCT</td>\n",
       "      <td>17</td>\n",
       "      <td>CCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTTCT</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11628</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAAGATGACAGGGGCCAGG</td>\n",
       "      <td>ACCGGCGGCCCCTGCACCAGCCCCCTCCTT</td>\n",
       "      <td>30</td>\n",
       "      <td>GGCCCCTGTC</td>\n",
       "      <td>10</td>\n",
       "      <td>ACCGGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTC</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11629</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAAGATGACAGGGGCCAGG</td>\n",
       "      <td>ACCGGCGGCCCCTGCACCAGCCCCCTCCTT</td>\n",
       "      <td>30</td>\n",
       "      <td>GGCCCCTGTCATC</td>\n",
       "      <td>13</td>\n",
       "      <td>ACCGGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATC</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11630</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAAGATGACAGGGGCCAGG</td>\n",
       "      <td>ACCGGCGGCCCCTGCACCAGCCCCCTCCTT</td>\n",
       "      <td>30</td>\n",
       "      <td>GGCCCCTGTCATCTT</td>\n",
       "      <td>15</td>\n",
       "      <td>ACCGGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTT</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11631</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAAGATGACAGGGGCCAGG</td>\n",
       "      <td>ACCGGCGGCCCCTGCACCAGCCCCCTCCTT</td>\n",
       "      <td>30</td>\n",
       "      <td>GGCCCCTGTCATCTTCT</td>\n",
       "      <td>17</td>\n",
       "      <td>ACCGGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTTCT</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11636</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCCTGCACCAGCCCCCTCC</td>\n",
       "      <td>GGGCCAAGGA</td>\n",
       "      <td>10</td>\n",
       "      <td>GGGGGCTGGT</td>\n",
       "      <td>10</td>\n",
       "      <td>GGGCCAAGGAGGGGGCTGGT</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11637</th>\n",
       "      <td>293</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579416</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCCTGCACCAGCCCCCTCC</td>\n",
       "      <td>GGGCCAAGGA</td>\n",
       "      <td>10</td>\n",
       "      <td>GGGGGCTGGTGCA</td>\n",
       "      <td>13</td>\n",
       "      <td>GGGCCAAGGAGGGGGCTGGTGCA</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation_idx Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "10176           259        TP53         17         7573991       7573992   \n",
       "10177           259        TP53         17         7573991       7573992   \n",
       "10178           259        TP53         17         7573991       7573992   \n",
       "10179           259        TP53         17         7573991       7573992   \n",
       "10180           259        TP53         17         7573991       7573992   \n",
       "10181           259        TP53         17         7573991       7573992   \n",
       "10182           259        TP53         17         7573991       7573992   \n",
       "10183           259        TP53         17         7573991       7573992   \n",
       "10184           259        TP53         17         7573991       7573992   \n",
       "10185           259        TP53         17         7573991       7573992   \n",
       "10186           259        TP53         17         7573991       7573992   \n",
       "10187           259        TP53         17         7573991       7573992   \n",
       "11548           292        TP53         17         7578508       7578509   \n",
       "11549           292        TP53         17         7578508       7578509   \n",
       "11550           292        TP53         17         7578508       7578509   \n",
       "11551           292        TP53         17         7578508       7578509   \n",
       "11552           292        TP53         17         7578508       7578509   \n",
       "11553           292        TP53         17         7578508       7578509   \n",
       "11554           292        TP53         17         7578508       7578509   \n",
       "11555           292        TP53         17         7578508       7578509   \n",
       "11556           292        TP53         17         7578508       7578509   \n",
       "11557           292        TP53         17         7578508       7578509   \n",
       "11558           292        TP53         17         7578508       7578509   \n",
       "11559           292        TP53         17         7578508       7578509   \n",
       "11608           293        TP53         17         7579415       7579416   \n",
       "11609           293        TP53         17         7579415       7579416   \n",
       "11610           293        TP53         17         7579415       7579416   \n",
       "11611           293        TP53         17         7579415       7579416   \n",
       "11612           293        TP53         17         7579415       7579416   \n",
       "11613           293        TP53         17         7579415       7579416   \n",
       "11614           293        TP53         17         7579415       7579416   \n",
       "11615           293        TP53         17         7579415       7579416   \n",
       "11616           293        TP53         17         7579415       7579416   \n",
       "11617           293        TP53         17         7579415       7579416   \n",
       "11618           293        TP53         17         7579415       7579416   \n",
       "11619           293        TP53         17         7579415       7579416   \n",
       "11620           293        TP53         17         7579415       7579416   \n",
       "11621           293        TP53         17         7579415       7579416   \n",
       "11622           293        TP53         17         7579415       7579416   \n",
       "11623           293        TP53         17         7579415       7579416   \n",
       "11624           293        TP53         17         7579415       7579416   \n",
       "11625           293        TP53         17         7579415       7579416   \n",
       "11626           293        TP53         17         7579415       7579416   \n",
       "11627           293        TP53         17         7579415       7579416   \n",
       "11628           293        TP53         17         7579415       7579416   \n",
       "11629           293        TP53         17         7579415       7579416   \n",
       "11630           293        TP53         17         7579415       7579416   \n",
       "11631           293        TP53         17         7579415       7579416   \n",
       "11636           293        TP53         17         7579415       7579416   \n",
       "11637           293        TP53         17         7579415       7579416   \n",
       "\n",
       "                          Consequence Variant_Classification Variant_Type  \\\n",
       "10176              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "10177              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "10178              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "10179              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "10180              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "10181              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "10182              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "10183              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "10184              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "10185              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "10186              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "10187              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11548  stop_gained,frameshift_variant      Nonsense_Mutation          INS   \n",
       "11549  stop_gained,frameshift_variant      Nonsense_Mutation          INS   \n",
       "11550  stop_gained,frameshift_variant      Nonsense_Mutation          INS   \n",
       "11551  stop_gained,frameshift_variant      Nonsense_Mutation          INS   \n",
       "11552  stop_gained,frameshift_variant      Nonsense_Mutation          INS   \n",
       "11553  stop_gained,frameshift_variant      Nonsense_Mutation          INS   \n",
       "11554  stop_gained,frameshift_variant      Nonsense_Mutation          INS   \n",
       "11555  stop_gained,frameshift_variant      Nonsense_Mutation          INS   \n",
       "11556  stop_gained,frameshift_variant      Nonsense_Mutation          INS   \n",
       "11557  stop_gained,frameshift_variant      Nonsense_Mutation          INS   \n",
       "11558  stop_gained,frameshift_variant      Nonsense_Mutation          INS   \n",
       "11559  stop_gained,frameshift_variant      Nonsense_Mutation          INS   \n",
       "11608              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11609              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11610              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11611              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11612              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11613              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11614              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11615              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11616              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11617              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11618              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11619              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11620              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11621              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11622              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11623              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11624              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11625              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11626              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11627              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11628              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11629              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11630              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11631              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11636              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "11637              frameshift_variant        Frame_Shift_Ins          INS   \n",
       "\n",
       "      Reference_Allele Tumor_Seq_Allele2  ...           Protospacer  \\\n",
       "10176                -                 A  ...  GATGTTCCGAGAGCTGAATG   \n",
       "10177                -                 A  ...  GATGTTCCGAGAGCTGAATG   \n",
       "10178                -                 A  ...  GATGTTCCGAGAGCTGAATG   \n",
       "10179                -                 A  ...  GATGTTCCGAGAGCTGAATG   \n",
       "10180                -                 A  ...  GATGTTCCGAGAGCTGAATG   \n",
       "10181                -                 A  ...  GATGTTCCGAGAGCTGAATG   \n",
       "10182                -                 A  ...  GATGTTCCGAGAGCTGAATG   \n",
       "10183                -                 A  ...  GATGTTCCGAGAGCTGAATG   \n",
       "10184                -                 A  ...  GATGTTCCGAGAGCTGAATG   \n",
       "10185                -                 A  ...  GATGTTCCGAGAGCTGAATG   \n",
       "10186                -                 A  ...  GATGTTCCGAGAGCTGAATG   \n",
       "10187                -                 A  ...  GATGTTCCGAGAGCTGAATG   \n",
       "11548                -                 T  ...  GACCCACAGCTGCACAGGGC   \n",
       "11549                -                 T  ...  GACCCACAGCTGCACAGGGC   \n",
       "11550                -                 T  ...  GACCCACAGCTGCACAGGGC   \n",
       "11551                -                 T  ...  GACCCACAGCTGCACAGGGC   \n",
       "11552                -                 T  ...  GACCCACAGCTGCACAGGGC   \n",
       "11553                -                 T  ...  GACCCACAGCTGCACAGGGC   \n",
       "11554                -                 T  ...  GACCCACAGCTGCACAGGGC   \n",
       "11555                -                 T  ...  GACCCACAGCTGCACAGGGC   \n",
       "11556                -                 T  ...  GACCCACAGCTGCACAGGGC   \n",
       "11557                -                 T  ...  GACCCACAGCTGCACAGGGC   \n",
       "11558                -                 T  ...  GACCCACAGCTGCACAGGGC   \n",
       "11559                -                 T  ...  GACCCACAGCTGCACAGGGC   \n",
       "11608                -                 A  ...  GACAGAAGATGACAGGGGCC   \n",
       "11609                -                 A  ...  GACAGAAGATGACAGGGGCC   \n",
       "11610                -                 A  ...  GACAGAAGATGACAGGGGCC   \n",
       "11611                -                 A  ...  GACAGAAGATGACAGGGGCC   \n",
       "11612                -                 A  ...  GACAGAAGATGACAGGGGCC   \n",
       "11613                -                 A  ...  GACAGAAGATGACAGGGGCC   \n",
       "11614                -                 A  ...  GACAGAAGATGACAGGGGCC   \n",
       "11615                -                 A  ...  GACAGAAGATGACAGGGGCC   \n",
       "11616                -                 A  ...  GACAGAAGATGACAGGGGCC   \n",
       "11617                -                 A  ...  GACAGAAGATGACAGGGGCC   \n",
       "11618                -                 A  ...  GACAGAAGATGACAGGGGCC   \n",
       "11619                -                 A  ...  GACAGAAGATGACAGGGGCC   \n",
       "11620                -                 A  ...  GGAAGATGACAGGGGCCAGG   \n",
       "11621                -                 A  ...  GGAAGATGACAGGGGCCAGG   \n",
       "11622                -                 A  ...  GGAAGATGACAGGGGCCAGG   \n",
       "11623                -                 A  ...  GGAAGATGACAGGGGCCAGG   \n",
       "11624                -                 A  ...  GGAAGATGACAGGGGCCAGG   \n",
       "11625                -                 A  ...  GGAAGATGACAGGGGCCAGG   \n",
       "11626                -                 A  ...  GGAAGATGACAGGGGCCAGG   \n",
       "11627                -                 A  ...  GGAAGATGACAGGGGCCAGG   \n",
       "11628                -                 A  ...  GGAAGATGACAGGGGCCAGG   \n",
       "11629                -                 A  ...  GGAAGATGACAGGGGCCAGG   \n",
       "11630                -                 A  ...  GGAAGATGACAGGGGCCAGG   \n",
       "11631                -                 A  ...  GGAAGATGACAGGGGCCAGG   \n",
       "11636                -                 A  ...  GCCCTGCACCAGCCCCCTCC   \n",
       "11637                -                 A  ...  GCCCTGCACCAGCCCCCTCC   \n",
       "\n",
       "                                  RTT  RTT_length                PBS  \\\n",
       "10176                      AGGCCTCAAT          10         TCAGCTCTCG   \n",
       "10177                      AGGCCTCAAT          10      TCAGCTCTCGGAA   \n",
       "10178                      AGGCCTCAAT          10    TCAGCTCTCGGAACA   \n",
       "10179                      AGGCCTCAAT          10  TCAGCTCTCGGAACATC   \n",
       "10180            TTGAGTTCCAAGGCCTCAAT          20         TCAGCTCTCG   \n",
       "10181            TTGAGTTCCAAGGCCTCAAT          20      TCAGCTCTCGGAA   \n",
       "10182            TTGAGTTCCAAGGCCTCAAT          20    TCAGCTCTCGGAACA   \n",
       "10183            TTGAGTTCCAAGGCCTCAAT          20  TCAGCTCTCGGAACATC   \n",
       "10184  CTGGGCATCCTTGAGTTCCAAGGCCTCAAT          30         TCAGCTCTCG   \n",
       "10185  CTGGGCATCCTTGAGTTCCAAGGCCTCAAT          30      TCAGCTCTCGGAA   \n",
       "10186  CTGGGCATCCTTGAGTTCCAAGGCCTCAAT          30    TCAGCTCTCGGAACA   \n",
       "10187  CTGGGCATCCTTGAGTTCCAAGGCCTCAAT          30  TCAGCTCTCGGAACATC   \n",
       "11548                      AGACCTAGCC          10         CTGTGCAGCT   \n",
       "11549                      AGACCTAGCC          10      CTGTGCAGCTGTG   \n",
       "11550                      AGACCTAGCC          10    CTGTGCAGCTGTGGG   \n",
       "11551                      AGACCTAGCC          10  CTGTGCAGCTGTGGGTT   \n",
       "11552            CAACTGGCCAAGACCTAGCC          20         CTGTGCAGCT   \n",
       "11553            CAACTGGCCAAGACCTAGCC          20      CTGTGCAGCTGTG   \n",
       "11554            CAACTGGCCAAGACCTAGCC          20    CTGTGCAGCTGTGGG   \n",
       "11555            CAACTGGCCAAGACCTAGCC          20  CTGTGCAGCTGTGGGTT   \n",
       "11556  GATGTTTTGCCAACTGGCCAAGACCTAGCC          30         CTGTGCAGCT   \n",
       "11557  GATGTTTTGCCAACTGGCCAAGACCTAGCC          30      CTGTGCAGCTGTG   \n",
       "11558  GATGTTTTGCCAACTGGCCAAGACCTAGCC          30    CTGTGCAGCTGTGGG   \n",
       "11559  GATGTTTTGCCAACTGGCCAAGACCTAGCC          30  CTGTGCAGCTGTGGGTT   \n",
       "11608                      CCTCCTTGGC          10         CCCTGTCATC   \n",
       "11609                      CCTCCTTGGC          10      CCCTGTCATCTTC   \n",
       "11610                      CCTCCTTGGC          10    CCCTGTCATCTTCTG   \n",
       "11611                      CCTCCTTGGC          10  CCCTGTCATCTTCTGTC   \n",
       "11612            GCACCAGCCCCCTCCTTGGC          20         CCCTGTCATC   \n",
       "11613            GCACCAGCCCCCTCCTTGGC          20      CCCTGTCATCTTC   \n",
       "11614            GCACCAGCCCCCTCCTTGGC          20    CCCTGTCATCTTCTG   \n",
       "11615            GCACCAGCCCCCTCCTTGGC          20  CCCTGTCATCTTCTGTC   \n",
       "11616  GGCGGCCCCTGCACCAGCCCCCTCCTTGGC          30         CCCTGTCATC   \n",
       "11617  GGCGGCCCCTGCACCAGCCCCCTCCTTGGC          30      CCCTGTCATCTTC   \n",
       "11618  GGCGGCCCCTGCACCAGCCCCCTCCTTGGC          30    CCCTGTCATCTTCTG   \n",
       "11619  GGCGGCCCCTGCACCAGCCCCCTCCTTGGC          30  CCCTGTCATCTTCTGTC   \n",
       "11620                      CCCCCTCCTT          10         GGCCCCTGTC   \n",
       "11621                      CCCCCTCCTT          10      GGCCCCTGTCATC   \n",
       "11622                      CCCCCTCCTT          10    GGCCCCTGTCATCTT   \n",
       "11623                      CCCCCTCCTT          10  GGCCCCTGTCATCTTCT   \n",
       "11624            CCTGCACCAGCCCCCTCCTT          20         GGCCCCTGTC   \n",
       "11625            CCTGCACCAGCCCCCTCCTT          20      GGCCCCTGTCATC   \n",
       "11626            CCTGCACCAGCCCCCTCCTT          20    GGCCCCTGTCATCTT   \n",
       "11627            CCTGCACCAGCCCCCTCCTT          20  GGCCCCTGTCATCTTCT   \n",
       "11628  ACCGGCGGCCCCTGCACCAGCCCCCTCCTT          30         GGCCCCTGTC   \n",
       "11629  ACCGGCGGCCCCTGCACCAGCCCCCTCCTT          30      GGCCCCTGTCATC   \n",
       "11630  ACCGGCGGCCCCTGCACCAGCCCCCTCCTT          30    GGCCCCTGTCATCTT   \n",
       "11631  ACCGGCGGCCCCTGCACCAGCCCCCTCCTT          30  GGCCCCTGTCATCTTCT   \n",
       "11636                      GGGCCAAGGA          10         GGGGGCTGGT   \n",
       "11637                      GGGCCAAGGA          10      GGGGGCTGGTGCA   \n",
       "\n",
       "      PBS_length                                          RTT_PBS  \\\n",
       "10176         10                             AGGCCTCAATTCAGCTCTCG   \n",
       "10177         13                          AGGCCTCAATTCAGCTCTCGGAA   \n",
       "10178         15                        AGGCCTCAATTCAGCTCTCGGAACA   \n",
       "10179         17                      AGGCCTCAATTCAGCTCTCGGAACATC   \n",
       "10180         10                   TTGAGTTCCAAGGCCTCAATTCAGCTCTCG   \n",
       "10181         13                TTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAA   \n",
       "10182         15              TTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAACA   \n",
       "10183         17            TTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAACATC   \n",
       "10184         10         CTGGGCATCCTTGAGTTCCAAGGCCTCAATTCAGCTCTCG   \n",
       "10185         13      CTGGGCATCCTTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAA   \n",
       "10186         15    CTGGGCATCCTTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAACA   \n",
       "10187         17  CTGGGCATCCTTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAACATC   \n",
       "11548         10                             AGACCTAGCCCTGTGCAGCT   \n",
       "11549         13                          AGACCTAGCCCTGTGCAGCTGTG   \n",
       "11550         15                        AGACCTAGCCCTGTGCAGCTGTGGG   \n",
       "11551         17                      AGACCTAGCCCTGTGCAGCTGTGGGTT   \n",
       "11552         10                   CAACTGGCCAAGACCTAGCCCTGTGCAGCT   \n",
       "11553         13                CAACTGGCCAAGACCTAGCCCTGTGCAGCTGTG   \n",
       "11554         15              CAACTGGCCAAGACCTAGCCCTGTGCAGCTGTGGG   \n",
       "11555         17            CAACTGGCCAAGACCTAGCCCTGTGCAGCTGTGGGTT   \n",
       "11556         10         GATGTTTTGCCAACTGGCCAAGACCTAGCCCTGTGCAGCT   \n",
       "11557         13      GATGTTTTGCCAACTGGCCAAGACCTAGCCCTGTGCAGCTGTG   \n",
       "11558         15    GATGTTTTGCCAACTGGCCAAGACCTAGCCCTGTGCAGCTGTGGG   \n",
       "11559         17  GATGTTTTGCCAACTGGCCAAGACCTAGCCCTGTGCAGCTGTGGGTT   \n",
       "11608         10                             CCTCCTTGGCCCCTGTCATC   \n",
       "11609         13                          CCTCCTTGGCCCCTGTCATCTTC   \n",
       "11610         15                        CCTCCTTGGCCCCTGTCATCTTCTG   \n",
       "11611         17                      CCTCCTTGGCCCCTGTCATCTTCTGTC   \n",
       "11612         10                   GCACCAGCCCCCTCCTTGGCCCCTGTCATC   \n",
       "11613         13                GCACCAGCCCCCTCCTTGGCCCCTGTCATCTTC   \n",
       "11614         15              GCACCAGCCCCCTCCTTGGCCCCTGTCATCTTCTG   \n",
       "11615         17            GCACCAGCCCCCTCCTTGGCCCCTGTCATCTTCTGTC   \n",
       "11616         10         GGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATC   \n",
       "11617         13      GGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTTC   \n",
       "11618         15    GGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTTCTG   \n",
       "11619         17  GGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTTCTGTC   \n",
       "11620         10                             CCCCCTCCTTGGCCCCTGTC   \n",
       "11621         13                          CCCCCTCCTTGGCCCCTGTCATC   \n",
       "11622         15                        CCCCCTCCTTGGCCCCTGTCATCTT   \n",
       "11623         17                      CCCCCTCCTTGGCCCCTGTCATCTTCT   \n",
       "11624         10                   CCTGCACCAGCCCCCTCCTTGGCCCCTGTC   \n",
       "11625         13                CCTGCACCAGCCCCCTCCTTGGCCCCTGTCATC   \n",
       "11626         15              CCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTT   \n",
       "11627         17            CCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTTCT   \n",
       "11628         10         ACCGGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTC   \n",
       "11629         13      ACCGGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATC   \n",
       "11630         15    ACCGGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTT   \n",
       "11631         17  ACCGGCGGCCCCTGCACCAGCCCCCTCCTTGGCCCCTGTCATCTTCT   \n",
       "11636         10                             GGGCCAAGGAGGGGGCTGGT   \n",
       "11637         13                          GGGCCAAGGAGGGGGCTGGTGCA   \n",
       "\n",
       "      Distance_to_nick RHA_size PAM_disrupted Proto_disrupted  \n",
       "10176                2        7          True            True  \n",
       "10177                2        7          True            True  \n",
       "10178                2        7          True            True  \n",
       "10179                2        7          True            True  \n",
       "10180                2       17          True            True  \n",
       "10181                2       17          True            True  \n",
       "10182                2       17          True            True  \n",
       "10183                2       17          True            True  \n",
       "10184                2       27          True            True  \n",
       "10185                2       27          True            True  \n",
       "10186                2       27          True            True  \n",
       "10187                2       27          True            True  \n",
       "11548                3        6          True           False  \n",
       "11549                3        6          True           False  \n",
       "11550                3        6          True           False  \n",
       "11551                3        6          True           False  \n",
       "11552                3       16          True           False  \n",
       "11553                3       16          True           False  \n",
       "11554                3       16          True           False  \n",
       "11555                3       16          True           False  \n",
       "11556                3       26          True           False  \n",
       "11557                3       26          True           False  \n",
       "11558                3       26          True           False  \n",
       "11559                3       26          True           False  \n",
       "11608                3        6          True           False  \n",
       "11609                3        6          True           False  \n",
       "11610                3        6          True           False  \n",
       "11611                3        6          True           False  \n",
       "11612                3       16          True           False  \n",
       "11613                3       16          True           False  \n",
       "11614                3       16          True           False  \n",
       "11615                3       16          True           False  \n",
       "11616                3       26          True           False  \n",
       "11617                3       26          True           False  \n",
       "11618                3       26          True           False  \n",
       "11619                3       26          True           False  \n",
       "11620                0        9          True            True  \n",
       "11621                0        9          True            True  \n",
       "11622                0        9          True            True  \n",
       "11623                0        9          True            True  \n",
       "11624                0       19          True            True  \n",
       "11625                0       19          True            True  \n",
       "11626                0       19          True            True  \n",
       "11627                0       19          True            True  \n",
       "11628                0       29          True            True  \n",
       "11629                0       29          True            True  \n",
       "11630                0       29          True            True  \n",
       "11631                0       29          True            True  \n",
       "11636                4        5          True           False  \n",
       "11637                4        5          True           False  \n",
       "\n",
       "[50 rows x 34 columns]"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df['PAM_disrupted']==True) & (df['Variant_Type']=='INS')][0:50]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Combined run function\n",
    "\n",
    "- Still need to add back in the ACTUAL PAM sequence locations...\n",
    "- Calculate it by converting from wt_w_context index to chromosomal location\n",
    "- Implement ClinVar IDs..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def input_formatter(input_df, input_format, chrom_dict, context_size):\n",
    "\n",
    "    #format the input dataframe appropriately...\n",
    "    if input_format == 'cBioPortal':\n",
    "        #add assert statement to make sure all necessary info is included\n",
    "        #MAKE COLS_TO_SAVE MORE FLEXIBLE??? to avoid weird errors...\n",
    "\n",
    "        input_df = df_formatter(input_df, chrom_dict, context_size)\n",
    "        input_df['mutation_idx'] = list(range(len(input_df)))\n",
    "\n",
    "    elif input_format == 'WT_ALT':\n",
    "\n",
    "        wt = []\n",
    "        mut = []\n",
    "        vts = []\n",
    "        l = []\n",
    "        ref = []\n",
    "        alt =[]\n",
    "        r = []\n",
    "        for i, val in input_df.iterrows():\n",
    "\n",
    "            var_type, left_seq, ref_seq, alt_seq, right_seq = mut_formatter(val['WT'], val['ALT'])\n",
    "            vts.append(var_type)\n",
    "            l.append(left_seq)\n",
    "            ref.append(ref_seq)\n",
    "            alt.append(alt_seq)\n",
    "            r.append(right_seq)\n",
    "            wt.append(val['WT'])\n",
    "            mut.append(val['ALT'])\n",
    "\n",
    "        mut_idx = list(range(len(wt)))\n",
    "        cols = [mut_idx, wt, mut, vts, ref, alt, l, r]\n",
    "        col_labels = [\"mutation_idx\", 'wt_w_context', 'alt_w_context', 'Variant_Type', 'REF', 'ALT', 'left_context', 'right_context']\n",
    "        input_df = pd.DataFrame(dict(zip(col_labels, cols)))        \n",
    "\n",
    "    elif input_format == 'PrimeDesign':\n",
    "\n",
    "        wt = []\n",
    "        mut = []\n",
    "        vts = []\n",
    "        l = []\n",
    "        ref = []\n",
    "        alt =[]\n",
    "        r = []\n",
    "        for i, val in input_df.iterrows():\n",
    "            var_type, wt_w_context, alt_w_context, left_seq, right_seq, ref_seq, alt_seq = primedesign_formatter(val['SEQ'])\n",
    "            vts.append(var_type)\n",
    "            l.append(left_seq)\n",
    "            ref.append(ref_seq)\n",
    "            alt.append(alt_seq)\n",
    "            r.append(right_seq)\n",
    "            wt.append(wt_w_context)\n",
    "            mut.append(alt_w_context)\n",
    "\n",
    "        mut_idx = list(range(len(wt)))\n",
    "        cols = [mut_idx, wt, mut, vts, ref, alt, l, r]\n",
    "        col_labels = [\"mutation_idx\", 'wt_w_context', 'alt_w_context', 'Variant_Type', 'REF', 'ALT', 'left_context', 'right_context']\n",
    "        input_df = pd.DataFrame(dict(zip(col_labels, cols)))     \n",
    "\n",
    "    elif input_format == 'ClinVar':\n",
    "        return \"ClinVar format input in development\"\n",
    "        #ADD FUNCTIONS...\n",
    "    \n",
    "    return input_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--params\n",
    "RTT_lengths = [10,20,30]\n",
    "PBS_lengths = [10,13, 15,17]\n",
    "proto_size = 19\n",
    "PAM = \"NGG\"\n",
    "\n",
    "#input_format options\n",
    "#cBioPortal, WT_ALT, PrimeDesign, ClinVarIDs\n",
    "\n",
    "def run(input_df, input_format, chrom_dict, PAM = \"NGG\", RTT_lengths = [5,10,15,25,30], PBS_lengths = [8,10,13,15], proto_size=19, context_size = 120, sensor=True):\n",
    "\n",
    "    for i in PBS_lengths:\n",
    "        assert i<=17, \"Max PBS Length = 17; rerun with smaller PBS length (or use preset parameters)\"\n",
    "\n",
    "    for i in RTT_lengths:\n",
    "        if i>30:\n",
    "            print(\"Warning: RTT lengths larger than 30 nt are not reccomended due to potential low PE efficiency\")\n",
    "            \n",
    "    #format the input df\n",
    "    #NEED TO UPDATE TO INCLUDE CLINVAR!!!\n",
    "    input_df = input_formatter(input_df, input_format, chrom_dict, context_size)\n",
    "\n",
    "    combined_peg_dfs = []\n",
    "\n",
    "    for i, val in input_df.iterrows():\n",
    "\n",
    "\n",
    "        mut = mutation(val['wt_w_context'], val['alt_w_context'], val['left_context'], val['right_context'], val['Variant_Type'], val['REF'], val['ALT'], chrom=None, genome=None)\n",
    "        mut.PAM_idx_forward, mut.PAM_idx_rc = eligible_PAM_finder(mut, PAM, max(RTT_lengths), proto_size)\n",
    "\n",
    "        peg_df_plus = pegRNA_generator(mut, PAM, '+', proto_size, RTT_lengths, PBS_lengths)\n",
    "        peg_df_minus = pegRNA_generator(mut, PAM, '-', proto_size, RTT_lengths, PBS_lengths)\n",
    "\n",
    "        #need to translate PAM indexing for the - strand!!!!\n",
    "        #convert to genome coordinates as well...\n",
    "        peg_df = pd.concat((peg_df_plus, peg_df_minus))\n",
    "        peg_df['mutation_idx'] = i\n",
    "\n",
    "        combined_peg_dfs.append(peg_df)\n",
    "\n",
    "    peg_df = pd.concat(combined_peg_dfs).set_index('mutation_idx').reset_index()\n",
    "\n",
    "    #COMBINE WITH INPUT_DF !!!!\n",
    "\n",
    "    #ADD IN SENSOR IF DESIRED!!!!\n",
    "    peg_df = pd.merge(input_df, peg_df, on='mutation_idx')\n",
    "\n",
    "    return peg_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>PAM_start</th>\n",
       "      <th>PAM_end</th>\n",
       "      <th>PAM</th>\n",
       "      <th>PAM_strand</th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>RTT</th>\n",
       "      <th>RTT_length</th>\n",
       "      <th>PBS</th>\n",
       "      <th>PBS_length</th>\n",
       "      <th>RTT_PBS</th>\n",
       "      <th>Distance_to_nick</th>\n",
       "      <th>RHA_size</th>\n",
       "      <th>PAM_disrupted</th>\n",
       "      <th>Proto_disrupted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "      <td>107</td>\n",
       "      <td>AGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>GGTGCATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>25</td>\n",
       "      <td>AGAGACCG</td>\n",
       "      <td>8</td>\n",
       "      <td>GGTGCATGTTTGTGCCTGTCCTGGGAGAGACCG</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "      <td>107</td>\n",
       "      <td>AGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>GGTGCATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>25</td>\n",
       "      <td>AGAGACCGGC</td>\n",
       "      <td>10</td>\n",
       "      <td>GGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGC</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "      <td>107</td>\n",
       "      <td>AGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>GGTGCATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>25</td>\n",
       "      <td>AGAGACCGGCGCA</td>\n",
       "      <td>13</td>\n",
       "      <td>GGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCA</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "      <td>107</td>\n",
       "      <td>AGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>GGTGCATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>25</td>\n",
       "      <td>AGAGACCGGCGCACA</td>\n",
       "      <td>15</td>\n",
       "      <td>GGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACA</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "      <td>107</td>\n",
       "      <td>AGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>TTTGAGGTGCATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>AGAGACCG</td>\n",
       "      <td>8</td>\n",
       "      <td>TTTGAGGTGCATGTTTGTGCCTGTCCTGGGAGAGACCG</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58127</th>\n",
       "      <td>999</td>\n",
       "      <td>119</td>\n",
       "      <td>116</td>\n",
       "      <td>CGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACCAGCAGCTCCTACAC</td>\n",
       "      <td>GGGGGCTGGTGCAGGGGCCGCCGGG</td>\n",
       "      <td>25</td>\n",
       "      <td>TAGGAGCTGCTGGTG</td>\n",
       "      <td>15</td>\n",
       "      <td>GGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGGTG</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58128</th>\n",
       "      <td>999</td>\n",
       "      <td>119</td>\n",
       "      <td>116</td>\n",
       "      <td>CGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACCAGCAGCTCCTACAC</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>TAGGAGCT</td>\n",
       "      <td>8</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCT</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58129</th>\n",
       "      <td>999</td>\n",
       "      <td>119</td>\n",
       "      <td>116</td>\n",
       "      <td>CGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACCAGCAGCTCCTACAC</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>TAGGAGCTGC</td>\n",
       "      <td>10</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGC</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58130</th>\n",
       "      <td>999</td>\n",
       "      <td>119</td>\n",
       "      <td>116</td>\n",
       "      <td>CGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACCAGCAGCTCCTACAC</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>TAGGAGCTGCTGG</td>\n",
       "      <td>13</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGG</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58131</th>\n",
       "      <td>999</td>\n",
       "      <td>119</td>\n",
       "      <td>116</td>\n",
       "      <td>CGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACCAGCAGCTCCTACAC</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>TAGGAGCTGCTGGTG</td>\n",
       "      <td>15</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGGTG</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58132 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation_idx  PAM_start  PAM_end  PAM PAM_strand           Protospacer  \\\n",
       "0                 0        104      107  AGG          +  GCTGTGCGCCGGTCTCTCCC   \n",
       "1                 0        104      107  AGG          +  GCTGTGCGCCGGTCTCTCCC   \n",
       "2                 0        104      107  AGG          +  GCTGTGCGCCGGTCTCTCCC   \n",
       "3                 0        104      107  AGG          +  GCTGTGCGCCGGTCTCTCCC   \n",
       "4                 0        104      107  AGG          +  GCTGTGCGCCGGTCTCTCCC   \n",
       "...             ...        ...      ...  ...        ...                   ...   \n",
       "58127           999        119      116  CGG          -  GGCACCAGCAGCTCCTACAC   \n",
       "58128           999        119      116  CGG          -  GGCACCAGCAGCTCCTACAC   \n",
       "58129           999        119      116  CGG          -  GGCACCAGCAGCTCCTACAC   \n",
       "58130           999        119      116  CGG          -  GGCACCAGCAGCTCCTACAC   \n",
       "58131           999        119      116  CGG          -  GGCACCAGCAGCTCCTACAC   \n",
       "\n",
       "                                  RTT  RTT_length              PBS  \\\n",
       "0           GGTGCATGTTTGTGCCTGTCCTGGG          25         AGAGACCG   \n",
       "1           GGTGCATGTTTGTGCCTGTCCTGGG          25       AGAGACCGGC   \n",
       "2           GGTGCATGTTTGTGCCTGTCCTGGG          25    AGAGACCGGCGCA   \n",
       "3           GGTGCATGTTTGTGCCTGTCCTGGG          25  AGAGACCGGCGCACA   \n",
       "4      TTTGAGGTGCATGTTTGTGCCTGTCCTGGG          30         AGAGACCG   \n",
       "...                               ...         ...              ...   \n",
       "58127       GGGGGCTGGTGCAGGGGCCGCCGGG          25  TAGGAGCTGCTGGTG   \n",
       "58128  CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG          30         TAGGAGCT   \n",
       "58129  CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG          30       TAGGAGCTGC   \n",
       "58130  CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG          30    TAGGAGCTGCTGG   \n",
       "58131  CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG          30  TAGGAGCTGCTGGTG   \n",
       "\n",
       "       PBS_length                                        RTT_PBS  \\\n",
       "0               8              GGTGCATGTTTGTGCCTGTCCTGGGAGAGACCG   \n",
       "1              10            GGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGC   \n",
       "2              13         GGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCA   \n",
       "3              15       GGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACA   \n",
       "4               8         TTTGAGGTGCATGTTTGTGCCTGTCCTGGGAGAGACCG   \n",
       "...           ...                                            ...   \n",
       "58127          15       GGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGGTG   \n",
       "58128           8         CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCT   \n",
       "58129          10       CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGC   \n",
       "58130          13    CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGG   \n",
       "58131          15  CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGGTG   \n",
       "\n",
       "       Distance_to_nick  RHA_size  PAM_disrupted  Proto_disrupted  \n",
       "0                    19         5          False            False  \n",
       "1                    19         5          False            False  \n",
       "2                    19         5          False            False  \n",
       "3                    19         5          False            False  \n",
       "4                    19        10          False            False  \n",
       "...                 ...       ...            ...              ...  \n",
       "58127                 1        24           True             True  \n",
       "58128                 1        29           True             True  \n",
       "58129                 1        29           True             True  \n",
       "58130                 1        29           True             True  \n",
       "58131                 1        29           True             True  \n",
       "\n",
       "[58132 rows x 15 columns]"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input = p53_filtered[0:1000]\n",
    "input_format = 'cBioPortal'\n",
    "peg_df = run(input, input_format, chrom_dict)\n",
    "peg_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs = ['AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAG(/ACAT)GGCGTCGCTAGCGATCGACTAGGCTAGCTACGACTGACTATCG',\n",
    "        'TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACCCTCG(A/CC)GGCCGGTGCTCGGGAGGAGAGCTAGCTACGAGCGACTACGATCGACTAGC']\n",
    "\n",
    "prime_design = pd.DataFrame(dict(zip(['SEQ'], [seqs])))\n",
    "input_format = 'PrimeDesign'\n",
    "peg_df = run(prime_design, input_format, chrom_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>left_context</th>\n",
       "      <th>right_context</th>\n",
       "      <th>PAM_start</th>\n",
       "      <th>PAM_end</th>\n",
       "      <th>...</th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>RTT</th>\n",
       "      <th>RTT_length</th>\n",
       "      <th>PBS</th>\n",
       "      <th>PBS_length</th>\n",
       "      <th>RTT_PBS</th>\n",
       "      <th>Distance_to_nick</th>\n",
       "      <th>RHA_size</th>\n",
       "      <th>PAM_disrupted</th>\n",
       "      <th>Proto_disrupted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGGG...</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGAC...</td>\n",
       "      <td>INS</td>\n",
       "      <td></td>\n",
       "      <td>ACAT</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAG</td>\n",
       "      <td>GGCGTCGCTAGCGATCGACTAGGCTAGCTACGACTGACTATCG</td>\n",
       "      <td>42</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>GTCGAGCTACGATCGACTCG</td>\n",
       "      <td>CATGTCTCGA</td>\n",
       "      <td>10</td>\n",
       "      <td>GTCGATCG</td>\n",
       "      <td>8</td>\n",
       "      <td>CATGTCTCGAGTCGATCG</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGGG...</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGAC...</td>\n",
       "      <td>INS</td>\n",
       "      <td></td>\n",
       "      <td>ACAT</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAG</td>\n",
       "      <td>GGCGTCGCTAGCGATCGACTAGGCTAGCTACGACTGACTATCG</td>\n",
       "      <td>42</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>GTCGAGCTACGATCGACTCG</td>\n",
       "      <td>CATGTCTCGA</td>\n",
       "      <td>10</td>\n",
       "      <td>GTCGATCGTA</td>\n",
       "      <td>10</td>\n",
       "      <td>CATGTCTCGAGTCGATCGTA</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGGG...</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGAC...</td>\n",
       "      <td>INS</td>\n",
       "      <td></td>\n",
       "      <td>ACAT</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAG</td>\n",
       "      <td>GGCGTCGCTAGCGATCGACTAGGCTAGCTACGACTGACTATCG</td>\n",
       "      <td>42</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>GTCGAGCTACGATCGACTCG</td>\n",
       "      <td>CATGTCTCGA</td>\n",
       "      <td>10</td>\n",
       "      <td>GTCGATCGTAGCT</td>\n",
       "      <td>13</td>\n",
       "      <td>CATGTCTCGAGTCGATCGTAGCT</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGGG...</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGAC...</td>\n",
       "      <td>INS</td>\n",
       "      <td></td>\n",
       "      <td>ACAT</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAG</td>\n",
       "      <td>GGCGTCGCTAGCGATCGACTAGGCTAGCTACGACTGACTATCG</td>\n",
       "      <td>42</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>GTCGAGCTACGATCGACTCG</td>\n",
       "      <td>CATGTCTCGA</td>\n",
       "      <td>10</td>\n",
       "      <td>GTCGATCGTAGCTCG</td>\n",
       "      <td>15</td>\n",
       "      <td>CATGTCTCGAGTCGATCGTAGCTCG</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGGG...</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGAC...</td>\n",
       "      <td>INS</td>\n",
       "      <td></td>\n",
       "      <td>ACAT</td>\n",
       "      <td>AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAG</td>\n",
       "      <td>GGCGTCGCTAGCGATCGACTAGGCTAGCTACGACTGACTATCG</td>\n",
       "      <td>42</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>GTCGAGCTACGATCGACTCG</td>\n",
       "      <td>GACGCCATGTCTCGA</td>\n",
       "      <td>15</td>\n",
       "      <td>GTCGATCG</td>\n",
       "      <td>8</td>\n",
       "      <td>GACGCCATGTCTCGAGTCGATCG</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>1</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>A</td>\n",
       "      <td>CC</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>GGCCGGTGCTCGGGAGGAGAGCTAGCTACGAGCGACTACGATCGAC...</td>\n",
       "      <td>48</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTCCCGAGCACCGGCCTCG</td>\n",
       "      <td>CGGCGCGCGCGTAGCGACCCTCGCC</td>\n",
       "      <td>25</td>\n",
       "      <td>GGCCGGTGCTCGGGA</td>\n",
       "      <td>15</td>\n",
       "      <td>CGGCGCGCGCGTAGCGACCCTCGCCGGCCGGTGCTCGGGA</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>1</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>A</td>\n",
       "      <td>CC</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>GGCCGGTGCTCGGGAGGAGAGCTAGCTACGAGCGACTACGATCGAC...</td>\n",
       "      <td>48</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTCCCGAGCACCGGCCTCG</td>\n",
       "      <td>CGGCGCGGCGCGCGCGTAGCGACCCTCGCC</td>\n",
       "      <td>30</td>\n",
       "      <td>GGCCGGTG</td>\n",
       "      <td>8</td>\n",
       "      <td>CGGCGCGGCGCGCGCGTAGCGACCCTCGCCGGCCGGTG</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>1</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>A</td>\n",
       "      <td>CC</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>GGCCGGTGCTCGGGAGGAGAGCTAGCTACGAGCGACTACGATCGAC...</td>\n",
       "      <td>48</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTCCCGAGCACCGGCCTCG</td>\n",
       "      <td>CGGCGCGGCGCGCGCGTAGCGACCCTCGCC</td>\n",
       "      <td>30</td>\n",
       "      <td>GGCCGGTGCT</td>\n",
       "      <td>10</td>\n",
       "      <td>CGGCGCGGCGCGCGCGTAGCGACCCTCGCCGGCCGGTGCT</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>1</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>A</td>\n",
       "      <td>CC</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>GGCCGGTGCTCGGGAGGAGAGCTAGCTACGAGCGACTACGATCGAC...</td>\n",
       "      <td>48</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTCCCGAGCACCGGCCTCG</td>\n",
       "      <td>CGGCGCGGCGCGCGCGTAGCGACCCTCGCC</td>\n",
       "      <td>30</td>\n",
       "      <td>GGCCGGTGCTCGG</td>\n",
       "      <td>13</td>\n",
       "      <td>CGGCGCGGCGCGCGCGTAGCGACCCTCGCCGGCCGGTGCTCGG</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>1</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>A</td>\n",
       "      <td>CC</td>\n",
       "      <td>TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...</td>\n",
       "      <td>GGCCGGTGCTCGGGAGGAGAGCTAGCTACGAGCGACTACGATCGAC...</td>\n",
       "      <td>48</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTCCCGAGCACCGGCCTCG</td>\n",
       "      <td>CGGCGCGGCGCGCGCGTAGCGACCCTCGCC</td>\n",
       "      <td>30</td>\n",
       "      <td>GGCCGGTGCTCGGGA</td>\n",
       "      <td>15</td>\n",
       "      <td>CGGCGCGGCGCGCGCGTAGCGACCCTCGCCGGCCGGTGCTCGGGA</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    mutation_idx                                       wt_w_context  \\\n",
       "0              0  AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGGG...   \n",
       "1              0  AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGGG...   \n",
       "2              0  AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGGG...   \n",
       "3              0  AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGGG...   \n",
       "4              0  AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGGG...   \n",
       "..           ...                                                ...   \n",
       "87             1  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...   \n",
       "88             1  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...   \n",
       "89             1  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...   \n",
       "90             1  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...   \n",
       "91             1  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...   \n",
       "\n",
       "                                        alt_w_context Variant_Type REF   ALT  \\\n",
       "0   AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGAC...          INS      ACAT   \n",
       "1   AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGAC...          INS      ACAT   \n",
       "2   AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGAC...          INS      ACAT   \n",
       "3   AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGAC...          INS      ACAT   \n",
       "4   AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGAC...          INS      ACAT   \n",
       "..                                                ...          ...  ..   ...   \n",
       "87  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...        INDEL   A    CC   \n",
       "88  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...        INDEL   A    CC   \n",
       "89  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...        INDEL   A    CC   \n",
       "90  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...        INDEL   A    CC   \n",
       "91  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...        INDEL   A    CC   \n",
       "\n",
       "                                         left_context  \\\n",
       "0        AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAG   \n",
       "1        AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAG   \n",
       "2        AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAG   \n",
       "3        AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAG   \n",
       "4        AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAG   \n",
       "..                                                ...   \n",
       "87  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...   \n",
       "88  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...   \n",
       "89  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...   \n",
       "90  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...   \n",
       "91  TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACC...   \n",
       "\n",
       "                                        right_context  PAM_start  PAM_end  \\\n",
       "0         GGCGTCGCTAGCGATCGACTAGGCTAGCTACGACTGACTATCG         42       45   \n",
       "1         GGCGTCGCTAGCGATCGACTAGGCTAGCTACGACTGACTATCG         42       45   \n",
       "2         GGCGTCGCTAGCGATCGACTAGGCTAGCTACGACTGACTATCG         42       45   \n",
       "3         GGCGTCGCTAGCGATCGACTAGGCTAGCTACGACTGACTATCG         42       45   \n",
       "4         GGCGTCGCTAGCGATCGACTAGGCTAGCTACGACTGACTATCG         42       45   \n",
       "..                                                ...        ...      ...   \n",
       "87  GGCCGGTGCTCGGGAGGAGAGCTAGCTACGAGCGACTACGATCGAC...         48       45   \n",
       "88  GGCCGGTGCTCGGGAGGAGAGCTAGCTACGAGCGACTACGATCGAC...         48       45   \n",
       "89  GGCCGGTGCTCGGGAGGAGAGCTAGCTACGAGCGACTACGATCGAC...         48       45   \n",
       "90  GGCCGGTGCTCGGGAGGAGAGCTAGCTACGAGCGACTACGATCGAC...         48       45   \n",
       "91  GGCCGGTGCTCGGGAGGAGAGCTAGCTACGAGCGACTACGATCGAC...         48       45   \n",
       "\n",
       "    ...           Protospacer                             RTT RTT_length  \\\n",
       "0   ...  GTCGAGCTACGATCGACTCG                      CATGTCTCGA         10   \n",
       "1   ...  GTCGAGCTACGATCGACTCG                      CATGTCTCGA         10   \n",
       "2   ...  GTCGAGCTACGATCGACTCG                      CATGTCTCGA         10   \n",
       "3   ...  GTCGAGCTACGATCGACTCG                      CATGTCTCGA         10   \n",
       "4   ...  GTCGAGCTACGATCGACTCG                 GACGCCATGTCTCGA         15   \n",
       "..  ...                   ...                             ...        ...   \n",
       "87  ...  GCTCCCGAGCACCGGCCTCG       CGGCGCGCGCGTAGCGACCCTCGCC         25   \n",
       "88  ...  GCTCCCGAGCACCGGCCTCG  CGGCGCGGCGCGCGCGTAGCGACCCTCGCC         30   \n",
       "89  ...  GCTCCCGAGCACCGGCCTCG  CGGCGCGGCGCGCGCGTAGCGACCCTCGCC         30   \n",
       "90  ...  GCTCCCGAGCACCGGCCTCG  CGGCGCGGCGCGCGCGTAGCGACCCTCGCC         30   \n",
       "91  ...  GCTCCCGAGCACCGGCCTCG  CGGCGCGGCGCGCGCGTAGCGACCCTCGCC         30   \n",
       "\n",
       "                PBS  PBS_length  \\\n",
       "0          GTCGATCG           8   \n",
       "1        GTCGATCGTA          10   \n",
       "2     GTCGATCGTAGCT          13   \n",
       "3   GTCGATCGTAGCTCG          15   \n",
       "4          GTCGATCG           8   \n",
       "..              ...         ...   \n",
       "87  GGCCGGTGCTCGGGA          15   \n",
       "88         GGCCGGTG           8   \n",
       "89       GGCCGGTGCT          10   \n",
       "90    GGCCGGTGCTCGG          13   \n",
       "91  GGCCGGTGCTCGGGA          15   \n",
       "\n",
       "                                          RTT_PBS  Distance_to_nick RHA_size  \\\n",
       "0                              CATGTCTCGAGTCGATCG                 5        1   \n",
       "1                            CATGTCTCGAGTCGATCGTA                 5        1   \n",
       "2                         CATGTCTCGAGTCGATCGTAGCT                 5        1   \n",
       "3                       CATGTCTCGAGTCGATCGTAGCTCG                 5        1   \n",
       "4                         GACGCCATGTCTCGAGTCGATCG                 5        6   \n",
       "..                                            ...               ...      ...   \n",
       "87       CGGCGCGCGCGTAGCGACCCTCGCCGGCCGGTGCTCGGGA                 0       23   \n",
       "88         CGGCGCGGCGCGCGCGTAGCGACCCTCGCCGGCCGGTG                 0       28   \n",
       "89       CGGCGCGGCGCGCGCGTAGCGACCCTCGCCGGCCGGTGCT                 0       28   \n",
       "90    CGGCGCGGCGCGCGCGTAGCGACCCTCGCCGGCCGGTGCTCGG                 0       28   \n",
       "91  CGGCGCGGCGCGCGCGTAGCGACCCTCGCCGGCCGGTGCTCGGGA                 0       28   \n",
       "\n",
       "    PAM_disrupted  Proto_disrupted  \n",
       "0            True            False  \n",
       "1            True            False  \n",
       "2            True            False  \n",
       "3            True            False  \n",
       "4            True            False  \n",
       "..            ...              ...  \n",
       "87           True             True  \n",
       "88           True             True  \n",
       "89           True             True  \n",
       "90           True             True  \n",
       "91           True             True  \n",
       "\n",
       "[92 rows x 22 columns]"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peg_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [],
   "source": [
    "wt_seqs = ['AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGACTCGAGGGCGTCGCTAGCGATCGACTAGGCTAGCTACGACTGACTATCG',\n",
    "        'TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGACCCTCGGGCCGGTGCTCGGGAGGAGAGCTAGCTACGAGCGACTACGATCGACTAGC']\n",
    "alt_seqs = ['AATCGCTCGTCGATCGACTAGCATCGAGCTACGATCGAGGCCTCGAGGGCGTCGCTAGCGATCGACTAGGCTAGCTACGACTGACTATCG',\n",
    "        'TGCTCGATCGGGCGCGCTAGCGCGGCGCGGCGCGCGCGTAGCGAGCCACGGGCCGGTGCTCGGGAGGAGAGCTAGCTACGAGCGACTACGATCGACTAGC']\n",
    "\n",
    "wt_alt = pd.DataFrame(dict(zip(['WT', 'ALT'], [wt_seqs, alt_seqs])))\n",
    "input_format = 'WT_ALT'\n",
    "peg_df = run(wt_alt, input_format, chrom_dict, PAM='NNN')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PEGG Score\n",
    "\n",
    "- Update PEGG score to reflect the results of my dataset (include PAM/protospacer disrupting edits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# pegRNA visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>...</th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>RTT</th>\n",
       "      <th>RTT_length</th>\n",
       "      <th>PBS</th>\n",
       "      <th>PBS_length</th>\n",
       "      <th>RTT_PBS</th>\n",
       "      <th>Distance_to_nick</th>\n",
       "      <th>RHA_size</th>\n",
       "      <th>PAM_disrupted</th>\n",
       "      <th>Proto_disrupted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGC</td>\n",
       "      <td>10</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGC</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGCGCA</td>\n",
       "      <td>13</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCA</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGCGCACA</td>\n",
       "      <td>15</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACA</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGCGCACAGA</td>\n",
       "      <td>17</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGA</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>TTTGAGGTGCATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>AGAGACCGGC</td>\n",
       "      <td>10</td>\n",
       "      <td>TTTGAGGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGC</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122567</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>CAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>20</td>\n",
       "      <td>CGGAAACCGTAGCTGCC</td>\n",
       "      <td>17</td>\n",
       "      <td>CAGAATGCAATAAGCCCAGACGGAAACCGTAGCTGCC</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122568</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>30</td>\n",
       "      <td>CGGAAACCGT</td>\n",
       "      <td>10</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGT</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122569</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>30</td>\n",
       "      <td>CGGAAACCGTAGC</td>\n",
       "      <td>13</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGC</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122570</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>30</td>\n",
       "      <td>CGGAAACCGTAGCTG</td>\n",
       "      <td>15</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGCTG</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122571</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>30</td>\n",
       "      <td>CGGAAACCGTAGCTGCC</td>\n",
       "      <td>17</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGCTGCC</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122572 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mutation_idx Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0                  0        TP53         17         7577120       7577120   \n",
       "1                  0        TP53         17         7577120       7577120   \n",
       "2                  0        TP53         17         7577120       7577120   \n",
       "3                  0        TP53         17         7577120       7577120   \n",
       "4                  0        TP53         17         7577120       7577120   \n",
       "...              ...         ...        ...             ...           ...   \n",
       "122567          3004        TP53         17         7579348       7579348   \n",
       "122568          3004        TP53         17         7579348       7579348   \n",
       "122569          3004        TP53         17         7579348       7579348   \n",
       "122570          3004        TP53         17         7579348       7579348   \n",
       "122571          3004        TP53         17         7579348       7579348   \n",
       "\n",
       "             Consequence Variant_Classification Variant_Type Reference_Allele  \\\n",
       "0       missense_variant      Missense_Mutation          SNP                C   \n",
       "1       missense_variant      Missense_Mutation          SNP                C   \n",
       "2       missense_variant      Missense_Mutation          SNP                C   \n",
       "3       missense_variant      Missense_Mutation          SNP                C   \n",
       "4       missense_variant      Missense_Mutation          SNP                C   \n",
       "...                  ...                    ...          ...              ...   \n",
       "122567  missense_variant      Missense_Mutation          SNP                G   \n",
       "122568  missense_variant      Missense_Mutation          SNP                G   \n",
       "122569  missense_variant      Missense_Mutation          SNP                G   \n",
       "122570  missense_variant      Missense_Mutation          SNP                G   \n",
       "122571  missense_variant      Missense_Mutation          SNP                G   \n",
       "\n",
       "       Tumor_Seq_Allele2  ...           Protospacer  \\\n",
       "0                      T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "1                      T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "2                      T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "3                      T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "4                      T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "...                  ...  ...                   ...   \n",
       "122567                 T  ...  GGCAGCTACGGTTTCCGTCT   \n",
       "122568                 T  ...  GGCAGCTACGGTTTCCGTCT   \n",
       "122569                 T  ...  GGCAGCTACGGTTTCCGTCT   \n",
       "122570                 T  ...  GGCAGCTACGGTTTCCGTCT   \n",
       "122571                 T  ...  GGCAGCTACGGTTTCCGTCT   \n",
       "\n",
       "                                   RTT  RTT_length                PBS  \\\n",
       "0                 ATGTTTGTGCCTGTCCTGGG          20         AGAGACCGGC   \n",
       "1                 ATGTTTGTGCCTGTCCTGGG          20      AGAGACCGGCGCA   \n",
       "2                 ATGTTTGTGCCTGTCCTGGG          20    AGAGACCGGCGCACA   \n",
       "3                 ATGTTTGTGCCTGTCCTGGG          20  AGAGACCGGCGCACAGA   \n",
       "4       TTTGAGGTGCATGTTTGTGCCTGTCCTGGG          30         AGAGACCGGC   \n",
       "...                                ...         ...                ...   \n",
       "122567            CAGAATGCAATAAGCCCAGA          20  CGGAAACCGTAGCTGCC   \n",
       "122568  TTGGCTGTCCCAGAATGCAATAAGCCCAGA          30         CGGAAACCGT   \n",
       "122569  TTGGCTGTCCCAGAATGCAATAAGCCCAGA          30      CGGAAACCGTAGC   \n",
       "122570  TTGGCTGTCCCAGAATGCAATAAGCCCAGA          30    CGGAAACCGTAGCTG   \n",
       "122571  TTGGCTGTCCCAGAATGCAATAAGCCCAGA          30  CGGAAACCGTAGCTGCC   \n",
       "\n",
       "       PBS_length                                          RTT_PBS  \\\n",
       "0              10                   ATGTTTGTGCCTGTCCTGGGAGAGACCGGC   \n",
       "1              13                ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCA   \n",
       "2              15              ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACA   \n",
       "3              17            ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGA   \n",
       "4              10         TTTGAGGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGC   \n",
       "...           ...                                              ...   \n",
       "122567         17            CAGAATGCAATAAGCCCAGACGGAAACCGTAGCTGCC   \n",
       "122568         10         TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGT   \n",
       "122569         13      TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGC   \n",
       "122570         15    TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGCTG   \n",
       "122571         17  TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGCTGCC   \n",
       "\n",
       "       Distance_to_nick RHA_size PAM_disrupted Proto_disrupted  \n",
       "0                    19        0         False           False  \n",
       "1                    19        0         False           False  \n",
       "2                    19        0         False           False  \n",
       "3                    19        0         False           False  \n",
       "4                    19       10         False           False  \n",
       "...                 ...      ...           ...             ...  \n",
       "122567                9       10         False           False  \n",
       "122568                9       20         False           False  \n",
       "122569                9       20         False           False  \n",
       "122570                9       20         False           False  \n",
       "122571                9       20         False           False  \n",
       "\n",
       "[122572 rows x 34 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_df = input_df.rename(columns = {'index':'mutation_idx'})\n",
    "\n",
    "df = pd.merge(input_df, peg_df, on='mutation_idx')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "class bcolors:\n",
    "    Red = '\\033[91m'\n",
    "    Green = '\\033[92m'\n",
    "    Blue = '\\033[94m'\n",
    "    Cyan = '\\033[96m'\n",
    "    White = '\\033[97m'\n",
    "    Yellow = '\\033[93m'\n",
    "    Magenta = '\\033[95m'\n",
    "    Grey = '\\033[90m'\n",
    "    ENDC = '\\033[0m'\n",
    "\n",
    "from Bio import pairwise2\n",
    "\n",
    "def visualize_pegRNA(df, i):\n",
    "    \"\"\"\n",
    "    This one visualizes the binding of the protospacer and the 3' extension to the ref seq\n",
    "    It also says the rank and PRIDICT score (intended/unintended editing)\n",
    "\n",
    "    Inputs:\n",
    "    df = datframe (loaded from pegRNA_dataframe.csv)\n",
    "    i = index (row/pegRNA that you want to visualize the mutation from;\n",
    "            will change depending on the strand (Reverse or Forward) of the pegRNA)\n",
    "    \"\"\"\n",
    "\n",
    "    if df['PAM_strand'][i] == '+':\n",
    "        refseq = df['wt_w_context'][i]\n",
    "    else:\n",
    "        refseq = df['wt_w_context'][i]\n",
    "        refseq = str(Bio.Seq.Seq(refseq).reverse_complement())\n",
    "        \n",
    "\n",
    "\n",
    "    proto = df['Protospacer'][i]\n",
    "\n",
    "    #print a bunch of info about the pegRNA\n",
    "    #print( bcolors.ENDC + 'pegRNA Rank = ' + str(df['rank'][i]))\n",
    "    print('Target Strand = ' + df['PAM_strand'][i])\n",
    "    print(str(df['Variant_Type'][i]) +': ' + str(df['REF'][i]) + '/' + str(df['ALT'][i]))\n",
    "    print('PBS length = ' + str(df['PBS_length'][i]) + ' ; RT length = '+ str(df['RTT_length'][i]) + '; RT overhang = ' + str(df['RHA_size'][i]))\n",
    "\n",
    "\n",
    "    proto_loc = refseq.find(proto[1:])\n",
    "\n",
    "    #RT = Bio.Seq.Seq(df['RTrevcomp'][i]).reverse_complement()\n",
    "    ext = Bio.Seq.Seq(df['RTT_PBS'][i]).reverse_complement()\n",
    "\n",
    "    first = pairwise2.align.localms(refseq, ext, 2,0, -3, -0.2)[0]\n",
    "    a=format_alignment(*first, full_sequences=True)\n",
    "\n",
    "    start = len(first.seqB) + 1\n",
    "    end = 2*len(first.seqB) + 1\n",
    "    pipes = a[start:end]\n",
    "\n",
    "    #translating first.seqB into complement (the true 3' extension)\n",
    "    dict_comp = {'-':'-', 'A':'T', 'T':'A', 'C':'G', 'G':'C'}\n",
    "    comp1 = [dict_comp[i] for i in first.seqB]\n",
    "    complement = \"\".join(comp1)\n",
    "\n",
    "    print(complement + '\\n' + pipes + '\\n' + first.seqA[0:proto_loc] + bcolors.Blue + first.seqA[proto_loc:proto_loc+19] + bcolors.Magenta + first.seqA[proto_loc+19:proto_loc+22] + bcolors.ENDC + first.seqA[proto_loc+22:] )\n",
    "    print(Bio.Seq.Seq(first.seqA).complement())\n",
    "\n",
    "    second = pairwise2.align.localms(refseq, proto, 2,0, -3, -0.2)[0]\n",
    "    b=format_alignment(*second, full_sequences=True)\n",
    "\n",
    "    start2 = len(second.seqB) + 1\n",
    "    end2 = 2*len(second.seqB) + 1\n",
    "    pipes2 = b[start:end]\n",
    "\n",
    "    #print(' ' + pipes2)\n",
    "    print(pipes2)\n",
    "    print(second.seqB)\n",
    "\n",
    "    print(bcolors.Blue + 'Protospacer (G+19 bp)')\n",
    "    print(bcolors.Magenta + 'PAM')\n",
    "    print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/5g/xxjcy2g531n02hnyb6d8qdsr0000gn/T/ipykernel_79410/3693921967.py\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mvisualize_pegRNA\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m20228\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'df' is not defined"
     ]
    }
   ],
   "source": [
    "visualize_pegRNA(df, 20228)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 799,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 799,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('ACGTGGTCGTCGAGGAT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 818,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>...</th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>RTT</th>\n",
       "      <th>RTT_length</th>\n",
       "      <th>PBS</th>\n",
       "      <th>PBS_length</th>\n",
       "      <th>RTT_PBS</th>\n",
       "      <th>Distance_to_nick</th>\n",
       "      <th>RHA_size</th>\n",
       "      <th>PAM_disrupted</th>\n",
       "      <th>Proto_disrupted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27228</th>\n",
       "      <td>792</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577035</td>\n",
       "      <td>7577036</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCTCGCTTAGTGCTCCCTG</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAG</td>\n",
       "      <td>20</td>\n",
       "      <td>GGAGCACTAA</td>\n",
       "      <td>10</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAGGGAGCACTAA</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27229</th>\n",
       "      <td>792</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577035</td>\n",
       "      <td>7577036</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCTCGCTTAGTGCTCCCTG</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAG</td>\n",
       "      <td>20</td>\n",
       "      <td>GGAGCACTAAGCG</td>\n",
       "      <td>13</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAGGGAGCACTAAGCG</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27230</th>\n",
       "      <td>792</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577035</td>\n",
       "      <td>7577036</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCTCGCTTAGTGCTCCCTG</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAG</td>\n",
       "      <td>20</td>\n",
       "      <td>GGAGCACTAAGCGAG</td>\n",
       "      <td>15</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAGGGAGCACTAAGCGAG</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27231</th>\n",
       "      <td>792</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577035</td>\n",
       "      <td>7577036</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCTCGCTTAGTGCTCCCTG</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAG</td>\n",
       "      <td>20</td>\n",
       "      <td>GGAGCACTAAGCGAGGT</td>\n",
       "      <td>17</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAGGGAGCACTAAGCGAGGT</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27232</th>\n",
       "      <td>792</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577035</td>\n",
       "      <td>7577036</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCTCGCTTAGTGCTCCCTG</td>\n",
       "      <td>AGGGGAGCCTCACCACGAGCTGCCCCCCAG</td>\n",
       "      <td>30</td>\n",
       "      <td>GGAGCACTAA</td>\n",
       "      <td>10</td>\n",
       "      <td>AGGGGAGCCTCACCACGAGCTGCCCCCCAGGGAGCACTAA</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33963</th>\n",
       "      <td>982</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578479</td>\n",
       "      <td>7578480</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCTGCCCTGTGCAGCTGT</td>\n",
       "      <td>GGATGTGGAATCAACCCACA</td>\n",
       "      <td>20</td>\n",
       "      <td>GCTGCACAGGGCAGGTC</td>\n",
       "      <td>17</td>\n",
       "      <td>GGATGTGGAATCAACCCACAGCTGCACAGGGCAGGTC</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33964</th>\n",
       "      <td>982</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578479</td>\n",
       "      <td>7578480</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCTGCCCTGTGCAGCTGT</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACA</td>\n",
       "      <td>30</td>\n",
       "      <td>GCTGCACAGG</td>\n",
       "      <td>10</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGG</td>\n",
       "      <td>17</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33965</th>\n",
       "      <td>982</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578479</td>\n",
       "      <td>7578480</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCTGCCCTGTGCAGCTGT</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACA</td>\n",
       "      <td>30</td>\n",
       "      <td>GCTGCACAGGGCA</td>\n",
       "      <td>13</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGGGCA</td>\n",
       "      <td>17</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33966</th>\n",
       "      <td>982</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578479</td>\n",
       "      <td>7578480</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCTGCCCTGTGCAGCTGT</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACA</td>\n",
       "      <td>30</td>\n",
       "      <td>GCTGCACAGGGCAGG</td>\n",
       "      <td>15</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGGGCAGG</td>\n",
       "      <td>17</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33967</th>\n",
       "      <td>982</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578479</td>\n",
       "      <td>7578480</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCTGCCCTGTGCAGCTGT</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACA</td>\n",
       "      <td>30</td>\n",
       "      <td>GCTGCACAGGGCAGGTC</td>\n",
       "      <td>17</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGGGCAGGTC</td>\n",
       "      <td>17</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation_idx Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "27228           792        TP53         17         7577035       7577036   \n",
       "27229           792        TP53         17         7577035       7577036   \n",
       "27230           792        TP53         17         7577035       7577036   \n",
       "27231           792        TP53         17         7577035       7577036   \n",
       "27232           792        TP53         17         7577035       7577036   \n",
       "...             ...         ...        ...             ...           ...   \n",
       "33963           982        TP53         17         7578479       7578480   \n",
       "33964           982        TP53         17         7578479       7578480   \n",
       "33965           982        TP53         17         7578479       7578480   \n",
       "33966           982        TP53         17         7578479       7578480   \n",
       "33967           982        TP53         17         7578479       7578480   \n",
       "\n",
       "              Consequence Variant_Classification Variant_Type  \\\n",
       "27228  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "27229  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "27230  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "27231  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "27232  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "...                   ...                    ...          ...   \n",
       "33963  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "33964  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "33965  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "33966  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "33967  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "\n",
       "      Reference_Allele Tumor_Seq_Allele2  ...           Protospacer  \\\n",
       "27228                -                 G  ...  GCCTCGCTTAGTGCTCCCTG   \n",
       "27229                -                 G  ...  GCCTCGCTTAGTGCTCCCTG   \n",
       "27230                -                 G  ...  GCCTCGCTTAGTGCTCCCTG   \n",
       "27231                -                 G  ...  GCCTCGCTTAGTGCTCCCTG   \n",
       "27232                -                 G  ...  GCCTCGCTTAGTGCTCCCTG   \n",
       "...                ...               ...  ...                   ...   \n",
       "33963                -                 A  ...  GACCTGCCCTGTGCAGCTGT   \n",
       "33964                -                 A  ...  GACCTGCCCTGTGCAGCTGT   \n",
       "33965                -                 A  ...  GACCTGCCCTGTGCAGCTGT   \n",
       "33966                -                 A  ...  GACCTGCCCTGTGCAGCTGT   \n",
       "33967                -                 A  ...  GACCTGCCCTGTGCAGCTGT   \n",
       "\n",
       "                                  RTT  RTT_length                PBS  \\\n",
       "27228            CACCACGAGCTGCCCCCCAG          20         GGAGCACTAA   \n",
       "27229            CACCACGAGCTGCCCCCCAG          20      GGAGCACTAAGCG   \n",
       "27230            CACCACGAGCTGCCCCCCAG          20    GGAGCACTAAGCGAG   \n",
       "27231            CACCACGAGCTGCCCCCCAG          20  GGAGCACTAAGCGAGGT   \n",
       "27232  AGGGGAGCCTCACCACGAGCTGCCCCCCAG          30         GGAGCACTAA   \n",
       "...                               ...         ...                ...   \n",
       "33963            GGATGTGGAATCAACCCACA          20  GCTGCACAGGGCAGGTC   \n",
       "33964  GCCGGGCGGGGGATGTGGAATCAACCCACA          30         GCTGCACAGG   \n",
       "33965  GCCGGGCGGGGGATGTGGAATCAACCCACA          30      GCTGCACAGGGCA   \n",
       "33966  GCCGGGCGGGGGATGTGGAATCAACCCACA          30    GCTGCACAGGGCAGG   \n",
       "33967  GCCGGGCGGGGGATGTGGAATCAACCCACA          30  GCTGCACAGGGCAGGTC   \n",
       "\n",
       "      PBS_length                                          RTT_PBS  \\\n",
       "27228         10                   CACCACGAGCTGCCCCCCAGGGAGCACTAA   \n",
       "27229         13                CACCACGAGCTGCCCCCCAGGGAGCACTAAGCG   \n",
       "27230         15              CACCACGAGCTGCCCCCCAGGGAGCACTAAGCGAG   \n",
       "27231         17            CACCACGAGCTGCCCCCCAGGGAGCACTAAGCGAGGT   \n",
       "27232         10         AGGGGAGCCTCACCACGAGCTGCCCCCCAGGGAGCACTAA   \n",
       "...          ...                                              ...   \n",
       "33963         17            GGATGTGGAATCAACCCACAGCTGCACAGGGCAGGTC   \n",
       "33964         10         GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGG   \n",
       "33965         13      GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGGGCA   \n",
       "33966         15    GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGGGCAGG   \n",
       "33967         17  GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGGGCAGGTC   \n",
       "\n",
       "      Distance_to_nick RHA_size PAM_disrupted Proto_disrupted  \n",
       "27228                2       17         False            True  \n",
       "27229                2       17         False            True  \n",
       "27230                2       17         False            True  \n",
       "27231                2       17         False            True  \n",
       "27232                2       27         False            True  \n",
       "...                ...      ...           ...             ...  \n",
       "33963               17        2         False           False  \n",
       "33964               17       12         False           False  \n",
       "33965               17       12         False           False  \n",
       "33966               17       12         False           False  \n",
       "33967               17       12         False           False  \n",
       "\n",
       "[1000 rows x 34 columns]"
      ]
     },
     "execution_count": 818,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Variant_Type']=='INS'][-1000:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 786,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 786,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 782,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TGCGCCGGTCTCTCCCAGGACAGGCACAAACAT')"
      ]
     },
     "execution_count": 782,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Bio.Seq.Seq(df.iloc[1]['RTT_PBS']).reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/5g/xxjcy2g531n02hnyb6d8qdsr0000gn/T/ipykernel_79410/1396537375.py\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'df' is not defined"
     ]
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 780,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 780,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('GCCGGTCTCTCCCAGGACAGGCACAAACAT')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Protospacer ranking\n",
    "\n",
    "https://github.com/gpp-rnd/rs3\n",
    "\n",
    "https://github.com/microsoftResearch//azimuth\n",
    "\n",
    "https://pypi.org/project/Biomatters-Azimuth-2/\n",
    "\n",
    "\n",
    "This seems like the best option: https://github.com/crisprVerse/crisprScore\n",
    "\n",
    "Use in python: https://willfondrie.com/2022/01/how-to-use-r-packages-in-python/\n",
    "\n",
    "May be able to get this to work: https://github.com/crisprVerse/crisprScore/blob/master/inst/python/rs3/targetfeat.py\n",
    "\n",
    "> Modify these and port them into my workspace..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This code is adopted from CRISPOR program with minor modifications\n",
    "# https://github.com/maximilianh/crisporWebsite\n",
    "# \"Evaluation of off-target and on-target scoring algorithms and integration into the guide RNA selection tool CRISPOR\"\n",
    "# by M. Haeussler et al. 2016\n",
    "\n",
    "import math\n",
    " \n",
    "params = [\n",
    "# pasted/typed table from PDF and converted to zero-based positions\n",
    "(1,'G',-0.2753771),(2,'A',-0.3238875),(2,'C',0.17212887),(3,'C',-0.1006662),\n",
    "(4,'C',-0.2018029),(4,'G',0.24595663),(5,'A',0.03644004),(5,'C',0.09837684),\n",
    "(6,'C',-0.7411813),(6,'G',-0.3932644),(11,'A',-0.466099),(14,'A',0.08537695),\n",
    "(14,'C',-0.013814),(15,'A',0.27262051),(15,'C',-0.1190226),(15,'T',-0.2859442),\n",
    "(16,'A',0.09745459),(16,'G',-0.1755462),(17,'C',-0.3457955),(17,'G',-0.6780964),\n",
    "(18,'A',0.22508903),(18,'C',-0.5077941),(19,'G',-0.4173736),(19,'T',-0.054307),\n",
    "(20,'G',0.37989937),(20,'T',-0.0907126),(21,'C',0.05782332),(21,'T',-0.5305673),\n",
    "(22,'T',-0.8770074),(23,'C',-0.8762358),(23,'G',0.27891626),(23,'T',-0.4031022),\n",
    "(24,'A',-0.0773007),(24,'C',0.28793562),(24,'T',-0.2216372),(27,'G',-0.6890167),\n",
    "(27,'T',0.11787758),(28,'C',-0.1604453),(29,'G',0.38634258),(1,'GT',-0.6257787),\n",
    "(4,'GC',0.30004332),(5,'AA',-0.8348362),(5,'TA',0.76062777),(6,'GG',-0.4908167),\n",
    "(11,'GG',-1.5169074),(11,'TA',0.7092612),(11,'TC',0.49629861),(11,'TT',-0.5868739),\n",
    "(12,'GG',-0.3345637),(13,'GA',0.76384993),(13,'GC',-0.5370252),(16,'TG',-0.7981461),\n",
    "(18,'GG',-0.6668087),(18,'TC',0.35318325),(19,'CC',0.74807209),(19,'TG',-0.3672668),\n",
    "(20,'AC',0.56820913),(20,'CG',0.32907207),(20,'GA',-0.8364568),(20,'GG',-0.7822076),\n",
    "(21,'TC',-1.029693),(22,'CG',0.85619782),(22,'CT',-0.4632077),(23,'AA',-0.5794924),\n",
    "(23,'AG',0.64907554),(24,'AG',-0.0773007),(24,'CG',0.28793562),(24,'TG',-0.2216372),\n",
    "(26,'GT',0.11787758),(28,'GG',-0.69774)]\n",
    " \n",
    "intercept =  0.59763615\n",
    "gc_high    = -0.1665878\n",
    "gc_low     = -0.2026259\n",
    " \n",
    "def calc_doench_score(seq):\n",
    "  score = intercept\n",
    " \n",
    "  #guide = seq[4:24]\n",
    "  guide = seq\n",
    "  gc_count = guide.count('G') + guide.count('C')\n",
    "  gc_weight = gc_low if gc_count <= 10 else gc_high\n",
    "  score += abs(10-gc_count)*gc_weight\n",
    " \n",
    "  for pos, model_seq, weight in params:\n",
    "    sub_seq = seq[pos:pos+len(model_seq)]\n",
    "    if sub_seq == model_seq:\n",
    "      score += weight\n",
    "  total_score = 1.0/(1.0+math.exp(-score)) * 100\n",
    "  return '%.3f' % total_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['3.911', '8.716', '15.631', '0.019']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs = ['AGAGAGATTCTTGCTCACCGTCTGTGGAGA',\n",
    "'AGAGAGATTCTTGCTCACCGTCTGCGGAGA',\n",
    "'AGAGAGATTCTTCATCACCGTCTGCGGAGA',\n",
    "'GGGGGGGGGGGGGGGGGGGGGGGGGGGGGG']\n",
    "\n",
    "[calc_doench_score(seq) for seq in seqs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[58]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#rs2/Azimuth\n",
    "\n",
    "import crisporEffScores\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[58, 63, 67, 9]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs = ['AGAGAGATTCTTGCTCACCGTCTGTGGAGA',\n",
    "'AGAGAGATTCTTGCTCACCGTCTGCGGAGA',\n",
    "'AGAGAGATTCTTCATCACCGTCTGCGGAGA',\n",
    "'GGGGGGGGGGGGGGGGGGGGGGGGGGGGGG']\n",
    "\n",
    "#3 extra on the PAM side; 4 extra on protospacer side\n",
    "\n",
    "\n",
    "scores = crisporEffScores.calcAziScore(seqs)\n",
    "scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "#checking how closely it matches up with previously calculated scores...\n",
    "\n",
    "p53_lib = pd.read_csv('/Users/samgould/Desktop/FSR Lab/Manuscripts/p53 PE sensor/final submission/Supplemental/Supplemental-Table-1_pegRNAs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['peg_id', 'variant_index', 'HGVSp', 'HGVSc', 'gene', 'chr', 'start',\n",
       "       'end', 'variant_type', 'classification', 'consequence', 'ref_allele',\n",
       "       'mut_allele', 'codon', 'Domain', 'num_occurences', 'WT_sensor',\n",
       "       'CorrectEdit_sensor', 'PAM_location', 'PAM_strand', 'PAM',\n",
       "       'protospacer', 'PBS', 'PBS_length', 'RTT', 'RTT_length', 'PBS_RTT_5to3',\n",
       "       'epegRNA_tevopreQ1_full', 'PBS_GC_content', 'distance_to_nick',\n",
       "       'homology_overhang', 'MIT_specificity', 'Doench16_Azimuth_percentile',\n",
       "       'PEGG_score', 'wt_qwc', 'edit_qwc'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p53_lib.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr17 = chrom_dict[17]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "PAM_locs = np.unique(p53_lib[p53_lib['PAM_strand']=='+']['PAM_location'])\n",
    "\n",
    "seq_holder = []\n",
    "for i in range(len(PAM_locs)):\n",
    "    a = str(chr17[PAM_locs[i]-20-4:PAM_locs[i]+3+3]).upper()\n",
    "    seq_holder.append(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plus = p53_lib[p53_lib['PAM_strand']=='+']\n",
    "score_ucsc = []\n",
    "for i in PAM_locs:\n",
    "    b = plus[plus['PAM_location']==i]['Doench16_Azimuth_percentile'].values[0]\n",
    "    score_ucsc.append(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "scores = crisporEffScores.calcAziScore(seq_holder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'Calcualted Azimuth Score')"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjMAAAGwCAYAAABcnuQpAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAA9hAAAPYQGoP6dpAABWm0lEQVR4nO3de1wU9foH8M9yW8BgFS8sGOKamiJp3vNSaAWaZpmnTkczNf35M+2iUalkJZRCaoesNDVOmR7zaP3S0jSTLFFT00RSwbyFaApxjiCgIAj7/f3B2Y3lOrPsZWb5vF+vfb3a2e/OPDO7LY/feeYZjRBCgIiIiEil3JwdABEREVFjMJkhIiIiVWMyQ0RERKrGZIaIiIhUjckMERERqRqTGSIiIlI1JjNERESkah7ODsDejEYjLl++DD8/P2g0GmeHQ0RERBIIIVBUVITg4GC4udU/9+Lyyczly5cREhLi7DCIiIjIChcvXsStt95a7xiXT2b8/PwAVB4Mf39/J0dDREREUhQWFiIkJMT8d7w+Lp/MmE4t+fv7M5khIiJSGSklIk4tAC4vL8err74Kg8EAHx8fdOjQAW+88QaMRqN5jBACsbGxCA4Oho+PD4YMGYL09HQnRk1ERERK4tRkZtGiRVi5ciWWLVuGkydPYvHixViyZAnef/9985jFixcjMTERy5Ytw+HDh6HX6xEZGYmioiInRk5ERERK4dRk5sCBA3j44YcxcuRItG/fHo8++iiioqLw888/A6iclVm6dCnmzZuHMWPGIDw8HGvWrEFxcTHWr1/vzNCJiIhIIZyazAwePBi7du3C6dOnAQC//PIL9u3bhxEjRgAAMjMzkZOTg6ioKPN7tFotIiIisH///lrXWVpaisLCQosHERERuS6nFgDPmTMHBQUF6NKlC9zd3VFRUYGFCxdi7NixAICcnBwAQGBgoMX7AgMDkZWVVes6ExISEBcXZ9/AiYiISDGcOjOzceNGrFu3DuvXr0dqairWrFmDt99+G2vWrLEYV72SWQhRZ3VzTEwMCgoKzI+LFy/aLX4iIiJyPqfOzLz88suYO3cu/va3vwEA7rjjDmRlZSEhIQETJ06EXq8HUDlDExQUZH5fbm5ujdkaE61WC61Wa//giYiISBGcOjNTXFxco0Wxu7u7+dJsg8EAvV6P5ORk8+tlZWVISUnBwIEDHRorERERKZNTZ2ZGjRqFhQsXol27dujWrRuOHj2KxMRETJ48GUDl6aVZs2YhPj4enTp1QqdOnRAfHw9fX1+MGzfOmaETERGRQjg1mXn//ffx2muvYcaMGcjNzUVwcDCmTZuG119/3Txm9uzZKCkpwYwZM5Cfn4/+/ftj586dktobExERkX1UGAUOZeYht+gG2vh5o58hAO5uzrmhs0YIIZyyZQcpLCyETqdDQUEBb2dARERkAztOZCNuawayC26YlwXpvDF/VBiGhwfV807p5Pz9dmrNDBEREanLjhPZmL4u1SKRAYCcghuYvi4VO05kOzwmJjNEREQkSYVRIG5rBmo7pWNaFrc1AxVGx570YTJDREREkhzKzKsxI1OVAJBdcAOHMvMcFxSYzBAREZFEuUV1JzLWjLMVJjNEREQkSRs/b5uOsxUmM0RERCRJP0MAgnTeqOsCbA0qr2rqZwhwZFhMZoiIiEgadzcN5o8KA4AaCY3p+fxRYQ7vN8NkhoiIiCQbHh6EFeN7Qa+zPJWk13ljxfheNuszI4dTOwATERGR+gwPD0JkmF4xHYCZzBAREZFs7m4aDLitpbPDAMDTTERERKRyTGaIiIhI1ZjMEBERkaoxmSEiIiJVYzJDREREqsZkhoiIiFSNyQwRERGpGpMZIiIiUjUmM0RERKRqTGaIiIhI1ZjMEBERkaoxmSEiIiJVYzJDREREqsZkhoiIiFSNyQwRERGpGpMZIiIiUjUmM0RERKRqTGaIiIhI1ZjMEBERkaoxmSEiIiJVYzJDREREqsZkhoiIiFSNyQwRERGpGpMZIiIiUjUmM0RERKRqTGaIiIhI1ZjMEBERkap5ODsAIiIiUr4Ko8ChzDzkFt1AGz9v9DMEwN1N4+ywADCZISIiogbsOJGNuK0ZyC64YV4WpPPG/FFhGB4e5MTIKjn1NFP79u2h0WhqPJ555hkAgBACsbGxCA4Oho+PD4YMGYL09HRnhkxERNSk7DiRjenrUi0SGQDIKbiB6etSseNEtpMi+5NTk5nDhw8jOzvb/EhOTgYAPPbYYwCAxYsXIzExEcuWLcPhw4eh1+sRGRmJoqIiZ4ZNRETUJFQYBeK2ZkDU8pppWdzWDFQYaxvhOE5NZlq3bg29Xm9+fP3117jtttsQEREBIQSWLl2KefPmYcyYMQgPD8eaNWtQXFyM9evX17nO0tJSFBYWWjyIiIhIvkOZeTVmZKoSALILbuBQZp7jgqqFYq5mKisrw7p16zB58mRoNBpkZmYiJycHUVFR5jFarRYRERHYv39/netJSEiATqczP0JCQhwRPhERkWpVGAUOnLuCr9Iu4cC5K+aZltyiuhOZqqSOsxfFFAB/+eWXuHr1KiZNmgQAyMnJAQAEBgZajAsMDERWVlad64mJiUF0dLT5eWFhIRMaIiKiOtRX3NvGz1vSOqSOsxfFzMx89NFHeOCBBxAcHGyxXKOxvOxLCFFjWVVarRb+/v4WDyIiIqqpoeLe/OtlCNJ5o66/uhpUJj79DAF2j7U+ikhmsrKy8N133+F//ud/zMv0ej2AP2doTHJzc2vM1hAREZE8Uop739yWgddGdgWAGgmN6fn8UWFO7zejiGRm9erVaNOmDUaOHGleZjAYoNfrzVc4AZV1NSkpKRg4cKAzwiQiInIZUot7WzTTYsX4XtDrLE8l6XXeWDG+lyL6zDi9ZsZoNGL16tWYOHEiPDz+DEej0WDWrFmIj49Hp06d0KlTJ8THx8PX1xfjxo1zYsRERETqJ6e49+E72yIyTM8OwHX57rvvcOHCBUyePLnGa7Nnz0ZJSQlmzJiB/Px89O/fHzt37oSfn58TIiUiInIdcot73d00GHBbS3uGZDWNEMK5nW7srLCwEDqdDgUFBSwGJiIi+q8Ko0DvBcm4WnyzzjHNfT1x5NVIp8zAyPn7rYiaGSIiIlIeZZxEahiTGSIioiboUGZevbMyAJBffNPp3X2lcHrNDBERETlOhVHgUGYevpF4g0hnd/eVgskMERFRE1Fbt9+GOLu7rxRMZoiIiJoAU7dfOVf9tPD1dHp3XylYM0NEROTi6uv2Wx+1XO7MZIaIiMjFNdTtty5XWQBMRERE1jIV6tqi425jinhZAExERESy1VaoG6TzxvxRYVbdC6kxRbxqKADmaSYiIiIFMRXqVj8tlFNwA9PXpWKHxEuqq+pnCECQzltWEzwNKhMoFgATERGRZPUV6pqWxW3NQIVRXmmuu5sG80eFAZDW1dc0Zv6oMMXcTLI+TGaIiIgUoqFCXQEgu+CGVUW5w8ODsGJ8L+h1lqeNWvh6ormvp8Uyvc4bK8b3suqUljOwZoaIiEghpBbb/nj235ILgqsXEqe8PBRHsvItCosB2KzY2BmYzBARESmE1GLbZT+cwxeplxosCK6vkPjhO9tajB1wW0vrglYAnmYiIiJSCDmFug0VBNujkFipmMwQEREphJxC3foKgu1VSKxUTGaIiIgUpK5C3drUVRBsz0JiJWIyQ0REVIsKo8CBc1fwVdolHDh3xaGzGMPDg7Bvzr14dmhHSeOrFw5LLSRWQ3dfKVgATEREVI2tO/Baw91Ng0EdW2HZD2cbHFu9cFhqIbEauvtKwZkZIiKiKpRUONtQQXBdXXqtfZ9aMZkhIiL6L6UVztZXEFxfl15r36dWTGaIiIj+S4mFs3UVBDfUpdfa96kRa2aIiMjlVe+CW1eHW0cVzpaVG/HPA+eRlVeM0ABfPDmgPbw86p5fGB4ehHu7BMp6j+l9kWF6VXf3lYLJDBERuTQ5xbyOKJxN2J6BpL2ZqHqmauH2k5h6twExI8JqfU9t+/CPfZmSCpLd3TSq7u4rBU8zERGRy5JbzGvvwtmE7RlYtccykQEAowBW7clEwvaMRu9DU8RkhoiIXJI1xbz2LJwtKzciaW9mvWOS9mairNxofq60gmSlYjJDREQuydpiXnsVzv7zwPkaMzLVGUXlOBMlFiQrEWtmiIjIJTWmmFdqwa3UwmIAyMorlhRP1XFNrZOvtZjMEBGRS2pMMa+Uglu5XYJDA3wlxVN1XFPr5GstnmYiIiKXZG0xr5SCW2uKcp8c0B4Nldq4aSrHNXYfmhomM0RE5JKsKeaVWnAbuyVddlGul4cbpt5tqDfmqXcbLE5lNbVOvtZiMkNERC5LbjGv1ILbnMLSBsfUVpQbMyIM0+4x1JihcdMA0+6pvc9MU+rkay3WzJBiyCmkI6KmyZrfCTndc21ZSFvXumJGhOHFqC6yOwBL6eTbVH9HrU5mysrKkJmZidtuuw0eHsyJqHHkFtIRUdNj7e+EnO65tiykrW9dXh5umHJ3B1nra6iTb1P+HZV9mqm4uBhTpkyBr68vunXrhgsXLgAAnn/+ebz11ls2D5BcH7tbElFDrP2dsFcHYL2/VlFFuU39d1R2MhMTE4NffvkFu3fvhrf3n1nn/fffj40bN9o0OHJ97G5JRA2x9nfCnh2AYx/q1uAYR53e4e+oFcnMl19+iWXLlmHw4MHQaP78oMLCwnDu3DmbBkeuj90tiagh1v5O2LMDsJKKcvk7akXNzL///W+0adOmxvLr169bJDdEUrC7JVHTVVZutGlRbvVxje0A3FDBrdSiXHvj76gVMzN9+/bFtm3bzM9NCUxSUhIGDBggO4BLly5h/PjxaNmyJXx9fXHnnXfiyJEj5teFEIiNjUVwcDB8fHwwZMgQpKeny94OKRO7WxI1TQnbM9DltW/w5raTWHsgC29uO4kur31T612jrf2daOzvi6ng9uE722LAbS1rTVKkjLE3/o5aMTOTkJCA4cOHIyMjA+Xl5Xj33XeRnp6OAwcOICUlRda68vPzMWjQIAwdOhTffPMN2rRpg3PnzqF58+bmMYsXL0ZiYiI++eQTdO7cGQsWLEBkZCROnToFPz8/ueGTwpiK7XIKbtR6vleDymnbpt7dksiVJGzPwKo9Ne8ebRQwL6/ab8Xa34mm8vvSVPazPrJnZgYOHIj9+/ejuLgYt912G3bu3InAwEAcOHAAvXv3lrWuRYsWISQkBKtXr0a/fv3Qvn173HfffbjtttsAVM7KLF26FPPmzcOYMWMQHh6ONWvWoLi4GOvXr5cbOikQu1sSNS1l5UYk7a2ZyFSVtDcTZeVG83Nrfyeayu9LU9nP+shKZm7evImnnnoKvr6+WLNmDU6cOIGMjAysW7cOd9xxh+yNb9myBX369MFjjz2GNm3aoGfPnkhKSjK/npmZiZycHERFRZmXabVaREREYP/+/bWus7S0FIWFhRYPUjYlFdIRkX3988B5NHRRjVFUjqvK2t+JpvL70lT2sy6yTjN5enpi8+bNeO2112yy8d9++w0rVqxAdHQ0XnnlFRw6dAjPP/88tFotJkyYgJycHABAYGCgxfsCAwORlZVV6zoTEhIQFxdnk/jIcZRSSEdE8snpOpuVVyxpnbWNk9PJt/r77PX7oqSOu035d1R2zcwjjzyCL7/8EtHR0Y3euNFoRJ8+fRAfHw8A6NmzJ9LT07FixQpMmDDBPK76VVJCiDqvnIqJibGIrbCwECEhIY2Oleyvoe6WRKQ8crvOhgb4SlpvbePkdPKtzh6/L0rsuNtUf0dlJzMdO3bEm2++if3796N3795o1qyZxevPP/+85HUFBQUhLMzyplpdu3bFF198AQDQ6/UAgJycHAQF/fnFyM3NrTFbY6LVaqHVaiXHQERE1jF1na1+1sjUdba20xtPDmiPhdtP1nuqyU1TOa6x27InpcXT1MlOZv7xj3+gefPmOHLkiMUl1EDlDIqcZGbQoEE4deqUxbLTp08jNDQUAGAwGKDX65GcnIyePXsCqLwnVEpKChYtWiQ3dCIispGGus5qUNl1NjJMb3Gaw8vDDVPvNtR6NZPJ1LsNFqeOrN2WvSgtHrIimcnMrL8KXY4XXngBAwcORHx8PP7617/i0KFD+PDDD/Hhhx8CqEyOZs2ahfj4eHTq1AmdOnVCfHw8fH19MW7cOJvFQURE8sjpOlv9tIfpsuukvZkWMzRumspEpupl2Y3dlj0oLR5qxF2zgcraFaBmTYtUffv2xebNmxETE4M33ngDBoMBS5cuxRNPPGEeM3v2bJSUlGDGjBnIz89H//79sXPnTvaYISKyA6kFrY3tOhszIgwvRnWxawdge1FaPGRlMrN27VosWbIEZ86cAQB07twZL7/8Mp588knZ63rwwQfx4IMP1vm6RqNBbGwsYmNjrQmViIgkklPQaouus14ebphyd4dGrcOacY2ltHjIiqZ5iYmJmD59OkaMGIHPPvsMGzduxPDhw/H000/jnXfesUeMRERkZ6aC1uqnT0wFrTtOZFssN3WdrWteXoPKRMgWXWcduS01xkNWJDPvv/8+VqxYgUWLFuGhhx7Cww8/jMWLF+ODDz7Ae++9Z48YiYjIjhoqaAUqC1orqhS4OLLrrNI63CotHrIimcnOzsbAgQNrLB84cCCys7NreQcRESmZnILWqhzZdVZpHW6VFk9TZ1Wfmc8++wyvvPKKxfKNGzeiU6dONguMiIgqlZRVIH57Bs5fKUb7lr54ZUQYfLzcbbb+xhS0OrLrrKM73DZUDN2UO+4qjexkJi4uDo8//jj27NmDQYMGQaPRYN++fdi1axc+++wze8RIRNRkTV17GMkZuebne88A/zx4AZFhbZA0oa9NttHYglZHdp111LakFkM31Y67SiP7NNNf/vIX/PTTT2jVqhW+/PJLbNq0Ca1atcKhQ4fwyCOP2CNGIqImqXoiU1VyRi6mrj1sk+2woNWS3GJocj6rLs3u3bs31q1bZ+tYiIjov0rKKupMZEySM3JRUlbR6FNOpoLW6etSoQEsCoGbWkEru/uqk+yZme3bt+Pbb7+tsfzbb7/FN998Y5OgiIiauvjtGTYd1xAWtFaythianEv2zMzcuXPx1ltv1VguhMDcuXPxwAMP2CQwIqKm7PyV4kaNk9rJtyprC1qt2Za1ysqNkroGW0vJ3X0deZzVRnYyc+bMmRp3ugaALl264OzZszYJioioqWvf0hd7z0gbV52cTr7VyS1obcy25ErYnlHjfk4Lt5+s9X5O1lJqd19HHmc1kp3O6nQ6/PbbbzWWnz17Fs2aNbNJUERETd0rEv84Vx/nyOJVR24rYXsGVu2xTGQAwCiAVXsykWCj021KLIZmQXLDZCczDz30EGbNmoVz586Zl509exYvvvgiHnroIZsGR0TUVHl5uEHbwOkTrYebxSkWazr5WsuR2yorNyJpb2a9Y5L2ZqKs3NjobSmtu68jj7OayU5mlixZgmbNmqFLly4wGAwwGAzo2rUrWrZsibffftseMRIRNTmHMvNQ2sAf59Jyo0UhqiOLVx25rX8eOF9jRqY6o6gcZwtKKoZmQbI0smtmdDod9u/fj+TkZPzyyy/w8fFB9+7dcc8999gjPiIi1WpMwaY1haiOLF61xbakHp+sPGnF0FLHSWHv7r5S991Rn6nai4ut6jOj0WgQFRWFqKgoW8dDROQSGluwaU0hqiOLVxu7LTnHJzSgZpFzbaSOk8pe3X3l7LsjPlNXKC6WfJrpp59+qtFHZu3atTAYDGjTpg3+93//F6WlpTYPkIhIbWxRsGlNIaoji1cbsy25x+fJAe3R0CSBm6ZynNLJ3Xd7f6auUlwsOZmJjY3FsWPHzM+PHz+OKVOm4P7778fcuXOxdetWJCQk2CVIIiK1sFXBpjWFqI4sXrV2W9YcHy8PN0y921BvPFPvNti034w9WLPv9vxMXam4WPInn5aWhvvuu8/8fMOGDejfvz+SkpIQHR2N9957jzeaJKImz5YFm9YUojqyeNWabVl7fGJGhGHaPYYaMzRuGmDaPbbrM2NP1u67vT5TVyoullwzk5+fj8DAQPPzlJQUDB8+3Py8b9++uHjxom2jIyJSADnFkbYu2LSmENXexauN2VZjjk/MiDC8GNXFrh2A7akx+26Pz1TJ3Y7lkpzMBAYGIjMzEyEhISgrK0Nqairi4uLMrxcVFcHT09MuQRIROYvc4kh7FGxaU4hqr+LVxm6rscfHy8MNU+7uIDk2JWnsvtv6M1Vqt2NrSE5nhw8fjrlz52Lv3r2IiYmBr68v7r77bvPrx44dw2233WaXIImInMGa4kgldpBVkqZ8fJS270qLpzEkJzMLFiyAu7s7IiIikJSUhKSkJHh5eZlf//jjj3mpNhG5DGuLI5XWQVZpmvLxUdq+Ky2extAIIWSVKRcUFOCWW26Bu7u7xfK8vDzccsstFgmOEhQWFkKn06GgoAD+/v7ODoeaALU3n6JKB85dwdikgw2O+9fUu2qd+rdl7w5X/E65Qm8Taylt35UWj4mcv9+ykxm1YTJDjqTUHwWS76u0S5i5Ia3Bce/+7U48fGfbWl+zRRLiyt8pV0zSpFLavistHoDJjAUmM+QopvqK6v9DmX4OHH1PF2qcxs7M2AK/U9SUyfn7rY7r2YgUzpWaT1ElZxdH8jtFJB2TGSIbcKXmU1TJ2cWR/E4RScdkhsgGXKn5lKurMAocOHcFX6VdwoFzV+qd2TB1Xg30t3833er4nSKSzqq7Zp8+fRq7d+9Gbm4ujEajxWuvv/66TQIjUhNXaj7lyqwvprVMeBxRasjvFJF0spOZpKQkTJ8+Ha1atYJer4dG8+cUq0ajYTJDTZKpviKn4EatNQ4aVP5rXg3Np1xVXcW0pgZ4tc201PWePwpL63yPrfA7RSSd7NNMCxYswMKFC5GTk4O0tDQcPXrU/EhNTbVHjESK5+z6CqqfNcW0zi7A5XeKSDrZyUx+fj4ee+wxe8RCpGqOvFsxyWNNMa0SCnD5nSKSRvZppsceeww7d+7E008/bY94iFTNkXcrdraycqNVdy92RnMua4pplVKA68rfKSU2aiN1kpTMvPfee+b/7tixI1577TUcPHgQd9xxR407ZT///PO2jZBIZRx5t2JnSdiegaS9mah6hmXh9pOYercBMSPC6nyfs7rZWlNMq6QCXFf8TrlyZ2NyPEkdgA0Gg7SVaTT47bffGh2ULbEDMJFtJWzPwKo9mXW+Pu2e2hMaZ3azrTAK9F6QjKvFN+sc08LXEz+/GmmeGbDmPSQNOxuTFHL+fkuamcnMrPuHi4iajrJyI5L21v97kLQ3Ey9GdbE45dRQMa0GlcW0kWF6pyUG1pTxsveufGr4LpD6yC4AfuONN1BcXFxjeUlJCd544w2bBEVEyvTPA+fR0MU7RlE5ripnF9Meysyrd4YFAK4W36xRACz3PdQwZ38XyDXJTmbi4uJw7dq1GsuLi4sRFxdnk6CIqCY5nWvtJSuv5j9kpIyzRzFtWbkRH+39Da9/dQIf7f0NZeXGOsc6qwBYToxNhVIKq8m1yL6aSQhh0SjP5JdffkFAAJs3EdmDUoolQwN8rRpn62JauQXIzigAtrZI2tUpqbCaXIfkmZkWLVogICAAGo0GnTt3RkBAgPmh0+kQGRmJv/71r7I2HhsbC41GY/HQ6/Xm14UQiI2NRXBwMHx8fDBkyBCkp6fL2gaR2pmKJatPzZs61+44ke2wWJ4c0B4NlTG4aSrHVdXPEIDmvp61v+G/Wvh6SupmaypArj4xZRTAqj2ZSNieUeM91twBuzF3zbYmxqZCynehucTvApGJ5JmZpUuXQgiByZMnIy4uDjqdzvyal5cX2rdvjwEDBsgOoFu3bvjuu+/Mz93d3c3/vXjxYiQmJuKTTz5B586dsWDBAkRGRuLUqVPw8/OTvS0itVFasaSXhxum3m2o92qmqXcbJPWbqU7KSTNrC5BN3XSnr0uFptq26uqma817GhMj/YllvySX5GRm4sSJACov0x44cGCN/jJWB+DhYTEbYyKEwNKlSzFv3jyMGTMGALBmzRoEBgZi/fr1mDZtWq3rKy0tRWlpqfl5YWGhTeIkcgY5xZKO6kNiOkVS/RSKmwZ1nkKRU0xb337IKUCecncHi+WmbrrVT9fp6zldZ817GhNjUyDlu5Av4btAVJXsmhmDwYDs7Lqntdu1aydrfWfOnEFwcDC0Wi369++P+Ph4dOjQAZmZmcjJyUFUVJR5rFarRUREBPbv319nMpOQkMBCZHIZSi2WjBkRhln334747Rk4f6UY7Vv64pURYfDxcq91vK32w9oCZBNruunKfU9jY3R1Sv1Ok7rJTmbat29fawGwSUVFheR19e/fH2vXrkXnzp3xxx9/YMGCBRg4cCDS09ORk5MDAAgMDLR4T2BgILKysupcZ0xMDKKjo83PCwsLERISIjkmIiVRarFk9YLkvWeA707m1jlbYav9sLYAuSpruunKeY8tYnRlSv1Ok7rJPmFruju26fHTTz9h5cqV6Ny5Mz7//HNZ63rggQfwl7/8BXfccQfuv/9+bNu2DUDl6SST6olTXVdTmWi1Wvj7+1s8iNSqMUWo9mJNQbKt9sPaAmRHUkOMzqTE7zSpn+xkpkePHhaPPn36YOrUqXj77bct7uFkjWbNmuGOO+7AmTNnzHU0phkak9zc3BqzNUSuylSECtQsiqyvCNVeGipIBioLkqv3wLHVfpgKkOtjbQGyraghRmdS2neaXIPN/m/q3LkzDh8+3Kh1lJaW4uTJkwgKCoLBYIBer0dycrL59bKyMqSkpGDgwIGNDZdINUxFqHqd5bS7Xuft8HvYNKZ7q632I2ZEGKbdY6gx++Gmqfu+UI6mhhidSUnfaXINkm40WVX1q4OEEMjOzkZsbCx+/fVXpKWlSV7XSy+9hFGjRqFdu3bIzc3FggULkJKSguPHjyM0NBSLFi1CQkICVq9ejU6dOiE+Ph67d++WdWk2bzRZtwqjkFUISc5VVm7EPw+cR1ZeMUIDfPHkgPYO/9f9V2mXMHNDWoPj3v3bnXj4zra1vmar711JWYXkAmRnUcJnpmT8DaL62PxGk1U1b9681jqWkJAQbNiwQda6fv/9d4wdOxb/+c9/0Lp1a9x11104ePAgQkNDAQCzZ89GSUkJZsyYgfz8fPTv3x87d+5kjxkbUEpHWZKmts/rH/syHf552aJ405oC3OrkFiA7i5eHW5O8/FoqW3wXiAArZmZSUlIsnru5uaF169bo2LEjPDxk50Z2x5mZmkwFnNU/eFOKymleZVHS51VhFOi9ILnePiHNfT1x5NVIu/0LW0nHg4jsx64zMxEREVYHRs6ntI6yVD81fl72jEKNx4OI7M+qqZRLly7hxx9/RG5uLoxGy7vAPv/88zYJjOxDiR1lqW5K+7yc3b1VaceDiJRBdjKzevVqPP300/Dy8kLLli0t6mc0Gg2TGYVj9011Udrn5ex4nL19IlIm2cnM66+/jtdffx0xMTFwc2NVvtqw+6a6KO3zcnY8zt4+ESmT7GykuLgYf/vb35jIqBS7b6qL0j4vZ8fTzxCA5r713+S2ua8nv79ETYzsjGTKlCmyb1tAysHum+qitM9LafHUht9coqZH9qXZFRUVePDBB1FSUoI77rgDnp6W/0pKTEy0aYCNxUuza8c+M+qitM/LWfEcOHcFY5MONjjuX1PvYgEwkcrZ9dLs+Ph4fPvtt7j99tsBoEYBMKnD8PAgRIbp2X1TJZT2eQ0PD8K9XQId3t3WWQXATalTbVPaV3IdspOZxMREfPzxx5g0aZIdwiFHYvdNdVHS5+WsjsTOKABW2qyYPTWlfSXXIvufUVqtFoMGDbJHLESkAqYOvNX7veQU3MD0danYcSLbbtt2dAGyM/fV0ZrSvpLrkZ3MzJw5E++//749YiEihWuoAy9Q2YG3wiirFE8yRxYgO3tfHakp7Su5JtmnmQ4dOoTvv/8eX3/9Nbp161ajAHjTpk02C46IlEUJHXiHhwdhxfheNU6H6G18OkQJ++ooTWlfyTVZddfsMWPG2CMWIrKDsnKjzQp1ldKB1xEF0UrZV0dgYTWpnVW3MyAidUjYnoGkvZmoenZg4faTmHq3ATEjwmSvT0kdeO1dEK2kfbU3FlaT2rGNL5GLStiegVV7LBMZADAKYNWeTCRsz5C9Tmd3AHYk7uufWFhNSicpmenVqxfy8/MBAD179kSvXr3qfBCR85WVG5G0N7PeMUl7M1FWbqx3THVq6ABsK9xXWDxnYTUpmaTTTA8//DC0Wq35v9kcj0jZ/nngfI0ZmeqMonLclLs7yFq3owpwlYD7ysJqUgdJycz8+fPN/x0bG2uvWIhcgpSiRnsXPmblFdt0XHVK60hsT9xXFlaT8skuAJ48eTIiIiIwceJEi+WFhYWYNWsWPv74Y5sFR6Q2UooaHVH4GBrga9NxtVFSR2J7477aTlMqrCbHkX2jSTc3N/j4+GDKlClYunQp3Nwqy27++OMPBAcHo6Kiwi6BWos3miRHMRU1Vv8fyvRv2hXjK2vKGhpji4SmrNyILq99U++pJjcN8OubD9j9fkpEVVUYBQYv+h45BTdqrZvRoPLU1r4597rk7BdJJ+fvt1W/Ytu2bcM333yDYcOGmQuDiZoyKUWNsVvSEbvFMYWPXh5umHq3od4xU+82MJEhh2tKhdXkOFb9koWFheHgwYO4efMm+vbti5MnT9o6LiJVkVLUmFNYipxCaYWPthAzIgzT7jGg+t8ENw0w7R7r+swQ2YKp2FivszyVpNd522x2kpoW2TUzpiuZWrZsie+++w5PP/007rrrLrz99ts2D45ILWxZrGjLdcWMCMOLUV0a7ABsTUGys7u3Onv71DhNqbCa7E92MlO1xMbDwwP/+Mc/EBYWhhkzZtg0MCI1sWWxoq0LH7083Oq9/NqagmRnd2919vbJNppSYTXZl+zTTD/88AMCAiy7QEZHR+Obb77B66+/brPAiNSknyEAzX096x3T3McDzX0aGOPr6dCOstZ0YnV291Znb5+IlEd2MhMREQEPj5oTOoGBgSwGJqqHAGpWPFbjyAl2azqxOrt7q7O3T0TK1KhLGQoLC7Fq1Sr069cPPXr0wO7du20UFpG6HMrMw9Xim/WOKSgpb3BMfvFNmxUAN0ROJ9bGvMeWnL19IlImq5KZlJQUTJgwAUFBQZgxYwbuvfdenD59GmlpaTYOj8h5KowCB85dwVdpl3Dg3JV6/7Wv1AJgW2yn6jhnd2919vaJSJkkFwBnZ2dj9erV+Pjjj3H9+nWMHTsWKSkpGDBgACZMmICOHTvaM04ih5JbYKrkAuDGbqfqOGd3b3X29olImSTPzBgMBpw8eRLLly/HpUuXkJiYiD59+tgzNiKnsKbAtJ8hAEE67zprXjQA9P5a6P3rHxOk83ZYAbCUmKvHY817bMnZ2yciZZKczISGhmLfvn3Ys2cPTp8+bc+YiJzG2gJTKV1NYx/qhtiHlNP51JpOrM7u3urs7RORMklOZk6dOoV169YhOzsbffv2Re/evfHOO+8A+LORHpHaNabAVEpXU6V1PrUmHmfvg7O3T0TKI/tGkwBw7do1/Otf/8LHH3+Mn376CRERERg3bhxGjx6N1q1b2yNOq/FGk43jql1W69qvr9IuYeaGtAbf/+7f7sTDd7aVtW65YxyJHYCJSGnk/P22Kpmp6uTJk/joo4/wz3/+E3l5ebh5s/5LTx2NyYz1XLXLan37pfPxwtikgw2u419T72LnUiIiO7L7XbOr6tq1K95++21cunQJGzdubOzqSCFctctqQ/uVf72UBaZERCrT6GTGxMPDA2PGjLHV6siJXLXLqpT9enPbSbw2kgWmRERqYrNkhlyHq3ZZlbpfLZp5scCUiEhFZN81m5TJlsWQSu6y2pjiWjn79fCdbREZpmeBKRGRCigmmUlISMArr7yCmTNnYunSpQAAIQTi4uLw4YcfIj8/H/3798fy5cvRrVs35warMLYu1FVql1Up+1nfGLn75e6mYZEvEZEKKOI00+HDh/Hhhx+ie/fuFssXL16MxMRELFu2DIcPH4Zer0dkZCSKioqcFKny2KNQV4ldVqXsZ8PFvWVo7utZ73aa+3qyuJeISGUkzczIKezdtGmTrACuXbuGJ554AklJSViwYIF5uRACS5cuxbx588zbX7NmDQIDA7F+/XpMmzZN1nZcUUMFrRpUFupGhullnR4xdVmdvi4VGsBi/c4ogpW6n0KIese88XV6g9viSSQiIvWRNDOj0+nMD39/f+zatQs///yz+fUjR45g165d0Ol0sgN45plnMHLkSNx///0WyzMzM5GTk4OoqCjzMq1Wi4iICOzfv7/O9ZWWlqKwsNDi4arsWairpC6rUvczp7C03jE5haW4Wlx/H6T84puqK2wmImrqJM3MrF692vzfc+bMwV//+lesXLkS7u7uAICKigrMmDFDdlO6DRs2IDU1FYcPH67xWk5ODgAgMDDQYnlgYCCysrLqXGdCQgLi4uJkxaFW9i7UHR4eZHURrC274Dq60Ni0PXaYJSJSB9kFwB9//DH27dtnTmQAwN3dHdHR0Rg4cCCWLFkiaT0XL17EzJkzsXPnTnh7112YWf2+T0KIeu8FFRMTg+joaPPzwsJChISESIpJbRxRqGtNEWxjC3Wrz/o4utC4jZ+3y3Y/JiJyRbILgMvLy3Hy5Mkay0+ePAmj0Sh5PUeOHEFubi569+4NDw8PeHh4ICUlBe+99x48PDzMMzKmGRqT3NzcGrM1VWm1Wvj7+1s8XJWrFupWL1qWup86n/pz8+Y+HtD7N7ye/OulLtn9mIjIVclOZp566ilMnjwZb7/9Nvbt24d9+/bh7bffxv/8z//gqaeekrye++67D8ePH0daWpr50adPHzzxxBNIS0tDhw4doNfrkZycbH5PWVkZUlJSMHDgQLlhuyRToS6gjG61UjsHx25Jl9VdWMp+vjYyrMG7t2s0Grz+YNcG1/PmtpMu1/2YiMiVyT7N9Pbbb0Ov1+Odd95Bdnblv1CDgoIwe/ZsvPjii5LX4+fnh/DwcItlzZo1Q8uWLc3LZ82ahfj4eHTq1AmdOnVCfHw8fH19MW7cOLlhuyxToW71UyJ6J5wSkVqoW5+qRctVT281tJ86Hy9Jxb0tmmkbXI/Uomr2oCEiUgbZyYybmxtmz56N2bNnm68UstepnNmzZ6OkpAQzZswwN83buXMn/Pz87LI9tWpMoa4t2bJQt7Z11befX6Vdkrze+rr7ylkPEREpg1UdgMvLy7F7926cO3fOPEty+fJl+Pv745ZbbrE6mN27d1s812g0iI2NRWxsrNXrbCqU0K3WloW6da2rrv20VXdfpXY/JiKiuslOZrKysjB8+HBcuHABpaWliIyMhJ+fHxYvXowbN25g5cqV9oiTVMBUqJtTcKPWmhMNKk/nCCHwR2FpvWPkFi1L3XZD67XVeoiIyHFkFwDPnDkTffr0QX5+Pnx8fMzLH3nkEezatcumwZG6SC1Ijn2oW4Nj5J4is1UxtNKKqomIqGGyk5l9+/bh1VdfhZeXl8Xy0NBQXLokrd6AXJeUzsH26i5sq/UqqfsxERE1TPZpJqPRiIqKihrLf//9dxbmEgBpBcn2Klq21XqVUlRNREQN0wghZDXMePzxx6HT6fDhhx/Cz88Px44dQ+vWrfHwww+jXbt2Frc+UILCwkLodDoUFBS4dAM9IiIiVyLn77fsZOby5csYOnQo3N3dcebMGfTp0wdnzpxBq1atsGfPHrRp06ZRwdsakxkiIiL1kfP3W/ZppuDgYKSlpWHDhg04cuQIjEYjpkyZgieeeMKiIJiIiIjIEWTPzOzZswcDBw6Eh4dlHlReXo79+/fjnnvusWmAjcWZGduz5R2xnUXp8RERNXV2nZkZOnQosrOza5xOKigowNChQ2stDibXYes7YjuD0uMjIiJ5ZF+aLYSo9YZ+V65cQbNmzWwSFCmTPe6I7WhKj4+IiOSTPDMzZswYAJW3GJg0aRK0Wq35tYqKChw7dox3s3ZhDd0RW4PKu0kLIRocExmmd8opHan74Kz4iIjIOpKTGZ1OB6ByZsbPz8+i2NfLywt33XUXpk6davsISRHseUdsR5G6D7wjNhGRukhOZkz9Y9q3b4+XXnqJp5SaGHvfEduW6irulbpd3hGbiEhdZBcAz58/3x5xkMI54o7YtlBfcS/viE1E5JokJTM9e/astei3NqmpqY0KiJTJmXfElspU3Ft926bi3uXjevGO2ERELkhSMjN69Gg7h0FKZ7qb9PR1qdAAFslA1btJA2hwjD2Ka6UU9765LQOvjeyKZ9YfdXh8RERkP7Kb5qkNm+bZllL7zBw4dwVjkw42OO5fU+9CQUkZ+8wQESmcXZvmkfM5s3ut1Dti39slEP88cB5ZecUIDfDFkwPaw8tDdlsjyeQU9z58Z1veEZuIyIXITmYqKirwzjvv4LPPPsOFCxdQVlZm8XpeXp7NgqOalNC91t1NU++ly7XF+I99mXaNUW5xb0P7QERE6iH7n8pxcXFITEzEX//6VxQUFCA6OhpjxoyBm5sbYmNj7RAimaihe62zYjQVKNc1t6JBZdLH4l4iItcjO5n59NNPkZSUhJdeegkeHh4YO3Ys/vGPf+D111/HwYMN1yyQdRoqcAUqu9dWGJ1XAuXMGE0FygBqJDQs7iUicm2yk5mcnBzccccdAIBbbrkFBQUFAIAHH3wQ27Zts210ZCane62zODvG4eFBWDG+F/Q6y1NOep03VozvxeJeIiIXJbtm5tZbb0V2djbatWuHjh07YufOnejVqxcOHz5scb8msi01dK9VQoxSCpSJiMi1yE5mHnnkEezatQv9+/fHzJkzMXbsWHz00Ue4cOECXnjhBXvESJBf4OoMSomRxb1ERE2L7GTmrbfeMv/3o48+iltvvRX79+9Hx44d8dBDD9k0OPqT1A68zixwVUOMRETkehrd+OOuu+5CdHQ0Exk7U0OBqxpiJCIi1yO7A/DatWvrfX3ChAmNCsjWXK0DsBL6zDREDTESEZGyyfn7LTuZadGihcXzmzdvori4GF5eXvD19VVc0zxXS2YA53YAlkoNMRIRkXLZ9XYG+fn5NZadOXMG06dPx8svvyx3dWQFNRS4qiFGIiJyDTa5WU6nTp3w1ltvYebMmbZYHREREZFkNrvzn7u7Oy5fvmyr1RERERFJIvs005YtWyyeCyGQnZ2NZcuWYdCgQTYLjIiIiEgK2cnM6NGjLZ5rNBq0bt0a9957L/7+97/bKi4iIiIiSWQnM0aj0R5xEBEREVnFZjUzRERERM4gO5l59NFHLW5pYLJkyRI89thjNgmKiIiISCrZyUxKSgpGjhxZY/nw4cOxZ88emwRFREREJJXsmplr167By8urxnJPT08UFhbaJCiqn7O76zp7+0RERFXJnpkJDw/Hxo0bayzfsGEDwsLCZK1rxYoV6N69O/z9/eHv748BAwbgm2++Mb8uhEBsbCyCg4Ph4+ODIUOGID09XW7ILmXHiWwMXvQ9xiYdxMwNaRibdBCDF32PHSeym8T2iYiIqpN9b6YtW7bgL3/5C8aNG4d7770XALBr1y7861//wueff17j0u36bN26Fe7u7ujYsSMAYM2aNViyZAmOHj2Kbt26YdGiRVi4cCE++eQTdO7cGQsWLMCePXtw6tQp+Pn5SdqGK92baceJbExfl4rqH5hpTmTF+F52vZGjs7dPRERNh11vNAkA27ZtQ3x8PNLS0uDj44Pu3btj/vz5iIiIsDpok4CAACxZsgSTJ09GcHAwZs2ahTlz5gAASktLERgYiEWLFmHatGmS1ucqyUyFUWDwou8t7kRdlQaAXueNfXPutcspH2dvn4iImha73mgSAEaOHFlrEXBjVFRU4PPPP8f169cxYMAAZGZmIicnB1FRUeYxWq0WERER2L9/f53JTGlpKUpLS83PXaWO51BmXp2JBAAIANkFN3AoM88uN3h09vaJiIjqYlUyY0vHjx/HgAEDcOPGDdxyyy3YvHkzwsLCsH//fgBAYGCgxfjAwEBkZWXVub6EhATExcXZNWZ7qqu4Nreo7kSiKtM4Wxfpyt0+ERGRo0hKZlq0aAGNRtofwry8PFkB3H777UhLS8PVq1fxxRdfYOLEiUhJSTG/Xn27Qoh6Y4mJiUF0dLT5eWFhIUJCQmTF5Cw7TmQjbmuGxQxIkM4b80eFoY2ft6R1tPHzrnc91ta0yNk+ERGRI0lKZpYuXWq3ALy8vMwFwH369MHhw4fx7rvvmutkcnJyEBT05x/g3NzcGrM1VWm1Wmi1WrvFay91FdfmFNzA9HWpWD6uJ4J03sgpuFFjDPBnzUr+9TI8s77u9VhbpNvPECBp+/0MAbLXTURE1BiSkpmJEyfaOw4zIQRKS0thMBig1+uRnJyMnj17AgDKysqQkpKCRYsWOSweR6gwCsRtzag1SRCoTBTe3HYSr40MwzPrU6H573IT0zzVayO74s1t9a8nbmsGIsP0sk85ubtpMH9UGKavq3v780eFsfiXiIgcrlH3ZiopKUFhYaHFQ45XXnkFe/fuxfnz53H8+HHMmzcPu3fvxhNPPAGNRoNZs2YhPj4emzdvxokTJzBp0iT4+vpi3LhxjQlbcaQW17Zo5oUV43tBr7M8laPXeWPF+F5o0UwruUjXGsPDg+rdPi/LJiIiZ5BdAHz9+nXMmTMHn332Ga5cuVLj9YqKCsnr+uOPP/Dkk08iOzsbOp0O3bt3x44dOxAZGQkAmD17NkpKSjBjxgzk5+ejf//+2Llzp+QeM2ohtWj2mxPZeCA8CCkvD8WRrPwaxb1fpV2y6fZqMzw8CJFhenYAJiIixZCdzMyePRs//PADPvjgA0yYMAHLly/HpUuXsGrVqlpvQFmfjz76qN7XNRoNYmNjERsbKzdMVZFaNLv2QBbWHsgyF/M+fGdbq9bT2CJddzcNL78mIiLFkH2aaevWrfjggw/w6KOPwsPDA3fffTdeffVVxMfH49NPP7VHjC7PVFwrdW7DVMxb/RYCDa1Hg8qrmlikS0RErkR2MpOXlweDwQAA8Pf3N1+KPXjwYN4120qm4loAkhIaU/Ft3NYMVBj/LMWtbz0s0iUiIlclO5np0KEDzp8/DwAICwvDZ599BqByxqZ58+a2jK1Jqau4ti51FfOySJeIiJoa2TUzTz31FH755RdEREQgJiYGI0eOxPvvv4/y8nIkJibaI8Ymo2px7TcnsrH2QN2djk1qK+ZlkS4RETUlspOZF154wfzfQ4cOxa+//oqff/4Zt912G3r06GHT4JqiqsW1UpKZuop5WaRLRERNRaPvzdSuXTu0a9fOFrFQFey4S0REJI3kmpnvv/8eYWFhtTbGKygoQLdu3bB3716bBteUsZiXiIhIGsnJzNKlSzF16lT4+/vXeE2n02HatGmsmbExFvMSERE1TCOEqO0sRg2hoaHYsWMHunbtWuvrv/76K6KionDhwgWbBthYhYWF0Ol0KCgoqDURU4MKo2AxLxERNSly/n5Lrpn5448/4OnpWfeKPDzw73//W3qUJBmLeYmIiOom+TRT27Ztcfz48TpfP3bsGIKCeNqDiIiIHEtyMjNixAi8/vrruHGjZl+TkpISzJ8/Hw8++KBNgyMiIiJqiOSamT/++AO9evWCu7s7nn32Wdx+++3QaDQ4efIkli9fjoqKCqSmpiIwMNDeMcviCjUzRERETY1damYCAwOxf/9+TJ8+HTExMTDlQBqNBsOGDcMHH3yguESGiIiIXJ+spnmhoaHYvn078vPzcfbsWQgh0KlTJ7Ro0cJe8RERERHVy6oOwC1atEDfvn1tHQsRERGRbLLvmk1ERESkJExmiIiISNUafaNJcjx2BCYiIvoTkxmV2XEiG3FbM5Bd8Ge/nyCdN+aPCuO9moiIqEniaSYV2XEiG9PXpVokMgCQU3AD09elYseJbCdFRkRE5DxMZlSiwigQtzUDtXU4NC2L25qBCqOkHohEREQug8mMShzKzKsxI1OVAJBdcAOHMvMcFxQREZECMJlRidyiuhMZa8YRERG5CiYzKtHGz9um44iIiFwFkxmV6GcIQJDOG3VdgK1B5VVN/QwBjgyLiIjI6ZjMqIS7mwbzR4UBQI2ExvR8/qgw9pshIqImh8mMigwPD8KK8b2g11meStLrvLFifC/2mSEioiaJTfOcTG433+HhQYgM07MDMBER0X8xmXEia7v5urtpMOC2lo4IkYiISPF4mslJ2M2XiIjINpjMOAG7+RIREdkOkxknYDdfIiIi22HNjI3IKeS1RzdfuYXEREREroLJjA3ILeS1dTdfawuJiYiIXAFPMzWSNYW8/QwBaO7rWe96W/h6Surmy0JiIiJq6pjMNII9C3mlvIOFxERERExmGsXaQt5DmXm4Wnyz3nVfLb7ZYAEwC4mJiIiYzDSKtYW8tioAtkchMRERkdo4NZlJSEhA37594efnhzZt2mD06NE4deqUxRghBGJjYxEcHAwfHx8MGTIE6enpTorYkrWFvLYqALZ1ITEREZEaOTWZSUlJwTPPPIODBw8iOTkZ5eXliIqKwvXr181jFi9ejMTERCxbtgyHDx+GXq9HZGQkioqKnBh5pX6GAATpvGvcxdpEg8qriqoX8lr7Plttn4iIyJU4NZnZsWMHJk2ahG7duqFHjx5YvXo1Lly4gCNHjgConJVZunQp5s2bhzFjxiA8PBxr1qxBcXEx1q9f78zQAVTeI2n+qDAAqJFQmJ7PHxVWo9+Lte+z1faJiIhciaJqZgoKCgAAAQGVMwmZmZnIyclBVFSUeYxWq0VERAT2799f6zpKS0tRWFho8bCn4eFBWDG+F/Q6y1M5ep03VozvVWefF2vfZ6/1EBERqZVimuYJIRAdHY3BgwcjPDwcAJCTkwMACAwMtBgbGBiIrKysWteTkJCAuLg4+wZbzfDwIESG6WV34LX2ffZaDxERkRopJpl59tlncezYMezbt6/GaxqN5R9lIUSNZSYxMTGIjo42Py8sLERISIhtg62Fu5sGA25r6bD32Ws9REREaqOIZOa5557Dli1bsGfPHtx6663m5Xq9HkDlDE1Q0J+nS3Jzc2vM1photVpotVr7BkxERESK4dSaGSEEnn32WWzatAnff/89DAaDxesGgwF6vR7JycnmZWVlZUhJScHAgQMdHS4REREpkFNnZp555hmsX78eX331Ffz8/Mw1MjqdDj4+PtBoNJg1axbi4+PRqVMndOrUCfHx8fD19cW4ceOcGToREREphFOTmRUrVgAAhgwZYrF89erVmDRpEgBg9uzZKCkpwYwZM5Cfn4/+/ftj586d8PPzc3C0REREpEQaIYRL34WwsLAQOp0OBQUF8Pf3d3Y4REREJIGcv9+K6jNDREREJBeTGSIiIlI1RVya3VRUGAUb2xEREdkYkxkH2XEiG3FbM5BdcMO8LEjnjfmjwnjLASIiokbgaSYH2HEiG9PXpVokMgCQU3AD09elYseJbCdFRkREpH5MZuyswigQtzUDtV0yZloWtzUDFUaXvqiMiIjIbpjM2NmhzLwaMzJVCQDZBTdwKDPPcUERERG5ECYzdpZbVHciY804IiIissRkxs7a+HnbdBwRERFZYjJjZ/0MAQjSeaOuC7A1qLyqqZ8hwJFhERERuQwmM3bm7qbB/FFhAFAjoTE9nz8qjP1miIiIrMRkxgGGhwdhxfhe0OssTyXpdd5YMb4X+8wQERE1ApvmOcjw8CBEhunZAZiIiMjGmMw4kLubBgNua+nsMIiIiFwKTzMRERGRqjGZISIiIlVjMkNERESqxmSGiIiIVI3JDBEREakakxkiIiJSNSYzREREpGpMZoiIiEjVmMwQERGRqjGZISIiIlVjMkNERESqxmSGiIiIVI3JDBEREakakxkiIiJSNSYzREREpGpMZoiIiEjVmMwQERGRqjGZISIiIlVjMkNERESqxmSGiIiIVI3JDBEREakakxkiIiJSNSYzREREpGoezg5ArSqMAocy85BbdANt/LzRzxAAdzeNs8MiIiJqcpw6M7Nnzx6MGjUKwcHB0Gg0+PLLLy1eF0IgNjYWwcHB8PHxwZAhQ5Cenu6cYKvYcSIbgxd9j7FJBzFzQxrGJh3E4EXfY8eJbGeHRkRE1OQ4NZm5fv06evTogWXLltX6+uLFi5GYmIhly5bh8OHD0Ov1iIyMRFFRkYMj/dOOE9mYvi4V2QU3LJbnFNzA9HWpTGiIiIgcTCOEEM4OAgA0Gg02b96M0aNHA6iclQkODsasWbMwZ84cAEBpaSkCAwOxaNEiTJs2TdJ6CwsLodPpUFBQAH9//0bFWGEUGLzo+xqJjHkfAOh13tg3516eciIiImoEOX+/FVsAnJmZiZycHERFRZmXabVaREREYP/+/XW+r7S0FIWFhRYPWzmUmVdnIgMAAkB2wQ0cysyz2TaJiIiofopNZnJycgAAgYGBFssDAwPNr9UmISEBOp3O/AgJCbFZTLlFdScy1owjIiKixlNsMmOi0VierhFC1FhWVUxMDAoKCsyPixcv2iyWNn7eNh1HREREjafYS7P1ej2AyhmaoKAg8/Lc3NwaszVVabVaaLVau8TUzxCAIJ03cgpuoLZCI1PNTD9DgF22T0RERDUpdmbGYDBAr9cjOTnZvKysrAwpKSkYOHCgU2Jyd9Ng/qgwAJWJS1Wm5/NHhbH4l4iIyIGcmsxcu3YNaWlpSEtLA1BZ9JuWloYLFy5Ao9Fg1qxZiI+Px+bNm3HixAlMmjQJvr6+GDdunNNiHh4ehBXje0GvszyVpNd5Y8X4XhgeHlTHO4mIiMgenHpp9u7duzF06NAayydOnIhPPvkEQgjExcVh1apVyM/PR//+/bF8+XKEh4dL3oYtL82uih2AiYiI7EfO32/F9JmxF3slM0RERGQ/LtFnhoiIiEgKJjNERESkakxmiIiISNWYzBAREZGqMZkhIiIiVWMyQ0RERKrGZIaIiIhUjckMERERqRqTGSIiIlI1xd4121ZMDY4LCwudHAkRERFJZfq7LeVGBS6fzBQVFQEAQkJCnBwJERERyVVUVASdTlfvGJe/N5PRaMTly5fh5+cHjca2N4IsLCxESEgILl68yPs+2RGPs2PwODsGj7Nj8Dg7hj2PsxACRUVFCA4Ohptb/VUxLj8z4+bmhltvvdWu2/D39+f/LA7A4+wYPM6OwePsGDzOjmGv49zQjIwJC4CJiIhI1ZjMEBERkaoxmWkErVaL+fPnQ6vVOjsUl8bj7Bg8zo7B4+wYPM6OoZTj7PIFwEREROTaODNDREREqsZkhoiIiFSNyQwRERGpGpMZIiIiUjUmM1b64IMPYDAY4O3tjd69e2Pv3r3ODknVEhIS0LdvX/j5+aFNmzYYPXo0Tp06ZTFGCIHY2FgEBwfDx8cHQ4YMQXp6upMidg0JCQnQaDSYNWuWeRmPs21cunQJ48ePR8uWLeHr64s777wTR44cMb/O49x45eXlePXVV2EwGODj44MOHTrgjTfegNFoNI/hcbbOnj17MGrUKAQHB0Oj0eDLL7+0eF3KcS0tLcVzzz2HVq1aoVmzZnjooYfw+++/2ydgQbJt2LBBeHp6iqSkJJGRkSFmzpwpmjVrJrKyspwdmmoNGzZMrF69Wpw4cUKkpaWJkSNHinbt2olr166Zx7z11lvCz89PfPHFF+L48ePi8ccfF0FBQaKwsNCJkavXoUOHRPv27UX37t3FzJkzzct5nBsvLy9PhIaGikmTJomffvpJZGZmiu+++06cPXvWPIbHufEWLFggWrZsKb7++muRmZkpPv/8c3HLLbeIpUuXmsfwOFtn+/btYt68eeKLL74QAMTmzZstXpdyXJ9++mnRtm1bkZycLFJTU8XQoUNFjx49RHl5uc3jZTJjhX79+omnn37aYlmXLl3E3LlznRSR68nNzRUAREpKihBCCKPRKPR6vXjrrbfMY27cuCF0Op1YuXKls8JUraKiItGpUyeRnJwsIiIizMkMj7NtzJkzRwwePLjO13mcbWPkyJFi8uTJFsvGjBkjxo8fL4TgcbaV6smMlON69epV4enpKTZs2GAec+nSJeHm5iZ27Nhh8xh5mkmmsrIyHDlyBFFRURbLo6KisH//fidF5XoKCgoAAAEBAQCAzMxM5OTkWBx3rVaLiIgIHncrPPPMMxg5ciTuv/9+i+U8zraxZcsW9OnTB4899hjatGmDnj17Iikpyfw6j7NtDB48GLt27cLp06cBAL/88gv27duHESNGAOBxthcpx/XIkSO4efOmxZjg4GCEh4fb5di7/I0mbe0///kPKioqEBgYaLE8MDAQOTk5TorKtQghEB0djcGDByM8PBwAzMe2tuOelZXl8BjVbMOGDUhNTcXhw4drvMbjbBu//fYbVqxYgejoaLzyyis4dOgQnn/+eWi1WkyYMIHH2UbmzJmDgoICdOnSBe7u7qioqMDChQsxduxYAPw+24uU45qTkwMvLy+0aNGixhh7/K1kMmMljUZj8VwIUWMZWefZZ5/FsWPHsG/fvhqv8bg3zsWLFzFz5kzs3LkT3t7edY7jcW4co9GIPn36ID4+HgDQs2dPpKenY8WKFZgwYYJ5HI9z42zcuBHr1q3D+vXr0a1bN6SlpWHWrFkIDg7GxIkTzeN4nO3DmuNqr2PP00wytWrVCu7u7jUyy9zc3BpZKsn33HPPYcuWLfjhhx9w6623mpfr9XoA4HFvpCNHjiA3Nxe9e/eGh4cHPDw8kJKSgvfeew8eHh7mY8nj3DhBQUEICwuzWNa1a1dcuHABAL/PtvLyyy9j7ty5+Nvf/oY77rgDTz75JF544QUkJCQA4HG2FynHVa/Xo6ysDPn5+XWOsSUmMzJ5eXmhd+/eSE5OtlienJyMgQMHOikq9RNC4Nlnn8WmTZvw/fffw2AwWLxuMBig1+stjntZWRlSUlJ43GW47777cPz4caSlpZkfffr0wRNPPIG0tDR06NCBx9kGBg0aVKO1wOnTpxEaGgqA32dbKS4uhpub5Z8xd3d386XZPM72IeW49u7dG56enhZjsrOzceLECfsce5uXFDcBpkuzP/roI5GRkSFmzZolmjVrJs6fP+/s0FRr+vTpQqfTid27d4vs7Gzzo7i42DzmrbfeEjqdTmzatEkcP35cjB07lpdY2kDVq5mE4HG2hUOHDgkPDw+xcOFCcebMGfHpp58KX19fsW7dOvMYHufGmzhxomjbtq350uxNmzaJVq1aidmzZ5vH8Dhbp6ioSBw9elQcPXpUABCJiYni6NGj5hYkUo7r008/LW699Vbx3XffidTUVHHvvffy0mylWb58uQgNDRVeXl6iV69e5kuIyToAan2sXr3aPMZoNIr58+cLvV4vtFqtuOeee8Tx48edF7SLqJ7M8DjbxtatW0V4eLjQarWiS5cu4sMPP7R4nce58QoLC8XMmTNFu3bthLe3t+jQoYOYN2+eKC0tNY/hcbbODz/8UOtv8sSJE4UQ0o5rSUmJePbZZ0VAQIDw8fERDz74oLhw4YJd4tUIIYTt53uIiIiIHIM1M0RERKRqTGaIiIhI1ZjMEBERkaoxmSEiIiJVYzJDREREqsZkhoiIiFSNyQwRERGpGpMZIiIiUjUmM0Qu6Pz589BoNEhLS5P8nk8++QTNmzd3ShynTp2CXq9HUVGRTbdvD3379sWmTZucHQYRVcFkhkihLl68iClTpiA4OBheXl4IDQ3FzJkzceXKlQbfGxISguzsbISHh0ve3uOPP47Tp083JmSrzZs3D8888wz8/PwA1J9YNW/eHJ988onFsh9++AEjRoxAy5Yt4evri7CwMLz44ou4dOmSecyqVavQo0cPNGvWDM2bN0fPnj2xaNEii/UUFhZi3rx56NKlC7y9vaHX63H//fdj06ZNMDVLf+211zB37lzzzQzr8sMPP2Do0KEICAiAr68vOnXqhIkTJ6K8vFzm0SGihjCZIVKg3377DX369MHp06fxr3/9C2fPnsXKlSuxa9cuDBgwAHl5eXW+t6ysDO7u7tDr9fDw8JC8TR8fH7Rp08YW4cvy+++/Y8uWLXjqqaesev+qVatw//33Q6/X44svvkBGRgZWrlyJgoIC/P3vfwcAfPTRR4iOjsbzzz+PX375BT/++CNmz56Na9eumddz9epVDBw4EGvXrkVMTAxSU1OxZ88ePP7445g9ezYKCgoAACNHjkRBQQG+/fbbOmNKT0/HAw88gL59+2LPnj04fvw43n//fXh6ejaYBFlLCMFEiZouu9zxiYgaZfjw4eLWW2+1uGu4EEJkZ2cLX19f8fTTT5uXhYaGijfffFNMnDhR+Pv7iwkTJojMzEwBQBw9etQ87quvvhIdO3YU3t7eYsiQIeKTTz4RAER+fr4QQojVq1cLnU5nHj9//nzRo0cPsXbtWhEaGir8/f3F448/bnFX3G+++UYMGjRI6HQ6ERAQIEaOHCnOnj1rfr22OKr7+9//Lvr06WOxrHosVel0OvMNSC9evCi8vLzErFmzah1r2reHH35YTJo0qc4YhKi8c3uzZs3EpUuXarxWVFQkbt68aX4+adIk8eSTT9a5rnfeeUe0b9++3u0JIcS+ffvEPffcI3x8fETz5s1FVFSUyMvLE0IIcePGDfHcc8+J1q1bC61WKwYNGiQOHTpkfq/pRoA7duwQvXv3Fp6enuL7778XRqNRLFq0SBgMBuHt7S26d+8uPv/88wZjIVIzzswQKUxeXh6+/fZbzJgxAz4+Phav6fV6PPHEE9i4caP5tAcALFmyBOHh4Thy5Ahee+21Gus8f/48Hn30UYwePRppaWmYNm0a5s2b12As586dw5dffomvv/4aX3/9NVJSUvDWW2+ZX79+/Tqio6Nx+PBh7Nq1C25ubnjkkUdkzT7s2bMHffr0kTy+qs8//xxlZWWYPXt2ra+bTlXp9XocPHgQWVlZtY4zGo3YsGEDnnjiCQQHB9d4/ZZbbrGY5erXrx/27t1bZ1x6vR7Z2dnYs2dPnWPS0tJw3333oVu3bjhw4AD27duHUaNGoaKiAgAwe/ZsfPHFF1izZg1SU1PRsWNHDBs2rMas3OzZs5GQkICTJ0+ie/fuePXVV7F69WqsWLEC6enpeOGFFzB+/HikpKTUGQuR6jk7myIiSwcPHhQAxObNm2t9PTExUQAQf/zxhxCicmZm9OjRFmOqz4jMmTNHhIeHW4yZN29egzMzvr6+FjMxL7/8sujfv3+dsefm5goA4vjx47XGUZsePXqIN954w2KZ1JmZ6dOnC39//zrXbXL58mVx1113CQCic+fOYuLEiWLjxo2ioqJCCCHEH3/8IQCIxMTEBtclROUsl5ubm/n91ZWXl4tJkyYJAEKv14vRo0eL999/XxQUFJjHjB07VgwaNKjW91+7dk14enqKTz/91LysrKxMBAcHi8WLFwsh/pyZ+fLLLy3e5+3tLfbv32+xvilTpoixY8dK2jciNeLMDJHKiP/OyGg0GvOyhmY2Tp06hb59+1os69evX4Pbat++vbkoFwCCgoKQm5trfn7u3DmMGzcOHTp0gL+/PwwGAwDgwoULDe/If5WUlMDb21vy+KqEEBbHoS5BQUE4cOAAjh8/jueffx43b97ExIkTMXz4cBiNxlqPaX18fHxgNBpRWlpa6+vu7u5YvXo1fv/9dyxevBjBwcFYuHAhunXrhuzsbAB/zszU5ty5c7h58yYGDRpkXubp6Yl+/frh5MmTFmOrfvYZGRm4ceMGIiMjccstt5gfa9euxblz5yTtG5EaMZkhUpiOHTtCo9EgIyOj1td//fVXtGjRAq1atTIva9asWb3rrO2Pvqhymqounp6eFs81Go3FKaRRo0bhypUrSEpKwk8//YSffvoJQGURslStWrVCfn6+xTJ/f39cu3bNfMrFpKKiAteuXYNOpwMAdO7cGQUFBeYEoSHh4eF45pln8OmnnyI5ORnJyclISUlB69at0aJFixqJQl3y8vLg6+tb4zRgdW3btsWTTz6J5cuXmxONlStXAkC9760ruartc6z62Zs+m23btiEtLc38yMjIwP/93/9J2jciNWIyQ6QwLVu2RGRkJD744AOUlJRYvJaTk4NPP/0Ujz/+uORZBADo0qULDh8+bLHs559/blScV65cwcmTJ/Hqq6/ivvvuQ9euXWskJVL07NmzRuLWpUsXVFRU4OjRoxbLU1NTUVFRgdtvvx0A8Oijj8LLywuLFy+udd1Xr16tc7thYWEAKut+3Nzc8Pjjj+PTTz/F5cuXa4y9fv26xZVCJ06cQK9evSTtn0mLFi0QFBSE69evAwC6d++OXbt21Tq2Y8eO8PLywr59+8zLbt68iZ9//hldu3atd5+0Wi0uXLiAjh07WjxCQkJkxUukJkxmiBRo2bJlKC0txbBhw7Bnzx5cvHgRO3bsQGRkJNq2bYuFCxfKWt+0adPw66+/Ys6cOTh9+jQ+++wzc68WOUlRVS1atEDLli3x4Ycf4uzZs/j+++8RHR0tez3Dhg3DgQMHLGZhwsLC8MADD2Dy5Mn47rvvkJmZie+++w5TpkzBAw88YE5EQkJC8M477+Ddd9/FlClTkJKSgqysLPz444+YNm0a3nzzTQDA9OnT8eabb+LHH39EVlYWDh48iAkTJqB169YYMGAAACA+Ph4hISHo378/1q5di4yMDJw5cwYff/wx7rzzTovLuPfu3YuoqKg692nVqlWYPn06du7ciXPnziE9PR1z5sxBeno6Ro0aBQCIiYnB4cOHMWPGDBw7dgy//vorVqxYgf/85z9o1qwZpk+fjpdffhk7duxARkYGpk6diuLiYkyZMqXO7fr5+eGll17CCy+8gDVr1uDcuXM4evQoli9fjjVr1sj+bIhUw4n1OkRUj/Pnz4tJkyYJvV4vPD09RUhIiHjuuefEf/7zH4txoaGh4p133rFYVt+l2VqtVgwZMkSsWLFCABAlJSVCiLovza7qnXfeEaGhoebnycnJomvXrkKr1Yru3buL3bt3WxQvSykALi8vF23bthU7duywWF5QUCBeeOEF8+XkHTt2FLNmzRJXr16tsY7k5GQxbNgw0aJFC+Ht7S26dOkiXnrpJXH58mUhhBD/93//J0aMGCGCgoKEl5eXCA4OFn/5y1/EsWPHLNZz9epVMXfuXNGpUyfh5eUlAgMDxf333y82b94sjEajEEKI33//XXh6eoqLFy/WuU+pqali/PjxwmAwCK1WK1q2bCnuuecesWXLFotxu3fvFgMHDhRarVY0b95cDBs2zFyQXVJSIp577jnRqlWrei/NNo03MRqN4t133xW333678PT0FK1btxbDhg0TKSkpdcZLpHYaISScOCcil7Nw4UKsXLkSFy9edHYo+OCDD/DVV1/V24hOKV5++WUUFBTgww8/dHYoRPRf0tuDEpGqffDBB+jbty9atmyJH3/8EUuWLMGzzz7r7LAAAP/7v/+L/Px8FBUVWVw9pURt2rTBSy+95OwwiKgKzswQNREvvPACNm7ciLy8PLRr1w5PPvkkYmJiZN3ygIhIiZjMEBERkarxaiYiIiJSNSYzREREpGpMZoiIiEjVmMwQERGRqjGZISIiIlVjMkNERESqxmSGiIiIVI3JDBEREana/wMR3WD3x7tZQwAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import scipy.stats\n",
    "plt.scatter(score_ucsc, scores)\n",
    "scipy.stats.spearmanr(score_ucsc, scores)\n",
    "plt.xlabel('Original (UCSC) Score')\n",
    "plt.ylabel('Calculated Azimuth Score')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Error checking\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Align import PairwiseAligner\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sensor_outcome_corrected(p53_lib, i):\n",
    "    #p53_lib = df containing pegRNA + target sequence + other info\n",
    "    #i = index\n",
    "\n",
    "    sensor_true = p53_lib.iloc[i]['wt_w_context']\n",
    "\n",
    "    #let's quantify the homology overhang as well...\n",
    "    PAM_strand = p53_lib.iloc[i]['PAM_strand']\n",
    "    var_type = p53_lib.iloc[i]['Variant_Type']\n",
    "\n",
    "\n",
    "    if PAM_strand=='+':\n",
    "        sensor = Bio.Seq.Seq(p53_lib.iloc[i]['wt_w_context']).reverse_complement()\n",
    "\n",
    "    elif PAM_strand=='-':\n",
    "        sensor = p53_lib.iloc[i]['wt_w_context']\n",
    "\n",
    "    pbs_length = p53_lib.iloc[i]['PBS_length']\n",
    "    pbs = p53_lib.iloc[i]['RTT_PBS'][-pbs_length:]\n",
    "    rtt = p53_lib.iloc[i]['RTT_PBS'][:-(pbs_length-1)] #FIXING [:-0] indexing issue...\n",
    "\n",
    "    #pbs_rtt = p53_lib.iloc[i]['PBS_RTT_5to3']\n",
    "\n",
    "    pbs_find = sensor.find(pbs)\n",
    "    #print(pbs_find)\n",
    "\n",
    "    #first compute the portion before the mutation for computing the right flank\n",
    "    for kk, val in enumerate(reversed(rtt)):\n",
    "        idx = pbs_find-kk\n",
    "        if val!= sensor[idx]: #find first mismatch\n",
    "            mut_start=kk\n",
    "            break\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "\n",
    "    try: mut_start\n",
    "    except: #case where's theres no identifiable change (because of homology overhang being too short/matching coincidentally...)\n",
    "        return sensor_true, sensor_true, 0, \"no homology overhang\"\n",
    "\n",
    "    #alignments = aligner.align(pbs, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(rtt, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #now we have the right flank...\n",
    "    right_flank = rtt[-mut_start:-1] + sensor[pbs_find:]\n",
    "\n",
    "    #alignments = aligner.align(right_flank, sensor)\n",
    "    #print(alignments[0])\n",
    "    \n",
    "    if var_type in ['SNP', 'ONP']:\n",
    "\n",
    "        matching_arr = np.zeros(len(rtt[:-mut_start]))\n",
    "\n",
    "        for kk, val in enumerate(reversed(rtt[:-mut_start])):\n",
    "            idx = pbs_find-mut_start-kk\n",
    "\n",
    "            if val== sensor[idx]: #find regions of homology\n",
    "                matching_arr[kk]+=1\n",
    "                \n",
    "            else:\n",
    "                continue\n",
    "        \n",
    "        #based on values in match array, determine where mutation vs. homology overhang is...\n",
    "        end_mut = np.where(matching_arr==0)[0][-1]\n",
    "\n",
    "        mut_allele = rtt[-(mut_start+end_mut+1):-mut_start]\n",
    "        homology_overhang = rtt[:-(mut_start+end_mut+1)]\n",
    "        #this formulation will NOT catch situations where the homology overhang doesn't match???????\n",
    "        #(though this shouldn't be an issue for SNPs/ONPs)\n",
    "\n",
    "        left_flank = sensor[:(pbs_find-(len(rtt)-1))]\n",
    "\n",
    "    elif var_type=='DEL':\n",
    "\n",
    "        #find the size of the deletion by shifting and checking for matches...\n",
    "        #will allow for detection of insufficient homology overhang...\n",
    "        length_remaining = len(rtt[:-mut_start])\n",
    "        homology_overhang = rtt[:-mut_start]\n",
    "        mut_allele = ''\n",
    "\n",
    "        start_idx = pbs_find-mut_start\n",
    "        for i in range(len(sensor[:start_idx])): #check rest of the sensor...\n",
    "        \n",
    "            #sliding window to find size of deletion/perfect match...\n",
    "            sensor_a = sensor[start_idx-i-length_remaining:start_idx-i]\n",
    "\n",
    "            if sensor_a==homology_overhang: #once match is found, record and break...\n",
    "                size_del = i\n",
    "                break\n",
    "        \n",
    "            else:\n",
    "                continue\n",
    "\n",
    "        #error where the full homology overhang region not included in sensor...\n",
    "        try: size_del\n",
    "        except: \n",
    "            return sensor_true, sensor_true, len(homology_overhang), 'sensor too short' #no excepted editing, return WT sensor\n",
    "\n",
    "        left_flank = sensor[:start_idx-size_del-length_remaining]\n",
    "\n",
    "    elif var_type=='INS':\n",
    "\n",
    "        length_remaining = len(rtt[:-mut_start])\n",
    "        start_idx = pbs_find-mut_start+1\n",
    "\n",
    "        for i in range(length_remaining):\n",
    "            rtt_region = rtt[:-(mut_start+i)]\n",
    "            sensor_a = sensor[start_idx-len(rtt_region):start_idx]\n",
    "\n",
    "            #print('rt: ' + rtt_region)\n",
    "            #print('s: ' + sensor_a)\n",
    "\n",
    "            if sensor_a==rtt_region:\n",
    "                mut_allele = rtt[-(mut_start+i):-mut_start]\n",
    "                break\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "        #in case of no homology overhang...\n",
    "        try: mut_allele\n",
    "        except: \n",
    "            return sensor_true, sensor_true, 0, 'no homology overhang' #no excepted editing, return WT sensor\n",
    "\n",
    "        homology_overhang = sensor_a\n",
    "        left_flank = sensor[:start_idx-len(rtt_region)]\n",
    "\n",
    "    edited_sensor = left_flank + homology_overhang + mut_allele + right_flank      \n",
    "\n",
    "    if PAM_strand=='+':\n",
    "        edited_sensor_true = str(Bio.Seq.Seq(edited_sensor).reverse_complement())\n",
    "\n",
    "    elif PAM_strand=='-':\n",
    "        edited_sensor_true=edited_sensor\n",
    "\n",
    "\n",
    "\n",
    "    #alignments = aligner.align(rtt[-mut_start:], sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(right_flank, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(homology_overhang, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(edited_sensor, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(edited_sensor_true, sensor_true)\n",
    "    #print(alignments[0])\n",
    "\n",
    "\n",
    "    return sensor_true, edited_sensor_true, len(homology_overhang), 'no error'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "p53_lib = df[:50000]\n",
    "\n",
    "#running it for every mutation in the p53 library\n",
    "edit_sens = []\n",
    "homo = []\n",
    "idx_holder = []\n",
    "classif = []\n",
    "#del_idx = p53_del.index\n",
    "\n",
    "#ins_idx = p53_lib[p53_lib['variant_type']=='SNP'].index\n",
    "\n",
    "error = []\n",
    "for i in range(len(p53_lib)):\n",
    "    try:\n",
    "        s, s_e, homo2, m = sensor_outcome_corrected(p53_lib, i)\n",
    "        idx_holder.append(i)\n",
    "        edit_sens.append(s_e)\n",
    "        homo.append(homo2)\n",
    "        classif.append(m)\n",
    "\n",
    "        assert s_e == p53_lib['alt_w_context']\n",
    "    \n",
    "    except:\n",
    "        error.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['no error', 'no homology overhang', 'sensor too short'],\n",
       "       dtype='<U20'),\n",
       " array([46234,  1972,    66]))"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(classif, return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19976"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "classif.index('sensor too short')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>...</th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>RTT</th>\n",
       "      <th>RTT_length</th>\n",
       "      <th>PBS</th>\n",
       "      <th>PBS_length</th>\n",
       "      <th>RTT_PBS</th>\n",
       "      <th>Distance_to_nick</th>\n",
       "      <th>RHA_size</th>\n",
       "      <th>PAM_disrupted</th>\n",
       "      <th>Proto_disrupted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19976</th>\n",
       "      <td>494</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578285</td>\n",
       "      <td>7578288</td>\n",
       "      <td>frameshift_variant,splice_region_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>CAGA</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>GTTCCTCACTGATTGCTCTT</td>\n",
       "      <td>TCGGATAAGATGCTGAGGAGGGGCCCTAAG</td>\n",
       "      <td>30</td>\n",
       "      <td>AGCAATCAGT</td>\n",
       "      <td>10</td>\n",
       "      <td>TCGGATAAGATGCTGAGGAGGGGCCCTAAGAGCAATCAGT</td>\n",
       "      <td>6</td>\n",
       "      <td>24</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation_idx Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "19976           494        TP53         17         7578285       7578288   \n",
       "\n",
       "                                    Consequence Variant_Classification  \\\n",
       "19976  frameshift_variant,splice_region_variant        Frame_Shift_Del   \n",
       "\n",
       "      Variant_Type Reference_Allele Tumor_Seq_Allele2  ...  \\\n",
       "19976          DEL             CAGA                 -  ...   \n",
       "\n",
       "                Protospacer                             RTT  RTT_length  \\\n",
       "19976  GTTCCTCACTGATTGCTCTT  TCGGATAAGATGCTGAGGAGGGGCCCTAAG          30   \n",
       "\n",
       "              PBS PBS_length                                   RTT_PBS  \\\n",
       "19976  AGCAATCAGT         10  TCGGATAAGATGCTGAGGAGGGGCCCTAAGAGCAATCAGT   \n",
       "\n",
       "      Distance_to_nick RHA_size PAM_disrupted Proto_disrupted  \n",
       "19976                6       24         False           False  \n",
       "\n",
       "[1 rows x 34 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[[19976]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
