{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PEGG 2.0\n",
    "\n",
    "This is the real notebook for making the next generation of PEGG...\n",
    "\n",
    "- Start with the flexible input + getting it to work WITH error checking, etc...\n",
    "- I should translate everything into the easiest possible format to work with that is the most flexible...\n",
    "\n",
    "\n",
    "- To do:\n",
    "    - Get the flexible input format in PrimeDesign format working as well!\n",
    "    - Have it detect what the input format is??\n",
    "    - Or not?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt \n",
    "import Bio.Seq\n",
    "import gzip\n",
    "from Bio import SeqIO\n",
    "from Bio.Align import PairwiseAligner\n",
    "from Bio.pairwise2 import format_alignment\n",
    "import warnings\n",
    "import regex as re\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "MATCH_SCORE = 1\n",
    "MISMATCH_SCORE = -0.5\n",
    "OPEN_GAP_SCORE = -5\n",
    "EXTEND_GAP_SCORE = -0.1\n",
    "TARGET_END_GAP_SCORE = 0\n",
    "QUERY_END_GAP_SCORE = 0\n",
    "\n",
    "def make_aligner():\n",
    "    aligner = PairwiseAligner()\n",
    "    aligner.mode = 'global'\n",
    "    aligner.match_score = MATCH_SCORE\n",
    "    aligner.mismatch_score = MISMATCH_SCORE\n",
    "    aligner.open_gap_score = OPEN_GAP_SCORE\n",
    "    aligner.extend_gap_score = EXTEND_GAP_SCORE\n",
    "    aligner.target_end_gap_score = TARGET_END_GAP_SCORE\n",
    "    aligner.query_end_gap_score = QUERY_END_GAP_SCORE\n",
    "    return aligner\n",
    "\n",
    "aligner = make_aligner()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class mutation:\n",
    "    def __init__(self, wt_w_context, alt_w_context, left_seq, right_seq, var_type, ref_seq, alt_seq, chrom=None, genome=None):\n",
    "        \"\"\" \n",
    "        change seq_start and seq_end to correspond with chromosome coordinates\n",
    "        or some other organizational method...\n",
    "        \"\"\"\n",
    "        \n",
    "        #WT sequence\n",
    "        self.wt_forward = wt_w_context.upper()\n",
    "        rc = str(Bio.Seq.Seq(self.wt_forward).reverse_complement())\n",
    "        self.wt_rc = rc\n",
    "\n",
    "        #ALT sequence\n",
    "        self.mut_forward = alt_w_context.upper()\n",
    "        rc_mut = str(Bio.Seq.Seq(self.mut_forward).reverse_complement())\n",
    "        self.mut_rc = rc_mut\n",
    "\n",
    "        #seq start refers to the WIDE sequence start (i.e. the WT sequence with e.g. 60 nt of context)\n",
    "        self.seq_start = 0\n",
    "        self.seq_end = len(self.wt_forward)\n",
    "        self.chrom = chrom\n",
    "        self.genome_build = genome\n",
    "\n",
    "        #and splitting it up into the constituent components\n",
    "        self.left_seq = left_seq.upper()\n",
    "        self.right_seq = right_seq.upper()\n",
    "        self.ref_seq = ref_seq.upper()\n",
    "        self.alt_seq = alt_seq.upper()\n",
    "\n",
    "        #and do the same to get it into the RC format\n",
    "\n",
    "        #RIGHT/LEFT ARE SWAPPED ON REVERSE COMPLEMENT\n",
    "        self.left_seq_rc = str(Bio.Seq.Seq(right_seq).reverse_complement())\n",
    "        self.right_seq_rc = str(Bio.Seq.Seq(left_seq).reverse_complement())\n",
    "        self.ref_seq_rc = str(Bio.Seq.Seq(ref_seq).reverse_complement())\n",
    "        self.alt_seq_rc = str(Bio.Seq.Seq(alt_seq).reverse_complement())\n",
    "\n",
    "        self.variant_type = var_type\n",
    "        #self.alt_size = (positive or negative integer)\n",
    "        \n",
    "        #information about the PAM sequence location\n",
    "        self.PAM_idx_forward = None\n",
    "        self.PAM_idx_rc = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAAACCCTCTCTC\n",
      "GAGAGAGGGTTTT\n",
      "GAGAGAGGGTTTT\n"
     ]
    }
   ],
   "source": [
    "left_seq = 'AAAA'\n",
    "ref_seq = 'TGT'\n",
    "alt_seq = 'CCC'\n",
    "right_seq = 'TCTCTC'\n",
    "\n",
    "wt_w_context = left_seq + ref_seq + right_seq\n",
    "alt_w_context = left_seq + alt_seq + right_seq\n",
    "print(alt_w_context)\n",
    "print(Bio.Seq.Seq(alt_w_context).reverse_complement())\n",
    "\n",
    "print(Bio.Seq.Seq(right_seq).reverse_complement() + Bio.Seq.Seq(alt_seq).reverse_complement() + Bio.Seq.Seq(left_seq).reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------functions------------\n",
    "def genome_loader(filepath_gz):\n",
    "    \"\"\"\n",
    "    Takes in filepath of human genome (GrCH37 or GrCh38) and returns records and index_list for PEGG parsing.\n",
    "    \n",
    "    Parameters\n",
    "    -----------\n",
    "    filepath_gz\n",
    "        *type = str*\n",
    "        \n",
    "        The filepath to the .gz file holding the reference genome file.\n",
    "    \n",
    "    \"\"\"\n",
    "    #------loading in reference genome and organizing it into a 2-d list by chromosome---------------------\n",
    "    wrong = [\"alternate\", \"unplaced\", \"unlocalized\", \"patch\", \"mitochondrion\"]\n",
    "    filtered = []\n",
    "    chrom_list = []\n",
    "    seq_list = []\n",
    "\n",
    "    with gzip.open(filepath_gz, \"rt\") as handle:\n",
    "        #records = list(SeqIO.parse(handle, \"fasta\")) #about 4 Gb in  memory\n",
    "\n",
    "        for i in SeqIO.parse(handle, \"fasta\"):\n",
    "            ii = i.description\n",
    "            ignore=False \n",
    "\n",
    "            for key in wrong:\n",
    "                if key in ii:\n",
    "                    ignore = True\n",
    "            \n",
    "            if ignore==False:\n",
    "\n",
    "                x = ii.find('chromosome')\n",
    "                chrom = ii[x:].split(' ')[1]\n",
    "                chrom = chrom.split(',')[0]\n",
    "                if chrom not in ['X', 'Y']:\n",
    "                    chrom = int(chrom)\n",
    "                chrom_list.append(chrom)\n",
    "                seq_list.append(i.seq)\n",
    "\n",
    "                filtered.append(ii)\n",
    "\n",
    "  \n",
    "    return dict(zip(chrom_list, seq_list)), filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath_37 = '/Users/samgould/Desktop/FSR Lab/reference files/GRCh37/ncbi-genomes-2022-03-17/GCF_000001405.25_GRCh37.p13_genomic.fna.gz'\n",
    "filepath_38 = '/Users/samgould/Desktop/FSR Lab/reference files/GRCh38/ncbi-genomes-2022-02-23/GCF_000001405.26_GRCh38_genomic.fna.gz'\n",
    "mouse = '/Users/samgould/Desktop/FSR Lab/reference files/GRCm38.p6 (mouse)/ncbi-genomes-2022-04-04/GCF_000001635.26_GRCm38.p6_genomic.fna.gz'\n",
    "chrom_dict, i = genome_loader(filepath_37)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#starting by bringing in the code for extracting sequence\n",
    "\n",
    "#need to modify this so that it works more generally (the chromosome shit in particular...)\n",
    "\n",
    "def df_formatter(df, chrom_dict, context_size = 60, \n",
    "                 cols_to_save = ['Hugo_Symbol', 'Chromosome', 'Start_Position', 'End_Position', 'Consequence', 'Variant_Classification', 'Variant_Type', 'Reference_Allele', 'Tumor_Seq_Allele2',  'HGVSc', 'HGVSp_Short']):\n",
    "\n",
    "    \"\"\" \n",
    "    Takes in variants (in cBioPortal format!)\n",
    "    and outputs dataframe with WT and ALT oligos with designated context_size\n",
    "    context_size = the amount of nt on either side of the variant e.g. AAA(A/G)AAA = context_size of 3\n",
    "    cols_to_save = information to save about the mutations...(can modify beyond defaults...)\n",
    "    \"\"\"\n",
    "\n",
    "    wt_w_context = []\n",
    "    alt_w_context = []\n",
    "    left_context_list = []\n",
    "    right_context_list = []\n",
    "    ref_allele = []\n",
    "    alt_allele = []\n",
    "\n",
    "    seq_start = []\n",
    "    seq_end = []\n",
    "\n",
    "    for i, val in df.iterrows():\n",
    "        vt = val['Variant_Type']\n",
    "        s = val['Start_Position']\n",
    "        e = val['End_Position']\n",
    "        ref = val['Reference_Allele']\n",
    "        alt = val['Tumor_Seq_Allele2']\n",
    "        chrom = val['Chromosome']\n",
    "\n",
    "        if chrom not in ['X','Y']:\n",
    "            chrom = int(chrom)\n",
    "       \n",
    "        chr_seq = chrom_dict[chrom].upper()\n",
    "\n",
    "        if vt in ['SNP', 'ONP', 'DNP']:\n",
    "            ref = ref\n",
    "            alt = alt\n",
    "            #assert ref == chr_seq[s-1:e], print(ref, chr_seq[s-1:e])\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        elif vt =='INS':\n",
    "            ref = ''\n",
    "            alt = alt\n",
    "            #left_context = chr_seq[s-1-context_size:s+1] #need to do this since INS reference alleles are blank\n",
    "            left_context = chr_seq[s-1-context_size:s]\n",
    "            right_context = chr_seq[e-1:e+context_size]\n",
    "\n",
    "        elif vt=='DEL':\n",
    "            ref = ref\n",
    "            alt = ''\n",
    "            left_context = chr_seq[s-1-context_size:s-1]\n",
    "            right_context = chr_seq[e:e+context_size]\n",
    "\n",
    "        wt_seq = left_context + ref + right_context\n",
    "        alt_seq = left_context + alt + right_context\n",
    "\n",
    "        wt_w_context.append(str(wt_seq))\n",
    "        alt_w_context.append(str(alt_seq))\n",
    "        left_context_list.append(str(left_context))\n",
    "        right_context_list.append(str(right_context))\n",
    "        ref_allele.append(str(ref))\n",
    "        alt_allele.append(str(alt))\n",
    "\n",
    "        start = s-context_size\n",
    "        end = e+context_size\n",
    "\n",
    "        seq_start.append(start)\n",
    "        seq_end.append(end)\n",
    "\n",
    "        assert str(chr_seq[start-1:end])==str(wt_seq), print(chr_seq[start-1:end] + '\\n' + str(wt_seq))\n",
    "                                                            \n",
    "\n",
    "    df_new = df[cols_to_save]\n",
    "\n",
    "    df_new['seq_start'] = seq_start\n",
    "    df_new['seq_end'] = seq_end\n",
    "    df_new['wt_w_context'] = wt_w_context\n",
    "    df_new['alt_w_context'] = alt_w_context\n",
    "    df_new['left_context'] = left_context_list\n",
    "    df_new['right_context'] = right_context_list\n",
    "    df_new['REF'] = ref_allele\n",
    "    df_new['ALT'] = alt_allele\n",
    "    df_new = df_new.reset_index()\n",
    "    \n",
    "    return df_new\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "impact = pd.read_csv('/Users/samgould/Desktop/FSR Lab/reference files/2020-06-16-MSK-IMPACT_EDITED.txt', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [],
   "source": [
    "p53 = impact[impact['Hugo_Symbol']=='TP53'].drop_duplicates(subset='HGVSc')\n",
    "\n",
    "#or change these to INDELs\n",
    "dels1 = p53[p53['Variant_Type']=='DEL']\n",
    "bad_del_idx = list(dels1[dels1['Tumor_Seq_Allele2']!='-'].index)\n",
    "\n",
    "\n",
    "ins1 = p53[p53['Variant_Type']=='INS']\n",
    "bad_ins_idx = list(ins1[ins1['Reference_Allele']!='-'].index)\n",
    "\n",
    "bad_idxs = bad_del_idx + bad_ins_idx\n",
    "\n",
    "p53_filtered = p53.drop(index=bad_idxs).reset_index().drop(columns=['index'])\n",
    "\n",
    "p53_formatted = df_formatter(p53_filtered, chrom_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "      <th>left_context</th>\n",
       "      <th>right_context</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.818G&gt;A</td>\n",
       "      <td>p.R273H</td>\n",
       "      <td>7577060</td>\n",
       "      <td>7577180</td>\n",
       "      <td>CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAG...</td>\n",
       "      <td>CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAG...</td>\n",
       "      <td>CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAG...</td>\n",
       "      <td>GCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGAT...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577108</td>\n",
       "      <td>7577108</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.830G&gt;T</td>\n",
       "      <td>p.C277F</td>\n",
       "      <td>7577048</td>\n",
       "      <td>7577168</td>\n",
       "      <td>TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCC...</td>\n",
       "      <td>TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCC...</td>\n",
       "      <td>TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCC...</td>\n",
       "      <td>AGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578239</td>\n",
       "      <td>7578239</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.610G&gt;T</td>\n",
       "      <td>p.E204*</td>\n",
       "      <td>7578179</td>\n",
       "      <td>7578299</td>\n",
       "      <td>CAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTTCT...</td>\n",
       "      <td>CAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTTCT...</td>\n",
       "      <td>CAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTTCT...</td>\n",
       "      <td>CACACGCAAATTTCCTTCCACTCGGATAAGATGCTGAGGAGGGGCC...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577121</td>\n",
       "      <td>7577121</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.817C&gt;T</td>\n",
       "      <td>p.R273C</td>\n",
       "      <td>7577061</td>\n",
       "      <td>7577181</td>\n",
       "      <td>CTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGG...</td>\n",
       "      <td>CTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGG...</td>\n",
       "      <td>CTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGG...</td>\n",
       "      <td>CACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATA...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577506</td>\n",
       "      <td>7577506</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.775G&gt;T</td>\n",
       "      <td>p.D259Y</td>\n",
       "      <td>7577446</td>\n",
       "      <td>7577566</td>\n",
       "      <td>GCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGCAAGTGGC...</td>\n",
       "      <td>GCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGCAAGTGGC...</td>\n",
       "      <td>GCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGCAAGTGGC...</td>\n",
       "      <td>TTCCAGTGTGATGATGGTGAGGATGGGCCTCCGGTTCATGCCGCCC...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3000</th>\n",
       "      <td>3000</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577539</td>\n",
       "      <td>7577541</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GGT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.740_742del</td>\n",
       "      <td>p.N247del</td>\n",
       "      <td>7577479</td>\n",
       "      <td>7577601</td>\n",
       "      <td>GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...</td>\n",
       "      <td>GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...</td>\n",
       "      <td>GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...</td>\n",
       "      <td>TCATGCCGCCCATGCAGGAACTGTTACACATGTAGTTGTAGTGGAT...</td>\n",
       "      <td>GGT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3001</th>\n",
       "      <td>3001</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578438</td>\n",
       "      <td>7578438</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>ENST00000269305.4:c.492G&gt;C</td>\n",
       "      <td>p.K164N</td>\n",
       "      <td>7578378</td>\n",
       "      <td>7578498</td>\n",
       "      <td>ATCTGAGCAGCGCTCATGGTGGGGGCAGCGCCTCACAACCTCCGTC...</td>\n",
       "      <td>ATCTGAGCAGCGCTCATGGTGGGGGCAGCGCCTCACAACCTCCGTC...</td>\n",
       "      <td>ATCTGAGCAGCGCTCATGGTGGGGGCAGCGCCTCACAACCTCCGTC...</td>\n",
       "      <td>TTGTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGG...</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3002</th>\n",
       "      <td>3002</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577531</td>\n",
       "      <td>7577532</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GG</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.749_750delCC</td>\n",
       "      <td>p.P250Hfs*13</td>\n",
       "      <td>7577471</td>\n",
       "      <td>7577592</td>\n",
       "      <td>GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...</td>\n",
       "      <td>GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...</td>\n",
       "      <td>GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...</td>\n",
       "      <td>GCCTCCGGTTCATGCCGCCCATGCAGGAACTGTTACACATGTAGTT...</td>\n",
       "      <td>GG</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3003</th>\n",
       "      <td>3003</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578421</td>\n",
       "      <td>7578440</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GTCATGTGCTGTGACTGCTT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.490_509del</td>\n",
       "      <td>p.K164Gfs*10</td>\n",
       "      <td>7578361</td>\n",
       "      <td>7578500</td>\n",
       "      <td>AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...</td>\n",
       "      <td>AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...</td>\n",
       "      <td>AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...</td>\n",
       "      <td>GTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGGAA...</td>\n",
       "      <td>GTCATGTGCTGTGACTGCTT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3004</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.339C&gt;A</td>\n",
       "      <td>p.F113L</td>\n",
       "      <td>7579288</td>\n",
       "      <td>7579408</td>\n",
       "      <td>AGCCAGCCCCTCAGGGCAACTGACCGTGCAAGTCACAGACTTGGCT...</td>\n",
       "      <td>AGCCAGCCCCTCAGGGCAACTGACCGTGCAAGTCACAGACTTGGCT...</td>\n",
       "      <td>AGCCAGCCCCTCAGGGCAACTGACCGTGCAAGTCACAGACTTGGCT...</td>\n",
       "      <td>AAGCCCAGACGGAAACCGTAGCTGCCCTGGTAGGTTTTCTGGGAAG...</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3005 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0         0        TP53         17         7577120       7577120   \n",
       "1         1        TP53         17         7577108       7577108   \n",
       "2         2        TP53         17         7578239       7578239   \n",
       "3         3        TP53         17         7577121       7577121   \n",
       "4         4        TP53         17         7577506       7577506   \n",
       "...     ...         ...        ...             ...           ...   \n",
       "3000   3000        TP53         17         7577539       7577541   \n",
       "3001   3001        TP53         17         7578438       7578438   \n",
       "3002   3002        TP53         17         7577531       7577532   \n",
       "3003   3003        TP53         17         7578421       7578440   \n",
       "3004   3004        TP53         17         7579348       7579348   \n",
       "\n",
       "             Consequence Variant_Classification Variant_Type  \\\n",
       "0       missense_variant      Missense_Mutation          SNP   \n",
       "1       missense_variant      Missense_Mutation          SNP   \n",
       "2            stop_gained      Nonsense_Mutation          SNP   \n",
       "3       missense_variant      Missense_Mutation          SNP   \n",
       "4       missense_variant      Missense_Mutation          SNP   \n",
       "...                  ...                    ...          ...   \n",
       "3000    inframe_deletion           In_Frame_Del          DEL   \n",
       "3001    missense_variant      Missense_Mutation          SNP   \n",
       "3002  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "3003  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "3004    missense_variant      Missense_Mutation          SNP   \n",
       "\n",
       "          Reference_Allele Tumor_Seq_Allele2  \\\n",
       "0                        C                 T   \n",
       "1                        C                 A   \n",
       "2                        C                 A   \n",
       "3                        G                 A   \n",
       "4                        C                 A   \n",
       "...                    ...               ...   \n",
       "3000                   GGT                 -   \n",
       "3001                     C                 G   \n",
       "3002                    GG                 -   \n",
       "3003  GTCATGTGCTGTGACTGCTT                 -   \n",
       "3004                     G                 T   \n",
       "\n",
       "                                 HGVSc   HGVSp_Short  seq_start  seq_end  \\\n",
       "0           ENST00000269305.4:c.818G>A       p.R273H    7577060  7577180   \n",
       "1           ENST00000269305.4:c.830G>T       p.C277F    7577048  7577168   \n",
       "2           ENST00000269305.4:c.610G>T       p.E204*    7578179  7578299   \n",
       "3           ENST00000269305.4:c.817C>T       p.R273C    7577061  7577181   \n",
       "4           ENST00000269305.4:c.775G>T       p.D259Y    7577446  7577566   \n",
       "...                                ...           ...        ...      ...   \n",
       "3000    ENST00000269305.4:c.740_742del     p.N247del    7577479  7577601   \n",
       "3001        ENST00000269305.4:c.492G>C       p.K164N    7578378  7578498   \n",
       "3002  ENST00000269305.4:c.749_750delCC  p.P250Hfs*13    7577471  7577592   \n",
       "3003    ENST00000269305.4:c.490_509del  p.K164Gfs*10    7578361  7578500   \n",
       "3004        ENST00000269305.4:c.339C>A       p.F113L    7579288  7579408   \n",
       "\n",
       "                                           wt_w_context  \\\n",
       "0     CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAG...   \n",
       "1     TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCC...   \n",
       "2     CAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTTCT...   \n",
       "3     CTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGG...   \n",
       "4     GCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGCAAGTGGC...   \n",
       "...                                                 ...   \n",
       "3000  GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...   \n",
       "3001  ATCTGAGCAGCGCTCATGGTGGGGGCAGCGCCTCACAACCTCCGTC...   \n",
       "3002  GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...   \n",
       "3003  AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...   \n",
       "3004  AGCCAGCCCCTCAGGGCAACTGACCGTGCAAGTCACAGACTTGGCT...   \n",
       "\n",
       "                                          alt_w_context  \\\n",
       "0     CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAG...   \n",
       "1     TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCC...   \n",
       "2     CAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTTCT...   \n",
       "3     CTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGG...   \n",
       "4     GCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGCAAGTGGC...   \n",
       "...                                                 ...   \n",
       "3000  GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...   \n",
       "3001  ATCTGAGCAGCGCTCATGGTGGGGGCAGCGCCTCACAACCTCCGTC...   \n",
       "3002  GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...   \n",
       "3003  AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...   \n",
       "3004  AGCCAGCCCCTCAGGGCAACTGACCGTGCAAGTCACAGACTTGGCT...   \n",
       "\n",
       "                                           left_context  \\\n",
       "0     CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAG...   \n",
       "1     TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCC...   \n",
       "2     CAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTTCT...   \n",
       "3     CTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGG...   \n",
       "4     GCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGCAAGTGGC...   \n",
       "...                                                 ...   \n",
       "3000  GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...   \n",
       "3001  ATCTGAGCAGCGCTCATGGTGGGGGCAGCGCCTCACAACCTCCGTC...   \n",
       "3002  GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...   \n",
       "3003  AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...   \n",
       "3004  AGCCAGCCCCTCAGGGCAACTGACCGTGCAAGTCACAGACTTGGCT...   \n",
       "\n",
       "                                          right_context                   REF  \\\n",
       "0     GCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGAT...                     C   \n",
       "1     AGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACC...                     C   \n",
       "2     CACACGCAAATTTCCTTCCACTCGGATAAGATGCTGAGGAGGGGCC...                     C   \n",
       "3     CACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATA...                     G   \n",
       "4     TTCCAGTGTGATGATGGTGAGGATGGGCCTCCGGTTCATGCCGCCC...                     C   \n",
       "...                                                 ...                   ...   \n",
       "3000  TCATGCCGCCCATGCAGGAACTGTTACACATGTAGTTGTAGTGGAT...                   GGT   \n",
       "3001  TTGTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGG...                     C   \n",
       "3002  GCCTCCGGTTCATGCCGCCCATGCAGGAACTGTTACACATGTAGTT...                    GG   \n",
       "3003  GTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGGAA...  GTCATGTGCTGTGACTGCTT   \n",
       "3004  AAGCCCAGACGGAAACCGTAGCTGCCCTGGTAGGTTTTCTGGGAAG...                     G   \n",
       "\n",
       "     ALT  \n",
       "0      T  \n",
       "1      A  \n",
       "2      A  \n",
       "3      A  \n",
       "4      A  \n",
       "...   ..  \n",
       "3000      \n",
       "3001   G  \n",
       "3002      \n",
       "3003      \n",
       "3004   T  \n",
       "\n",
       "[3005 rows x 20 columns]"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p53_formatted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Flexible input format\n",
    "\n",
    "- Get this working so that you can also put in sequences in PrimeDesign format AND in for just inputting the ref and mutant seq\n",
    "\n",
    "- Need to get out the indeces as well? or can just count the size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "      <th>left_context</th>\n",
       "      <th>right_context</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>257</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578398</td>\n",
       "      <td>7578398</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.532delC</td>\n",
       "      <td>p.H178Tfs*69</td>\n",
       "      <td>7578338</td>\n",
       "      <td>7578458</td>\n",
       "      <td>AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...</td>\n",
       "      <td>AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...</td>\n",
       "      <td>AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...</td>\n",
       "      <td>GGGGCAGCGCCTCACAACCTCCGTCATGTGCTGTGACTGCTTGTAG...</td>\n",
       "      <td>G</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>258</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578217</td>\n",
       "      <td>7578240</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GTGTTTCTGTCATCCAAATACTCC</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.609_632del</td>\n",
       "      <td>p.E204_T211del</td>\n",
       "      <td>7578157</td>\n",
       "      <td>7578300</td>\n",
       "      <td>ACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATAGGGCACCACCA...</td>\n",
       "      <td>ACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATAGGGCACCACCA...</td>\n",
       "      <td>ACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATAGGGCACCACCA...</td>\n",
       "      <td>ACACGCAAATTTCCTTCCACTCGGATAAGATGCTGAGGAGGGGCCA...</td>\n",
       "      <td>GTGTTTCTGTCATCCAAATACTCC</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>270</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578206</td>\n",
       "      <td>7578207</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>TA</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.642_643delTA</td>\n",
       "      <td>p.H214Qfs*7</td>\n",
       "      <td>7578146</td>\n",
       "      <td>7578267</td>\n",
       "      <td>TCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATA...</td>\n",
       "      <td>TCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATA...</td>\n",
       "      <td>TCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATA...</td>\n",
       "      <td>TGTCGAAAAGTGTTTCTGTCATCCAAATACTCCACACGCAAATTTC...</td>\n",
       "      <td>TA</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>274</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578237</td>\n",
       "      <td>7578253</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>CTCCACACGCAAATTTC</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.596_612del</td>\n",
       "      <td>p.G199Vfs*4</td>\n",
       "      <td>7578177</td>\n",
       "      <td>7578313</td>\n",
       "      <td>CTCAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTT...</td>\n",
       "      <td>CTCAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTT...</td>\n",
       "      <td>CTCAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTT...</td>\n",
       "      <td>CTTCCACTCGGATAAGATGCTGAGGAGGGGCCAGACCTAAGAGCAA...</td>\n",
       "      <td>CTCCACACGCAAATTTC</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>276</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577525</td>\n",
       "      <td>7577527</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GAG</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.754_756del</td>\n",
       "      <td>p.L252del</td>\n",
       "      <td>7577465</td>\n",
       "      <td>7577587</td>\n",
       "      <td>AGGCCAGTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCC...</td>\n",
       "      <td>AGGCCAGTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCC...</td>\n",
       "      <td>AGGCCAGTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCC...</td>\n",
       "      <td>GATGGGCCTCCGGTTCATGCCGCCCATGCAGGAACTGTTACACATG...</td>\n",
       "      <td>GAG</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2996</th>\n",
       "      <td>2996</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579319</td>\n",
       "      <td>7579320</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.367_368del</td>\n",
       "      <td>p.T123Lfs*25</td>\n",
       "      <td>7579259</td>\n",
       "      <td>7579380</td>\n",
       "      <td>GATACGGCCAGGCATTGAAGTCTCATGGAAGCCAGCCCCTCAGGGC...</td>\n",
       "      <td>GATACGGCCAGGCATTGAAGTCTCATGGAAGCCAGCCCCTCAGGGC...</td>\n",
       "      <td>GATACGGCCAGGCATTGAAGTCTCATGGAAGCCAGCCCCTCAGGGC...</td>\n",
       "      <td>CACAGACTTGGCTGTCCCAGAATGCAAGAAGCCCAGACGGAAACCG...</td>\n",
       "      <td>GT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2999</th>\n",
       "      <td>2999</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579535</td>\n",
       "      <td>7579556</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>TCAATATCGTCCGGGGACAGCA</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.131_152del</td>\n",
       "      <td>p.M44Nfs*72</td>\n",
       "      <td>7579475</td>\n",
       "      <td>7579616</td>\n",
       "      <td>GGAGCAGCCTCTGGCATTCTGGGAGCTTCATCTGGACCTGGGTCTT...</td>\n",
       "      <td>GGAGCAGCCTCTGGCATTCTGGGAGCTTCATCTGGACCTGGGTCTT...</td>\n",
       "      <td>GGAGCAGCCTCTGGCATTCTGGGAGCTTCATCTGGACCTGGGTCTT...</td>\n",
       "      <td>TCAAATCATCCATTGCTTGGGACGGCAAGGGGGACTGTAGATGGGT...</td>\n",
       "      <td>TCAATATCGTCCGGGGACAGCA</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3000</th>\n",
       "      <td>3000</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577539</td>\n",
       "      <td>7577541</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GGT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.740_742del</td>\n",
       "      <td>p.N247del</td>\n",
       "      <td>7577479</td>\n",
       "      <td>7577601</td>\n",
       "      <td>GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...</td>\n",
       "      <td>GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...</td>\n",
       "      <td>GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...</td>\n",
       "      <td>TCATGCCGCCCATGCAGGAACTGTTACACATGTAGTTGTAGTGGAT...</td>\n",
       "      <td>GGT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3002</th>\n",
       "      <td>3002</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577531</td>\n",
       "      <td>7577532</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GG</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.749_750delCC</td>\n",
       "      <td>p.P250Hfs*13</td>\n",
       "      <td>7577471</td>\n",
       "      <td>7577592</td>\n",
       "      <td>GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...</td>\n",
       "      <td>GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...</td>\n",
       "      <td>GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...</td>\n",
       "      <td>GCCTCCGGTTCATGCCGCCCATGCAGGAACTGTTACACATGTAGTT...</td>\n",
       "      <td>GG</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3003</th>\n",
       "      <td>3003</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578421</td>\n",
       "      <td>7578440</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GTCATGTGCTGTGACTGCTT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.490_509del</td>\n",
       "      <td>p.K164Gfs*10</td>\n",
       "      <td>7578361</td>\n",
       "      <td>7578500</td>\n",
       "      <td>AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...</td>\n",
       "      <td>AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...</td>\n",
       "      <td>AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...</td>\n",
       "      <td>GTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGGAA...</td>\n",
       "      <td>GTCATGTGCTGTGACTGCTT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1327 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "257     257        TP53         17         7578398       7578398   \n",
       "258     258        TP53         17         7578217       7578240   \n",
       "270     270        TP53         17         7578206       7578207   \n",
       "274     274        TP53         17         7578237       7578253   \n",
       "276     276        TP53         17         7577525       7577527   \n",
       "...     ...         ...        ...             ...           ...   \n",
       "2996   2996        TP53         17         7579319       7579320   \n",
       "2999   2999        TP53         17         7579535       7579556   \n",
       "3000   3000        TP53         17         7577539       7577541   \n",
       "3002   3002        TP53         17         7577531       7577532   \n",
       "3003   3003        TP53         17         7578421       7578440   \n",
       "\n",
       "             Consequence Variant_Classification Variant_Type  \\\n",
       "257   frameshift_variant        Frame_Shift_Del          DEL   \n",
       "258     inframe_deletion           In_Frame_Del          DEL   \n",
       "270   frameshift_variant        Frame_Shift_Del          DEL   \n",
       "274   frameshift_variant        Frame_Shift_Del          DEL   \n",
       "276     inframe_deletion           In_Frame_Del          DEL   \n",
       "...                  ...                    ...          ...   \n",
       "2996  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "2999  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "3000    inframe_deletion           In_Frame_Del          DEL   \n",
       "3002  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "3003  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "\n",
       "              Reference_Allele Tumor_Seq_Allele2  \\\n",
       "257                          G                 -   \n",
       "258   GTGTTTCTGTCATCCAAATACTCC                 -   \n",
       "270                         TA                 -   \n",
       "274          CTCCACACGCAAATTTC                 -   \n",
       "276                        GAG                 -   \n",
       "...                        ...               ...   \n",
       "2996                        GT                 -   \n",
       "2999    TCAATATCGTCCGGGGACAGCA                 -   \n",
       "3000                       GGT                 -   \n",
       "3002                        GG                 -   \n",
       "3003      GTCATGTGCTGTGACTGCTT                 -   \n",
       "\n",
       "                                 HGVSc     HGVSp_Short  seq_start  seq_end  \\\n",
       "257        ENST00000269305.4:c.532delC    p.H178Tfs*69    7578338  7578458   \n",
       "258     ENST00000269305.4:c.609_632del  p.E204_T211del    7578157  7578300   \n",
       "270   ENST00000269305.4:c.642_643delTA     p.H214Qfs*7    7578146  7578267   \n",
       "274     ENST00000269305.4:c.596_612del     p.G199Vfs*4    7578177  7578313   \n",
       "276     ENST00000269305.4:c.754_756del       p.L252del    7577465  7577587   \n",
       "...                                ...             ...        ...      ...   \n",
       "2996    ENST00000269305.4:c.367_368del    p.T123Lfs*25    7579259  7579380   \n",
       "2999    ENST00000269305.4:c.131_152del     p.M44Nfs*72    7579475  7579616   \n",
       "3000    ENST00000269305.4:c.740_742del       p.N247del    7577479  7577601   \n",
       "3002  ENST00000269305.4:c.749_750delCC    p.P250Hfs*13    7577471  7577592   \n",
       "3003    ENST00000269305.4:c.490_509del    p.K164Gfs*10    7578361  7578500   \n",
       "\n",
       "                                           wt_w_context  \\\n",
       "257   AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...   \n",
       "258   ACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATAGGGCACCACCA...   \n",
       "270   TCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATA...   \n",
       "274   CTCAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTT...   \n",
       "276   AGGCCAGTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCC...   \n",
       "...                                                 ...   \n",
       "2996  GATACGGCCAGGCATTGAAGTCTCATGGAAGCCAGCCCCTCAGGGC...   \n",
       "2999  GGAGCAGCCTCTGGCATTCTGGGAGCTTCATCTGGACCTGGGTCTT...   \n",
       "3000  GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...   \n",
       "3002  GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...   \n",
       "3003  AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...   \n",
       "\n",
       "                                          alt_w_context  \\\n",
       "257   AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...   \n",
       "258   ACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATAGGGCACCACCA...   \n",
       "270   TCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATA...   \n",
       "274   CTCAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTT...   \n",
       "276   AGGCCAGTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCC...   \n",
       "...                                                 ...   \n",
       "2996  GATACGGCCAGGCATTGAAGTCTCATGGAAGCCAGCCCCTCAGGGC...   \n",
       "2999  GGAGCAGCCTCTGGCATTCTGGGAGCTTCATCTGGACCTGGGTCTT...   \n",
       "3000  GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...   \n",
       "3002  GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...   \n",
       "3003  AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...   \n",
       "\n",
       "                                           left_context  \\\n",
       "257   AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...   \n",
       "258   ACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATAGGGCACCACCA...   \n",
       "270   TCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATA...   \n",
       "274   CTCAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTT...   \n",
       "276   AGGCCAGTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCC...   \n",
       "...                                                 ...   \n",
       "2996  GATACGGCCAGGCATTGAAGTCTCATGGAAGCCAGCCCCTCAGGGC...   \n",
       "2999  GGAGCAGCCTCTGGCATTCTGGGAGCTTCATCTGGACCTGGGTCTT...   \n",
       "3000  GGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTGATGATGGT...   \n",
       "3002  GTGTGCAGGGTGGCAAGTGGCTCCTGACCTGGAGTCTTCCAGTGTG...   \n",
       "3003  AGCTGCTCACCATCGCTATCTGAGCAGCGCTCATGGTGGGGGCAGC...   \n",
       "\n",
       "                                          right_context  \\\n",
       "257   GGGGCAGCGCCTCACAACCTCCGTCATGTGCTGTGACTGCTTGTAG...   \n",
       "258   ACACGCAAATTTCCTTCCACTCGGATAAGATGCTGAGGAGGGGCCA...   \n",
       "270   TGTCGAAAAGTGTTTCTGTCATCCAAATACTCCACACGCAAATTTC...   \n",
       "274   CTTCCACTCGGATAAGATGCTGAGGAGGGGCCAGACCTAAGAGCAA...   \n",
       "276   GATGGGCCTCCGGTTCATGCCGCCCATGCAGGAACTGTTACACATG...   \n",
       "...                                                 ...   \n",
       "2996  CACAGACTTGGCTGTCCCAGAATGCAAGAAGCCCAGACGGAAACCG...   \n",
       "2999  TCAAATCATCCATTGCTTGGGACGGCAAGGGGGACTGTAGATGGGT...   \n",
       "3000  TCATGCCGCCCATGCAGGAACTGTTACACATGTAGTTGTAGTGGAT...   \n",
       "3002  GCCTCCGGTTCATGCCGCCCATGCAGGAACTGTTACACATGTAGTT...   \n",
       "3003  GTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGGAA...   \n",
       "\n",
       "                           REF ALT  \n",
       "257                          G      \n",
       "258   GTGTTTCTGTCATCCAAATACTCC      \n",
       "270                         TA      \n",
       "274          CTCCACACGCAAATTTC      \n",
       "276                        GAG      \n",
       "...                        ...  ..  \n",
       "2996                        GT      \n",
       "2999    TCAATATCGTCCGGGGACAGCA      \n",
       "3000                       GGT      \n",
       "3002                        GG      \n",
       "3003      GTCATGTGCTGTGACTGCTT      \n",
       "\n",
       "[1327 rows x 20 columns]"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p53_formatted[p53_formatted['Variant_Type']=='DEL']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGTGAGCCCTGCTCCCCCCTGGCTCCTTCCCAGCCTGGGCATCCTTGAGTTCCAAGGCCTC-ATTCAGCTCTCGGAACATCTCGAAGCGCTCACGCCCACGGATCTGCAGCAACAGAGGAGGG\n",
      "|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "AGTGAGCCCTGCTCCCCCCTGGCTCCTTCCCAGCCTGGGCATCCTTGAGTTCCAAGGCCTCAATTCAGCTCTCGGAACATCTCGAAGCGCTCACGCCCACGGATCTGCAGCAACAGAGGAGGG\n",
      "\n"
     ]
    }
   ],
   "source": [
    "i=259\n",
    "wt = p53_formatted.iloc[i]['wt_w_context']\n",
    "alt = p53_formatted.iloc[i]['alt_w_context']\n",
    "\n",
    "#wt = 'TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGGACAGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATAG'\n",
    "#alt = 'TGGTCAGGCTCCCCTTTCTTGCGGAGGTTCTCTTCGTCTGTGCGCCGGTCTCTCCCAGGACAGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATAG'\n",
    "alignments = aligner.align(wt, alt)\n",
    "\n",
    "print(alignments[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mut_formatter(wt, alt):\n",
    "\n",
    "    alignments = aligner.align(wt, alt)\n",
    "\n",
    "    WT_align = alignments[0].aligned[0]\n",
    "    alt_align = alignments[0].aligned[1]\n",
    "\n",
    "    if len(WT_align) == 1:\n",
    "        #Substitution (SNP or ONP)\n",
    "\n",
    "        #find indexes of non-matching values to get right/left context sequences and mutant sequence\n",
    "        non_matching_idx = []\n",
    "\n",
    "        for i, val in enumerate(list(zip(wt, alt))):\n",
    "            if val[0] != val[1]:\n",
    "                non_matching_idx.append(i)\n",
    "        \n",
    "        assert len(non_matching_idx) !=0, f'No mutation detected in alt sequence provided | row index = {i}'\n",
    "\n",
    "        if len(non_matching_idx)==1:\n",
    "            var_type = 'SNP'\n",
    "        else:\n",
    "            var_type = 'ONP'\n",
    "\n",
    "        left_seq = wt[:non_matching_idx[0]]\n",
    "        ref_seq = wt[non_matching_idx[0]:non_matching_idx[-1]+1]\n",
    "        alt_seq = alt[non_matching_idx[0]:non_matching_idx[-1]+1]\n",
    "        right_seq = wt[non_matching_idx[-1]+1:]\n",
    "\n",
    "\n",
    "    elif len(WT_align)>1:\n",
    "        if len(alt)>len(wt):\n",
    "            var_type = 'INS'\n",
    "        elif len(alt)<len(wt):\n",
    "            var_type='DEL'\n",
    "\n",
    "        left_seq = wt[WT_align[0][0]:WT_align[0][1]]\n",
    "        right_seq = wt[WT_align[1][0]:WT_align[1][1]]\n",
    "\n",
    "        if var_type=='INS':\n",
    "        #INS or DEL\n",
    "            ref_seq = ''\n",
    "            alt_seq = alt[alt_align[0][1]:alt_align[1][0]]\n",
    "            \n",
    "        elif var_type =='DEL':\n",
    "            ref_seq = wt[WT_align[0][1]:WT_align[1][0]]\n",
    "            alt_seq = ''\n",
    "            \n",
    "\n",
    "\n",
    "    print(var_type)\n",
    "    print(left_seq)\n",
    "    print(f'REF: {ref_seq}')\n",
    "    print(f'ALT: {alt_seq}')\n",
    "    print(right_seq)\n",
    "\n",
    "    #print(alignments[0])\n",
    "    #this program will break for complex substitutions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATAGGGCACCACCACACTATGTCGAAAAGTGTTTCTGTCATCCAAATACTCCACACGCAAATTTCCTTCCACTCGGATA\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||--||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "TCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATAGGGCACCACCACAC--TGTCGAAAAGTGTTTCTGTCATCCAAATACTCCACACGCAAATTTCCTTCCACTCGGATA\n",
      "\n",
      "DEL\n",
      "TCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCGGCTCATAGGGCACCACCACAC\n",
      "REF: TA\n",
      "ALT: \n",
      "TGTCGAAAAGTGTTTCTGTCATCCAAATACTCCACACGCAAATTTCCTTCCACTCGGATA\n"
     ]
    }
   ],
   "source": [
    "i=270\n",
    "wt = p53_formatted.iloc[i]['wt_w_context']\n",
    "alt = p53_formatted.iloc[i]['alt_w_context']\n",
    "\n",
    "#wt = 'TGGTGAGGCTCCCCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGGACAGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATAG'\n",
    "#alt = 'TGGTCAGGCTCCCCTTTCTTGCGGAGGTTCTCTTCGTCTGTGCGCCGGTCTCTCCCAGGACAGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATAG'\n",
    "alignments = aligner.align(wt, alt)\n",
    "\n",
    "print(alignments[0])\n",
    "\n",
    "mut_formatter(wt, alt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def primedesign_formatter(seq):\n",
    "    \"\"\" \n",
    "    takes as input sequence in prime design format \n",
    "    e.g. AATTCCG(G/C)AATTCGCT\n",
    "    \"\"\"\n",
    "\n",
    "    start = seq.find(\"(\")\n",
    "\n",
    "    end = seq.find(\")\")\n",
    "\n",
    "    replace_seq = seq[start:end+1]\n",
    "    #if '/' in replace_seq:\n",
    "        #if '+' in replace_seq:\n",
    "            #throw an error\n",
    "\n",
    "    loc_replace = replace_seq.find('/')\n",
    "    ref_seq = replace_seq[1:loc_replace]\n",
    "    alt_seq = replace_seq[loc_replace+1:-1]\n",
    "\n",
    "    left_seq = seq[:start]\n",
    "    right_seq = seq[end+1:]\n",
    "\n",
    "    wt_w_context = left_seq + ref_seq + right_seq\n",
    "    alt_w_context = left_seq + alt_seq + right_seq\n",
    "\n",
    "    if len(ref_seq)==0:\n",
    "        var_type='INS'\n",
    "\n",
    "    else:\n",
    "        if len(ref_seq)==len(alt_seq):\n",
    "            if len(ref_seq)==1:\n",
    "                var_type='SNP'\n",
    "            else:\n",
    "                var_type='ONP'\n",
    "\n",
    "        else:\n",
    "            if len(alt_seq) == 0:\n",
    "                var_type = 'DEL'\n",
    "            else:\n",
    "                var_type = 'INDEL'\n",
    "\n",
    "\n",
    "    \n",
    "    print(var_type)\n",
    "\n",
    "    print(wt_w_context)\n",
    "    print(alt_w_context)\n",
    "    print(f'LEFT: {left_seq}')\n",
    "    print(f'REF: {ref_seq}')\n",
    "    print(f'ALT: {alt_seq}')\n",
    "    print(f'RIGHT: {right_seq}')\n",
    "\n",
    "    return var_type, wt_w_context, alt_w_context, left_seq, right_seq, ref_seq, alt_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEL\n",
      "AATTCGTCGTAGCGTCGGCTAGCTAGCTGGCA\n",
      "AATTCGTCGTAGCTCGGCTAGCTAGCTGGCA\n",
      "LEFT: AATTCGTCGTAGC\n",
      "REF: G\n",
      "ALT: \n",
      "RIGHT: TCGGCTAGCTAGCTGGCA\n"
     ]
    }
   ],
   "source": [
    "seq = 'AATTCGTCGTAGC(G/)TCGGCTAGCTAGCTGGCA'\n",
    "var_type, wt_w_context, alt_w_context, left_seq, right_seq, ref_seq, alt_seq = primedesign_formatter(seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PAM searching\n",
    "\n",
    "- Idea is to only do everything with respect to the forward strand\n",
    "- Not code a bespoke solution for the reverse complement strand\n",
    "- And just fix the coordinates afterward (translate them to the forward strand coordinates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PAM_finder(seq, PAM):\n",
    "    \"\"\" \n",
    "    Finds indeces of PAM sequences\n",
    "    Formatting: e.g. \"NGG\"\n",
    "    N = [A|T|C|G]\n",
    "    R = [A|G]\n",
    "    Y = [C|T]\n",
    "    S = [G|C]\n",
    "    W = [A|T]\n",
    "    K = [G|T]\n",
    "    M = [A|C]\n",
    "    B = [C|G|T]\n",
    "    D = [A|G|T]\n",
    "    H = [A|C|T]\n",
    "    V = [A|C|G]\n",
    "    \"\"\"\n",
    "    PAM_dict = {\"A\":\"A\",\n",
    "                \"G\":\"G\",\n",
    "                \"C\":\"C\",\n",
    "                \"T\":\"T\",\n",
    "                \"N\":\"[A|T|C|G]\",\n",
    "                \"R\":\"[A|G]\",\n",
    "                \"Y\":\"[C|T]\",\n",
    "                \"S\":\"[G|C]\",\n",
    "                \"W\":\"[A|T]\",\n",
    "                \"K\":\"[G|T]\",\n",
    "                \"M\":\"[A|C]\",\n",
    "                \"B\":\"[C|G|T]\",\n",
    "                \"D\":\"[A|G|T]\",\n",
    "                \"H\":\"[A|C|T]\",\n",
    "                \"V\":\"[A|C|G]\"}\n",
    "\n",
    "    PAM_new = \"\"\n",
    "    for i in PAM:\n",
    "        PAM_new += PAM_dict[i]\n",
    "\n",
    "    p = re.compile(PAM_new)\n",
    "\n",
    "    iterator = p.finditer(seq, overlapped=True)\n",
    "    PAM_holder = []\n",
    "    for match in iterator:\n",
    "        PAM_holder.append(match.span())\n",
    "\n",
    "    return PAM_holder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 410,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGG\n",
      "CGG\n",
      "AGG\n",
      "GGG\n",
      "CGG\n"
     ]
    }
   ],
   "source": [
    "#use left_seq length as well?\n",
    "#to filter which are eligible PAM sequences???\n",
    "#and then do the reverse complement...\n",
    "\n",
    "i=270\n",
    "wt = p53_formatted.iloc[i]['wt_w_context']\n",
    "alt = p53_formatted.iloc[i]['alt_w_context']\n",
    "\n",
    "PAM = 'NGG'\n",
    "PAM_holder = PAM_finder(wt, PAM)\n",
    "\n",
    "for i in PAM_holder:\n",
    "    print(wt[i[0]:i[1]])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 414,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 3 GGG\n",
      "1 4 GGG\n",
      "4 7 TGG\n",
      "5 8 GGG\n"
     ]
    }
   ],
   "source": [
    "PAM = 'NGG'\n",
    "seq = 'GGGGTGGG'\n",
    "PAM_holder = PAM_finder(seq, PAM)\n",
    "\n",
    "for i in PAM_holder:\n",
    "    print(i[0], i[1], seq[i[0]:i[1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(18, 21), (30, 33)]"
      ]
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PAM_holder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 501,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ONP\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTACGTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTGCATGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "LEFT: GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTT\n",
      "REF: ACG\n",
      "ALT: GCA\n",
      "RIGHT: TGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "GGG\n",
      "TGG\n",
      "TGG\n",
      "TGG\n",
      "TGG\n",
      "TGG\n",
      "TGG\n",
      "TGG\n"
     ]
    }
   ],
   "source": [
    "PAM = 'NGG'\n",
    "\n",
    "#seq = 'ATGATGCGCGGCTAGAATTCGTTGGCGTAGCGTAGCGT(G/)TCTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA'\n",
    "seq = 'GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTT(ACG/GCA)TGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA'\n",
    "\n",
    "var_type, wt_w_context, alt_w_context, left_seq, right_seq, ref_seq, alt_seq = primedesign_formatter(seq)\n",
    "\n",
    "\n",
    "PAM_holder = PAM_finder(wt_w_context, PAM)\n",
    "\n",
    "for i in PAM_holder:\n",
    "    print(wt_w_context[i[0]:i[1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 502,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 502,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('CGTAAGCGT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 505,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(26, 29), (43, 46)]"
      ]
     },
     "execution_count": 505,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#searching for PAM sequences on the FORWARD STRAND\n",
    "\n",
    "#search on max_RTT_length - \n",
    "\n",
    "left_len = len(left_seq)\n",
    "right_len = len(right_seq)\n",
    "max_RTT_length = 20\n",
    "\n",
    "if var_type =='INS':\n",
    "    alt_size = len(alt_seq)\n",
    "    m_start = left_len - (max_RTT_length - alt_size - 3)\n",
    "    m_end = left_len + 4\n",
    "\n",
    "else:\n",
    "    ref_size = len(ref_seq)\n",
    "    alt_size = len(alt_seq)\n",
    "    m_start = left_len - (max_RTT_length - alt_size - 3)\n",
    "    m_end = left_len + 3\n",
    "\n",
    "proto_size = 19\n",
    "\n",
    "eligible_PAM_start = max(m_start, proto_size + 3) #\n",
    "eligible_PAM_end = m_end\n",
    "\n",
    "eligible_PAMS = []\n",
    "for i in PAM_holder:\n",
    "    #check start of PAM site\n",
    "    s = i[0]\n",
    "    if (s>=eligible_PAM_start) and (s<=eligible_PAM_end):\n",
    "        eligible_PAMS.append(i)\n",
    "\n",
    "eligible_PAMS\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 506,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------GGGCGTAAGCGTTTACGTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "--------------------------|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTACGTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTACGTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "||||||||||||||||||||||||||||||||||||||||.|.||||||||||||||||||||||||||||||||||||||||||\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTGCATGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "\n",
      "-------------------------------------------TGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "-------------------------------------------||||||||||||||||||||||||||||||||||||||||||\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTACGTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTACGTGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "||||||||||||||||||||||||||||||||||||||||.|.||||||||||||||||||||||||||||||||||||||||||\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTGCATGGCTAGCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in eligible_PAMS:\n",
    "    s = i[0]\n",
    "    alignments = aligner.align(wt_w_context[s:], wt_w_context)\n",
    "\n",
    "    print(alignments[0])\n",
    "\n",
    "    alignments2 = aligner.align(wt_w_context, alt_w_context)\n",
    "    print(alignments2[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 492,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 492,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('GGGCGTAAGCGTTTACG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCTTTCTTGCGGAGATTCTCTTCCTCTGTGCGCCGGTCTCTCCCAGGACAGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATAGGAAAAGAGAAGC\n",
      "CGCC\n",
      "GACA\n",
      "GGCA\n",
      "AACA\n",
      "CGCA\n",
      "AGCT\n",
      "TACC\n",
      "TACT\n"
     ]
    }
   ],
   "source": [
    "seq = p53_formatted.iloc[0]['wt_w_context']\n",
    "print(seq)\n",
    "PAM = \"NRCH\"\n",
    "PAM_holder = PAM_finder(seq, PAM)\n",
    "\n",
    "for i in PAM_holder:\n",
    "    print(seq[i[0]:i[1]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Putting some things together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 734,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEL\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTATGGTAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "LEFT: GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTT\n",
      "REF: ATGG\n",
      "ALT: \n",
      "RIGHT: TAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n"
     ]
    }
   ],
   "source": [
    "seq = 'GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTT(ATGG/)TAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA'\n",
    "\n",
    "var_type, wt_w_context, alt_w_context, left_seq, right_seq, ref_seq, alt_seq = primedesign_formatter(seq)\n",
    "mut = mutation(wt_w_context, alt_w_context, left_seq, right_seq, var_type, ref_seq, alt_seq, chrom=None, genome=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def eligible_PAM_finder(mut, PAM, max_RTT_length, proto_size=19):\n",
    "    \"\"\" \n",
    "    Determines eligible PAM sequences for creating pegRNAs...\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    forward = PAM_finder(mut.wt_forward, PAM)\n",
    "    reverse = PAM_finder(mut.wt_rc, PAM)\n",
    "\n",
    "\n",
    "    #first do the forward sequence\n",
    "    left_len_F = len(mut.left_seq)\n",
    "    left_len_R = len(mut.left_seq_rc)\n",
    "\n",
    "    if var_type =='INS':\n",
    "        alt_size = len(mut.alt_seq)\n",
    "        m_start_F = left_len_F - (max_RTT_length - alt_size - 3)\n",
    "        m_end_F = left_len_F + 4\n",
    "        \n",
    "        m_start_R = left_len_R - (max_RTT_length - alt_size - 3)\n",
    "        m_end_R = left_len_R + 4\n",
    "\n",
    "    else:\n",
    "        alt_size = len(mut.alt_seq)\n",
    "        m_start_F = left_len_F - (max_RTT_length - alt_size - 3)\n",
    "        m_end_F = left_len_F + 3\n",
    "\n",
    "        m_start_R = left_len_R - (max_RTT_length - alt_size - 3)\n",
    "        m_end_R = left_len_R + 3\n",
    "\n",
    "\n",
    "    eligible_PAM_start_F = max(m_start_F, proto_size + 3) #\n",
    "    eligible_PAM_end_F = m_end_F\n",
    "    \n",
    "    eligible_PAM_start_R = max(m_start_R, proto_size + 3) #\n",
    "    eligible_PAM_end_R = m_end_R\n",
    "\n",
    "    eligible_PAMS_F = []\n",
    "    eligible_PAMS_R = []\n",
    "\n",
    "    for i in forward:\n",
    "        #check start of PAM site\n",
    "        s = i[0]\n",
    "        if (s>=eligible_PAM_start_F) and (s<=eligible_PAM_end_F):\n",
    "            eligible_PAMS_F.append(i)\n",
    "\n",
    "    for i in reverse:\n",
    "        #check start of PAM site\n",
    "        s = i[0]\n",
    "        if (s>=eligible_PAM_start_R) and (s<=eligible_PAM_end_R):\n",
    "            eligible_PAMS_R.append(i)\n",
    "\n",
    "    return eligible_PAMS_F, eligible_PAMS_R\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 622,
   "metadata": {},
   "outputs": [],
   "source": [
    "#eligible PAM finding\n",
    "PAM = \"NGG\"\n",
    "max_RTT_length = 20\n",
    "proto_size = 10\n",
    "\n",
    "mut.PAM_idx_forward = PAM_finder(mut.wt_forward, PAM)\n",
    "mut.PAM_idx_rc = PAM_finder(mut.wt_rc, PAM)\n",
    "\n",
    "mut.PAM_idx_forward, mut.PAM_idx_rc = eligible_PAM_finder(mut, PAM, max_RTT_length, proto_size)\n",
    "#mut.PAM_idx_forward\n",
    "#mut.PAM_idx_rc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 623,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(23, 26), (24, 27), (25, 28), (26, 29), (41, 44)]"
      ]
     },
     "execution_count": 623,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut.PAM_idx_forward"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 624,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(33, 36)]"
      ]
     },
     "execution_count": 624,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut.PAM_idx_rc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# pegRNA design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pegRNA_generator(mut, PAM, orientation, proto_size, RTT_lengths, PBS_lengths):\n",
    "    #------function----\n",
    "    protospacer_seqs = []\n",
    "    PAM_start_list = []\n",
    "    PAM_end_list = []\n",
    "    PAM_list = []\n",
    "\n",
    "    RTT_seqs = []\n",
    "    RTT_size = []\n",
    "    PBS_seqs = []\n",
    "    PBS_size = []\n",
    "    RTT_PBS_seqs = []\n",
    "\n",
    "    #descriptors of pegRNAs...\n",
    "    distance_to_nick = [] #distance from the start of the mutation to the nick introduced by nCas9 (-3 from PAM)\n",
    "    RHA = [] #right homology arm (distance from end of mutation to end of RTT)\n",
    "    PAM_disrupted_list = []\n",
    "    proto_disrupted_list = []\n",
    "\n",
    "    #Get mutation info, depending on whether we're looking on the + or the - strand\n",
    "    if orientation == '+':\n",
    "        left_len = len(mut.left_seq)\n",
    "        ref_len = len(mut.ref_seq)\n",
    "        alt_len = len(mut.alt_seq)\n",
    "        alt_seq = mut.alt_seq\n",
    "        seq_F = mut.wt_forward\n",
    "\n",
    "        PAM_seqs = mut.PAM_idx_forward\n",
    "\n",
    "    elif orientation == '-':\n",
    "        left_len = len(mut.left_seq_rc)\n",
    "        ref_len = len(mut.ref_seq_rc)\n",
    "        alt_len = len(mut.alt_seq_rc)\n",
    "        alt_seq = mut.alt_seq_rc\n",
    "        seq_F = mut.wt_rc\n",
    "\n",
    "        PAM_seqs = mut.PAM_idx_rc\n",
    "\n",
    "    #iterate over the PAM sequences:\n",
    "    for i in PAM_seqs:\n",
    "        PAM_start = i[0]\n",
    "        PAM_end = i[1]\n",
    "\n",
    "        RTT_start = PAM_start - 3\n",
    "\n",
    "        left_RTT = seq_F[RTT_start:left_len]\n",
    "\n",
    "        for RTT_length in RTT_lengths:\n",
    "\n",
    "            remaining_length = RTT_length - (len(left_RTT) + alt_len)\n",
    "\n",
    "            #make sure RTT is long enough\n",
    "            #and make sure that there's sufficient sequence to actually pull from...\n",
    "            if (remaining_length <0) or ((left_len+ref_len + remaining_length) > len(seq_F)):\n",
    "                RTT_seqs.append(None)\n",
    "                distance_to_nick.append(None)\n",
    "                RHA.append(None)\n",
    "                PAM_disrupted_list.append(None)\n",
    "                proto_disrupted_list.append(None)\n",
    "                PAM_start_list.append(None)\n",
    "                PAM_end_list.append(None)\n",
    "                PAM_list.append(None)\n",
    "                protospacer_seqs.append(None)\n",
    "                PBS_seqs.append(None)\n",
    "                RTT_PBS_seqs.append(None)\n",
    "                RTT_size.append(None)\n",
    "                PBS_size.append(None)\n",
    "\n",
    "            else: #able to design the pegRNA\n",
    "\n",
    "                right_RTT = seq_F[left_len+ref_len:left_len+ref_len + remaining_length]\n",
    "                RTT = left_RTT + alt_seq + right_RTT\n",
    "\n",
    "                #and then determine if the PAM sequence is disrupted\n",
    "                #protospacer disrupted would just be if the distance to nick is < 3\n",
    "                #fish out the PAM sequence from RTT\n",
    "                PAM_new = RTT[3:3+len(PAM)]\n",
    "                pam_list = PAM_finder(PAM_new, PAM)\n",
    "\n",
    "                PAM_disrupted = False\n",
    "                if len(pam_list)==0:\n",
    "                    PAM_disrupted = True\n",
    "\n",
    "                proto_disrupted = False\n",
    "                if len(left_RTT) < 3:\n",
    "                    proto_disrupted = True\n",
    "\n",
    "                #and finally reverse complement the RTT\n",
    "                RTT = str(Bio.Seq.Seq(RTT).reverse_complement())\n",
    "\n",
    "                #pull out the PAM sequence and protospacer as well\n",
    "                PAM_sequence = seq_F[PAM_start:PAM_end]\n",
    "                protospacer =  'G' + seq_F[PAM_start - proto_size:PAM_start]\n",
    "\n",
    "\n",
    "                #iterate over PBS lengths for adding things to list\n",
    "                for PBS_length in PBS_lengths:\n",
    "                    RTT_seqs.append(RTT)\n",
    "                    RTT_size.append(RTT_length)\n",
    "                    PBS_size.append(PBS_length)\n",
    "                    distance_to_nick.append(len(left_RTT))\n",
    "                    RHA.append(len(right_RTT))\n",
    "                    PAM_disrupted_list.append(PAM_disrupted)\n",
    "                    proto_disrupted_list.append(proto_disrupted)\n",
    "\n",
    "                    #------add in PAM info, protospacer, + PBS/RTT_PBS\n",
    "                    #record info about PAM sequences\n",
    "                    if orientation == '+':\n",
    "                        PAM_start_list.append(PAM_start)\n",
    "                        PAM_end_list.append(PAM_end)\n",
    "                    elif orientation == '-':\n",
    "                        PAM_start_list.append(len(seq_F)-PAM_start)\n",
    "                        PAM_end_list.append(len(seq_F)-PAM_end)\n",
    "\n",
    "                    PAM_list.append(PAM_sequence)\n",
    "                    protospacer_seqs.append(protospacer)\n",
    "\n",
    "                    #----and finally design the PBS sequences----\n",
    "                    #NEED TO HAVE AN ERROR MESSAGE FOR PBS SEQUENCES LARGER THAN 17\n",
    "                    PBS = seq_F[RTT_start-PBS_length:RTT_start]\n",
    "                    PBS = str(Bio.Seq.Seq(PBS).reverse_complement())\n",
    "                    PBS_seqs.append(PBS)\n",
    "\n",
    "                    #---and then add it to the RTT to make the full 3' extension\n",
    "                    RTT_PBS_seqs.append(RTT+PBS)\n",
    "\n",
    "    cols = [PAM_start_list, PAM_end_list, PAM_list, orientation, protospacer_seqs, RTT_seqs, RTT_size, PBS_seqs, PBS_size, RTT_PBS_seqs, distance_to_nick, RHA, PAM_disrupted_list, proto_disrupted_list]\n",
    "    col_labels = [\"PAM_start\", \"PAM_end\", \"PAM\", \"PAM_strand\",\"Protospacer\", \"RTT\", \"RTT_length\", \"PBS\", \"PBS_length\", \"RTT_PBS\", \"Distance_to_nick\", \"RHA_size\", \"PAM_disrupted\", \"Proto_disrupted\"]\n",
    "    dtypes = ['int', 'int', 'str', 'str', 'str', 'str', 'int', 'str', 'int', 'str', 'int', 'int', 'bool', 'bool']\n",
    "\n",
    "    dtype_dict = dict(zip(col_labels, dtypes))\n",
    "\n",
    "    peg_df = pd.DataFrame(dict(zip(col_labels, cols))).dropna().reset_index().drop(columns='index').astype(dtype_dict)\n",
    "\n",
    "    return peg_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEL\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTATGGTAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n",
      "LEFT: GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTT\n",
      "REF: ATGG\n",
      "ALT: \n",
      "RIGHT: TAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PAM_start</th>\n",
       "      <th>PAM_end</th>\n",
       "      <th>PAM</th>\n",
       "      <th>PAM_strand</th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>RTT</th>\n",
       "      <th>RTT_length</th>\n",
       "      <th>PBS</th>\n",
       "      <th>PBS_length</th>\n",
       "      <th>RTT_PBS</th>\n",
       "      <th>Distance_to_nick</th>\n",
       "      <th>RHA_size</th>\n",
       "      <th>PAM_disrupted</th>\n",
       "      <th>Proto_disrupted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>25</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>25</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>25</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>25</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>AAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>AAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>TAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>20</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>CTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCC</td>\n",
       "      <td>10</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>GGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGGGGGGGGGGGGG</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>30</td>\n",
       "      <td>CCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>AGCTACTAAA</td>\n",
       "      <td>10</td>\n",
       "      <td>ACGCTTACGC</td>\n",
       "      <td>10</td>\n",
       "      <td>AGCTACTAAAACGCTTACGC</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>AGCTACTAAA</td>\n",
       "      <td>10</td>\n",
       "      <td>ACGCTTACGCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>AGCTACTAAAACGCTTACGCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>AGCTACTAAA</td>\n",
       "      <td>10</td>\n",
       "      <td>ACGCTTACGCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>AGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>AGCTACTAAA</td>\n",
       "      <td>10</td>\n",
       "      <td>ACGCTTACGCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>AGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>20</td>\n",
       "      <td>ACGCTTACGC</td>\n",
       "      <td>10</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAAACGCTTACGC</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>20</td>\n",
       "      <td>ACGCTTACGCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAAACGCTTACGCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>20</td>\n",
       "      <td>ACGCTTACGCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>20</td>\n",
       "      <td>ACGCTTACGCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>CAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>30</td>\n",
       "      <td>ACGCTTACGC</td>\n",
       "      <td>10</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGC</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>30</td>\n",
       "      <td>ACGCTTACGCCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>30</td>\n",
       "      <td>ACGCTTACGCCCCCC</td>\n",
       "      <td>15</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>TGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGGGGGGCGTAAGCGTTTA</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAA</td>\n",
       "      <td>30</td>\n",
       "      <td>ACGCTTACGCCCCCCCC</td>\n",
       "      <td>17</td>\n",
       "      <td>CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCCCC</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>20</td>\n",
       "      <td>TGGCATGGCA</td>\n",
       "      <td>10</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGCTGGCATGGCA</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>20</td>\n",
       "      <td>TGGCATGGCATGG</td>\n",
       "      <td>13</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGCTGGCATGGCATGG</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>20</td>\n",
       "      <td>TGGCATGGCATGGCA</td>\n",
       "      <td>15</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCA</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>20</td>\n",
       "      <td>TGGCATGGCATGGCATG</td>\n",
       "      <td>17</td>\n",
       "      <td>GTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCATG</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>30</td>\n",
       "      <td>TGGCATGGCA</td>\n",
       "      <td>10</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCA</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>30</td>\n",
       "      <td>TGGCATGGCATGG</td>\n",
       "      <td>13</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGG</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>30</td>\n",
       "      <td>TGGCATGGCATGGCA</td>\n",
       "      <td>15</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCA</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>AGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GATGCCATGCCATGCCAGCT</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC</td>\n",
       "      <td>30</td>\n",
       "      <td>TGGCATGGCATGGCATG</td>\n",
       "      <td>17</td>\n",
       "      <td>GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCATG</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    PAM_start  PAM_end  PAM PAM_strand           Protospacer  \\\n",
       "0          22       25  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "1          22       25  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "2          22       25  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "3          22       25  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "4          23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "5          23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "6          23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "7          23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "8          23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "9          23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "10         23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "11         23       26  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "12         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "13         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "14         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "15         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "16         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "17         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "18         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "19         24       27  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "20         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "21         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "22         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "23         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "24         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "25         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "26         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "27         25       28  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "28         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "29         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "30         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "31         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "32         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "33         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "34         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "35         26       29  GGG          +  GGGGGGGGGGGGGGGGGGGG   \n",
       "36         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "37         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "38         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "39         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "40         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "41         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "42         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "43         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "44         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "45         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "46         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "47         41       44  TGG          +  GGGGGGGGCGTAAGCGTTTA   \n",
       "48         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "49         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "50         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "51         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "52         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "53         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "54         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "55         57       54  AGG          -  GATGCCATGCCATGCCAGCT   \n",
       "\n",
       "                               RTT  RTT_length                PBS  PBS_length  \\\n",
       "0   TAGCTACTAAAACGCTTACGCCCCCCCCCC          30         CCCCCCCCCC          10   \n",
       "1   TAGCTACTAAAACGCTTACGCCCCCCCCCC          30      CCCCCCCCCCCCC          13   \n",
       "2   TAGCTACTAAAACGCTTACGCCCCCCCCCC          30    CCCCCCCCCCCCCCC          15   \n",
       "3   TAGCTACTAAAACGCTTACGCCCCCCCCCC          30  CCCCCCCCCCCCCCCCC          17   \n",
       "4             AAACGCTTACGCCCCCCCCC          20         CCCCCCCCCC          10   \n",
       "5             AAACGCTTACGCCCCCCCCC          20      CCCCCCCCCCCCC          13   \n",
       "6             AAACGCTTACGCCCCCCCCC          20    CCCCCCCCCCCCCCC          15   \n",
       "7             AAACGCTTACGCCCCCCCCC          20  CCCCCCCCCCCCCCCCC          17   \n",
       "8   CTAGCTACTAAAACGCTTACGCCCCCCCCC          30         CCCCCCCCCC          10   \n",
       "9   CTAGCTACTAAAACGCTTACGCCCCCCCCC          30      CCCCCCCCCCCCC          13   \n",
       "10  CTAGCTACTAAAACGCTTACGCCCCCCCCC          30    CCCCCCCCCCCCCCC          15   \n",
       "11  CTAGCTACTAAAACGCTTACGCCCCCCCCC          30  CCCCCCCCCCCCCCCCC          17   \n",
       "12            AAAACGCTTACGCCCCCCCC          20         CCCCCCCCCC          10   \n",
       "13            AAAACGCTTACGCCCCCCCC          20      CCCCCCCCCCCCC          13   \n",
       "14            AAAACGCTTACGCCCCCCCC          20    CCCCCCCCCCCCCCC          15   \n",
       "15            AAAACGCTTACGCCCCCCCC          20  CCCCCCCCCCCCCCCCC          17   \n",
       "16  GCTAGCTACTAAAACGCTTACGCCCCCCCC          30         CCCCCCCCCC          10   \n",
       "17  GCTAGCTACTAAAACGCTTACGCCCCCCCC          30      CCCCCCCCCCCCC          13   \n",
       "18  GCTAGCTACTAAAACGCTTACGCCCCCCCC          30    CCCCCCCCCCCCCCC          15   \n",
       "19  GCTAGCTACTAAAACGCTTACGCCCCCCCC          30  CCCCCCCCCCCCCCCCC          17   \n",
       "20            TAAAACGCTTACGCCCCCCC          20         CCCCCCCCCC          10   \n",
       "21            TAAAACGCTTACGCCCCCCC          20      CCCCCCCCCCCCC          13   \n",
       "22            TAAAACGCTTACGCCCCCCC          20    CCCCCCCCCCCCCCC          15   \n",
       "23            TAAAACGCTTACGCCCCCCC          20  CCCCCCCCCCCCCCCCC          17   \n",
       "24  GGCTAGCTACTAAAACGCTTACGCCCCCCC          30         CCCCCCCCCC          10   \n",
       "25  GGCTAGCTACTAAAACGCTTACGCCCCCCC          30      CCCCCCCCCCCCC          13   \n",
       "26  GGCTAGCTACTAAAACGCTTACGCCCCCCC          30    CCCCCCCCCCCCCCC          15   \n",
       "27  GGCTAGCTACTAAAACGCTTACGCCCCCCC          30  CCCCCCCCCCCCCCCCC          17   \n",
       "28            CTAAAACGCTTACGCCCCCC          20         CCCCCCCCCC          10   \n",
       "29            CTAAAACGCTTACGCCCCCC          20      CCCCCCCCCCCCC          13   \n",
       "30            CTAAAACGCTTACGCCCCCC          20    CCCCCCCCCCCCCCC          15   \n",
       "31            CTAAAACGCTTACGCCCCCC          20  CCCCCCCCCCCCCCCCC          17   \n",
       "32  AGGCTAGCTACTAAAACGCTTACGCCCCCC          30         CCCCCCCCCC          10   \n",
       "33  AGGCTAGCTACTAAAACGCTTACGCCCCCC          30      CCCCCCCCCCCCC          13   \n",
       "34  AGGCTAGCTACTAAAACGCTTACGCCCCCC          30    CCCCCCCCCCCCCCC          15   \n",
       "35  AGGCTAGCTACTAAAACGCTTACGCCCCCC          30  CCCCCCCCCCCCCCCCC          17   \n",
       "36                      AGCTACTAAA          10         ACGCTTACGC          10   \n",
       "37                      AGCTACTAAA          10      ACGCTTACGCCCC          13   \n",
       "38                      AGCTACTAAA          10    ACGCTTACGCCCCCC          15   \n",
       "39                      AGCTACTAAA          10  ACGCTTACGCCCCCCCC          17   \n",
       "40            CAGCTAGGCTAGCTACTAAA          20         ACGCTTACGC          10   \n",
       "41            CAGCTAGGCTAGCTACTAAA          20      ACGCTTACGCCCC          13   \n",
       "42            CAGCTAGGCTAGCTACTAAA          20    ACGCTTACGCCCCCC          15   \n",
       "43            CAGCTAGGCTAGCTACTAAA          20  ACGCTTACGCCCCCCCC          17   \n",
       "44  CATGCCATGCCAGCTAGGCTAGCTACTAAA          30         ACGCTTACGC          10   \n",
       "45  CATGCCATGCCAGCTAGGCTAGCTACTAAA          30      ACGCTTACGCCCC          13   \n",
       "46  CATGCCATGCCAGCTAGGCTAGCTACTAAA          30    ACGCTTACGCCCCCC          15   \n",
       "47  CATGCCATGCCAGCTAGGCTAGCTACTAAA          30  ACGCTTACGCCCCCCCC          17   \n",
       "48            GTTTTAGTAGCTAGCCTAGC          20         TGGCATGGCA          10   \n",
       "49            GTTTTAGTAGCTAGCCTAGC          20      TGGCATGGCATGG          13   \n",
       "50            GTTTTAGTAGCTAGCCTAGC          20    TGGCATGGCATGGCA          15   \n",
       "51            GTTTTAGTAGCTAGCCTAGC          20  TGGCATGGCATGGCATG          17   \n",
       "52  GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC          30         TGGCATGGCA          10   \n",
       "53  GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC          30      TGGCATGGCATGG          13   \n",
       "54  GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC          30    TGGCATGGCATGGCA          15   \n",
       "55  GGGCGTAAGCGTTTTAGTAGCTAGCCTAGC          30  TGGCATGGCATGGCATG          17   \n",
       "\n",
       "                                            RTT_PBS  Distance_to_nick  \\\n",
       "0          TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCC                21   \n",
       "1       TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC                21   \n",
       "2     TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCC                21   \n",
       "3   TAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCCC                21   \n",
       "4                    AAACGCTTACGCCCCCCCCCCCCCCCCCCC                20   \n",
       "5                 AAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC                20   \n",
       "6               AAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC                20   \n",
       "7             AAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCC                20   \n",
       "8          CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCC                20   \n",
       "9       CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC                20   \n",
       "10    CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC                20   \n",
       "11  CTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCC                20   \n",
       "12                   AAAACGCTTACGCCCCCCCCCCCCCCCCCC                19   \n",
       "13                AAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC                19   \n",
       "14              AAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC                19   \n",
       "15            AAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCC                19   \n",
       "16         GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCC                19   \n",
       "17      GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC                19   \n",
       "18    GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC                19   \n",
       "19  GCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCC                19   \n",
       "20                   TAAAACGCTTACGCCCCCCCCCCCCCCCCC                18   \n",
       "21                TAAAACGCTTACGCCCCCCCCCCCCCCCCCCCC                18   \n",
       "22              TAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC                18   \n",
       "23            TAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC                18   \n",
       "24         GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCC                18   \n",
       "25      GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCC                18   \n",
       "26    GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCC                18   \n",
       "27  GGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCC                18   \n",
       "28                   CTAAAACGCTTACGCCCCCCCCCCCCCCCC                17   \n",
       "29                CTAAAACGCTTACGCCCCCCCCCCCCCCCCCCC                17   \n",
       "30              CTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC                17   \n",
       "31            CTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC                17   \n",
       "32         AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCC                17   \n",
       "33      AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCC                17   \n",
       "34    AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCC                17   \n",
       "35  AGGCTAGCTACTAAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCC                17   \n",
       "36                             AGCTACTAAAACGCTTACGC                 2   \n",
       "37                          AGCTACTAAAACGCTTACGCCCC                 2   \n",
       "38                        AGCTACTAAAACGCTTACGCCCCCC                 2   \n",
       "39                      AGCTACTAAAACGCTTACGCCCCCCCC                 2   \n",
       "40                   CAGCTAGGCTAGCTACTAAAACGCTTACGC                 2   \n",
       "41                CAGCTAGGCTAGCTACTAAAACGCTTACGCCCC                 2   \n",
       "42              CAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCC                 2   \n",
       "43            CAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCCCC                 2   \n",
       "44         CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGC                 2   \n",
       "45      CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGCCCC                 2   \n",
       "46    CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCC                 2   \n",
       "47  CATGCCATGCCAGCTAGGCTAGCTACTAAAACGCTTACGCCCCCCCC                 2   \n",
       "48                   GTTTTAGTAGCTAGCCTAGCTGGCATGGCA                16   \n",
       "49                GTTTTAGTAGCTAGCCTAGCTGGCATGGCATGG                16   \n",
       "50              GTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCA                16   \n",
       "51            GTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCATG                16   \n",
       "52         GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCA                16   \n",
       "53      GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGG                16   \n",
       "54    GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCA                16   \n",
       "55  GGGCGTAAGCGTTTTAGTAGCTAGCCTAGCTGGCATGGCATGGCATG                16   \n",
       "\n",
       "    RHA_size  PAM_disrupted  Proto_disrupted  \n",
       "0          9          False            False  \n",
       "1          9          False            False  \n",
       "2          9          False            False  \n",
       "3          9          False            False  \n",
       "4          0          False            False  \n",
       "5          0          False            False  \n",
       "6          0          False            False  \n",
       "7          0          False            False  \n",
       "8         10          False            False  \n",
       "9         10          False            False  \n",
       "10        10          False            False  \n",
       "11        10          False            False  \n",
       "12         1          False            False  \n",
       "13         1          False            False  \n",
       "14         1          False            False  \n",
       "15         1          False            False  \n",
       "16        11          False            False  \n",
       "17        11          False            False  \n",
       "18        11          False            False  \n",
       "19        11          False            False  \n",
       "20         2          False            False  \n",
       "21         2          False            False  \n",
       "22         2          False            False  \n",
       "23         2          False            False  \n",
       "24        12          False            False  \n",
       "25        12          False            False  \n",
       "26        12          False            False  \n",
       "27        12          False            False  \n",
       "28         3          False            False  \n",
       "29         3          False            False  \n",
       "30         3          False            False  \n",
       "31         3          False            False  \n",
       "32        13          False            False  \n",
       "33        13          False            False  \n",
       "34        13          False            False  \n",
       "35        13          False            False  \n",
       "36         8           True             True  \n",
       "37         8           True             True  \n",
       "38         8           True             True  \n",
       "39         8           True             True  \n",
       "40        18           True             True  \n",
       "41        18           True             True  \n",
       "42        18           True             True  \n",
       "43        18           True             True  \n",
       "44        28           True             True  \n",
       "45        28           True             True  \n",
       "46        28           True             True  \n",
       "47        28           True             True  \n",
       "48         4          False            False  \n",
       "49         4          False            False  \n",
       "50         4          False            False  \n",
       "51         4          False            False  \n",
       "52        14          False            False  \n",
       "53        14          False            False  \n",
       "54        14          False            False  \n",
       "55        14          False            False  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#mutation info\n",
    "seq = 'GGGGGGGGGGGGGGGGGGGGGGGGGGGGGCGTAAGCGTTT(ATGG/)TAGTAGCTAGCCTAGCTGGCATGGCATGGCATGGCATGGCATGGCA'\n",
    "PAM = \"NGG\"\n",
    "#--params\n",
    "RTT_lengths = [10,20,30]\n",
    "PBS_lengths = [10,13, 15,17]\n",
    "orientation = '-'\n",
    "proto_size = 19\n",
    "\n",
    "max_RTT_length = max(RTT_lengths)\n",
    "\n",
    "var_type, wt_w_context, alt_w_context, left_seq, right_seq, ref_seq, alt_seq = primedesign_formatter(seq)\n",
    "mut = mutation(wt_w_context, alt_w_context, left_seq, right_seq, var_type, ref_seq, alt_seq, chrom=None, genome=None)\n",
    "mut.PAM_idx_forward, mut.PAM_idx_rc = eligible_PAM_finder(mut, PAM, max_RTT_length, proto_size)\n",
    "\n",
    "\n",
    "\n",
    "peg_df_plus = pegRNA_generator(mut, PAM, '+', proto_size, RTT_lengths, PBS_lengths)\n",
    "peg_df_minus = pegRNA_generator(mut, PAM, '-', proto_size, RTT_lengths, PBS_lengths)\n",
    "\n",
    "#need to translate PAM indexing for the - strand!!!!\n",
    "#convert to genome coordinates as well...\n",
    "pd.concat((peg_df_plus, peg_df_minus)).reset_index().drop(columns='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 738,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CCT'"
      ]
     },
     "execution_count": 738,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut.wt_forward[54:57]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 681,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGCCATGCCATGCCATGCCATGCCATGCCAGCTAGGCTAGCTACTACCATAAACGCTTACGCCCCCCCCCCCCCCCCCCCCCCCCCCCCC\n",
      "-------------.|||||||||||||||||||---------------------------------------------------------\n",
      "-------------GATGCCATGCCATGCCAGCT---------------------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "alignments2 = aligner.align(mut.wt_rc, \n",
    "                            'GATGCCATGCCATGCCAGCT')\n",
    "print(alignments2[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing on larger set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "p53 = impact[impact['Hugo_Symbol']=='TP53'].drop_duplicates(subset='HGVSc')\n",
    "\n",
    "#or change these to INDELs\n",
    "dels1 = p53[p53['Variant_Type']=='DEL']\n",
    "bad_del_idx = list(dels1[dels1['Tumor_Seq_Allele2']!='-'].index)\n",
    "\n",
    "\n",
    "ins1 = p53[p53['Variant_Type']=='INS']\n",
    "bad_ins_idx = list(ins1[ins1['Reference_Allele']!='-'].index)\n",
    "\n",
    "bad_idxs = bad_del_idx + bad_ins_idx\n",
    "\n",
    "p53_filtered = p53.drop(index=bad_idxs).reset_index().drop(columns=['index'])\n",
    "\n",
    "p53_formatted = df_formatter(p53_filtered, chrom_dict, context_size=100)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_df = p53_formatted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>wt_w_context</th>\n",
       "      <th>alt_w_context</th>\n",
       "      <th>left_context</th>\n",
       "      <th>right_context</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.818G&gt;A</td>\n",
       "      <td>p.R273H</td>\n",
       "      <td>7577020</td>\n",
       "      <td>7577220</td>\n",
       "      <td>TCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTC...</td>\n",
       "      <td>TCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTC...</td>\n",
       "      <td>TCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTC...</td>\n",
       "      <td>GCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGAT...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577108</td>\n",
       "      <td>7577108</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.830G&gt;T</td>\n",
       "      <td>p.C277F</td>\n",
       "      <td>7577008</td>\n",
       "      <td>7577208</td>\n",
       "      <td>TGCTTGCTTACCTCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGA...</td>\n",
       "      <td>TGCTTGCTTACCTCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGA...</td>\n",
       "      <td>TGCTTGCTTACCTCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGA...</td>\n",
       "      <td>AGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACC...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578239</td>\n",
       "      <td>7578239</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.610G&gt;T</td>\n",
       "      <td>p.E204*</td>\n",
       "      <td>7578139</td>\n",
       "      <td>7578339</td>\n",
       "      <td>TAACCCCTCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCG...</td>\n",
       "      <td>TAACCCCTCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCG...</td>\n",
       "      <td>TAACCCCTCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCG...</td>\n",
       "      <td>CACACGCAAATTTCCTTCCACTCGGATAAGATGCTGAGGAGGGGCC...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577121</td>\n",
       "      <td>7577121</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.817C&gt;T</td>\n",
       "      <td>p.R273C</td>\n",
       "      <td>7577021</td>\n",
       "      <td>7577221</td>\n",
       "      <td>CGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCT...</td>\n",
       "      <td>CGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCT...</td>\n",
       "      <td>CGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCT...</td>\n",
       "      <td>CACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATA...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577506</td>\n",
       "      <td>7577506</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.775G&gt;T</td>\n",
       "      <td>p.D259Y</td>\n",
       "      <td>7577406</td>\n",
       "      <td>7577606</td>\n",
       "      <td>GAAGAAATCGGTAAGAGGTGGGCCCAGGGGTCAGAGGCAAGCAGAG...</td>\n",
       "      <td>GAAGAAATCGGTAAGAGGTGGGCCCAGGGGTCAGAGGCAAGCAGAG...</td>\n",
       "      <td>GAAGAAATCGGTAAGAGGTGGGCCCAGGGGTCAGAGGCAAGCAGAG...</td>\n",
       "      <td>TTCCAGTGTGATGATGGTGAGGATGGGCCTCCGGTTCATGCCGCCC...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3000</th>\n",
       "      <td>3000</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577539</td>\n",
       "      <td>7577541</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GGT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.740_742del</td>\n",
       "      <td>p.N247del</td>\n",
       "      <td>7577439</td>\n",
       "      <td>7577641</td>\n",
       "      <td>GAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGC...</td>\n",
       "      <td>GAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGC...</td>\n",
       "      <td>GAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGC...</td>\n",
       "      <td>TCATGCCGCCCATGCAGGAACTGTTACACATGTAGTTGTAGTGGAT...</td>\n",
       "      <td>GGT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3001</th>\n",
       "      <td>3001</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578438</td>\n",
       "      <td>7578438</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>ENST00000269305.4:c.492G&gt;C</td>\n",
       "      <td>p.K164N</td>\n",
       "      <td>7578338</td>\n",
       "      <td>7578538</td>\n",
       "      <td>AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...</td>\n",
       "      <td>AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...</td>\n",
       "      <td>AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...</td>\n",
       "      <td>TTGTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGG...</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3002</th>\n",
       "      <td>3002</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577531</td>\n",
       "      <td>7577532</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GG</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.749_750delCC</td>\n",
       "      <td>p.P250Hfs*13</td>\n",
       "      <td>7577431</td>\n",
       "      <td>7577632</td>\n",
       "      <td>AGGGGTCAGAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGC...</td>\n",
       "      <td>AGGGGTCAGAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGC...</td>\n",
       "      <td>AGGGGTCAGAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGC...</td>\n",
       "      <td>GCCTCCGGTTCATGCCGCCCATGCAGGAACTGTTACACATGTAGTT...</td>\n",
       "      <td>GG</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3003</th>\n",
       "      <td>3003</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578421</td>\n",
       "      <td>7578440</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>GTCATGTGCTGTGACTGCTT</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4:c.490_509del</td>\n",
       "      <td>p.K164Gfs*10</td>\n",
       "      <td>7578321</td>\n",
       "      <td>7578540</td>\n",
       "      <td>GGGGACCCTGGGCAACCAGCCCTGTCGTCTCTCCAGCCCCAGCTGC...</td>\n",
       "      <td>GGGGACCCTGGGCAACCAGCCCTGTCGTCTCTCCAGCCCCAGCTGC...</td>\n",
       "      <td>GGGGACCCTGGGCAACCAGCCCTGTCGTCTCTCCAGCCCCAGCTGC...</td>\n",
       "      <td>GTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGGAA...</td>\n",
       "      <td>GTCATGTGCTGTGACTGCTT</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3004</th>\n",
       "      <td>3004</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579348</td>\n",
       "      <td>7579348</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.339C&gt;A</td>\n",
       "      <td>p.F113L</td>\n",
       "      <td>7579248</td>\n",
       "      <td>7579448</td>\n",
       "      <td>AAATGCAGGGGGATACGGCCAGGCATTGAAGTCTCATGGAAGCCAG...</td>\n",
       "      <td>AAATGCAGGGGGATACGGCCAGGCATTGAAGTCTCATGGAAGCCAG...</td>\n",
       "      <td>AAATGCAGGGGGATACGGCCAGGCATTGAAGTCTCATGGAAGCCAG...</td>\n",
       "      <td>AAGCCCAGACGGAAACCGTAGCTGCCCTGGTAGGTTTTCTGGGAAG...</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3005 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0         0        TP53         17         7577120       7577120   \n",
       "1         1        TP53         17         7577108       7577108   \n",
       "2         2        TP53         17         7578239       7578239   \n",
       "3         3        TP53         17         7577121       7577121   \n",
       "4         4        TP53         17         7577506       7577506   \n",
       "...     ...         ...        ...             ...           ...   \n",
       "3000   3000        TP53         17         7577539       7577541   \n",
       "3001   3001        TP53         17         7578438       7578438   \n",
       "3002   3002        TP53         17         7577531       7577532   \n",
       "3003   3003        TP53         17         7578421       7578440   \n",
       "3004   3004        TP53         17         7579348       7579348   \n",
       "\n",
       "             Consequence Variant_Classification Variant_Type  \\\n",
       "0       missense_variant      Missense_Mutation          SNP   \n",
       "1       missense_variant      Missense_Mutation          SNP   \n",
       "2            stop_gained      Nonsense_Mutation          SNP   \n",
       "3       missense_variant      Missense_Mutation          SNP   \n",
       "4       missense_variant      Missense_Mutation          SNP   \n",
       "...                  ...                    ...          ...   \n",
       "3000    inframe_deletion           In_Frame_Del          DEL   \n",
       "3001    missense_variant      Missense_Mutation          SNP   \n",
       "3002  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "3003  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "3004    missense_variant      Missense_Mutation          SNP   \n",
       "\n",
       "          Reference_Allele Tumor_Seq_Allele2  \\\n",
       "0                        C                 T   \n",
       "1                        C                 A   \n",
       "2                        C                 A   \n",
       "3                        G                 A   \n",
       "4                        C                 A   \n",
       "...                    ...               ...   \n",
       "3000                   GGT                 -   \n",
       "3001                     C                 G   \n",
       "3002                    GG                 -   \n",
       "3003  GTCATGTGCTGTGACTGCTT                 -   \n",
       "3004                     G                 T   \n",
       "\n",
       "                                 HGVSc   HGVSp_Short  seq_start  seq_end  \\\n",
       "0           ENST00000269305.4:c.818G>A       p.R273H    7577020  7577220   \n",
       "1           ENST00000269305.4:c.830G>T       p.C277F    7577008  7577208   \n",
       "2           ENST00000269305.4:c.610G>T       p.E204*    7578139  7578339   \n",
       "3           ENST00000269305.4:c.817C>T       p.R273C    7577021  7577221   \n",
       "4           ENST00000269305.4:c.775G>T       p.D259Y    7577406  7577606   \n",
       "...                                ...           ...        ...      ...   \n",
       "3000    ENST00000269305.4:c.740_742del     p.N247del    7577439  7577641   \n",
       "3001        ENST00000269305.4:c.492G>C       p.K164N    7578338  7578538   \n",
       "3002  ENST00000269305.4:c.749_750delCC  p.P250Hfs*13    7577431  7577632   \n",
       "3003    ENST00000269305.4:c.490_509del  p.K164Gfs*10    7578321  7578540   \n",
       "3004        ENST00000269305.4:c.339C>A       p.F113L    7579248  7579448   \n",
       "\n",
       "                                           wt_w_context  \\\n",
       "0     TCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTC...   \n",
       "1     TGCTTGCTTACCTCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGA...   \n",
       "2     TAACCCCTCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCG...   \n",
       "3     CGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCT...   \n",
       "4     GAAGAAATCGGTAAGAGGTGGGCCCAGGGGTCAGAGGCAAGCAGAG...   \n",
       "...                                                 ...   \n",
       "3000  GAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGC...   \n",
       "3001  AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...   \n",
       "3002  AGGGGTCAGAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGC...   \n",
       "3003  GGGGACCCTGGGCAACCAGCCCTGTCGTCTCTCCAGCCCCAGCTGC...   \n",
       "3004  AAATGCAGGGGGATACGGCCAGGCATTGAAGTCTCATGGAAGCCAG...   \n",
       "\n",
       "                                          alt_w_context  \\\n",
       "0     TCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTC...   \n",
       "1     TGCTTGCTTACCTCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGA...   \n",
       "2     TAACCCCTCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCG...   \n",
       "3     CGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCT...   \n",
       "4     GAAGAAATCGGTAAGAGGTGGGCCCAGGGGTCAGAGGCAAGCAGAG...   \n",
       "...                                                 ...   \n",
       "3000  GAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGC...   \n",
       "3001  AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...   \n",
       "3002  AGGGGTCAGAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGC...   \n",
       "3003  GGGGACCCTGGGCAACCAGCCCTGTCGTCTCTCCAGCCCCAGCTGC...   \n",
       "3004  AAATGCAGGGGGATACGGCCAGGCATTGAAGTCTCATGGAAGCCAG...   \n",
       "\n",
       "                                           left_context  \\\n",
       "0     TCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTC...   \n",
       "1     TGCTTGCTTACCTCGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGA...   \n",
       "2     TAACCCCTCCTCCCAGAGACCCCAGTTGCAAACCAGACCTCAGGCG...   \n",
       "3     CGCTTAGTGCTCCCTGGGGGCAGCTCGTGGTGAGGCTCCCCTTTCT...   \n",
       "4     GAAGAAATCGGTAAGAGGTGGGCCCAGGGGTCAGAGGCAAGCAGAG...   \n",
       "...                                                 ...   \n",
       "3000  GAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGCAGGGTGGC...   \n",
       "3001  AGCCCTGTCGTCTCTCCAGCCCCAGCTGCTCACCATCGCTATCTGA...   \n",
       "3002  AGGGGTCAGAGGCAAGCAGAGGCTGGGGCACAGCAGGCCAGTGTGC...   \n",
       "3003  GGGGACCCTGGGCAACCAGCCCTGTCGTCTCTCCAGCCCCAGCTGC...   \n",
       "3004  AAATGCAGGGGGATACGGCCAGGCATTGAAGTCTCATGGAAGCCAG...   \n",
       "\n",
       "                                          right_context                   REF  \\\n",
       "0     GCACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGAT...                     C   \n",
       "1     AGGCACAAACACGCACCTCAAAGCTGTTCCGTCCCAGTAGATTACC...                     C   \n",
       "2     CACACGCAAATTTCCTTCCACTCGGATAAGATGCTGAGGAGGGGCC...                     C   \n",
       "3     CACCTCAAAGCTGTTCCGTCCCAGTAGATTACCACTACTCAGGATA...                     G   \n",
       "4     TTCCAGTGTGATGATGGTGAGGATGGGCCTCCGGTTCATGCCGCCC...                     C   \n",
       "...                                                 ...                   ...   \n",
       "3000  TCATGCCGCCCATGCAGGAACTGTTACACATGTAGTTGTAGTGGAT...                   GGT   \n",
       "3001  TTGTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGG...                     C   \n",
       "3002  GCCTCCGGTTCATGCCGCCCATGCAGGAACTGTTACACATGTAGTT...                    GG   \n",
       "3003  GTAGATGGCCATGGCGCGGACGCGGGTGCCGGGCGGGGGTGTGGAA...  GTCATGTGCTGTGACTGCTT   \n",
       "3004  AAGCCCAGACGGAAACCGTAGCTGCCCTGGTAGGTTTTCTGGGAAG...                     G   \n",
       "\n",
       "     ALT  \n",
       "0      T  \n",
       "1      A  \n",
       "2      A  \n",
       "3      A  \n",
       "4      A  \n",
       "...   ..  \n",
       "3000      \n",
       "3001   G  \n",
       "3002      \n",
       "3003      \n",
       "3004   T  \n",
       "\n",
       "[3005 rows x 20 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--params\n",
    "RTT_lengths = [10,20,30]\n",
    "PBS_lengths = [10,13, 15,17]\n",
    "orientation = '-'\n",
    "proto_size = 19\n",
    "PAM = \"NGG\"\n",
    "\n",
    "combined_peg_dfs = []\n",
    "\n",
    "for i, val in input_df.iterrows():\n",
    "\n",
    "\n",
    "    mut = mutation(val['wt_w_context'], val['alt_w_context'], val['left_context'], val['right_context'], val['Variant_Type'], val['REF'], val['ALT'], chrom=None, genome=None)\n",
    "    mut.PAM_idx_forward, mut.PAM_idx_rc = eligible_PAM_finder(mut, PAM, max(RTT_lengths), proto_size)\n",
    "\n",
    "    peg_df_plus = pegRNA_generator(mut, PAM, '+', proto_size, RTT_lengths, PBS_lengths)\n",
    "    peg_df_minus = pegRNA_generator(mut, PAM, '-', proto_size, RTT_lengths, PBS_lengths)\n",
    "\n",
    "    #need to translate PAM indexing for the - strand!!!!\n",
    "    #convert to genome coordinates as well...\n",
    "    peg_df = pd.concat((peg_df_plus, peg_df_minus))\n",
    "    peg_df['mutation_idx'] = i\n",
    "\n",
    "    combined_peg_dfs.append(peg_df)\n",
    "\n",
    "peg_df = pd.concat(combined_peg_dfs).set_index('mutation_idx').reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>PAM_start</th>\n",
       "      <th>PAM_end</th>\n",
       "      <th>PAM</th>\n",
       "      <th>PAM_strand</th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>RTT</th>\n",
       "      <th>RTT_length</th>\n",
       "      <th>PBS</th>\n",
       "      <th>PBS_length</th>\n",
       "      <th>RTT_PBS</th>\n",
       "      <th>Distance_to_nick</th>\n",
       "      <th>RHA_size</th>\n",
       "      <th>PAM_disrupted</th>\n",
       "      <th>Proto_disrupted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>84</td>\n",
       "      <td>87</td>\n",
       "      <td>AGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGC</td>\n",
       "      <td>10</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGC</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>84</td>\n",
       "      <td>87</td>\n",
       "      <td>AGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGCGCA</td>\n",
       "      <td>13</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCA</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>84</td>\n",
       "      <td>87</td>\n",
       "      <td>AGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGCGCACA</td>\n",
       "      <td>15</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACA</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>84</td>\n",
       "      <td>87</td>\n",
       "      <td>AGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGCGCACAGA</td>\n",
       "      <td>17</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGA</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>84</td>\n",
       "      <td>87</td>\n",
       "      <td>AGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>TTTGAGGTGCATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>AGAGACCGGC</td>\n",
       "      <td>10</td>\n",
       "      <td>TTTGAGGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGC</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122567</th>\n",
       "      <td>3004</td>\n",
       "      <td>107</td>\n",
       "      <td>104</td>\n",
       "      <td>GGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>CAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>20</td>\n",
       "      <td>CGGAAACCGTAGCTGCC</td>\n",
       "      <td>17</td>\n",
       "      <td>CAGAATGCAATAAGCCCAGACGGAAACCGTAGCTGCC</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122568</th>\n",
       "      <td>3004</td>\n",
       "      <td>107</td>\n",
       "      <td>104</td>\n",
       "      <td>GGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>30</td>\n",
       "      <td>CGGAAACCGT</td>\n",
       "      <td>10</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGT</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122569</th>\n",
       "      <td>3004</td>\n",
       "      <td>107</td>\n",
       "      <td>104</td>\n",
       "      <td>GGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>30</td>\n",
       "      <td>CGGAAACCGTAGC</td>\n",
       "      <td>13</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGC</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122570</th>\n",
       "      <td>3004</td>\n",
       "      <td>107</td>\n",
       "      <td>104</td>\n",
       "      <td>GGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>30</td>\n",
       "      <td>CGGAAACCGTAGCTG</td>\n",
       "      <td>15</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGCTG</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122571</th>\n",
       "      <td>3004</td>\n",
       "      <td>107</td>\n",
       "      <td>104</td>\n",
       "      <td>GGG</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCAGCTACGGTTTCCGTCT</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGA</td>\n",
       "      <td>30</td>\n",
       "      <td>CGGAAACCGTAGCTGCC</td>\n",
       "      <td>17</td>\n",
       "      <td>TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGCTGCC</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122572 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mutation_idx  PAM_start  PAM_end  PAM PAM_strand  \\\n",
       "0                  0         84       87  AGG          +   \n",
       "1                  0         84       87  AGG          +   \n",
       "2                  0         84       87  AGG          +   \n",
       "3                  0         84       87  AGG          +   \n",
       "4                  0         84       87  AGG          +   \n",
       "...              ...        ...      ...  ...        ...   \n",
       "122567          3004        107      104  GGG          -   \n",
       "122568          3004        107      104  GGG          -   \n",
       "122569          3004        107      104  GGG          -   \n",
       "122570          3004        107      104  GGG          -   \n",
       "122571          3004        107      104  GGG          -   \n",
       "\n",
       "                 Protospacer                             RTT  RTT_length  \\\n",
       "0       GCTGTGCGCCGGTCTCTCCC            ATGTTTGTGCCTGTCCTGGG          20   \n",
       "1       GCTGTGCGCCGGTCTCTCCC            ATGTTTGTGCCTGTCCTGGG          20   \n",
       "2       GCTGTGCGCCGGTCTCTCCC            ATGTTTGTGCCTGTCCTGGG          20   \n",
       "3       GCTGTGCGCCGGTCTCTCCC            ATGTTTGTGCCTGTCCTGGG          20   \n",
       "4       GCTGTGCGCCGGTCTCTCCC  TTTGAGGTGCATGTTTGTGCCTGTCCTGGG          30   \n",
       "...                      ...                             ...         ...   \n",
       "122567  GGCAGCTACGGTTTCCGTCT            CAGAATGCAATAAGCCCAGA          20   \n",
       "122568  GGCAGCTACGGTTTCCGTCT  TTGGCTGTCCCAGAATGCAATAAGCCCAGA          30   \n",
       "122569  GGCAGCTACGGTTTCCGTCT  TTGGCTGTCCCAGAATGCAATAAGCCCAGA          30   \n",
       "122570  GGCAGCTACGGTTTCCGTCT  TTGGCTGTCCCAGAATGCAATAAGCCCAGA          30   \n",
       "122571  GGCAGCTACGGTTTCCGTCT  TTGGCTGTCCCAGAATGCAATAAGCCCAGA          30   \n",
       "\n",
       "                      PBS  PBS_length  \\\n",
       "0              AGAGACCGGC          10   \n",
       "1           AGAGACCGGCGCA          13   \n",
       "2         AGAGACCGGCGCACA          15   \n",
       "3       AGAGACCGGCGCACAGA          17   \n",
       "4              AGAGACCGGC          10   \n",
       "...                   ...         ...   \n",
       "122567  CGGAAACCGTAGCTGCC          17   \n",
       "122568         CGGAAACCGT          10   \n",
       "122569      CGGAAACCGTAGC          13   \n",
       "122570    CGGAAACCGTAGCTG          15   \n",
       "122571  CGGAAACCGTAGCTGCC          17   \n",
       "\n",
       "                                                RTT_PBS  Distance_to_nick  \\\n",
       "0                        ATGTTTGTGCCTGTCCTGGGAGAGACCGGC                19   \n",
       "1                     ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCA                19   \n",
       "2                   ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACA                19   \n",
       "3                 ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGA                19   \n",
       "4              TTTGAGGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGC                19   \n",
       "...                                                 ...               ...   \n",
       "122567            CAGAATGCAATAAGCCCAGACGGAAACCGTAGCTGCC                 9   \n",
       "122568         TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGT                 9   \n",
       "122569      TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGC                 9   \n",
       "122570    TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGCTG                 9   \n",
       "122571  TTGGCTGTCCCAGAATGCAATAAGCCCAGACGGAAACCGTAGCTGCC                 9   \n",
       "\n",
       "        RHA_size  PAM_disrupted  Proto_disrupted  \n",
       "0              0          False            False  \n",
       "1              0          False            False  \n",
       "2              0          False            False  \n",
       "3              0          False            False  \n",
       "4             10          False            False  \n",
       "...          ...            ...              ...  \n",
       "122567        10          False            False  \n",
       "122568        20          False            False  \n",
       "122569        20          False            False  \n",
       "122570        20          False            False  \n",
       "122571        20          False            False  \n",
       "\n",
       "[122572 rows x 15 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peg_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sensor design\n",
    "\n",
    "- Put it in reverse complement orientation\n",
    "- 5 nt of context prior to protospacer (adjustable)\n",
    "- Also output the edited sensor\n",
    "- Include whether the sensor is large enough to accomodate the desired edit..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PEGG Score\n",
    "\n",
    "- Update PEGG score to reflect the results of my dataset (include PAM/protospacer disrupting edits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# pegRNA visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 758,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>...</th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>RTT</th>\n",
       "      <th>RTT_length</th>\n",
       "      <th>PBS</th>\n",
       "      <th>PBS_length</th>\n",
       "      <th>RTT_PBS</th>\n",
       "      <th>Distance_to_nick</th>\n",
       "      <th>RHA_size</th>\n",
       "      <th>PAM_disrupted</th>\n",
       "      <th>Proto_disrupted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGC</td>\n",
       "      <td>10</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGC</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGCGCA</td>\n",
       "      <td>13</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCA</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGCGCACA</td>\n",
       "      <td>15</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACA</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>AGAGACCGGCGCACAGA</td>\n",
       "      <td>17</td>\n",
       "      <td>ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGA</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGTGCGCCGGTCTCTCCC</td>\n",
       "      <td>TTTGAGGTGCATGTTTGTGCCTGTCCTGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>AGAGACCGGC</td>\n",
       "      <td>10</td>\n",
       "      <td>TTTGAGGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGC</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34639</th>\n",
       "      <td>999</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579444</td>\n",
       "      <td>7579444</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCACCAGCAGCTCCTACAC</td>\n",
       "      <td>CTGGTGCAGGGGCCGCCGGG</td>\n",
       "      <td>20</td>\n",
       "      <td>TAGGAGCTGCTGGTGCA</td>\n",
       "      <td>17</td>\n",
       "      <td>CTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGGTGCA</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34640</th>\n",
       "      <td>999</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579444</td>\n",
       "      <td>7579444</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCACCAGCAGCTCCTACAC</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>TAGGAGCTGC</td>\n",
       "      <td>10</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGC</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34641</th>\n",
       "      <td>999</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579444</td>\n",
       "      <td>7579444</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCACCAGCAGCTCCTACAC</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>TAGGAGCTGCTGG</td>\n",
       "      <td>13</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGG</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34642</th>\n",
       "      <td>999</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579444</td>\n",
       "      <td>7579444</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCACCAGCAGCTCCTACAC</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>TAGGAGCTGCTGGTG</td>\n",
       "      <td>15</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGGTG</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34643</th>\n",
       "      <td>999</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7579444</td>\n",
       "      <td>7579444</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>GGCACCAGCAGCTCCTACAC</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>TAGGAGCTGCTGGTGCA</td>\n",
       "      <td>17</td>\n",
       "      <td>CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGGTGCA</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34644 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation_idx Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0                 0        TP53         17         7577120       7577120   \n",
       "1                 0        TP53         17         7577120       7577120   \n",
       "2                 0        TP53         17         7577120       7577120   \n",
       "3                 0        TP53         17         7577120       7577120   \n",
       "4                 0        TP53         17         7577120       7577120   \n",
       "...             ...         ...        ...             ...           ...   \n",
       "34639           999        TP53         17         7579444       7579444   \n",
       "34640           999        TP53         17         7579444       7579444   \n",
       "34641           999        TP53         17         7579444       7579444   \n",
       "34642           999        TP53         17         7579444       7579444   \n",
       "34643           999        TP53         17         7579444       7579444   \n",
       "\n",
       "              Consequence Variant_Classification Variant_Type  \\\n",
       "0        missense_variant      Missense_Mutation          SNP   \n",
       "1        missense_variant      Missense_Mutation          SNP   \n",
       "2        missense_variant      Missense_Mutation          SNP   \n",
       "3        missense_variant      Missense_Mutation          SNP   \n",
       "4        missense_variant      Missense_Mutation          SNP   \n",
       "...                   ...                    ...          ...   \n",
       "34639  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "34640  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "34641  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "34642  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "34643  frameshift_variant        Frame_Shift_Del          DEL   \n",
       "\n",
       "      Reference_Allele Tumor_Seq_Allele2  ...           Protospacer  \\\n",
       "0                    C                 T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "1                    C                 T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "2                    C                 T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "3                    C                 T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "4                    C                 T  ...  GCTGTGCGCCGGTCTCTCCC   \n",
       "...                ...               ...  ...                   ...   \n",
       "34639                T                 -  ...  GGCACCAGCAGCTCCTACAC   \n",
       "34640                T                 -  ...  GGCACCAGCAGCTCCTACAC   \n",
       "34641                T                 -  ...  GGCACCAGCAGCTCCTACAC   \n",
       "34642                T                 -  ...  GGCACCAGCAGCTCCTACAC   \n",
       "34643                T                 -  ...  GGCACCAGCAGCTCCTACAC   \n",
       "\n",
       "                                  RTT  RTT_length                PBS  \\\n",
       "0                ATGTTTGTGCCTGTCCTGGG          20         AGAGACCGGC   \n",
       "1                ATGTTTGTGCCTGTCCTGGG          20      AGAGACCGGCGCA   \n",
       "2                ATGTTTGTGCCTGTCCTGGG          20    AGAGACCGGCGCACA   \n",
       "3                ATGTTTGTGCCTGTCCTGGG          20  AGAGACCGGCGCACAGA   \n",
       "4      TTTGAGGTGCATGTTTGTGCCTGTCCTGGG          30         AGAGACCGGC   \n",
       "...                               ...         ...                ...   \n",
       "34639            CTGGTGCAGGGGCCGCCGGG          20  TAGGAGCTGCTGGTGCA   \n",
       "34640  CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG          30         TAGGAGCTGC   \n",
       "34641  CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG          30      TAGGAGCTGCTGG   \n",
       "34642  CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG          30    TAGGAGCTGCTGGTG   \n",
       "34643  CAGGAGGGGGCTGGTGCAGGGGCCGCCGGG          30  TAGGAGCTGCTGGTGCA   \n",
       "\n",
       "      PBS_length                                          RTT_PBS  \\\n",
       "0             10                   ATGTTTGTGCCTGTCCTGGGAGAGACCGGC   \n",
       "1             13                ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCA   \n",
       "2             15              ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACA   \n",
       "3             17            ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGA   \n",
       "4             10         TTTGAGGTGCATGTTTGTGCCTGTCCTGGGAGAGACCGGC   \n",
       "...          ...                                              ...   \n",
       "34639         17            CTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGGTGCA   \n",
       "34640         10         CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGC   \n",
       "34641         13      CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGG   \n",
       "34642         15    CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGGTG   \n",
       "34643         17  CAGGAGGGGGCTGGTGCAGGGGCCGCCGGGTAGGAGCTGCTGGTGCA   \n",
       "\n",
       "      Distance_to_nick RHA_size PAM_disrupted Proto_disrupted  \n",
       "0                   19        0         False           False  \n",
       "1                   19        0         False           False  \n",
       "2                   19        0         False           False  \n",
       "3                   19        0         False           False  \n",
       "4                   19       10         False           False  \n",
       "...                ...      ...           ...             ...  \n",
       "34639                1       19          True            True  \n",
       "34640                1       29          True            True  \n",
       "34641                1       29          True            True  \n",
       "34642                1       29          True            True  \n",
       "34643                1       29          True            True  \n",
       "\n",
       "[34644 rows x 34 columns]"
      ]
     },
     "execution_count": 758,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_df = input_df.rename(columns = {'index':'mutation_idx'})\n",
    "\n",
    "df = pd.merge(input_df, peg_df, on='mutation_idx')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "class bcolors:\n",
    "    Red = '\\033[91m'\n",
    "    Green = '\\033[92m'\n",
    "    Blue = '\\033[94m'\n",
    "    Cyan = '\\033[96m'\n",
    "    White = '\\033[97m'\n",
    "    Yellow = '\\033[93m'\n",
    "    Magenta = '\\033[95m'\n",
    "    Grey = '\\033[90m'\n",
    "    ENDC = '\\033[0m'\n",
    "\n",
    "from Bio import pairwise2\n",
    "\n",
    "def visualize_pegRNA(df, i):\n",
    "    \"\"\"\n",
    "    This one visualizes the binding of the protospacer and the 3' extension to the ref seq\n",
    "    It also says the rank and PRIDICT score (intended/unintended editing)\n",
    "\n",
    "    Inputs:\n",
    "    df = datframe (loaded from pegRNA_dataframe.csv)\n",
    "    i = index (row/pegRNA that you want to visualize the mutation from;\n",
    "            will change depending on the strand (Reverse or Forward) of the pegRNA)\n",
    "    \"\"\"\n",
    "\n",
    "    if df['PAM_strand'][i] == '+':\n",
    "        refseq = df['wt_w_context'][i]\n",
    "    else:\n",
    "        refseq = df['wt_w_context'][i]\n",
    "        refseq = str(Bio.Seq.Seq(refseq).reverse_complement())\n",
    "        \n",
    "\n",
    "\n",
    "    proto = df['Protospacer'][i]\n",
    "\n",
    "    #print a bunch of info about the pegRNA\n",
    "    #print( bcolors.ENDC + 'pegRNA Rank = ' + str(df['rank'][i]))\n",
    "    print('Target Strand = ' + df['PAM_strand'][i])\n",
    "    print(str(df['Variant_Type'][i]) +': ' + str(df['REF'][i]) + '/' + str(df['ALT'][i]))\n",
    "    print('PBS length = ' + str(df['PBS_length'][i]) + ' ; RT length = '+ str(df['RTT_length'][i]) + '; RT overhang = ' + str(df['RHA_size'][i]))\n",
    "\n",
    "\n",
    "    proto_loc = refseq.find(proto[1:])\n",
    "\n",
    "    #RT = Bio.Seq.Seq(df['RTrevcomp'][i]).reverse_complement()\n",
    "    ext = Bio.Seq.Seq(df['RTT_PBS'][i]).reverse_complement()\n",
    "\n",
    "    first = pairwise2.align.localms(refseq, ext, 2,0, -3, -0.2)[0]\n",
    "    a=format_alignment(*first, full_sequences=True)\n",
    "\n",
    "    start = len(first.seqB) + 1\n",
    "    end = 2*len(first.seqB) + 1\n",
    "    pipes = a[start:end]\n",
    "\n",
    "    #translating first.seqB into complement (the true 3' extension)\n",
    "    dict_comp = {'-':'-', 'A':'T', 'T':'A', 'C':'G', 'G':'C'}\n",
    "    comp1 = [dict_comp[i] for i in first.seqB]\n",
    "    complement = \"\".join(comp1)\n",
    "\n",
    "    print(complement + '\\n' + pipes + '\\n' + first.seqA[0:proto_loc] + bcolors.Blue + first.seqA[proto_loc:proto_loc+19] + bcolors.Magenta + first.seqA[proto_loc+19:proto_loc+22] + bcolors.ENDC + first.seqA[proto_loc+22:] )\n",
    "    print(Bio.Seq.Seq(first.seqA).complement())\n",
    "\n",
    "    second = pairwise2.align.localms(refseq, proto, 2,0, -3, -0.2)[0]\n",
    "    b=format_alignment(*second, full_sequences=True)\n",
    "\n",
    "    start2 = len(second.seqB) + 1\n",
    "    end2 = 2*len(second.seqB) + 1\n",
    "    pipes2 = b[start:end]\n",
    "\n",
    "    #print(' ' + pipes2)\n",
    "    print(pipes2)\n",
    "    print(second.seqB)\n",
    "\n",
    "    print(bcolors.Blue + 'Protospacer (G+19 bp)')\n",
    "    print(bcolors.Magenta + 'PAM')\n",
    "    print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/5g/xxjcy2g531n02hnyb6d8qdsr0000gn/T/ipykernel_79410/3693921967.py\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mvisualize_pegRNA\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m20228\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'df' is not defined"
     ]
    }
   ],
   "source": [
    "visualize_pegRNA(df, 20228)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 799,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 799,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('ACGTGGTCGTCGAGGAT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 818,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>...</th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>RTT</th>\n",
       "      <th>RTT_length</th>\n",
       "      <th>PBS</th>\n",
       "      <th>PBS_length</th>\n",
       "      <th>RTT_PBS</th>\n",
       "      <th>Distance_to_nick</th>\n",
       "      <th>RHA_size</th>\n",
       "      <th>PAM_disrupted</th>\n",
       "      <th>Proto_disrupted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27228</th>\n",
       "      <td>792</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577035</td>\n",
       "      <td>7577036</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCTCGCTTAGTGCTCCCTG</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAG</td>\n",
       "      <td>20</td>\n",
       "      <td>GGAGCACTAA</td>\n",
       "      <td>10</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAGGGAGCACTAA</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27229</th>\n",
       "      <td>792</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577035</td>\n",
       "      <td>7577036</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCTCGCTTAGTGCTCCCTG</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAG</td>\n",
       "      <td>20</td>\n",
       "      <td>GGAGCACTAAGCG</td>\n",
       "      <td>13</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAGGGAGCACTAAGCG</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27230</th>\n",
       "      <td>792</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577035</td>\n",
       "      <td>7577036</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCTCGCTTAGTGCTCCCTG</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAG</td>\n",
       "      <td>20</td>\n",
       "      <td>GGAGCACTAAGCGAG</td>\n",
       "      <td>15</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAGGGAGCACTAAGCGAG</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27231</th>\n",
       "      <td>792</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577035</td>\n",
       "      <td>7577036</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCTCGCTTAGTGCTCCCTG</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAG</td>\n",
       "      <td>20</td>\n",
       "      <td>GGAGCACTAAGCGAGGT</td>\n",
       "      <td>17</td>\n",
       "      <td>CACCACGAGCTGCCCCCCAGGGAGCACTAAGCGAGGT</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27232</th>\n",
       "      <td>792</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577035</td>\n",
       "      <td>7577036</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCTCGCTTAGTGCTCCCTG</td>\n",
       "      <td>AGGGGAGCCTCACCACGAGCTGCCCCCCAG</td>\n",
       "      <td>30</td>\n",
       "      <td>GGAGCACTAA</td>\n",
       "      <td>10</td>\n",
       "      <td>AGGGGAGCCTCACCACGAGCTGCCCCCCAGGGAGCACTAA</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33963</th>\n",
       "      <td>982</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578479</td>\n",
       "      <td>7578480</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCTGCCCTGTGCAGCTGT</td>\n",
       "      <td>GGATGTGGAATCAACCCACA</td>\n",
       "      <td>20</td>\n",
       "      <td>GCTGCACAGGGCAGGTC</td>\n",
       "      <td>17</td>\n",
       "      <td>GGATGTGGAATCAACCCACAGCTGCACAGGGCAGGTC</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33964</th>\n",
       "      <td>982</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578479</td>\n",
       "      <td>7578480</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCTGCCCTGTGCAGCTGT</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACA</td>\n",
       "      <td>30</td>\n",
       "      <td>GCTGCACAGG</td>\n",
       "      <td>10</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGG</td>\n",
       "      <td>17</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33965</th>\n",
       "      <td>982</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578479</td>\n",
       "      <td>7578480</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCTGCCCTGTGCAGCTGT</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACA</td>\n",
       "      <td>30</td>\n",
       "      <td>GCTGCACAGGGCA</td>\n",
       "      <td>13</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGGGCA</td>\n",
       "      <td>17</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33966</th>\n",
       "      <td>982</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578479</td>\n",
       "      <td>7578480</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCTGCCCTGTGCAGCTGT</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACA</td>\n",
       "      <td>30</td>\n",
       "      <td>GCTGCACAGGGCAGG</td>\n",
       "      <td>15</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGGGCAGG</td>\n",
       "      <td>17</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33967</th>\n",
       "      <td>982</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578479</td>\n",
       "      <td>7578480</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>GACCTGCCCTGTGCAGCTGT</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACA</td>\n",
       "      <td>30</td>\n",
       "      <td>GCTGCACAGGGCAGGTC</td>\n",
       "      <td>17</td>\n",
       "      <td>GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGGGCAGGTC</td>\n",
       "      <td>17</td>\n",
       "      <td>12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation_idx Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "27228           792        TP53         17         7577035       7577036   \n",
       "27229           792        TP53         17         7577035       7577036   \n",
       "27230           792        TP53         17         7577035       7577036   \n",
       "27231           792        TP53         17         7577035       7577036   \n",
       "27232           792        TP53         17         7577035       7577036   \n",
       "...             ...         ...        ...             ...           ...   \n",
       "33963           982        TP53         17         7578479       7578480   \n",
       "33964           982        TP53         17         7578479       7578480   \n",
       "33965           982        TP53         17         7578479       7578480   \n",
       "33966           982        TP53         17         7578479       7578480   \n",
       "33967           982        TP53         17         7578479       7578480   \n",
       "\n",
       "              Consequence Variant_Classification Variant_Type  \\\n",
       "27228  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "27229  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "27230  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "27231  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "27232  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "...                   ...                    ...          ...   \n",
       "33963  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "33964  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "33965  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "33966  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "33967  frameshift_variant        Frame_Shift_Ins          INS   \n",
       "\n",
       "      Reference_Allele Tumor_Seq_Allele2  ...           Protospacer  \\\n",
       "27228                -                 G  ...  GCCTCGCTTAGTGCTCCCTG   \n",
       "27229                -                 G  ...  GCCTCGCTTAGTGCTCCCTG   \n",
       "27230                -                 G  ...  GCCTCGCTTAGTGCTCCCTG   \n",
       "27231                -                 G  ...  GCCTCGCTTAGTGCTCCCTG   \n",
       "27232                -                 G  ...  GCCTCGCTTAGTGCTCCCTG   \n",
       "...                ...               ...  ...                   ...   \n",
       "33963                -                 A  ...  GACCTGCCCTGTGCAGCTGT   \n",
       "33964                -                 A  ...  GACCTGCCCTGTGCAGCTGT   \n",
       "33965                -                 A  ...  GACCTGCCCTGTGCAGCTGT   \n",
       "33966                -                 A  ...  GACCTGCCCTGTGCAGCTGT   \n",
       "33967                -                 A  ...  GACCTGCCCTGTGCAGCTGT   \n",
       "\n",
       "                                  RTT  RTT_length                PBS  \\\n",
       "27228            CACCACGAGCTGCCCCCCAG          20         GGAGCACTAA   \n",
       "27229            CACCACGAGCTGCCCCCCAG          20      GGAGCACTAAGCG   \n",
       "27230            CACCACGAGCTGCCCCCCAG          20    GGAGCACTAAGCGAG   \n",
       "27231            CACCACGAGCTGCCCCCCAG          20  GGAGCACTAAGCGAGGT   \n",
       "27232  AGGGGAGCCTCACCACGAGCTGCCCCCCAG          30         GGAGCACTAA   \n",
       "...                               ...         ...                ...   \n",
       "33963            GGATGTGGAATCAACCCACA          20  GCTGCACAGGGCAGGTC   \n",
       "33964  GCCGGGCGGGGGATGTGGAATCAACCCACA          30         GCTGCACAGG   \n",
       "33965  GCCGGGCGGGGGATGTGGAATCAACCCACA          30      GCTGCACAGGGCA   \n",
       "33966  GCCGGGCGGGGGATGTGGAATCAACCCACA          30    GCTGCACAGGGCAGG   \n",
       "33967  GCCGGGCGGGGGATGTGGAATCAACCCACA          30  GCTGCACAGGGCAGGTC   \n",
       "\n",
       "      PBS_length                                          RTT_PBS  \\\n",
       "27228         10                   CACCACGAGCTGCCCCCCAGGGAGCACTAA   \n",
       "27229         13                CACCACGAGCTGCCCCCCAGGGAGCACTAAGCG   \n",
       "27230         15              CACCACGAGCTGCCCCCCAGGGAGCACTAAGCGAG   \n",
       "27231         17            CACCACGAGCTGCCCCCCAGGGAGCACTAAGCGAGGT   \n",
       "27232         10         AGGGGAGCCTCACCACGAGCTGCCCCCCAGGGAGCACTAA   \n",
       "...          ...                                              ...   \n",
       "33963         17            GGATGTGGAATCAACCCACAGCTGCACAGGGCAGGTC   \n",
       "33964         10         GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGG   \n",
       "33965         13      GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGGGCA   \n",
       "33966         15    GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGGGCAGG   \n",
       "33967         17  GCCGGGCGGGGGATGTGGAATCAACCCACAGCTGCACAGGGCAGGTC   \n",
       "\n",
       "      Distance_to_nick RHA_size PAM_disrupted Proto_disrupted  \n",
       "27228                2       17         False            True  \n",
       "27229                2       17         False            True  \n",
       "27230                2       17         False            True  \n",
       "27231                2       17         False            True  \n",
       "27232                2       27         False            True  \n",
       "...                ...      ...           ...             ...  \n",
       "33963               17        2         False           False  \n",
       "33964               17       12         False           False  \n",
       "33965               17       12         False           False  \n",
       "33966               17       12         False           False  \n",
       "33967               17       12         False           False  \n",
       "\n",
       "[1000 rows x 34 columns]"
      ]
     },
     "execution_count": 818,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Variant_Type']=='INS'][-1000:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 786,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 786,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 782,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TGCGCCGGTCTCTCCCAGGACAGGCACAAACAT')"
      ]
     },
     "execution_count": 782,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Bio.Seq.Seq(df.iloc[1]['RTT_PBS']).reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/5g/xxjcy2g531n02hnyb6d8qdsr0000gn/T/ipykernel_79410/1396537375.py\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'df' is not defined"
     ]
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 780,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 780,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('GCCGGTCTCTCCCAGGACAGGCACAAACAT')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Protospacer ranking\n",
    "\n",
    "https://github.com/gpp-rnd/rs3\n",
    "\n",
    "https://github.com/microsoftResearch//azimuth\n",
    "\n",
    "\n",
    "This seems like the best option: https://github.com/crisprVerse/crisprScore\n",
    "\n",
    "Use in python: https://willfondrie.com/2022/01/how-to-use-r-packages-in-python/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This code is adopted from CRISPOR program with minor modifications\n",
    "# https://github.com/maximilianh/crisporWebsite\n",
    "# \"Evaluation of off-target and on-target scoring algorithms and integration into the guide RNA selection tool CRISPOR\"\n",
    "# by M. Haeussler et al. 2016\n",
    "\n",
    "import math\n",
    " \n",
    "params = [\n",
    "# pasted/typed table from PDF and converted to zero-based positions\n",
    "(1,'G',-0.2753771),(2,'A',-0.3238875),(2,'C',0.17212887),(3,'C',-0.1006662),\n",
    "(4,'C',-0.2018029),(4,'G',0.24595663),(5,'A',0.03644004),(5,'C',0.09837684),\n",
    "(6,'C',-0.7411813),(6,'G',-0.3932644),(11,'A',-0.466099),(14,'A',0.08537695),\n",
    "(14,'C',-0.013814),(15,'A',0.27262051),(15,'C',-0.1190226),(15,'T',-0.2859442),\n",
    "(16,'A',0.09745459),(16,'G',-0.1755462),(17,'C',-0.3457955),(17,'G',-0.6780964),\n",
    "(18,'A',0.22508903),(18,'C',-0.5077941),(19,'G',-0.4173736),(19,'T',-0.054307),\n",
    "(20,'G',0.37989937),(20,'T',-0.0907126),(21,'C',0.05782332),(21,'T',-0.5305673),\n",
    "(22,'T',-0.8770074),(23,'C',-0.8762358),(23,'G',0.27891626),(23,'T',-0.4031022),\n",
    "(24,'A',-0.0773007),(24,'C',0.28793562),(24,'T',-0.2216372),(27,'G',-0.6890167),\n",
    "(27,'T',0.11787758),(28,'C',-0.1604453),(29,'G',0.38634258),(1,'GT',-0.6257787),\n",
    "(4,'GC',0.30004332),(5,'AA',-0.8348362),(5,'TA',0.76062777),(6,'GG',-0.4908167),\n",
    "(11,'GG',-1.5169074),(11,'TA',0.7092612),(11,'TC',0.49629861),(11,'TT',-0.5868739),\n",
    "(12,'GG',-0.3345637),(13,'GA',0.76384993),(13,'GC',-0.5370252),(16,'TG',-0.7981461),\n",
    "(18,'GG',-0.6668087),(18,'TC',0.35318325),(19,'CC',0.74807209),(19,'TG',-0.3672668),\n",
    "(20,'AC',0.56820913),(20,'CG',0.32907207),(20,'GA',-0.8364568),(20,'GG',-0.7822076),\n",
    "(21,'TC',-1.029693),(22,'CG',0.85619782),(22,'CT',-0.4632077),(23,'AA',-0.5794924),\n",
    "(23,'AG',0.64907554),(24,'AG',-0.0773007),(24,'CG',0.28793562),(24,'TG',-0.2216372),\n",
    "(26,'GT',0.11787758),(28,'GG',-0.69774)]\n",
    " \n",
    "intercept =  0.59763615\n",
    "gc_high    = -0.1665878\n",
    "gc_low     = -0.2026259\n",
    " \n",
    "def calc_doench_score(seq):\n",
    "  score = intercept\n",
    " \n",
    "  #guide = seq[4:24]\n",
    "  guide = seq\n",
    "  gc_count = guide.count('G') + guide.count('C')\n",
    "  gc_weight = gc_low if gc_count <= 10 else gc_high\n",
    "  score += abs(10-gc_count)*gc_weight\n",
    " \n",
    "  for pos, model_seq, weight in params:\n",
    "    sub_seq = seq[pos:pos+len(model_seq)]\n",
    "    if sub_seq == model_seq:\n",
    "      score += weight\n",
    "  total_score = 1.0/(1.0+math.exp(-score)) * 100\n",
    "  return '%.3f' % total_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'33.621'"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "calc_doench_score('AGATTCTTGCTCACCGTCTG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "ename": "OSError",
     "evalue": "dlopen(/Users/samgould/opt/anaconda3/lib/python3.9/site-packages/lightgbm/lib/lib_lightgbm.so, 0x0006): Library not loaded: /usr/local/opt/libomp/lib/libomp.dylib\n  Referenced from: /Users/samgould/opt/anaconda3/lib/python3.9/site-packages/lightgbm/lib/lib_lightgbm.so\n  Reason: tried: '/usr/local/opt/libomp/lib/libomp.dylib' (no such file), '/usr/local/lib/libomp.dylib' (no such file), '/usr/lib/libomp.dylib' (no such file)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/5g/xxjcy2g531n02hnyb6d8qdsr0000gn/T/ipykernel_79410/4176615838.py\u001b[0m in \u001b[0;36m<cell line: 4>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mrs3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseq\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpredict_seq\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mcontext_seqs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'GACGAAAGCGACAACGCGTTCATCCGGGCA'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'AGAAAACACTAGCATCCCCACCCGCGGACT'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mpredict_seq\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcontext_seqs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msequence_tracr\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'Chen2013'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/rs3/seq.py\u001b[0m in \u001b[0;36mpredict_seq\u001b[0;34m(context_sequences, sequence_tracr, ref_tracrs, n_jobs)\u001b[0m\n\u001b[1;32m     47\u001b[0m     \u001b[0;34m:\u001b[0m\u001b[0;32mreturn\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mlist\u001b[0m \u001b[0mof\u001b[0m \u001b[0mfloat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpredictions\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m     \"\"\"\n\u001b[0;32m---> 49\u001b[0;31m     \u001b[0mmodel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mload_seq_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     50\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Calculating sequence-based features'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     51\u001b[0m     featurized_sgrnas = featurize_context(context_sequences, sequence_tracr=sequence_tracr, ref_tracrs=ref_tracrs,\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/rs3/seq.py\u001b[0m in \u001b[0;36mload_seq_model\u001b[0;34m()\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mload_seq_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0;34m\"\"\"Load rule set 3 sequence model\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m     \u001b[0mmodel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mjoblib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdirname\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m__file__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'RuleSet3.pkl'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/joblib/numpy_pickle.py\u001b[0m in \u001b[0;36mload\u001b[0;34m(filename, mmap_mode)\u001b[0m\n\u001b[1;32m    656\u001b[0m                     \u001b[0;32mreturn\u001b[0m \u001b[0mload_compatibility\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    657\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 658\u001b[0;31m                 \u001b[0mobj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_unpickle\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfilename\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmmap_mode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    659\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/joblib/numpy_pickle.py\u001b[0m in \u001b[0;36m_unpickle\u001b[0;34m(fobj, filename, mmap_mode)\u001b[0m\n\u001b[1;32m    575\u001b[0m     \u001b[0mobj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    576\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 577\u001b[0;31m         \u001b[0mobj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0munpickler\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    578\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0munpickler\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompat_mode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    579\u001b[0m             warnings.warn(\"The file '%s' has been generated with a \"\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/pickle.py\u001b[0m in \u001b[0;36mload\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1210\u001b[0m                     \u001b[0;32mraise\u001b[0m \u001b[0mEOFError\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1211\u001b[0m                 \u001b[0;32massert\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbytes_types\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1212\u001b[0;31m                 \u001b[0mdispatch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1213\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0m_Stop\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mstopinst\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1214\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mstopinst\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/pickle.py\u001b[0m in \u001b[0;36mload_stack_global\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1535\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mstr\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodule\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1536\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mUnpicklingError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"STACK_GLOBAL requires str\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1537\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind_class\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodule\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1538\u001b[0m     \u001b[0mdispatch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mSTACK_GLOBAL\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mload_stack_global\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1539\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/pickle.py\u001b[0m in \u001b[0;36mfind_class\u001b[0;34m(self, module, name)\u001b[0m\n\u001b[1;32m   1577\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mmodule\u001b[0m \u001b[0;32min\u001b[0m \u001b[0m_compat_pickle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mIMPORT_MAPPING\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1578\u001b[0m                 \u001b[0mmodule\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_compat_pickle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mIMPORT_MAPPING\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmodule\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1579\u001b[0;31m         \u001b[0m__import__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodule\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1580\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mproto\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0;36m4\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1581\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0m_getattribute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodules\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmodule\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/lightgbm/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mpathlib\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mPath\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mbasic\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mBooster\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mDataset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSequence\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mregister_logger\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mcallback\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mearly_stopping\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlog_evaluation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrecord_evaluation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreset_parameter\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mCVBooster\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcv\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtrain\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/lightgbm/basic.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    224\u001b[0m     \u001b[0m_LIB\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mMock\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mctypes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCDLL\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    225\u001b[0m \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 226\u001b[0;31m     \u001b[0m_LIB\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_load_lib\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    227\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    228\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/lightgbm/basic.py\u001b[0m in \u001b[0;36m_load_lib\u001b[0;34m()\u001b[0m\n\u001b[1;32m    209\u001b[0m     \u001b[0;34m\"\"\"Load LightGBM library.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    210\u001b[0m     \u001b[0mlib_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfind_lib_path\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 211\u001b[0;31m     \u001b[0mlib\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mctypes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcdll\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mLoadLibrary\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlib_path\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    212\u001b[0m     \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mLGBM_GetLastError\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrestype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mctypes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mc_char_p\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    213\u001b[0m     \u001b[0mcallback\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mctypes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCFUNCTYPE\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mctypes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mc_char_p\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/ctypes/__init__.py\u001b[0m in \u001b[0;36mLoadLibrary\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m    458\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    459\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mLoadLibrary\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 460\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_dlltype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    461\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    462\u001b[0m     \u001b[0m__class_getitem__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclassmethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_types\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGenericAlias\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/ctypes/__init__.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, name, mode, handle, use_errno, use_last_error, winmode)\u001b[0m\n\u001b[1;32m    380\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    381\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mhandle\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 382\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_handle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_dlopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    383\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    384\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_handle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhandle\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: dlopen(/Users/samgould/opt/anaconda3/lib/python3.9/site-packages/lightgbm/lib/lib_lightgbm.so, 0x0006): Library not loaded: /usr/local/opt/libomp/lib/libomp.dylib\n  Referenced from: /Users/samgould/opt/anaconda3/lib/python3.9/site-packages/lightgbm/lib/lib_lightgbm.so\n  Reason: tried: '/usr/local/opt/libomp/lib/libomp.dylib' (no such file), '/usr/local/lib/libomp.dylib' (no such file), '/usr/lib/libomp.dylib' (no such file)"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from rs3.seq import predict_seq\n",
    "context_seqs = ['GACGAAAGCGACAACGCGTTCATCCGGGCA', 'AGAAAACACTAGCATCCCCACCCGCGGACT']\n",
    "predict_seq(context_seqs, sequence_tracr='Chen2013')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 821,
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mCannot execute code, session has been disposed. Please try restarting the Kernel."
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "from genet.predict import SpCas9\n",
    "\n",
    "# Put the target context (30bp) that you want to find Cas9 activity in the list.\n",
    "# Input seq: 4bp 5' context + 20 guide + 3bp PAM + 3bp 3' context\n",
    "\n",
    "spcas = SpCas9()\n",
    "\n",
    "list_target = [\n",
    "                'TCACCTTCGTTTTTTTCCTTCTGCAGGAGG',\n",
    "                'CCTTCGTTTTTTTCCTTCTGCAGGAGGACA',\n",
    "                'CTTTCAAGAACTCTTCCACCTCCATGGTGT',\n",
    "                ]\n",
    "                \n",
    "df_out = spcas.predict(list_target)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Error checking\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Align import PairwiseAligner\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sensor_outcome_corrected(p53_lib, i):\n",
    "    #p53_lib = df containing pegRNA + target sequence + other info\n",
    "    #i = index\n",
    "\n",
    "    sensor_true = p53_lib.iloc[i]['target sequence']\n",
    "\n",
    "    #let's quantify the homology overhang as well...\n",
    "    PAM_strand = p53_lib.iloc[i]['PAM strand']\n",
    "    var_type = p53_lib.iloc[i]['variant_type']\n",
    "\n",
    "\n",
    "    if PAM_strand=='+':\n",
    "        sensor = Bio.Seq.Seq(p53_lib.iloc[i]['target sequence']).reverse_complement()\n",
    "\n",
    "    elif PAM_strand=='-':\n",
    "        sensor = p53_lib.iloc[i]['target sequence']\n",
    "\n",
    "    pbs_length = p53_lib.iloc[i]['PBS length']\n",
    "    pbs = p53_lib.iloc[i]['PBS_RTT_5to3'][-pbs_length:]\n",
    "    rtt = p53_lib.iloc[i]['PBS_RTT_5to3'][:-(pbs_length-1)] #FIXING [:-0] indexing issue...\n",
    "\n",
    "    #pbs_rtt = p53_lib.iloc[i]['PBS_RTT_5to3']\n",
    "\n",
    "    pbs_find = sensor.find(pbs)\n",
    "\n",
    "    #first compute the portion before the mutation for computing the right flank\n",
    "    for kk, val in enumerate(reversed(rtt)):\n",
    "        idx = pbs_find-kk\n",
    "        if val!= sensor[idx]: #find first mismatch\n",
    "            mut_start=kk\n",
    "            break\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "\n",
    "    try: mut_start\n",
    "    except: #case where's theres no identifiable change (because of homology overhang being too short/matching coincidentally...)\n",
    "        return sensor_true, sensor_true, 0, \"no homology overhang\"\n",
    "\n",
    "    #alignments = aligner.align(pbs, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(rtt, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #now we have the right flank...\n",
    "    right_flank = rtt[-mut_start:-1] + sensor[pbs_find:]\n",
    "\n",
    "    #alignments = aligner.align(right_flank, sensor)\n",
    "    #print(alignments[0])\n",
    "    \n",
    "    if var_type in ['SNP', 'ONP']:\n",
    "\n",
    "        matching_arr = np.zeros(len(rtt[:-mut_start]))\n",
    "\n",
    "        for kk, val in enumerate(reversed(rtt[:-mut_start])):\n",
    "            idx = pbs_find-mut_start-kk\n",
    "\n",
    "            if val== sensor[idx]: #find regions of homology\n",
    "                matching_arr[kk]+=1\n",
    "                \n",
    "            else:\n",
    "                continue\n",
    "        \n",
    "        #based on values in match array, determine where mutation vs. homology overhang is...\n",
    "        end_mut = np.where(matching_arr==0)[0][-1]\n",
    "\n",
    "        mut_allele = rtt[-(mut_start+end_mut+1):-mut_start]\n",
    "        homology_overhang = rtt[:-(mut_start+end_mut+1)]\n",
    "        #this formulation will NOT catch situations where the homology overhang doesn't match???????\n",
    "        #(though this shouldn't be an issue for SNPs/ONPs)\n",
    "\n",
    "        left_flank = sensor[:(pbs_find-(len(rtt)-1))]\n",
    "\n",
    "    elif var_type=='DEL':\n",
    "\n",
    "        #find the size of the deletion by shifting and checking for matches...\n",
    "        #will allow for detection of insufficient homology overhang...\n",
    "        length_remaining = len(rtt[:-mut_start])\n",
    "        homology_overhang = rtt[:-mut_start]\n",
    "        mut_allele = ''\n",
    "\n",
    "        start_idx = pbs_find-mut_start\n",
    "        for i in range(len(sensor[:start_idx])): #check rest of the sensor...\n",
    "        \n",
    "            #sliding window to find size of deletion/perfect match...\n",
    "            sensor_a = sensor[start_idx-i-length_remaining:start_idx-i]\n",
    "\n",
    "            if sensor_a==homology_overhang: #once match is found, record and break...\n",
    "                size_del = i\n",
    "                break\n",
    "        \n",
    "            else:\n",
    "                continue\n",
    "\n",
    "        #error where the full homology overhang region not included in sensor...\n",
    "        try: size_del\n",
    "        except: \n",
    "            return sensor_true, sensor_true, len(homology_overhang), 'sensor too short' #no excepted editing, return WT sensor\n",
    "\n",
    "        left_flank = sensor[:start_idx-size_del-length_remaining]\n",
    "\n",
    "    elif var_type=='INS':\n",
    "\n",
    "        length_remaining = len(rtt[:-mut_start])\n",
    "        start_idx = pbs_find-mut_start+1\n",
    "\n",
    "        for i in range(length_remaining):\n",
    "            rtt_region = rtt[:-(mut_start+i)]\n",
    "            sensor_a = sensor[start_idx-len(rtt_region):start_idx]\n",
    "\n",
    "            #print('rt: ' + rtt_region)\n",
    "            #print('s: ' + sensor_a)\n",
    "\n",
    "            if sensor_a==rtt_region:\n",
    "                mut_allele = rtt[-(mut_start+i):-mut_start]\n",
    "                break\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "        #in case of no homology overhang...\n",
    "        try: mut_allele\n",
    "        except: \n",
    "            return sensor_true, sensor_true, 0, 'no homology overhang' #no excepted editing, return WT sensor\n",
    "\n",
    "        homology_overhang = sensor_a\n",
    "        left_flank = sensor[:start_idx-len(rtt_region)]\n",
    "\n",
    "    edited_sensor = left_flank + homology_overhang + mut_allele + right_flank      \n",
    "\n",
    "    if PAM_strand=='+':\n",
    "        edited_sensor_true = str(Bio.Seq.Seq(edited_sensor).reverse_complement())\n",
    "\n",
    "    elif PAM_strand=='-':\n",
    "        edited_sensor_true=edited_sensor\n",
    "\n",
    "\n",
    "\n",
    "    #alignments = aligner.align(rtt[-mut_start:], sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(right_flank, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(homology_overhang, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(edited_sensor, sensor)\n",
    "    #print(alignments[0])\n",
    "\n",
    "    #alignments = aligner.align(edited_sensor_true, sensor_true)\n",
    "    #print(alignments[0])\n",
    "\n",
    "\n",
    "    return sensor_true, edited_sensor_true, len(homology_overhang), 'no error'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "p53_lib = ....\n",
    "\n",
    "#running it for every mutation in the p53 library\n",
    "edit_sens = []\n",
    "homo = []\n",
    "idx_holder = []\n",
    "classif = []\n",
    "#del_idx = p53_del.index\n",
    "\n",
    "#ins_idx = p53_lib[p53_lib['variant_type']=='SNP'].index\n",
    "\n",
    "error = []\n",
    "for i in range(len(p53_lib)):\n",
    "    try:\n",
    "        s, s_e, homo2, m = sensor_outcome_corrected(p53_lib, i)\n",
    "        idx_holder.append(i)\n",
    "        edit_sens.append(s_e)\n",
    "        homo.append(homo2)\n",
    "        classif.append(m)\n",
    "    \n",
    "    except:\n",
    "        error.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
