<!DOCTYPE html>
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="theme-color" content="#2D2D2D" />
  
  <title>PEGG :: pegg.prime package</title>
  
  <link rel="icon" type="image/png" href="_static/PEGG_notext.ico">
  
  <link rel="index" title="Index" href="genindex.html"/>

  <link rel="stylesheet" href="_static/css/nltk_theme.css"/>
  <link rel="stylesheet" href="_static/css/custom.css"/>

  <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
      <script type="text/javascript" src="_static/documentation_options.js"></script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/_sphinx_javascript_frameworks_compat.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
  

  <script src="https://email.tl.fortawesome.com/c/eJxNjUEOgyAQAF8jR7Kw6wIHDh7sP1Cw2mgxgmn6-3JsMqc5zEQfE8dkxOY1KKMUOI3ACFKRJpSW2AAp7ontYIaxI6i7XPJVwyeVfCQ550Os3jLrGSNOLgbdAy6s0PBk2TFNjEbsfq31LB0OnX407pJa5v2faRadwSW63mn5KuLyR9j2tgx3zecanl-55R_-jjPs"></script> 
</head>

<body>
  <div id="nltk-theme-container">
    <header>
      <div id="logo-container">
          
          <a href="index.html"><img src="_static/PEGG_notext.png"></a>
          
      </div>
      <div id="project-container">
        
        <h1>PEGG Documentation</h1>
        
      </div>

      <a id="menu-toggle" class="fa fa-bars" aria-hidden="true"></a>

      <script type="text/javascript">
        $("#menu-toggle").click(function() {
          $("#menu-toggle").toggleClass("toggled");
          $("#side-menu-container").slideToggle(300);
        });
      </script>
    </header>

    <div id="content-container">

      <div id="side-menu-container">

        <div id="search" role="search">
        <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
            <input type="text" name="q" placeholder="Search" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
        </form>
</div>

        <div id="side-menu" role="navigation">
          
  
    
  
  
    <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">ğŸš€ Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#required-input-files">Required Input Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#reference-file-formatting">Reference File Formatting</a><ul>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#input-mutations">(1) Input mutations</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#a-compatible-input-files">(1a) Compatible Input Files</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#b-clinvar-variants">(1b) ClinVar variants</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#reference-genome">(2) Reference Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#on-off-target-scores">(3) On/Off Target scores</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#using-pegg">Using PEGG</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#visualization-tools">Visualization Tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#oligo-generation">Oligo Generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#library-generation">Library Generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#jupyter-notebook-tutorial">Jupyter Notebook Tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#a-note-on-ram">A Note on RAM</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="jupyter_tutorial.html">ğŸª Jupyter Notebook Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="PEGG.html">ğŸ“š Complete Function Documentation</a></li>
</ul>

  

        </div>

        
      </div>

      <div id="main-content-container">
        <div id="main-content" role="main">
          
  <div class="section" id="pegg-prime-package">
<h1>pegg.prime package<a class="headerlink" href="#pegg-prime-package" title="Permalink to this heading">Â¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading">Â¶</a></h2>
</div>
<div class="section" id="module-pegg.prime.crisporEffScores">
<span id="pegg-prime-crisporeffscores-module"></span><h2>pegg.prime.crisporEffScores module<a class="headerlink" href="#module-pegg.prime.crisporEffScores" title="Permalink to this heading">Â¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.ScoreCache">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">ScoreCache</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scoreName</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.ScoreCache" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>a cache of eff scores, kept on disk. Can avoid slow calculations by keeping
the value of the score in a tab-sep file.</p>
<dl class="py method">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.ScoreCache.findNewSeqs">
<span class="sig-name descname"><span class="pre">findNewSeqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.ScoreCache.findNewSeqs" title="Permalink to this definition">Â¶</a></dt>
<dd><p>get seqs that are not in cache. If all are, return the list of scores.
Otherwise return None for the scores.
Returns tuple (seqs, scores)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.ScoreCache.mergeIntoCache">
<span class="sig-name descname"><span class="pre">mergeIntoCache</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">newScores</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.ScoreCache.mergeIntoCache" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.cacheScores">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">cacheScores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scoreName</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scoreFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.cacheScores" title="Permalink to this definition">Â¶</a></dt>
<dd><p>run scoreFunc on seqs, using an on-disk score cache to improve speed</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcAllBaeScores">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcAllBaeScores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcAllBaeScores" title="Permalink to this definition">Â¶</a></dt>
<dd><p>run seqs through calcMicroHomolScore()
PAM-site has to start at the nucleotide exactly in the middle of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">calcAllBaeScores</span><span class="p">([</span><span class="s2">&quot;AGCAGGATAGTCCTTCCGAGTGGAGGGAGGAGCAGGATAGTCCTTCCGAGTGGAGGGAGGAGCAGGATAGTCCTTCCGAGTGGAGGGAGG&quot;</span><span class="p">])[:</span><span class="mi">2</span><span class="p">]</span>
<span class="go">([7829], [46])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">calcAllBaeScores</span><span class="p">([</span><span class="s2">&quot;AGCAGGATAGTCCTTCCGAGTGGANNNAGGAGCAGGATAGTCCTTCCGAGTGGAGGGAGGAGCAGGATAGTCCTTCCGAGTGGAGGGAGG&quot;</span><span class="p">])[:</span><span class="mi">2</span><span class="p">]</span>
<span class="go">([6646], [45])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcAllScores">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcAllScores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">addOpt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skipScores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enzyme</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scoreNames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcAllScores" title="Permalink to this definition">Â¶</a></dt>
<dd><p>given 100bp sequences (50bp 5â€™ of PAM, 50bp 3â€™ of PAM) calculate all efficiency scores
and return as a dict scoreName -&gt; list of scores (same order).
&gt;&gt;&gt; sorted(calcAllScores([â€œCCACGTCTCCACACATCAGCACAACTACGCAGCGCCTCCCTCCACTCGGAAGGACTATCCTGCTGCCAAGAGGGTCAAGTTGGACAGTGTCAGAGTCCTGâ€]).items())
[(â€˜aziInVitroâ€™, [39]), (â€˜ccTopâ€™, [64.53235600000001]), (â€˜chariRankâ€™, [54]), (â€˜chariRawâ€™, [-0.15504833]), (â€˜crisprScanâ€™, [39]), (â€˜doenchâ€™, [10]), (â€˜fusiâ€™, [55]), (â€˜fusiOldâ€™, [56]), (â€˜housdenâ€™, [6.3]), (â€˜sscâ€™, [-0.035894]), (â€˜wangâ€™, [66]), (â€˜wuCrisprâ€™, [0])]
&gt;&gt;&gt; sorted(calcAllScores([â€œCCACGTCTCCACACATCAGCACAACTACGCAGCGCCTCCCTCCACTCGGAAGGACTANCCTGCTGCCAAGAGGGTCAAGTTGGACAGTGTCAGAGTCCTGâ€]).items())
[(â€˜aziInVitroâ€™, [39]), (â€˜ccTopâ€™, [64.53235600000001]), (â€˜chariRankâ€™, [54]), (â€˜chariRawâ€™, [-0.15504833]), (â€˜crisprScanâ€™, [40]), (â€˜doenchâ€™, [10]), (â€˜fusiâ€™, [55]), (â€˜fusiOldâ€™, [56]), (â€˜housdenâ€™, [6.3]), (â€˜sscâ€™, [-0.035894]), (â€˜wangâ€™, [66]), (â€˜wuCrisprâ€™, [0])]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcAziInVitro">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcAziInVitro</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcAziInVitro" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Another score: Azimuth trained on the Moreno-Mateos data, see README, received from J. Listgarden</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcAziScore">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcAziScore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcAziScore" title="Permalink to this definition">Â¶</a></dt>
<dd><p>the official implementation of the Doench2016 (aka Fusi) score from Microsoft</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcCctopScore">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcCctopScore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcCctopScore" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcChariScores">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcChariScores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">baseDir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcChariScores" title="Permalink to this definition">Â¶</a></dt>
<dd><p>return dict with chari 2015 scores, returns two lists (rawScores, rankPercent)
input seqs have lengths 21bp: 20 bp guide + 1bp first from PAM
&gt;&gt;&gt; calcChariScores([â€œCTTCTTCAAGGTAACTGCAGAâ€, â€œCTTCTTCAAGGTAACTGGGGGâ€])
([0.54947621, 0.58604487], [80, 81])
&gt;&gt;&gt; calcChariScores([â€œCTTCTTCAAGGNAACTGCAGAâ€])
([0.9025848], [88])</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcCrisprScanScores">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcCrisprScanScores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcCrisprScanScores" title="Permalink to this definition">Â¶</a></dt>
<dd><p>input is a 35bp long sequence: 6bp 5â€™, 20bp guide, 3 bp PAM and 6bp 3â€™
&gt;&gt;&gt; calcCrisprScanScores([â€œTCCTCTGGTGGCGCTGCTGGATGGACGGGACTGTAâ€])
[77]
&gt;&gt;&gt; calcCrisprScanScores([â€œTCCTCTNGTGGCGCTGCTGGATGGACGGGACTGTAâ€])
[77]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcDeepCpf1Scores">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcDeepCpf1Scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcDeepCpf1Scores" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcDoenchScores">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcDoenchScores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcDoenchScores" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Code reproduced following paperâ€™s methods section. Thanks to Daniel McPherson for fixing it.
Input is a 30mer: 4bp 5â€™, 20bp guide, 3bp PAM, 3bp 5â€™</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcFreeEnergy">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcFreeEnergy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcFreeEnergy" title="Permalink to this definition">Â¶</a></dt>
<dd><p>runs a list of 20bp guide sequences through mfold and returns their gibbs free energy
&gt;&gt;&gt; calcFreeEnergy([â€œGGGTGGGGGGAGTTTGCTCCTGGâ€])
0</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcFusiDoench">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcFusiDoench</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcFusiDoench" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcHousden">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcHousden</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcHousden" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Calc housden score and round to one decimal point.
Based on java file Crispr.java received from the authors.
&gt;&gt;&gt; calcHousden([â€œATCTGACCTCCCGGCTAATTâ€])
[6.9]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcLindelScore">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcLindelScore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqIds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcLindelScore" title="Permalink to this definition">Â¶</a></dt>
<dd><p>run model by Wei Chen. seqs is 100bp long sequences around beginning of PAM, like all other code.
returns dict with seqId -&gt; (probability of frameshift, mutSeqs)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcMicroHomolScore">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcMicroHomolScore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcMicroHomolScore" title="Permalink to this definition">Â¶</a></dt>
<dd><p>calculate the micro homology and out-of-frame score for a breakpoint in a 60-80mer
See <a class="reference external" href="http://www.nature.com/nmeth/journal/v11/n7/full/nmeth.3015.html">http://www.nature.com/nmeth/journal/v11/n7/full/nmeth.3015.html</a>
Source code adapted from Supp File 1
returns micro-homology score, out-of-frame score and a list of tuples:
(sequence, score)</p>
<p>From the manuscript:
â€œOn the basis of these observations, we developed a simple formula and a
computer program (Supplementary Fig. 3) to predict the deletion patterns
at a given nuclease target site that are associated with microhomology of
at least two bases (Fig. 1b and Supplementary Note). We assigned a pattern
score to each deletion pattern and a microhomology score (equaling the sum
of pattern scores) to each target site. We then obtained an out-of-frame
score at a given site by dividing the sum of pattern scores assigned to
frameshifting deletions by the microhomology score.â€</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcMutSeqs">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcMutSeqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqIds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enzyme</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scoreNames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcMutSeqs" title="Permalink to this definition">Â¶</a></dt>
<dd><p>run predictors that return the result of DSB repair. seqs is a list of 100bp long sequences.</p>
<p>returns dict with the score name as the key.
Each value is a dict again. The keys are the name of the resulting score e.g. â€œoofâ€ (out of frame score) and
at least the key â€œseqsâ€, a dict with seqId -&gt; (score, sequence).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcNajmScore">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcNajmScore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcNajmScore" title="Permalink to this definition">Â¶</a></dt>
<dd><p>The score of the Najm 2018 paper, for SaCas9, using Azimuth</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcRs3Scores">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcRs3Scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcRs3Scores" title="Permalink to this definition">Â¶</a></dt>
<dd><p>calc Doench RS3 scores, from <a class="reference external" href="https://github.com/gpp-rnd/rs3">https://github.com/gpp-rnd/rs3</a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcSscScores">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcSscScores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcSscScores" title="Permalink to this definition">Â¶</a></dt>
<dd><p>calc the SSC scores from the paper Xu Xiao Chen Li Meyer Brown Lui Gen Res 2015 
Input is a 30mer, 20bp for the guide, 3bp PAM, 7bp 3â€™ flanking
&gt;&gt;&gt; calcSscScores([â€œAGCAGGATAGTCCTTCCGAGTGGAGGGAGGâ€])
[0.182006]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcWangSvmScores">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcWangSvmScores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcWangSvmScores" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Use the wang.model file to score sequences. Input is only the 20bp guide sequence.
Uses libsvmâ€™s svm-predict program, V2.6.
The score is inversed, so higher scores are better, like all other scores</p>
<p>The results here are off mostly by 1-5% from the results returned by the Wang et al source code.
I never found out why, there are no parameters for â€œsvm_predictâ€. Should not be due to a version
difference either, Iâ€™m using the same libsvm version as the e1071 R module.
This is necessary for a web server implementation as e1071 in R cannot read the model from a file.</p>
<p>The original implementation from the paper can be called with calcWangSvmScoresUsingR()</p>
<p>See compareWangScores.py:</p>
<p>The Pearson correlation between both ways to calculate the score is 97%.</p>
<p>Histogram of the score differences:
0.000000 <strong>********************************************************</strong> 3074
0.050000 <strong>*****************************</strong> 1674
0.100000 <strong>********</strong> 612
0.150000 <a href="#id1"><span class="problematic" id="id2">**</span></a>** 191
0.200000 * 52
0.250000  7
0.300000  1
cat out/wangDiffs.tsv | cut -f4 | tr -d â€˜-â€˜ | grep -v diff | textHistogram stdin stdout -real -binSize=0.05</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">calcWangSvmScores</span><span class="p">([</span><span class="s2">&quot;ATAGACCTACCTTGTTGAAG&quot;</span><span class="p">])</span>
<span class="go">[60]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">calcWangSvmScores</span><span class="p">([</span><span class="s2">&quot;NTAGACCTACCTTGTTGAAG&quot;</span><span class="p">])</span>
<span class="go">[60]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcWangSvmScoresUsingR">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcWangSvmScoresUsingR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcWangSvmScoresUsingR" title="Permalink to this definition">Â¶</a></dt>
<dd><p>returns the SVM-calculated efficiency scores from the Wang/Sabatini/Lander paper
This is using their original source code
Weâ€™re inversing the score so higher scores are better</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcWeiChenScores">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcWeiChenScores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcWeiChenScores" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Calc weiChen score</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.calcWuCrisprScore">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">calcWuCrisprScore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.calcWuCrisprScore" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Input is a list of 30mers:
20bp guide, 3bp PAM, 7bp 3â€™ sequence.
&gt;&gt;&gt; calcWuCrisprScore([â€œggtgcagctcgagcaacaggcggcâ€])
[93]
&gt;&gt;&gt; calcWuCrisprScore([â€œggtgcagctngagcaacaggcggcâ€])
[0]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.convertChariToRankPerc">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">convertChariToRankPerc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">score</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.convertChariToRankPerc" title="Permalink to this definition">Â¶</a></dt>
<dd><p>convert chari score to rank percent using only 2 digits precision. A lot faster than
the version in the paper.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.forceWrapper">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">forceWrapper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">func</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.forceWrapper" title="Permalink to this definition">Â¶</a></dt>
<dd><p>run func over seqs. If any exception occurs, return a list of -1s for all seqs.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.getBinPath">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">getBinPath</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isDir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.getBinPath" title="Permalink to this definition">Â¶</a></dt>
<dd><p>get the full pathname of a platform-specific binary, in the bin/ directory relative to this directory</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.getGrafType">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">getGrafType</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.getGrafType" title="Permalink to this definition">Â¶</a></dt>
<dd><p>check if a guide fulfills the criteria described in Graf et al, Cell Reports 2019
returns â€œttâ€ or â€œgccâ€ depending on the motif found or None if none of them were found.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.inList">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">inList</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.inList" title="Permalink to this definition">Â¶</a></dt>
<dd><p>return true if name is in list l</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.isCas9">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">isCas9</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">enzyme</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.isCas9" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.iterSvmRows">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">iterSvmRows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.iterSvmRows" title="Permalink to this definition">Â¶</a></dt>
<dd><p>translate sequences to wang/sabatini/lander paper representation
&gt;&gt;&gt; list(iterSvmRows([â€œATAGACCTACCTTGTTGAAGâ€]))
[[â€˜SEQâ€™, â€˜BP1Aâ€™, â€˜BP1Câ€™, â€˜BP1Tâ€™, â€˜BP1Gâ€™, â€˜BP2Aâ€™, â€˜BP2Câ€™, â€˜BP2Tâ€™, â€˜BP2Gâ€™, â€˜BP3Aâ€™, â€˜BP3Câ€™, â€˜BP3Tâ€™, â€˜BP3Gâ€™, â€˜BP4Aâ€™, â€˜BP4Câ€™, â€˜BP4Tâ€™, â€˜BP4Gâ€™, â€˜BP5Aâ€™, â€˜BP5Câ€™, â€˜BP5Tâ€™, â€˜BP5Gâ€™, â€˜BP6Aâ€™, â€˜BP6Câ€™, â€˜BP6Tâ€™, â€˜BP6Gâ€™, â€˜BP7Aâ€™, â€˜BP7Câ€™, â€˜BP7Tâ€™, â€˜BP7Gâ€™, â€˜BP8Aâ€™, â€˜BP8Câ€™, â€˜BP8Tâ€™, â€˜BP8Gâ€™, â€˜BP9Aâ€™, â€˜BP9Câ€™, â€˜BP9Tâ€™, â€˜BP9Gâ€™, â€˜BP10Aâ€™, â€˜BP10Câ€™, â€˜BP10Tâ€™, â€˜BP10Gâ€™, â€˜BP11Aâ€™, â€˜BP11Câ€™, â€˜BP11Tâ€™, â€˜BP11Gâ€™, â€˜BP12Aâ€™, â€˜BP12Câ€™, â€˜BP12Tâ€™, â€˜BP12Gâ€™, â€˜BP13Aâ€™, â€˜BP13Câ€™, â€˜BP13Tâ€™, â€˜BP13Gâ€™, â€˜BP14Aâ€™, â€˜BP14Câ€™, â€˜BP14Tâ€™, â€˜BP14Gâ€™, â€˜BP15Aâ€™, â€˜BP15Câ€™, â€˜BP15Tâ€™, â€˜BP15Gâ€™, â€˜BP16Aâ€™, â€˜BP16Câ€™, â€˜BP16Tâ€™, â€˜BP16Gâ€™, â€˜BP17Aâ€™, â€˜BP17Câ€™, â€˜BP17Tâ€™, â€˜BP17Gâ€™, â€˜BP18Aâ€™, â€˜BP18Câ€™, â€˜BP18Tâ€™, â€˜BP18Gâ€™, â€˜BP19Aâ€™, â€˜BP19Câ€™, â€˜BP19Tâ€™, â€˜BP19Gâ€™, â€˜BP20Aâ€™, â€˜BP20Câ€™, â€˜BP20Tâ€™, â€˜BP20Gâ€™], [â€˜ATAGACCTACCTTGTTGAAGâ€™, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1]]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.listToSvml">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">listToSvml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vec</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">res</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.listToSvml" title="Permalink to this definition">Â¶</a></dt>
<dd><p>convert a list of values to a line in svml format line like â€œ+1 1:0.5 2:1.5 â€¦</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.parseArgs">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">parseArgs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.parseArgs" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.parseSvmOut">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">parseSvmOut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.parseSvmOut" title="Permalink to this definition">Â¶</a></dt>
<dd><p>parse R SVM output file, return as dict seq -&gt; score</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.printScoreTabSep">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">printScoreTabSep</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enzyme</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.printScoreTabSep" title="Permalink to this definition">Â¶</a></dt>
<dd><p>print tab-sep rows with all seqs</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.readDict">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">readDict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isFloat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.readDict" title="Permalink to this definition">Â¶</a></dt>
<dd><p>read dict from a tab sep file</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.readSeqs">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">readSeqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inFname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.readSeqs" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.runLindel">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">runLindel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqIds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.runLindel" title="Permalink to this definition">Â¶</a></dt>
<dd><p>based on Lindel_prediction.py sent by Wei Chen
runtime 1-2 seconds.
Return: a dict with seqId -&gt; list of (fs-score, list of (seq-illustration, score, indel-desc))
&gt;&gt;&gt; ret = runLindel([â€œtestâ€], [â€œCCCTGGCGGCCTAAGGACTCGGCGCGCCGGAAGTGGCCAGGGCGGGGGCGACCTCGGCTCACAGâ€])
&gt;&gt;&gt; ret[â€œtestâ€][0]
70.36
&gt;&gt;&gt; ret[â€œtestâ€][1][1]
(â€˜4.91554921â€™, â€˜CCCTGGCGGCCTAAGGACTCGGCGCGCCGG | â€”â€”CCAGGGCGGGGGCGACCTCGGCTCACAGâ€™, â€˜D6  0â€™)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.sendFusiRequest">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">sendFusiRequest</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.sendFusiRequest" title="Permalink to this definition">Â¶</a></dt>
<dd><p>obtain the fusi score as calculated by Fusi et alâ€™s webservice
Needs a file fusiKey.txt in current directory or a file ~/.fusiKey.txt in userâ€™s home directory.</p>
<p># test deactivated - server is not working
&gt;&gt; sendFusiRequest([ â€œGGGAGGCTGCTTTACCCGCTGTGGGGGCGCâ€, â€œGGGAGGCTGCTTTACCCGCTGTGGGGGCGCâ€])
[60, 60]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.seqToVec">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">seqToVec</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offsets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'A':</span> <span class="pre">0,</span> <span class="pre">'C':</span> <span class="pre">1,</span> <span class="pre">'G':</span> <span class="pre">2,</span> <span class="pre">'T':</span> <span class="pre">3}</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.seqToVec" title="Permalink to this definition">Â¶</a></dt>
<dd><p>convert a x bp sequence to a 4 * x 0/1 vector
&gt;&gt;&gt; seqToVec(â€œAAAAATTTTTGGGGGCCCCCâ€)
[1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.seqsToChariSvml">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">seqsToChariSvml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.seqsToChariSvml" title="Permalink to this definition">Â¶</a></dt>
<dd><p>partially copied from generateSVMFile.FASTA.py in the Chari et al source code
&gt;&gt;&gt; seqsToChariSvml([â€œCTTCTTCAAGGTAACTGCAGAâ€, â€œCTTCTTCAAGGTAACTGGGGGâ€])
â€˜0 13:1 22:1 32:1 43:1 52:1 62:1 73:1 84:1 94:1 101:1 111:1 122:1 134:1 144:1 153:1 162:1 171:1 183:1 194:1 201:1 214:1n0 13:1 22:1 32:1 43:1 52:1 62:1 73:1 84:1 94:1 101:1 111:1 122:1 134:1 144:1 153:1 162:1 171:1 181:1 191:1 201:1 211:1â€™</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.setBinDir">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">setBinDir</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.setBinDir" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.setCacheDir">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">setCacheDir</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.setCacheDir" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.test">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">test</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.test" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.trimSeqs">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">trimSeqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fiveFlank</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threeFlank</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.trimSeqs" title="Permalink to this definition">Â¶</a></dt>
<dd><p>given a list of 100bp sequences, return a list of sequences with the
given number of basepairs 5â€™ and 3â€™ added from the middle position (pos 50) of
the sequences.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.trimSeqsForGuides">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">trimSeqsForGuides</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fiveFlank</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threeFlank</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.trimSeqsForGuides" title="Permalink to this definition">Â¶</a></dt>
<dd><p>like trimSeqs, but yield  (guide, trimmedSeq)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.vecToSeqDicts">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">vecToSeqDicts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coefs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.vecToSeqDicts" title="Permalink to this definition">Â¶</a></dt>
<dd><p>convert a list of 80 floats to 20 dictionaries with A/C/T/G -&gt; float</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.writeDict">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">writeDict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.writeDict" title="Permalink to this definition">Â¶</a></dt>
<dd><p>write dict as a tab file</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.crisporEffScores.writeSvmRows">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.crisporEffScores.</span></span><span class="sig-name descname"><span class="pre">writeSvmRows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.crisporEffScores.writeSvmRows" title="Permalink to this definition">Â¶</a></dt>
<dd><p>write the seqs in wang/sabatini SVM format to a file
#&gt;&gt;&gt; writeSvmRows([â€œATAGACCTACCTTGTTGAAGâ€])</p>
</dd></dl>

</div>
<div class="section" id="module-pegg.prime.pegg2">
<span id="pegg-prime-pegg2-module"></span><h2>pegg.prime.pegg2 module<a class="headerlink" href="#module-pegg.prime.pegg2" title="Permalink to this heading">Â¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.PAM_finder">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">PAM_finder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.PAM_finder" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Finds indeces of PAM sequences
Formatting: e.g. â€œNGGâ€
N = [A|T|C|G]
R = [A|G]
Y = [C|T]
S = [G|C]
W = [A|T]
K = [G|T]
M = [A|C]
B = [C|G|T]
D = [A|G|T]
H = [A|C|T]
V = [A|C|G]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.RF_score">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">RF_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.RF_score" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.clinvar_VCF_translator">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">clinvar_VCF_translator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variation_ids</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.clinvar_VCF_translator" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Function that takes a clinvar.vcf.gz file containing information about variants, as well as a list of variation ID numbers,
and returns a pandas dataframe containing the variants in a format that can be used by PEGG to design pegRNAs.</p>
<dl>
<dt>filepath</dt><dd><p><em>type = str</em></p>
<p>Filepath to the clinvar.vcf.gz file.</p>
</dd>
<dt>variation_ids</dt><dd><p><em>type = list</em></p>
<p>List of variation IDs that the user wants to convert to .</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.df_formatter">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">df_formatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">120</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.df_formatter" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Takes in variants (in cBioPortal format!)
and outputs dataframe with WT and ALT oligos with designated context_size
context_size = the amount of nt on either side of the variant e.g. AAA(A/G)AAA = context_size of 3
cols_to_save = information to save about the mutationsâ€¦(can modify beyond defaultsâ€¦)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.eligible_PAM_finder">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">eligible_PAM_finder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_RTT_length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">19</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.eligible_PAM_finder" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Determines eligible PAM sequences for creating pegRNAsâ€¦</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.genome_loader">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">genome_loader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath_gz</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.genome_loader" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Takes in filepath of human genome (GrCH37 or GrCh38) and returns records and index_list for PEGG parsing.</p>
<dl>
<dt>filepath_gz</dt><dd><p><em>type = str</em></p>
<p>The filepath to the .gz file holding the reference genome file.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.input_formatter">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">input_formatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context_size</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.input_formatter" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.make_aligner">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">make_aligner</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.make_aligner" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.mut_formatter">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">mut_formatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.mut_formatter" title="Permalink to this definition">Â¶</a></dt>
<dd><p>This program will break for more complex mutations (e.g. INDELs)</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pegg.prime.pegg2.mutation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">mutation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wt_w_context</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt_w_context</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.mutation" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.ontarget_score">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">ontarget_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.ontarget_score" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.other_filtration">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">other_filtration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pegRNA_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RE_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polyT_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.other_filtration" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Determines whether pegRNAs contain polyT termination sites or Restriction Enzyme Sites. Returns pegRNA_df with properties added.
Note: RE_sites requires only A, T, C, or G bases (Ns will not work currently; need to do custom filtration)â€¦
These properties are checking for:</p>
<ol class="arabic simple">
<li><p>u6 terminator (polyT sequence)</p></li>
<li><p>RE site presence</p></li>
</ol>
<dl>
<dt>pegRNA_df</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>A dataframe containing the pegRNAs for the selected input mutations. Generated by run() or pegRNA_generator()</p>
</dd>
<dt>RE_sites</dt><dd><p><em>type = list or None</em></p>
<p>A list containing the RE recognition sites to filter (e.g. [â€˜CGTCTCâ€™, â€˜GAATTCâ€™] for Esp3I and EcoRI). Default = None (no filtration).</p>
</dd>
<dt>polyT_threshold</dt><dd><p><em>type = int</em></p>
<p>The length of the polyT sequence to classify as a terminator. Default = 4.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.pegRNA_generator">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">pegRNA_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RTT_lengths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PBS_lengths</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.pegRNA_generator" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.peggscore2">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">peggscore2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.peggscore2" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.primedesign_formatter">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">primedesign_formatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.primedesign_formatter" title="Permalink to this definition">Â¶</a></dt>
<dd><p>takes as input sequence in prime design format 
e.g. AATTCCG(G/C)AATTCGCT</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.run">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'NGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rankby</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PEGG2_Score'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pegRNAs_per_mut</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'All'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RTT_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[5,</span> <span class="pre">10,</span> <span class="pre">15,</span> <span class="pre">25,</span> <span class="pre">30]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PBS_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[8,</span> <span class="pre">10,</span> <span class="pre">13,</span> <span class="pre">15]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_RHA_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RE_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polyT_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">19</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">120</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before_proto_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reverse-complement'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.run" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.sensor_generator">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">sensor_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before_proto_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reverse-complement'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.sensor_generator" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Generates sensor sequence for quantification of pegRNA editing outcomes
Automatically puts sensor in reverse complement orientation with respect to protospacer
This is highly reccomended to reduce recombination during cloning/library preparation</p>
<p>options for sensor_orienation = â€˜reverse_compâ€™; â€˜forwardâ€™</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.sensor_viz">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">sensor_viz</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_w_sensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.sensor_viz" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.split">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">split</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">word</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.split" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Simple function for splitting string into component characters</p>
</dd></dl>

</div>
<div class="section" id="module-pegg.prime">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pegg.prime" title="Permalink to this heading">Â¶</a></h2>
</div>
</div>


        </div>
      </div>

    </div>

<footer>
    <div id="footer-info">
        <ul id="build-details">
            
                <li class="footer-element">
                    
                        <a href="_sources/pegg.prime.rst.txt" rel="nofollow"> source</a>
                    
                </li>
            

            

            
        </ul>

        
            <div id="copyright">
                &copy; 2023, Samuel Gould
            </div>
        

        <div id="credit">
            created with <a href="http://sphinx-doc.org/">Sphinx</a> and <a href="https://github.com/tomaarsen/nltk_theme">NLTK Theme</a>
        </div>
    </div>
</footer> 

</div>

</body>
</html>