<!DOCTYPE html>
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="theme-color" content="#2D2D2D" />
  
  <title>PEGG :: 📚 Complete Function Documentation</title>
  
  <link rel="icon" type="image/png" href="_static/PEGG_notext.ico">
  
  <link rel="index" title="Index" href="genindex.html"/>

  <link rel="stylesheet" href="_static/css/nltk_theme.css"/>
  <link rel="stylesheet" href="_static/css/custom.css"/>

  <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
      <script type="text/javascript" src="_static/documentation_options.js"></script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/_sphinx_javascript_frameworks_compat.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
  

  <script src="https://email.tl.fortawesome.com/c/eJxNjUEOgyAQAF8jR7Kw6wIHDh7sP1Cw2mgxgmn6-3JsMqc5zEQfE8dkxOY1KKMUOI3ACFKRJpSW2AAp7ontYIaxI6i7XPJVwyeVfCQ550Os3jLrGSNOLgbdAy6s0PBk2TFNjEbsfq31LB0OnX407pJa5v2faRadwSW63mn5KuLyR9j2tgx3zecanl-55R_-jjPs"></script> 
</head>

<body>
  <div id="nltk-theme-container">
    <header>
      <div id="logo-container">
          
          <a href="index.html"><img src="_static/PEGG_notext.png"></a>
          
      </div>
      <div id="project-container">
        
        <h1>PEGG Documentation</h1>
        
      </div>

      <a id="menu-toggle" class="fa fa-bars" aria-hidden="true"></a>

      <script type="text/javascript">
        $("#menu-toggle").click(function() {
          $("#menu-toggle").toggleClass("toggled");
          $("#side-menu-container").slideToggle(300);
        });
      </script>
    </header>

    <div id="content-container">

      <div id="side-menu-container">

        <div id="search" role="search">
        <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
            <input type="text" name="q" placeholder="Search" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
        </form>
</div>

        <div id="side-menu" role="navigation">
          
  
    
  
  
    <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">🚀 Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#designing-pegrnas-with-pegg">Designing pegRNAs with PEGG</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#wt-alt-format">(1) WT-ALT Format</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#primedesign-format">(2) PrimeDesign Format</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#cbioportal-format">(3) cBioPortal Format</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#clinvar-mutations">(4) ClinVar mutations</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#generating-pegrnas">Generating pegRNAs</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#reference-file-formatting">Reference File Formatting</a><ul>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#input-mutations">(1) Input mutations</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#a-compatible-input-files">(1a) Compatible Input Files</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#b-clinvar-variants">(1b) ClinVar variants</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#reference-genome">(2) Reference Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#on-off-target-scores">(3) On/Off Target scores</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#using-pegg">Using PEGG</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#visualization-tools">Visualization Tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#oligo-generation">Oligo Generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#library-generation">Library Generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#jupyter-notebook-tutorial">Jupyter Notebook Tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#a-note-on-ram">A Note on RAM</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="jupyter_tutorial.html">🪐 Jupyter Notebook Tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">📚 Complete Function Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#pegg-prime-module">pegg.prime module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pegg-base-module">pegg.base module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pegg.library_design.library">pegg.library_design module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="about.html">❓ About PEGG</a></li>
<li class="toctree-l1"><a class="reference internal" href="other_design_tools.html">🚩Other pegRNA &amp; gRNA Design Tools</a><ul>
<li class="toctree-l2"><a class="reference internal" href="other_design_tools.html#pegrna-design-tools">pegRNA Design Tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="other_design_tools.html#base-editing-design-tools">base editing design tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="other_design_tools.html#off-target-analysis">off-target analysis</a></li>
</ul>
</li>
</ul>

  

        </div>

        
      </div>

      <div id="main-content-container">
        <div id="main-content" role="main">
          
  <div class="section" id="complete-function-documentation">
<h1>📚 Complete Function Documentation<a class="headerlink" href="#complete-function-documentation" title="Permalink to this heading">¶</a></h1>
<p>This page contains the complete documentation for each function in the pegg module.</p>
<div class="section" id="pegg-prime-module">
<h2>pegg.prime module<a class="headerlink" href="#pegg-prime-module" title="Permalink to this heading">¶</a></h2>
</div>
<div class="section" id="pegg-base-module">
<h2>pegg.base module<a class="headerlink" href="#pegg-base-module" title="Permalink to this heading">¶</a></h2>
</div>
<div class="section" id="module-pegg.library_design.library">
<span id="pegg-library-design-module"></span><h2>pegg.library_design module<a class="headerlink" href="#module-pegg.library_design.library" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.aavs1_muts">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">aavs1_muts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_muts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'GRCh37'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.aavs1_muts" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for generating list of “mutations” within AAVS1 locus as negative controls.
Generates a list of SNPs that contain idential reference and alternate alleles.
Intron 1-2 of PPP1R12C (AAVS1 locus): GRCh37 transcript = ENST00000263433.3 | GRCh38 transcript = ENST00000263433.8</p>
<p>Not currently used in the pipeline, but provided anyway.</p>
<dl>
<dt>chrom_dict</dt><dd><p><em>type = dict</em></p>
<p>Dictionary containing chromosomes in dictionary format for parsing by PEGG.</p>
</dd>
<dt>num_muts</dt><dd><p><em>type = int</em></p>
<p>Number of safe-targetting mutations to select/generate.</p>
</dd>
<dt>genome_version</dt><dd><p><em>type = str</em></p>
<p>Options = ‘GRCh37’, ‘GRCh38’</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.library_maker">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">library_maker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutant_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_safetarget</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'human'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_silent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_end_cds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.library_maker" title="Permalink to this definition">¶</a></dt>
<dd><p>Compiles the different library design functions to aggregate all of the variants for a particular gene,
include silent substitution controls at a designated %, and non-targeting controls. This can then be fed
into the pegg2.run() or base_editing.run_base() functions.</p>
<p>NOTE: THIS ONLY WORKS FOR GRCh37 and GRCm38!!!
Alternatively, generate these mutations/guides using these reference builds, and then generate your own guides
targeting mutations in your desired reference build.</p>
<dl>
<dt>mutant_input</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>DataFrame containing all of the input mutations. Doesn’t require filtration for gene of interest.</p>
</dd>
<dt>gene_name</dt><dd><p><em>type = str</em></p>
<p>Gene name to select from the mutant_input dataframe.</p>
</dd>
<dt>chrom_dict</dt><dd><p><em>type = dict</em></p>
<p>Dictionary containing the reference genome. See genome_loader()</p>
</dd>
<dt>fraction_safetarget</dt><dd><p><em>type = float</em></p>
<p>Value from 0 to 1 that corresponds to the fraction of safe-targetting mutations to include in the library.
(i.e. what fraction of the mutant_input)</p>
</dd>
<dt>organism</dt><dd><p><em>type = str</em></p>
<p>Options = ‘human’ or ‘mouse’. Determines which of the organisms to generate safe targeting guides for.</p>
</dd>
<dt>fraction_silent</dt><dd><p><em>type = float</em></p>
<p>Value from 0 to 1 that corresponds to the fraction of silent mutations to include in the library.
(i.e. what fraction of the mutant_input)</p>
</dd>
<dt>chrom</dt><dd><p><em>type = int or str</em></p>
<p>Chromosome that the gene of interest falls on.</p>
</dd>
<dt>strand</dt><dd><p><em>type = str</em></p>
<p>‘+’ or ‘-‘ – corresponds with which strand the transcript falls on.</p>
</dd>
<dt>start_end_cds</dt><dd><p><em>type = list</em></p>
<p>Nested list that contains the CDS locations of the gene transcript of interest, in the + strand orientation.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.matrix_rep_library">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">matrix_rep_library</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_end_cds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.matrix_rep_library" title="Permalink to this definition">¶</a></dt>
<dd><p>LEGACY FUNCTION FROM PEGG1.0 (NOT UPDATED)! – untested; will likely throw errors</p>
<p>Visualization function for looking at the number of pegRNAs generated in each codon of target gene.
Also visualizes the identitity of mutant amino acid for SNPs in the input dataframe.</p>
<p>FOR PROPER FUNCTION, REQUIRES USER INCLUSION OF HGVSp INFORMATION IN RF. See documentation for more info.</p>
<p>Returns matrix_snp, matrix_ins, matrix_del, matrix_neutral, fig. First four return objects correspond with
matrices used for plotting. Fig = figure, which can be saved/exported.</p>
<dl>
<dt>rf</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>A dataframe containing the pegRNAs generated by run().</p>
</dd>
<dt>gene_name</dt><dd><p><em>type = str</em></p>
<p>Gene’s Hugo_Symbol (i.e. name).</p>
</dd>
<dt>start_end_cds</dt><dd><p><em>type = list</em></p>
<p>A 2-d list containing the start/end locations of each region of the coding sequence (CDS) for the gene’s selected transcript.
See documentation for example and precise specifications of format.</p>
</dd>
<dt>strand</dt><dd><p><em>type = str</em></p>
<p>Strand that transcript is on. Options are ‘+’ or ‘-‘.</p>
</dd>
<dt>plot</dt><dd><p><em>type = boolean (‘True’ or ‘False’)</em></p>
<p>If true, show stacked bar plot. If false, simply return the matrices.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.mutation_aggregator">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">mutation_aggregator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutant_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.mutation_aggregator" title="Permalink to this definition">¶</a></dt>
<dd><p>Selects the mutations that correspond to the desired gene name.
Removes duplicates (checking at DNA level; not amino acid level).
Returns dataframe containing these aggregated mutants occuring in gene_name.</p>
<dl>
<dt>mutant_input</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>A dataframe containing the input mutations from which selections are made to generate pegRNAs.
See documentation for precise qualities of this dataframe</p>
</dd>
<dt>gene_name</dt><dd><p><em>type = str</em></p>
<p>Gene’s Hugo_Symbol (i.e. name).</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.neutral_substitutions">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">neutral_substitutions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_end_cds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.neutral_substitutions" title="Permalink to this definition">¶</a></dt>
<dd><p>A function for generating all possible synonymous codon substitutions (i.e. silent mutations) of a given gene.
See documentation for more information about format of start_end_cds &amp; example usage.</p>
<dl>
<dt>gene_name</dt><dd><p><em>type = str</em></p>
<p>Gene’s Hugo_Symbol (i.e. name).</p>
</dd>
<dt>chrom</dt><dd><p><em>type = str</em></p>
<p>Chromosome that gene occurs on. Format = e.g. ‘chr17’, ‘chrX’, etc.</p>
</dd>
<dt>strand</dt><dd><p><em>type = str</em></p>
<p>Strand that transcript is on. Options are ‘+’ or ‘-‘.</p>
</dd>
<dt>start_end_cds</dt><dd><p><em>type = list</em></p>
<p>A 2-d list containing the start/end locations of each region of the coding sequence (CDS) for the gene’s selected transcript.
See documentation for example and precise specifications of format.</p>
</dd>
<dt>chrom_dict</dt><dd><p><em>type = dict</em></p>
<p>Dictionary containing the reference genome. See genome_loader()</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.nontargeting_guides">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">nontargeting_guides</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_guides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edit_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'prime'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.nontargeting_guides" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for generating a list of non-targetting guides (in the human genome).
Future versions will expand to mouse as well…Unprocessed files for mouse located within this module.
Guides taken from: <a class="reference external" href="https://www.nature.com/articles/nmeth.4423">https://www.nature.com/articles/nmeth.4423</a> (<a class="reference external" href="https://doi.org/10.1038/nmeth.4423">https://doi.org/10.1038/nmeth.4423</a>)</p>
<dl>
<dt>num_guides</dt><dd><p><em>type = int</em></p>
<p>Number of guides to non-targetting generate. Max = 1000.</p>
</dd>
<dt>edit_type</dt><dd><p><em>type = str</em></p>
<p>Options = ‘base’, ‘prime’</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.safe_muts">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">safe_muts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_muts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'human'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.safe_muts" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for generating a list of safe-targetting pegRNAs/gRNAs.
NOTE: THIS ONLY WORKS FOR GRCh37 and GRCm38!!!
Alternatively, generate these mutations/guides using these reference builds, and then generate your own guides
targeting mutations in your desired reference build.</p>
<p>This is a random subset of 100,000 “safe regions” taken from Morgens et al., 2017 (<a class="reference external" href="https://doi.org/10.1038/ncomms15178">https://doi.org/10.1038/ncomms15178</a>)</p>
<dl>
<dt>num_muts</dt><dd><p><em>type = int</em></p>
<p>Number of safe-targetting mutations to select/generate.</p>
</dd>
<dt>chrom_dict</dt><dd><p><em>type = dict</em></p>
<p>Dictionary containing chromosomes in dictionary format for parsing by PEGG.</p>
</dd>
<dt>organism</dt><dd><p><em>type = str</em></p>
<p>Choices = “human” or “mouse”. Generates safe targeting guides in human or mouse genome.</p>
</dd>
</dl>
</dd></dl>

</div>
</div>


        </div>
      </div>

    </div>

<footer>
    <div id="footer-info">
        <ul id="build-details">
            
                <li class="footer-element">
                    
                        <a href="_sources/PEGG.rst.txt" rel="nofollow"> source</a>
                    
                </li>
            

            

            
        </ul>

        
            <div id="copyright">
                &copy; 2023, Samuel Gould
            </div>
        

        <div id="credit">
            created with <a href="http://sphinx-doc.org/">Sphinx</a> and <a href="https://github.com/tomaarsen/nltk_theme">NLTK Theme</a>
        </div>
    </div>
</footer> 

</div>

</body>
</html>