<!DOCTYPE html>
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="theme-color" content="#2D2D2D" />
  
  <title>PEGG :: ğŸ“š Complete Function Documentation</title>
  
  <link rel="icon" type="image/png" href="_static/PEGG_notext.ico">
  
  <link rel="index" title="Index" href="genindex.html"/>

  <link rel="stylesheet" href="_static/css/nltk_theme.css"/>
  <link rel="stylesheet" href="_static/css/custom.css"/>

  <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
      <script type="text/javascript" src="_static/documentation_options.js"></script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/_sphinx_javascript_frameworks_compat.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
  

  <script src="https://email.tl.fortawesome.com/c/eJxNjUEOgyAQAF8jR7Kw6wIHDh7sP1Cw2mgxgmn6-3JsMqc5zEQfE8dkxOY1KKMUOI3ACFKRJpSW2AAp7ontYIaxI6i7XPJVwyeVfCQ550Os3jLrGSNOLgbdAy6s0PBk2TFNjEbsfq31LB0OnX407pJa5v2faRadwSW63mn5KuLyR9j2tgx3zecanl-55R_-jjPs"></script> 
</head>

<body>
  <div id="nltk-theme-container">
    <header>
      <div id="logo-container">
          
          <a href="index.html"><img src="_static/PEGG_notext.png"></a>
          
      </div>
      <div id="project-container">
        
        <h1>PEGG Documentation</h1>
        
      </div>

      <a id="menu-toggle" class="fa fa-bars" aria-hidden="true"></a>

      <script type="text/javascript">
        $("#menu-toggle").click(function() {
          $("#menu-toggle").toggleClass("toggled");
          $("#side-menu-container").slideToggle(300);
        });
      </script>
    </header>

    <div id="content-container">

      <div id="side-menu-container">

        <div id="search" role="search">
        <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
            <input type="text" name="q" placeholder="Search" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
        </form>
</div>

        <div id="side-menu" role="navigation">
          
  
    
  
  
    <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">ğŸš€ Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#required-input-files">Required Input Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#reference-file-formatting">Reference File Formatting</a><ul>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#input-mutations">(1) Input mutations</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#a-compatible-input-files">(1a) Compatible Input Files</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#b-clinvar-variants">(1b) ClinVar variants</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#reference-genome">(2) Reference Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#on-off-target-scores">(3) On/Off Target scores</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#using-pegg">Using PEGG</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#visualization-tools">Visualization Tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#oligo-generation">Oligo Generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#library-generation">Library Generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#jupyter-notebook-tutorial">Jupyter Notebook Tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#a-note-on-ram">A Note on RAM</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="jupyter_tutorial.html">ğŸª Jupyter Notebook Tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">ğŸ“š Complete Function Documentation</a></li>
</ul>

  

        </div>

        
      </div>

      <div id="main-content-container">
        <div id="main-content" role="main">
          
  <div class="section" id="complete-function-documentation">
<h1>ğŸ“š Complete Function Documentation<a class="headerlink" href="#complete-function-documentation" title="Permalink to this heading">Â¶</a></h1>
<p>This page contains the complete documentation for each function in the pegg module.</p>
<span class="target" id="module-pegg.prime.pegg2"></span><dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.PAM_finder">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">PAM_finder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.PAM_finder" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Finds indeces of PAM sequences
Formatting: e.g. â€œNGGâ€
N = [A|T|C|G]
R = [A|G]
Y = [C|T]
S = [G|C]
W = [A|T]
K = [G|T]
M = [A|C]
B = [C|G|T]
D = [A|G|T]
H = [A|C|T]
V = [A|C|G]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.RF_score">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">RF_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.RF_score" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.clinvar_VCF_translator">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">clinvar_VCF_translator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variation_ids</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.clinvar_VCF_translator" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Function that takes a clinvar.vcf.gz file containing information about variants, as well as a list of variation ID numbers,
and returns a pandas dataframe containing the variants in a format that can be used by PEGG to design pegRNAs.</p>
<dl>
<dt>filepath</dt><dd><p><em>type = str</em></p>
<p>Filepath to the clinvar.vcf.gz file.</p>
</dd>
<dt>variation_ids</dt><dd><p><em>type = list</em></p>
<p>List of variation IDs that the user wants to convert to .</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.df_formatter">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">df_formatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">120</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.df_formatter" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Takes in variants (in cBioPortal format!)
and outputs dataframe with WT and ALT oligos with designated context_size
context_size = the amount of nt on either side of the variant e.g. AAA(A/G)AAA = context_size of 3
cols_to_save = information to save about the mutationsâ€¦(can modify beyond defaultsâ€¦)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.eligible_PAM_finder">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">eligible_PAM_finder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_RTT_length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">19</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.eligible_PAM_finder" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Determines eligible PAM sequences for creating pegRNAsâ€¦</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.genome_loader">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">genome_loader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath_gz</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.genome_loader" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Takes in filepath of human genome (GrCH37 or GrCh38) and returns records and index_list for PEGG parsing.</p>
<dl>
<dt>filepath_gz</dt><dd><p><em>type = str</em></p>
<p>The filepath to the .gz file holding the reference genome file.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.input_formatter">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">input_formatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context_size</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.input_formatter" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.make_aligner">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">make_aligner</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.make_aligner" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.mut_formatter">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">mut_formatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.mut_formatter" title="Permalink to this definition">Â¶</a></dt>
<dd><p>This program will break for more complex mutations (e.g. INDELs)</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pegg.prime.pegg2.mutation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">mutation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wt_w_context</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt_w_context</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.mutation" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.ontarget_score">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">ontarget_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.ontarget_score" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.other_filtration">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">other_filtration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pegRNA_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RE_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polyT_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.other_filtration" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Determines whether pegRNAs contain polyT termination sites or Restriction Enzyme Sites. Returns pegRNA_df with properties added.
Note: RE_sites requires only A, T, C, or G bases (Ns will not work currently; need to do custom filtration)â€¦
These properties are checking for:</p>
<ol class="arabic simple">
<li><p>u6 terminator (polyT sequence)</p></li>
<li><p>RE site presence</p></li>
</ol>
<dl>
<dt>pegRNA_df</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>A dataframe containing the pegRNAs for the selected input mutations. Generated by run() or pegRNA_generator()</p>
</dd>
<dt>RE_sites</dt><dd><p><em>type = list or None</em></p>
<p>A list containing the RE recognition sites to filter (e.g. [â€˜CGTCTCâ€™, â€˜GAATTCâ€™] for Esp3I and EcoRI). Default = None (no filtration).</p>
</dd>
<dt>polyT_threshold</dt><dd><p><em>type = int</em></p>
<p>The length of the polyT sequence to classify as a terminator. Default = 4.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.pegRNA_generator">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">pegRNA_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RTT_lengths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PBS_lengths</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.pegRNA_generator" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.peggscore2">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">peggscore2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.peggscore2" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.primedesign_formatter">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">primedesign_formatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.primedesign_formatter" title="Permalink to this definition">Â¶</a></dt>
<dd><p>takes as input sequence in prime design format 
e.g. AATTCCG(G/C)AATTCGCT</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.run">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'NGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rankby</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PEGG2_Score'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pegRNAs_per_mut</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'All'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RTT_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[5,</span> <span class="pre">10,</span> <span class="pre">15,</span> <span class="pre">25,</span> <span class="pre">30]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PBS_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[8,</span> <span class="pre">10,</span> <span class="pre">13,</span> <span class="pre">15]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_RHA_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RE_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polyT_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">19</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">120</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before_proto_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reverse-complement'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.run" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.sensor_generator">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">sensor_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before_proto_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reverse-complement'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.sensor_generator" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Generates sensor sequence for quantification of pegRNA editing outcomes
Automatically puts sensor in reverse complement orientation with respect to protospacer
This is highly reccomended to reduce recombination during cloning/library preparation</p>
<p>options for sensor_orienation = â€˜reverse_compâ€™; â€˜forwardâ€™</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.sensor_viz">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">sensor_viz</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_w_sensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.sensor_viz" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegg2.split">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.pegg2.</span></span><span class="sig-name descname"><span class="pre">split</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">word</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegg2.split" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Simple function for splitting string into component characters</p>
</dd></dl>

<span class="target" id="module-pegg.base.base_editing"></span><dl class="py function">
<dt class="sig sig-object py" id="pegg.base.base_editing.eligible_PAM_finder_base">
<span class="sig-prename descclassname"><span class="pre">pegg.base.base_editing.</span></span><span class="sig-name descname"><span class="pre">eligible_PAM_finder_base</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">19</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.base.base_editing.eligible_PAM_finder_base" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Determines eligible PAM sequences for creating pegRNAsâ€¦</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.base.base_editing.gRNA_generator">
<span class="sig-prename descclassname"><span class="pre">pegg.base.base_editing.</span></span><span class="sig-name descname"><span class="pre">gRNA_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ideal_edit_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[4,</span> <span class="pre">8]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.base.base_editing.gRNA_generator" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.base.base_editing.run_base">
<span class="sig-prename descclassname"><span class="pre">pegg.base.base_editing.</span></span><span class="sig-name descname"><span class="pre">run_base</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'NGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filtration</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ABE+CBE'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ideal_edit_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[4,</span> <span class="pre">8]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auto_SNP_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">19</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">120</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RE_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polyT_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before_proto_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">40</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reverse-complement'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.base.base_editing.run_base" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.base.base_editing.sensor_generator_base">
<span class="sig-prename descclassname"><span class="pre">pegg.base.base_editing.</span></span><span class="sig-name descname"><span class="pre">sensor_generator_base</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before_proto_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reverse-complement'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.base.base_editing.sensor_generator_base" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Generates sensor sequence for quantification of pegRNA editing outcomes
Automatically puts sensor in reverse complement orientation with respect to protospacer
This is highly reccomended to reduce recombination during cloning/library preparation</p>
<p>options for sensor_orienation = â€˜reverse_compâ€™; â€˜forwardâ€™</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.base.base_editing.sensor_viz_base">
<span class="sig-prename descclassname"><span class="pre">pegg.base.base_editing.</span></span><span class="sig-name descname"><span class="pre">sensor_viz_base</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_w_sensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.base.base_editing.sensor_viz_base" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.base.base_editing.split">
<span class="sig-prename descclassname"><span class="pre">pegg.base.base_editing.</span></span><span class="sig-name descname"><span class="pre">split</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">word</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.base.base_editing.split" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Simple function for splitting string into component characters</p>
</dd></dl>

<span class="target" id="module-pegg.library_design.library"></span><dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.aavs1_muts">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">aavs1_muts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_muts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'GRCh37'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.aavs1_muts" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Function for generating list of â€œmutationsâ€ within AAVS1 locus as negative controls.
Generates a list of SNPs that contain idential reference and alternate alleles.
Intron 1-2 of PPP1R12C (AAVS1 locus): GRCh37 transcript = ENST00000263433.3 | GRCh38 transcript = ENST00000263433.8</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.library_maker">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">library_maker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutant_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_safetarget</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'human'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_silent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_end_cds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.library_maker" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Compiles the different library design functions to aggregate all of the variants for a particular gene,
include silent substitution controls at a designated %, and non-targeting controls. This can then be fed
into the pegg2.run() or base_editing.run_base() functions.</p>
<p>NOTE: THIS ONLY WORKS FOR GRCh37 and GRCm38!!!
Alternatively, generate these mutations/guides using these reference builds, and then generate your own guides
targeting mutations in your desired reference build.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.matrix_rep_library">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">matrix_rep_library</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_end_cds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.matrix_rep_library" title="Permalink to this definition">Â¶</a></dt>
<dd><p>LEGACY FUNCTION FROM PEGG1.0 (NOT UPDATED)! â€“ untested; will likely throw errors</p>
<p>Visualization function for looking at the number of pegRNAs generated in each codon of target gene.
Also visualizes the identitity of mutant amino acid for SNPs in the input dataframe.</p>
<p>FOR PROPER FUNCTION, REQUIRES USER INCLUSION OF HGVSp INFORMATION IN RF. See documentation for more info.</p>
<p>Returns matrix_snp, matrix_ins, matrix_del, matrix_neutral, fig. First four return objects correspond with
matrices used for plotting. Fig = figure, which can be saved/exported.</p>
<dl>
<dt>rf</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>A dataframe containing the pegRNAs generated by run().</p>
</dd>
<dt>gene_name</dt><dd><p><em>type = str</em></p>
<p>Geneâ€™s Hugo_Symbol (i.e. name).</p>
</dd>
<dt>start_end_cds</dt><dd><p><em>type = list</em></p>
<p>A 2-d list containing the start/end locations of each region of the coding sequence (CDS) for the geneâ€™s selected transcript.
See documentation for example and precise specifications of format.</p>
</dd>
<dt>strand</dt><dd><p><em>type = str</em></p>
<p>Strand that transcript is on. Options are â€˜+â€™ or â€˜-â€˜.</p>
</dd>
<dt>plot</dt><dd><p><em>type = boolean (â€˜Trueâ€™ or â€˜Falseâ€™)</em></p>
<p>If true, show stacked bar plot. If false, simply return the matrices.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.mutation_aggregator">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">mutation_aggregator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutant_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.mutation_aggregator" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Selects the mutations that correspond to the desired gene name.
Removes duplicates (checking at DNA level; not amino acid level).
Returns dataframe containing these aggregated mutants occuring in gene_name.</p>
<dl>
<dt>mutant_input</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>A dataframe containing the input mutations from which selections are made to generate pegRNAs.
See documentation for precise qualities of this dataframe</p>
</dd>
<dt>gene_name</dt><dd><p><em>type = str</em></p>
<p>Geneâ€™s Hugo_Symbol (i.e. name).</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.neutral_substitutions">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">neutral_substitutions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_end_cds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.neutral_substitutions" title="Permalink to this definition">Â¶</a></dt>
<dd><p>A function for generating all possible synonymous codon substitutions (i.e. silent mutations) of a given gene.
See documentation for more information about format of start_end_cds &amp; example usage.</p>
<dl>
<dt>gene_name</dt><dd><p><em>type = str</em></p>
<p>Geneâ€™s Hugo_Symbol (i.e. name).</p>
</dd>
<dt>chrom</dt><dd><p><em>type = str</em></p>
<p>Chromosome that gene occurs on. Format = e.g. â€˜chr17â€™, â€˜chrXâ€™, etc.</p>
</dd>
<dt>strand</dt><dd><p><em>type = str</em></p>
<p>Strand that transcript is on. Options are â€˜+â€™ or â€˜-â€˜.</p>
</dd>
<dt>start_end_cds</dt><dd><p><em>type = list</em></p>
<p>A 2-d list containing the start/end locations of each region of the coding sequence (CDS) for the geneâ€™s selected transcript.
See documentation for example and precise specifications of format.</p>
</dd>
<dt>chrom_dict</dt><dd><p><em>type = dict</em></p>
<p>Dictionary containing the reference genome. See genome_loader()</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.nontargeting_guides">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">nontargeting_guides</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_guides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edit_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'prime'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.nontargeting_guides" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Function for generating a list of non-targetting guides (in the human genome).
Future versions will expand to mouse as wellâ€¦</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library_design.library.safe_muts">
<span class="sig-prename descclassname"><span class="pre">pegg.library_design.library.</span></span><span class="sig-name descname"><span class="pre">safe_muts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_muts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'human'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library_design.library.safe_muts" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Function for generating a list of safe-targetting pegRNAs/gRNAs.
NOTE: THIS ONLY WORKS FOR GRCh37 and GRCm38!!!
Alternatively, generate these mutations/guides using these reference builds, and then generate your own guides
targeting mutations in your desired reference build.</p>
<p>This is a random subset of 100,000 â€œsafe regionsâ€ taken from Morgens et al., 2017 (<a class="reference external" href="https://doi.org/10.1038/ncomms15178">https://doi.org/10.1038/ncomms15178</a>)</p>
</dd></dl>

</div>


        </div>
      </div>

    </div>

<footer>
    <div id="footer-info">
        <ul id="build-details">
            
                <li class="footer-element">
                    
                        <a href="_sources/PEGG.rst.txt" rel="nofollow"> source</a>
                    
                </li>
            

            

            
        </ul>

        
            <div id="copyright">
                &copy; 2023, Samuel Gould
            </div>
        

        <div id="credit">
            created with <a href="http://sphinx-doc.org/">Sphinx</a> and <a href="https://github.com/tomaarsen/nltk_theme">NLTK Theme</a>
        </div>
    </div>
</footer> 

</div>

</body>
</html>