<!DOCTYPE html>
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="theme-color" content="#2D2D2D" />
  
  <title>PEGG :: 📚 Complete Function Documentation</title>
  
  <link rel="icon" type="image/png" href="_static/PEGG_notext.ico">
  
  <link rel="index" title="Index" href="genindex.html"/>

  <link rel="stylesheet" href="_static/css/nltk_theme.css"/>
  <link rel="stylesheet" href="_static/css/custom.css"/>

  <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
      <script type="text/javascript" src="_static/documentation_options.js"></script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/_sphinx_javascript_frameworks_compat.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
  

  <script src="https://email.tl.fortawesome.com/c/eJxNjUEOgyAQAF8jR7Kw6wIHDh7sP1Cw2mgxgmn6-3JsMqc5zEQfE8dkxOY1KKMUOI3ACFKRJpSW2AAp7ontYIaxI6i7XPJVwyeVfCQ550Os3jLrGSNOLgbdAy6s0PBk2TFNjEbsfq31LB0OnX407pJa5v2faRadwSW63mn5KuLyR9j2tgx3zecanl-55R_-jjPs"></script> 
</head>

<body>
  <div id="nltk-theme-container">
    <header>
      <div id="logo-container">
          
          <a href="index.html"><img src="_static/PEGG_notext.png"></a>
          
      </div>
      <div id="project-container">
        
        <h1>PEGG Documentation</h1>
        
      </div>

      <a id="menu-toggle" class="fa fa-bars" aria-hidden="true"></a>

      <script type="text/javascript">
        $("#menu-toggle").click(function() {
          $("#menu-toggle").toggleClass("toggled");
          $("#side-menu-container").slideToggle(300);
        });
      </script>
    </header>

    <div id="content-container">

      <div id="side-menu-container">

        <div id="search" role="search">
        <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
            <input type="text" name="q" placeholder="Search" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
        </form>
</div>

        <div id="side-menu" role="navigation">
          
  
    
  
  
    <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">🚀 Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#input-formatting">Input Formatting</a><ul>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#wt-alt-format">(1) WT-ALT Format</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#primedesign-format">(2) PrimeDesign Format</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#cbioportal-format">(3) cBioPortal Format</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#b-reference-genome">(3b) Reference Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#clinvar-mutations">(4) ClinVar mutations</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#generating-pegrnas">Generating pegRNAs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#design-options">Design Options</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#base-editing">Base Editing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="quickstart.html#id2">Design Options</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#sensor-design">Sensor Design</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#oligo-generation">Oligo Generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#library-design">Library Design</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart.html#visualizations">Visualizations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="jupyter_tutorial.html">🪐 Jupyter Notebook Tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">📚 Complete Function Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-pegg.prime">pegg.prime</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pegg.base">pegg.base</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pegg.library">pegg.library</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="about.html">❓ About PEGG</a></li>
<li class="toctree-l1"><a class="reference internal" href="other_design_tools.html">🚩Other pegRNA &amp; gRNA Design Tools</a><ul>
<li class="toctree-l2"><a class="reference internal" href="other_design_tools.html#pegrna-design-tools">pegRNA Design Tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="other_design_tools.html#base-editing-design-tools">base editing design tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="other_design_tools.html#off-target-analysis">off-target analysis</a></li>
</ul>
</li>
</ul>

  

        </div>

        
      </div>

      <div id="main-content-container">
        <div id="main-content" role="main">
          
  <div class="section" id="complete-function-documentation">
<h1>📚 Complete Function Documentation<a class="headerlink" href="#complete-function-documentation" title="Permalink to this heading">¶</a></h1>
<p>This page contains the complete documentation for each function in the pegg module.</p>
<div class="section" id="module-pegg.prime">
<span id="pegg-prime"></span><h2>pegg.prime<a class="headerlink" href="#module-pegg.prime" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.PAM_finder">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">PAM_finder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.PAM_finder" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds indeces of PAM sequences in sequence.</p>
<dl>
<dt>seq</dt><dd><p><em>type = str</em></p>
<p>Sequence to search for PAM sequences.</p>
</dd>
<dt>PAM</dt><dd><p><em>type = str</em></p>
<p>PAM sequence for searching. Formatting: e.g. “NGG”</p>
<p>N = [A|T|C|G]
R = [A|G]
Y = [C|T]
S = [G|C]
W = [A|T]
K = [G|T]
M = [A|C]
B = [C|G|T]
D = [A|G|T]
H = [A|C|T]
V = [A|C|G]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.RF_score">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">RF_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.RF_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for calculating the Random Forest score from pegRNA parameters.</p>
<dl>
<dt>df</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>Dataframe containing the pegRNAs generated by run().</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.clinvar_VCF_translator">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">clinvar_VCF_translator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variation_ids</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.clinvar_VCF_translator" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that takes a clinvar.vcf.gz file containing information about variants, as well as a list of variation ID numbers,
and returns a pandas dataframe containing the variants in a format that can be used by PEGG to design pegRNAs.</p>
<dl>
<dt>filepath</dt><dd><p><em>type = str</em></p>
<p>Filepath to the clinvar.vcf.gz file.</p>
</dd>
<dt>variation_ids</dt><dd><p><em>type = list</em></p>
<p>List of variation IDs that the user wants to convert to .</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.df_formatter">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">df_formatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">120</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.df_formatter" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes in variants (in cBioPortal format!) and outputs dataframe with REF and ALT oligos with designated context_size
that can be used by PEGG for creating pegRNAs.</p>
<dl>
<dt>df</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>The dataframe of input variants in cBioPortal format.</p>
</dd>
<dt>chrom_dict</dt><dd><p><em>type = dict</em></p>
<p>Dictionary generated by genome_loader() that holds the chromosome sequences.</p>
</dd>
<dt>context_size</dt><dd><p><em>type = int</em></p>
<p>The amount of context/flanking sequence on either side of the mutation to generate. For larger variants, set this larger.
Default = 120. e.g. AAA(A/G)AAA = context_size of 3</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.eligible_PAM_finder">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">eligible_PAM_finder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_RTT_length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">19</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.eligible_PAM_finder" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines eligible PAM sequences for creating pegRNAs.
Returns PAM sequence locations on (1) forward, and (2) reverse-complement strand</p>
<dl>
<dt>mut</dt><dd><p><em>type = mutation class</em></p>
<p>See class: mutation</p>
</dd>
<dt>PAM</dt><dd><p><em>type = str</em></p>
<p>PAM sequence for searching. Formatting: e.g. “NGG”</p>
<p>N = [A|T|C|G]
R = [A|G]
Y = [C|T]
S = [G|C]
W = [A|T]
K = [G|T]
M = [A|C]
B = [C|G|T]
D = [A|G|T]
H = [A|C|T]
V = [A|C|G]</p>
</dd>
<dt>max_RTT_length</dt><dd><p><em>type = int</em></p>
<p>Max RTT length for pegRNAs being designed.</p>
</dd>
<dt>proto_size</dt><dd><p><em>type = int</em></p>
<p>Size of protospacer being used. Default = 19 (G+19).</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.genome_loader">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">genome_loader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath_gz</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.genome_loader" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes in filepath of human or mouse genome and returns dictionary of chromosome sequences that PEGG can parse.
Tested only on human and mouse genomes.</p>
<p>Returns (1) chromosome dictionary, and (2) file names of the chromosomes that are stored (for manual checking of errors).</p>
<dl>
<dt>filepath_gz</dt><dd><p><em>type = str</em></p>
<p>The filepath to the .gz file holding the reference genome file.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.goldengate_oligos">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">goldengate_oligos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peg_df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.goldengate_oligos" title="Permalink to this definition">¶</a></dt>
<dd><p>Currently a place-holder. Going to implement automated Golden Gate oligo generator (so scaffold doesn’t need to be synthesized).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.input_formatter">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">input_formatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context_size</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.input_formatter" title="Permalink to this definition">¶</a></dt>
<dd><p>Master function for putting input mutations in the correct format for pegRNA/gRNA design.</p>
<dl>
<dt>input_df</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>Pandas dataframe that contains the input mutations.</p>
</dd>
<dt>input_format</dt><dd><p><em>type = str</em></p>
<p>Options = ‘cBioPortal’, ‘WT_ALT’, ‘PrimeDesign’. For ‘WT_ALT’, make sure to put headers as “WT” and “ALT”. 
For “PrimeDesign”, put the header as “SEQ”.</p>
</dd>
<dt>chrom_dict</dt><dd><p><em>type = dict</em></p>
<p>Dictionary generated by genome_loader() that holds the chromosome sequences.</p>
</dd>
<dt>context_size</dt><dd><p><em>type = int</em></p>
<p>The amount of context/flanking sequence on either side of the mutation to generate. For larger variants, set this larger.
Default = 120. e.g. AAA(A/G)AAA = context_size of 3</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.make_aligner">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">make_aligner</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.make_aligner" title="Permalink to this definition">¶</a></dt>
<dd><p>Aligner function from Bio.Align.PairwiseAligner with custom parameters.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.mut_formatter">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">mut_formatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.mut_formatter" title="Permalink to this definition">¶</a></dt>
<dd><p>Formats mutations for pegRNA generation. Takes in WT and ALT sequences, returns necessary parameters for gRNA/pegRNA generation.
Note: This function will break with INDELs and complex variants since it relies on a simple alignment.</p>
<dl>
<dt>wt</dt><dd><p><em>type = str</em></p>
<p>WT sequence</p>
</dd>
<dt>alt</dt><dd><p><em>type = str</em></p>
<p>Mutant/alternate sequence.</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pegg.prime.mutation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">mutation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wt_w_context</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt_w_context</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.mutation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class for storing information about individual mutations. Used by functions throughout.
Prevents the use of disorganized lists.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.ontarget_score">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">ontarget_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.ontarget_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Calls to crisporEffScores (taken from CRISPOR github) to generate on-target scores using Rule Set 2.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.other_filtration">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">other_filtration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pegRNA_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RE_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polyT_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.other_filtration" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines whether pegRNAs contain polyT termination sites or Restriction Enzyme Sites. Returns pegRNA_df with properties added.
Note: RE_sites requires only A, T, C, or G bases (Ns will not work currently; need to do custom filtration)…
These properties are checking for:</p>
<ol class="arabic simple">
<li><p>u6 terminator (polyT sequence)</p></li>
<li><p>RE site presence</p></li>
</ol>
<dl>
<dt>pegRNA_df</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>A dataframe containing the pegRNAs for the selected input mutations. Generated by run() or pegRNA_generator()</p>
</dd>
<dt>RE_sites</dt><dd><p><em>type = list or None</em></p>
<p>A list containing the RE recognition sites to filter (e.g. [‘CGTCTC’, ‘GAATTC’] for Esp3I and EcoRI). Default = None (no filtration).</p>
</dd>
<dt>polyT_threshold</dt><dd><p><em>type = int</em></p>
<p>The length of the polyT sequence to classify as a terminator. Default = 4.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.pegRNA_generator">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">pegRNA_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RTT_lengths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PBS_lengths</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.pegRNA_generator" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates pegRNAs for a given mutation.</p>
<dl>
<dt>mut</dt><dd><p><em>type = mutation class</em></p>
<p>See class: mutation</p>
</dd>
<dt>PAM</dt><dd><p><em>type = str</em></p>
<p>PAM sequence for searching. Formatting: e.g. “NGG”</p>
<p>N = [A|T|C|G]
R = [A|G]
Y = [C|T]
S = [G|C]
W = [A|T]
K = [G|T]
M = [A|C]
B = [C|G|T]
D = [A|G|T]
H = [A|C|T]
V = [A|C|G]</p>
</dd>
<dt>orientation</dt><dd><p><em>type = str</em></p>
<p>‘+’ or ‘-‘</p>
</dd>
<dt>proto_size</dt><dd><p><em>type = int</em></p>
<p>Size of protospacer being used. Default = 19 (G+19).</p>
</dd>
<dt>RTT_lengths</dt><dd><p><em>type = list</em></p>
<p>List containing RTT lengths to design pegRNAs for.</p>
</dd>
<dt>PBS_length</dt><dd><p><em>type = list</em></p>
<p>List containing PBS lengths to desing pegRNAs for.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.peggscore2">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">peggscore2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.peggscore2" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for calculating the PEGG2 Score. A multiple linear regression score based on pegRNA parameters.</p>
<dl>
<dt>df</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>Dataframe containing the pegRNAs generated by run().</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.prime_oligo_generator">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">prime_oligo_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peg_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epeg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epeg_motif</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tevopreQ1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">five_prime_adapter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'AGCGTACACGTCTCACACC'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">three_prime_adapter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'GAATTCTAGATCCGGTCGTCAAC'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gRNA_scaff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGC'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.prime_oligo_generator" title="Permalink to this definition">¶</a></dt>
<dd><p>A tool for automatically generating oligos from the output of run().
Returns input dataframe with new columns containing the pegRNA oligo.</p>
<dl>
<dt>peg_df</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>A dataframe containing the pegRNAs for the selected input mutations. Generated by run() or pegRNA_generator()</p>
</dd>
<dt>epeg</dt><dd><p><em>type = bool</em></p>
<p>True/False whether to include an epegRNA motif at the end of the 3’ extension.</p>
</dd>
<dt>epeg_motif</dt><dd><p><em>type = str</em></p>
<p>Which epeg motif to include at end of 3’ extension. Default = “tevopreQ1”; Other option = “mpknot” (both from Chen et al.).
Can also input a custom motif sequence (just put in the sequence in 5’ to 3’ orientation)!</p>
</dd>
<dt>five_prime_adapter</dt><dd><p><em>type = str</em></p>
<p>5’ Prime Adapter. The automatically provided 5’ adapter contains an Esp3I (BsmBI) site. Can be swapped with 
whatever input string user wants.</p>
</dd>
<dt>three_prime_adapter</dt><dd><p><em>type = str</em></p>
<p>5’ Prime Adapter. The automatically provided 5’ adapter contains an Esp3I (BsmBI) site. Can be swapped with 
whatever input string user wants.</p>
</dd>
<dt>gRNA_scaff</dt><dd><p><em>type = str</em></p>
<p>gRNA scaffold region. Automatically set to a functional gRNA scaffold. Can be swapped with 
whatever input string user wants.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.primedesign_formatter">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">primedesign_formatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.primedesign_formatter" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes as input sequence in prime design format e.g. AATTCCG(G/C)AATTCGCT. Returns necessary parameters for pegRNA generation.</p>
<dl>
<dt>seq</dt><dd><p><em>type = str</em></p>
<p>The PrimeDesign formatted sequence.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.run">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'NGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rankby</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PEGG2_Score'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pegRNAs_per_mut</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'All'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RTT_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[5,</span> <span class="pre">10,</span> <span class="pre">15,</span> <span class="pre">25,</span> <span class="pre">30]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PBS_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[8,</span> <span class="pre">10,</span> <span class="pre">13,</span> <span class="pre">15]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_RHA_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RE_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polyT_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">19</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">120</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before_proto_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reverse-complement'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.run" title="Permalink to this definition">¶</a></dt>
<dd><p>Master function for generating pegRNAs. Takes as input a dataframe containing mutations in one of the acceptable formats.
Returns a dataframe with pegRNAs with desired design parameters.</p>
<dl>
<dt>input_df</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>Pandas dataframe that contains the input mutations.</p>
</dd>
<dt>input_format</dt><dd><p><em>type = str</em></p>
<p>Options = ‘cBioPortal’, ‘WT_ALT’, ‘PrimeDesign’. For ‘WT_ALT’, make sure to put headers as “WT” and “ALT”. 
For “PrimeDesign”, put the header as “SEQ”.</p>
</dd>
<dt>chrom_dict</dt><dd><p><em>type = dict or None</em></p>
<p>Dictionary generated by genome_loader() that holds the chromosome sequences.</p>
</dd>
<dt>PAM</dt><dd><p><em>type = str</em></p>
<p>PAM sequence for searching. Default = “NGG”. Can include any nucleic acid code (e.g. PAM = “NRCH”).</p>
</dd>
<dt>rank_by</dt><dd><p><em>type = str</em></p>
<p>What pegRNA parameter to rank pegRNAs by. Options = “PEGG2_Score” (weighted linear regression of different pegRNA parameters)
or “RF_Score” (random forest predictor of pegRNA efficiency).</p>
</dd>
<dt>pegRNAs_per_mut</dt><dd><p><em>type = ‘All’ or int</em></p>
<p>How many pegRNAs to produce per mutation. Default = ‘All’ (all possible pegRNAs with parameters). Otherwise, choose an integer value (e.g. 5).</p>
</dd>
<dt>RTT_lengths</dt><dd><p><em>type = list</em></p>
<p>List containing RTT lengths to design pegRNAs for.</p>
</dd>
<dt>PBS_length</dt><dd><p><em>type = list</em></p>
<p>List containing PBS lengths to desing pegRNAs for.</p>
</dd>
<dt>min_RHA_size</dt><dd><p><em>type = int</em></p>
<p>Minimum size of the RHA (Right homology arm). Default = 1. Generally pegRNAs with smaller RHA perform poorly.</p>
</dd>
<dt>RE_sites</dt><dd><p><em>type = list or None</em></p>
<p>A list containing the RE recognition sites to filter (e.g. [‘CGTCTC’, ‘GAATTC’] for Esp3I and EcoRI). Default = None (no filtration).</p>
</dd>
<dt>polyT_threshold</dt><dd><p><em>type = int</em></p>
<p>The length of the polyT sequence to classify as a terminator. Default = 4.</p>
</dd>
<dt>proto_size</dt><dd><p><em>type = int</em></p>
<p>The length of the protospacer (excluding the appended G at the 5’ end). Default = 19 (G+19).</p>
</dd>
<dt>context_size</dt><dd><p><em>type = int</em></p>
<p>The amount of context/flanking sequence on either side of the mutation to generate. For larger variants, set this larger.
Default = 120. e.g. AAA(A/G)AAA = context_size of 3</p>
</dd>
<dt>before_proto_context</dt><dd><p><em>type = int</em></p>
<p>Default = 5. Amount of nucleotide context to put before the protospacer in the sensor</p>
</dd>
<dt>sensor_length</dt><dd><p><em>type = int</em></p>
<p>Total length of the sensor in nt. Default = 60.</p>
</dd>
<dt>sensor_orientation</dt><dd><p><em>type = str</em></p>
<p>Options for sensor_orientation = ‘reverse-complement’ or’forward’.</p>
</dd>
<dt>sensor</dt><dd><p><em>type = bool</em></p>
<p>True/False whether to include a sensor in the pegRNA design or not.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.sensor_generator">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">sensor_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before_proto_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reverse-complement'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.sensor_generator" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates sensor sequence for quantification of pegRNA editing outcomes.
Automatically puts sensor in reverse complement orientation with respect to protospacer.
This is highly reccomended to reduce recombination during cloning/library preparation.</p>
<dl>
<dt>df</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>Dataframe containing pegRNAs.</p>
</dd>
<dt>proto_size</dt><dd><p><em>type = int</em></p>
<p>Size of protospacer.</p>
</dd>
<dt>before_proto_context</dt><dd><p><em>type = int</em></p>
<p>Default = 5. Amount of nucleotide context to put before the protospacer in the sensor</p>
</dd>
<dt>sensor_length</dt><dd><p><em>type = int</em></p>
<p>Total length of the sensor in nt. Default = 60.</p>
</dd>
<dt>sensor_orientation</dt><dd><p><em>type = str</em></p>
<p>Options for sensor_orientation = ‘reverse-complement’ or’forward’.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.sensor_viz">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">sensor_viz</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_w_sensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.sensor_viz" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for visualizing the pegRNA aligned to the sensor. Input = dataframe with pegRNA-sensor pairs
and the row index (i) to visualize.</p>
<dl>
<dt>df_w_sensor</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>Dataframe containing the pegRNA-sensor pairs generated by run().</p>
</dd>
<dt>i</dt><dd><p><em>type = int</em></p>
<p>Row index from the dataframe that you want to visualize.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.prime.split_word">
<span class="sig-prename descclassname"><span class="pre">pegg.prime.</span></span><span class="sig-name descname"><span class="pre">split_word</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">word</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.prime.split_word" title="Permalink to this definition">¶</a></dt>
<dd><p>Simple function for splitting string into component characters</p>
</dd></dl>

</div>
<div class="section" id="module-pegg.base">
<span id="pegg-base"></span><h2>pegg.base<a class="headerlink" href="#module-pegg.base" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pegg.base.base_oligo_generator">
<span class="sig-prename descclassname"><span class="pre">pegg.base.</span></span><span class="sig-name descname"><span class="pre">base_oligo_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peg_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">five_prime_adapter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'AGCGTACACGTCTCACACC'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">three_prime_adapter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'GAATTCTAGATCCGGTCGTCAAC'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gRNA_scaff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGC'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.base.base_oligo_generator" title="Permalink to this definition">¶</a></dt>
<dd><p>A tool for automatically generating oligos from the output of run().
Returns input dataframe with new columns containing gRNA oligo with or without sensor.</p>
<dl>
<dt>peg_df</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>A dataframe containing the gRNAs for the selected input mutations. Generated by run_base() or gRNA_generator()</p>
</dd>
<dt>five_prime_adapter</dt><dd><p><em>type = str</em></p>
<p>5’ Prime Adapter. The automatically provided 5’ adapter contains an Esp3I (BsmBI) site. Can be swapped with 
whatever input string user wants.</p>
</dd>
<dt>three_prime_adapter</dt><dd><p><em>type = str</em></p>
<p>5’ Prime Adapter. The automatically provided 5’ adapter contains an Esp3I (BsmBI) site. Can be swapped with 
whatever input string user wants.</p>
</dd>
<dt>gRNA_scaff</dt><dd><p><em>type = str</em></p>
<p>gRNA scaffold region. Automatically set to a functional gRNA scaffold. Can be swapped with 
whatever input string user wants.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.base.eligible_PAM_finder_base">
<span class="sig-prename descclassname"><span class="pre">pegg.base.</span></span><span class="sig-name descname"><span class="pre">eligible_PAM_finder_base</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">19</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.base.eligible_PAM_finder_base" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines eligible PAM sequences for creating gRNAs.
Returns PAM sequence locations on (1) forward, and (2) reverse-complement strand</p>
<dl>
<dt>mut</dt><dd><p><em>type = mutation class</em></p>
<p>See class: mutation</p>
</dd>
<dt>PAM</dt><dd><p><em>type = str</em></p>
<p>PAM sequence for searching. Formatting: e.g. “NGG”</p>
<p>N = [A|T|C|G]
R = [A|G]
Y = [C|T]
S = [G|C]
W = [A|T]
K = [G|T]
M = [A|C]
B = [C|G|T]
D = [A|G|T]
H = [A|C|T]
V = [A|C|G]</p>
</dd>
<dt>proto_size</dt><dd><p><em>type = int</em></p>
<p>Size of protospacer being used. Default = 19 (G+19).</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.base.gRNA_generator">
<span class="sig-prename descclassname"><span class="pre">pegg.base.</span></span><span class="sig-name descname"><span class="pre">gRNA_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ideal_edit_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[4,</span> <span class="pre">8]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.base.gRNA_generator" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates pegRNAs for a given mutation.</p>
<dl>
<dt>mut</dt><dd><p><em>type = mutation class</em></p>
<p>See class: mutation</p>
</dd>
<dt>PAM</dt><dd><p><em>type = str</em></p>
<p>PAM sequence for searching. Formatting: e.g. “NGG”</p>
<p>N = [A|T|C|G]
R = [A|G]
Y = [C|T]
S = [G|C]
W = [A|T]
K = [G|T]
M = [A|C]
B = [C|G|T]
D = [A|G|T]
H = [A|C|T]
V = [A|C|G]</p>
</dd>
<dt>orientation</dt><dd><p><em>type = str</em></p>
<p>‘+’ or ‘-‘</p>
</dd>
<dt>proto_size</dt><dd><p><em>type = int</em></p>
<p>Size of protospacer being used. Default = 19 (G+19).</p>
</dd>
<dt>ideal_edit_window</dt><dd><p><em>type = list</em></p>
<p>Ideal editing window for the editor being used. Default = [4,8]. Labels mutations that fall in this window for 
future filtration if desired.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.base.run_base">
<span class="sig-prename descclassname"><span class="pre">pegg.base.</span></span><span class="sig-name descname"><span class="pre">run_base</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'NGG'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filtration</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ABE+CBE'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ideal_edit_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[4,</span> <span class="pre">8]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auto_SNP_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">19</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">120</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RE_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polyT_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before_proto_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">40</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reverse-complement'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.base.run_base" title="Permalink to this definition">¶</a></dt>
<dd><p>Master function for generating base editing gRNAs. Takes as input a dataframe containing mutations in one of the acceptable formats.
Returns a dataframe with gRNAs with desired design parameters.</p>
<dl>
<dt>input_df</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>Pandas dataframe that contains the input mutations.</p>
</dd>
<dt>input_format</dt><dd><p><em>type = str</em></p>
<p>Options = ‘cBioPortal’, ‘WT_ALT’, ‘PrimeDesign’. For ‘WT_ALT’, make sure to put headers as “WT” and “ALT”. 
For “PrimeDesign”, put the header as “SEQ”.</p>
</dd>
<dt>chrom_dict</dt><dd><p><em>type = dict or None</em></p>
<p>Dictionary generated by genome_loader() that holds the chromosome sequences.</p>
</dd>
<dt>PAM</dt><dd><p><em>type = str</em></p>
<p>PAM sequence for searching. Default = “NGG”. Can include any nucleic acid code (e.g. PAM = “NRCH”).</p>
</dd>
<dt>filtration</dt><dd><p><em>type = str or list</em></p>
<p>Filters the mutation input list to only include the desired SNPs. Options = “No filter”, “ABE”, “CBE”, “ABE+CBE”,
or a list containing the desired SNPs to model (e.g. [‘C&gt;A’, ‘T&gt;C’]).</p>
</dd>
<dt>ideal_edit_window</dt><dd><p><em>type = list</em></p>
<p>Ideal editing window for the editor being used. Default = [4,8]. Labels mutations that fall in this window for 
future filtration if desired.</p>
</dd>
<dt>auto_SNP_filter</dt><dd><p><em>type = bool</em></p>
<p>True/False for whether to filter mutant input to exclude mutations that are NOT SNPs (and thus not BE amenable).</p>
</dd>
<dt>proto_size</dt><dd><p><em>type = int</em></p>
<p>The length of the protospacer (excluding the appended G at the 5’ end). Default = 19 (G+19).</p>
</dd>
<dt>context_size</dt><dd><p><em>type = int</em></p>
<p>The amount of context/flanking sequence on either side of the mutation to generate. For larger variants, set this larger.
Default = 120. e.g. AAA(A/G)AAA = context_size of 3</p>
</dd>
<dt>RE_sites</dt><dd><p><em>type = list or None</em></p>
<p>A list containing the RE recognition sites to filter (e.g. [‘CGTCTC’, ‘GAATTC’] for Esp3I and EcoRI). Default = None (no filtration).</p>
</dd>
<dt>polyT_threshold</dt><dd><p><em>type = int</em></p>
<p>The length of the polyT sequence to classify as a terminator. Default = 4.</p>
</dd>
<dt>before_proto_context</dt><dd><p><em>type = int</em></p>
<p>Default = 5. Amount of nucleotide context to put before the protospacer in the sensor</p>
</dd>
<dt>sensor_length</dt><dd><p><em>type = int</em></p>
<p>Total length of the sensor in nt. Default = 60.</p>
</dd>
<dt>sensor_orientation</dt><dd><p><em>type = str</em></p>
<p>Options for sensor_orientation = ‘reverse-complement’ or’forward’.</p>
</dd>
<dt>sensor</dt><dd><p><em>type = bool</em></p>
<p>True/False whether to include a sensor in the pegRNA design or not.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.base.sensor_generator_base">
<span class="sig-prename descclassname"><span class="pre">pegg.base.</span></span><span class="sig-name descname"><span class="pre">sensor_generator_base</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proto_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before_proto_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensor_orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reverse-complement'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.base.sensor_generator_base" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates sensor sequence for quantification of gRNA editing outcomes.
Automatically puts sensor in reverse complement orientation with respect to protospacer.
This is highly reccomended to reduce recombination during cloning/library preparation.</p>
<dl>
<dt>df</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>Dataframe containing gRNAs.</p>
</dd>
<dt>proto_size</dt><dd><p><em>type = int</em></p>
<p>Size of protospacer.</p>
</dd>
<dt>before_proto_context</dt><dd><p><em>type = int</em></p>
<p>Default = 5. Amount of nucleotide context to put before the protospacer in the sensor</p>
</dd>
<dt>sensor_length</dt><dd><p><em>type = int</em></p>
<p>Total length of the sensor in nt. Default = 60.</p>
</dd>
<dt>sensor_orientation</dt><dd><p><em>type = str</em></p>
<p>Options for sensor_orientation = ‘reverse-complement’ or’forward’.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.base.sensor_viz_base">
<span class="sig-prename descclassname"><span class="pre">pegg.base.</span></span><span class="sig-name descname"><span class="pre">sensor_viz_base</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_w_sensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.base.sensor_viz_base" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for visualizing the gRNA aligned to the sensor. Input = dataframe with pegRNA-sensor pairs
and the row index (i) to visualize.</p>
<dl>
<dt>df_w_sensor</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>Dataframe containing the gRNA-sensor pairs generated by run().</p>
</dd>
<dt>i</dt><dd><p><em>type = int</em></p>
<p>Row index from the dataframe that you want to visualize.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.base.split_word">
<span class="sig-prename descclassname"><span class="pre">pegg.base.</span></span><span class="sig-name descname"><span class="pre">split_word</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">word</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.base.split_word" title="Permalink to this definition">¶</a></dt>
<dd><p>Simple function for splitting string into component characters</p>
</dd></dl>

</div>
<div class="section" id="module-pegg.library">
<span id="pegg-library"></span><h2>pegg.library<a class="headerlink" href="#module-pegg.library" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pegg.library.aavs1_muts">
<span class="sig-prename descclassname"><span class="pre">pegg.library.</span></span><span class="sig-name descname"><span class="pre">aavs1_muts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_muts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'GRCh37'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library.aavs1_muts" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for generating list of “mutations” within AAVS1 locus as negative controls.
Generates a list of SNPs that contain idential reference and alternate alleles.
Intron 1-2 of PPP1R12C (AAVS1 locus): GRCh37 transcript = ENST00000263433.3 | GRCh38 transcript = ENST00000263433.8</p>
<p>Not currently used in the pipeline, but provided anyway.</p>
<dl>
<dt>chrom_dict</dt><dd><p><em>type = dict</em></p>
<p>Dictionary containing chromosomes in dictionary format for parsing by PEGG.</p>
</dd>
<dt>num_muts</dt><dd><p><em>type = int</em></p>
<p>Number of safe-targetting mutations to select/generate.</p>
</dd>
<dt>genome_version</dt><dd><p><em>type = str</em></p>
<p>Options = ‘GRCh37’, ‘GRCh38’</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library.library_maker">
<span class="sig-prename descclassname"><span class="pre">pegg.library.</span></span><span class="sig-name descname"><span class="pre">library_maker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutant_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_safetarget</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'human'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_silent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_end_cds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library.library_maker" title="Permalink to this definition">¶</a></dt>
<dd><p>Compiles the different library design functions to aggregate all of the variants for a particular gene,
include silent substitution controls at a designated %, and non-targeting controls. This can then be fed
into the pegg2.run() or base_editing.run_base() functions.</p>
<p>NOTE: THIS ONLY WORKS FOR GRCh37 and GRCm38!!!
Alternatively, generate these mutations/guides using these reference builds, and then generate your own guides
targeting mutations in your desired reference build.</p>
<dl>
<dt>mutant_input</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>DataFrame containing all of the input mutations. Doesn’t require filtration for gene of interest.</p>
</dd>
<dt>gene_name</dt><dd><p><em>type = str</em></p>
<p>Gene name to select from the mutant_input dataframe.</p>
</dd>
<dt>chrom_dict</dt><dd><p><em>type = dict</em></p>
<p>Dictionary containing the reference genome. See genome_loader()</p>
</dd>
<dt>fraction_safetarget</dt><dd><p><em>type = float</em></p>
<p>Value from 0 to 1 that corresponds to the fraction of safe-targetting mutations to include in the library.
(i.e. what fraction of the mutant_input)</p>
</dd>
<dt>organism</dt><dd><p><em>type = str</em></p>
<p>Options = ‘human’ or ‘mouse’. Determines which of the organisms to generate safe targeting guides for.</p>
</dd>
<dt>fraction_silent</dt><dd><p><em>type = float</em></p>
<p>Value from 0 to 1 that corresponds to the fraction of silent mutations to include in the library.
(i.e. what fraction of the mutant_input)</p>
</dd>
<dt>chrom</dt><dd><p><em>type = int or str</em></p>
<p>Chromosome that the gene of interest falls on.</p>
</dd>
<dt>strand</dt><dd><p><em>type = str</em></p>
<p>‘+’ or ‘-‘ – corresponds with which strand the transcript falls on.</p>
</dd>
<dt>start_end_cds</dt><dd><p><em>type = list</em></p>
<p>Nested list that contains the CDS locations of the gene transcript of interest, in the + strand orientation.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library.mutation_aggregator">
<span class="sig-prename descclassname"><span class="pre">pegg.library.</span></span><span class="sig-name descname"><span class="pre">mutation_aggregator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutant_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library.mutation_aggregator" title="Permalink to this definition">¶</a></dt>
<dd><p>Selects the mutations that correspond to the desired gene name.
Removes duplicates (checking at DNA level; not amino acid level).
Returns dataframe containing these aggregated mutants occuring in gene_name.</p>
<dl>
<dt>mutant_input</dt><dd><p><em>type = pd.DataFrame</em></p>
<p>A dataframe containing the input mutations from which selections are made to generate pegRNAs.
See documentation for precise qualities of this dataframe</p>
</dd>
<dt>gene_name</dt><dd><p><em>type = str</em></p>
<p>Gene’s Hugo_Symbol (i.e. name).</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library.neutral_substitutions">
<span class="sig-prename descclassname"><span class="pre">pegg.library.</span></span><span class="sig-name descname"><span class="pre">neutral_substitutions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_end_cds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library.neutral_substitutions" title="Permalink to this definition">¶</a></dt>
<dd><p>A function for generating all possible synonymous codon substitutions (i.e. silent mutations) of a given gene.
See documentation for more information about format of start_end_cds &amp; example usage.</p>
<dl>
<dt>gene_name</dt><dd><p><em>type = str</em></p>
<p>Gene’s Hugo_Symbol (i.e. name).</p>
</dd>
<dt>chrom</dt><dd><p><em>type = str</em></p>
<p>Chromosome that gene occurs on. Format = e.g. ‘chr17’, ‘chrX’, etc.</p>
</dd>
<dt>strand</dt><dd><p><em>type = str</em></p>
<p>Strand that transcript is on. Options are ‘+’ or ‘-‘.</p>
</dd>
<dt>start_end_cds</dt><dd><p><em>type = list</em></p>
<p>A 2-d list containing the start/end locations of each region of the coding sequence (CDS) for the gene’s selected transcript.
See documentation for example and precise specifications of format.</p>
</dd>
<dt>chrom_dict</dt><dd><p><em>type = dict</em></p>
<p>Dictionary containing the reference genome. See genome_loader()</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library.nontargeting_guides">
<span class="sig-prename descclassname"><span class="pre">pegg.library.</span></span><span class="sig-name descname"><span class="pre">nontargeting_guides</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_guides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edit_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'prime'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library.nontargeting_guides" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for generating a list of non-targetting guides (in the human genome).
Future versions will expand to mouse as well…Unprocessed files for mouse located within this module.
Guides taken from: <a class="reference external" href="https://www.nature.com/articles/nmeth.4423">https://www.nature.com/articles/nmeth.4423</a> (<a class="reference external" href="https://doi.org/10.1038/nmeth.4423">https://doi.org/10.1038/nmeth.4423</a>)</p>
<dl>
<dt>num_guides</dt><dd><p><em>type = int</em></p>
<p>Number of guides to non-targetting generate. Max = 1000.</p>
</dd>
<dt>edit_type</dt><dd><p><em>type = str</em></p>
<p>Options = ‘base’, ‘prime’</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pegg.library.safe_muts">
<span class="sig-prename descclassname"><span class="pre">pegg.library.</span></span><span class="sig-name descname"><span class="pre">safe_muts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_muts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'human'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pegg.library.safe_muts" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for generating a list of safe-targetting pegRNAs/gRNAs.
NOTE: THIS ONLY WORKS FOR GRCh37 and GRCm38!!!
Alternatively, generate these mutations/guides using these reference builds, and then generate your own guides
targeting mutations in your desired reference build.</p>
<p>This is a random subset of 100,000 “safe regions” taken from Morgens et al., 2017 (<a class="reference external" href="https://doi.org/10.1038/ncomms15178">https://doi.org/10.1038/ncomms15178</a>)</p>
<dl>
<dt>num_muts</dt><dd><p><em>type = int</em></p>
<p>Number of safe-targetting mutations to select/generate.</p>
</dd>
<dt>chrom_dict</dt><dd><p><em>type = dict</em></p>
<p>Dictionary containing chromosomes in dictionary format for parsing by PEGG.</p>
</dd>
<dt>organism</dt><dd><p><em>type = str</em></p>
<p>Choices = “human” or “mouse”. Generates safe targeting guides in human or mouse genome.</p>
</dd>
</dl>
</dd></dl>

</div>
</div>


        </div>
      </div>

    </div>

<footer>
    <div id="footer-info">
        <ul id="build-details">
            
                <li class="footer-element">
                    
                        <a href="_sources/PEGG.rst.txt" rel="nofollow"> source</a>
                    
                </li>
            

            

            
        </ul>

        
            <div id="copyright">
                &copy; 2023, Samuel Gould
            </div>
        

        <div id="credit">
            created with <a href="http://sphinx-doc.org/">Sphinx</a> and <a href="https://github.com/tomaarsen/nltk_theme">NLTK Theme</a>
        </div>
    </div>
</footer> 

</div>

</body>
</html>